---
title: "Independent Study to prepare for workshop"
subtitle: "Transcriptomics 1: 👋 Hello data!"
author: "Emma Rand"
format: 
  revealjs:
    footer: "🔗 [About Transcriptomics 1: Hello data!](https://3mmarand.github.io/BIO00088H-data/transcriptomics/week-3/overview.html)"
    slide-number: true
    chalkboard: true
    code-link: true
    code-line-numbers: false
bibliography: ../../references.bib
editor: 
  markdown: 
    wrap: 72
---

## Overview

::: incremental
-   Concise summary of the experimental design and aims

-   What the raw data consist of

-   What has been done to the data so far

-   What steps we will take in the workshop
:::

## The Data

There are three transcriptomic datasets

-   🐸 bulk RNA-seq from *Xenopus laevis* embryos.

-   🐭 single cell RNA-seq from mouse stemcells

-   🎄 bulk RNA-seq from *Arabidopsis thaliana*

-   💉 bulk RNA-seq from *Leishmania mexicana*

# Experimental design

## 🐸 Experimental design {auto-animate="true"}

![Schematic of frog development
experiment](images/88H-exp-design-betsy.png){fig-align="center"
width="700"}

## 🐸 Experimental design {auto-animate="true"}

![Schematic of frog development
experiment](images/88H-exp-design-betsy.png){fig-align="left"
width="200"}

::: incremental
-   3 fertilisations

-   two siblings from each fertilisation one control, one FGF treated

-   sequenced at three time points

-   3 x 2 x 3 = 18 groups
:::

## 🐸 Experimental design {auto-animate="true"}

![Schematic of frog development
experiment](images/88H-exp-design-betsy.png){fig-align="left"
width="200"}

::: incremental
-   3 fertilisations. [These are the replicates, `1`, `2`,
    `3`]{style="color:#009900"}

-   two siblings from each fertilisation one control, one FGF treated.
    [The treatments are paired]{style="color:#009900"}

-   sequenced at three time points. [S14, S20,
    S30]{style="color:#009900"}

-   3 x 2 x 3 = 18 groups
:::

## 🐸 Aim

::: incremental
-   Find genes that are "differentially expressed" between
     control-treated and FGF-treated siblings

-   Differentially expressed means the expression in one group is
    significantly higher than in the other
:::

## 🐸 Guided analysis

::: incremental
-   The workshops will take you through comparing the control and FGF
    treated sibling at S30
    
-   You will make other comparisons independently

-   You will be guided to carefully document your work so you can apply
    the same methods to other comparisons
    
-   Do the independent study before and after the workshop!
:::

## 🐭 Experimental design {auto-animate="true"}

![Schematic of stem cell
experiment](images/88H-exp-design-jillian.png){fig-align="center"
width="700"}

## 🐭 Experimental design {auto-animate="true"}

![Schematic of stem cell
experiment](images/88H-exp-design-jillian.png){fig-align="left"
width="200"}

::: incremental
-   Cells were sorted using flow cytometry on the basis of cell surface
    markers

-   There are three cell types: LT-HSCs, HSPCs, Progs

-   Many cells of each cell type were sequenced
:::

## 🐭 Experimental design {auto-animate="true"}

![Schematic of stem cell
experiment](images/88H-exp-design-jillian.png){fig-align="left"
width="200"}

::: incremental
-   There are three cell types: LT-HSCs, HSPCs, Progs [These are the
    "treaments"]{style="color:#009900"}

-   Many cells of each type were sequenced: [These are the
    replicates]{style="color:#009900"}

-   [155 LT-HSCs, 701 HSPCs, 798 Progs]{style="color:#009900"}
:::

## 🐭 Aim

::: incremental
-   find genes that are "differentially expressed" between at
    least two cell types

-   Differentially expressed means the expression in one group is
    significantly higher than in the other
:::

## 🐭 Guided analysis

::: incremental
-   The workshops will take you through comparing the HSPC and Prog
    cells

-   You will make other comparisons independently

-   You will be guided to carefully document your work so you can apply
    the same methods to other comparisons
    
-   Do the independent study before and after the workshop!
:::

# Where do the data come from?

## Raw Sequence data

::: incremental
-   The raw data are "reads" from a sequencing machine in FASTQ files

-   A read is sequence of RNA which is shorter than the whole 
    transcriptome

-   The length of the reads depends on the type of sequencing machine

    -   Short-read technologies (e.g. Illumina) have higher base accuracy
        but are harder to align
    -   Long-read technologies (e.g. Nanopore) have lower base accuracy
        but are easier to align

-   Optional: You can read more about Sequencing technologies in
    [Statistically useful experimental
    design](https://cloud-span.github.io/experimental_design00-overview/)
    [@rand_statistically_2022]        

# What has been done to the data so far

## General steps

::: incremental
-   Reads are filtered and trimmed on the basis of a quality score

-   They are then aligned/pseudo-aligned to a reference
    genome/transcriptome 

-   And then counted to quantify the expression 

-   Counts need to be normalised to account for differences 
    in sequencing depth and transcript length before, or as part 
    of, statistical analysis. 
:::

## 🐸 Data

-   Unpublished (so far!)

-   Expression for the whole transcriptome [*X. laevis* v10.1 genome
    assembly](https://www.xenbase.org/xenbase/static-xenbase/ftpDatafiles.jsp)

-   Values are raw counts

-   The statistical analysis method we will use `DESeq2` [@DESeq2]
    requires raw counts and performs the normalisation itself

## 🐭 Data

-   Published in @nestorowa2016

-   Expression for a **subset of genes**, the surfaceome

-   Values are log2 normalised values

-   The statistical analysis method we will use `scran` [@scran]
    requires normalised values

# Workshops

## Workshops

-   Transcriptomics 1: Hello data Getting to know the data. Checking the
    distributions of values overall, across rows and columns to
    check things are as we expect and detect rows/columns that need to
    be removed

-   Transcriptomics 2: Statistical Analysis. Identifying which genes are
    differentially expressed between treatments. This is the main
    analysis step. We will use different methods for bulk and single
    cell data.

-   Transcriptomics 3: Visualising. Principal Component 
    Analysis (PCA) volcano plots to visualise the results of the 


## References
Pages made with R [@R-core], Quarto [@Allaire_Quarto_2024], `knitr` [@knitr1; @knitr2; @knitr3], `kableExtra` [@kableExtra]