<!DOCTYPE html>
<html lang="en"><head>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.33">

  <meta name="author" content="Emma Rand">
  <title>Data Analysis for Group Project – Independent Study to prepare for workshop</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-f563837468303362081e247dddd440d0.css">
  <script async="" src="https://www.googletagmanager.com/gtag/js?id=G-TVF95J593T"></script>

  <script type="text/javascript">

  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-TVF95J593T', { 'anonymize_ip': true});
  </script>
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Independent Study to prepare for workshop</h1>
  <p class="subtitle">Transcriptomics 1: 👋 Hello data!</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Emma Rand 
</div>
</div>
</div>

  <p class="date">19 August, 2025</p>
</section>
<section id="overview" class="slide level2">
<h2>Overview</h2>
<ul>
<li class="fragment"><p>Concise summary of the experimental design and aims</p></li>
<li class="fragment"><p>What the raw data consist of</p></li>
<li class="fragment"><p>What has been done to the data so far</p></li>
<li class="fragment"><p>What steps we will take in the workshop</p></li>
</ul>
</section>
<section id="the-data" class="slide level2">
<h2>The Data</h2>
<p>There are 3 transcriptomic datasets</p>
<ul>
<li><p>🎄 bulk RNA-seq from <em>Arabidopsis thaliana</em></p></li>
<li><p>💉 bulk RNA-seq from <em>Leishmania mexicana</em></p></li>
<li><p>🐭 single cell RNA-seq from mouse stemcells</p></li>
</ul>
</section>
<section>
<section id="experimental-design" class="title-slide slide level1 center">
<h1>Experimental design</h1>

</section>
<section id="experimental-design-1" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">🎄 Experimental design</h2>

<img data-src="images/88H-exp-design-liz.png" width="700" class="r-stretch quarto-figure-center"><p class="caption">Schematic of arabidopsis experiment</p></section>
<section id="experimental-design-2" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">🎄 Experimental design</h2>

<img data-src="images/88H-exp-design-liz.png" width="200" class="r-stretch quarto-figure-left"><p class="caption">Schematic of arabidopsis experiment</p><ul>
<li class="fragment"><p>2 plant tissues</p></li>
<li class="fragment"><p>2 nickel conditions</p></li>
<li class="fragment"><p>6 replicates</p></li>
<li class="fragment"><p>2 x 2 x 6 = 24 samples</p></li>
</ul>
</section>
<section id="experimental-design-3" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">🎄 Experimental design</h2>

<img data-src="images/88H-exp-design-liz.png" width="200" class="r-stretch quarto-figure-left"><p class="caption">Schematic of arabidopsis experiment</p><ul>
<li class="fragment"><p>2 plant tissues: root and aerial. <span style="color:#009900">This is the tissue treatment</span></p></li>
<li class="fragment"><p>2 nickel conditions: control and low Ni. <span style="color:#009900">This is the Ni treatment</span></p></li>
<li class="fragment"><p>6 replicates. <span style="color:#009900">These are the replicates</span></p></li>
<li class="fragment"><p>2 x 2 x 6 = 24 samples</p></li>
</ul>
</section>
<section id="aim" class="slide level2">
<h2>🎄 Aim</h2>
<ul>
<li class="fragment"><p>Find genes that are “differentially expressed” between tissue types and nickel conditions e.g.&nbsp;root tissue grown under control and low Ni conditions</p></li>
<li class="fragment"><p>Differentially expressed means the expression in one group is significantly higher than in the other</p></li>
</ul>
</section>
<section id="guided-analysis" class="slide level2">
<h2>🎄 Guided analysis</h2>
<ul>
<li class="fragment"><p>The workshops will take you through comparing the root tissue grown under control and low Ni conditions</p></li>
<li class="fragment"><p>You will make other comparisons independently</p></li>
<li class="fragment"><p>You will be guided to carefully document your work so you can apply the same methods to other comparisons</p></li>
<li class="fragment"><p>Do the independent study before and after the workshop!</p></li>
</ul>
</section>
<section id="experimental-design-4" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">💉 Experimental design</h2>

<img data-src="images/88H-exp-design-pegine.png" width="700" class="r-stretch quarto-figure-center"><p class="caption">Schematic of leishmania experiment</p></section>
<section id="experimental-design-5" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">💉 Experimental design</h2>

<img data-src="images/88H-exp-design-pegine.png" width="200" class="r-stretch quarto-figure-left"><p class="caption">Schematic of leishmania experiment</p><ul>
<li class="fragment"><p>3 stages</p></li>
<li class="fragment"><p>3 samples</p></li>
<li class="fragment"><p>3 x 3 = 9 samples</p></li>
</ul>
</section>
<section id="experimental-design-6" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">💉 Experimental design</h2>

<img data-src="images/88H-exp-design-pegine.png" width="200" class="r-stretch quarto-figure-left"><p class="caption">Schematic of leishmania experiment</p><ul>
<li class="fragment"><p>three stages: procyclic promastigotes, metacyclic promastigotes and amastigotes. <span style="color:#009900">This is the stage treatment</span></p></li>
<li class="fragment"><p>three samples. <span style="color:#009900">These are the replicates</span></p></li>
<li class="fragment"><p>3 x 3 = 9 samples</p></li>
</ul>
</section>
<section id="aim-1" class="slide level2">
<h2>💉 Aim</h2>
<ul>
<li class="fragment"><p>Find genes that are “differentially expressed” between stages e.g., procyclic promastigotes and the metacyclic promastigotes</p></li>
<li class="fragment"><p>Differentially expressed means the expression in one group is significantly higher than in the other</p></li>
</ul>
</section>
<section id="guided-analysis-1" class="slide level2">
<h2>💉 Guided analysis</h2>
<ul>
<li class="fragment"><p>The workshops will take you through comparing the procyclic promastigotes and the metacyclic promastigotes</p></li>
<li class="fragment"><p>You will make other comparisons independently</p></li>
<li class="fragment"><p>You will be guided to carefully document your work so you can apply the same methods to other comparisons</p></li>
<li class="fragment"><p>Do the independent study before and after the workshop!</p></li>
</ul>
</section>
<section id="experimental-design-7" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">🐭 Experimental design</h2>

<img data-src="images/88H-exp-design-jillian.png" width="700" class="r-stretch quarto-figure-center"><p class="caption">Schematic of stem cell experiment</p></section>
<section id="experimental-design-8" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">🐭 Experimental design</h2>

<img data-src="images/88H-exp-design-jillian.png" width="200" class="r-stretch quarto-figure-left"><p class="caption">Schematic of stem cell experiment</p><ul>
<li class="fragment"><p>Cells were sorted using flow cytometry on the basis of cell surface markers</p></li>
<li class="fragment"><p>There are 3 cell types</p></li>
<li class="fragment"><p>Many cells of each cell type were sequenced</p></li>
</ul>
</section>
<section id="experimental-design-9" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">🐭 Experimental design</h2>

<img data-src="images/88H-exp-design-jillian.png" width="200" class="r-stretch quarto-figure-left"><p class="caption">Schematic of stem cell experiment</p><ul>
<li class="fragment"><p>There are three cell types: LT-HSCs, HSPCs, Progs <span style="color:#009900">This is the cell “treatment”</span></p></li>
<li class="fragment"><p>Many cells of each type were sequenced: <span style="color:#009900">These are the replicates</span></p></li>
<li class="fragment"><p><span style="color:#009900">155 LT-HSCs, 701 HSPCs, 798 Progs</span></p></li>
</ul>
</section>
<section id="aim-2" class="slide level2">
<h2>🐭 Aim</h2>
<ul>
<li class="fragment"><p>find genes that are “differentially expressed” between at least two cell types</p></li>
<li class="fragment"><p>Differentially expressed means the expression in one group is significantly higher than in the other</p></li>
</ul>
</section>
<section id="guided-analysis-2" class="slide level2">
<h2>🐭 Guided analysis</h2>
<ul>
<li class="fragment"><p>The workshops will take you through comparing the HSPC and Prog cells</p></li>
<li class="fragment"><p>You will make other comparisons independently</p></li>
<li class="fragment"><p>You will be guided to carefully document your work so you can apply the same methods to other comparisons</p></li>
<li class="fragment"><p>Do the independent study before and after the workshop!</p></li>
</ul>
</section></section>
<section>
<section id="where-do-the-data-come-from" class="title-slide slide level1 center">
<h1>Where do the data come from?</h1>

</section>
<section id="raw-sequence-data" class="slide level2">
<h2>Raw Sequence data</h2>
<ul>
<li class="fragment"><p>The raw data are “reads” from a sequencing machine in FASTQ files</p></li>
<li class="fragment"><p>A read is sequence of RNA which is shorter than the whole transcriptome</p></li>
<li class="fragment"><p>The length of the reads depends on the type of sequencing machine</p>
<ul>
<li class="fragment">Short-read technologies (e.g.&nbsp;Illumina) have higher base accuracy but are harder to align</li>
<li class="fragment">Long-read technologies (e.g.&nbsp;Nanopore) have lower base accuracy but are easier to align</li>
</ul></li>
</ul>
</section>
<section id="raw-sequence-data-1" class="slide level2">
<h2>Raw Sequence data</h2>
<h3 id="optional">Optional</h3>
<p>You can read more about Sequencing technologies in <a href="https://cloud-span.github.io/experimental_design00-overview/">Statistically useful experimental design</a><span class="citation" data-cites="rand_statistically_2022">(<a href="#/references" role="doc-biblioref" onclick="">Rand and Forrester 2022</a>)</span></p>
</section></section>
<section>
<section id="what-has-been-done-to-the-data-so-far" class="title-slide slide level1 center">
<h1>What has been done to the data so far</h1>

</section>
<section id="general-steps" class="slide level2">
<h2>General steps</h2>
<ul>
<li class="fragment"><p>Reads are filtered and trimmed on the basis of a quality score</p></li>
<li class="fragment"><p>They are then aligned/pseudo-aligned to a reference genome/transcriptome (or assembled <em>de novo</em>)</p></li>
<li class="fragment"><p>And then counted to quantify the expression</p></li>
<li class="fragment"><p>Counts need to be normalised to account for differences in sequencing depth and transcript length before, or as part of, statistical analysis.</p></li>
</ul>
</section>
<section id="data" class="slide level2">
<h2>🎄 Data</h2>
<ul>
<li><p>Brand spanking new! Provided by Alex Marks <span class="citation" data-cites="marks_nickel_nodate">(<a href="#/references" role="doc-biblioref" onclick="">Marks and Rylott, n.d.</a>)</span></p></li>
<li><p>Expression for the whole transcriptome <a href="https://plants.ensembl.org/Arabidopsis_thaliana/Info/Index">ENSEMBL <em>Arabidopsis</em> TAIR10</a><span class="citation" data-cites="yates_ensembl_2022">(<a href="#/references" role="doc-biblioref" onclick="">Yates et al. 2022</a>)</span></p></li>
<li><p>Values are raw counts</p></li>
<li><p>The statistical analysis method we will use is <code>DESeq2</code> <span class="citation" data-cites="DESeq2">(<a href="#/references" role="doc-biblioref" onclick="">Love, Huber, and Anders 2014</a>)</span>. It requires raw counts and performs the normalisation itself.</p></li>
</ul>
</section>
<section id="data-1" class="slide level2">
<h2>💉 Data</h2>
<ul>
<li><p>Brand spanking new!</p></li>
<li><p>Expression for the whole transcriptome <a href="https://tritrypdb.org/tritrypdb/app/record/dataset/DS_e318592c8e"><em>L. mexicana</em> MHOM/GT/2001/U1103</a><span class="citation" data-cites="rogers_chromosome_2011">(<a href="#/references" role="doc-biblioref" onclick="">Rogers et al. 2011</a>)</span></p></li>
<li><p>Values are raw counts</p></li>
<li><p>The statistical analysis method we will use is <code>DESeq2</code> <span class="citation" data-cites="DESeq2">(<a href="#/references" role="doc-biblioref" onclick="">Love, Huber, and Anders 2014</a>)</span>. It requires raw counts and performs the normalisation itself.</p></li>
</ul>
</section>
<section id="data-2" class="slide level2">
<h2>🐭 Data</h2>
<ul>
<li><p>Published in <span class="citation" data-cites="nestorowa2016">Nestorowa et al. (<a href="#/references" role="doc-biblioref" onclick="">2016</a>)</span></p></li>
<li><p>Expression for a <strong>subset of genes</strong>, the surfaceome</p></li>
<li><p>Values are log<sub>2</sub> normalised values</p></li>
<li><p>The statistical analysis method we will use is <code>scran</code> <span class="citation" data-cites="scran">(<a href="#/references" role="doc-biblioref" onclick="">Lun, McCarthy, and Marioni 2016</a>)</span> and it requires normalised values</p></li>
</ul>
</section></section>
<section>
<section id="workshops" class="title-slide slide level1 center">
<h1>Workshops</h1>

</section>
<section id="workshops-1" class="slide level2">
<h2>Workshops</h2>
<ul>
<li><p>Transcriptomics 1: Hello data Getting to know the data. Checking the distributions of values overall, across rows and columns to check things are as we expect and detect rows/columns that need to be removed</p></li>
<li><p>Transcriptomics 2: Statistical Analysis. Identifying which genes are differentially expressed between treatments. This is the main analysis step. We will use different methods for bulk and single cell data.</p></li>
<li><p>Transcriptomics 3: Visualising. Principal Component Analysis (PCA) and volcano plots to visualise the results of the analysis.</p></li>
</ul>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>
<p>Pages made with R <span class="citation" data-cites="R-core">(<a href="#/references" role="doc-biblioref" onclick="">R Core Team 2024</a>)</span>, Quarto <span class="citation" data-cites="Allaire_Quarto_2024">(<a href="#/references" role="doc-biblioref" onclick="">Allaire et al. 2024</a>)</span>, <code>knitr</code> <span class="citation" data-cites="knitr1 knitr2 knitr3">(<a href="#/references" role="doc-biblioref" onclick="">Xie 2024</a>, <a href="#/references" role="doc-biblioref" onclick="">2015</a>, <a href="#/references" role="doc-biblioref" onclick="">2014</a>)</span>, <code>kableExtra</code> <span class="citation" data-cites="kableExtra">(<a href="#/references" role="doc-biblioref" onclick="">Zhu 2021</a>)</span></p>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Allaire_Quarto_2024" class="csl-entry" role="listitem">
Allaire, J. J., Charles Teague, Carlos Scheidegger, Yihui Xie, and Christophe Dervieux. 2024. <span>“<span>Quarto</span>.”</span> <a href="https://doi.org/10.5281/zenodo.5960048">https://doi.org/10.5281/zenodo.5960048</a>.
</div>
<div id="ref-DESeq2" class="csl-entry" role="listitem">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-scran" class="csl-entry" role="listitem">
Lun, Aaron T. L., Davis J. McCarthy, and John C. Marioni. 2016. <span>“A Step-by-Step Workflow for Low-Level Analysis of Single-Cell RNA-Seq Data with Bioconductor.”</span> <em>F1000Res.</em> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-marks_nickel_nodate" class="csl-entry" role="listitem">
Marks, Alex, and Elizabeth Rylott. n.d. <span>“Nickel <span>Dosing</span> <span>Arabidopsis</span> Thaliana <span>RNA</span>-Seq.”</span>
</div>
<div id="ref-nestorowa2016" class="csl-entry" role="listitem">
Nestorowa, Sonia, Fiona K. Hamey, Blanca Pijuan Sala, Evangelia Diamanti, Mairi Shepherd, Elisa Laurenti, Nicola K. Wilson, David G. Kent, and Berthold Göttgens. 2016. <span>“A Single-Cell Resolution Map of Mouse Hematopoietic Stem and Progenitor Cell Differentiation.”</span> <em>Blood</em> 128 (8): e20–31. <a href="https://doi.org/10.1182/blood-2016-05-716480">https://doi.org/10.1182/blood-2016-05-716480</a>.
</div>
<div id="ref-R-core" class="csl-entry" role="listitem">
R Core Team. 2024. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-rand_statistically_2022" class="csl-entry" role="listitem">
Rand, Emma, and Sarah Forrester. 2022. <span>“Statistically Useful Experimental Design.”</span> <a href="https://cloud-span.github.io/experimental_design00-overview/">https://cloud-span.github.io/experimental_design00-overview/</a>.
</div>
<div id="ref-rogers_chromosome_2011" class="csl-entry" role="listitem">
Rogers, Matthew B., James D. Hilley, Nicholas J. Dickens, Jon Wilkes, Paul A. Bates, Daniel P. Depledge, David Harris, et al. 2011. <span>“Chromosome and Gene Copy Number Variation Allow Major Structural Change Between Species and Strains of <span>Leishmania</span>.”</span> <em>Genome Research</em> 21 (12): 2129–42. <a href="https://doi.org/10.1101/gr.122945.111">https://doi.org/10.1101/gr.122945.111</a>.
</div>
<div id="ref-knitr3" class="csl-entry" role="listitem">
Xie, Yihui. 2014. <span>“Knitr: A Comprehensive Tool for Reproducible Research in <span>R</span>.”</span> In <em>Implementing Reproducible Computational Research</em>, edited by Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman; Hall/CRC.
</div>
<div id="ref-knitr2" class="csl-entry" role="listitem">
———. 2015. <em>Dynamic Documents with <span>R</span> and Knitr</em>. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-knitr1" class="csl-entry" role="listitem">
———. 2024. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in r</em>. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-yates_ensembl_2022" class="csl-entry" role="listitem">
Yates, Andrew D, James Allen, Ridwan M Amode, Andrey G Azov, Matthieu Barba, Andrés Becerra, Jyothish Bhai, et al. 2022. <span>“Ensembl <span>Genomes</span> 2022: An Expanding Genome Resource for Non-Vertebrates.”</span> <em>Nucleic Acids Research</em> 50 (D1): D996–1003. <a href="https://doi.org/10.1093/nar/gkab1007">https://doi.org/10.1093/nar/gkab1007</a>.
</div>
<div id="ref-kableExtra" class="csl-entry" role="listitem">
Zhu, Hao. 2021. <span>“kableExtra: Construct Complex Table with ’Kable’ and Pipe Syntax.”</span> <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>.
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">
<p>🔗 <a href="https://3mmarand.github.io/BIO00088H-data/transcriptomics/week-3/overview.html">About Transcriptomics 1: Hello data!</a></p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>