---
title: "Independent Study to prepare for workshop"
subtitle: "Transcriptomics 1: 游녦 Hello data!"
author: "Emma Rand"
format: 
  revealjs:
    footer: "游댕 [About Transcriptomics 1: Hello data!](https://3mmarand.github.io/BIO00088H-data/transcriptomics/week-3/overview.html)"
    slide-number: true
    chalkboard: true
    code-link: true
    code-line-numbers: false
bibliography: ../../references.bib
editor: 
  markdown: 
    wrap: 72
---

## Overview

::: incremental
-   Concise summary of the experimental design and aims

-   What the raw data consist of

-   What has been done to the data so far

-   What steps we will take in the workshop
:::

## The Data

There are 3 transcriptomic datasets

-   游꾻 bulk RNA-seq from *Arabidopsis thaliana*

-   游눌 bulk RNA-seq from *Leishmania mexicana*

-   游냜 single cell RNA-seq from mouse stemcells

# Experimental design

## 游꾻 Experimental design {auto-animate="true"}

![Schematic of arabidopsis
experiment](images/88H-exp-design-liz.png){fig-align="center"
width="700"}

## 游꾻 Experimental design {auto-animate="true"}

![Schematic of arabidopsis
experiment](images/88H-exp-design-liz.png){fig-align="left" width="200"}

::: incremental
-   2 plant tissues

-   2 nickel conditions

-   6 replicates

-   2 x 2 x 6 = 24 samples
:::

## 游꾻 Experimental design {auto-animate="true"}

![Schematic of arabidopsis
experiment](images/88H-exp-design-liz.png){fig-align="left" width="200"}

::: incremental
-   2 plant tissues: root and aerial. [This is the tissue
    treatment]{style="color:#009900"}

-   2 nickel conditions: control and low Ni. [This is the Ni
    treatment]{style="color:#009900"}

-   6 replicates. [These are the replicates]{style="color:#009900"}

-   2 x 2 x 6 = 24 samples
:::

## 游꾻 Aim

::: incremental
-   Find genes that are "differentially expressed" between tissue types
    and nickel conditions e.g. root tissue grown under control and low
    Ni conditions

-   Differentially expressed means the expression in one group is
    significantly higher than in the other
:::

## 游꾻 Guided analysis

::: incremental
-   The workshops will take you through comparing the root tissue grown
    under control and low Ni conditions

-   You will make other comparisons independently

-   You will be guided to carefully document your work so you can apply
    the same methods to other comparisons

-   Do the independent study before and after the workshop!
:::

## 游눌 Experimental design {auto-animate="true"}

![Schematic of leishmania
experiment](images/88H-exp-design-pegine.png){fig-align="center"
width="700"}

## 游눌 Experimental design {auto-animate="true"}

![Schematic of leishmania
experiment](images/88H-exp-design-pegine.png){fig-align="left"
width="200"}

::: incremental
-   3 stages

-   3 samples

-   3 x 3 = 9 samples
:::

## 游눌 Experimental design {auto-animate="true"}

![Schematic of leishmania
experiment](images/88H-exp-design-pegine.png){fig-align="left"
width="200"}

::: incremental
-   three stages: procyclic promastigotes, metacyclic promastigotes and
    amastigotes. [This is the stage treatment]{style="color:#009900"}

-   three samples. [These are the replicates]{style="color:#009900"}

-   3 x 3 = 9 samples
:::

## 游눌 Aim

::: incremental
-   Find genes that are "differentially expressed" between stages e.g.,
    procyclic promastigotes and the metacyclic promastigotes

-   Differentially expressed means the expression in one group is
    significantly higher than in the other
:::

## 游눌 Guided analysis

::: incremental
-   The workshops will take you through comparing the procyclic
    promastigotes and the metacyclic promastigotes

-   You will make other comparisons independently

-   You will be guided to carefully document your work so you can apply
    the same methods to other comparisons

-   Do the independent study before and after the workshop!
:::

## 游냜 Experimental design {auto-animate="true"}

![Schematic of stem cell
experiment](images/88H-exp-design-jillian.png){fig-align="center"
width="700"}

## 游냜 Experimental design {auto-animate="true"}

![Schematic of stem cell
experiment](images/88H-exp-design-jillian.png){fig-align="left"
width="200"}

::: incremental
-   Cells were sorted using flow cytometry on the basis of cell surface
    markers

-   There are 3 cell types

-   Many cells of each cell type were sequenced
:::

## 游냜 Experimental design {auto-animate="true"}

![Schematic of stem cell
experiment](images/88H-exp-design-jillian.png){fig-align="left"
width="200"}

::: incremental
-   There are three cell types: LT-HSCs, HSPCs, Progs [This is the cell
    "treatment"]{style="color:#009900"}

-   Many cells of each type were sequenced: [These are the
    replicates]{style="color:#009900"}

-   [155 LT-HSCs, 701 HSPCs, 798 Progs]{style="color:#009900"}
:::

## 游냜 Aim

::: incremental
-   find genes that are "differentially expressed" between at least two
    cell types

-   Differentially expressed means the expression in one group is
    significantly higher than in the other
:::

## 游냜 Guided analysis

::: incremental
-   The workshops will take you through comparing the HSPC and Prog
    cells

-   You will make other comparisons independently

-   You will be guided to carefully document your work so you can apply
    the same methods to other comparisons

-   Do the independent study before and after the workshop!
:::

# Where do the data come from?

## Raw Sequence data

::: incremental
-   The raw data are "reads" from a sequencing machine in FASTQ files

-   A read is sequence of RNA which is shorter than the whole
    transcriptome

-   The length of the reads depends on the type of sequencing machine

    -   Short-read technologies (e.g. Illumina) have higher base
        accuracy but are harder to align
    -   Long-read technologies (e.g. Nanopore) have lower base accuracy
        but are easier to align
:::

## Raw Sequence data

### Optional

You can read more about Sequencing technologies in [Statistically useful
experimental
design](https://cloud-span.github.io/experimental_design00-overview/)[@rand_statistically_2022]

# What has been done to the data so far

## General steps

::: incremental
-   Reads are filtered and trimmed on the basis of a quality score

-   They are then aligned/pseudo-aligned to a reference
    genome/transcriptome (or assembled *de novo*)

-   And then counted to quantify the expression

-   Counts need to be normalised to account for differences in
    sequencing depth and transcript length before, or as part of,
    statistical analysis.
:::

## 游꾻 Data

-   Brand spanking new! Provided by Alex Marks [@marks_nickel_nodate]

-   Expression for the whole transcriptome [ENSEMBL *Arabidopsis*
    TAIR10](https://plants.ensembl.org/Arabidopsis_thaliana/Info/Index)[@yates_ensembl_2022]

-   Values are raw counts

-   The statistical analysis method we will use is `DESeq2` [@DESeq2].
    It requires raw counts and performs the normalisation itself.

## 游눌 Data

-   Brand spanking new!

-   Expression for the whole transcriptome [*L. mexicana*
    MHOM/GT/2001/U1103](https://tritrypdb.org/tritrypdb/app/record/dataset/DS_e318592c8e)[@rogers_chromosome_2011]

-   Values are raw counts

-   The statistical analysis method we will use is `DESeq2` [@DESeq2].
    It requires raw counts and performs the normalisation itself.

## 游냜 Data

-   Published in @nestorowa2016

::: {.content-visible when-profile="surfaceome"}
-   Expression for a **subset of genes**, the surfaceome
:::

::: {.content-visible when-profile="secretome"}
-   Expression for a **subset of genes**, the secretome
:::

-   Values are log~2~ normalised values

-   The statistical analysis method we will use is `scran` [@scran] and
    it requires normalised values

# Workshops

## Workshops

-   Transcriptomics 1: Hello data Getting to know the data. Checking the
    distributions of values overall, across rows and columns to check
    things are as we expect and detect rows/columns that need to be
    removed

-   Transcriptomics 2: Statistical Analysis. Identifying which genes are
    differentially expressed between treatments. This is the main
    analysis step. We will use different methods for bulk and single
    cell data.

-   Transcriptomics 3: Visualising. Principal Component Analysis (PCA)
    and volcano plots to visualise the results of the analysis.

## References

Pages made with R [@R-core], Quarto [@Allaire_Quarto_2024], `knitr`
[@knitr1; @knitr2; @knitr3], `kableExtra` [@kableExtra]
