<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.24">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Emma Rand">
<title>Workshop ‚Äì Data Analysis for Group Project</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dc55a5b9e770e841cd82e46aadbfb9b0.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-b651517ce65839d647a86e2780455cfb.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dc55a5b9e770e841cd82e46aadbfb9b0.css" rel="stylesheet" class="quarto-color-scheme-extra" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-bd31597842da8e70aa6de660987a7f7b.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark-0724b92bd66eafb47a2b216599690d49.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="../../site_libs/bootstrap/bootstrap-bd31597842da8e70aa6de660987a7f7b.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme-extra" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-TVF95J593T"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-TVF95J593T', { 'anonymize_ip': true});
</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar floating nav-fixed quarto-light"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = false;
    const darkModeDefault = authorPrefersDark;
      document.querySelector('link#quarto-text-highlighting-styles.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
      document.querySelector('link#quarto-bootstrap.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
      window.dispatchEvent(new Event('resize'));
    };
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Data Analysis for Group Project</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" aria-current="page"> 
<span class="menu-text">Welcome!</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../core/core.html"> 
<span class="menu-text">Core Supporting Info</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../transcriptomics/transcriptomics.html"> 
<span class="menu-text">Transcriptomics</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../images/images.html"> 
<span class="menu-text">Image Analysis</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../structures/structures.html"> 
<span class="menu-text">Structure Analysis</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <a href="https://github.com/3mmaRand/BIO00088H-data" title="BIO00088H-data GitHub" class="quarto-navigation-tool px-1" aria-label="BIO00088H-data GitHub"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/week-3/overview.html">Week 3: Hello data!</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/week-3/workshop.html">Workshop</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="../../index.html" class="sidebar-logo-link">
      </a>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/transcriptomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Transcriptomics Data Analysis for Group Project</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 3: Hello data!</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/workshop.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 4: Statistical Analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 5: Visualising and Interpreting</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Semester 2: Extras</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/semester-2/stem_cell_variant.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Additional üê≠ stem cell labels</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Kelly‚Äôs Project</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/kelly/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul>
<li><a href="#session-overview" id="toc-session-overview" class="nav-link" data-scroll-target="#session-overview">Session overview</a></li>
  </ul>
</li>
  <li><a href="#set-up-a-project" id="toc-set-up-a-project" class="nav-link" data-scroll-target="#set-up-a-project">Set up a Project</a></li>
  <li><a href="#examine-data-using-a-spreadsheet" id="toc-examine-data-using-a-spreadsheet" class="nav-link" data-scroll-target="#examine-data-using-a-spreadsheet">Examine data using a spreadsheet</a></li>
  <li>
<a href="#explore-data-in-r" id="toc-explore-data-in-r" class="nav-link" data-scroll-target="#explore-data-in-r">Explore data in R</a>
  <ul>
<li>
<a href="#arabidopsis" id="toc-arabidopsis" class="nav-link" data-scroll-target="#arabidopsis">üéÑ <em>Arabidopsis</em></a>
  <ul class="collapse">
<li><a href="#import" id="toc-import" class="nav-link" data-scroll-target="#import">Import</a></li>
  <li><a href="#distribution-of-values-across-all-the-data-in-the-file" id="toc-distribution-of-values-across-all-the-data-in-the-file" class="nav-link" data-scroll-target="#distribution-of-values-across-all-the-data-in-the-file">Distribution of values across all the data in the file</a></li>
  <li><a href="#distribution-of-values-across-the-samples" id="toc-distribution-of-values-across-the-samples" class="nav-link" data-scroll-target="#distribution-of-values-across-the-samples">Distribution of values across the samples</a></li>
  <li><a href="#distribution-of-values-across-the-genes" id="toc-distribution-of-values-across-the-genes" class="nav-link" data-scroll-target="#distribution-of-values-across-the-genes">Distribution of values across the genes</a></li>
  </ul>
</li>
  <li>
<a href="#leishmania" id="toc-leishmania" class="nav-link" data-scroll-target="#leishmania">üíâ <em>Leishmania</em></a>
  <ul class="collapse">
<li><a href="#import-1" id="toc-import-1" class="nav-link" data-scroll-target="#import-1">Import</a></li>
  <li><a href="#distribution-of-values-across-all-the-data-in-the-file-1" id="toc-distribution-of-values-across-all-the-data-in-the-file-1" class="nav-link" data-scroll-target="#distribution-of-values-across-all-the-data-in-the-file-1">Distribution of values across all the data in the file</a></li>
  <li><a href="#distribution-of-values-across-the-samples-1" id="toc-distribution-of-values-across-the-samples-1" class="nav-link" data-scroll-target="#distribution-of-values-across-the-samples-1">Distribution of values across the samples</a></li>
  <li><a href="#distribution-of-values-across-the-genes-1" id="toc-distribution-of-values-across-the-genes-1" class="nav-link" data-scroll-target="#distribution-of-values-across-the-genes-1">Distribution of values across the genes</a></li>
  </ul>
</li>
  <li>
<a href="#stem-cells" id="toc-stem-cells" class="nav-link" data-scroll-target="#stem-cells">üê≠ Stem cells</a>
  <ul class="collapse">
<li><a href="#import-2" id="toc-import-2" class="nav-link" data-scroll-target="#import-2">Import</a></li>
  <li><a href="#distribution-of-values-across-all-the-data-in-the-file-2" id="toc-distribution-of-values-across-all-the-data-in-the-file-2" class="nav-link" data-scroll-target="#distribution-of-values-across-all-the-data-in-the-file-2">Distribution of values across all the data in the file</a></li>
  <li><a href="#distribution-of-values-across-the-samples-2" id="toc-distribution-of-values-across-the-samples-2" class="nav-link" data-scroll-target="#distribution-of-values-across-the-samples-2">Distribution of values across the samples</a></li>
  <li><a href="#distribution-of-values-across-the-genes-2" id="toc-distribution-of-values-across-the-genes-2" class="nav-link" data-scroll-target="#distribution-of-values-across-the-genes-2">Distribution of values across the genes</a></li>
  </ul>
</li>
  </ul>
</li>
  <li>
<a href="#filtering-for-qc" id="toc-filtering-for-qc" class="nav-link" data-scroll-target="#filtering-for-qc">Filtering for QC</a>
  <ul>
<li><a href="#arabidopsis-filterqc" id="toc-arabidopsis-filterqc" class="nav-link" data-scroll-target="#arabidopsis-filterqc">üéÑ <em>Arabidopsis</em></a></li>
  <li><a href="#leishmania-filterqc" id="toc-leishmania-filterqc" class="nav-link" data-scroll-target="#leishmania-filterqc">üíâ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells-filterqc" id="toc-stem-cells-filterqc" class="nav-link" data-scroll-target="#stem-cells-filterqc">üê≠ Stem cells</a></li>
  </ul>
</li>
  <li>
<a href="#look-after-future-you" id="toc-look-after-future-you" class="nav-link" data-scroll-target="#look-after-future-you">ü§ó Look after future you!</a>
  <ul>
<li><a href="#arabidopsis-and-future-you" id="toc-arabidopsis-and-future-you" class="nav-link" data-scroll-target="#arabidopsis-and-future-you">üéÑ <em>Arabidopsis</em> and future you</a></li>
  <li><a href="#leishmania-and-future-you" id="toc-leishmania-and-future-you" class="nav-link" data-scroll-target="#leishmania-and-future-you">üíâ <em>Leishmania</em> and future you</a></li>
  <li><a href="#stem-cells-and-future-you" id="toc-stem-cells-and-future-you" class="nav-link" data-scroll-target="#stem-cells-and-future-you">üê≠ Stem cells and future you</a></li>
  </ul>
</li>
  <li><a href="#finished" id="toc-finished" class="nav-link" data-scroll-target="#finished">ü•≥ Finished</a></li>
  <li><a href="#independent-study-following-the-workshop" id="toc-independent-study-following-the-workshop" class="nav-link" data-scroll-target="#independent-study-following-the-workshop">Independent study following the workshop</a></li>
  <li><a href="#the-code-file" id="toc-the-code-file" class="nav-link" data-scroll-target="#the-code-file">The Code file</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/3mmaRand/BIO00088H-data/edit/main/transcriptomics/week-3/workshop.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li></ul></div><div class="quarto-other-links"><h2>Other Links</h2><ul><li><a href="#"><i class="bi bi-chat-left-dots"></i>Make a suggestion</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/week-3/overview.html">Week 3: Hello data!</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/week-3/workshop.html">Workshop</a></li></ol></nav><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Workshop</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
<p class="subtitle lead">Transcriptomics 1: Hello data!</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Emma Rand </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">29 September, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header><section id="introduction" class="level1"><h1>Introduction</h1>
<section id="session-overview" class="level2"><h2 class="anchored" data-anchor-id="session-overview">Session overview</h2>
<p>In this workshop you will learn what steps to take to get a good understanding of your transcriptomics data before you consider any statistical analysis. This is an often overlooked, but very valuable and informative, part of any data pipeline. It gives you the deep understanding of the data structures and values that you will need to code and trouble-shoot code, allows you to spot failed or problematic samples and informs your decisions on quality control.</p>
<p>In this session, you should examine <strong>all three data sets</strong> because the comparisons will give you a much stronger understanding of your own project data. Compare and contrast is a very useful way to build understanding.</p>
</section></section><section id="set-up-a-project" class="level1"><h1>Set up a Project</h1>
<p>üé¨ Start RStudio from the Start menu</p>
<p>üé¨ Make an RStudio project. Be deliberate about where you create it so that it is a good place for you</p>
<p>üé¨ Use the Files pane to make new folders for the data. I suggest <code>data-raw</code> and <code>data-processed</code></p>
<p>üé¨ Make a new script called <code>workshop-1.R</code> to carry out the rest of the work.</p>
<p>üé¨ Record what you do and what you find out. All of it!</p>
<p>üé¨ Load <code>tidyverse</code> <span class="citation" data-cites="tidyverse">(<a href="#ref-tidyverse" role="doc-biblioref">Wickham et al. 2019</a>)</span> for importing, summarising, plotting and filtering.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="examine-data-using-a-spreadsheet" class="level1"><h1>Examine data using a spreadsheet</h1>
<p>These are the three datasets. Each set compromises several files.</p>
<p>üéÑ <em>Arabidopsis</em>:</p>
<ul>
<li><a href="data-raw/arabidopsis-root.csv">arabidopsis-root.csv</a></li>
<li><a href="data-raw/arabidopsis-aerial.csv">arabidopsis-aerial.csv</a></li>
</ul>
<p>üíâ <em>Leishmania</em>:</p>
<ul>
<li><a href="data-raw/leishmania-mex-ama.csv">leishmania-mex-ama.csv</a></li>
<li><a href="data-raw/leishmania-mex-pro.csv">leishmania-mex-pro.csv</a></li>
<li><a href="data-raw/leishmania-mex-meta.csv">leishmania-mex-meta.csv</a></li>
</ul>
<p>üê≠ Stem cells:</p>
<ul>
<li><a href="data-raw/secretome_hspc.csv">secretome_hspc.csv</a></li>
<li><a href="data-raw/secretome_prog.csv">secretome_prog.csv</a></li>
<li><a href="data-raw/secretome_lthsc.csv">secretome_lthsc.csv</a></li>
</ul>
<p>üé¨ Save the files to <code>data-raw</code> and open them in Excel</p>
<p>üé¨ Answer the following questions:</p>
<ul>
<li>Describe how the sets of data are similar and how they are different.</li>
<li>What is in the rows and columns of each file?</li>
<li>How many rows and columns are there in each file? Are these the same? In all cases or some cases? Why?</li>
<li>Google gene id. Where does your search take you? How much information is available?</li>
</ul>
<p>üé¨ Did you record all that??</p>
</section><section id="explore-data-in-r" class="level1"><h1>Explore data in R</h1>
<p>The first task is to get an overview. We want to know</p>
<ul>
<li>how may zeros are there and how are they distributed?</li>
<li>what is the distribution of non-zero values?</li>
<li>does it look as though all the samples/cells were equally ‚Äúsuccessful‚Äù? Can we spot any problems?</li>
<li>can we determine a quality threshold for filtering out genes with very low counts?</li>
</ul>
<p>Genes which are zero in <em>every</em> cell/sample, i.e., genes that are not expressed at all give us no information. We will want to to filter those out. We will also want to filter out genes with very low counts for quality control. These are mostly likely false positives. If our data collection has gone well we would hope to see approximately the same average expression in each sample/cell of the same type. That is, replicates should be similar. In contrast, we would expect to see that the average expression of genes varies.</p>
<p>We get this overview by looking at:</p>
<ul>
<li><p>The <strong>distribution of values across all the data in the file</strong></p></li>
<li><p>The <strong>distribution of values across the samples/cells</strong> (i.e., averaged across genes). This allows us to see variation between samples/cells:</p></li>
<li><p>The <strong>distribution of values across the genes</strong> (i.e., averaged across samples/cells). This allows us to see variation between genes.</p></li>
</ul>
<p>The next sections will guide you through each of these for each of the data sets. Start with your own data set then move on to at least one of the others.</p>
<section id="arabidopsis" class="level2"><h2 class="anchored" data-anchor-id="arabidopsis">üéÑ <em>Arabidopsis</em>
</h2>
<section id="import" class="level3"><h3 class="anchored" data-anchor-id="import">Import</h3>
<p>Import the data for root tissue.</p>
<p>üé¨ Import <a href="data-raw/arabidopsis-root.csv">arabidopsis-root.csv</a></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># üéÑ import the root data</span></span>
<span><span class="va">root</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data-raw/arabidopsis-root.csv"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>üé¨ Check the dataframe has the number of rows and columns you were expecting and that column types and names are as expected.</p>
</section><section id="distribution-of-values-across-all-the-data-in-the-file" class="level3"><h3 class="anchored" data-anchor-id="distribution-of-values-across-all-the-data-in-the-file">Distribution of values across all the data in the file</h3>
<p>The values are spread over multiple columns so in order to plot the distribution as a whole, we will need to first use <code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer()</a></code> to put the data in <a href="https://3mmarand.github.io/R4BABS/r4babs1/week-9/workshop.html#tidying-data">‚Äòtidy‚Äô format</a> <span class="citation" data-cites="wickham2014">(<a href="#ref-wickham2014" role="doc-biblioref">Wickham 2014</a>)</span> by stacking the columns. We <em>could</em> save a copy of the stacked data and then plot it, but here, I have just piped the stacked data straight into <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code>. This helps me avoid cluttering my R environment with temporary objects.</p>
<p>üé¨ Pivot the counts (stack the columns) so all the counts are in a single column (<code>count</code>) labelled in <code>sample</code> by the column it came from and pipe into <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code> to create a histogram:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="va">gene_id</span>, <span class="op">-</span><span class="va">gene_name</span><span class="op">)</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-3-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="workshop_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>This data is very skewed - there are very many low counts and a very few higher numbers. It is hard to see the very low bars for the higher values. Logging the counts is a way to make the distribution more visible. You cannot take the log of 0 so we add 1 to the count before logging. The log of 1 is zero so we will be able to see how many zeros we had.</p>
<p>üé¨ Repeat the plot of log of the counts.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="va">gene_id</span>, <span class="op">-</span><span class="va">gene_name</span><span class="op">)</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-4-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="workshop_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>I‚Äôve used base 10 only because it easy to convert to the original scale (1 is 10, 2 is 100, 3 is 1000 etc). Notice we have a peak at zero indicating there are many zeros. We would expect the distribution of counts to be roughly log normal because this is expression of <em>all</em> the genes in the genome<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. The number of low counts is inflated (small peak near the low end). This suggests that these lower counts might be false positives. The removal of low counts is a common processing step in ‚Äôomic data. We will revisit this after we have considered the distribution of counts across samples and genes.</p>
</section><section id="distribution-of-values-across-the-samples" class="level3"><h3 class="anchored" data-anchor-id="distribution-of-values-across-the-samples">Distribution of values across the samples</h3>
<p>Summary statistics including the the number of NAs can be seen using the <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code>. It is most helpful which you have up to about 25 columns. There is nothing special about the number 25, it is just that summaries of a larger number of columns are difficult to grasp.</p>
<p>üé¨ Get a quick overview of the 14 columns:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># examine all the columns quickly</span></span>
<span><span class="co"># works well with smaller numbers of column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">root</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   gene_id           gene_name              CTR1               CTR2         
 Length:32833       Length:32833       Min.   :     0.0   Min.   :     0.0  
 Class :character   Class :character   1st Qu.:     0.0   1st Qu.:     0.0  
 Mode  :character   Mode  :character   Median :   100.0   Median :   131.0  
                                       Mean   :   705.9   Mean   :   930.5  
                                       3rd Qu.:   609.0   3rd Qu.:   811.0  
                                       Max.   :133872.0   Max.   :256188.0  
      CTR3               CTR4               CTR5               CTR6         
 Min.   :     0.0   Min.   :     0.0   Min.   :     0.0   Min.   :     0.0  
 1st Qu.:     0.0   1st Qu.:     0.0   1st Qu.:     0.0   1st Qu.:     0.0  
 Median :    94.0   Median :   114.0   Median :    97.0   Median :    95.0  
 Mean   :   715.2   Mean   :   786.5   Mean   :   731.4   Mean   :   703.7  
 3rd Qu.:   609.0   3rd Qu.:   686.0   3rd Qu.:   623.0   3rd Qu.:   608.0  
 Max.   :137167.0   Max.   :121160.0   Max.   :168910.0   Max.   :134192.0  
      LWR1               LWR2               LWR3               LWR4         
 Min.   :     0.0   Min.   :     0.0   Min.   :     0.0   Min.   :     0.0  
 1st Qu.:     0.0   1st Qu.:     0.0   1st Qu.:     0.0   1st Qu.:     0.0  
 Median :    81.0   Median :    77.0   Median :    78.0   Median :    80.0  
 Mean   :   730.1   Mean   :   622.2   Mean   :   647.1   Mean   :   662.3  
 3rd Qu.:   606.0   3rd Qu.:   524.0   3rd Qu.:   538.0   3rd Qu.:   566.0  
 Max.   :183421.0   Max.   :139241.0   Max.   :121856.0   Max.   :139044.0  
      LWR5               LWR6         
 Min.   :     0.0   Min.   :     0.0  
 1st Qu.:     0.0   1st Qu.:     0.0  
 Median :    93.0   Median :    89.0  
 Mean   :   739.7   Mean   :   742.8  
 3rd Qu.:   645.0   3rd Qu.:   625.0  
 Max.   :161872.0   Max.   :179613.0  </code></pre>
</div>
</div>
<p>Notice that:</p>
<ul>
<li>the minimum count is 0 and the maximums are very high in all the columns</li>
<li>the medians are quite a lot lower than the means so the data are skewed (hump to the left, tail to the right) and there must be quite a lot of zeros</li>
</ul>
<p>We want to know how many zeros there are in each a column. To achieve this, we can make use of the fact that <code>TRUE</code> evaluates to 1 and <code>FALSE</code> evaluates to 0. Consequently, summing a column of TRUE/FALSE values will give you the number of TRUE values. For example, <code>sum(CTR1 &gt; 0)</code> gives the number of values above zero in the <code>CTR1</code> column. If you wanted the number of zeros, you could use <code>sum(CTR1 == 0)</code>.</p>
<p>üé¨ Find the number values above zero in all six columns:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">CTR1</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">CTR2</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">CTR3</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">CTR4</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">CTR5</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">CTR6</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">LWR1</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">LWR2</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">LWR3</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">LWR4</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">LWR5</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">LWR6</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 √ó 12
  `sum(CTR1 &gt; 0)` `sum(CTR2 &gt; 0)` `sum(CTR3 &gt; 0)` `sum(CTR4 &gt; 0)`
            &lt;int&gt;           &lt;int&gt;           &lt;int&gt;           &lt;int&gt;
1           23119           23378           22789           23153
# ‚Ñπ 8 more variables: `sum(CTR5 &gt; 0)` &lt;int&gt;, `sum(CTR6 &gt; 0)` &lt;int&gt;,
#   `sum(LWR1 &gt; 0)` &lt;int&gt;, `sum(LWR2 &gt; 0)` &lt;int&gt;, `sum(LWR3 &gt; 0)` &lt;int&gt;,
#   `sum(LWR4 &gt; 0)` &lt;int&gt;, `sum(LWR5 &gt; 0)` &lt;int&gt;, `sum(LWR6 &gt; 0)` &lt;int&gt;</code></pre>
</div>
</div>
<p>There is a better way of doing this that saves you having to repeat so much code - very useful if you have a lot more than 6 columns! We can use <code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer()</a></code> to put the data in tidy format and then use the <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code> approach we have used extensively before.</p>
<p>üé¨ Find the number of zeros in all columns:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="va">gene_id</span>, <span class="op">-</span><span class="va">gene_name</span><span class="op">)</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>n_above_zero <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">count</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 √ó 2
   sample n_above_zero
   &lt;chr&gt;         &lt;int&gt;
 1 CTR1          23119
 2 CTR2          23378
 3 CTR3          22789
 4 CTR4          23153
 5 CTR5          22941
 6 CTR6          22982
 7 LWR1          23103
 8 LWR2          23111
 9 LWR3          22955
10 LWR4          22852
11 LWR5          23122
12 LWR6          23034</code></pre>
</div>
</div>
<p>You could expand this code to get get other useful summary information</p>
<p>üé¨ Summarise all the samples:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="va">gene_id</span>, <span class="op">-</span><span class="va">gene_name</span><span class="op">)</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>min <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>,</span>
<span>            lowerq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">count</span>, <span class="fl">0.25</span><span class="op">)</span>,</span>
<span>            mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>,</span>
<span>            median <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>,</span>
<span>            upperq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">count</span>, <span class="fl">0.75</span><span class="op">)</span>,</span>
<span>            max <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>,</span>
<span>            n_above_zero <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">count</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 √ó 8
   sample   min lowerq  mean median upperq    max n_above_zero
   &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;        &lt;int&gt;
 1 CTR1       0      0  706.    100    609 133872        23119
 2 CTR2       0      0  931.    131    811 256188        23378
 3 CTR3       0      0  715.     94    609 137167        22789
 4 CTR4       0      0  786.    114    686 121160        23153
 5 CTR5       0      0  731.     97    623 168910        22941
 6 CTR6       0      0  704.     95    608 134192        22982
 7 LWR1       0      0  730.     81    606 183421        23103
 8 LWR2       0      0  622.     77    524 139241        23111
 9 LWR3       0      0  647.     78    538 121856        22955
10 LWR4       0      0  662.     80    566 139044        22852
11 LWR5       0      0  740.     93    645 161872        23122
12 LWR6       0      0  743.     89    625 179613        23034</code></pre>
</div>
</div>
<p>The mean count ranges from 704 to 931. <code>CTR2</code> stands out a little - having a higher mean and maximum than the others. When we have a good number of replicates ‚Äì 6 is good in these experiments ‚Äì this is unlikely to be a problem. The potential effect of having an odd replicate when you have only two or three replicates, is reduced statistical power. Differences between genes with lower average expression and or more variable expression might be missed. Whether this matters depends on the biological question you are asking. In this case, it does not matter because a) we have 6 replicates and b) because the major differences in gene expression will be enough.</p>
<p>üé¨ Save the summary as a dataframe, <code>root_summary_samp</code> (using assignment).</p>
<p>We can also plot the distribution of counts across samples. We have many values (32833) so we are not limited to using <code><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram()</a></code>. <code><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density()</a></code> gives us a smooth distribution.</p>
<p>üé¨ Plot the log10 of the counts + 1 again but this time facet by the sample:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="va">gene_id</span>, <span class="op">-</span><span class="va">gene_name</span><span class="op">)</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">sample</span>, nrow <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-10-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="workshop_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>The key information to take from these plots is:</p>
<ul>
<li>the peak at zero suggests quite a few counts of 1.</li>
<li>we would expect the distribution of counts in each sample to be roughly log normal so that the rise near the low end, even before the peak at zero, suggests that these lower counts might be anomalies.</li>
</ul>
<p>We have found the distribution across samples to be similar to that over all. This is good because it means that the samples are fairly consistent with each other. We can now move on to the next step.</p>
</section><section id="distribution-of-values-across-the-genes" class="level3"><h3 class="anchored" data-anchor-id="distribution-of-values-across-the-genes">Distribution of values across the genes</h3>
<p>There are lots of genes in this dataset therefore we will take a slightly different approach. We would not want to use plot a distribution for each gene in the same way. Will pivot the data to tidy and then summarise the counts for each gene.</p>
<p>üé¨ Summarise the counts for each gene and save the result as <code>root_summary_gene</code>. Include the same columns as we had in the by sample summary (<code>root_summary_samp</code>) <em>and</em> an additional column, <code>total</code> for the total number of counts for each gene.</p>
<p>üé¨ View the <code>root_summary_gene</code> dataframe.</p>
<p>Notice that we have:</p>
<ul>
<li>a lot of genes with counts of zero in every sample</li>
<li>a lot of genes with zero counts in several of the samples</li>
<li>some very very low counts.</li>
</ul>
<p>Genes with very low counts should be filtered out because they are unreliable - or, at the least, uninformative. The goal of our downstream analysis will be to see if there is a significant difference in gene expression between the control and FGF-treated sibling. Since we have only three replicates in each group, having one or two unreliable, missing or zero values, makes such a determination impossible for a particular gene. We will use the total counts (<code>total</code>) and the number of samples with non-zero values (<code>n_above_zero</code>) in this dataframe to filter our genes later.</p>
<p>As we have a lot of genes, it is helpful to plot the mean counts with <code><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange()</a></code> to get an overview of the distributions. We will again plot the log of the mean counts. We will also order the genes from lowest to highest mean count.</p>
<p>üé¨ Plot the logged mean counts for each gene in order of size using <code><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange()</a></code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root_summary_gene</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="va">mean</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">mean</span> <span class="op">-</span> <span class="va">sd</span><span class="op">)</span>, </span>
<span>                      ymax <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">mean</span> <span class="op">+</span> <span class="va">sd</span> <span class="op">)</span><span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-12-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="workshop_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>(Note the warning is expected since we have zero means).</p>
<p>You can see we also have quite a few genes with means less than 1 (log below zero). Note that the variability between genes (average counts between 0 and 43348) is far greater than between samples (average counts from 70 to 296) which is exactly what we would expect to see.</p>
<p>Now go to <a href="#arabidopsis-filterqc">Filtering for QC</a>.</p>
</section></section><section id="leishmania" class="level2"><h2 class="anchored" data-anchor-id="leishmania">üíâ <em>Leishmania</em>
</h2>
<section id="import-1" class="level3"><h3 class="anchored" data-anchor-id="import-1">Import</h3>
<p>Import the data for <em>L.mexicana</em> procyclic promastigote (pro) and the metacyclic promastigotes (meta)</p>
<p>üé¨ Import <a href="data-raw/leishmania-mex-pro.csv">leishmania-mex-pro.csv</a> and <a href="data-raw/leishmania-mex-meta.csv">leishmania-mex-meta.csv</a></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># üíâ import the pro and meta leish data</span></span>
<span><span class="va">pro</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data-raw/leishmania-mex-pro.csv"</span><span class="op">)</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data-raw/leishmania-mex-meta.csv"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>We will need to combine the two sets of columns (datasets) so we can compare the two stages. We will join them using <code>gene_id</code> to match the rows. The column names differ so we don‚Äôt need to worry about renaming any of them.</p>
<p>üé¨ Combine the two datasets by <code>gene_id</code> and save the result as <code>pro_meta</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#  combine the two datasets</span></span>
<span><span class="va">pro_meta</span> <span class="op">&lt;-</span> <span class="va">pro</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">meta</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>üé¨ Check the dataframe has the number of rows and columns you were expecting and that column types and names are as expected.</p>
</section><section id="distribution-of-values-across-all-the-data-in-the-file-1" class="level3"><h3 class="anchored" data-anchor-id="distribution-of-values-across-all-the-data-in-the-file-1">Distribution of values across all the data in the file</h3>
<p>The values are spread over multiple columns so in order to plot the distribution as a whole, we will need to first use <code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer()</a></code> to put the data in <a href="https://3mmarand.github.io/R4BABS/r4babs1/week-9/workshop.html#tidying-data">‚Äòtidy‚Äô format</a> <span class="citation" data-cites="wickham2014">(<a href="#ref-wickham2014" role="doc-biblioref">Wickham 2014</a>)</span> by stacking the columns. We <em>could</em> save a copy of the stacked data and then plot it, but here, I have just piped the stacked data straight into <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code>. This helps me avoid cluttering my R environment with temporary objects.</p>
<p>üé¨ Pivot the counts (stack the columns) so all the counts are in a single column (<code>count</code>) labelled in <code>sample</code> by the column it came from and pipe into <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code> to create a histogram:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">gene_id</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-15-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="workshop_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>This data is very skewed - there are very many low counts and a very few higher numbers. It is hard to see the very low bars for the higher values. Logging the counts is a way to make the distribution more visible. You cannot take the log of 0 so we add 1 to the count before logging. The log of 1 is zero so we will be able to see how many zeros we had.</p>
<p>üé¨ Repeat the plot of log of the counts.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">gene_id</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-16-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="workshop_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>I‚Äôve used base 10 only because it easy to convert to the original scale (1 is 10, 2 is 100, 3 is 1000 etc). Notice we have a peak at zero indicating there are many zeros. We would expect the distribution of counts to be roughly log normal because this is expression of <em>all</em> the genes in the genome<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>. The number of low counts is inflated (small peak near the low end). This suggests that these lower counts might be false positives. The removal of low counts is a common processing step in ‚Äôomic data. We will revisit this after we have considered the distribution of counts across samples and genes.</p>
</section><section id="distribution-of-values-across-the-samples-1" class="level3"><h3 class="anchored" data-anchor-id="distribution-of-values-across-the-samples-1">Distribution of values across the samples</h3>
<p>Summary statistics including the the number of NAs can be seen using the <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code>. It is most helpful which you have up to about 25 columns. There is nothing special about the number 25, it is just that summaries of a larger number of columns are difficult to grasp.</p>
<p>üé¨ Get a quick overview of the 7 columns:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># examine all the columns quickly</span></span>
<span><span class="co"># works well with smaller numbers of column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">pro_meta</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   gene_id             lm_pro_1           lm_pro_2           lm_pro_3       
 Length:8677        Min.   :     0.0   Min.   :     0.0   Min.   :     0.0  
 Class :character   1st Qu.:    77.0   1st Qu.:    53.0   1st Qu.:    59.0  
 Mode  :character   Median :   191.0   Median :   135.0   Median :   145.0  
                    Mean   :   364.5   Mean   :   255.7   Mean   :   281.4  
                    3rd Qu.:   332.0   3rd Qu.:   238.0   3rd Qu.:   256.0  
                    Max.   :442477.0   Max.   :295423.0   Max.   :411663.0  
   lm_meta_1          lm_meta_2          lm_meta_3       
 Min.   :     0.0   Min.   :     0.0   Min.   :     0.0  
 1st Qu.:    48.0   1st Qu.:    51.0   1st Qu.:    78.0  
 Median :   110.0   Median :   120.0   Median :   187.0  
 Mean   :   220.3   Mean   :   221.9   Mean   :   355.9  
 3rd Qu.:   197.0   3rd Qu.:   215.0   3rd Qu.:   341.0  
 Max.   :244569.0   Max.   :205203.0   Max.   :498303.0  </code></pre>
</div>
</div>
<p>Notice that:</p>
<ul>
<li>the minimum count is 0 and the maximums are very high in all the columns</li>
<li>the medians are quite a lot lower than the means so the data are skewed (hump to the left, tail to the right) and there must be quite a lot of zeros</li>
</ul>
<p>We want to know how many zeros there are in each a column. To achieve this, we can make use of the fact that <code>TRUE</code> evaluates to 1 and <code>FALSE</code> evaluates to 0. Consequently, summing a column of TRUE/FALSE values will give you the number of TRUE values. For example, <code>sum(lm_pro_1 &gt; 0)</code> gives the number of values above zero in the <code>lm_pro_1</code> column. If you wanted the number of zeros, you could use <code>sum(lm_pro_1 == 0)</code>.</p>
<p>üé¨ Find the number values above zero in all six columns:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">lm_pro_1</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">lm_pro_2</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">lm_pro_3</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">lm_meta_1</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">lm_meta_2</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">lm_meta_3</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 √ó 6
  `sum(lm_pro_1 &gt; 0)` `sum(lm_pro_2 &gt; 0)` `sum(lm_pro_3 &gt; 0)`
                &lt;int&gt;               &lt;int&gt;               &lt;int&gt;
1                8549                8522                8509
# ‚Ñπ 3 more variables: `sum(lm_meta_1 &gt; 0)` &lt;int&gt;, `sum(lm_meta_2 &gt; 0)` &lt;int&gt;,
#   `sum(lm_meta_3 &gt; 0)` &lt;int&gt;</code></pre>
</div>
</div>
<p>There is a better way of doing this that saves you having to repeat so much code - very useful if you have a lot more than 6 columns! We can use <code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer()</a></code> to put the data in tidy format and then use the <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code> approach we have used extensively before.</p>
<p>üé¨ Find the number of zeros in all columns:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">gene_id</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>n_above_zero <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">count</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 √ó 2
  sample    n_above_zero
  &lt;chr&gt;            &lt;int&gt;
1 lm_meta_1         8535
2 lm_meta_2         8535
3 lm_meta_3         8530
4 lm_pro_1          8549
5 lm_pro_2          8522
6 lm_pro_3          8509</code></pre>
</div>
</div>
<p>You could expand this code to get get other useful summary information</p>
<p>üé¨ Summarise all the samples:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">gene_id</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>min <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>,</span>
<span>            lowerq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">count</span>, <span class="fl">0.25</span><span class="op">)</span>,</span>
<span>            mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>,</span>
<span>            median <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>,</span>
<span>            upperq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">count</span>, <span class="fl">0.75</span><span class="op">)</span>,</span>
<span>            max <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>,</span>
<span>            n_above_zero <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">count</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 √ó 8
  sample      min lowerq  mean median upperq    max n_above_zero
  &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;        &lt;int&gt;
1 lm_meta_1     0     48  220.    110    197 244569         8535
2 lm_meta_2     0     51  222.    120    215 205203         8535
3 lm_meta_3     0     78  356.    187    341 498303         8530
4 lm_pro_1      0     77  364.    191    332 442477         8549
5 lm_pro_2      0     53  256.    135    238 295423         8522
6 lm_pro_3      0     59  281.    145    256 411663         8509</code></pre>
</div>
</div>
<p>The mean count ranges from 220 to 364. We do not appear to have any outlying (odd) replicates. The potential effect of an odd replicate is reduced statistical power. Major differences in gene expression will still be uncovered. Differences between genes with lower average expression and or more variable expression might be missed. Whether this matters depends on the biological question you are asking.</p>
<p>üé¨ Save the summary as a dataframe, <code>pro_meta_summary_samp</code> (using assignment).</p>
<p>We can also plot the distribution of counts across samples. We have many values (8677) so we are not limited to using <code><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram()</a></code>. <code><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density()</a></code> gives us a smooth distribution.</p>
<p>üé¨ Plot the log10 of the counts + 1 again but this time facet by the sample:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">gene_id</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">sample</span>, nrow <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-22-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="workshop_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>The key information to take from these plots is:</p>
<ul>
<li>the distributions are roughly similar</li>
<li>the peak at zero suggests quite a few counts of 1.</li>
<li>we would expect the distribution of counts in each sample to be roughly log normal so that the small rise near the low end, even before the peak at zero, suggests that these lower counts might be anomalies.</li>
</ul>
<p>We have found the distribution across samples to be similar to that over all. This is good because it means that the samples are fairly consistent with each other. We can now move on to the next step.</p>
</section><section id="distribution-of-values-across-the-genes-1" class="level3"><h3 class="anchored" data-anchor-id="distribution-of-values-across-the-genes-1">Distribution of values across the genes</h3>
<p>There are lots of genes in this dataset therefore we will take a slightly different approach. We would not want to use plot a distribution for each gene in the same way. Will pivot the data to tidy and then summarise the counts for each gene.</p>
<p>üé¨ Summarise the counts for each gene and save the result as <code>pro_meta_summary_gene</code>. Include the same columns as we had in the by sample summary (<code>pro_meta_summary_samp</code>) <em>and</em> an additional column, <code>total</code> for the total number of counts for each gene.</p>
<p>üé¨ View the <code>pro_meta_summary_gene</code> dataframe.</p>
<p>Notice that we have:</p>
<ul>
<li>a lot of genes with counts of zero in every sample</li>
<li>a lot of genes with zero counts in several of the samples</li>
<li>some very very low counts.</li>
</ul>
<p>Genes with very low counts should be filtered out because they are unreliable - or, at the least, uninformative. The goal of our downstream analysis will be to see if there is a significant difference in gene expression between the stages. Since we have only three replicates in each group, having one or two unreliable, missing or zero values, makes such a determination impossible for a particular gene. We will use the total counts (<code>total</code>) and the number of samples with non-zero values (<code>n_above_zero</code>) in this dataframe to filter our genes later.</p>
<p>As we have a lot of genes, it is helpful to plot the mean counts with <code><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange()</a></code> to get an overview of the distributions. We will again plot the log of the mean counts. We will also order the genes from lowest to highest mean count.</p>
<p>üé¨ Plot the logged mean counts for each gene in order of size using <code><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange()</a></code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_summary_gene</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="va">mean</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">mean</span> <span class="op">-</span> <span class="va">sd</span><span class="op">)</span>, </span>
<span>                      ymax <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">mean</span> <span class="op">+</span> <span class="va">sd</span> <span class="op">)</span><span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-24-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="workshop_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>(Note the warning is expected since we have zero means).</p>
<p>You can see we also have quite a few genes with means less than 1 (log below zero). Note that the variability between genes (average counts between 0 and 349606) is far greater than between samples (average counts from 220 to 364) which is exactly what we would expect to see.</p>
<p>Now go to <a href="#leishmania-filterqc">Filtering for QC</a>.</p>
</section></section><section id="stem-cells" class="level2"><h2 class="anchored" data-anchor-id="stem-cells">üê≠ Stem cells</h2>
<section id="import-2" class="level3"><h3 class="anchored" data-anchor-id="import-2">Import</h3>
<p>Import the data for the HSPC and the Progenitor cells.</p>
<p>üé¨ Import <a href="data-raw/secretome_hspc.csv">secretome_hspc.csv</a> and <a href="data-raw/secretome_prog.csv">secretome_prog.csv</a></p>
<div class="cell" data-cache.extra="secretome">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># üê≠ import the hspc and prog data</span></span>
<span><span class="va">hspc</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data-raw/secretome_hspc.csv"</span><span class="op">)</span></span>
<span><span class="va">prog</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data-raw/secretome_prog.csv"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>We will need to combine the two sets of columns (datasets) so we can compare the two stages. We will join them using <code>ensembl_gene_id</code> to match the rows. The column names differ so we don‚Äôt need to worry about renaming any of them.</p>
<p>üé¨ Combine the two datasets by <code>ensembl_gene_id</code> and save the result as <code>hspc_prog</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#  combine the two datasets</span></span>
<span><span class="va">hspc_prog</span> <span class="op">&lt;-</span> <span class="va">hspc</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">prog</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>üé¨ Check the dataframe has the number of rows and columns you were expecting and that column types and names are as expected.</p>
</section><section id="distribution-of-values-across-all-the-data-in-the-file-2" class="level3"><h3 class="anchored" data-anchor-id="distribution-of-values-across-all-the-data-in-the-file-2">Distribution of values across all the data in the file</h3>
<p>The values are spread over multiple columns so in order to plot the distribution as a whole, we will need to first use <code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer()</a></code> to put the data in <a href="https://3mmarand.github.io/R4BABS/r4babs1/week-9/workshop.html#tidying-data">‚Äòtidy‚Äô format</a> <span class="citation" data-cites="wickham2014">(<a href="#ref-wickham2014" role="doc-biblioref">Wickham 2014</a>)</span> by stacking the columns. We <em>could</em> save a copy of the stacked data and then plot it, but here, I have just piped the stacked data straight into <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code>. This helps me avoid cluttering my R environment with temporary objects.</p>
<p>üé¨ Pivot the counts (stack the columns) so all the counts are in a single column (<code>expr</code>) labelled in <code>cell</code> by the column it came from and pipe into <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code> to create a histogram:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">ensembl_gene_id</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"expr"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">expr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-26-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="workshop_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>This is a very striking distribution. Is it what we are expecting? Notice we have a peak at zero indicating there are low values zeros. This inflation of low values suggests some are anomalous - they will have been derived from low counts which are likely false positives. As inaccurate measures, we will want to exclude expression values below (about) 1. We will revisit this after we have considered the distribution of expression across cells and genes.</p>
<p>What about the bimodal appearance of the the ‚Äòreal‚Äô values? If we had the whole transcriptome we would <em>not</em> expect to see such a pattern - we‚Äôd expect to see a roughly normal distribution<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>. However, this is a <em>subset</em> of the genome and the nature of the subsetting has had an influence here. These are a subset of cell surface proteins that show a significant difference between at least two of twelve cell subtypes. That is, all of these genes are either ‚Äúhigh‚Äù or ‚Äúlow‚Äù leading to a bimodal distribution.</p>
<p>Unlike the other three datasets, which count raw counts, these data are normalised and log<sub>2</sub> transformed. We do not need to plot the log of the values to see the distribution - they are already logged.</p>
</section><section id="distribution-of-values-across-the-samples-2" class="level3"><h3 class="anchored" data-anchor-id="distribution-of-values-across-the-samples-2">Distribution of values across the samples</h3>
<p>For the other three datasets, we used the <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> function to get an overview of the columns. This works well when you have upto about 25 columns but it is not helpful here because we have a lot of cells! Feel free to try it!</p>
<p>In this data set, there is even more of an advantage of using the <code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code> approach. We will be able to open the dataframe in the Viewer and make plots to examine whether the distributions are similar across cells. The mean and the standard deviation are useful to see the distributions across cells in a plot but we will also examine the interquartile values, maximums and the number of non-zero values.</p>
<p>üé¨ Summarise all the cells:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_summary_cell</span> <span class="op">&lt;-</span> <span class="va">hspc_prog</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">ensembl_gene_id</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"expr"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cell</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>min <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span>,</span>
<span>            lowerq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">expr</span>, <span class="fl">0.25</span><span class="op">)</span>,</span>
<span>            sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span>,</span>
<span>            mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span>,</span>
<span>            median <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span>,</span>
<span>            upperq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">expr</span>, <span class="fl">0.75</span><span class="op">)</span>,</span>
<span>            max <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span>,</span>
<span>            total <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span>,</span>
<span>            n_above_zero <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">expr</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>üé¨ View the <code>hspc_prog_summary_cell</code> dataframe (click on it in the environment).</p>
<p>Notice that: - a minimum value of 0 appears in all 1499 cells - the lower quartiles are all zero and so are many of the medians - there are no cells with above 0 expression in all 423 of the gene subset - the highest number of genes expressed is 287, the lowest is 139</p>
<p>In short, there are quite a lot of zeros.</p>
<p>To get a better understanding of the distribution of expressions in cells we can create a ggplot using the pointrange geom. Pointrange puts a dot at the mean and a line between a minimum and a maximum such as +/- one standard deviation. Not unlike a boxplot, but when you need the boxes too be very narrow!</p>
<p>üé¨ Create a pointrange plot.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_summary_cell</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cell</span>, y <span class="op">=</span> <span class="va">mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">mean</span> <span class="op">-</span> <span class="va">sd</span>, </span>
<span>                      ymax <span class="op">=</span> <span class="va">mean</span> <span class="op">+</span> <span class="va">sd</span> <span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-28-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="workshop_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>You will need to use the Zoom button to pop the plot window out so you can make it as wide as possible</p>
<p>The things to notice are:</p>
<ul>
<li>the average expression in cells is similar for all cells. This is good to know - if some cells had much lower expression perhaps there is something wrong with them, or their sequencing, and they should be excluded.</li>
<li>the distributions are roughly similar in width too</li>
</ul>
<p>The default order of <code>cell</code> is alphabetical. It can be easier to judge if there are unusual cells if we order the lines by the size of the mean.</p>
<p>üé¨ Order a pointrange plot with <code>reorder(variable_to_order, order_by)</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_summary_cell</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">mean</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">mean</span> <span class="op">-</span> <span class="va">sd</span>, </span>
<span>                      ymax <span class="op">=</span> <span class="va">mean</span> <span class="op">+</span> <span class="va">sd</span> <span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-29-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="workshop_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p><code><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder()</a></code> arranges <code>cell</code> in increasing size of <code>mean</code></p>
<p>As we thought, the distributions are similar across cells - there are not any cells that are obviously different from the others (only incrementally).</p>
</section><section id="distribution-of-values-across-the-genes-2" class="level3"><h3 class="anchored" data-anchor-id="distribution-of-values-across-the-genes-2">Distribution of values across the genes</h3>
<p>We will use the same approach to summarise the genes.</p>
<p>üé¨ Summarise the expression for each gene and save the result as <code>hspc_prog_summary_gene</code>. Include the same columns as we had in the by cell summary (<code>hspc_prog_summary_cell</code>) and an additional column, <code>total</code> for the total expression for each gene.</p>
<p>üé¨ View the <code>hspc_prog_summary_gene</code> dataframe. Remember these are normalised and logged (base 2) so we should not see very large values.</p>
<p>Notice that:</p>
<ul>
<li>some genes are expressed in every cell, and many are expressed in most cells</li>
<li>quite a few genes are zero in many cells. This this matters less when we have many cells (samples) than when we have few samples.</li>
<li>no genes have zeros in every cell - the lowest number of cells is is 8</li>
</ul>
<p>It is again helpful to plot the ordered mean expression with pointrange to get an overview.</p>
<p>üé¨ Plot the logged mean counts for each gene in order of size using <code><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange()</a></code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_summary_gene</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">ensembl_gene_id</span>, <span class="va">mean</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">mean</span> <span class="op">-</span> <span class="va">sd</span>, </span>
<span>                      ymax <span class="op">=</span> <span class="va">mean</span> <span class="op">+</span> <span class="va">sd</span><span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="workshop_files/figure-html/unnamed-chunk-31-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="workshop_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Note that the variability between genes (average expression between 0.0074 and 11.1016) is greater than between cells (average expression between 1.4083 and 3.4678) which is just what we would expect.</p>
<p>Now go to <a href="#stem-cells-filterqc">Filtering for QC</a>.</p>
</section></section></section><section id="filtering-for-qc" class="level1"><h1>Filtering for QC</h1>
<section id="arabidopsis-filterqc" class="level2"><h2 class="anchored" data-anchor-id="arabidopsis-filterqc">üéÑ <em>Arabidopsis</em>
</h2>
<p>Our samples look to be similarly well sequenced although this is difficult to determine with only two replicates. However, some genes are not expressed or the expression values are so low in for a gene that they are uninformative. We will filter the <code>root_summary_gene</code> dataframe to obtain a list of <code>gene_id</code> we can use to filter <code>root</code>.</p>
<p>My suggestion is to include only the genes with counts in at least 6 samples, and those with total counts above 20. I chose 6 because that would keep genes expressed only in one treatment. I chose 20 based on</p>
<p>üé¨ Filter the summary by gene dataframe:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root_summary_gene_filtered</span> <span class="op">&lt;-</span> <span class="va">root_summary_gene</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">total</span> <span class="op">&gt;</span> <span class="fl">20</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">n_above_zero</span> <span class="op">&gt;=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>‚ùì How many genes do you have left</p>
<!-- #---THINKING ANSWER--- -->
<!-- there are 22361 genes left. 9472 have total counts less than 20 and or -->
<!-- have counts in fewer the 6 samples. -->
<p>üé¨ Use the list of <code>gene_id</code> in the filtered summary to filter the original dataset:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root_filtered</span> <span class="op">&lt;-</span> <span class="va">root</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gene_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span>  <span class="va">root_summary_gene_filtered</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>üé¨ Write the filtered data to file:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">root_filtered</span>, </span>
<span>          file <span class="op">=</span> <span class="st">"data-processed/root_filtered.csv"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Now go to <a href="#arabidopsis-and-future-you">Look after future you</a></p>
</section><section id="leishmania-filterqc" class="level2"><h2 class="anchored" data-anchor-id="leishmania-filterqc">üíâ <em>Leishmania</em>
</h2>
<p>Our samples look to be similarly well sequenced. There are no samples we should remove. However, some genes are not expressed or the expression values are so low in for a gene that they are uninformative. We will filter the <code>pro_meta_summary_gene</code> dataframe to obtain a list of <code>gene_id</code> we can use to filter <code>pro_meta</code>.</p>
<p>My suggestion is to include only the genes with counts in at least 3 samples and those with total counts above 20. I chose 3 because that would keep genes expressed only in one treatment: [0, 0, 0] [#,#,#]. This is a difference we cannot test statistically, but which matters biologically.</p>
<p>üé¨ Filter the summary by gene dataframe:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_summary_gene_filtered</span> <span class="op">&lt;-</span> <span class="va">pro_meta_summary_gene</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">total</span> <span class="op">&gt;</span> <span class="fl">20</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">n_above_zero</span> <span class="op">&gt;=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>‚ùì How many genes do you have left</p>
<!-- #---THINKING ANSWER--- -->
<!-- there are 8437 genes left. 240 have total counts less than 20 and or -->
<!-- have counts in fewer than 3 samples. -->
<p>üé¨ Use the list of <code>gene_id</code> in the filtered summary to filter the original dataset:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_filtered</span> <span class="op">&lt;-</span> <span class="va">pro_meta</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gene_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span>  <span class="va">pro_meta_summary_gene_filtered</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>üé¨ Write the filtered data to file:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">pro_meta_filtered</span>, </span>
<span>          file <span class="op">=</span> <span class="st">"data-processed/pro_meta_filtered.csv"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Now go to <a href="#leishmania-and-future-you">Look after future you</a></p>
</section><section id="stem-cells-filterqc" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-filterqc">üê≠ Stem cells</h2>
<p>In this dataset, we will not see and genes that are not expressed in any of the cells because we are using a specific subset of the transcriptome that was deliberately selected. In other words unexpressed genes have already been filtered out. However, it is good practices to verify there are no unexpressed genes before we embark on our analysis.</p>
<p>Where the sum of all the values in the rows is zero, all the entries must be zero. We can use this to find any the genes that are not expressed in any of the cells. To do row wise aggregates such as the sum across rows we can use the <code><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise()</a></code> function. <code><a href="https://dplyr.tidyverse.org/reference/c_across.html">c_across()</a></code> allows us to use the colon notation to select columns. This is very useful when you have a lot of columns because it would be annoying to have to list all of them <code>HSPC_001:Prog_852</code> means all the columns from <code>HSPC_001</code> to <code>Prog_852</code>.</p>
<p>üé¨ Find the genes that are 0 in every column of the hspc_prog dataframe:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/c_across.html">c_across</a></span><span class="op">(</span><span class="va">HSPC_001</span><span class="op">:</span><span class="va">Prog_852</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 0 √ó 1,500
# Rowwise: 
# ‚Ñπ 1,500 variables: ensembl_gene_id &lt;chr&gt;, HSPC_001 &lt;dbl&gt;, HSPC_002 &lt;dbl&gt;,
#   HSPC_003 &lt;dbl&gt;, HSPC_004 &lt;dbl&gt;, HSPC_006 &lt;dbl&gt;, HSPC_008 &lt;dbl&gt;,
#   HSPC_009 &lt;dbl&gt;, HSPC_011 &lt;dbl&gt;, HSPC_012 &lt;dbl&gt;, HSPC_014 &lt;dbl&gt;,
#   HSPC_015 &lt;dbl&gt;, HSPC_016 &lt;dbl&gt;, HSPC_017 &lt;dbl&gt;, HSPC_018 &lt;dbl&gt;,
#   HSPC_020 &lt;dbl&gt;, HSPC_021 &lt;dbl&gt;, HSPC_022 &lt;dbl&gt;, HSPC_023 &lt;dbl&gt;,
#   HSPC_024 &lt;dbl&gt;, HSPC_025 &lt;dbl&gt;, HSPC_026 &lt;dbl&gt;, HSPC_027 &lt;dbl&gt;,
#   HSPC_028 &lt;dbl&gt;, HSPC_030 &lt;dbl&gt;, HSPC_031 &lt;dbl&gt;, HSPC_033 &lt;dbl&gt;, ‚Ä¶</code></pre>
</div>
</div>
<p>Notice that we have summed across all the columns.</p>
<p>‚ùì What do you conclude?</p>
<!-- #---THINKING ANSWER--- -->
<!-- The are no genes that are completely unexpressed in  -->
<!-- this set of genes -->
<p>üé¨ Write combined data to file:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">hspc_prog</span>, </span>
<span>          file <span class="op">=</span> <span class="st">"data-processed/hspc_prog.csv"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Now go to <a href="#stem-cells-and-future-you">Look after future you</a></p>
</section></section><section id="look-after-future-you" class="level1"><h1>ü§ó Look after future you!</h1>
<p>You need only do the section for your own project data but <strong>completing this step is important</strong>.</p>
<section id="arabidopsis-and-future-you" class="level2"><h2 class="anchored" data-anchor-id="arabidopsis-and-future-you">üéÑ <em>Arabidopsis</em> and future you</h2>
<p>üé¨ Create a new Project, <code>arab-88H</code>, populated with folders and your data. Make a script file called <code>cont-low-root.R</code>. This will a be commented analysis of comparison between control and low Ni root tissue. You will build on this each workshop and be able to use it as a template to examine other comparisons. Copy in the <em>appropriate code and comments</em> from <code>workshop-1.R</code>. Edit to improve your comments where your understanding has developed since you made them. Make sure you can close down RStudio, reopen it and run your whole script again.</p>
</section><section id="leishmania-and-future-you" class="level2"><h2 class="anchored" data-anchor-id="leishmania-and-future-you">üíâ <em>Leishmania</em> and future you</h2>
<p>üé¨ Create a new Project, <code>leish-88H</code>, populated with folders and your data. Make a script file called <code>pro_meta.R</code>. This will a be commented analysis of comparison procyclic promastigote and metacyclic promastigotes. You will build on this each workshop and be able to use it as a template to examine other comparisons. Copy in the <em>appropriate code and comments</em> from <code>workshop-1.R</code>. Edit to improve your comments where your understanding has developed since you made them. Make sure you can close down RStudio, reopen it and run your whole script again.</p>
</section><section id="stem-cells-and-future-you" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-and-future-you">üê≠ Stem cells and future you</h2>
<p>üé¨ Create a new Project, <code>mice-88H</code>, populated with folders and your data. Make a script file called <code>hspc-prog.R</code>. This will a be commented analysis of the hspc cells vs the prog cells. You will build on this each workshop and be able to use it as a template to examine other comparisons. Copy in the <em>appropriate code and comments</em> from <code>workshop-1.R</code>. Edit to improve your comments where your understanding has developed since you made them. Make sure you can close down RStudio, reopen it and run your whole script again.</p>
</section></section><section id="finished" class="level1"><h1>ü•≥ Finished</h1>
<p>Well Done!</p>
</section><section id="independent-study-following-the-workshop" class="level1"><h1>Independent study following the workshop</h1>
<p><a href="../../transcriptomics/week-3/study_after_workshop.html">Consolidate</a></p>
</section><section id="the-code-file" class="level1"><h1>The Code file</h1>
<p>These contain all the code needed in the workshop even where it is not visible on the webpage.</p>
<p>The <a href="../../transcriptomics/week-3/workshop.html">workshop.qmd</a> file is the file I use to compile the practical. Qmd stands for Quarto markdown. It allows code and ordinary text to be interleaved to produce well-formatted reports including webpages. Right-click on the link and choose View the source code for this workshop using the <code>&lt;/&gt; Code</code> button at the top of the page. Coding and thinking answers are marked with <code>#---CODING ANSWER---</code> and <code>#---THINKING ANSWER---</code></p>
<p>Pages made with R <span class="citation" data-cites="R-core">(<a href="#ref-R-core" role="doc-biblioref">R Core Team 2024</a>)</span>, Quarto <span class="citation" data-cites="Allaire_Quarto_2024">(<a href="#ref-Allaire_Quarto_2024" role="doc-biblioref">Allaire et al. 2024</a>)</span>, <code>knitr</code> <span class="citation" data-cites="knitr1 knitr2 knitr3">(<a href="#ref-knitr1" role="doc-biblioref">Xie 2024</a>, <a href="#ref-knitr2" role="doc-biblioref">2015</a>, <a href="#ref-knitr3" role="doc-biblioref">2014</a>)</span>, <code>kableExtra</code> <span class="citation" data-cites="kableExtra">(<a href="#ref-kableExtra" role="doc-biblioref">Zhu 2021</a>)</span></p>
</section><section id="references" class="level1">


<!-- -->


</section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Allaire_Quarto_2024" class="csl-entry" role="listitem">
Allaire, J. J., Charles Teague, Carlos Scheidegger, Yihui Xie, and Christophe Dervieux. 2024. <span>‚Äú<span>Quarto</span>.‚Äù</span> <a href="https://doi.org/10.5281/zenodo.5960048">https://doi.org/10.5281/zenodo.5960048</a>.
</div>
<div id="ref-R-core" class="csl-entry" role="listitem">
R Core Team. 2024. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-wickham2014" class="csl-entry" role="listitem">
Wickham, Hadley. 2014. <span>‚ÄúTidy Data.‚Äù</span> <em>Journal of Statistical Software, Articles</em> 59 (10): 1‚Äì23. <a href="https://vita.had.co.nz/papers/tidy-data.pdf">https://vita.had.co.nz/papers/tidy-data.pdf</a>.
</div>
<div id="ref-tidyverse" class="csl-entry" role="listitem">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D‚ÄôAgostino McGowan, Romain Fran√ßois, Garrett Grolemund, et al. 2019. <span>‚ÄúWelcome to the <span class="nocase">tidyverse</span>.‚Äù</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-knitr3" class="csl-entry" role="listitem">
Xie, Yihui. 2014. <span>‚ÄúKnitr: A Comprehensive Tool for Reproducible Research in <span>R</span>.‚Äù</span> In <em>Implementing Reproducible Computational Research</em>, edited by Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman; Hall/CRC.
</div>
<div id="ref-knitr2" class="csl-entry" role="listitem">
‚Äî‚Äî‚Äî. 2015. <em>Dynamic Documents with <span>R</span> and Knitr</em>. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-knitr1" class="csl-entry" role="listitem">
‚Äî‚Äî‚Äî. 2024. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in r</em>. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-kableExtra" class="csl-entry" role="listitem">
Zhu, Hao. 2021. <span>‚ÄúkableExtra: Construct Complex Table with ‚ÄôKable‚Äô and Pipe Syntax.‚Äù</span> <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>
<ol>
<li id="fn1"><p>This a result of the <a href="https://en.wikipedia.org/wiki/Central_limit_theorem">Central limit theorem</a>,one consequence of which is that adding together lots of distributions - whatever distributions they are - will tend to a normal distribution.<a href="#fnref1" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn2"><p>This a result of the <a href="https://en.wikipedia.org/wiki/Central_limit_theorem">Central limit theorem</a>,one consequence of which is that adding together lots of distributions - whatever distributions they are - will tend to a normal distribution.<a href="#fnref2" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn3"><p>This a result of the <a href="https://en.wikipedia.org/wiki/Central_limit_theorem">Central limit theorem</a>,one consequence of which is that adding together lots of distributions - whatever distributions they are - will tend to a normal distribution.<a href="#fnref3" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
</ol></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "Óßã";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("^(?:http:|https:)\/\/3mmarand\.github\.io\/BIO00088H-data");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb45" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Workshop"</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="an">subtitle:</span><span class="co"> "Transcriptomics 1: Hello data!"</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Emma Rand"</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="an">toc:</span><span class="co"> true</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a><span class="an">toc-depth:</span><span class="co"> 4</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="an">toc-location:</span><span class="co"> right</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a><span class="co">  echo: true</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a><span class="co">  include: true</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a><span class="co">  error: true</span></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a><span class="an">bibliography:</span><span class="co"> ../../references.bib</span></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span><span class="co"> </span></span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown: </span></span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a><span class="an">params:</span></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a><span class="co">  mouse: secretome  # or: surfaceome</span></span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a><span class="fu"># Introduction</span></span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session overview</span></span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-24"><a href="#cb45-24" aria-hidden="true" tabindex="-1"></a>In this workshop you will learn what steps to take to get a good</span>
<span id="cb45-25"><a href="#cb45-25" aria-hidden="true" tabindex="-1"></a>understanding of your transcriptomics data before you consider any</span>
<span id="cb45-26"><a href="#cb45-26" aria-hidden="true" tabindex="-1"></a>statistical analysis. This is an often overlooked, but very valuable and</span>
<span id="cb45-27"><a href="#cb45-27" aria-hidden="true" tabindex="-1"></a>informative, part of any data pipeline. It gives you the deep</span>
<span id="cb45-28"><a href="#cb45-28" aria-hidden="true" tabindex="-1"></a>understanding of the data structures and values that you will need to</span>
<span id="cb45-29"><a href="#cb45-29" aria-hidden="true" tabindex="-1"></a>code and trouble-shoot code, allows you to spot failed or problematic</span>
<span id="cb45-30"><a href="#cb45-30" aria-hidden="true" tabindex="-1"></a>samples and informs your decisions on quality control.</span>
<span id="cb45-31"><a href="#cb45-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-32"><a href="#cb45-32" aria-hidden="true" tabindex="-1"></a>In this session, you should examine **all three data sets** because the</span>
<span id="cb45-33"><a href="#cb45-33" aria-hidden="true" tabindex="-1"></a>comparisons will give you a much stronger understanding of your own</span>
<span id="cb45-34"><a href="#cb45-34" aria-hidden="true" tabindex="-1"></a>project data. Compare and contrast is a very useful way to build</span>
<span id="cb45-35"><a href="#cb45-35" aria-hidden="true" tabindex="-1"></a>understanding.</span>
<span id="cb45-36"><a href="#cb45-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-37"><a href="#cb45-37" aria-hidden="true" tabindex="-1"></a><span class="fu"># Set up a Project</span></span>
<span id="cb45-38"><a href="#cb45-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-39"><a href="#cb45-39" aria-hidden="true" tabindex="-1"></a>üé¨ Start RStudio from the Start menu</span>
<span id="cb45-40"><a href="#cb45-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-41"><a href="#cb45-41" aria-hidden="true" tabindex="-1"></a>üé¨ Make an RStudio project. Be deliberate about where you create it so</span>
<span id="cb45-42"><a href="#cb45-42" aria-hidden="true" tabindex="-1"></a>that it is a good place for you</span>
<span id="cb45-43"><a href="#cb45-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-44"><a href="#cb45-44" aria-hidden="true" tabindex="-1"></a>üé¨ Use the Files pane to make new folders for the data. I suggest</span>
<span id="cb45-45"><a href="#cb45-45" aria-hidden="true" tabindex="-1"></a><span class="in">`data-raw`</span> and <span class="in">`data-processed`</span></span>
<span id="cb45-46"><a href="#cb45-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-47"><a href="#cb45-47" aria-hidden="true" tabindex="-1"></a>üé¨ Make a new script called <span class="in">`workshop-1.R`</span> to carry out the rest of the</span>
<span id="cb45-48"><a href="#cb45-48" aria-hidden="true" tabindex="-1"></a>work.</span>
<span id="cb45-49"><a href="#cb45-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-50"><a href="#cb45-50" aria-hidden="true" tabindex="-1"></a>üé¨ Record what you do and what you find out. All of it!</span>
<span id="cb45-51"><a href="#cb45-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-52"><a href="#cb45-52" aria-hidden="true" tabindex="-1"></a>üé¨ Load <span class="in">`tidyverse`</span> <span class="co">[</span><span class="ot">@tidyverse</span><span class="co">]</span> for importing, summarising, plotting</span>
<span id="cb45-53"><a href="#cb45-53" aria-hidden="true" tabindex="-1"></a>and filtering.</span>
<span id="cb45-54"><a href="#cb45-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-57"><a href="#cb45-57" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-58"><a href="#cb45-58" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb45-59"><a href="#cb45-59" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-60"><a href="#cb45-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-61"><a href="#cb45-61" aria-hidden="true" tabindex="-1"></a><span class="fu"># Examine data using a spreadsheet</span></span>
<span id="cb45-62"><a href="#cb45-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-63"><a href="#cb45-63" aria-hidden="true" tabindex="-1"></a>These are the three datasets. Each set compromises several files.</span>
<span id="cb45-64"><a href="#cb45-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-65"><a href="#cb45-65" aria-hidden="true" tabindex="-1"></a>üéÑ *Arabidopsis*:</span>
<span id="cb45-66"><a href="#cb45-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-67"><a href="#cb45-67" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">arabidopsis-root.csv</span><span class="co">](data-raw/arabidopsis-root.csv)</span></span>
<span id="cb45-68"><a href="#cb45-68" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">arabidopsis-aerial.csv</span><span class="co">](data-raw/arabidopsis-aerial.csv)</span></span>
<span id="cb45-69"><a href="#cb45-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-70"><a href="#cb45-70" aria-hidden="true" tabindex="-1"></a>üíâ *Leishmania*:</span>
<span id="cb45-71"><a href="#cb45-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-72"><a href="#cb45-72" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">leishmania-mex-ama.csv</span><span class="co">](data-raw/leishmania-mex-ama.csv)</span></span>
<span id="cb45-73"><a href="#cb45-73" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">leishmania-mex-pro.csv</span><span class="co">](data-raw/leishmania-mex-pro.csv)</span></span>
<span id="cb45-74"><a href="#cb45-74" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">leishmania-mex-meta.csv</span><span class="co">](data-raw/leishmania-mex-meta.csv)</span></span>
<span id="cb45-75"><a href="#cb45-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-76"><a href="#cb45-76" aria-hidden="true" tabindex="-1"></a>üê≠ Stem cells:</span>
<span id="cb45-77"><a href="#cb45-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-78"><a href="#cb45-78" aria-hidden="true" tabindex="-1"></a>::: {.content-visible when-profile="secretome"}</span>
<span id="cb45-79"><a href="#cb45-79" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">secretome_hspc.csv</span><span class="co">](data-raw/secretome_hspc.csv)</span></span>
<span id="cb45-80"><a href="#cb45-80" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">secretome_prog.csv</span><span class="co">](data-raw/secretome_prog.csv)</span></span>
<span id="cb45-81"><a href="#cb45-81" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">secretome_lthsc.csv</span><span class="co">](data-raw/secretome_lthsc.csv)</span></span>
<span id="cb45-82"><a href="#cb45-82" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb45-83"><a href="#cb45-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-84"><a href="#cb45-84" aria-hidden="true" tabindex="-1"></a>::: {.content-visible when-profile="surfaceome"}</span>
<span id="cb45-85"><a href="#cb45-85" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">surfaceome_hspc.csv</span><span class="co">](data-raw/surfaceome_hspc.csv)</span></span>
<span id="cb45-86"><a href="#cb45-86" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">surfaceome_prog.csv</span><span class="co">](data-raw/surfaceome_prog.csv)</span></span>
<span id="cb45-87"><a href="#cb45-87" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">surfaceome_lthsc.csv</span><span class="co">](data-raw/surfaceome_lthsc.csv)</span></span>
<span id="cb45-88"><a href="#cb45-88" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb45-89"><a href="#cb45-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-90"><a href="#cb45-90" aria-hidden="true" tabindex="-1"></a>üé¨ Save the files to <span class="in">`data-raw`</span> and open them in Excel</span>
<span id="cb45-91"><a href="#cb45-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-92"><a href="#cb45-92" aria-hidden="true" tabindex="-1"></a>üé¨ Answer the following questions:</span>
<span id="cb45-93"><a href="#cb45-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-94"><a href="#cb45-94" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Describe how the sets of data are similar and how they are</span>
<span id="cb45-95"><a href="#cb45-95" aria-hidden="true" tabindex="-1"></a>    different.</span>
<span id="cb45-96"><a href="#cb45-96" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What is in the rows and columns of each file?</span>
<span id="cb45-97"><a href="#cb45-97" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How many rows and columns are there in each file? Are these the</span>
<span id="cb45-98"><a href="#cb45-98" aria-hidden="true" tabindex="-1"></a>    same? In all cases or some cases? Why?</span>
<span id="cb45-99"><a href="#cb45-99" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Google gene id. Where does your search take you? How much</span>
<span id="cb45-100"><a href="#cb45-100" aria-hidden="true" tabindex="-1"></a>    information is available?</span>
<span id="cb45-101"><a href="#cb45-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-102"><a href="#cb45-102" aria-hidden="true" tabindex="-1"></a>üé¨ Did you record all that??</span>
<span id="cb45-103"><a href="#cb45-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-104"><a href="#cb45-104" aria-hidden="true" tabindex="-1"></a><span class="fu"># Explore data in R</span></span>
<span id="cb45-105"><a href="#cb45-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-106"><a href="#cb45-106" aria-hidden="true" tabindex="-1"></a>The first task is to get an overview. We want to know</span>
<span id="cb45-107"><a href="#cb45-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-108"><a href="#cb45-108" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>how may zeros are there and how are they distributed?</span>
<span id="cb45-109"><a href="#cb45-109" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>what is the distribution of non-zero values?</span>
<span id="cb45-110"><a href="#cb45-110" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>does it look as though all the samples/cells were equally</span>
<span id="cb45-111"><a href="#cb45-111" aria-hidden="true" tabindex="-1"></a>    "successful"? Can we spot any problems?</span>
<span id="cb45-112"><a href="#cb45-112" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>can we determine a quality threshold for filtering out genes with</span>
<span id="cb45-113"><a href="#cb45-113" aria-hidden="true" tabindex="-1"></a>    very low counts?</span>
<span id="cb45-114"><a href="#cb45-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-115"><a href="#cb45-115" aria-hidden="true" tabindex="-1"></a>Genes which are zero in *every* cell/sample, i.e., genes that are not</span>
<span id="cb45-116"><a href="#cb45-116" aria-hidden="true" tabindex="-1"></a>expressed at all give us no information. We will want to to filter those</span>
<span id="cb45-117"><a href="#cb45-117" aria-hidden="true" tabindex="-1"></a>out. We will also want to filter out genes with very low counts for</span>
<span id="cb45-118"><a href="#cb45-118" aria-hidden="true" tabindex="-1"></a>quality control. These are mostly likely false positives. If our data</span>
<span id="cb45-119"><a href="#cb45-119" aria-hidden="true" tabindex="-1"></a>collection has gone well we would hope to see approximately the same</span>
<span id="cb45-120"><a href="#cb45-120" aria-hidden="true" tabindex="-1"></a>average expression in each sample/cell of the same type. That is,</span>
<span id="cb45-121"><a href="#cb45-121" aria-hidden="true" tabindex="-1"></a>replicates should be similar. In contrast, we would expect to see that</span>
<span id="cb45-122"><a href="#cb45-122" aria-hidden="true" tabindex="-1"></a>the average expression of genes varies.</span>
<span id="cb45-123"><a href="#cb45-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-124"><a href="#cb45-124" aria-hidden="true" tabindex="-1"></a>We get this overview by looking at:</span>
<span id="cb45-125"><a href="#cb45-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-126"><a href="#cb45-126" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The **distribution of values across all the data in the file**</span>
<span id="cb45-127"><a href="#cb45-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-128"><a href="#cb45-128" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The **distribution of values across the samples/cells** (i.e.,</span>
<span id="cb45-129"><a href="#cb45-129" aria-hidden="true" tabindex="-1"></a>    averaged across genes). This allows us to see variation between</span>
<span id="cb45-130"><a href="#cb45-130" aria-hidden="true" tabindex="-1"></a>    samples/cells:</span>
<span id="cb45-131"><a href="#cb45-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-132"><a href="#cb45-132" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The **distribution of values across the genes** (i.e., averaged</span>
<span id="cb45-133"><a href="#cb45-133" aria-hidden="true" tabindex="-1"></a>    across samples/cells). This allows us to see variation between</span>
<span id="cb45-134"><a href="#cb45-134" aria-hidden="true" tabindex="-1"></a>    genes.</span>
<span id="cb45-135"><a href="#cb45-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-136"><a href="#cb45-136" aria-hidden="true" tabindex="-1"></a>The next sections will guide you through each of these for each of the</span>
<span id="cb45-137"><a href="#cb45-137" aria-hidden="true" tabindex="-1"></a>data sets. Start with your own data set then move on to at least one of</span>
<span id="cb45-138"><a href="#cb45-138" aria-hidden="true" tabindex="-1"></a>the others.</span>
<span id="cb45-139"><a href="#cb45-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-140"><a href="#cb45-140" aria-hidden="true" tabindex="-1"></a><span class="fu">## üéÑ *Arabidopsis*</span></span>
<span id="cb45-141"><a href="#cb45-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-142"><a href="#cb45-142" aria-hidden="true" tabindex="-1"></a><span class="fu">### Import</span></span>
<span id="cb45-143"><a href="#cb45-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-144"><a href="#cb45-144" aria-hidden="true" tabindex="-1"></a>Import the data for root tissue.</span>
<span id="cb45-145"><a href="#cb45-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-146"><a href="#cb45-146" aria-hidden="true" tabindex="-1"></a>üé¨ Import <span class="co">[</span><span class="ot">arabidopsis-root.csv</span><span class="co">](data-raw/arabidopsis-root.csv)</span></span>
<span id="cb45-147"><a href="#cb45-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-150"><a href="#cb45-150" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-151"><a href="#cb45-151" aria-hidden="true" tabindex="-1"></a><span class="co"># üéÑ import the root data</span></span>
<span id="cb45-152"><a href="#cb45-152" aria-hidden="true" tabindex="-1"></a>root <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/arabidopsis-root.csv"</span>)</span>
<span id="cb45-153"><a href="#cb45-153" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-154"><a href="#cb45-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-155"><a href="#cb45-155" aria-hidden="true" tabindex="-1"></a>üé¨ Check the dataframe has the number of rows and columns you were</span>
<span id="cb45-156"><a href="#cb45-156" aria-hidden="true" tabindex="-1"></a>expecting and that column types and names are as expected.</span>
<span id="cb45-157"><a href="#cb45-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-158"><a href="#cb45-158" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distribution of values across all the data in the file</span></span>
<span id="cb45-159"><a href="#cb45-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-160"><a href="#cb45-160" aria-hidden="true" tabindex="-1"></a>The values are spread over multiple columns so in order to plot the</span>
<span id="cb45-161"><a href="#cb45-161" aria-hidden="true" tabindex="-1"></a>distribution as a whole, we will need to first use <span class="in">`pivot_longer()`</span> to</span>
<span id="cb45-162"><a href="#cb45-162" aria-hidden="true" tabindex="-1"></a>put the data in ['tidy'</span>
<span id="cb45-163"><a href="#cb45-163" aria-hidden="true" tabindex="-1"></a>format](https://3mmarand.github.io/R4BABS/r4babs1/week-9/workshop.html#tidying-data)</span>
<span id="cb45-164"><a href="#cb45-164" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@wickham2014</span><span class="co">]</span> by stacking the columns. We *could* save a copy of the</span>
<span id="cb45-165"><a href="#cb45-165" aria-hidden="true" tabindex="-1"></a>stacked data and then plot it, but here, I have just piped the stacked</span>
<span id="cb45-166"><a href="#cb45-166" aria-hidden="true" tabindex="-1"></a>data straight into <span class="in">`ggplot()`</span>. This helps me avoid cluttering my R</span>
<span id="cb45-167"><a href="#cb45-167" aria-hidden="true" tabindex="-1"></a>environment with temporary objects.</span>
<span id="cb45-168"><a href="#cb45-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-169"><a href="#cb45-169" aria-hidden="true" tabindex="-1"></a>üé¨ Pivot the counts (stack the columns) so all the counts are in a</span>
<span id="cb45-170"><a href="#cb45-170" aria-hidden="true" tabindex="-1"></a>single column (<span class="in">`count`</span>) labelled in <span class="in">`sample`</span> by the column it came from</span>
<span id="cb45-171"><a href="#cb45-171" aria-hidden="true" tabindex="-1"></a>and pipe into <span class="in">`ggplot()`</span> to create a histogram:</span>
<span id="cb45-172"><a href="#cb45-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-175"><a href="#cb45-175" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-176"><a href="#cb45-176" aria-hidden="true" tabindex="-1"></a>root <span class="sc">|&gt;</span></span>
<span id="cb45-177"><a href="#cb45-177" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="sc">-</span>gene_id, <span class="sc">-</span>gene_name),</span>
<span id="cb45-178"><a href="#cb45-178" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-179"><a href="#cb45-179" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-180"><a href="#cb45-180" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> count)) <span class="sc">+</span></span>
<span id="cb45-181"><a href="#cb45-181" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span>
<span id="cb45-182"><a href="#cb45-182" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-183"><a href="#cb45-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-184"><a href="#cb45-184" aria-hidden="true" tabindex="-1"></a>This data is very skewed - there are very many low counts and a very few</span>
<span id="cb45-185"><a href="#cb45-185" aria-hidden="true" tabindex="-1"></a>higher numbers. It is hard to see the very low bars for the higher</span>
<span id="cb45-186"><a href="#cb45-186" aria-hidden="true" tabindex="-1"></a>values. Logging the counts is a way to make the distribution more</span>
<span id="cb45-187"><a href="#cb45-187" aria-hidden="true" tabindex="-1"></a>visible. You cannot take the log of 0 so we add 1 to the count before</span>
<span id="cb45-188"><a href="#cb45-188" aria-hidden="true" tabindex="-1"></a>logging. The log of 1 is zero so we will be able to see how many zeros</span>
<span id="cb45-189"><a href="#cb45-189" aria-hidden="true" tabindex="-1"></a>we had.</span>
<span id="cb45-190"><a href="#cb45-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-191"><a href="#cb45-191" aria-hidden="true" tabindex="-1"></a>üé¨ Repeat the plot of log of the counts.</span>
<span id="cb45-192"><a href="#cb45-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-195"><a href="#cb45-195" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-196"><a href="#cb45-196" aria-hidden="true" tabindex="-1"></a>root <span class="sc">|&gt;</span></span>
<span id="cb45-197"><a href="#cb45-197" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="sc">-</span>gene_id, <span class="sc">-</span>gene_name),</span>
<span id="cb45-198"><a href="#cb45-198" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-199"><a href="#cb45-199" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-200"><a href="#cb45-200" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log10</span>(count <span class="sc">+</span> <span class="dv">1</span>))) <span class="sc">+</span></span>
<span id="cb45-201"><a href="#cb45-201" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span>
<span id="cb45-202"><a href="#cb45-202" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-203"><a href="#cb45-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-204"><a href="#cb45-204" aria-hidden="true" tabindex="-1"></a>I've used base 10 only because it easy to convert to the original scale</span>
<span id="cb45-205"><a href="#cb45-205" aria-hidden="true" tabindex="-1"></a>(1 is 10, 2 is 100, 3 is 1000 etc). Notice we have a peak at zero</span>
<span id="cb45-206"><a href="#cb45-206" aria-hidden="true" tabindex="-1"></a>indicating there are many zeros. We would expect the distribution of</span>
<span id="cb45-207"><a href="#cb45-207" aria-hidden="true" tabindex="-1"></a>counts to be roughly log normal because this is expression of *all* the</span>
<span id="cb45-208"><a href="#cb45-208" aria-hidden="true" tabindex="-1"></a>genes in the genome<span class="ot">[^1]</span>. The number of low counts is inflated (small</span>
<span id="cb45-209"><a href="#cb45-209" aria-hidden="true" tabindex="-1"></a>peak near the low end). This suggests that these lower counts might be</span>
<span id="cb45-210"><a href="#cb45-210" aria-hidden="true" tabindex="-1"></a>false positives. The removal of low counts is a common processing step</span>
<span id="cb45-211"><a href="#cb45-211" aria-hidden="true" tabindex="-1"></a>in 'omic data. We will revisit this after we have considered the</span>
<span id="cb45-212"><a href="#cb45-212" aria-hidden="true" tabindex="-1"></a>distribution of counts across samples and genes.</span>
<span id="cb45-213"><a href="#cb45-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-214"><a href="#cb45-214" aria-hidden="true" tabindex="-1"></a><span class="ot">[^1]: </span>This a result of the [Central limit</span>
<span id="cb45-215"><a href="#cb45-215" aria-hidden="true" tabindex="-1"></a>    theorem](https://en.wikipedia.org/wiki/Central_limit_theorem),one</span>
<span id="cb45-216"><a href="#cb45-216" aria-hidden="true" tabindex="-1"></a>    consequence of which is that adding together lots of distributions -</span>
<span id="cb45-217"><a href="#cb45-217" aria-hidden="true" tabindex="-1"></a>    whatever distributions they are - will tend to a normal</span>
<span id="cb45-218"><a href="#cb45-218" aria-hidden="true" tabindex="-1"></a>    distribution.</span>
<span id="cb45-219"><a href="#cb45-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-220"><a href="#cb45-220" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distribution of values across the samples</span></span>
<span id="cb45-221"><a href="#cb45-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-222"><a href="#cb45-222" aria-hidden="true" tabindex="-1"></a>Summary statistics including the the number of NAs can be seen using the</span>
<span id="cb45-223"><a href="#cb45-223" aria-hidden="true" tabindex="-1"></a><span class="in">`summary()`</span>. It is most helpful which you have up to about 25 columns.</span>
<span id="cb45-224"><a href="#cb45-224" aria-hidden="true" tabindex="-1"></a>There is nothing special about the number 25, it is just that summaries</span>
<span id="cb45-225"><a href="#cb45-225" aria-hidden="true" tabindex="-1"></a>of a larger number of columns are difficult to grasp.</span>
<span id="cb45-226"><a href="#cb45-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-227"><a href="#cb45-227" aria-hidden="true" tabindex="-1"></a>üé¨ Get a quick overview of the 14 columns:</span>
<span id="cb45-228"><a href="#cb45-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-231"><a href="#cb45-231" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-232"><a href="#cb45-232" aria-hidden="true" tabindex="-1"></a><span class="co"># examine all the columns quickly</span></span>
<span id="cb45-233"><a href="#cb45-233" aria-hidden="true" tabindex="-1"></a><span class="co"># works well with smaller numbers of column</span></span>
<span id="cb45-234"><a href="#cb45-234" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(root)</span>
<span id="cb45-235"><a href="#cb45-235" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-236"><a href="#cb45-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-237"><a href="#cb45-237" aria-hidden="true" tabindex="-1"></a>Notice that:</span>
<span id="cb45-238"><a href="#cb45-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-239"><a href="#cb45-239" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the minimum count is 0 and the maximums are very high in all the</span>
<span id="cb45-240"><a href="#cb45-240" aria-hidden="true" tabindex="-1"></a>    columns</span>
<span id="cb45-241"><a href="#cb45-241" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the medians are quite a lot lower than the means so the data are</span>
<span id="cb45-242"><a href="#cb45-242" aria-hidden="true" tabindex="-1"></a>    skewed (hump to the left, tail to the right) and there must be quite</span>
<span id="cb45-243"><a href="#cb45-243" aria-hidden="true" tabindex="-1"></a>    a lot of zeros</span>
<span id="cb45-244"><a href="#cb45-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-245"><a href="#cb45-245" aria-hidden="true" tabindex="-1"></a>We want to know how many zeros there are in each a column. To achieve</span>
<span id="cb45-246"><a href="#cb45-246" aria-hidden="true" tabindex="-1"></a>this, we can make use of the fact that <span class="in">`TRUE`</span> evaluates to 1 and <span class="in">`FALSE`</span></span>
<span id="cb45-247"><a href="#cb45-247" aria-hidden="true" tabindex="-1"></a>evaluates to 0. Consequently, summing a column of TRUE/FALSE values will</span>
<span id="cb45-248"><a href="#cb45-248" aria-hidden="true" tabindex="-1"></a>give you the number of TRUE values. For example, <span class="in">`sum(CTR1 &gt; 0)`</span> gives</span>
<span id="cb45-249"><a href="#cb45-249" aria-hidden="true" tabindex="-1"></a>the number of values above zero in the <span class="in">`CTR1`</span> column. If you wanted the</span>
<span id="cb45-250"><a href="#cb45-250" aria-hidden="true" tabindex="-1"></a>number of zeros, you could use <span class="in">`sum(CTR1 == 0)`</span>.</span>
<span id="cb45-251"><a href="#cb45-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-252"><a href="#cb45-252" aria-hidden="true" tabindex="-1"></a>üé¨ Find the number values above zero in all six columns:</span>
<span id="cb45-253"><a href="#cb45-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-256"><a href="#cb45-256" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-257"><a href="#cb45-257" aria-hidden="true" tabindex="-1"></a>root <span class="sc">|&gt;</span></span>
<span id="cb45-258"><a href="#cb45-258" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">sum</span>(CTR1 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-259"><a href="#cb45-259" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(CTR2 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-260"><a href="#cb45-260" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(CTR3 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-261"><a href="#cb45-261" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(CTR4 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-262"><a href="#cb45-262" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(CTR5 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-263"><a href="#cb45-263" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(CTR6 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-264"><a href="#cb45-264" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(LWR1 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-265"><a href="#cb45-265" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(LWR2 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-266"><a href="#cb45-266" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(LWR3 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-267"><a href="#cb45-267" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(LWR4 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-268"><a href="#cb45-268" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(LWR5 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-269"><a href="#cb45-269" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(LWR6 <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-270"><a href="#cb45-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-271"><a href="#cb45-271" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-272"><a href="#cb45-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-273"><a href="#cb45-273" aria-hidden="true" tabindex="-1"></a>There is a better way of doing this that saves you having to repeat so</span>
<span id="cb45-274"><a href="#cb45-274" aria-hidden="true" tabindex="-1"></a>much code - very useful if you have a lot more than 6 columns! We can</span>
<span id="cb45-275"><a href="#cb45-275" aria-hidden="true" tabindex="-1"></a>use <span class="in">`pivot_longer()`</span> to put the data in tidy format and then use the</span>
<span id="cb45-276"><a href="#cb45-276" aria-hidden="true" tabindex="-1"></a><span class="in">`group_by()`</span> and <span class="in">`summarise()`</span> approach we have used extensively before.</span>
<span id="cb45-277"><a href="#cb45-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-278"><a href="#cb45-278" aria-hidden="true" tabindex="-1"></a>üé¨ Find the number of zeros in all columns:</span>
<span id="cb45-279"><a href="#cb45-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-282"><a href="#cb45-282" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-283"><a href="#cb45-283" aria-hidden="true" tabindex="-1"></a>root <span class="sc">|&gt;</span></span>
<span id="cb45-284"><a href="#cb45-284" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="sc">-</span>gene_id, <span class="sc">-</span>gene_name),</span>
<span id="cb45-285"><a href="#cb45-285" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-286"><a href="#cb45-286" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-287"><a href="#cb45-287" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sample) <span class="sc">|&gt;</span></span>
<span id="cb45-288"><a href="#cb45-288" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n_above_zero =</span> <span class="fu">sum</span>(count <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-289"><a href="#cb45-289" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-290"><a href="#cb45-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-291"><a href="#cb45-291" aria-hidden="true" tabindex="-1"></a>You could expand this code to get get other useful summary information</span>
<span id="cb45-292"><a href="#cb45-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-293"><a href="#cb45-293" aria-hidden="true" tabindex="-1"></a>üé¨ Summarise all the samples:</span>
<span id="cb45-294"><a href="#cb45-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-297"><a href="#cb45-297" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-298"><a href="#cb45-298" aria-hidden="true" tabindex="-1"></a>root <span class="sc">|&gt;</span></span>
<span id="cb45-299"><a href="#cb45-299" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="sc">-</span>gene_id, <span class="sc">-</span>gene_name),</span>
<span id="cb45-300"><a href="#cb45-300" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-301"><a href="#cb45-301" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-302"><a href="#cb45-302" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sample) <span class="sc">|&gt;</span></span>
<span id="cb45-303"><a href="#cb45-303" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">min =</span> <span class="fu">min</span>(count),</span>
<span id="cb45-304"><a href="#cb45-304" aria-hidden="true" tabindex="-1"></a>            <span class="at">lowerq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.25</span>),</span>
<span id="cb45-305"><a href="#cb45-305" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean =</span> <span class="fu">mean</span>(count),</span>
<span id="cb45-306"><a href="#cb45-306" aria-hidden="true" tabindex="-1"></a>            <span class="at">median =</span> <span class="fu">median</span>(count),</span>
<span id="cb45-307"><a href="#cb45-307" aria-hidden="true" tabindex="-1"></a>            <span class="at">upperq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.75</span>),</span>
<span id="cb45-308"><a href="#cb45-308" aria-hidden="true" tabindex="-1"></a>            <span class="at">max =</span> <span class="fu">max</span>(count),</span>
<span id="cb45-309"><a href="#cb45-309" aria-hidden="true" tabindex="-1"></a>            <span class="at">n_above_zero =</span> <span class="fu">sum</span>(count <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-310"><a href="#cb45-310" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-311"><a href="#cb45-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-312"><a href="#cb45-312" aria-hidden="true" tabindex="-1"></a>The mean count ranges from 704 to 931. <span class="in">`CTR2`</span> stands out a little -</span>
<span id="cb45-313"><a href="#cb45-313" aria-hidden="true" tabindex="-1"></a>having a higher mean and maximum than the others. When we have a good</span>
<span id="cb45-314"><a href="#cb45-314" aria-hidden="true" tabindex="-1"></a>number of replicates -- 6 is good in these experiments -- this is</span>
<span id="cb45-315"><a href="#cb45-315" aria-hidden="true" tabindex="-1"></a>unlikely to be a problem. The potential effect of having an odd</span>
<span id="cb45-316"><a href="#cb45-316" aria-hidden="true" tabindex="-1"></a>replicate when you have only two or three replicates, is reduced</span>
<span id="cb45-317"><a href="#cb45-317" aria-hidden="true" tabindex="-1"></a>statistical power. Differences between genes with lower average</span>
<span id="cb45-318"><a href="#cb45-318" aria-hidden="true" tabindex="-1"></a>expression and or more variable expression might be missed. Whether this</span>
<span id="cb45-319"><a href="#cb45-319" aria-hidden="true" tabindex="-1"></a>matters depends on the biological question you are asking. In this case,</span>
<span id="cb45-320"><a href="#cb45-320" aria-hidden="true" tabindex="-1"></a>it does not matter because a) we have 6 replicates and b) because the</span>
<span id="cb45-321"><a href="#cb45-321" aria-hidden="true" tabindex="-1"></a>major differences in gene expression will be enough.</span>
<span id="cb45-322"><a href="#cb45-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-323"><a href="#cb45-323" aria-hidden="true" tabindex="-1"></a>üé¨ Save the summary as a dataframe, <span class="in">`root_summary_samp`</span> (using</span>
<span id="cb45-324"><a href="#cb45-324" aria-hidden="true" tabindex="-1"></a>assignment).</span>
<span id="cb45-325"><a href="#cb45-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-328"><a href="#cb45-328" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-329"><a href="#cb45-329" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb45-330"><a href="#cb45-330" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb45-331"><a href="#cb45-331" aria-hidden="true" tabindex="-1"></a>root_summary_samp <span class="ot">&lt;-</span> root <span class="sc">|&gt;</span></span>
<span id="cb45-332"><a href="#cb45-332" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="sc">-</span>gene_id, <span class="sc">-</span>gene_name),</span>
<span id="cb45-333"><a href="#cb45-333" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-334"><a href="#cb45-334" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-335"><a href="#cb45-335" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sample) <span class="sc">|&gt;</span></span>
<span id="cb45-336"><a href="#cb45-336" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">min =</span> <span class="fu">min</span>(count),</span>
<span id="cb45-337"><a href="#cb45-337" aria-hidden="true" tabindex="-1"></a>            <span class="at">lowerq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.25</span>),</span>
<span id="cb45-338"><a href="#cb45-338" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean =</span> <span class="fu">mean</span>(count),</span>
<span id="cb45-339"><a href="#cb45-339" aria-hidden="true" tabindex="-1"></a>            <span class="at">median =</span> <span class="fu">median</span>(count),</span>
<span id="cb45-340"><a href="#cb45-340" aria-hidden="true" tabindex="-1"></a>            <span class="at">upperq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.75</span>),</span>
<span id="cb45-341"><a href="#cb45-341" aria-hidden="true" tabindex="-1"></a>            <span class="at">max =</span> <span class="fu">max</span>(count),</span>
<span id="cb45-342"><a href="#cb45-342" aria-hidden="true" tabindex="-1"></a>            <span class="at">n_above_zero =</span> <span class="fu">sum</span>(count <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-343"><a href="#cb45-343" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-344"><a href="#cb45-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-345"><a href="#cb45-345" aria-hidden="true" tabindex="-1"></a>We can also plot the distribution of counts across samples. We have many</span>
<span id="cb45-346"><a href="#cb45-346" aria-hidden="true" tabindex="-1"></a>values (<span class="in">`r length(root$gene_id)`</span>) so we are not limited to using</span>
<span id="cb45-347"><a href="#cb45-347" aria-hidden="true" tabindex="-1"></a><span class="in">`geom_histogram()`</span>. <span class="in">`geom_density()`</span> gives us a smooth distribution.</span>
<span id="cb45-348"><a href="#cb45-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-349"><a href="#cb45-349" aria-hidden="true" tabindex="-1"></a>üé¨ Plot the log10 of the counts + 1 again but this time facet by the</span>
<span id="cb45-350"><a href="#cb45-350" aria-hidden="true" tabindex="-1"></a>sample:</span>
<span id="cb45-351"><a href="#cb45-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-354"><a href="#cb45-354" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-355"><a href="#cb45-355" aria-hidden="true" tabindex="-1"></a>root <span class="sc">|&gt;</span></span>
<span id="cb45-356"><a href="#cb45-356" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="sc">-</span>gene_id, <span class="sc">-</span>gene_name),</span>
<span id="cb45-357"><a href="#cb45-357" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-358"><a href="#cb45-358" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-359"><a href="#cb45-359" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="fu">log10</span>(count <span class="sc">+</span> <span class="dv">1</span>))) <span class="sc">+</span></span>
<span id="cb45-360"><a href="#cb45-360" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb45-361"><a href="#cb45-361" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(. <span class="sc">~</span> sample, <span class="at">nrow =</span> <span class="dv">3</span>)</span>
<span id="cb45-362"><a href="#cb45-362" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-363"><a href="#cb45-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-364"><a href="#cb45-364" aria-hidden="true" tabindex="-1"></a>The key information to take from these plots is:</span>
<span id="cb45-365"><a href="#cb45-365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-366"><a href="#cb45-366" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the peak at zero suggests quite a few counts of 1.</span>
<span id="cb45-367"><a href="#cb45-367" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>we would expect the distribution of counts in each sample to be</span>
<span id="cb45-368"><a href="#cb45-368" aria-hidden="true" tabindex="-1"></a>    roughly log normal so that the rise near the low end, even before</span>
<span id="cb45-369"><a href="#cb45-369" aria-hidden="true" tabindex="-1"></a>    the peak at zero, suggests that these lower counts might be</span>
<span id="cb45-370"><a href="#cb45-370" aria-hidden="true" tabindex="-1"></a>    anomalies.</span>
<span id="cb45-371"><a href="#cb45-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-372"><a href="#cb45-372" aria-hidden="true" tabindex="-1"></a>We have found the distribution across samples to be similar to that over</span>
<span id="cb45-373"><a href="#cb45-373" aria-hidden="true" tabindex="-1"></a>all. This is good because it means that the samples are fairly</span>
<span id="cb45-374"><a href="#cb45-374" aria-hidden="true" tabindex="-1"></a>consistent with each other. We can now move on to the next step.</span>
<span id="cb45-375"><a href="#cb45-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-376"><a href="#cb45-376" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distribution of values across the genes</span></span>
<span id="cb45-377"><a href="#cb45-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-378"><a href="#cb45-378" aria-hidden="true" tabindex="-1"></a>There are lots of genes in this dataset therefore we will take a</span>
<span id="cb45-379"><a href="#cb45-379" aria-hidden="true" tabindex="-1"></a>slightly different approach. We would not want to use plot a</span>
<span id="cb45-380"><a href="#cb45-380" aria-hidden="true" tabindex="-1"></a>distribution for each gene in the same way. Will pivot the data to tidy</span>
<span id="cb45-381"><a href="#cb45-381" aria-hidden="true" tabindex="-1"></a>and then summarise the counts for each gene.</span>
<span id="cb45-382"><a href="#cb45-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-383"><a href="#cb45-383" aria-hidden="true" tabindex="-1"></a>üé¨ Summarise the counts for each gene and save the result as</span>
<span id="cb45-384"><a href="#cb45-384" aria-hidden="true" tabindex="-1"></a><span class="in">`root_summary_gene`</span>. Include the same columns as we had in the by sample</span>
<span id="cb45-385"><a href="#cb45-385" aria-hidden="true" tabindex="-1"></a>summary (<span class="in">`root_summary_samp`</span>) *and* an additional column, <span class="in">`total`</span> for</span>
<span id="cb45-386"><a href="#cb45-386" aria-hidden="true" tabindex="-1"></a>the total number of counts for each gene.</span>
<span id="cb45-387"><a href="#cb45-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-390"><a href="#cb45-390" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-391"><a href="#cb45-391" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb45-392"><a href="#cb45-392" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb45-393"><a href="#cb45-393" aria-hidden="true" tabindex="-1"></a><span class="co"># you need to group by `gene_id` rather than `sample_id`</span></span>
<span id="cb45-394"><a href="#cb45-394" aria-hidden="true" tabindex="-1"></a><span class="co"># sum(count) will find the sum of counts</span></span>
<span id="cb45-395"><a href="#cb45-395" aria-hidden="true" tabindex="-1"></a>root_summary_gene <span class="ot">&lt;-</span> root <span class="sc">|&gt;</span></span>
<span id="cb45-396"><a href="#cb45-396" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="sc">-</span>gene_id, <span class="sc">-</span>gene_name),</span>
<span id="cb45-397"><a href="#cb45-397" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-398"><a href="#cb45-398" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-399"><a href="#cb45-399" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene_id) <span class="sc">|&gt;</span></span>
<span id="cb45-400"><a href="#cb45-400" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">min =</span> <span class="fu">min</span>(count),</span>
<span id="cb45-401"><a href="#cb45-401" aria-hidden="true" tabindex="-1"></a>            <span class="at">lowerq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.25</span>),</span>
<span id="cb45-402"><a href="#cb45-402" aria-hidden="true" tabindex="-1"></a>            <span class="at">sd =</span> <span class="fu">sd</span>(count),</span>
<span id="cb45-403"><a href="#cb45-403" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean =</span> <span class="fu">mean</span>(count),</span>
<span id="cb45-404"><a href="#cb45-404" aria-hidden="true" tabindex="-1"></a>            <span class="at">median =</span> <span class="fu">median</span>(count),</span>
<span id="cb45-405"><a href="#cb45-405" aria-hidden="true" tabindex="-1"></a>            <span class="at">upperq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.75</span>),</span>
<span id="cb45-406"><a href="#cb45-406" aria-hidden="true" tabindex="-1"></a>            <span class="at">max =</span> <span class="fu">max</span>(count),</span>
<span id="cb45-407"><a href="#cb45-407" aria-hidden="true" tabindex="-1"></a>            <span class="at">total =</span> <span class="fu">sum</span>(count),</span>
<span id="cb45-408"><a href="#cb45-408" aria-hidden="true" tabindex="-1"></a>            <span class="at">n_above_zero =</span> <span class="fu">sum</span>(count <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-409"><a href="#cb45-409" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-410"><a href="#cb45-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-411"><a href="#cb45-411" aria-hidden="true" tabindex="-1"></a>üé¨ View the <span class="in">`root_summary_gene`</span> dataframe.</span>
<span id="cb45-412"><a href="#cb45-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-413"><a href="#cb45-413" aria-hidden="true" tabindex="-1"></a>Notice that we have:</span>
<span id="cb45-414"><a href="#cb45-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-415"><a href="#cb45-415" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>a lot of genes with counts of zero in every sample</span>
<span id="cb45-416"><a href="#cb45-416" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>a lot of genes with zero counts in several of the samples</span>
<span id="cb45-417"><a href="#cb45-417" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>some very very low counts.</span>
<span id="cb45-418"><a href="#cb45-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-419"><a href="#cb45-419" aria-hidden="true" tabindex="-1"></a>Genes with very low counts should be filtered out because they are</span>
<span id="cb45-420"><a href="#cb45-420" aria-hidden="true" tabindex="-1"></a>unreliable - or, at the least, uninformative. The goal of our downstream</span>
<span id="cb45-421"><a href="#cb45-421" aria-hidden="true" tabindex="-1"></a>analysis will be to see if there is a significant difference in gene</span>
<span id="cb45-422"><a href="#cb45-422" aria-hidden="true" tabindex="-1"></a>expression between the control and FGF-treated sibling. Since we have</span>
<span id="cb45-423"><a href="#cb45-423" aria-hidden="true" tabindex="-1"></a>only three replicates in each group, having one or two unreliable,</span>
<span id="cb45-424"><a href="#cb45-424" aria-hidden="true" tabindex="-1"></a>missing or zero values, makes such a determination impossible for a</span>
<span id="cb45-425"><a href="#cb45-425" aria-hidden="true" tabindex="-1"></a>particular gene. We will use the total counts (<span class="in">`total`</span>) and the number</span>
<span id="cb45-426"><a href="#cb45-426" aria-hidden="true" tabindex="-1"></a>of samples with non-zero values (<span class="in">`n_above_zero`</span>) in this dataframe to</span>
<span id="cb45-427"><a href="#cb45-427" aria-hidden="true" tabindex="-1"></a>filter our genes later.</span>
<span id="cb45-428"><a href="#cb45-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-429"><a href="#cb45-429" aria-hidden="true" tabindex="-1"></a>As we have a lot of genes, it is helpful to plot the mean counts with</span>
<span id="cb45-430"><a href="#cb45-430" aria-hidden="true" tabindex="-1"></a><span class="in">`geom_pointrange()`</span> to get an overview of the distributions. We will</span>
<span id="cb45-431"><a href="#cb45-431" aria-hidden="true" tabindex="-1"></a>again plot the log of the mean counts. We will also order the genes from</span>
<span id="cb45-432"><a href="#cb45-432" aria-hidden="true" tabindex="-1"></a>lowest to highest mean count.</span>
<span id="cb45-433"><a href="#cb45-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-434"><a href="#cb45-434" aria-hidden="true" tabindex="-1"></a>üé¨ Plot the logged mean counts for each gene in order of size using</span>
<span id="cb45-435"><a href="#cb45-435" aria-hidden="true" tabindex="-1"></a><span class="in">`geom_pointrange()`</span>:</span>
<span id="cb45-436"><a href="#cb45-436" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-439"><a href="#cb45-439" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-440"><a href="#cb45-440" aria-hidden="true" tabindex="-1"></a>root_summary_gene <span class="sc">|&gt;</span> </span>
<span id="cb45-441"><a href="#cb45-441" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">reorder</span>(gene_id, mean), <span class="at">y =</span> <span class="fu">log10</span>(mean))) <span class="sc">+</span></span>
<span id="cb45-442"><a href="#cb45-442" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> <span class="fu">log10</span>(mean <span class="sc">-</span> sd), </span>
<span id="cb45-443"><a href="#cb45-443" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ymax =</span> <span class="fu">log10</span>(mean <span class="sc">+</span> sd )),</span>
<span id="cb45-444"><a href="#cb45-444" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="fl">0.1</span>)</span>
<span id="cb45-445"><a href="#cb45-445" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-446"><a href="#cb45-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-447"><a href="#cb45-447" aria-hidden="true" tabindex="-1"></a>(Note the warning is expected since we have zero means).</span>
<span id="cb45-448"><a href="#cb45-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-449"><a href="#cb45-449" aria-hidden="true" tabindex="-1"></a>You can see we also have quite a few genes with means less than 1 (log</span>
<span id="cb45-450"><a href="#cb45-450" aria-hidden="true" tabindex="-1"></a>below zero). Note that the variability between genes (average counts</span>
<span id="cb45-451"><a href="#cb45-451" aria-hidden="true" tabindex="-1"></a>between 0 and 43348) is far greater than between samples (average counts</span>
<span id="cb45-452"><a href="#cb45-452" aria-hidden="true" tabindex="-1"></a>from 70 to 296) which is exactly what we would expect to see.</span>
<span id="cb45-453"><a href="#cb45-453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-454"><a href="#cb45-454" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Filtering for QC</span><span class="co">](#arabidopsis-filterqc)</span>.</span>
<span id="cb45-455"><a href="#cb45-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-456"><a href="#cb45-456" aria-hidden="true" tabindex="-1"></a><span class="fu">## üíâ *Leishmania*</span></span>
<span id="cb45-457"><a href="#cb45-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-458"><a href="#cb45-458" aria-hidden="true" tabindex="-1"></a><span class="fu">### Import</span></span>
<span id="cb45-459"><a href="#cb45-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-460"><a href="#cb45-460" aria-hidden="true" tabindex="-1"></a>Import the data for *L.mexicana* procyclic promastigote (pro) and the</span>
<span id="cb45-461"><a href="#cb45-461" aria-hidden="true" tabindex="-1"></a>metacyclic promastigotes (meta)</span>
<span id="cb45-462"><a href="#cb45-462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-463"><a href="#cb45-463" aria-hidden="true" tabindex="-1"></a>üé¨ Import <span class="co">[</span><span class="ot">leishmania-mex-pro.csv</span><span class="co">](data-raw/leishmania-mex-pro.csv)</span> and</span>
<span id="cb45-464"><a href="#cb45-464" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">leishmania-mex-meta.csv</span><span class="co">](data-raw/leishmania-mex-meta.csv)</span></span>
<span id="cb45-465"><a href="#cb45-465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-468"><a href="#cb45-468" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-469"><a href="#cb45-469" aria-hidden="true" tabindex="-1"></a><span class="co"># üíâ import the pro and meta leish data</span></span>
<span id="cb45-470"><a href="#cb45-470" aria-hidden="true" tabindex="-1"></a>pro <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/leishmania-mex-pro.csv"</span>)</span>
<span id="cb45-471"><a href="#cb45-471" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/leishmania-mex-meta.csv"</span>)</span>
<span id="cb45-472"><a href="#cb45-472" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-473"><a href="#cb45-473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-474"><a href="#cb45-474" aria-hidden="true" tabindex="-1"></a>We will need to combine the two sets of columns (datasets) so we can</span>
<span id="cb45-475"><a href="#cb45-475" aria-hidden="true" tabindex="-1"></a>compare the two stages. We will join them using <span class="in">`gene_id`</span> to match the</span>
<span id="cb45-476"><a href="#cb45-476" aria-hidden="true" tabindex="-1"></a>rows. The column names differ so we don't need to worry about renaming</span>
<span id="cb45-477"><a href="#cb45-477" aria-hidden="true" tabindex="-1"></a>any of them.</span>
<span id="cb45-478"><a href="#cb45-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-479"><a href="#cb45-479" aria-hidden="true" tabindex="-1"></a>üé¨ Combine the two datasets by <span class="in">`gene_id`</span> and save the result as</span>
<span id="cb45-480"><a href="#cb45-480" aria-hidden="true" tabindex="-1"></a><span class="in">`pro_meta`</span>.</span>
<span id="cb45-481"><a href="#cb45-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-484"><a href="#cb45-484" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-485"><a href="#cb45-485" aria-hidden="true" tabindex="-1"></a><span class="co">#  combine the two datasets</span></span>
<span id="cb45-486"><a href="#cb45-486" aria-hidden="true" tabindex="-1"></a>pro_meta <span class="ot">&lt;-</span> pro <span class="sc">|&gt;</span></span>
<span id="cb45-487"><a href="#cb45-487" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(meta, </span>
<span id="cb45-488"><a href="#cb45-488" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"gene_id"</span>)</span>
<span id="cb45-489"><a href="#cb45-489" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-490"><a href="#cb45-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-491"><a href="#cb45-491" aria-hidden="true" tabindex="-1"></a>üé¨ Check the dataframe has the number of rows and columns you were</span>
<span id="cb45-492"><a href="#cb45-492" aria-hidden="true" tabindex="-1"></a>expecting and that column types and names are as expected.</span>
<span id="cb45-493"><a href="#cb45-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-494"><a href="#cb45-494" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distribution of values across all the data in the file</span></span>
<span id="cb45-495"><a href="#cb45-495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-496"><a href="#cb45-496" aria-hidden="true" tabindex="-1"></a>The values are spread over multiple columns so in order to plot the</span>
<span id="cb45-497"><a href="#cb45-497" aria-hidden="true" tabindex="-1"></a>distribution as a whole, we will need to first use <span class="in">`pivot_longer()`</span> to</span>
<span id="cb45-498"><a href="#cb45-498" aria-hidden="true" tabindex="-1"></a>put the data in ['tidy'</span>
<span id="cb45-499"><a href="#cb45-499" aria-hidden="true" tabindex="-1"></a>format](https://3mmarand.github.io/R4BABS/r4babs1/week-9/workshop.html#tidying-data)</span>
<span id="cb45-500"><a href="#cb45-500" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@wickham2014</span><span class="co">]</span> by stacking the columns. We *could* save a copy of the</span>
<span id="cb45-501"><a href="#cb45-501" aria-hidden="true" tabindex="-1"></a>stacked data and then plot it, but here, I have just piped the stacked</span>
<span id="cb45-502"><a href="#cb45-502" aria-hidden="true" tabindex="-1"></a>data straight into <span class="in">`ggplot()`</span>. This helps me avoid cluttering my R</span>
<span id="cb45-503"><a href="#cb45-503" aria-hidden="true" tabindex="-1"></a>environment with temporary objects.</span>
<span id="cb45-504"><a href="#cb45-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-505"><a href="#cb45-505" aria-hidden="true" tabindex="-1"></a>üé¨ Pivot the counts (stack the columns) so all the counts are in a</span>
<span id="cb45-506"><a href="#cb45-506" aria-hidden="true" tabindex="-1"></a>single column (<span class="in">`count`</span>) labelled in <span class="in">`sample`</span> by the column it came from</span>
<span id="cb45-507"><a href="#cb45-507" aria-hidden="true" tabindex="-1"></a>and pipe into <span class="in">`ggplot()`</span> to create a histogram:</span>
<span id="cb45-508"><a href="#cb45-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-511"><a href="#cb45-511" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-512"><a href="#cb45-512" aria-hidden="true" tabindex="-1"></a>pro_meta <span class="sc">|&gt;</span></span>
<span id="cb45-513"><a href="#cb45-513" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>gene_id,</span>
<span id="cb45-514"><a href="#cb45-514" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-515"><a href="#cb45-515" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-516"><a href="#cb45-516" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> count)) <span class="sc">+</span></span>
<span id="cb45-517"><a href="#cb45-517" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span>
<span id="cb45-518"><a href="#cb45-518" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-519"><a href="#cb45-519" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-520"><a href="#cb45-520" aria-hidden="true" tabindex="-1"></a>This data is very skewed - there are very many low counts and a very few</span>
<span id="cb45-521"><a href="#cb45-521" aria-hidden="true" tabindex="-1"></a>higher numbers. It is hard to see the very low bars for the higher</span>
<span id="cb45-522"><a href="#cb45-522" aria-hidden="true" tabindex="-1"></a>values. Logging the counts is a way to make the distribution more</span>
<span id="cb45-523"><a href="#cb45-523" aria-hidden="true" tabindex="-1"></a>visible. You cannot take the log of 0 so we add 1 to the count before</span>
<span id="cb45-524"><a href="#cb45-524" aria-hidden="true" tabindex="-1"></a>logging. The log of 1 is zero so we will be able to see how many zeros</span>
<span id="cb45-525"><a href="#cb45-525" aria-hidden="true" tabindex="-1"></a>we had.</span>
<span id="cb45-526"><a href="#cb45-526" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-527"><a href="#cb45-527" aria-hidden="true" tabindex="-1"></a>üé¨ Repeat the plot of log of the counts.</span>
<span id="cb45-528"><a href="#cb45-528" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-531"><a href="#cb45-531" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-532"><a href="#cb45-532" aria-hidden="true" tabindex="-1"></a>pro_meta <span class="sc">|&gt;</span></span>
<span id="cb45-533"><a href="#cb45-533" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>gene_id,</span>
<span id="cb45-534"><a href="#cb45-534" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-535"><a href="#cb45-535" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-536"><a href="#cb45-536" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log10</span>(count <span class="sc">+</span> <span class="dv">1</span>))) <span class="sc">+</span></span>
<span id="cb45-537"><a href="#cb45-537" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span>
<span id="cb45-538"><a href="#cb45-538" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-539"><a href="#cb45-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-540"><a href="#cb45-540" aria-hidden="true" tabindex="-1"></a>I've used base 10 only because it easy to convert to the original scale</span>
<span id="cb45-541"><a href="#cb45-541" aria-hidden="true" tabindex="-1"></a>(1 is 10, 2 is 100, 3 is 1000 etc). Notice we have a peak at zero</span>
<span id="cb45-542"><a href="#cb45-542" aria-hidden="true" tabindex="-1"></a>indicating there are many zeros. We would expect the distribution of</span>
<span id="cb45-543"><a href="#cb45-543" aria-hidden="true" tabindex="-1"></a>counts to be roughly log normal because this is expression of *all* the</span>
<span id="cb45-544"><a href="#cb45-544" aria-hidden="true" tabindex="-1"></a>genes in the genome<span class="ot">[^2]</span>. The number of low counts is inflated (small</span>
<span id="cb45-545"><a href="#cb45-545" aria-hidden="true" tabindex="-1"></a>peak near the low end). This suggests that these lower counts might be</span>
<span id="cb45-546"><a href="#cb45-546" aria-hidden="true" tabindex="-1"></a>false positives. The removal of low counts is a common processing step</span>
<span id="cb45-547"><a href="#cb45-547" aria-hidden="true" tabindex="-1"></a>in 'omic data. We will revisit this after we have considered the</span>
<span id="cb45-548"><a href="#cb45-548" aria-hidden="true" tabindex="-1"></a>distribution of counts across samples and genes.</span>
<span id="cb45-549"><a href="#cb45-549" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-550"><a href="#cb45-550" aria-hidden="true" tabindex="-1"></a><span class="ot">[^2]: </span>This a result of the [Central limit</span>
<span id="cb45-551"><a href="#cb45-551" aria-hidden="true" tabindex="-1"></a>    theorem](https://en.wikipedia.org/wiki/Central_limit_theorem),one</span>
<span id="cb45-552"><a href="#cb45-552" aria-hidden="true" tabindex="-1"></a>    consequence of which is that adding together lots of distributions -</span>
<span id="cb45-553"><a href="#cb45-553" aria-hidden="true" tabindex="-1"></a>    whatever distributions they are - will tend to a normal</span>
<span id="cb45-554"><a href="#cb45-554" aria-hidden="true" tabindex="-1"></a>    distribution.</span>
<span id="cb45-555"><a href="#cb45-555" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-556"><a href="#cb45-556" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distribution of values across the samples</span></span>
<span id="cb45-557"><a href="#cb45-557" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-558"><a href="#cb45-558" aria-hidden="true" tabindex="-1"></a>Summary statistics including the the number of NAs can be seen using the</span>
<span id="cb45-559"><a href="#cb45-559" aria-hidden="true" tabindex="-1"></a><span class="in">`summary()`</span>. It is most helpful which you have up to about 25 columns.</span>
<span id="cb45-560"><a href="#cb45-560" aria-hidden="true" tabindex="-1"></a>There is nothing special about the number 25, it is just that summaries</span>
<span id="cb45-561"><a href="#cb45-561" aria-hidden="true" tabindex="-1"></a>of a larger number of columns are difficult to grasp.</span>
<span id="cb45-562"><a href="#cb45-562" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-563"><a href="#cb45-563" aria-hidden="true" tabindex="-1"></a>üé¨ Get a quick overview of the 7 columns:</span>
<span id="cb45-564"><a href="#cb45-564" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-567"><a href="#cb45-567" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-568"><a href="#cb45-568" aria-hidden="true" tabindex="-1"></a><span class="co"># examine all the columns quickly</span></span>
<span id="cb45-569"><a href="#cb45-569" aria-hidden="true" tabindex="-1"></a><span class="co"># works well with smaller numbers of column</span></span>
<span id="cb45-570"><a href="#cb45-570" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pro_meta)</span>
<span id="cb45-571"><a href="#cb45-571" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-572"><a href="#cb45-572" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-573"><a href="#cb45-573" aria-hidden="true" tabindex="-1"></a>Notice that:</span>
<span id="cb45-574"><a href="#cb45-574" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-575"><a href="#cb45-575" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the minimum count is 0 and the maximums are very high in all the</span>
<span id="cb45-576"><a href="#cb45-576" aria-hidden="true" tabindex="-1"></a>    columns</span>
<span id="cb45-577"><a href="#cb45-577" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the medians are quite a lot lower than the means so the data are</span>
<span id="cb45-578"><a href="#cb45-578" aria-hidden="true" tabindex="-1"></a>    skewed (hump to the left, tail to the right) and there must be quite</span>
<span id="cb45-579"><a href="#cb45-579" aria-hidden="true" tabindex="-1"></a>    a lot of zeros</span>
<span id="cb45-580"><a href="#cb45-580" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-581"><a href="#cb45-581" aria-hidden="true" tabindex="-1"></a>We want to know how many zeros there are in each a column. To achieve</span>
<span id="cb45-582"><a href="#cb45-582" aria-hidden="true" tabindex="-1"></a>this, we can make use of the fact that <span class="in">`TRUE`</span> evaluates to 1 and <span class="in">`FALSE`</span></span>
<span id="cb45-583"><a href="#cb45-583" aria-hidden="true" tabindex="-1"></a>evaluates to 0. Consequently, summing a column of TRUE/FALSE values will</span>
<span id="cb45-584"><a href="#cb45-584" aria-hidden="true" tabindex="-1"></a>give you the number of TRUE values. For example, <span class="in">`sum(lm_pro_1 &gt; 0)`</span></span>
<span id="cb45-585"><a href="#cb45-585" aria-hidden="true" tabindex="-1"></a>gives the number of values above zero in the <span class="in">`lm_pro_1`</span> column. If you</span>
<span id="cb45-586"><a href="#cb45-586" aria-hidden="true" tabindex="-1"></a>wanted the number of zeros, you could use <span class="in">`sum(lm_pro_1 == 0)`</span>.</span>
<span id="cb45-587"><a href="#cb45-587" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-588"><a href="#cb45-588" aria-hidden="true" tabindex="-1"></a>üé¨ Find the number values above zero in all six columns:</span>
<span id="cb45-589"><a href="#cb45-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-592"><a href="#cb45-592" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-593"><a href="#cb45-593" aria-hidden="true" tabindex="-1"></a>pro_meta <span class="sc">|&gt;</span></span>
<span id="cb45-594"><a href="#cb45-594" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">sum</span>(lm_pro_1 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-595"><a href="#cb45-595" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(lm_pro_2 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-596"><a href="#cb45-596" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(lm_pro_3 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-597"><a href="#cb45-597" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(lm_meta_1 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-598"><a href="#cb45-598" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(lm_meta_2 <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb45-599"><a href="#cb45-599" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(lm_meta_3 <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-600"><a href="#cb45-600" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-601"><a href="#cb45-601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-602"><a href="#cb45-602" aria-hidden="true" tabindex="-1"></a>There is a better way of doing this that saves you having to repeat so</span>
<span id="cb45-603"><a href="#cb45-603" aria-hidden="true" tabindex="-1"></a>much code - very useful if you have a lot more than 6 columns! We can</span>
<span id="cb45-604"><a href="#cb45-604" aria-hidden="true" tabindex="-1"></a>use <span class="in">`pivot_longer()`</span> to put the data in tidy format and then use the</span>
<span id="cb45-605"><a href="#cb45-605" aria-hidden="true" tabindex="-1"></a><span class="in">`group_by()`</span> and <span class="in">`summarise()`</span> approach we have used extensively before.</span>
<span id="cb45-606"><a href="#cb45-606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-607"><a href="#cb45-607" aria-hidden="true" tabindex="-1"></a>üé¨ Find the number of zeros in all columns:</span>
<span id="cb45-608"><a href="#cb45-608" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-611"><a href="#cb45-611" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-612"><a href="#cb45-612" aria-hidden="true" tabindex="-1"></a>pro_meta <span class="sc">|&gt;</span></span>
<span id="cb45-613"><a href="#cb45-613" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>gene_id,</span>
<span id="cb45-614"><a href="#cb45-614" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-615"><a href="#cb45-615" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-616"><a href="#cb45-616" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sample) <span class="sc">|&gt;</span></span>
<span id="cb45-617"><a href="#cb45-617" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n_above_zero =</span> <span class="fu">sum</span>(count <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-618"><a href="#cb45-618" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-619"><a href="#cb45-619" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-620"><a href="#cb45-620" aria-hidden="true" tabindex="-1"></a>You could expand this code to get get other useful summary information</span>
<span id="cb45-621"><a href="#cb45-621" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-622"><a href="#cb45-622" aria-hidden="true" tabindex="-1"></a>üé¨ Summarise all the samples:</span>
<span id="cb45-623"><a href="#cb45-623" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-626"><a href="#cb45-626" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-627"><a href="#cb45-627" aria-hidden="true" tabindex="-1"></a>pro_meta <span class="sc">|&gt;</span></span>
<span id="cb45-628"><a href="#cb45-628" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>gene_id,</span>
<span id="cb45-629"><a href="#cb45-629" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-630"><a href="#cb45-630" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-631"><a href="#cb45-631" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sample) <span class="sc">|&gt;</span></span>
<span id="cb45-632"><a href="#cb45-632" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">min =</span> <span class="fu">min</span>(count),</span>
<span id="cb45-633"><a href="#cb45-633" aria-hidden="true" tabindex="-1"></a>            <span class="at">lowerq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.25</span>),</span>
<span id="cb45-634"><a href="#cb45-634" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean =</span> <span class="fu">mean</span>(count),</span>
<span id="cb45-635"><a href="#cb45-635" aria-hidden="true" tabindex="-1"></a>            <span class="at">median =</span> <span class="fu">median</span>(count),</span>
<span id="cb45-636"><a href="#cb45-636" aria-hidden="true" tabindex="-1"></a>            <span class="at">upperq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.75</span>),</span>
<span id="cb45-637"><a href="#cb45-637" aria-hidden="true" tabindex="-1"></a>            <span class="at">max =</span> <span class="fu">max</span>(count),</span>
<span id="cb45-638"><a href="#cb45-638" aria-hidden="true" tabindex="-1"></a>            <span class="at">n_above_zero =</span> <span class="fu">sum</span>(count <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-639"><a href="#cb45-639" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-640"><a href="#cb45-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-641"><a href="#cb45-641" aria-hidden="true" tabindex="-1"></a>The mean count ranges from 220 to 364. We do not appear to have any</span>
<span id="cb45-642"><a href="#cb45-642" aria-hidden="true" tabindex="-1"></a>outlying (odd) replicates. The potential effect of an odd replicate is</span>
<span id="cb45-643"><a href="#cb45-643" aria-hidden="true" tabindex="-1"></a>reduced statistical power. Major differences in gene expression will</span>
<span id="cb45-644"><a href="#cb45-644" aria-hidden="true" tabindex="-1"></a>still be uncovered. Differences between genes with lower average</span>
<span id="cb45-645"><a href="#cb45-645" aria-hidden="true" tabindex="-1"></a>expression and or more variable expression might be missed. Whether this</span>
<span id="cb45-646"><a href="#cb45-646" aria-hidden="true" tabindex="-1"></a>matters depends on the biological question you are asking.</span>
<span id="cb45-647"><a href="#cb45-647" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-648"><a href="#cb45-648" aria-hidden="true" tabindex="-1"></a>üé¨ Save the summary as a dataframe, <span class="in">`pro_meta_summary_samp`</span> (using</span>
<span id="cb45-649"><a href="#cb45-649" aria-hidden="true" tabindex="-1"></a>assignment).</span>
<span id="cb45-650"><a href="#cb45-650" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-653"><a href="#cb45-653" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-654"><a href="#cb45-654" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb45-655"><a href="#cb45-655" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb45-656"><a href="#cb45-656" aria-hidden="true" tabindex="-1"></a>pro_meta_summary_samp <span class="ot">&lt;-</span> pro_meta <span class="sc">|&gt;</span></span>
<span id="cb45-657"><a href="#cb45-657" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>gene_id,</span>
<span id="cb45-658"><a href="#cb45-658" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-659"><a href="#cb45-659" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-660"><a href="#cb45-660" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sample) <span class="sc">|&gt;</span></span>
<span id="cb45-661"><a href="#cb45-661" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">min =</span> <span class="fu">min</span>(count),</span>
<span id="cb45-662"><a href="#cb45-662" aria-hidden="true" tabindex="-1"></a>            <span class="at">lowerq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.25</span>),</span>
<span id="cb45-663"><a href="#cb45-663" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean =</span> <span class="fu">mean</span>(count),</span>
<span id="cb45-664"><a href="#cb45-664" aria-hidden="true" tabindex="-1"></a>            <span class="at">median =</span> <span class="fu">median</span>(count),</span>
<span id="cb45-665"><a href="#cb45-665" aria-hidden="true" tabindex="-1"></a>            <span class="at">upperq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.75</span>),</span>
<span id="cb45-666"><a href="#cb45-666" aria-hidden="true" tabindex="-1"></a>            <span class="at">max =</span> <span class="fu">max</span>(count),</span>
<span id="cb45-667"><a href="#cb45-667" aria-hidden="true" tabindex="-1"></a>            <span class="at">n_above_zero =</span> <span class="fu">sum</span>(count <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-668"><a href="#cb45-668" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-669"><a href="#cb45-669" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-670"><a href="#cb45-670" aria-hidden="true" tabindex="-1"></a>We can also plot the distribution of counts across samples. We have many</span>
<span id="cb45-671"><a href="#cb45-671" aria-hidden="true" tabindex="-1"></a>values (<span class="in">`r length(pro_meta$gene_id)`</span>) so we are not limited to using</span>
<span id="cb45-672"><a href="#cb45-672" aria-hidden="true" tabindex="-1"></a><span class="in">`geom_histogram()`</span>. <span class="in">`geom_density()`</span> gives us a smooth distribution.</span>
<span id="cb45-673"><a href="#cb45-673" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-674"><a href="#cb45-674" aria-hidden="true" tabindex="-1"></a>üé¨ Plot the log10 of the counts + 1 again but this time facet by the</span>
<span id="cb45-675"><a href="#cb45-675" aria-hidden="true" tabindex="-1"></a>sample:</span>
<span id="cb45-676"><a href="#cb45-676" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-679"><a href="#cb45-679" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-680"><a href="#cb45-680" aria-hidden="true" tabindex="-1"></a>pro_meta <span class="sc">|&gt;</span></span>
<span id="cb45-681"><a href="#cb45-681" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>gene_id,</span>
<span id="cb45-682"><a href="#cb45-682" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-683"><a href="#cb45-683" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-684"><a href="#cb45-684" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="fu">log10</span>(count <span class="sc">+</span> <span class="dv">1</span>))) <span class="sc">+</span></span>
<span id="cb45-685"><a href="#cb45-685" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb45-686"><a href="#cb45-686" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(. <span class="sc">~</span> sample, <span class="at">nrow =</span> <span class="dv">3</span>)</span>
<span id="cb45-687"><a href="#cb45-687" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-688"><a href="#cb45-688" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-689"><a href="#cb45-689" aria-hidden="true" tabindex="-1"></a>The key information to take from these plots is:</span>
<span id="cb45-690"><a href="#cb45-690" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-691"><a href="#cb45-691" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the distributions are roughly similar</span>
<span id="cb45-692"><a href="#cb45-692" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the peak at zero suggests quite a few counts of 1.</span>
<span id="cb45-693"><a href="#cb45-693" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>we would expect the distribution of counts in each sample to be</span>
<span id="cb45-694"><a href="#cb45-694" aria-hidden="true" tabindex="-1"></a>    roughly log normal so that the small rise near the low end, even</span>
<span id="cb45-695"><a href="#cb45-695" aria-hidden="true" tabindex="-1"></a>    before the peak at zero, suggests that these lower counts might be</span>
<span id="cb45-696"><a href="#cb45-696" aria-hidden="true" tabindex="-1"></a>    anomalies.</span>
<span id="cb45-697"><a href="#cb45-697" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-698"><a href="#cb45-698" aria-hidden="true" tabindex="-1"></a>We have found the distribution across samples to be similar to that over</span>
<span id="cb45-699"><a href="#cb45-699" aria-hidden="true" tabindex="-1"></a>all. This is good because it means that the samples are fairly</span>
<span id="cb45-700"><a href="#cb45-700" aria-hidden="true" tabindex="-1"></a>consistent with each other. We can now move on to the next step.</span>
<span id="cb45-701"><a href="#cb45-701" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-702"><a href="#cb45-702" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distribution of values across the genes</span></span>
<span id="cb45-703"><a href="#cb45-703" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-704"><a href="#cb45-704" aria-hidden="true" tabindex="-1"></a>There are lots of genes in this dataset therefore we will take a</span>
<span id="cb45-705"><a href="#cb45-705" aria-hidden="true" tabindex="-1"></a>slightly different approach. We would not want to use plot a</span>
<span id="cb45-706"><a href="#cb45-706" aria-hidden="true" tabindex="-1"></a>distribution for each gene in the same way. Will pivot the data to tidy</span>
<span id="cb45-707"><a href="#cb45-707" aria-hidden="true" tabindex="-1"></a>and then summarise the counts for each gene.</span>
<span id="cb45-708"><a href="#cb45-708" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-709"><a href="#cb45-709" aria-hidden="true" tabindex="-1"></a>üé¨ Summarise the counts for each gene and save the result as</span>
<span id="cb45-710"><a href="#cb45-710" aria-hidden="true" tabindex="-1"></a><span class="in">`pro_meta_summary_gene`</span>. Include the same columns as we had in the by</span>
<span id="cb45-711"><a href="#cb45-711" aria-hidden="true" tabindex="-1"></a>sample summary (<span class="in">`pro_meta_summary_samp`</span>) *and* an additional column,</span>
<span id="cb45-712"><a href="#cb45-712" aria-hidden="true" tabindex="-1"></a><span class="in">`total`</span> for the total number of counts for each gene.</span>
<span id="cb45-713"><a href="#cb45-713" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-716"><a href="#cb45-716" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-717"><a href="#cb45-717" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb45-718"><a href="#cb45-718" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb45-719"><a href="#cb45-719" aria-hidden="true" tabindex="-1"></a><span class="co"># you need to group by `gene_id` rather than `sample_id`</span></span>
<span id="cb45-720"><a href="#cb45-720" aria-hidden="true" tabindex="-1"></a><span class="co"># sum(count) will find the sum of counts</span></span>
<span id="cb45-721"><a href="#cb45-721" aria-hidden="true" tabindex="-1"></a>pro_meta_summary_gene <span class="ot">&lt;-</span> pro_meta <span class="sc">|&gt;</span></span>
<span id="cb45-722"><a href="#cb45-722" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>gene_id,</span>
<span id="cb45-723"><a href="#cb45-723" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"sample"</span>,</span>
<span id="cb45-724"><a href="#cb45-724" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-725"><a href="#cb45-725" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene_id) <span class="sc">|&gt;</span></span>
<span id="cb45-726"><a href="#cb45-726" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">min =</span> <span class="fu">min</span>(count),</span>
<span id="cb45-727"><a href="#cb45-727" aria-hidden="true" tabindex="-1"></a>            <span class="at">lowerq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.25</span>),</span>
<span id="cb45-728"><a href="#cb45-728" aria-hidden="true" tabindex="-1"></a>            <span class="at">sd =</span> <span class="fu">sd</span>(count),</span>
<span id="cb45-729"><a href="#cb45-729" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean =</span> <span class="fu">mean</span>(count),</span>
<span id="cb45-730"><a href="#cb45-730" aria-hidden="true" tabindex="-1"></a>            <span class="at">median =</span> <span class="fu">median</span>(count),</span>
<span id="cb45-731"><a href="#cb45-731" aria-hidden="true" tabindex="-1"></a>            <span class="at">upperq =</span> <span class="fu">quantile</span>(count, <span class="fl">0.75</span>),</span>
<span id="cb45-732"><a href="#cb45-732" aria-hidden="true" tabindex="-1"></a>            <span class="at">max =</span> <span class="fu">max</span>(count),</span>
<span id="cb45-733"><a href="#cb45-733" aria-hidden="true" tabindex="-1"></a>            <span class="at">total =</span> <span class="fu">sum</span>(count),</span>
<span id="cb45-734"><a href="#cb45-734" aria-hidden="true" tabindex="-1"></a>            <span class="at">n_above_zero =</span> <span class="fu">sum</span>(count <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-735"><a href="#cb45-735" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-736"><a href="#cb45-736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-737"><a href="#cb45-737" aria-hidden="true" tabindex="-1"></a>üé¨ View the <span class="in">`pro_meta_summary_gene`</span> dataframe.</span>
<span id="cb45-738"><a href="#cb45-738" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-739"><a href="#cb45-739" aria-hidden="true" tabindex="-1"></a>Notice that we have:</span>
<span id="cb45-740"><a href="#cb45-740" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-741"><a href="#cb45-741" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>a lot of genes with counts of zero in every sample</span>
<span id="cb45-742"><a href="#cb45-742" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>a lot of genes with zero counts in several of the samples</span>
<span id="cb45-743"><a href="#cb45-743" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>some very very low counts.</span>
<span id="cb45-744"><a href="#cb45-744" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-745"><a href="#cb45-745" aria-hidden="true" tabindex="-1"></a>Genes with very low counts should be filtered out because they are</span>
<span id="cb45-746"><a href="#cb45-746" aria-hidden="true" tabindex="-1"></a>unreliable - or, at the least, uninformative. The goal of our downstream</span>
<span id="cb45-747"><a href="#cb45-747" aria-hidden="true" tabindex="-1"></a>analysis will be to see if there is a significant difference in gene</span>
<span id="cb45-748"><a href="#cb45-748" aria-hidden="true" tabindex="-1"></a>expression between the stages. Since we have only three replicates in</span>
<span id="cb45-749"><a href="#cb45-749" aria-hidden="true" tabindex="-1"></a>each group, having one or two unreliable, missing or zero values, makes</span>
<span id="cb45-750"><a href="#cb45-750" aria-hidden="true" tabindex="-1"></a>such a determination impossible for a particular gene. We will use the</span>
<span id="cb45-751"><a href="#cb45-751" aria-hidden="true" tabindex="-1"></a>total counts (<span class="in">`total`</span>) and the number of samples with non-zero values</span>
<span id="cb45-752"><a href="#cb45-752" aria-hidden="true" tabindex="-1"></a>(<span class="in">`n_above_zero`</span>) in this dataframe to filter our genes later.</span>
<span id="cb45-753"><a href="#cb45-753" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-754"><a href="#cb45-754" aria-hidden="true" tabindex="-1"></a>As we have a lot of genes, it is helpful to plot the mean counts with</span>
<span id="cb45-755"><a href="#cb45-755" aria-hidden="true" tabindex="-1"></a><span class="in">`geom_pointrange()`</span> to get an overview of the distributions. We will</span>
<span id="cb45-756"><a href="#cb45-756" aria-hidden="true" tabindex="-1"></a>again plot the log of the mean counts. We will also order the genes from</span>
<span id="cb45-757"><a href="#cb45-757" aria-hidden="true" tabindex="-1"></a>lowest to highest mean count.</span>
<span id="cb45-758"><a href="#cb45-758" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-759"><a href="#cb45-759" aria-hidden="true" tabindex="-1"></a>üé¨ Plot the logged mean counts for each gene in order of size using</span>
<span id="cb45-760"><a href="#cb45-760" aria-hidden="true" tabindex="-1"></a><span class="in">`geom_pointrange()`</span>:</span>
<span id="cb45-761"><a href="#cb45-761" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-764"><a href="#cb45-764" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-765"><a href="#cb45-765" aria-hidden="true" tabindex="-1"></a>pro_meta_summary_gene <span class="sc">|&gt;</span> </span>
<span id="cb45-766"><a href="#cb45-766" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">reorder</span>(gene_id, mean), <span class="at">y =</span> <span class="fu">log10</span>(mean))) <span class="sc">+</span></span>
<span id="cb45-767"><a href="#cb45-767" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> <span class="fu">log10</span>(mean <span class="sc">-</span> sd), </span>
<span id="cb45-768"><a href="#cb45-768" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ymax =</span> <span class="fu">log10</span>(mean <span class="sc">+</span> sd )),</span>
<span id="cb45-769"><a href="#cb45-769" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="fl">0.1</span>)</span>
<span id="cb45-770"><a href="#cb45-770" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-771"><a href="#cb45-771" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-772"><a href="#cb45-772" aria-hidden="true" tabindex="-1"></a>(Note the warning is expected since we have zero means).</span>
<span id="cb45-773"><a href="#cb45-773" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-774"><a href="#cb45-774" aria-hidden="true" tabindex="-1"></a>You can see we also have quite a few genes with means less than 1 (log</span>
<span id="cb45-775"><a href="#cb45-775" aria-hidden="true" tabindex="-1"></a>below zero). Note that the variability between genes (average counts</span>
<span id="cb45-776"><a href="#cb45-776" aria-hidden="true" tabindex="-1"></a>between 0 and 349606) is far greater than between samples (average</span>
<span id="cb45-777"><a href="#cb45-777" aria-hidden="true" tabindex="-1"></a>counts from 220 to 364) which is exactly what we would expect to see.</span>
<span id="cb45-778"><a href="#cb45-778" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-779"><a href="#cb45-779" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Filtering for QC</span><span class="co">](#leishmania-filterqc)</span>.</span>
<span id="cb45-780"><a href="#cb45-780" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-781"><a href="#cb45-781" aria-hidden="true" tabindex="-1"></a><span class="fu">## üê≠ Stem cells</span></span>
<span id="cb45-782"><a href="#cb45-782" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-783"><a href="#cb45-783" aria-hidden="true" tabindex="-1"></a><span class="fu">### Import</span></span>
<span id="cb45-784"><a href="#cb45-784" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-785"><a href="#cb45-785" aria-hidden="true" tabindex="-1"></a>Import the data for the HSPC and the Progenitor cells.</span>
<span id="cb45-786"><a href="#cb45-786" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-787"><a href="#cb45-787" aria-hidden="true" tabindex="-1"></a>::: {.content-visible when-profile="surfaceome"}</span>
<span id="cb45-788"><a href="#cb45-788" aria-hidden="true" tabindex="-1"></a>üé¨ Import <span class="co">[</span><span class="ot">surfaceome_hspc.csv</span><span class="co">](data-raw/surfaceome_hspc.csv)</span> and</span>
<span id="cb45-789"><a href="#cb45-789" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">surfaceome_prog.csv</span><span class="co">](data-raw/surfaceome_prog.csv)</span></span>
<span id="cb45-790"><a href="#cb45-790" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb45-791"><a href="#cb45-791" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-792"><a href="#cb45-792" aria-hidden="true" tabindex="-1"></a>::: {.content-visible when-profile="secretome"}</span>
<span id="cb45-793"><a href="#cb45-793" aria-hidden="true" tabindex="-1"></a>üé¨ Import <span class="co">[</span><span class="ot">secretome_hspc.csv</span><span class="co">](data-raw/secretome_hspc.csv)</span> and</span>
<span id="cb45-794"><a href="#cb45-794" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">secretome_prog.csv</span><span class="co">](data-raw/secretome_prog.csv)</span></span>
<span id="cb45-795"><a href="#cb45-795" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb45-796"><a href="#cb45-796" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-799"><a href="#cb45-799" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-800"><a href="#cb45-800" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-secretome</span></span>
<span id="cb45-801"><a href="#cb45-801" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: !expr params$mouse == "secretome"</span></span>
<span id="cb45-802"><a href="#cb45-802" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr params$mouse == "secretome"</span></span>
<span id="cb45-803"><a href="#cb45-803" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache: true</span></span>
<span id="cb45-804"><a href="#cb45-804" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache.extra: secretome</span></span>
<span id="cb45-805"><a href="#cb45-805" aria-hidden="true" tabindex="-1"></a><span class="co"># üê≠ import the hspc and prog data</span></span>
<span id="cb45-806"><a href="#cb45-806" aria-hidden="true" tabindex="-1"></a>hspc  <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/secretome_hspc.csv"</span>)</span>
<span id="cb45-807"><a href="#cb45-807" aria-hidden="true" tabindex="-1"></a>prog  <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/secretome_prog.csv"</span>)</span>
<span id="cb45-808"><a href="#cb45-808" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-809"><a href="#cb45-809" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-810"><a href="#cb45-810" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-813"><a href="#cb45-813" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-814"><a href="#cb45-814" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-surfaceome</span></span>
<span id="cb45-815"><a href="#cb45-815" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: !expr params$mouse == "surfaceome"</span></span>
<span id="cb45-816"><a href="#cb45-816" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr params$mouse == "surfaceome"</span></span>
<span id="cb45-817"><a href="#cb45-817" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache: true</span></span>
<span id="cb45-818"><a href="#cb45-818" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache.extra: surfaceome</span></span>
<span id="cb45-819"><a href="#cb45-819" aria-hidden="true" tabindex="-1"></a><span class="co"># üê≠ import the hspc and prog data</span></span>
<span id="cb45-820"><a href="#cb45-820" aria-hidden="true" tabindex="-1"></a>hspc  <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/surfaceome_hspc.csv"</span>)</span>
<span id="cb45-821"><a href="#cb45-821" aria-hidden="true" tabindex="-1"></a>prog  <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/surfaceome_prog.csv"</span>)</span>
<span id="cb45-822"><a href="#cb45-822" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-823"><a href="#cb45-823" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-824"><a href="#cb45-824" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-825"><a href="#cb45-825" aria-hidden="true" tabindex="-1"></a>We will need to combine the two sets of columns (datasets) so we can</span>
<span id="cb45-826"><a href="#cb45-826" aria-hidden="true" tabindex="-1"></a>compare the two stages. We will join them using <span class="in">`ensembl_gene_id`</span> to</span>
<span id="cb45-827"><a href="#cb45-827" aria-hidden="true" tabindex="-1"></a>match the rows. The column names differ so we don't need to worry about</span>
<span id="cb45-828"><a href="#cb45-828" aria-hidden="true" tabindex="-1"></a>renaming any of them.</span>
<span id="cb45-829"><a href="#cb45-829" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-830"><a href="#cb45-830" aria-hidden="true" tabindex="-1"></a>üé¨ Combine the two datasets by <span class="in">`ensembl_gene_id`</span> and save the result as</span>
<span id="cb45-831"><a href="#cb45-831" aria-hidden="true" tabindex="-1"></a><span class="in">`hspc_prog`</span>.</span>
<span id="cb45-832"><a href="#cb45-832" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-835"><a href="#cb45-835" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-836"><a href="#cb45-836" aria-hidden="true" tabindex="-1"></a><span class="co">#  combine the two datasets</span></span>
<span id="cb45-837"><a href="#cb45-837" aria-hidden="true" tabindex="-1"></a>hspc_prog <span class="ot">&lt;-</span> hspc <span class="sc">|&gt;</span></span>
<span id="cb45-838"><a href="#cb45-838" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(prog, </span>
<span id="cb45-839"><a href="#cb45-839" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"ensembl_gene_id"</span>)</span>
<span id="cb45-840"><a href="#cb45-840" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-841"><a href="#cb45-841" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-842"><a href="#cb45-842" aria-hidden="true" tabindex="-1"></a>üé¨ Check the dataframe has the number of rows and columns you were</span>
<span id="cb45-843"><a href="#cb45-843" aria-hidden="true" tabindex="-1"></a>expecting and that column types and names are as expected.</span>
<span id="cb45-844"><a href="#cb45-844" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-845"><a href="#cb45-845" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distribution of values across all the data in the file</span></span>
<span id="cb45-846"><a href="#cb45-846" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-847"><a href="#cb45-847" aria-hidden="true" tabindex="-1"></a>The values are spread over multiple columns so in order to plot the</span>
<span id="cb45-848"><a href="#cb45-848" aria-hidden="true" tabindex="-1"></a>distribution as a whole, we will need to first use <span class="in">`pivot_longer()`</span> to</span>
<span id="cb45-849"><a href="#cb45-849" aria-hidden="true" tabindex="-1"></a>put the data in ['tidy'</span>
<span id="cb45-850"><a href="#cb45-850" aria-hidden="true" tabindex="-1"></a>format](https://3mmarand.github.io/R4BABS/r4babs1/week-9/workshop.html#tidying-data)</span>
<span id="cb45-851"><a href="#cb45-851" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@wickham2014</span><span class="co">]</span> by stacking the columns. We *could* save a copy of the</span>
<span id="cb45-852"><a href="#cb45-852" aria-hidden="true" tabindex="-1"></a>stacked data and then plot it, but here, I have just piped the stacked</span>
<span id="cb45-853"><a href="#cb45-853" aria-hidden="true" tabindex="-1"></a>data straight into <span class="in">`ggplot()`</span>. This helps me avoid cluttering my R</span>
<span id="cb45-854"><a href="#cb45-854" aria-hidden="true" tabindex="-1"></a>environment with temporary objects.</span>
<span id="cb45-855"><a href="#cb45-855" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-856"><a href="#cb45-856" aria-hidden="true" tabindex="-1"></a>üé¨ Pivot the counts (stack the columns) so all the counts are in a</span>
<span id="cb45-857"><a href="#cb45-857" aria-hidden="true" tabindex="-1"></a>single column (<span class="in">`expr`</span>) labelled in <span class="in">`cell`</span> by the column it came from and</span>
<span id="cb45-858"><a href="#cb45-858" aria-hidden="true" tabindex="-1"></a>pipe into <span class="in">`ggplot()`</span> to create a histogram:</span>
<span id="cb45-859"><a href="#cb45-859" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-862"><a href="#cb45-862" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-863"><a href="#cb45-863" aria-hidden="true" tabindex="-1"></a>hspc_prog <span class="sc">|&gt;</span></span>
<span id="cb45-864"><a href="#cb45-864" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>ensembl_gene_id,</span>
<span id="cb45-865"><a href="#cb45-865" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"cell"</span>,</span>
<span id="cb45-866"><a href="#cb45-866" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"expr"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb45-867"><a href="#cb45-867" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> expr)) <span class="sc">+</span></span>
<span id="cb45-868"><a href="#cb45-868" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span>
<span id="cb45-869"><a href="#cb45-869" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-870"><a href="#cb45-870" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-871"><a href="#cb45-871" aria-hidden="true" tabindex="-1"></a>This is a very striking distribution. Is it what we are expecting?</span>
<span id="cb45-872"><a href="#cb45-872" aria-hidden="true" tabindex="-1"></a>Notice we have a peak at zero indicating there are low values zeros.</span>
<span id="cb45-873"><a href="#cb45-873" aria-hidden="true" tabindex="-1"></a>This inflation of low values suggests some are anomalous - they will</span>
<span id="cb45-874"><a href="#cb45-874" aria-hidden="true" tabindex="-1"></a>have been derived from low counts which are likely false positives. As</span>
<span id="cb45-875"><a href="#cb45-875" aria-hidden="true" tabindex="-1"></a>inaccurate measures, we will want to exclude expression values below</span>
<span id="cb45-876"><a href="#cb45-876" aria-hidden="true" tabindex="-1"></a>(about) 1. We will revisit this after we have considered the</span>
<span id="cb45-877"><a href="#cb45-877" aria-hidden="true" tabindex="-1"></a>distribution of expression across cells and genes.</span>
<span id="cb45-878"><a href="#cb45-878" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-879"><a href="#cb45-879" aria-hidden="true" tabindex="-1"></a>What about the bimodal appearance of the the 'real' values? If we had</span>
<span id="cb45-880"><a href="#cb45-880" aria-hidden="true" tabindex="-1"></a>the whole transcriptome we would *not* expect to see such a pattern -</span>
<span id="cb45-881"><a href="#cb45-881" aria-hidden="true" tabindex="-1"></a>we'd expect to see a roughly normal distribution<span class="ot">[^3]</span>. However, this is a</span>
<span id="cb45-882"><a href="#cb45-882" aria-hidden="true" tabindex="-1"></a>*subset* of the genome and the nature of the subsetting has had an</span>
<span id="cb45-883"><a href="#cb45-883" aria-hidden="true" tabindex="-1"></a>influence here. These are a subset of cell surface proteins that show a</span>
<span id="cb45-884"><a href="#cb45-884" aria-hidden="true" tabindex="-1"></a>significant difference between at least two of twelve cell subtypes.</span>
<span id="cb45-885"><a href="#cb45-885" aria-hidden="true" tabindex="-1"></a>That is, all of these genes are either "high" or "low" leading to a</span>
<span id="cb45-886"><a href="#cb45-886" aria-hidden="true" tabindex="-1"></a>bimodal distribution.</span>
<span id="cb45-887"><a href="#cb45-887" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-888"><a href="#cb45-888" aria-hidden="true" tabindex="-1"></a><span class="ot">[^3]: </span>This a result of the [Central limit</span>
<span id="cb45-889"><a href="#cb45-889" aria-hidden="true" tabindex="-1"></a>    theorem](https://en.wikipedia.org/wiki/Central_limit_theorem),one</span>
<span id="cb45-890"><a href="#cb45-890" aria-hidden="true" tabindex="-1"></a>    consequence of which is that adding together lots of distributions -</span>
<span id="cb45-891"><a href="#cb45-891" aria-hidden="true" tabindex="-1"></a>    whatever distributions they are - will tend to a normal</span>
<span id="cb45-892"><a href="#cb45-892" aria-hidden="true" tabindex="-1"></a>    distribution.</span>
<span id="cb45-893"><a href="#cb45-893" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-894"><a href="#cb45-894" aria-hidden="true" tabindex="-1"></a>Unlike the other three datasets, which count raw counts, these data are</span>
<span id="cb45-895"><a href="#cb45-895" aria-hidden="true" tabindex="-1"></a>normalised and log~2~ transformed. We do not need to plot the log of the</span>
<span id="cb45-896"><a href="#cb45-896" aria-hidden="true" tabindex="-1"></a>values to see the distribution - they are already logged.</span>
<span id="cb45-897"><a href="#cb45-897" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-898"><a href="#cb45-898" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distribution of values across the samples</span></span>
<span id="cb45-899"><a href="#cb45-899" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-900"><a href="#cb45-900" aria-hidden="true" tabindex="-1"></a>For the other three datasets, we used the <span class="in">`summary()`</span> function to get an</span>
<span id="cb45-901"><a href="#cb45-901" aria-hidden="true" tabindex="-1"></a>overview of the columns. This works well when you have upto about 25</span>
<span id="cb45-902"><a href="#cb45-902" aria-hidden="true" tabindex="-1"></a>columns but it is not helpful here because we have a lot of cells! Feel</span>
<span id="cb45-903"><a href="#cb45-903" aria-hidden="true" tabindex="-1"></a>free to try it!</span>
<span id="cb45-904"><a href="#cb45-904" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-905"><a href="#cb45-905" aria-hidden="true" tabindex="-1"></a>In this data set, there is even more of an advantage of using the</span>
<span id="cb45-906"><a href="#cb45-906" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_longer()`</span>, <span class="in">`group_by()`</span> and <span class="in">`summarise()`</span> approach. We will be</span>
<span id="cb45-907"><a href="#cb45-907" aria-hidden="true" tabindex="-1"></a>able to open the dataframe in the Viewer and make plots to examine</span>
<span id="cb45-908"><a href="#cb45-908" aria-hidden="true" tabindex="-1"></a>whether the distributions are similar across cells. The mean and the</span>
<span id="cb45-909"><a href="#cb45-909" aria-hidden="true" tabindex="-1"></a>standard deviation are useful to see the distributions across cells in a</span>
<span id="cb45-910"><a href="#cb45-910" aria-hidden="true" tabindex="-1"></a>plot but we will also examine the interquartile values, maximums and the</span>
<span id="cb45-911"><a href="#cb45-911" aria-hidden="true" tabindex="-1"></a>number of non-zero values.</span>
<span id="cb45-912"><a href="#cb45-912" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-913"><a href="#cb45-913" aria-hidden="true" tabindex="-1"></a>üé¨ Summarise all the cells:</span>
<span id="cb45-914"><a href="#cb45-914" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-917"><a href="#cb45-917" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-918"><a href="#cb45-918" aria-hidden="true" tabindex="-1"></a>hspc_prog_summary_cell <span class="ot">&lt;-</span> hspc_prog <span class="sc">|&gt;</span></span>
<span id="cb45-919"><a href="#cb45-919" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>ensembl_gene_id,</span>
<span id="cb45-920"><a href="#cb45-920" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"cell"</span>,</span>
<span id="cb45-921"><a href="#cb45-921" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"expr"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-922"><a href="#cb45-922" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cell) <span class="sc">|&gt;</span></span>
<span id="cb45-923"><a href="#cb45-923" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">min =</span> <span class="fu">min</span>(expr),</span>
<span id="cb45-924"><a href="#cb45-924" aria-hidden="true" tabindex="-1"></a>            <span class="at">lowerq =</span> <span class="fu">quantile</span>(expr, <span class="fl">0.25</span>),</span>
<span id="cb45-925"><a href="#cb45-925" aria-hidden="true" tabindex="-1"></a>            <span class="at">sd =</span> <span class="fu">sd</span>(expr),</span>
<span id="cb45-926"><a href="#cb45-926" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean =</span> <span class="fu">mean</span>(expr),</span>
<span id="cb45-927"><a href="#cb45-927" aria-hidden="true" tabindex="-1"></a>            <span class="at">median =</span> <span class="fu">median</span>(expr),</span>
<span id="cb45-928"><a href="#cb45-928" aria-hidden="true" tabindex="-1"></a>            <span class="at">upperq =</span> <span class="fu">quantile</span>(expr, <span class="fl">0.75</span>),</span>
<span id="cb45-929"><a href="#cb45-929" aria-hidden="true" tabindex="-1"></a>            <span class="at">max =</span> <span class="fu">max</span>(expr),</span>
<span id="cb45-930"><a href="#cb45-930" aria-hidden="true" tabindex="-1"></a>            <span class="at">total =</span> <span class="fu">sum</span>(expr),</span>
<span id="cb45-931"><a href="#cb45-931" aria-hidden="true" tabindex="-1"></a>            <span class="at">n_above_zero =</span> <span class="fu">sum</span>(expr <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-932"><a href="#cb45-932" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-933"><a href="#cb45-933" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-934"><a href="#cb45-934" aria-hidden="true" tabindex="-1"></a>üé¨ View the <span class="in">`hspc_prog_summary_cell`</span> dataframe (click on it in the</span>
<span id="cb45-935"><a href="#cb45-935" aria-hidden="true" tabindex="-1"></a>environment).</span>
<span id="cb45-936"><a href="#cb45-936" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-937"><a href="#cb45-937" aria-hidden="true" tabindex="-1"></a>Notice that: - a minimum value of 0 appears in all 1499 cells - the</span>
<span id="cb45-938"><a href="#cb45-938" aria-hidden="true" tabindex="-1"></a>lower quartiles are all zero and so are many of the medians - there are</span>
<span id="cb45-939"><a href="#cb45-939" aria-hidden="true" tabindex="-1"></a>no cells with above 0 expression in all <span class="in">`r nrow(hspc)`</span> of the gene</span>
<span id="cb45-940"><a href="#cb45-940" aria-hidden="true" tabindex="-1"></a>subset - the highest number of genes expressed is</span>
<span id="cb45-941"><a href="#cb45-941" aria-hidden="true" tabindex="-1"></a><span class="in">`r max(hspc_prog_summary_cell$n_above_zero)`</span>, the lowest is</span>
<span id="cb45-942"><a href="#cb45-942" aria-hidden="true" tabindex="-1"></a><span class="in">`r min(hspc_prog_summary_cell$n_above_zero)`</span></span>
<span id="cb45-943"><a href="#cb45-943" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-944"><a href="#cb45-944" aria-hidden="true" tabindex="-1"></a>In short, there are quite a lot of zeros.</span>
<span id="cb45-945"><a href="#cb45-945" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-946"><a href="#cb45-946" aria-hidden="true" tabindex="-1"></a>To get a better understanding of the distribution of expressions in</span>
<span id="cb45-947"><a href="#cb45-947" aria-hidden="true" tabindex="-1"></a>cells we can create a ggplot using the pointrange geom. Pointrange puts</span>
<span id="cb45-948"><a href="#cb45-948" aria-hidden="true" tabindex="-1"></a>a dot at the mean and a line between a minimum and a maximum such as +/-</span>
<span id="cb45-949"><a href="#cb45-949" aria-hidden="true" tabindex="-1"></a>one standard deviation. Not unlike a boxplot, but when you need the</span>
<span id="cb45-950"><a href="#cb45-950" aria-hidden="true" tabindex="-1"></a>boxes too be very narrow!</span>
<span id="cb45-951"><a href="#cb45-951" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-952"><a href="#cb45-952" aria-hidden="true" tabindex="-1"></a>üé¨ Create a pointrange plot.</span>
<span id="cb45-953"><a href="#cb45-953" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-956"><a href="#cb45-956" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-957"><a href="#cb45-957" aria-hidden="true" tabindex="-1"></a>hspc_prog_summary_cell <span class="sc">|&gt;</span> </span>
<span id="cb45-958"><a href="#cb45-958" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> cell, <span class="at">y =</span> mean)) <span class="sc">+</span></span>
<span id="cb45-959"><a href="#cb45-959" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> mean <span class="sc">-</span> sd, </span>
<span id="cb45-960"><a href="#cb45-960" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ymax =</span> mean <span class="sc">+</span> sd ),</span>
<span id="cb45-961"><a href="#cb45-961" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="fl">0.1</span>)</span>
<span id="cb45-962"><a href="#cb45-962" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-963"><a href="#cb45-963" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-964"><a href="#cb45-964" aria-hidden="true" tabindex="-1"></a>You will need to use the Zoom button to pop the plot window out so you</span>
<span id="cb45-965"><a href="#cb45-965" aria-hidden="true" tabindex="-1"></a>can make it as wide as possible</span>
<span id="cb45-966"><a href="#cb45-966" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-967"><a href="#cb45-967" aria-hidden="true" tabindex="-1"></a>The things to notice are:</span>
<span id="cb45-968"><a href="#cb45-968" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-969"><a href="#cb45-969" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the average expression in cells is similar for all cells. This is</span>
<span id="cb45-970"><a href="#cb45-970" aria-hidden="true" tabindex="-1"></a>    good to know - if some cells had much lower expression perhaps there</span>
<span id="cb45-971"><a href="#cb45-971" aria-hidden="true" tabindex="-1"></a>    is something wrong with them, or their sequencing, and they should</span>
<span id="cb45-972"><a href="#cb45-972" aria-hidden="true" tabindex="-1"></a>    be excluded.</span>
<span id="cb45-973"><a href="#cb45-973" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the distributions are roughly similar in width too</span>
<span id="cb45-974"><a href="#cb45-974" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-975"><a href="#cb45-975" aria-hidden="true" tabindex="-1"></a>The default order of <span class="in">`cell`</span> is alphabetical. It can be easier to judge</span>
<span id="cb45-976"><a href="#cb45-976" aria-hidden="true" tabindex="-1"></a>if there are unusual cells if we order the lines by the size of the</span>
<span id="cb45-977"><a href="#cb45-977" aria-hidden="true" tabindex="-1"></a>mean.</span>
<span id="cb45-978"><a href="#cb45-978" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-979"><a href="#cb45-979" aria-hidden="true" tabindex="-1"></a>üé¨ Order a pointrange plot with <span class="in">`reorder(variable_to_order, order_by)`</span>.</span>
<span id="cb45-980"><a href="#cb45-980" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-983"><a href="#cb45-983" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-984"><a href="#cb45-984" aria-hidden="true" tabindex="-1"></a>hspc_prog_summary_cell <span class="sc">|&gt;</span> </span>
<span id="cb45-985"><a href="#cb45-985" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">reorder</span>(cell, mean), <span class="at">y =</span> mean)) <span class="sc">+</span></span>
<span id="cb45-986"><a href="#cb45-986" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> mean <span class="sc">-</span> sd, </span>
<span id="cb45-987"><a href="#cb45-987" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ymax =</span> mean <span class="sc">+</span> sd ),</span>
<span id="cb45-988"><a href="#cb45-988" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="fl">0.1</span>)</span>
<span id="cb45-989"><a href="#cb45-989" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-990"><a href="#cb45-990" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-991"><a href="#cb45-991" aria-hidden="true" tabindex="-1"></a><span class="in">`reorder()`</span> arranges <span class="in">`cell`</span> in increasing size of <span class="in">`mean`</span></span>
<span id="cb45-992"><a href="#cb45-992" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-993"><a href="#cb45-993" aria-hidden="true" tabindex="-1"></a>As we thought, the distributions are similar across cells - there are</span>
<span id="cb45-994"><a href="#cb45-994" aria-hidden="true" tabindex="-1"></a>not any cells that are obviously different from the others (only</span>
<span id="cb45-995"><a href="#cb45-995" aria-hidden="true" tabindex="-1"></a>incrementally).</span>
<span id="cb45-996"><a href="#cb45-996" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-997"><a href="#cb45-997" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distribution of values across the genes</span></span>
<span id="cb45-998"><a href="#cb45-998" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-999"><a href="#cb45-999" aria-hidden="true" tabindex="-1"></a>We will use the same approach to summarise the genes.</span>
<span id="cb45-1000"><a href="#cb45-1000" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1001"><a href="#cb45-1001" aria-hidden="true" tabindex="-1"></a>üé¨ Summarise the expression for each gene and save the result as</span>
<span id="cb45-1002"><a href="#cb45-1002" aria-hidden="true" tabindex="-1"></a><span class="in">`hspc_prog_summary_gene`</span>. Include the same columns as we had in the by</span>
<span id="cb45-1003"><a href="#cb45-1003" aria-hidden="true" tabindex="-1"></a>cell summary (<span class="in">`hspc_prog_summary_cell`</span>) and an additional column,</span>
<span id="cb45-1004"><a href="#cb45-1004" aria-hidden="true" tabindex="-1"></a><span class="in">`total`</span> for the total expression for each gene.</span>
<span id="cb45-1005"><a href="#cb45-1005" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1008"><a href="#cb45-1008" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1009"><a href="#cb45-1009" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb45-1010"><a href="#cb45-1010" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb45-1011"><a href="#cb45-1011" aria-hidden="true" tabindex="-1"></a><span class="co"># you need to group by `ensembl_gene_id` rather than `cell`</span></span>
<span id="cb45-1012"><a href="#cb45-1012" aria-hidden="true" tabindex="-1"></a><span class="co"># sum(expr) will find the total expression in a gene</span></span>
<span id="cb45-1013"><a href="#cb45-1013" aria-hidden="true" tabindex="-1"></a>hspc_prog_summary_gene <span class="ot">&lt;-</span> hspc_prog <span class="sc">|&gt;</span></span>
<span id="cb45-1014"><a href="#cb45-1014" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>ensembl_gene_id,</span>
<span id="cb45-1015"><a href="#cb45-1015" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"cell"</span>,</span>
<span id="cb45-1016"><a href="#cb45-1016" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"expr"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-1017"><a href="#cb45-1017" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(ensembl_gene_id) <span class="sc">|&gt;</span></span>
<span id="cb45-1018"><a href="#cb45-1018" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">min =</span> <span class="fu">min</span>(expr),</span>
<span id="cb45-1019"><a href="#cb45-1019" aria-hidden="true" tabindex="-1"></a>            <span class="at">lowerq =</span> <span class="fu">quantile</span>(expr, <span class="fl">0.25</span>),</span>
<span id="cb45-1020"><a href="#cb45-1020" aria-hidden="true" tabindex="-1"></a>            <span class="at">sd =</span> <span class="fu">sd</span>(expr),</span>
<span id="cb45-1021"><a href="#cb45-1021" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean =</span> <span class="fu">mean</span>(expr),</span>
<span id="cb45-1022"><a href="#cb45-1022" aria-hidden="true" tabindex="-1"></a>            <span class="at">median =</span> <span class="fu">median</span>(expr),</span>
<span id="cb45-1023"><a href="#cb45-1023" aria-hidden="true" tabindex="-1"></a>            <span class="at">upperq =</span> <span class="fu">quantile</span>(expr, <span class="fl">0.75</span>),</span>
<span id="cb45-1024"><a href="#cb45-1024" aria-hidden="true" tabindex="-1"></a>            <span class="at">max =</span> <span class="fu">max</span>(expr),</span>
<span id="cb45-1025"><a href="#cb45-1025" aria-hidden="true" tabindex="-1"></a>            <span class="at">total =</span> <span class="fu">sum</span>(expr),</span>
<span id="cb45-1026"><a href="#cb45-1026" aria-hidden="true" tabindex="-1"></a>            <span class="at">n_above_zero =</span> <span class="fu">sum</span>(expr <span class="sc">&gt;</span> <span class="dv">0</span>))</span>
<span id="cb45-1027"><a href="#cb45-1027" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1028"><a href="#cb45-1028" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1029"><a href="#cb45-1029" aria-hidden="true" tabindex="-1"></a>üé¨ View the <span class="in">`hspc_prog_summary_gene`</span> dataframe. Remember these are</span>
<span id="cb45-1030"><a href="#cb45-1030" aria-hidden="true" tabindex="-1"></a>normalised and logged (base 2) so we should not see very large values.</span>
<span id="cb45-1031"><a href="#cb45-1031" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1032"><a href="#cb45-1032" aria-hidden="true" tabindex="-1"></a>Notice that:</span>
<span id="cb45-1033"><a href="#cb45-1033" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1034"><a href="#cb45-1034" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>some genes are expressed in every cell, and many are expressed in</span>
<span id="cb45-1035"><a href="#cb45-1035" aria-hidden="true" tabindex="-1"></a>    most cells</span>
<span id="cb45-1036"><a href="#cb45-1036" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>quite a few genes are zero in many cells. This this matters less</span>
<span id="cb45-1037"><a href="#cb45-1037" aria-hidden="true" tabindex="-1"></a>    when we have many cells (samples) than when we have few samples.</span>
<span id="cb45-1038"><a href="#cb45-1038" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>no genes have zeros in every cell - the lowest number of cells is is</span>
<span id="cb45-1039"><a href="#cb45-1039" aria-hidden="true" tabindex="-1"></a>    <span class="in">`r min(hspc_prog_summary_gene$n_above_zero)`</span></span>
<span id="cb45-1040"><a href="#cb45-1040" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1041"><a href="#cb45-1041" aria-hidden="true" tabindex="-1"></a>It is again helpful to plot the ordered mean expression with pointrange</span>
<span id="cb45-1042"><a href="#cb45-1042" aria-hidden="true" tabindex="-1"></a>to get an overview.</span>
<span id="cb45-1043"><a href="#cb45-1043" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1044"><a href="#cb45-1044" aria-hidden="true" tabindex="-1"></a>üé¨ Plot the logged mean counts for each gene in order of size using</span>
<span id="cb45-1045"><a href="#cb45-1045" aria-hidden="true" tabindex="-1"></a><span class="in">`geom_pointrange()`</span>:</span>
<span id="cb45-1046"><a href="#cb45-1046" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1049"><a href="#cb45-1049" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1050"><a href="#cb45-1050" aria-hidden="true" tabindex="-1"></a>hspc_prog_summary_gene <span class="sc">|&gt;</span> </span>
<span id="cb45-1051"><a href="#cb45-1051" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">reorder</span>(ensembl_gene_id, mean), <span class="at">y =</span> mean)) <span class="sc">+</span></span>
<span id="cb45-1052"><a href="#cb45-1052" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> mean <span class="sc">-</span> sd, </span>
<span id="cb45-1053"><a href="#cb45-1053" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ymax =</span> mean <span class="sc">+</span> sd),</span>
<span id="cb45-1054"><a href="#cb45-1054" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="fl">0.1</span>)</span>
<span id="cb45-1055"><a href="#cb45-1055" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1056"><a href="#cb45-1056" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1057"><a href="#cb45-1057" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1058"><a href="#cb45-1058" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1059"><a href="#cb45-1059" aria-hidden="true" tabindex="-1"></a>Note that the variability between genes (average expression between</span>
<span id="cb45-1060"><a href="#cb45-1060" aria-hidden="true" tabindex="-1"></a><span class="in">`r min(hspc_prog_summary_gene$mean) |&gt; round(4)`</span> and</span>
<span id="cb45-1061"><a href="#cb45-1061" aria-hidden="true" tabindex="-1"></a><span class="in">`r max(hspc_prog_summary_gene$mean) |&gt; round(4)`</span>) is greater than</span>
<span id="cb45-1062"><a href="#cb45-1062" aria-hidden="true" tabindex="-1"></a>between cells (average expression between</span>
<span id="cb45-1063"><a href="#cb45-1063" aria-hidden="true" tabindex="-1"></a><span class="in">`r min(hspc_prog_summary_cell$mean) |&gt; round(4)`</span> and</span>
<span id="cb45-1064"><a href="#cb45-1064" aria-hidden="true" tabindex="-1"></a><span class="in">`r max(hspc_prog_summary_cell$mean) |&gt; round(4)`</span>) which is just what we</span>
<span id="cb45-1065"><a href="#cb45-1065" aria-hidden="true" tabindex="-1"></a>would expect.</span>
<span id="cb45-1066"><a href="#cb45-1066" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1067"><a href="#cb45-1067" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Filtering for QC</span><span class="co">](#stem-cells-filterqc)</span>.</span>
<span id="cb45-1068"><a href="#cb45-1068" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1069"><a href="#cb45-1069" aria-hidden="true" tabindex="-1"></a><span class="fu"># Filtering for QC</span></span>
<span id="cb45-1070"><a href="#cb45-1070" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1071"><a href="#cb45-1071" aria-hidden="true" tabindex="-1"></a><span class="fu">## üéÑ *Arabidopsis* {#arabidopsis-filterqc}</span></span>
<span id="cb45-1072"><a href="#cb45-1072" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1073"><a href="#cb45-1073" aria-hidden="true" tabindex="-1"></a>Our samples look to be similarly well sequenced although this is</span>
<span id="cb45-1074"><a href="#cb45-1074" aria-hidden="true" tabindex="-1"></a>difficult to determine with only two replicates. However, some genes are</span>
<span id="cb45-1075"><a href="#cb45-1075" aria-hidden="true" tabindex="-1"></a>not expressed or the expression values are so low in for a gene that</span>
<span id="cb45-1076"><a href="#cb45-1076" aria-hidden="true" tabindex="-1"></a>they are uninformative. We will filter the <span class="in">`root_summary_gene`</span> dataframe</span>
<span id="cb45-1077"><a href="#cb45-1077" aria-hidden="true" tabindex="-1"></a>to obtain a list of <span class="in">`gene_id`</span> we can use to filter <span class="in">`root`</span>.</span>
<span id="cb45-1078"><a href="#cb45-1078" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1079"><a href="#cb45-1079" aria-hidden="true" tabindex="-1"></a>My suggestion is to include only the genes with counts in at least 6</span>
<span id="cb45-1080"><a href="#cb45-1080" aria-hidden="true" tabindex="-1"></a>samples, and those with total counts above 20. I chose 6 because that</span>
<span id="cb45-1081"><a href="#cb45-1081" aria-hidden="true" tabindex="-1"></a>would keep genes expressed only in one treatment. I chose 20 based on</span>
<span id="cb45-1082"><a href="#cb45-1082" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1083"><a href="#cb45-1083" aria-hidden="true" tabindex="-1"></a>üé¨ Filter the summary by gene dataframe:</span>
<span id="cb45-1084"><a href="#cb45-1084" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1087"><a href="#cb45-1087" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1088"><a href="#cb45-1088" aria-hidden="true" tabindex="-1"></a>root_summary_gene_filtered <span class="ot">&lt;-</span> root_summary_gene <span class="sc">|&gt;</span> </span>
<span id="cb45-1089"><a href="#cb45-1089" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(total <span class="sc">&gt;</span> <span class="dv">20</span>) <span class="sc">|&gt;</span> </span>
<span id="cb45-1090"><a href="#cb45-1090" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(n_above_zero <span class="sc">&gt;=</span> <span class="dv">6</span>)</span>
<span id="cb45-1091"><a href="#cb45-1091" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1092"><a href="#cb45-1092" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1093"><a href="#cb45-1093" aria-hidden="true" tabindex="-1"></a>‚ùì How many genes do you have left</span>
<span id="cb45-1094"><a href="#cb45-1094" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1095"><a href="#cb45-1095" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb45-1096"><a href="#cb45-1096" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1097"><a href="#cb45-1097" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- there are 22361 genes left. 9472 have total counts less than 20 and or --&gt;</span></span>
<span id="cb45-1098"><a href="#cb45-1098" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1099"><a href="#cb45-1099" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- have counts in fewer the 6 samples. --&gt;</span></span>
<span id="cb45-1100"><a href="#cb45-1100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1101"><a href="#cb45-1101" aria-hidden="true" tabindex="-1"></a>üé¨ Use the list of <span class="in">`gene_id`</span> in the filtered summary to filter the</span>
<span id="cb45-1102"><a href="#cb45-1102" aria-hidden="true" tabindex="-1"></a>original dataset:</span>
<span id="cb45-1103"><a href="#cb45-1103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1106"><a href="#cb45-1106" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1107"><a href="#cb45-1107" aria-hidden="true" tabindex="-1"></a>root_filtered <span class="ot">&lt;-</span> root <span class="sc">|&gt;</span> </span>
<span id="cb45-1108"><a href="#cb45-1108" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene_id <span class="sc">%in%</span>  root_summary_gene_filtered<span class="sc">$</span>gene_id)</span>
<span id="cb45-1109"><a href="#cb45-1109" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1110"><a href="#cb45-1110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1111"><a href="#cb45-1111" aria-hidden="true" tabindex="-1"></a>üé¨ Write the filtered data to file:</span>
<span id="cb45-1112"><a href="#cb45-1112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1115"><a href="#cb45-1115" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1116"><a href="#cb45-1116" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(root_filtered, </span>
<span id="cb45-1117"><a href="#cb45-1117" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"data-processed/root_filtered.csv"</span>)</span>
<span id="cb45-1118"><a href="#cb45-1118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1119"><a href="#cb45-1119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1122"><a href="#cb45-1122" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1123"><a href="#cb45-1123" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb45-1124"><a href="#cb45-1124" aria-hidden="true" tabindex="-1"></a><span class="co">#---UNDER THE HOOD CODE---</span></span>
<span id="cb45-1125"><a href="#cb45-1125" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(root_filtered, </span>
<span id="cb45-1126"><a href="#cb45-1126" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"../week-4/data-processed/root_filtered.csv"</span>)</span>
<span id="cb45-1127"><a href="#cb45-1127" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1128"><a href="#cb45-1128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1129"><a href="#cb45-1129" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Look after future you</span><span class="co">](#arabidopsis-and-future-you)</span></span>
<span id="cb45-1130"><a href="#cb45-1130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1131"><a href="#cb45-1131" aria-hidden="true" tabindex="-1"></a><span class="fu">## üíâ *Leishmania* {#leishmania-filterqc}</span></span>
<span id="cb45-1132"><a href="#cb45-1132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1133"><a href="#cb45-1133" aria-hidden="true" tabindex="-1"></a>Our samples look to be similarly well sequenced. There are no samples we</span>
<span id="cb45-1134"><a href="#cb45-1134" aria-hidden="true" tabindex="-1"></a>should remove. However, some genes are not expressed or the expression</span>
<span id="cb45-1135"><a href="#cb45-1135" aria-hidden="true" tabindex="-1"></a>values are so low in for a gene that they are uninformative. We will</span>
<span id="cb45-1136"><a href="#cb45-1136" aria-hidden="true" tabindex="-1"></a>filter the <span class="in">`pro_meta_summary_gene`</span> dataframe to obtain a list of</span>
<span id="cb45-1137"><a href="#cb45-1137" aria-hidden="true" tabindex="-1"></a><span class="in">`gene_id`</span> we can use to filter <span class="in">`pro_meta`</span>.</span>
<span id="cb45-1138"><a href="#cb45-1138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1139"><a href="#cb45-1139" aria-hidden="true" tabindex="-1"></a>My suggestion is to include only the genes with counts in at least 3</span>
<span id="cb45-1140"><a href="#cb45-1140" aria-hidden="true" tabindex="-1"></a>samples and those with total counts above 20. I chose 3 because that</span>
<span id="cb45-1141"><a href="#cb45-1141" aria-hidden="true" tabindex="-1"></a>would keep genes expressed only in one treatment: <span class="sc">\[</span>0, 0, 0<span class="sc">\]</span> <span class="sc">\[</span>#,#,#<span class="sc">\]</span>.</span>
<span id="cb45-1142"><a href="#cb45-1142" aria-hidden="true" tabindex="-1"></a>This is a difference we cannot test statistically, but which matters</span>
<span id="cb45-1143"><a href="#cb45-1143" aria-hidden="true" tabindex="-1"></a>biologically.</span>
<span id="cb45-1144"><a href="#cb45-1144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1145"><a href="#cb45-1145" aria-hidden="true" tabindex="-1"></a>üé¨ Filter the summary by gene dataframe:</span>
<span id="cb45-1146"><a href="#cb45-1146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1149"><a href="#cb45-1149" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1150"><a href="#cb45-1150" aria-hidden="true" tabindex="-1"></a>pro_meta_summary_gene_filtered <span class="ot">&lt;-</span> pro_meta_summary_gene <span class="sc">|&gt;</span> </span>
<span id="cb45-1151"><a href="#cb45-1151" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(total <span class="sc">&gt;</span> <span class="dv">20</span>) <span class="sc">|&gt;</span> </span>
<span id="cb45-1152"><a href="#cb45-1152" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(n_above_zero <span class="sc">&gt;=</span> <span class="dv">3</span>)</span>
<span id="cb45-1153"><a href="#cb45-1153" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1154"><a href="#cb45-1154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1155"><a href="#cb45-1155" aria-hidden="true" tabindex="-1"></a>‚ùì How many genes do you have left</span>
<span id="cb45-1156"><a href="#cb45-1156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1157"><a href="#cb45-1157" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb45-1158"><a href="#cb45-1158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1159"><a href="#cb45-1159" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- there are 8437 genes left. 240 have total counts less than 20 and or --&gt;</span></span>
<span id="cb45-1160"><a href="#cb45-1160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1161"><a href="#cb45-1161" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- have counts in fewer than 3 samples. --&gt;</span></span>
<span id="cb45-1162"><a href="#cb45-1162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1163"><a href="#cb45-1163" aria-hidden="true" tabindex="-1"></a>üé¨ Use the list of <span class="in">`gene_id`</span> in the filtered summary to filter the</span>
<span id="cb45-1164"><a href="#cb45-1164" aria-hidden="true" tabindex="-1"></a>original dataset:</span>
<span id="cb45-1165"><a href="#cb45-1165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1168"><a href="#cb45-1168" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1169"><a href="#cb45-1169" aria-hidden="true" tabindex="-1"></a>pro_meta_filtered <span class="ot">&lt;-</span> pro_meta <span class="sc">|&gt;</span> </span>
<span id="cb45-1170"><a href="#cb45-1170" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene_id <span class="sc">%in%</span>  pro_meta_summary_gene_filtered<span class="sc">$</span>gene_id)</span>
<span id="cb45-1171"><a href="#cb45-1171" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1172"><a href="#cb45-1172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1173"><a href="#cb45-1173" aria-hidden="true" tabindex="-1"></a>üé¨ Write the filtered data to file:</span>
<span id="cb45-1174"><a href="#cb45-1174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1177"><a href="#cb45-1177" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1178"><a href="#cb45-1178" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(pro_meta_filtered, </span>
<span id="cb45-1179"><a href="#cb45-1179" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"data-processed/pro_meta_filtered.csv"</span>)</span>
<span id="cb45-1180"><a href="#cb45-1180" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1181"><a href="#cb45-1181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1184"><a href="#cb45-1184" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1185"><a href="#cb45-1185" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb45-1186"><a href="#cb45-1186" aria-hidden="true" tabindex="-1"></a><span class="co">#---UNDER THE HOOD CODE---</span></span>
<span id="cb45-1187"><a href="#cb45-1187" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(pro_meta_filtered,  </span>
<span id="cb45-1188"><a href="#cb45-1188" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"../week-4/data-processed/pro_meta_filtered.csv"</span>)</span>
<span id="cb45-1189"><a href="#cb45-1189" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1190"><a href="#cb45-1190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1191"><a href="#cb45-1191" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Look after future you</span><span class="co">](#leishmania-and-future-you)</span></span>
<span id="cb45-1192"><a href="#cb45-1192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1193"><a href="#cb45-1193" aria-hidden="true" tabindex="-1"></a><span class="fu">## üê≠ Stem cells {#stem-cells-filterqc}</span></span>
<span id="cb45-1194"><a href="#cb45-1194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1195"><a href="#cb45-1195" aria-hidden="true" tabindex="-1"></a>In this dataset, we will not see and genes that are not expressed in any</span>
<span id="cb45-1196"><a href="#cb45-1196" aria-hidden="true" tabindex="-1"></a>of the cells because we are using a specific subset of the transcriptome</span>
<span id="cb45-1197"><a href="#cb45-1197" aria-hidden="true" tabindex="-1"></a>that was deliberately selected. In other words unexpressed genes have</span>
<span id="cb45-1198"><a href="#cb45-1198" aria-hidden="true" tabindex="-1"></a>already been filtered out. However, it is good practices to verify there</span>
<span id="cb45-1199"><a href="#cb45-1199" aria-hidden="true" tabindex="-1"></a>are no unexpressed genes before we embark on our analysis.</span>
<span id="cb45-1200"><a href="#cb45-1200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1201"><a href="#cb45-1201" aria-hidden="true" tabindex="-1"></a>Where the sum of all the values in the rows is zero, all the entries</span>
<span id="cb45-1202"><a href="#cb45-1202" aria-hidden="true" tabindex="-1"></a>must be zero. We can use this to find any the genes that are not</span>
<span id="cb45-1203"><a href="#cb45-1203" aria-hidden="true" tabindex="-1"></a>expressed in any of the cells. To do row wise aggregates such as the sum</span>
<span id="cb45-1204"><a href="#cb45-1204" aria-hidden="true" tabindex="-1"></a>across rows we can use the <span class="in">`rowwise()`</span> function. <span class="in">`c_across()`</span> allows us</span>
<span id="cb45-1205"><a href="#cb45-1205" aria-hidden="true" tabindex="-1"></a>to use the colon notation to select columns. This is very useful when</span>
<span id="cb45-1206"><a href="#cb45-1206" aria-hidden="true" tabindex="-1"></a>you have a lot of columns because it would be annoying to have to list</span>
<span id="cb45-1207"><a href="#cb45-1207" aria-hidden="true" tabindex="-1"></a>all of them <span class="in">`HSPC_001:Prog_852`</span> means all the columns from <span class="in">`HSPC_001`</span> to</span>
<span id="cb45-1208"><a href="#cb45-1208" aria-hidden="true" tabindex="-1"></a><span class="in">`Prog_852`</span>.</span>
<span id="cb45-1209"><a href="#cb45-1209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1210"><a href="#cb45-1210" aria-hidden="true" tabindex="-1"></a>üé¨ Find the genes that are 0 in every column of the hspc_prog dataframe:</span>
<span id="cb45-1211"><a href="#cb45-1211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1214"><a href="#cb45-1214" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1215"><a href="#cb45-1215" aria-hidden="true" tabindex="-1"></a>hspc_prog <span class="sc">|&gt;</span> </span>
<span id="cb45-1216"><a href="#cb45-1216" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">|&gt;</span> </span>
<span id="cb45-1217"><a href="#cb45-1217" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">sum</span>(<span class="fu">c_across</span>(HSPC_001<span class="sc">:</span>Prog_852)) <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb45-1218"><a href="#cb45-1218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1219"><a href="#cb45-1219" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1220"><a href="#cb45-1220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1221"><a href="#cb45-1221" aria-hidden="true" tabindex="-1"></a>Notice that we have summed across all the columns.</span>
<span id="cb45-1222"><a href="#cb45-1222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1223"><a href="#cb45-1223" aria-hidden="true" tabindex="-1"></a>‚ùì What do you conclude?</span>
<span id="cb45-1224"><a href="#cb45-1224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1225"><a href="#cb45-1225" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb45-1226"><a href="#cb45-1226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1227"><a href="#cb45-1227" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- The are no genes that are completely unexpressed in  --&gt;</span></span>
<span id="cb45-1228"><a href="#cb45-1228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1229"><a href="#cb45-1229" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- this set of genes --&gt;</span></span>
<span id="cb45-1230"><a href="#cb45-1230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1231"><a href="#cb45-1231" aria-hidden="true" tabindex="-1"></a>üé¨ Write combined data to file:</span>
<span id="cb45-1232"><a href="#cb45-1232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1235"><a href="#cb45-1235" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1236"><a href="#cb45-1236" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(hspc_prog, </span>
<span id="cb45-1237"><a href="#cb45-1237" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"data-processed/hspc_prog.csv"</span>)</span>
<span id="cb45-1238"><a href="#cb45-1238" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1239"><a href="#cb45-1239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1242"><a href="#cb45-1242" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-1243"><a href="#cb45-1243" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb45-1244"><a href="#cb45-1244" aria-hidden="true" tabindex="-1"></a><span class="co">#---UNDER THE HOOD CODE---</span></span>
<span id="cb45-1245"><a href="#cb45-1245" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(hspc_prog, </span>
<span id="cb45-1246"><a href="#cb45-1246" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"../week-4/data-processed/hspc_prog.csv"</span>)</span>
<span id="cb45-1247"><a href="#cb45-1247" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-1248"><a href="#cb45-1248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1249"><a href="#cb45-1249" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Look after future you</span><span class="co">](#stem-cells-and-future-you)</span></span>
<span id="cb45-1250"><a href="#cb45-1250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1251"><a href="#cb45-1251" aria-hidden="true" tabindex="-1"></a><span class="fu"># ü§ó Look after future you!</span></span>
<span id="cb45-1252"><a href="#cb45-1252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1253"><a href="#cb45-1253" aria-hidden="true" tabindex="-1"></a>You need only do the section for your own project data but **completing</span>
<span id="cb45-1254"><a href="#cb45-1254" aria-hidden="true" tabindex="-1"></a>this step is important**.</span>
<span id="cb45-1255"><a href="#cb45-1255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1256"><a href="#cb45-1256" aria-hidden="true" tabindex="-1"></a><span class="fu">## üéÑ *Arabidopsis* and future you {#arabidopsis-and-future-you}</span></span>
<span id="cb45-1257"><a href="#cb45-1257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1258"><a href="#cb45-1258" aria-hidden="true" tabindex="-1"></a>üé¨ Create a new Project, <span class="in">`arab-88H`</span>, populated with folders and your</span>
<span id="cb45-1259"><a href="#cb45-1259" aria-hidden="true" tabindex="-1"></a>data. Make a script file called <span class="in">`cont-low-root.R`</span>. This will a be</span>
<span id="cb45-1260"><a href="#cb45-1260" aria-hidden="true" tabindex="-1"></a>commented analysis of comparison between control and low Ni root tissue.</span>
<span id="cb45-1261"><a href="#cb45-1261" aria-hidden="true" tabindex="-1"></a>You will build on this each workshop and be able to use it as a template</span>
<span id="cb45-1262"><a href="#cb45-1262" aria-hidden="true" tabindex="-1"></a>to examine other comparisons. Copy in the *appropriate code and</span>
<span id="cb45-1263"><a href="#cb45-1263" aria-hidden="true" tabindex="-1"></a>comments* from <span class="in">`workshop-1.R`</span>. Edit to improve your comments where your</span>
<span id="cb45-1264"><a href="#cb45-1264" aria-hidden="true" tabindex="-1"></a>understanding has developed since you made them. Make sure you can close</span>
<span id="cb45-1265"><a href="#cb45-1265" aria-hidden="true" tabindex="-1"></a>down RStudio, reopen it and run your whole script again.</span>
<span id="cb45-1266"><a href="#cb45-1266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1267"><a href="#cb45-1267" aria-hidden="true" tabindex="-1"></a><span class="fu">## üíâ *Leishmania* and future you {#leishmania-and-future-you}</span></span>
<span id="cb45-1268"><a href="#cb45-1268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1269"><a href="#cb45-1269" aria-hidden="true" tabindex="-1"></a>üé¨ Create a new Project, <span class="in">`leish-88H`</span>, populated with folders and your</span>
<span id="cb45-1270"><a href="#cb45-1270" aria-hidden="true" tabindex="-1"></a>data. Make a script file called <span class="in">`pro_meta.R`</span>. This will a be commented</span>
<span id="cb45-1271"><a href="#cb45-1271" aria-hidden="true" tabindex="-1"></a>analysis of comparison procyclic promastigote and metacyclic</span>
<span id="cb45-1272"><a href="#cb45-1272" aria-hidden="true" tabindex="-1"></a>promastigotes. You will build on this each workshop and be able to use</span>
<span id="cb45-1273"><a href="#cb45-1273" aria-hidden="true" tabindex="-1"></a>it as a template to examine other comparisons. Copy in the *appropriate</span>
<span id="cb45-1274"><a href="#cb45-1274" aria-hidden="true" tabindex="-1"></a>code and comments* from <span class="in">`workshop-1.R`</span>. Edit to improve your comments</span>
<span id="cb45-1275"><a href="#cb45-1275" aria-hidden="true" tabindex="-1"></a>where your understanding has developed since you made them. Make sure</span>
<span id="cb45-1276"><a href="#cb45-1276" aria-hidden="true" tabindex="-1"></a>you can close down RStudio, reopen it and run your whole script again.</span>
<span id="cb45-1277"><a href="#cb45-1277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1278"><a href="#cb45-1278" aria-hidden="true" tabindex="-1"></a><span class="fu">## üê≠ Stem cells and future you {#stem-cells-and-future-you}</span></span>
<span id="cb45-1279"><a href="#cb45-1279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1280"><a href="#cb45-1280" aria-hidden="true" tabindex="-1"></a>üé¨ Create a new Project, <span class="in">`mice-88H`</span>, populated with folders and your</span>
<span id="cb45-1281"><a href="#cb45-1281" aria-hidden="true" tabindex="-1"></a>data. Make a script file called <span class="in">`hspc-prog.R`</span>. This will a be commented</span>
<span id="cb45-1282"><a href="#cb45-1282" aria-hidden="true" tabindex="-1"></a>analysis of the hspc cells vs the prog cells. You will build on this</span>
<span id="cb45-1283"><a href="#cb45-1283" aria-hidden="true" tabindex="-1"></a>each workshop and be able to use it as a template to examine other</span>
<span id="cb45-1284"><a href="#cb45-1284" aria-hidden="true" tabindex="-1"></a>comparisons. Copy in the *appropriate code and comments* from</span>
<span id="cb45-1285"><a href="#cb45-1285" aria-hidden="true" tabindex="-1"></a><span class="in">`workshop-1.R`</span>. Edit to improve your comments where your understanding</span>
<span id="cb45-1286"><a href="#cb45-1286" aria-hidden="true" tabindex="-1"></a>has developed since you made them. Make sure you can close down RStudio,</span>
<span id="cb45-1287"><a href="#cb45-1287" aria-hidden="true" tabindex="-1"></a>reopen it and run your whole script again.</span>
<span id="cb45-1288"><a href="#cb45-1288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1289"><a href="#cb45-1289" aria-hidden="true" tabindex="-1"></a><span class="fu"># ü•≥ Finished</span></span>
<span id="cb45-1290"><a href="#cb45-1290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1291"><a href="#cb45-1291" aria-hidden="true" tabindex="-1"></a>Well Done!</span>
<span id="cb45-1292"><a href="#cb45-1292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1293"><a href="#cb45-1293" aria-hidden="true" tabindex="-1"></a><span class="fu"># Independent study following the workshop</span></span>
<span id="cb45-1294"><a href="#cb45-1294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1295"><a href="#cb45-1295" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Consolidate</span><span class="co">](study_after_workshop.qmd)</span></span>
<span id="cb45-1296"><a href="#cb45-1296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1297"><a href="#cb45-1297" aria-hidden="true" tabindex="-1"></a><span class="fu"># The Code file</span></span>
<span id="cb45-1298"><a href="#cb45-1298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1299"><a href="#cb45-1299" aria-hidden="true" tabindex="-1"></a>These contain all the code needed in the workshop even where it is not</span>
<span id="cb45-1300"><a href="#cb45-1300" aria-hidden="true" tabindex="-1"></a>visible on the webpage.</span>
<span id="cb45-1301"><a href="#cb45-1301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1302"><a href="#cb45-1302" aria-hidden="true" tabindex="-1"></a>The <span class="co">[</span><span class="ot">workshop.qmd</span><span class="co">](workshop.qmd)</span> file is the file I use to compile the</span>
<span id="cb45-1303"><a href="#cb45-1303" aria-hidden="true" tabindex="-1"></a>practical. Qmd stands for Quarto markdown. It allows code and ordinary</span>
<span id="cb45-1304"><a href="#cb45-1304" aria-hidden="true" tabindex="-1"></a>text to be interleaved to produce well-formatted reports including</span>
<span id="cb45-1305"><a href="#cb45-1305" aria-hidden="true" tabindex="-1"></a>webpages. Right-click on the link and choose View the source code for</span>
<span id="cb45-1306"><a href="#cb45-1306" aria-hidden="true" tabindex="-1"></a>this workshop using the <span class="in">`&lt;/&gt; Code`</span> button at the top of the page. Coding</span>
<span id="cb45-1307"><a href="#cb45-1307" aria-hidden="true" tabindex="-1"></a>and thinking answers are marked with <span class="in">`#---CODING ANSWER---`</span> and</span>
<span id="cb45-1308"><a href="#cb45-1308" aria-hidden="true" tabindex="-1"></a><span class="in">`#---THINKING ANSWER---`</span></span>
<span id="cb45-1309"><a href="#cb45-1309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1310"><a href="#cb45-1310" aria-hidden="true" tabindex="-1"></a>Pages made with R <span class="co">[</span><span class="ot">@R-core</span><span class="co">]</span>, Quarto <span class="co">[</span><span class="ot">@Allaire_Quarto_2024</span><span class="co">]</span>, <span class="in">`knitr`</span></span>
<span id="cb45-1311"><a href="#cb45-1311" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@knitr1; @knitr2; @knitr3</span><span class="co">]</span>, <span class="in">`kableExtra`</span> <span class="co">[</span><span class="ot">@kableExtra</span><span class="co">]</span></span>
<span id="cb45-1312"><a href="#cb45-1312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-1313"><a href="#cb45-1313" aria-hidden="true" tabindex="-1"></a><span class="fu"># References</span></span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Rand E (2025). Data Analysis for Group Project. <a href="https://3mmarand.github.io/BIO00088H-data/" class="uri">https://3mmarand.github.io/BIO00088H-data/</a>.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/3mmaRand/BIO00088H-data/edit/main/transcriptomics/week-3/workshop.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li></ul></div></div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
    <a class="nav-link" href="https://github.com/3mmaRand">
<p>Emma on GitHub</p>
</a>
  </li>  
</ul>
</div>
  </div>
</footer><script>
document.addEventListener("DOMContentLoaded", function() {
  // grab all links under ‚ÄúOther Links‚Äù in the right-hand TOC
  const links = document.querySelectorAll('.quarto-other-links ul li a');
  links.forEach(function(a) {
    // find the one whose text is exactly ‚ÄúMake a suggestion‚Äù
    if (a.textContent.trim() === 'Make a suggestion') {
      const pageUrl   = encodeURIComponent(window.location.href);
      const pageTitle = encodeURIComponent(document.title);
      a.href = 'https://github.com/3mmaRand/BIO00088H-data/issues/new'
             + '?title=' + pageTitle
             + '&body='  + pageUrl;
      a.target = '_blank';
    }
  });
});
</script><script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>


</body></html>