<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Emma Rand">
<title>Data Analysis for Group Project - Workshop</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="../../styles.css">
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Data Analysis for Group Project</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Welcome!</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../core/core.html" rel="" target="">
 <span class="menu-text">Core</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../omics/omics.html" rel="" target="" aria-current="page">
 <span class="menu-text">Omics</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools ms-auto">
    <a href="https://twitter.com/er_13" rel="" title="Emma's Twitter" class="quarto-navigation-tool px-1" aria-label="Emma's Twitter"><i class="bi bi-twitter"></i></a>
    <a href="https://github.com/3mmaRand/BIO00088H-data" rel="" title="BIO00088H-data GitHub" class="quarto-navigation-tool px-1" aria-label="BIO00088H-data GitHub"><i class="bi bi-github"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../omics/week-4/overview.html">Week 4: Statistical Analysis</a></li><li class="breadcrumb-item"><a href="../../omics/week-4/workshop.html">Workshop</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/omics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Omics Data Analysis for Group Project</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Week 3: Hello data!</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-3/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-3/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-3/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-3/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Week 4: Statistical Analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-4/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-4/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-4/workshop.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-4/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Week 5: Visualising and Interpreting</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-5/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-5/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-5/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../omics/week-5/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul>
<li><a href="#session-overview" id="toc-session-overview" class="nav-link" data-scroll-target="#session-overview">Session overview</a></li>
  </ul>
</li>
  <li><a href="#set-up" id="toc-set-up" class="nav-link" data-scroll-target="#set-up">Set up</a></li>
  <li>
<a href="#analysis" id="toc-analysis" class="nav-link" data-scroll-target="#analysis">ğŸ¸ Analysis</a>
  <ul>
<li><a href="#import" id="toc-import" class="nav-link" data-scroll-target="#import">Import</a></li>
  <li><a href="#genes-expressed-in-one-treatment" id="toc-genes-expressed-in-one-treatment" class="nav-link" data-scroll-target="#genes-expressed-in-one-treatment">Genes expressed in one treatment</a></li>
  <li><a href="#create-deseqdataset-object" id="toc-create-deseqdataset-object" class="nav-link" data-scroll-target="#create-deseqdataset-object">Create DESeqDataSet object</a></li>
  <li><a href="#prepare-the-normalised-counts" id="toc-prepare-the-normalised-counts" class="nav-link" data-scroll-target="#prepare-the-normalised-counts">Prepare the normalised counts</a></li>
  <li><a href="#differential-expression-analysis" id="toc-differential-expression-analysis" class="nav-link" data-scroll-target="#differential-expression-analysis">Differential expression analysis</a></li>
  </ul>
</li>
  <li>
<a href="#analysis-1" id="toc-analysis-1" class="nav-link" data-scroll-target="#analysis-1">ğŸ­ Analysis</a>
  <ul>
<li><a href="#import-1" id="toc-import-1" class="nav-link" data-scroll-target="#import-1">Import</a></li>
  <li><a href="#combine-the-two-datasets" id="toc-combine-the-two-datasets" class="nav-link" data-scroll-target="#combine-the-two-datasets">Combine the two datasets</a></li>
  <li><a href="#filter-to-remove-unexpressed-genes" id="toc-filter-to-remove-unexpressed-genes" class="nav-link" data-scroll-target="#filter-to-remove-unexpressed-genes">Filter to remove unexpressed genes</a></li>
  <li><a href="#find-the-genes-that-are-expressed-in-only-one-cell-type" id="toc-find-the-genes-that-are-expressed-in-only-one-cell-type" class="nav-link" data-scroll-target="#find-the-genes-that-are-expressed-in-only-one-cell-type">Find the genes that are expressed in only one cell type</a></li>
  <li><a href="#differential-expression-analysis-1" id="toc-differential-expression-analysis-1" class="nav-link" data-scroll-target="#differential-expression-analysis-1">Differential expression analysis</a></li>
  </ul>
</li>
  <li><a href="#look-after-future-you" id="toc-look-after-future-you" class="nav-link" data-scroll-target="#look-after-future-you">ğŸ¤— Look after future you!</a></li>
  <li><a href="#finished" id="toc-finished" class="nav-link" data-scroll-target="#finished">ğŸ¥³ Finished</a></li>
  <li><a href="#independent-study-following-the-workshop" id="toc-independent-study-following-the-workshop" class="nav-link" data-scroll-target="#independent-study-following-the-workshop">Independent study following the workshop</a></li>
  <li><a href="#the-code-file" id="toc-the-code-file" class="nav-link" data-scroll-target="#the-code-file">The Code file</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Workshop</h1>
<p class="subtitle lead">Omics 2: Statistical Analysis</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Emma Rand </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">3 November, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header><section id="introduction" class="level1"><h1>Introduction</h1>
<section id="session-overview" class="level2"><h2 class="anchored" data-anchor-id="session-overview">Session overview</h2>
<p>In the workshop, you will learn how to perform differential expression analysis on raw counts using <strong><code>DESeq2</code></strong> <span class="citation" data-cites="DESeq2">(<a href="#ref-DESeq2" role="doc-biblioref">Love, Huber, and Anders 2014</a>)</span> or on logged normalised expression values using <strong><code>scran</code></strong> <span class="citation" data-cites="scran">(<a href="#ref-scran" role="doc-biblioref">Lun, McCarthy, and Marioni 2016</a>)</span> or both.</p>
</section></section><section id="set-up" class="level1"><h1>Set up</h1>
<p>Either:</p>
<p>ğŸ¬ Open ğŸ¸ <code>frogs-88H</code> Project and the <code>cont-fgf-s30.R</code> script.</p>
<p>Or</p>
<p>ğŸ¬ Open ğŸ­ <code>mice-88H</code> Project and the <code>hspc-prog.R</code> script.</p>
<p>ğŸ¬ Make a new folder <code>results</code> in the project directory. You can use the New Folder button in the Files pane but here I have used the <strong><code>fs</code></strong> <span class="citation" data-cites="fs">(<a href="#ref-fs" role="doc-biblioref">Hester, Wickham, and CsÃ¡rdi 2023</a>)</span> package</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/create.html">dir_create</a></span><span class="op">(</span><span class="st">"results"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is where we will save our results.</p>
<p>ğŸ¬ Load <strong><code>tidyverse</code></strong> <span class="citation" data-cites="tidyverse">(<a href="#ref-tidyverse" role="doc-biblioref">Wickham et al. 2019</a>)</span> You most likely have this code at the top of <code>cont-fgf-s30.R</code> or <code>hspc-prog.R</code> already.</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€
âœ” dplyr     1.1.3     âœ” readr     2.1.4
âœ” forcats   1.0.0     âœ” stringr   1.5.0
âœ” ggplot2   3.4.3     âœ” tibble    3.2.1
âœ” lubridate 1.9.3     âœ” tidyr     1.3.0
âœ” purrr     1.0.2     
â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€
âœ– dplyr::filter() masks stats::filter()
âœ– dplyr::lag()    masks stats::lag()
â„¹ Use the conflicted package to force all conflicts to become errors</code></pre>
<p>Have you ever stopped to think about this message? It is telling us that there are functions in the <code>dplyr</code> package that have the same name as functions in the <code>stats</code> package and that R will use the <code>dplyr</code> version. As this is what you want, this has always been fine. It still is fine in this case. However, as you start to load more packages, you will want to know if you are using a function from a package that has the same name as a function in another loaded package. This is where the <strong><code>conflicted</code></strong> <span class="citation" data-cites="conflicted">(<a href="#ref-conflicted" role="doc-biblioref">Wickham 2023</a>)</span> package comes in. <strong><code>Conflicted</code></strong> will warn you when you are using a function that has the same name as a function in another package. You can then choose which function to use.</p>
<p>ğŸ¬ Load the <strong><code>conflicted</code></strong> package:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://conflicted.r-lib.org/">conflicted</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Instead of getting a warning every time you are using a function that has a function with the same name in another package, we can declare a preference for one function over another. This is useful for the functions you use a lot or ones where you are certain you always want to use a particular function.</p>
<p>ğŸ¬ Use the <strong><code>dplyr</code></strong> version of <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> by default:</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://conflicted.r-lib.org/reference/conflict_prefer.html">conflict_prefer</a></span><span class="op">(</span><span class="st">"filter"</span>, <span class="st">"dplyr"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="analysis" class="level1"><h1>ğŸ¸ Analysis</h1>
<p>We will carry out several steps</p>
<ol type="1">
<li>Import the data</li>
<li>Find the genes that are expressed in only one treatment (the control or the FGF-treated)</li>
<li>Create a DESeqDataSet object. This is a special object that is used by the <strong><code>DESeq2</code></strong> package</li>
<li>Prepare the normalised counts from the DESeqDataSet object.</li>
<li>Do differential expression analysis on the genes. This needs to be done on the raw counts.</li>
</ol>
<p>The creation of the DESeqDataSet object, the preparation of the normalised counts and the differential expression analysis are all done with the <strong><code>DESeq2</code></strong> package</p>
<section id="import" class="level2"><h2 class="anchored" data-anchor-id="import">Import</h2>
<p>We need to import the S30 data that were filtered to remove genes with 4, 5 or 6 zeros and those where the total counts was less than 20.</p>
<p>ğŸ¬ Import the data from the <code>data-processed</code> folder.</p>
</section><section id="genes-expressed-in-one-treatment" class="level2"><h2 class="anchored" data-anchor-id="genes-expressed-in-one-treatment">Genes expressed in one treatment</h2>
<p>The genes expressed in only one treatment group are those with zeros in all three replicates in one group and non-zero values in all three replicates in the other group. For example, those shown here:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 Ã— 7
  xenbase_gene_id  S30_C_5 S30_C_6 S30_C_A S30_F_5 S30_F_6 S30_F_A
  &lt;chr&gt;              &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1 XB-GENE-1018260        0       0       0      10       2      16
2 XB-GENE-17330117       0       0       0      13       4      17
3 XB-GENE-17332184       0       0       0       6      19       6</code></pre>
</div>
</div>
<p>We will use <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> to find these genes.</p>
<p>ğŸ¬ Find the genes that are expressed only in the FGF-treated group:</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">s30_fgf_only</span> <span class="op">&lt;-</span> <span class="va">s30_filtered</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">S30_C_5</span> <span class="op">==</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_C_6</span> <span class="op">==</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_C_A</span> <span class="op">==</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_F_5</span> <span class="op">&gt;</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_F_6</span> <span class="op">&gt;</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_F_A</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>â“ How many genes are expressed only in the FGF-treated group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 26 genes expressed only in the FGF-treated group. -->
<p>ğŸ¬ Now you find any genes that are expressed only in the control group.</p>
<p>â“ Do the results make sense to you in light of what you know about the biology?</p>
<!-- #---THINKING ANSWER--- -->
<!-- there are no genes expressed in the control group but not in the FGF-treated group -->
<!-- Since FGF is a growth factor, it is likely to induce the expression of genes. Therefore,  -->
<!-- it is not surprising to me that there are no genes expressed in the control group but  -->
<!-- not in the FGF-treated group. However, I'm not an expert in developmental biology.  -->
<!-- Domain expertise is important when analysing data so discuss your results with Betsy  -->
<!-- and your group. -->
<p>ğŸ¬ Write to file (saved in <code>results</code>) all the genes that are expressed one group only.</p>
</section><section id="create-deseqdataset-object" class="level2"><h2 class="anchored" data-anchor-id="create-deseqdataset-object">Create DESeqDataSet object</h2>
<p>ğŸ¬ Load the DESeq2 package:</p>
<p>A DEseqDataSet object is a custom data type that is used by <strong><code>DESeq2</code></strong>. Custom data types are common in the Bioconductor<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> packages. They are used to store data in a way that is useful for the analysis. These data types typically have data, transformed data, metadata and experimental designs within them.</p>
<p>To create a DESeqDataSet object, we need to provide three things:</p>
<ol type="1">
<li>The raw counts - these are what we imported into <code>s30_filtered</code>
</li>
<li>The meta data which gives information about the samples and which treatment groups they belong to</li>
<li>A design matrix which captures the design of the statistical model.</li>
</ol>
<p>The counts must in a matrix rather than a dataframe. Unlike a dataframe, a matrix has columns of all the same type. That is, it will contain only the counts. The gene ids are given as row names rather than a column. The <code><a href="https://rdrr.io/r/base/matrix.html">matrix()</a></code> function will create a matrix from a dataframe of columns of the same type and the <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> function can be used to remove the gene ids column.</p>
<p>ğŸ¬ Create a matrix of the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">s30_count_mat</span> <span class="op">&lt;-</span> <span class="va">s30_filtered</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">xenbase_gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Add the gene ids as row names to the matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># add the row names to the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">s30_count_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">s30_filtered</span><span class="op">$</span><span class="va">xenbase_gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You might want to view the matrix.</p>
<p>The metadata are in a file, <a href="meta/frog_meta_data.txt">frog_meta_data.txt</a>. This is a tab-delimited file. The first column is the sample name and the second column is the treatment group.</p>
<p>ğŸ¬ Make a folder called <code>meta</code> and save the file to it.</p>
<p>ğŸ¬ Read the metadata into a dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_table.html">read_table</a></span><span class="op">(</span><span class="st">"meta/frog_meta_data.txt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Examine the resulting dataframe.</p>
<p>We need to add the sample names as row names to the metadata dataframe. This is because the DESeqDataSet object will use the row names to match the samples in the metadata to the samples in the counts matrix.</p>
<p>ğŸ¬ Add the sample names as row names to the metadata dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">sample_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>(you will get a warning message but you can ignore it)</p>
<p>We are dealing only with the S30 data so we need to remove the samples that are not in the S30 data.</p>
<p>ğŸ¬ Filter the metadata to keep only the S30 information:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta_S30</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">stage</span> <span class="op">==</span> <span class="st">"stage_30"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 Ã— 4
  sample_id stage    treatment sibling_rep
* &lt;chr&gt;     &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;      
1 S30_C_5   stage_30 control   five       
2 S30_C_6   stage_30 control   six        
3 S30_C_A   stage_30 control   A          
4 S30_F_5   stage_30 FGF       five       
5 S30_F_6   stage_30 FGF       six        
6 S30_F_A   stage_30 FGF       A          </code></pre>
</div>
</div>
<p>We can now create the DESeqDataSet object. The design formula describes the statistical model You should notice that it is the same sort of formula we used in <code><a href="https://rdrr.io/r/stats/t.test.html">t.test()</a></code>, <code><a href="https://rdrr.io/r/stats/lm.html">lm()</a></code>,<code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> etc. The <code>~</code> indicates that the left hand side is the response variable (in this case counts) and the right hand side are the explanatory variables. We are interested in the difference between the treatments but we include <code>sibling_rep</code> to account for the fact that the data are paired. The names of the columns in the count matrix have to match the names in the metadata dataframe and the names of the explanatory variables in the design formula have to match the names of columns in the metadata.</p>
<p>ğŸ¬ Create the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="va">s30_count_mat</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">meta_S30</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span> <span class="va">treatment</span> <span class="op">+</span> <span class="va">sibling_rep</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The warning â€œWarning: some variables in design formula are characters, converting to factorsâ€ just means that the variable type of treatment and sibling_rep in the metadata dataframe are char. This is not a as <code>DESeqDataSetFromMatrix()</code> has made them into the factors it needs.</p>
<p>ğŸ¬ Examine the DESeqDataSet object.</p>
<p>The counts are in <code>dds@assays@data@listData[["counts"]]</code> and the metadata are in <code>dds@colData</code> but the easiest way to see them is to use the <code>counts()</code> and <code>colData()</code> functions from the <strong><code>DESeq2</code></strong> package.</p>
<p>ğŸ¬ View the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in .External2(C_dataviewer, x, title): unable to start data viewer</code></pre>
</div>
</div>
<p>You should be able to see that this is the same as in <code>s30_count_mat</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 4 columns
          sample_id       stage treatment sibling_rep
        &lt;character&gt; &lt;character&gt;  &lt;factor&gt;    &lt;factor&gt;
S30_C_5     S30_C_5    stage_30   control        five
S30_C_6     S30_C_6    stage_30   control        six 
S30_C_A     S30_C_A    stage_30   control        A   
S30_F_5     S30_F_5    stage_30   FGF            five
S30_F_6     S30_F_6    stage_30   FGF            six 
S30_F_A     S30_F_A    stage_30   FGF            A   </code></pre>
</div>
</div>
</section><section id="prepare-the-normalised-counts" class="level2"><h2 class="anchored" data-anchor-id="prepare-the-normalised-counts">Prepare the normalised counts</h2>
<p>The normalised counts are the counts that have been transformed to account for the library size (i.e., the total number of reads in a sample) and the gene length. We have to first estimate the normalisation factors and store them in the DESeqDataSet object and then we can get the normalised counts.</p>
<p>ğŸ¬ Estimate the factors for normalisation and store them in the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Look at the factors (just for information):</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  S30_C_5   S30_C_6   S30_C_A   S30_F_5   S30_F_6   S30_F_A 
0.8812200 0.9454600 1.2989886 1.0881870 1.0518961 0.8322894 </code></pre>
</div>
</div>
<p>To get the normalised counts we again used the <code>counts()</code> function but this time we use the <code>normalized=TRUE</code> argument.</p>
<p>ğŸ¬ Save the normalised to a matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">normalised_counts</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will write the normalised counts to a file so that we can use them in the future.</p>
<p>ğŸ¬ Make a dataframe of the normalised counts, add a column for the gene ids and write to file:</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">normalised_counts</span>,</span>
<span>           xenbase_gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">normalised_counts</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"results/S30_normalised_counts.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="differential-expression-analysis" class="level2"><h2 class="anchored" data-anchor-id="differential-expression-analysis">Differential expression analysis</h2>
<p>We used the <code>DESeq()</code> function to do the differential expression analysis. This function fits the statistical model to the data and then uses the model to calculate the significance of the difference between the treatments. It again stored the results in the DESseqDataSet object. Note that the differential expression needs the raw (unnormalised counts) as it does its own normalisation as part of the process.</p>
<p>ğŸ¬ Run the differential expression analysis:</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function will take only a few moments to run on this data but can take longer for bigger datasets.</p>
<p>We need to define the contrasts we want to test. We want to test the difference between the treatments so we will define the contrast as <code>FGF</code> and <code>control</code>.</p>
<p>ğŸ¬ Define the contrast:</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">contrast_fgf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"treatment"</span>, <span class="st">"FGF"</span>, <span class="st">"control"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that <code>treatment</code> is the name of the column in the metadata dataframe and <code>FGF</code> and <code>control</code> are the names of the levels in the <code>treatment</code> column. By putting them in the order <code>FGF</code> , <code>control</code> we are saying the fold change will be FGF / control. If we had put them in the order <code>control</code>, <code>FGF</code> we would have got the fold change as control / FGF. This means:</p>
<ul>
<li>positive log fold changes indicate FGF &gt; control and</li>
<li>negative log fold changes indicates control &gt; FGF.</li>
</ul>
<p>ğŸ¬ Extract the results from the DESseqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results_fgf</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                       contrast <span class="op">=</span> <span class="va">contrast_fgf</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This will give us the log2 fold change and p-value for the contrast.</p>
<p>ğŸ¬ Save the results to a file:</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">results_fgf</span>,</span>
<span>           xenbase_gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">results_fgf</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"results/S30_results.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="analysis-1" class="level1"><h1>ğŸ­ Analysis</h1>
<p>We will carry out several steps</p>
<ol type="1">
<li>Import the prog and hspc data</li>
<li>Combine the two datasets ready for analysis</li>
<li>Filter the data to remove genes that are not expressed in any cell</li>
<li>Find the genes that are expressed in only one cell type (the prog or the hspc)</li>
<li>Do differential expression analysis on the genes using the <strong><code>scran</code></strong> package. This needs to be done on the logged normalised counts.</li>
</ol>
<section id="import-1" class="level2"><h2 class="anchored" data-anchor-id="import-1">Import</h2>
<p>ğŸ¬ Import surfaceome_hspc.csv and surfaceome_prog.csv into dataframes called <code>hspc</code> and <code>prog</code> respectively.</p>
</section><section id="combine-the-two-datasets" class="level2"><h2 class="anchored" data-anchor-id="combine-the-two-datasets">Combine the two datasets</h2>
<p>We need to combine the two datasets of 701 and 798 cells into one dataset of 1499 cells, i.e., 1499 columns. The number of rows is the number of genes, 280. Before combining, we must make sure genes in the same order in both dataframes or we would be comparing the expression of one gene in one cell type to the expression of a different gene in the other cell type!</p>
<p>ğŸ¬ Check the gene ids are in the same order in both dataframes:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span><span class="op">(</span><span class="va">prog</span><span class="op">$</span><span class="va">ensembl_gene_id</span>, <span class="va">hspc</span><span class="op">$</span><span class="va">ensembl_gene_id</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p><strong><code>scran</code></strong> can use a matrix or a dataframe of counts but theses must be log normalised counts. If using a dataframe, the columns must only contain the expression values (not the gene ids).</p>
<p>ğŸ¬ Combine the two dataframes (minus the gene ids) into one dataframe called <code>prog_hspc</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prog_hspc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html">bind_cols</a></span><span class="op">(</span><span class="va">prog</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>, <span class="va">hspc</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Now add the gene ids as the row names:</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">prog_hspc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">prog</span><span class="op">$</span><span class="va">ensembl_gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="filter-to-remove-unexpressed-genes" class="level2"><h2 class="anchored" data-anchor-id="filter-to-remove-unexpressed-genes">Filter to remove unexpressed genes</h2>
<p>In this dataset, we will not see and genes that are not expressed in any of the cells because we are using a specific subset of the transcriptome that was deliberately selected. However, we will go through how to do this because it is an important step in most analyses.</p>
<p>For the ğŸ¸ frog data you should remember that we were able to filter out our unexpressed genes in <a href="../week-3/workshop.html">Omics 1</a> because we were examining both groups to be compared. In that workshop, <a href="../week-3/workshop.html#frog-filtering">we discussed</a> that we could not filter out unexpressed genes in the ğŸ­ mouse data because we only had one cell types at that time. During the <a href="../week-3/study_after_workshop.html">Consolidate Independent Study</a> you examined the hspc cells.</p>
<p>Where the sum of all the values in the rows is zero, all the entries must be zero. We can use this to find the filter the genes that are not expressed in any of the cells. To do row wise aggregates such as the sum across rows we can use the <code><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise()</a></code> function. <code><a href="https://dplyr.tidyverse.org/reference/c_across.html">c_across()</a></code> allows us to use the colon notation <code>Prog_001:HSPC_852</code> in <code><a href="https://rdrr.io/r/base/sum.html">sum()</a></code> rather than having to list all the column names: <code>sum(Prog_001, Prog_002, Prog_002, Prog_004,.....)</code></p>
<p>ğŸ¬ Find the genes that are 0 in every column of the prog_hspc dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prog_hspc</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/c_across.html">c_across</a></span><span class="op">(</span><span class="va">Prog_001</span><span class="op">:</span><span class="va">HSPC_852</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 0 Ã— 1,499
# Rowwise: 
# â„¹ 1,499 variables: Prog_001 &lt;dbl&gt;, Prog_002 &lt;dbl&gt;, Prog_003 &lt;dbl&gt;,
#   Prog_004 &lt;dbl&gt;, Prog_006 &lt;dbl&gt;, Prog_007 &lt;dbl&gt;, Prog_008 &lt;dbl&gt;,
#   Prog_009 &lt;dbl&gt;, Prog_010 &lt;dbl&gt;, Prog_011 &lt;dbl&gt;, Prog_012 &lt;dbl&gt;,
#   Prog_013 &lt;dbl&gt;, Prog_014 &lt;dbl&gt;, Prog_015 &lt;dbl&gt;, Prog_016 &lt;dbl&gt;,
#   Prog_017 &lt;dbl&gt;, Prog_018 &lt;dbl&gt;, Prog_019 &lt;dbl&gt;, Prog_020 &lt;dbl&gt;,
#   Prog_021 &lt;dbl&gt;, Prog_022 &lt;dbl&gt;, Prog_023 &lt;dbl&gt;, Prog_024 &lt;dbl&gt;,
#   Prog_025 &lt;dbl&gt;, Prog_026 &lt;dbl&gt;, Prog_027 &lt;dbl&gt;, Prog_028 &lt;dbl&gt;, â€¦</code></pre>
</div>
</div>
<p>Notice that we have summed across all the columns.</p>
<p>â“ What do you conclude?</p>
<!-- #---THINKING ANSWER--- -->
<!-- The are no genes that are completely unexpressed in this set of 280 genes -->
<p>We might also examine the genes which are least expressed.</p>
<p>ğŸ¬ Find ten least expressed genes:</p>
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">prog_hspc</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ENSMUSG00000041046 ENSMUSG00000012428 ENSMUSG00000022225 ENSMUSG00000027863 
          30.70322           35.35796           50.45975           61.27461 
ENSMUSG00000019359 ENSMUSG00000020701 ENSMUSG00000030772 ENSMUSG00000027376 
          68.90961           77.95594           84.11234           97.69333 
ENSMUSG00000023132 ENSMUSG00000026285 
         120.43065          126.95425 </code></pre>
</div>
</div>
<p>â“ What do you conclude?</p>
<!-- #---THINKING ANSWER--- -->
<!-- When you consider that there are 1499 cells, a values of 30 are low even  -->
<!-- considering these are already logged and normalised (ie., the range of  -->
<!-- values is less that it would be for raw counts) -->
</section><section id="find-the-genes-that-are-expressed-in-only-one-cell-type" class="level2"><h2 class="anchored" data-anchor-id="find-the-genes-that-are-expressed-in-only-one-cell-type">Find the genes that are expressed in only one cell type</h2>
<p>To find the genes that are expressed in only one cell type, we can use the same approach as above but only sum the columns for one cell type.</p>
<p>ğŸ¬ Find the genes that are 0 in every column for the HSPC cells:</p>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prog_hspc</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/c_across.html">c_across</a></span><span class="op">(</span><span class="va">HSPC_001</span><span class="op">:</span><span class="va">HSPC_852</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 0 Ã— 1,499
# Rowwise: 
# â„¹ 1,499 variables: Prog_001 &lt;dbl&gt;, Prog_002 &lt;dbl&gt;, Prog_003 &lt;dbl&gt;,
#   Prog_004 &lt;dbl&gt;, Prog_006 &lt;dbl&gt;, Prog_007 &lt;dbl&gt;, Prog_008 &lt;dbl&gt;,
#   Prog_009 &lt;dbl&gt;, Prog_010 &lt;dbl&gt;, Prog_011 &lt;dbl&gt;, Prog_012 &lt;dbl&gt;,
#   Prog_013 &lt;dbl&gt;, Prog_014 &lt;dbl&gt;, Prog_015 &lt;dbl&gt;, Prog_016 &lt;dbl&gt;,
#   Prog_017 &lt;dbl&gt;, Prog_018 &lt;dbl&gt;, Prog_019 &lt;dbl&gt;, Prog_020 &lt;dbl&gt;,
#   Prog_021 &lt;dbl&gt;, Prog_022 &lt;dbl&gt;, Prog_023 &lt;dbl&gt;, Prog_024 &lt;dbl&gt;,
#   Prog_025 &lt;dbl&gt;, Prog_026 &lt;dbl&gt;, Prog_027 &lt;dbl&gt;, Prog_028 &lt;dbl&gt;, â€¦</code></pre>
</div>
</div>
<p>We have summed across the HSPC cells only. Note that if we knew there were some rows that were all zero across both cell types, we would need to add <code>|&gt; filter(sum(c_across(Prog_001:Prog_852)) != 0)</code></p>
<p>meaning zero in all the HSPC but not zero in all the Prog</p>
<p>ğŸ¬ Now you find the genes that are 0 in every column for the Prog cells:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 0 Ã— 1,499
# Rowwise: 
# â„¹ 1,499 variables: Prog_001 &lt;dbl&gt;, Prog_002 &lt;dbl&gt;, Prog_003 &lt;dbl&gt;,
#   Prog_004 &lt;dbl&gt;, Prog_006 &lt;dbl&gt;, Prog_007 &lt;dbl&gt;, Prog_008 &lt;dbl&gt;,
#   Prog_009 &lt;dbl&gt;, Prog_010 &lt;dbl&gt;, Prog_011 &lt;dbl&gt;, Prog_012 &lt;dbl&gt;,
#   Prog_013 &lt;dbl&gt;, Prog_014 &lt;dbl&gt;, Prog_015 &lt;dbl&gt;, Prog_016 &lt;dbl&gt;,
#   Prog_017 &lt;dbl&gt;, Prog_018 &lt;dbl&gt;, Prog_019 &lt;dbl&gt;, Prog_020 &lt;dbl&gt;,
#   Prog_021 &lt;dbl&gt;, Prog_022 &lt;dbl&gt;, Prog_023 &lt;dbl&gt;, Prog_024 &lt;dbl&gt;,
#   Prog_025 &lt;dbl&gt;, Prog_026 &lt;dbl&gt;, Prog_027 &lt;dbl&gt;, Prog_028 &lt;dbl&gt;, â€¦</code></pre>
</div>
</div>
<p>â“ What do you conclude?</p>
<!-- #---THINKING ANSWER--- -->
<!-- there are no genes that are expressed in only one cell type -->
</section><section id="differential-expression-analysis-1" class="level2"><h2 class="anchored" data-anchor-id="differential-expression-analysis-1">Differential expression analysis</h2>
<p>Like <strong><code>DESeq2</code></strong>, <strong><code>scran</code></strong> uses a statistical model to calculate the significance of the difference between the treatments and needs metadata to define the treatments.</p>
<p>ğŸ¬ Load the <strong><code>scran</code></strong> package:</p>
<p>The meta data needed for the frog data was information about which columns were in which treatment group and which sibling group and we had that information in a file. Similarly, here we need information on which columns are from which cell type. Instead of having this is a file, we will create a vector that indicates which column belongs to which cell type.</p>
<p>ğŸ¬ Create a vector that indicates which column belongs to which cell type:</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"prog"</span>,<span class="st">"hspc"</span><span class="op">)</span>, </span>
<span>                 times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">prog</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>                           <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">hspc</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The number of times each cell type is repeated is the number of columns in that cell type minus 1. This is because we have removed the column with the gene ids. Do check that the length of the <code>cell_type</code> vector is the same as the number of columns in the <code>prog_hspc</code> dataframe.</p>
<p>ğŸ¬ Run the differential expression analysis:</p>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">res_prog_hspc</span> <span class="op">&lt;-</span> <span class="fu">findMarkers</span><span class="op">(</span><span class="va">prog_hspc</span>, </span>
<span>                             <span class="va">cell_type</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>findMarkers()</code> is the function that runs the differential expression analysis. The first argument is the dataframe containing the data. The second argument is the vector indicating which columns are in which cell type. It gives us two dataframes of the results - rather unnecessarily. One is the results with fold changes that are Prog/HSPC and the other is the results with fold changes that are HSPC/Prog. These have the same magnitude, just a different sign</p>
<p>The dataframe <code>res_prog_hspc$prog</code> is log prog - log hspc (i.e.,Prog/HSPC). This means - Positive fold change: prog is higher than hspc - Negative fold change: hspc is higher than prog</p>
<p>The dataframe <code>res_prog_hspc$hspc</code> is log hspc - log prog (i.e., HSPC/Prog). . This means - Positive fold change: hspc is higher than prog - Negative fold change: prog is higher than hspc</p>
<div class="cell">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>The res_prog_hspc$prog dataframe</caption>
<colgroup>
<col style="width: 24%">
<col style="width: 5%">
<col style="width: 10%">
<col style="width: 5%">
<col style="width: 17%">
<col style="width: 13%">
<col style="width: 24%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Top</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">FDR</th>
<th style="text-align: right;">summary.logFC</th>
<th style="text-align: right;">logFC.hspc</th>
<th style="text-align: left;">ensembl_gene_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000028639</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1.596910</td>
<td style="text-align: right;">1.596910</td>
<td style="text-align: left;">ENSMUSG00000028639</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000024053</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.035165</td>
<td style="text-align: right;">3.035165</td>
<td style="text-align: left;">ENSMUSG00000024053</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000041329</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.261056</td>
<td style="text-align: right;">3.261056</td>
<td style="text-align: left;">ENSMUSG00000041329</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000030336</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-2.146491</td>
<td style="text-align: right;">-2.146491</td>
<td style="text-align: left;">ENSMUSG00000030336</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000016494</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.056730</td>
<td style="text-align: right;">-3.056730</td>
<td style="text-align: left;">ENSMUSG00000016494</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000002808</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.000810</td>
<td style="text-align: right;">3.000810</td>
<td style="text-align: left;">ENSMUSG00000002808</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>The res_prog_hspc$hspc dataframe. Notice the sign of the fold change is the other way</caption>
<colgroup>
<col style="width: 24%">
<col style="width: 5%">
<col style="width: 10%">
<col style="width: 5%">
<col style="width: 17%">
<col style="width: 13%">
<col style="width: 24%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Top</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">FDR</th>
<th style="text-align: right;">summary.logFC</th>
<th style="text-align: right;">logFC.prog</th>
<th style="text-align: left;">ensembl_gene_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000028639</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-1.596910</td>
<td style="text-align: right;">-1.596910</td>
<td style="text-align: left;">ENSMUSG00000028639</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000024053</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.035165</td>
<td style="text-align: right;">-3.035165</td>
<td style="text-align: left;">ENSMUSG00000024053</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000041329</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.261056</td>
<td style="text-align: right;">-3.261056</td>
<td style="text-align: left;">ENSMUSG00000041329</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000030336</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2.146491</td>
<td style="text-align: right;">2.146491</td>
<td style="text-align: left;">ENSMUSG00000030336</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000016494</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.056730</td>
<td style="text-align: right;">3.056730</td>
<td style="text-align: left;">ENSMUSG00000016494</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000002808</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.000810</td>
<td style="text-align: right;">-3.000810</td>
<td style="text-align: left;">ENSMUSG00000002808</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>ğŸ¬ Write the results to file:</p>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">res_prog_hspc</span><span class="op">$</span><span class="va">prog</span>, </span>
<span>           ensembl_gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">res_prog_hspc</span><span class="op">$</span><span class="va">prog</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="st">"results/prog_hspc_results.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="look-after-future-you" class="level1"><h1>ğŸ¤— Look after future you!</h1>
<p>ğŸ¬ Go through you script (<code>cont-fgf-s30.R</code> or <code>hspc-prog.R</code>) and tidy up. You might need to :</p>
<ul>
<li>collect together library statements at the top of the script</li>
<li>remove code that you needed to start today but which wouldnâ€™t be needed running the script from the top (e.g., importing)</li>
<li>edit your comments for clarity</li>
<li>rename variables for consistency or clarity</li>
<li>remove house keeping or exploratory code</li>
<li>restyle code, add code section headers etc</li>
</ul></section><section id="finished" class="level1"><h1>ğŸ¥³ Finished</h1>
<p>Well Done!</p>
</section><section id="independent-study-following-the-workshop" class="level1"><h1>Independent study following the workshop</h1>
<p><a href="../../omics/week-4/study_after_workshop.html">Consolidate</a></p>
</section><section id="the-code-file" class="level1"><h1>The Code file</h1>
<p>These contain all the code needed in the workshop even where it is not visible on the webpage.</p>
<p>The <a href="../../omics/week-4/workshop.html">workshop.qmd</a> file is the file I use to compile the practical. Qmd stands for Quarto markdown. It allows code and ordinary text to be interleaved to produce well-formatted reports including webpages. Right-click on the link and choose Save-As to download. You will be able to open the Qmd file in RStudio. Alternatively, <a href="https://github.com/3mmaRand/BIO00088H-data/blob/main/omics/week-4/workshop.qmd">View in Browser</a>. Coding and thinking answers are marked with <code>#---CODING ANSWER---</code> and <code>#---THINKING ANSWER---</code></p>
<p>Pages made with R <span class="citation" data-cites="R-core">(<a href="#ref-R-core" role="doc-biblioref">R Core Team 2023</a>)</span>, Quarto <span class="citation" data-cites="allaire2022">(<a href="#ref-allaire2022" role="doc-biblioref">Allaire et al. 2022</a>)</span>, <code>knitr</code> <span class="citation" data-cites="knitr">(<a href="#ref-knitr" role="doc-biblioref">Xie 2022</a>)</span>, <code>kableExtra</code> <span class="citation" data-cites="kableExtra">(<a href="#ref-kableExtra" role="doc-biblioref">Zhu 2021</a>)</span></p>
</section><section id="references" class="level1">



</section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-allaire2022" class="csl-entry" role="listitem">
Allaire, J. J., Charles Teague, Carlos Scheidegger, Yihui Xie, and Christophe Dervieux. 2022. <em>Quarto</em>. <a href="https://doi.org/10.5281/zenodo.5960048">https://doi.org/10.5281/zenodo.5960048</a>.
</div>
<div id="ref-fs" class="csl-entry" role="listitem">
Hester, Jim, Hadley Wickham, and GÃ¡bor CsÃ¡rdi. 2023. <span>â€œFs: Cross-Platform File System Operations Based on â€™Libuvâ€™.â€</span>
</div>
<div id="ref-DESeq2" class="csl-entry" role="listitem">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>â€œModerated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2â€</span> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-scran" class="csl-entry" role="listitem">
Lun, Aaron T. L., Davis J. McCarthy, and John C. Marioni. 2016. <span>â€œA Step-by-Step Workflow for Low-Level Analysis of Single-Cell RNA-Seq Data with Bioconductorâ€</span> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-R-core" class="csl-entry" role="listitem">
R Core Team. 2023. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-conflicted" class="csl-entry" role="listitem">
Wickham, Hadley. 2023. <span>â€œConflicted: An Alternative Conflict Resolution Strategy.â€</span>
</div>
<div id="ref-tidyverse" class="csl-entry" role="listitem">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy Dâ€™Agostino McGowan, Romain FranÃ§ois, Garrett Grolemund, et al. 2019. <span>â€œWelcome to the <span></span>Tidyverse<span></span>â€</span> 4: 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-knitr" class="csl-entry" role="listitem">
Xie, Yihui. 2022. <span>â€œKnitr: A General-Purpose Package for Dynamic Report Generation in r.â€</span> <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-kableExtra" class="csl-entry" role="listitem">
Zhu, Hao. 2021. <span>â€œkableExtra: Construct Complex Table with â€™Kableâ€™ and Pipe Syntax.â€</span> <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>
<ol>
<li id="fn1"><p><a href="https://www.bioconductor.org/">Bioconductor</a> is a project that develops and supports R packages for bioinformatics.<a href="#fnref1" class="footnote-back" role="doc-backlink">â†©ï¸</a></p></li>
</ol></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("^(?:http:|https:)\/\/3mmarand\.github\.io\/BIO00088H-data");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Rand E (2023). Data Analysis for Group Project. <a href="https://3mmarand.github.io/BIO00088H-data/" class="uri">https://3mmarand.github.io/BIO00088H-data/</a>.</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/er_13">
      <i class="bi bi-twitter" role="img" aria-label="Twitter">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/3mmaRand/BIO00088H-data">
      <i class="bi bi-github" role="img" aria-label="BIO00088H-data GitHub">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>