<!DOCTYPE html>
<html lang="en"><head>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">
<meta name="author" content="Emma Rand">
<title>Data Analysis for Group Project ‚Äì Independent Study to prepare for workshop</title>
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
<link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
<link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
<style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
<link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-bbe7401fe57d4b791b917637bb662036.css">
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-TVF95J593T"></script><script type="text/javascript">

  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-TVF95J593T', { 'anonymize_ip': true});
  </script><link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
<link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
<link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
<link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
<link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
<link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
<style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center"><h1 class="title">Independent Study to prepare for workshop</h1>
  <p class="subtitle">Transcriptomics 2: Statistical Analysis</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Emma Rand 
</div>
</div>
</div>

  <p class="date">17 December, 2024</p>
</section><section id="overview" class="slide level2"><h2>Overview</h2>
<p>In these slides we will:</p>
<div>
<ul>
<li class="fragment"><p>Check where you are following week 3</p></li>
<li class="fragment">
<p>learn some concepts in differential expression</p>
<ul>
<li class="fragment">log<sub>2</sub> fold changes</li>
<li class="fragment">Multiple correction</li>
<li class="fragment">normalisation</li>
<li class="fragment">statistical model</li>
</ul>
</li>
<li class="fragment"><p>Find out what packages we will use</p></li>
</ul>
</div>
</section><section><section id="where-should-you-be" class="title-slide slide level1 center"><h1>Where should you be?</h1>

</section><section id="what-we-did-in-transcriptomics-1-hello-data" class="slide level2"><h2>What we did in Transcriptomics 1: üëã Hello data!</h2>
<div>
<div style="font-size: 90%;">
<ul>
<li class="fragment">Discovered how many rows and columns we had in our datasets and what these were.</li>
<li class="fragment">Examined the distribution of values
<ul>
<li class="fragment">across the whole dataset</li>
<li class="fragment">across the samples/cells (i.e., averaged over genes) to see variation between samples/cells</li>
<li class="fragment">across the genes (i.e., averaged over samples/cells) to see variation between genes</li>
</ul>
</li>
<li class="fragment">Filtered data for quality control and wrote to file (except üê≠)</li>
</ul>
</div>
</div>
</section><section id="where-should-you-be-1" class="slide level2"><h2>Where should you be?</h2>
<p>After the <a href="../week-3/workshop.html">Transcriptomics 1: üëã Hello data! Workshop</a> including:</p>
<ul>
<li><p><a href="../week-3/workshop.html#look-after-future-you">ü§ó Look after future you!</a> and</p></li>
<li><p>the <a href="../week-3/study_after_workshop.html">Independent Study to consolidate</a>, you should have:</p></li>
</ul></section><section id="frog-development" class="slide level2"><h2>üê∏ Frog development</h2>
<ul>
<li>
<p>An RStudio Project called <code>frogs-88H</code> which contains:</p>
<ul>
<li>data-raw: <code>xlaevis_counts_S14.csv</code>, <code>xlaevis_counts_S20.csv</code>, <code>xlaevis_counts_S30.csv</code>
</li>
<li>data-processed: <code>s30_filtered.csv</code>, <code>s20_filtered.csv</code>
</li>
<li>Two scripts: <code>cont-fgf-s30.R</code>, <code>cont-fgf-s20.R</code>
</li>
</ul>
</li>
</ul></section><section id="arabidopsis" class="slide level2"><h2>üéÑ <em>Arabidopsis</em>
</h2>
<ul>
<li>
<p>An RStudio Project called <code>arab-88H</code> which contains:</p>
<ul>
<li>data-raw: <code>arabidopsis-wild.csv</code>, <code>arabidopsis-spl7.csv</code>
</li>
<li>data-processed: <code>wild_filtered.csv</code>, <code>spl7_filtered.csv</code>
</li>
<li>Two scripts: <code>suff-def-wild.R</code>, <code>suff-def-spl7.R</code>
</li>
</ul>
</li>
</ul></section><section id="leishmania" class="slide level2"><h2>üíâ <em>Leishmania</em>
</h2>
<ul>
<li>
<p>An RStudio Project called <code>leish-88H</code> which contains:</p>
<ul>
<li>data-raw: <code>leishmania-mex-ama.csv</code>, <code>leishmania-mex-pro.csv</code>, <code>leishmania-mex-meta.csv</code>
</li>
<li>data-processed: <code>pro_meta_filtered.csv</code>, <code>pro_ama_filtered.csv</code>
</li>
<li>Two scripts: <code>pro_meta.R</code>, <code>pro_ama.R</code>
</li>
</ul>
</li>
</ul></section><section id="stem-cells" class="slide level2"><h2>üê≠ Stem cells</h2>
<ul>
<li>
<p>An RStudio Project called <code>mice-88H</code> which contains</p>
<ul>
<li>data-raw: <code>surfaceome_hspc.csv</code>, <code>surfaceome_prog.csv</code>, <code>surfaceome_lthsc.csv</code>
</li>
<li>data-processed: <code>hspc_prog.csv</code>, <code>hspc_lthsc.csv</code>
</li>
<li>Two scripts: <code>hspc-prog.R</code>, <code>hspc-lthsc.R</code>
</li>
</ul>
</li>
</ul></section><section id="additionally" class="slide level2"><h2>Additionally‚Ä¶</h2>
<p>Files should be organised into folders. Code should well commented and easy to read. You should have curated your code to remove unnecessary commands that were useful to troubleshoot or understand objects in your environment but which are not needed for the final analysis.</p>
<p>If you are missing files, go through:</p>
<ul>
<li><p><a href="../week-3/workshop.html">Transcriptomics 1: üëã Hello data! Workshop</a> including:</p></li>
<li><p><a href="../week-3/workshop.html#look-after-future-you">ü§ó Look after future you!</a> and</p></li>
<li><p>the <a href="../week-3/study_after_workshop.html">Independent Study to consolidate</a></p></li>
</ul></section></section><section><section id="differential-expression" class="title-slide slide level1 center"><h1>Differential expression</h1>

</section><section id="differential-expression-1" class="slide level2"><h2>Differential expression</h2>
<div>
<ul>
<li class="fragment"><p>The goal of differential expression is to test whether there is a significant difference in gene expression between groups.</p></li>
<li class="fragment"><p>A large number of computational methods have been developed for differential expression analysis</p></li>
<li class="fragment"><p>R is the leading language for differential expression analysis</p></li>
</ul>
</div>
</section><section id="differential-expression-2" class="slide level2"><h2>Differential expression</h2>
<div>
<ul>
<li class="fragment"><p>the statistical concepts are very similar to those you have already encountered in stages 1 and 2</p></li>
<li class="fragment"><p>you are essentially doing paired- or independent-samples tests</p></li>
<li class="fragment"><p>but you are doing a lot of them! One for every gene</p></li>
<li class="fragment"><p>data need normalisation before comparison</p></li>
</ul>
</div>
</section><section id="statistical-concepts" class="slide level2"><h2>Statistical concepts</h2>
<p>Like familiar tests:</p>
<div>
<ul>
<li class="fragment"><p>the type of test (the function) you use depends on the type of data you have and the type of assumptions you want to make</p></li>
<li class="fragment"><p>the tests work by comparing the variation between groups to the variation within groups.</p></li>
<li class="fragment"><p>you will get: the difference between groups, a test statistic, and a <em>p</em>-value</p></li>
<li class="fragment"><p>you also get an <strong>adjusted <em>p</em>-value</strong> which is the ‚Äòcorrection‚Äô for multiple testing</p></li>
</ul>
</div>
</section><section id="the-difference-between-groups" class="slide level2"><h2>The difference between groups</h2>
<div>
<ul>
<li class="fragment"><p>The difference between groups is given as the log<sub>2</sub> fold change in expression between groups</p></li>
<li class="fragment"><p>A fold change is the expression in one group divided by the expression in the other group: <span class="math inline">\(\frac{A}{B}\)</span></p></li>
<li class="fragment"><p>we use fold changes because the absolute expression values may not be accurate and relative changes are what matters</p></li>
<li class="fragment"><p>we use log<sub>2</sub> fold changes because they are symmetrical around 0</p></li>
</ul>
</div>
</section><section id="why-log2-fold-change" class="slide level2"><h2>Why log<sub>2</sub> fold change?</h2>
<div>
<ul>
<li class="fragment"><p>log<sub>2</sub> means log to the base 2</p></li>
<li class="fragment"><p>Suppose the expression in group A is 5 and the expression in group B is 8</p></li>
<li class="fragment"><p><span class="math inline">\(\frac{A}{B} = \frac{5}{8}\)</span> = 0.625 and <span class="math inline">\(\frac{B}{A} = \frac{8}{5}\)</span> = 1.6</p></li>
<li class="fragment"><p>If B &gt; A the range of <span class="math inline">\(\frac{A}{B}\)</span> is 0 - 1 but the range of <span class="math inline">\(\frac{B}{A}\)</span> is 1 - <span class="math inline">\(\infty\)</span></p></li>
<li class="fragment"><p>However, if we take the log<sub>2</sub> of <span class="math inline">\(\frac{A}{B}\)</span> we get -0.678 and the log<sub>2</sub> of <span class="math inline">\(\frac{B}{A}\)</span> is 0.678.</p></li>
</ul>
</div>
</section><section id="adjusted-p-value" class="slide level2"><h2>Adjusted <em>p</em>-value</h2>
<div>
<ul>
<li class="fragment"><p>The <em>p</em>-value has to be adjusted because of the number of tested being done</p></li>
<li class="fragment"><p>In stage 1, we used Tukey‚Äôs HSD to adjust for multiple testing following an ANOVA</p></li>
<li class="fragment"><p>Here the Benjamini-Hochberg procedure <span class="citation" data-cites="benjamini1995">(<a href="#/references" role="doc-biblioref" onclick="">Benjamini and Hochberg 1995</a>)</span> is used to adjust for multiple testing</p></li>
<li class="fragment"><p>BH controls the <a href="https://en.wikipedia.org/wiki/False_discovery_rate">False Discovery Rate</a> (FDR)</p></li>
<li class="fragment"><p>The FDR is the proportion of false positives among the genes called significant</p></li>
</ul>
</div>
</section><section id="normalisation" class="slide level2"><h2>Normalisation</h2>
<div>
<ul>
<li class="fragment"><p>Normalisation adjusts raw counts to account for factors that prevent direct comparisons</p></li>
<li class="fragment"><p>Normalisation usually influences the experimental design as well as the analysis</p></li>
</ul>
</div>
</section><section id="normalisation-1" class="slide level2"><h2>Normalisation</h2>
<div>
<ul>
<li class="fragment"><p>üê≠ mice data are normalised</p></li>
<li class="fragment"><p>üê∏ frog, üéÑ Arabidopisis and üíâ Leishmania data are raw counts (not normalised) because the differential expression method will do this.</p></li>
<li class="fragment"><p>Normalisation is a big topic. See <span class="citation" data-cites="d√ºren2022">D√ºren, Lederer, and Qin (<a href="#/references" role="doc-biblioref" onclick="">2022</a>)</span>; <span class="citation" data-cites="bullard2010">Bullard et al. (<a href="#/references" role="doc-biblioref" onclick="">2010</a>)</span>; <span class="citation" data-cites="lytal2020">Lytal, Ran, and An (<a href="#/references" role="doc-biblioref" onclick="">2020</a>)</span>; <span class="citation" data-cites="abrams2019">Abrams et al. (<a href="#/references" role="doc-biblioref" onclick="">2019</a>)</span>; <span class="citation" data-cites="vallejos2017">Vallejos et al. (<a href="#/references" role="doc-biblioref" onclick="">2017</a>)</span>; <span class="citation" data-cites="evans2017">Evans, Hardin, and Stoebel (<a href="#/references" role="doc-biblioref" onclick="">2017</a>)</span></p></li>
</ul>
</div>
</section><section id="type-of-de-tests" class="slide level2"><h2>Type of DE tests</h2>
<div>
<ul>
<li class="fragment"><p>A large number of computational methods have been developed for differential expression analysis</p></li>
<li class="fragment"><p>Methods vary in the types of normalisation they do, the statistical model they use, and the assumptions they make</p></li>
<li class="fragment"><p>Some of the most well-known methods are provided by: <strong><code>DESeq2</code></strong> <span class="citation" data-cites="DESeq2">(<a href="#/references" role="doc-biblioref" onclick="">Love, Huber, and Anders 2014</a>)</span>, <strong><code>edgeR</code></strong> <span class="citation" data-cites="robinson2010 mccarthy2012 chen2016">(<a href="#/references" role="doc-biblioref" onclick="">Robinson, McCarthy, and Smyth 2010</a>; <a href="#/references" role="doc-biblioref" onclick="">McCarthy, Chen, and Smyth 2012</a>; <a href="#/references" role="doc-biblioref" onclick="">Chen, Lun, and Smyth 2016</a>)</span>, <strong><code>limma</code></strong> <span class="citation" data-cites="ritchie2015">(<a href="#/references" role="doc-biblioref" onclick="">Ritchie et al. 2015</a>)</span> and <strong><code>scran</code></strong> <span class="citation" data-cites="scran">(<a href="#/references" role="doc-biblioref" onclick="">Lun, McCarthy, and Marioni 2016</a>)</span></p></li>
</ul>
</div>
</section><section id="type-of-test-the-function" class="slide level2"><h2>Type of test (the function)</h2>
<div>
<ul>
<li class="fragment">
<strong><code>DESeq2</code></strong> and <strong><code>edgeR</code></strong>
<ul>
<li class="fragment">both require raw counts as input</li>
<li class="fragment">both assume that most genes are not DE</li>
<li class="fragment">both use a negative binomial distribution<sup>1</sup> to model the data</li>
<li class="fragment">use slightly different normalisation methods: <strong><code>DESeq2</code></strong> uses the median of ratios method; <strong><code>edgeR</code></strong> uses the trimmed mean of M values (TMM) method</li>
</ul>
</li>
</ul>
</div>
<aside><ol class="aside-footnotes"><li id="fn1"><p>A discrete distribution for counts, similar to the Poisson distribution</p></li></ol></aside></section><section id="type-of-test-the-function-1" class="slide level2"><h2>Type of test (the function)</h2>
<ul>
<li>
<strong><code>scran</code></strong>
<ul>
<li>works on normalized log-expression values</li>
<li>performs Welch t-tests</li>
</ul>
</li>
</ul></section><section id="meta-data" class="slide level2"><h2>Meta data</h2>
<div>
<ul>
<li class="fragment"><p>DE methods require two types of data: the expression data and the meta data</p></li>
<li class="fragment"><p>The meta data gives the information about the samples</p></li>
<li class="fragment"><p>It says which samples (which columns of data) are in which treatment group (s)</p></li>
<li class="fragment"><p>Meta data is usually stored in a separate file</p></li>
</ul>
</div>
</section><section id="frog-development-1" class="slide level2"><h2>üê∏ Frog development</h2>
<ul>
<li><p>Expression for the whole transcriptome <a href="https://www.xenbase.org/xenbase/static-xenbase/ftpDatafiles.jsp"><em>X. laevis</em> v10.1 genome assembly</a></p></li>
<li><p>Values are raw counts</p></li>
<li><p>The statistical analysis method we will use <strong><code>DESeq2</code></strong> <span class="citation" data-cites="DESeq2">(<a href="#/references" role="doc-biblioref" onclick="">Love, Huber, and Anders 2014</a>)</span> requires raw counts and performs the normalisation itself</p></li>
</ul></section><section id="arabidopisis" class="slide level2"><h2>üéÑ <em>Arabidopisis</em>
</h2>
<ul>
<li><p>Expression for the whole transcriptome <a href="https://plants.ensembl.org/Arabidopsis_thaliana/Info/Index">ENSEMBL <em>Arabidopsis</em> TAIR10</a><span class="citation" data-cites="yates_ensembl_2022">(<a href="#/references" role="doc-biblioref" onclick="">Yates et al. 2022</a>)</span></p></li>
<li><p>Values are raw counts</p></li>
<li><p>The statistical analysis method we will use <strong><code>DESeq2</code></strong> <span class="citation" data-cites="DESeq2">(<a href="#/references" role="doc-biblioref" onclick="">Love, Huber, and Anders 2014</a>)</span> requires raw counts and performs the normalisation itself</p></li>
</ul></section><section id="leishmania-1" class="slide level2"><h2>üíâ <em>Leishmania</em>
</h2>
<ul>
<li><p>Expression for the whole transcriptome <a href="https://tritrypdb.org/tritrypdb/app/record/dataset/DS_e318592c8e"><em>L. mexicana</em> MHOM/GT/2001/U1103</a><span class="citation" data-cites="rogers_chromosome_2011">(<a href="#/references" role="doc-biblioref" onclick="">Rogers et al. 2011</a>)</span></p></li>
<li><p>Values are raw counts</p></li>
<li><p>The statistical analysis method we will use <strong><code>DESeq2</code></strong> <span class="citation" data-cites="DESeq2">(<a href="#/references" role="doc-biblioref" onclick="">Love, Huber, and Anders 2014</a>)</span> requires raw counts and performs the normalisation itself</p></li>
</ul></section><section id="stem-cells-1" class="slide level2"><h2>üê≠ Stem cells</h2>
<ul>
<li><p>Expression for a <em>subset</em> of the transcriptome, the surfaceome</p></li>
<li><p>Values are log2 normalised values</p></li>
<li><p>The statistical analysis method we will use <strong><code>scran</code></strong> <span class="citation" data-cites="scran">(<a href="#/references" role="doc-biblioref" onclick="">Lun, McCarthy, and Marioni 2016</a>)</span> requires normalised values</p></li>
</ul></section></section><section><section id="adding-gene-information" class="title-slide slide level1 center"><h1>Adding gene information</h1>

</section><section id="adding-gene-information-1" class="slide level2"><h2>Adding gene information</h2>
<div>
<ul>
<li class="fragment"><p>The gene id is difficult to interpret</p></li>
<li class="fragment"><p>Therefore we need to add information such as the gene name and a description to the results</p></li>
</ul>
<div style="font-size: 70%;">
<ul>
<li class="fragment"><p>üê∏ Frog data information comes from Xenbase <span class="citation" data-cites="fisher2023">(<a href="#/references" role="doc-biblioref" onclick="">Fisher et al. 2023</a>)</span></p></li>
<li class="fragment"><p>üéÑ <em>Arabidopisis</em> information comes from TAIR10 <span class="citation" data-cites="yates_ensembl_2022">(<a href="#/references" role="doc-biblioref" onclick="">Yates et al. 2022</a>)</span></p></li>
<li class="fragment"><p>üíâ <em>Leishmania</em> information comes TriTrypDB <span class="citation" data-cites="rogers_chromosome_2011">(<a href="#/references" role="doc-biblioref" onclick="">Rogers et al. 2011</a>)</span></p></li>
<li class="fragment"><p>üê≠ Stem cell information comes from Ensembl <span class="citation" data-cites="birney2004">(<a href="#/references" role="doc-biblioref" onclick="">Birney et al. 2004</a>)</span></p></li>
</ul>
</div>
</div>
</section><section id="xenbase" class="slide level2"><h2>üê∏ Xenbase</h2>

<img data-src="images/Xenbase-Logo-Medium.png" width="800" class="r-stretch quarto-figure-center"><p class="caption">xenbase logo</p><p><a href="http://www.xenbase.org/">Xenbase</a> is a model organism database that provides genomic, molecular, and developmental biology information about <em>Xenopus laevis</em> and <em>Xenopus tropicalis</em>.</p>
<div class="fragment">
<p>It took me some time to find the information you need.</p>
</div>
</section><section id="xenbase-1" class="slide level2"><h2>üê∏ Xenbase</h2>
<div>
<ul>
<li class="fragment"><p>I got the information from the <a href="https://www.xenbase.org/xenbase/static-xenbase/ftpDatafiles.jsp">Xenbase information pages</a> under Data Reports | Gene Information</p></li>
<li class="fragment"><p>This is listed: Xenbase Gene Product Information [readme] <a href="https://download.xenbase.org/xenbase/GenePageReports/xenbase.gpi.gz">gzipped gpi (tab separated)</a></p></li>
<li class="fragment"><p>Click on the readme link to see the file format and columns</p></li>
<li class="fragment"><p>I downloaded <a href="https://download.xenbase.org/xenbase/GenePageReports/xenbase.gpi.gz">xenbase.gpi.gz</a>, unzipped it, removed header lines and the <em>Xenopus tropicalis</em> (taxon:8364) entries and saved it as <a href="meta/xenbase_info.xlsx">xenbase_info.xlsx</a></p></li>
<li class="fragment"><p>In the workshop you will import this file and merge the information with the results file</p></li>
</ul>
</div>
</section><section id="tair10-through-ensembl" class="slide level2"><h2>üéÑ TAIR10 through Ensembl</h2>
<div>
<ul>
<li class="fragment"><p><a href="https://www.ensembl.org/index.html">Ensembl</a> creates, integrates and distributes reference datasets and analysis tools that enable genomics</p></li>
<li class="fragment"><p><a href="https://grch37.ensembl.org/info/data/biomart/index.html">BioMart</a> <span class="citation" data-cites="smedley2009">(<a href="#/references" role="doc-biblioref" onclick="">Smedley et al. 2009</a>)</span> provides uniform access to these large datasets</p></li>
<li class="fragment"><p><strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#/references" role="doc-biblioref" onclick="">Durinck et al. 2009</a>, <a href="#/references" role="doc-biblioref" onclick="">2005</a>)</span> is a Bioconductor package gives you programmatic access to BioMart.</p></li>
<li class="fragment"><p>In the workshop you use this package to get information you can merge with the results file</p></li>
</ul>
</div>
</section><section id="tritrypdb" class="slide level2"><h2>üíâ TriTrypDB</h2>
<div>
<div style="font-size: 80%;">
<ul>
<li class="fragment"><p>I got the information from <a href="https://tritrypdb.org/tritrypdb/app/">TriTrypDB</a><br>
which is a functional genomic resource for the Trypanosomatidae and Plasmodidae</p></li>
<li class="fragment"><p><a href="https://tritrypdb.org/tritrypdb/app/downloads">https://tritrypdb.org/tritrypdb/app/downloads</a> section</p></li>
<li class="fragment"><p>I downloaded the <a href="https://tritrypdb.org/common/downloads/release-68/LmexicanaMHOMGT2001U1103/gff/data/TriTrypDB-68_LmexicanaMHOMGT2001U1103.gff"><em>L. mexicana</em> MHOM/GT/2001/U1103 Full GFF</a> and extracted the gene information and saved it as <a href="meta/leishmania_mex.xlsx">leishmania_mex.xlsx</a></p></li>
<li class="fragment"><p>In the workshop you will import this file and merge the information with the results file</p></li>
</ul>
</div>
</div>
</section><section id="ensembl" class="slide level2"><h2>üê≠ Ensembl</h2>
<div>
<ul>
<li class="fragment"><p><a href="https://www.ensembl.org/index.html">Ensembl</a> creates, integrates and distributes reference datasets and analysis tools that enable genomics</p></li>
<li class="fragment"><p><a href="https://grch37.ensembl.org/info/data/biomart/index.html">BioMart</a> <span class="citation" data-cites="smedley2009">(<a href="#/references" role="doc-biblioref" onclick="">Smedley et al. 2009</a>)</span> provides uniform access to these large datasets</p></li>
<li class="fragment"><p><strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#/references" role="doc-biblioref" onclick="">Durinck et al. 2009</a>, <a href="#/references" role="doc-biblioref" onclick="">2005</a>)</span> is a Bioconductor package gives you programmatic access to BioMart.</p></li>
<li class="fragment"><p>In the workshop you use this package to get information you can merge with the results file</p></li>
</ul>
</div>
</section><section id="packages" class="slide level2"><h2>Packages</h2>
<p>These packages are all on the University computers which you can access on campus or <a href="https://www.york.ac.uk/it-services/tools/vds/">remotely using the VDS</a></p>
<p>If you want to use your own machine you will need to install the packages.</p>
<div style="font-size: 60%;">
<p>Install <strong><code>BiocManager</code></strong> from CRAN in the the normal way and set the version of Bioconductor packages to install:</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.19"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Install <strong><code>DESeq2</code></strong> from Bioconductor using BiocManager:</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"DESeq2"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Install <strong><code>scran</code></strong> from Bioconductor using BiocManager:</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"scran"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Install <strong><code>biomaRt</code></strong> from Bioconductor using BiocManager:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section></section><section><section id="workshops" class="title-slide slide level1 center"><h1>Workshops</h1>

</section><section id="workshops-1" class="slide level2"><h2>Workshops</h2>
<ul>
<li><p>Transcriptomics 1: Hello data. Getting to know the data. Checking the distributions of values overall, across rows and columns to check things are as we expect and detect rows/columns that need to be removed</p></li>
<li><p>Transcriptomics 2: Statistical Analysis. Identifying which genes are differentially expressed between treatments. This is the main analysis step. We will use different methods for bulk and single cell data.</p></li>
<li><p>Transcriptomics 3: Visualising. Principal Component Analysis (PCA) volcano plots to visualise the results of the</p></li>
</ul></section><section id="references" class="slide level2 smaller scrollable"><h2>References</h2>
<p>Pages made with R <span class="citation" data-cites="R-core">(<a href="#/references" role="doc-biblioref" onclick="">R Core Team 2024</a>)</span>, Quarto <span class="citation" data-cites="Allaire_Quarto_2024">(<a href="#/references" role="doc-biblioref" onclick="">Allaire et al. 2024</a>)</span>, <code>knitr</code> <span class="citation" data-cites="knitr1 knitr2 knitr3">(<a href="#/references" role="doc-biblioref" onclick="">Xie 2024</a>, <a href="#/references" role="doc-biblioref" onclick="">2015</a>, <a href="#/references" role="doc-biblioref" onclick="">2014</a>)</span>, <code>kableExtra</code> <span class="citation" data-cites="kableExtra">(<a href="#/references" role="doc-biblioref" onclick="">Zhu 2021</a>)</span></p>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-abrams2019" class="csl-entry" role="listitem">
Abrams, Zachary B., Travis S. Johnson, Kun Huang, Philip R. O. Payne, and Kevin Coombes. 2019. <span>‚ÄúA Protocol to Evaluate RNA Sequencing Normalization Methods.‚Äù</span> <em>BMC Bioinformatics</em> 20 (24): 679. <a href="https://doi.org/10.1186/s12859-019-3247-x">https://doi.org/10.1186/s12859-019-3247-x</a>.
</div>
<div id="ref-Allaire_Quarto_2024" class="csl-entry" role="listitem">
Allaire, J. J., Charles Teague, Carlos Scheidegger, Yihui Xie, and Christophe Dervieux. 2024. <span>‚Äú<span>Quarto</span>.‚Äù</span> <a href="https://doi.org/10.5281/zenodo.5960048">https://doi.org/10.5281/zenodo.5960048</a>.
</div>
<div id="ref-benjamini1995" class="csl-entry" role="listitem">
Benjamini, Yoav, and Yosef Hochberg. 1995. <span>‚ÄúControlling the False Discovery Rate: A Practical and Powerful Approach to Multiple Testing.‚Äù</span> <em>J. R. Stat. Soc. Series B Stat. Methodol.</em> 57 (1): 289‚Äì300. <a href="http://www.jstor.org/stable/2346101">http://www.jstor.org/stable/2346101</a>.
</div>
<div id="ref-birney2004" class="csl-entry" role="listitem">
Birney, Ewan, T. Daniel Andrews, Paul Bevan, Mario Caccamo, Yuan Chen, Laura Clarke, Guy Coates, et al. 2004. <span>‚ÄúAn Overview of Ensembl.‚Äù</span> <em>Genome Research</em> 14 (5): 925‚Äì28. <a href="https://doi.org/10.1101/gr.1860604">https://doi.org/10.1101/gr.1860604</a>.
</div>
<div id="ref-bullard2010" class="csl-entry" role="listitem">
Bullard, James H., Elizabeth Purdom, Kasper D. Hansen, and Sandrine Dudoit. 2010. <span>‚ÄúEvaluation of Statistical Methods for Normalization and Differential Expression in mRNA-Seq Experiments.‚Äù</span> <em>BMC Bioinformatics</em> 11 (1): 94. <a href="https://doi.org/10.1186/1471-2105-11-94">https://doi.org/10.1186/1471-2105-11-94</a>.
</div>
<div id="ref-chen2016" class="csl-entry" role="listitem">
Chen, Yunshun, Aaron T. L. Lun, and Gordon K. Smyth. 2016. <span>‚ÄúFrom Reads to Genes to Pathways: Differential Expression Analysis of RNA-Seq Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.‚Äù</span> <a href="https://doi.org/10.12688/f1000research.8987.2">https://doi.org/10.12688/f1000research.8987.2</a>.
</div>
<div id="ref-d√ºren2022" class="csl-entry" role="listitem">
D√ºren, Yannick, Johannes Lederer, and Li-Xuan Qin. 2022. <span>‚ÄúDepth Normalization of Small RNA Sequencing: Using Data and Biology to Select a Suitable Method.‚Äù</span> <em>Nucleic Acids Research</em> 50 (10): e56. <a href="https://doi.org/10.1093/nar/gkac064">https://doi.org/10.1093/nar/gkac064</a>.
</div>
<div id="ref-biomaRt2" class="csl-entry" role="listitem">
Durinck, Steffen, Yves Moreau, Arek Kasprzyk, Sean Davis, Bart De Moor, Alvis Brazma, and Wolfgang Huber. 2005. <span>‚ÄúBioMart and Bioconductor: A Powerful Link Between Biological Databases and Microarray Data Analysis.‚Äù</span> <em>Bioinformatics</em> 21: 3439‚Äì40.
</div>
<div id="ref-biomaRt1" class="csl-entry" role="listitem">
Durinck, Steffen, Paul T. Spellman, Ewan Birney, and Wolfgang Huber. 2009. <span>‚ÄúMapping Identifiers for the Integration of Genomic Datasets with the r/Bioconductor Package biomaRt.‚Äù</span> <em>Nature Protocols</em> 4: 1184‚Äì91.
</div>
<div id="ref-evans2017" class="csl-entry" role="listitem">
Evans, Ciaran, Johanna Hardin, and Daniel M Stoebel. 2017. <span>‚ÄúSelecting Between-Sample RNA-Seq Normalization Methods from the Perspective of Their Assumptions.‚Äù</span> <em>Briefings in Bioinformatics</em> 19 (5): 776‚Äì92. <a href="https://doi.org/10.1093/bib/bbx008">https://doi.org/10.1093/bib/bbx008</a>.
</div>
<div id="ref-fisher2023" class="csl-entry" role="listitem">
Fisher, Malcolm, Christina James-Zorn, Virgilio Ponferrada, Andrew J Bell, Nivitha Sundararaj, Erik Segerdell, Praneet Chaturvedi, et al. 2023. <span>‚ÄúXenbase: Key Features and Resources of the Xenopus Model Organism Knowledgebase.‚Äù</span> <em>Genetics</em> 224 (1): iyad018. <a href="https://doi.org/10.1093/genetics/iyad018">https://doi.org/10.1093/genetics/iyad018</a>.
</div>
<div id="ref-DESeq2" class="csl-entry" role="listitem">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>‚ÄúModerated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.‚Äù</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-scran" class="csl-entry" role="listitem">
Lun, Aaron T. L., Davis J. McCarthy, and John C. Marioni. 2016. <span>‚ÄúA Step-by-Step Workflow for Low-Level Analysis of Single-Cell RNA-Seq Data with Bioconductor.‚Äù</span> <em>F1000Res.</em> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-lytal2020" class="csl-entry" role="listitem">
Lytal, Nicholas, Di Ran, and Lingling An. 2020. <span>‚ÄúNormalization Methods on Single-Cell RNA-Seq Data: An Empirical Survey.‚Äù</span> <em>Frontiers in Genetics</em> 11. <a href="https://www.frontiersin.org/articles/10.3389/fgene.2020.00041">https://www.frontiersin.org/articles/10.3389/fgene.2020.00041</a>.
</div>
<div id="ref-mccarthy2012" class="csl-entry" role="listitem">
McCarthy, Davis J., Yunshun Chen, and Gordon K. Smyth. 2012. <span>‚ÄúDifferential Expression Analysis of Multifactor RNA-Seq Experiments with Respect to Biological Variation.‚Äù</span> <em>Nucleic Acids Research</em> 40 (10): 4288‚Äì97. <a href="https://doi.org/10.1093/nar/gks042">https://doi.org/10.1093/nar/gks042</a>.
</div>
<div id="ref-R-core" class="csl-entry" role="listitem">
R Core Team. 2024. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-ritchie2015" class="csl-entry" role="listitem">
Ritchie, Matthew E., Belinda Phipson, Di Wu, Yifang Hu, Charity W. Law, Wei Shi, and Gordon K. Smyth. 2015. <span>‚ÄúLimma Powers Differential Expression Analyses for RNA-Sequencing and Microarray Studies.‚Äù</span> <em>Nucleic Acids Research</em> 43 (7): e47. <a href="https://doi.org/10.1093/nar/gkv007">https://doi.org/10.1093/nar/gkv007</a>.
</div>
<div id="ref-robinson2010" class="csl-entry" role="listitem">
Robinson, Mark D., Davis J. McCarthy, and Gordon K. Smyth. 2010. <span>‚ÄúedgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.‚Äù</span> <em>Bioinformatics</em> 26 (1): 139‚Äì40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-rogers_chromosome_2011" class="csl-entry" role="listitem">
Rogers, Matthew B., James D. Hilley, Nicholas J. Dickens, Jon Wilkes, Paul A. Bates, Daniel P. Depledge, David Harris, et al. 2011. <span>‚ÄúChromosome and Gene Copy Number Variation Allow Major Structural Change Between Species and Strains of <span>Leishmania</span>.‚Äù</span> <em>Genome Research</em> 21 (12): 2129‚Äì42. <a href="https://doi.org/10.1101/gr.122945.111">https://doi.org/10.1101/gr.122945.111</a>.
</div>
<div id="ref-smedley2009" class="csl-entry" role="listitem">
Smedley, Damian, Syed Haider, Benoit Ballester, Richard Holland, Darin London, Gudmundur Thorisson, and Arek Kasprzyk. 2009. <span>‚ÄúBioMart <span></span> Biological Queries Made Easy.‚Äù</span> <em>BMC Genomics</em> 10 (1): 22. <a href="https://doi.org/10.1186/1471-2164-10-22">https://doi.org/10.1186/1471-2164-10-22</a>.
</div>
<div id="ref-vallejos2017" class="csl-entry" role="listitem">
Vallejos, Catalina A., Davide Risso, Antonio Scialdone, Sandrine Dudoit, and John C. Marioni. 2017. <span>‚ÄúNormalizing Single-Cell RNA Sequencing Data: Challenges and Opportunities.‚Äù</span> <em>Nature Methods</em> 14 (6): 565‚Äì71. <a href="https://doi.org/10.1038/nmeth.4292">https://doi.org/10.1038/nmeth.4292</a>.
</div>
<div id="ref-knitr3" class="csl-entry" role="listitem">
Xie, Yihui. 2014. <span>‚ÄúKnitr: A Comprehensive Tool for Reproducible Research in <span>R</span>.‚Äù</span> In <em>Implementing Reproducible Computational Research</em>, edited by Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman; Hall/CRC.
</div>
<div id="ref-knitr2" class="csl-entry" role="listitem">
‚Äî‚Äî‚Äî. 2015. <em>Dynamic Documents with <span>R</span> and Knitr</em>. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-knitr1" class="csl-entry" role="listitem">
‚Äî‚Äî‚Äî. 2024. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in r</em>. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-yates_ensembl_2022" class="csl-entry" role="listitem">
Yates, Andrew D, James Allen, Ridwan M Amode, Andrey G Azov, Matthieu Barba, Andr√©s Becerra, Jyothish Bhai, et al. 2022. <span>‚ÄúEnsembl <span>Genomes</span> 2022: An Expanding Genome Resource for Non-Vertebrates.‚Äù</span> <em>Nucleic Acids Research</em> 50 (D1): D996‚Äì1003. <a href="https://doi.org/10.1093/nar/gkab1007">https://doi.org/10.1093/nar/gkab1007</a>.
</div>
<div id="ref-kableExtra" class="csl-entry" role="listitem">
Zhu, Hao. 2021. <span>‚ÄúkableExtra: Construct Complex Table with ‚ÄôKable‚Äô and Pipe Syntax.‚Äù</span> <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>.
</div>
</div>
</section></section>
</div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">
<p>üîó <a href="https://3mmarand.github.io/BIO00088H-data/transcriptomics/week-4/overview.html">About Transcriptomics 2: Statistical Analysis</a></p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="../../site_libs/revealjs/dist/reveal.js"></script><!-- reveal.js plugins --><script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script><script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script><script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script><script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script><script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script><script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script><script src="../../site_libs/revealjs/plugin/notes/notes.js"></script><script src="../../site_libs/revealjs/plugin/search/search.js"></script><script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script><script src="../../site_libs/revealjs/plugin/math/math.js"></script><script>window.define = window.backupDefine; window.backupDefine = undefined;</script><script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script><script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script><script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>


</body></html>