<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Emma Rand">
<title>Workshop â€“ Data Analysis for Group Project</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-TVF95J593T"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-TVF95J593T', { 'anonymize_ip': true});
</script><link rel="stylesheet" href="../../styles.css">
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Data Analysis for Group Project</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Welcome!</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../core/core.html"> 
<span class="menu-text">Core Supporting Info</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../transcriptomics/transcriptomics.html" aria-current="page"> 
<span class="menu-text">Transcriptomics</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../images/images.html"> 
<span class="menu-text">Image Analysis</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../structures/structures.html"> 
<span class="menu-text">Structure Analysis</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <a href="https://twitter.com/er_13" title="Emma's Twitter" class="quarto-navigation-tool px-1" aria-label="Emma's Twitter"><i class="bi bi-twitter"></i></a>
    <a href="https://github.com/3mmaRand/BIO00088H-data" title="BIO00088H-data GitHub" class="quarto-navigation-tool px-1" aria-label="BIO00088H-data GitHub"><i class="bi bi-github"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/overview.html">Week 4: Statistical Analysis</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/workshop.html">Workshop</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/transcriptomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Transcriptomics Data Analysis for Group Project</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 3: Hello data!</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 4: Statistical Analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/workshop.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 5: Visualising and Interpreting</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul>
<li><a href="#session-overview" id="toc-session-overview" class="nav-link" data-scroll-target="#session-overview">Session overview</a></li>
  </ul>
</li>
  <li>
<a href="#set-up" id="toc-set-up" class="nav-link" data-scroll-target="#set-up">Set up</a>
  <ul>
<li><a href="#frog-development" id="toc-frog-development" class="nav-link" data-scroll-target="#frog-development">ğŸ¸ Frog development</a></li>
  <li><a href="#arabidopisis" id="toc-arabidopisis" class="nav-link" data-scroll-target="#arabidopisis">ğŸ„ Arabidopisis</a></li>
  <li><a href="#leishmania-mexicana" id="toc-leishmania-mexicana" class="nav-link" data-scroll-target="#leishmania-mexicana">ğŸ’‰ Leishmania mexicana</a></li>
  <li><a href="#stem-cells" id="toc-stem-cells" class="nav-link" data-scroll-target="#stem-cells">ğŸ­ Stem cells</a></li>
  <li><a href="#everyone" id="toc-everyone" class="nav-link" data-scroll-target="#everyone">Everyone</a></li>
  </ul>
</li>
  <li>
<a href="#import" id="toc-import" class="nav-link" data-scroll-target="#import">Import</a>
  <ul>
<li><a href="#frog-development-1" id="toc-frog-development-1" class="nav-link" data-scroll-target="#frog-development-1">ğŸ¸ Frog development</a></li>
  <li><a href="#arabidopisis-1" id="toc-arabidopisis-1" class="nav-link" data-scroll-target="#arabidopisis-1">ğŸ„ Arabidopisis</a></li>
  <li><a href="#leishmania-mexicana-1" id="toc-leishmania-mexicana-1" class="nav-link" data-scroll-target="#leishmania-mexicana-1">ğŸ’‰ Leishmania mexicana</a></li>
  <li><a href="#stem-cells-1" id="toc-stem-cells-1" class="nav-link" data-scroll-target="#stem-cells-1">ğŸ­ Stem cells</a></li>
  </ul>
</li>
  <li>
<a href="#differential-expression-analysis" id="toc-differential-expression-analysis" class="nav-link" data-scroll-target="#differential-expression-analysis">Differential Expression Analysis</a>
  <ul>
<li>
<a href="#frog-development-2" id="toc-frog-development-2" class="nav-link" data-scroll-target="#frog-development-2">ğŸ¸ Frog development</a>
  <ul class="collapse">
<li><a href="#genes-expressed-in-one-treatment" id="toc-genes-expressed-in-one-treatment" class="nav-link" data-scroll-target="#genes-expressed-in-one-treatment">1. Genes expressed in one treatment</a></li>
  <li><a href="#create-deseqdataset-object" id="toc-create-deseqdataset-object" class="nav-link" data-scroll-target="#create-deseqdataset-object">2. Create DESeqDataSet object</a></li>
  <li><a href="#prepare-the-normalised-counts" id="toc-prepare-the-normalised-counts" class="nav-link" data-scroll-target="#prepare-the-normalised-counts">3. Prepare the normalised counts</a></li>
  <li><a href="#differential-expression-analysis-1" id="toc-differential-expression-analysis-1" class="nav-link" data-scroll-target="#differential-expression-analysis-1">4. Differential expression analysis</a></li>
  </ul>
</li>
  <li><a href="#arabidopisis-2" id="toc-arabidopisis-2" class="nav-link" data-scroll-target="#arabidopisis-2">ğŸ„ Arabidopisis</a></li>
  <li><a href="#leishmania-mexicana-2" id="toc-leishmania-mexicana-2" class="nav-link" data-scroll-target="#leishmania-mexicana-2">ğŸ’‰ Leishmania mexicana</a></li>
  <li><a href="#stem-cells-2" id="toc-stem-cells-2" class="nav-link" data-scroll-target="#stem-cells-2">ğŸ­ Stem cells</a></li>
  </ul>
</li>
  <li>
<a href="#add-gene-information" id="toc-add-gene-information" class="nav-link" data-scroll-target="#add-gene-information">Add gene information</a>
  <ul>
<li><a href="#frog-development-3" id="toc-frog-development-3" class="nav-link" data-scroll-target="#frog-development-3">ğŸ¸ Frog development</a></li>
  <li><a href="#arabidopisis-3" id="toc-arabidopisis-3" class="nav-link" data-scroll-target="#arabidopisis-3">ğŸ„ Arabidopisis</a></li>
  <li><a href="#leishmania-mexicana-3" id="toc-leishmania-mexicana-3" class="nav-link" data-scroll-target="#leishmania-mexicana-3">ğŸ’‰ Leishmania mexicana</a></li>
  <li><a href="#stem-cells-3" id="toc-stem-cells-3" class="nav-link" data-scroll-target="#stem-cells-3">ğŸ­ Stem cells</a></li>
  </ul>
</li>
  <li><a href="#look-after-future-you" id="toc-look-after-future-you" class="nav-link" data-scroll-target="#look-after-future-you">ğŸ¤— Look after future you!</a></li>
  <li><a href="#finished" id="toc-finished" class="nav-link" data-scroll-target="#finished">ğŸ¥³ Finished</a></li>
  <li><a href="#independent-study-following-the-workshop" id="toc-independent-study-following-the-workshop" class="nav-link" data-scroll-target="#independent-study-following-the-workshop">Independent study following the workshop</a></li>
  <li><a href="#the-code-file" id="toc-the-code-file" class="nav-link" data-scroll-target="#the-code-file">The Code file</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/overview.html">Week 4: Statistical Analysis</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/workshop.html">Workshop</a></li></ol></nav><div class="quarto-title">
<h1 class="title">Workshop</h1>
<p class="subtitle lead">Transcripttranscriptomics 2: Statistical Analysis</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Emma Rand </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">2 October, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header><section id="introduction" class="level1"><h1>Introduction</h1>
<section id="session-overview" class="level2"><h2 class="anchored" data-anchor-id="session-overview">Session overview</h2>
<p>In the workshop, you will learn how to perform differential expression analysis on raw counts using <strong><code>DESeq2</code></strong> <span class="citation" data-cites="DESeq2">(<a href="#ref-DESeq2" role="doc-biblioref">Love, Huber, and Anders 2014</a>)</span> or on logged normalised expression values using <strong><code>scran</code></strong> <span class="citation" data-cites="scran">(<a href="#ref-scran" role="doc-biblioref">Lun, McCarthy, and Marioni 2016</a>)</span> or both.</p>
</section></section><section id="set-up" class="level1"><h1>Set up</h1>
<section id="frog-development" class="level2"><h2 class="anchored" data-anchor-id="frog-development">ğŸ¸ Frog development</h2>
<p>ğŸ¬ Open the <code>frogs-88H</code> RStudio Project and the <code>cont-fgf-s30.R</code> script.</p>
</section><section id="arabidopisis" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis">ğŸ„ Arabidopisis</h2>
<p>ğŸ¬ Open the <code>arabi-88H</code> RStudio Project and the <code>wildsuf-wilddef-s30.R</code> script.</p>
</section><section id="leishmania-mexicana" class="level2"><h2 class="anchored" data-anchor-id="leishmania-mexicana">ğŸ’‰ Leishmania mexicana</h2>
<p>ğŸ¬ Open the <code>leish-88H</code> RStudio Project and the <code>pro-meta-s30.R</code> script.</p>
</section><section id="stem-cells" class="level2"><h2 class="anchored" data-anchor-id="stem-cells">ğŸ­ Stem cells</h2>
<p>ğŸ¬ Open the <code>mice-88H</code> RStudio Project and the <code>hspc-prog.R</code> script.</p>
</section><section id="everyone" class="level2"><h2 class="anchored" data-anchor-id="everyone">Everyone</h2>
<p>ğŸ¬ Make a new folder <code>results</code> in the project directory.</p>
<p>This is where we will save our results.</p>
<p>ğŸ¬ Load <strong><code>tidyverse</code></strong> <span class="citation" data-cites="tidyverse">(<a href="#ref-tidyverse" role="doc-biblioref">Wickham et al. 2019</a>)</span> You most likely have this code at the top of `your script already.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€
âœ” dplyr     1.1.3     âœ” readr     2.1.4
âœ” forcats   1.0.0     âœ” stringr   1.5.0
âœ” ggplot2   3.4.3     âœ” tibble    3.2.1
âœ” lubridate 1.9.3     âœ” tidyr     1.3.0
âœ” purrr     1.0.2     
â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€
âœ– dplyr::filter() masks stats::filter()
âœ– dplyr::lag()    masks stats::lag()
â„¹ Use the conflicted package to force all conflicts to become errors</code></pre>
<p>Have you ever stopped to think about this message? It is telling us that there are functions in the <code>dplyr</code> package that have the same name as functions in the <code>stats</code> package and that R will use the <code>dplyr</code> version. As this is what you want, this has always been fine. It still is fine in this case. However, as you start to load more packages, you will want to know if you are using a function from a package that has the same name as a function in another loaded package. This is where the <strong><code>conflicted</code></strong> <span class="citation" data-cites="conflicted">(<a href="#ref-conflicted" role="doc-biblioref">Wickham 2023</a>)</span> package comes in. <strong><code>Conflicted</code></strong> will warn you when you are using a function that has the same name as a function in another package. You can then choose which function to use.</p>
<p>ğŸ¬ Load the <strong><code>conflicted</code></strong> package:</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://conflicted.r-lib.org/">conflicted</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Instead of getting a warning every time you are using a function that has a function with the same name in another package, we can declare a preference for one function over another. This is useful for the functions you use a lot or ones where you are certain you always want to use a particular function.</p>
<p>For example, to always use the <strong><code>dplyr</code></strong> version of <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> by default you can add this to the top of your script:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://conflicted.r-lib.org/reference/conflicts_prefer.html">conflicts_prefer</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will also want to ensure that we are using the <code><a href="https://generics.r-lib.org/reference/setops.html">setdiff()</a></code> function from the <code>GenomicRanges</code> package.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://conflicted.r-lib.org/reference/conflicts_prefer.html">conflicts_prefer</a></span><span class="op">(</span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html">setdiff</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="import" class="level1"><h1>Import</h1>
<section id="frog-development-1" class="level2"><h2 class="anchored" data-anchor-id="frog-development-1">ğŸ¸ Frog development</h2>
<p>We need to import the S30 data that were filtered to remove genes with 4, 5 or 6 zeros and those where the total counts was less than 20.</p>
<p>ğŸ¬ Import the data from the <code>data-processed</code> folder.</p>
<p>Now go to <a href="#differential-expression-analysis">Differential Expression Analysis</a>.</p>
</section><section id="arabidopisis-1" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis-1">ğŸ„ Arabidopisis</h2>
<p>Import</p>
<p>Now go to <a href="#differential-expression-analysis">Differential Expression Analysis</a>.</p>
</section><section id="leishmania-mexicana-1" class="level2"><h2 class="anchored" data-anchor-id="leishmania-mexicana-1">ğŸ’‰ Leishmania mexicana</h2>
<p>Import</p>
<p>Now go to <a href="#differential-expression-analysis">Differential Expression Analysis</a>.</p>
</section><section id="stem-cells-1" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-1">ğŸ­ Stem cells</h2>
<p>Import</p>
<p>Now go to <a href="#differential-expression-analysis">Differential Expression Analysis</a>.</p>
</section></section><section id="differential-expression-analysis" class="level1"><h1>Differential Expression Analysis</h1>
<section id="frog-development-2" class="level2"><h2 class="anchored" data-anchor-id="frog-development-2">ğŸ¸ Frog development</h2>
<p>These are the steps we will take</p>
<ol type="1">
<li>Find the genes that are expressed in only one treatment group.</li>
<li>Create a DESeqDataSet object. This is a special object that is used by the <strong><code>DESeq2</code></strong> package</li>
<li>Prepare the normalised counts from the DESeqDataSet object.</li>
<li>Do differential expression analysis on the genes. This needs to be done on the raw counts.</li>
</ol>
<p>The all but the first step are done with the <strong><code>DESeq2</code></strong> package</p>
<section id="genes-expressed-in-one-treatment" class="level3"><h3 class="anchored" data-anchor-id="genes-expressed-in-one-treatment">1. Genes expressed in one treatment</h3>
<p>The genes expressed in only one treatment group are those with zeros in all three replicates in one group and non-zero values in all three replicates in the other group. For example, those shown here:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 26%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">xenbase_gene_id</th>
<th style="text-align: right;">S30_C_1</th>
<th style="text-align: right;">S30_C_2</th>
<th style="text-align: right;">S30_C_3</th>
<th style="text-align: right;">S30_F_1</th>
<th style="text-align: right;">S30_F_2</th>
<th style="text-align: right;">S30_F_3</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">XB-GENE-1018260</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">16</td>
</tr>
<tr class="even">
<td style="text-align: left;">XB-GENE-17330117</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">17</td>
</tr>
<tr class="odd">
<td style="text-align: left;">XB-GENE-17332184</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">6</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We will use <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> to find these genes.</p>
<p>ğŸ¬ Find the genes that are expressed only in the FGF-treated group:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">s30_fgf_only</span> <span class="op">&lt;-</span> <span class="va">s30_filtered</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">S30_C_1</span> <span class="op">==</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_C_2</span> <span class="op">==</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_C_3</span> <span class="op">==</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_F_1</span> <span class="op">&gt;</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_F_2</span> <span class="op">&gt;</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_F_3</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>â“ How many genes are expressed only in the FGF-treated group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 26 genes expressed only in the FGF-treated group. -->
<p>ğŸ¬ Now you find any genes that are expressed only in the control group.</p>
<p>â“ How many genes are expressed only in the control group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 0 genes expressed only in the control group. -->
<p>â“ Do the results make sense to you in light of what you know about the biology?</p>
<!-- #---THINKING ANSWER--- -->
<!-- Since FGF is a growth factor, it is likely to induce the expression  -->
<!-- of genes. Therefore, it is not surprising to me that there are no  -->
<!-- genes expressed in the control group but not in the FGF-treated  -->
<!-- group. However, I'm not an expert in developmental biology. -->
<!-- Domain expertise is important when analysing data so  -->
<!-- discuss your results with Betsy and your group. -->
<p>ğŸ¬ Write all the genes that are expressed one group only to file (saved in <code>results</code>)</p>
</section><section id="create-deseqdataset-object" class="level3"><h3 class="anchored" data-anchor-id="create-deseqdataset-object">2. Create DESeqDataSet object</h3>
<p>ğŸ¬ Load the DESeq2 package:</p>
<p>A DEseqDataSet object is a custom data type that is used by <strong><code>DESeq2</code></strong>. Custom data types are common in the Bioconductor<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> packages. They are used to store data in a way that is useful for the analysis. These data types typically have data, transformed data, metadata and experimental designs within them.</p>
<p>To create a DESeqDataSet object, we need to provide three things:</p>
<ul>
<li>The raw counts - these are in <code>s30_filtered</code>
</li>
<li>The meta data which gives information about the samples and which treatment groups they belong to</li>
<li>A design matrix which captures the design of the statistical model.</li>
</ul>
<p>The counts must in a <em>matrix</em> rather than a dataframe. Unlike a dataframe, a matrix has columns of all the same type. That is, it will contain only the counts. The gene ids are given as row names rather than a column. The <code><a href="https://rdrr.io/r/base/matrix.html">matrix()</a></code> function will create a matrix from a dataframe of columns of the same type and the <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> function can be used to remove the gene ids column.</p>
<p>ğŸ¬ Create a matrix of the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">s30_count_mat</span> <span class="op">&lt;-</span> <span class="va">s30_filtered</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">xenbase_gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Add the gene ids as row names to the matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># add the row names to the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">s30_count_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">s30_filtered</span><span class="op">$</span><span class="va">xenbase_gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You might want to view the matrix (click on it in your environment pane).</p>
<p>The metadata are in a file, <a href="meta/frog_meta_data.txt">frog_meta_data.txt</a>. This is a tab-delimited file. The first column is the sample name and the second column is the treatment group.</p>
<p>ğŸ¬ Make a folder called <code>meta</code> and save the file to it.</p>
<p>ğŸ¬ Read the metadata into a dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_table.html">read_table</a></span><span class="op">(</span><span class="st">"meta/frog_meta_data.txt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Examine the resulting dataframe.</p>
<p>We need to add the sample names as row names to the metadata dataframe. This is because the DESeqDataSet object will use the row names to match the samples in the metadata to the samples in the counts matrix.</p>
<p>ğŸ¬ Add the sample names as row names to the metadata dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">sample_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>(you will get a warning message but you can ignore it)</p>
<p>We are dealing only with the S30 data so we need to remove the samples that are not in the S30 data.</p>
<p>ğŸ¬ Filter the metadata to keep only the S30 information:</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta_s30</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">stage</span> <span class="op">==</span> <span class="st">"stage_30"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now create the DESeqDataSet object. The design formula describes the statistical model. You should the form from previous work. The <code>~</code> indicates that the left hand side is the response variable (in this case the counts) and the right hand side are the explanatory variables. We are interested in the difference between the treatments but we include <code>sibling_rep</code> to account for the fact that the data are paired.</p>
<p>Note that:</p>
<ul>
<li>The names of the columns in the count matrix have to exactly match the names of the rows in the metadata dataframe. They also need to be in the same order.</li>
<li>The names of the explanatory variables in the design formula have to match the names of columns in the metadata.</li>
</ul>
<p>ğŸ¬ Create the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="va">s30_count_mat</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">meta_s30</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span> <span class="va">treatment</span> <span class="op">+</span> <span class="va">sibling_rep</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The warning â€œWarning: some variables in design formula are characters, converting to factorsâ€ just means that the variable type of treatment and sibling_rep in the metadata dataframe are â€œcharâ€ and they have been converted into the factors.</p>
<p>To help you understand what the <code>DESeqDataSet</code> object we have called <code>dds</code> contains, we can look its contents</p>
<p>The counts are in <code>dds@assays@data@listData[["counts"]]</code> and the metadata are in <code>dds@colData</code> but the easiest way to see them is to use the <code>counts()</code> and <code>colData()</code> functions from the <strong><code>DESeq2</code></strong> package.</p>
<p>ğŸ¬ View the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in .External2(C_dataviewer, x, title): unable to start data viewer</code></pre>
</div>
</div>
<p>You should be able to see that this is the same as in <code>s30_count_mat</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 4 columns
          sample_id       stage treatment sibling_rep
        &lt;character&gt; &lt;character&gt;  &lt;factor&gt;    &lt;factor&gt;
S30_C_1     S30_C_1    stage_30   control       one  
S30_C_2     S30_C_2    stage_30   control       two  
S30_C_3     S30_C_3    stage_30   control       three
S30_F_1     S30_F_1    stage_30   FGF           one  
S30_F_2     S30_F_2    stage_30   FGF           two  
S30_F_3     S30_F_3    stage_30   FGF           three</code></pre>
</div>
</div>
</section><section id="prepare-the-normalised-counts" class="level3"><h3 class="anchored" data-anchor-id="prepare-the-normalised-counts">3. Prepare the normalised counts</h3>
<p>The normalised counts are the counts that have been transformed to account for the library size (i.e., the total number of reads in a sample) and the gene length. We have to first estimate the normalisation factors and store them in the DESeqDataSet object and then we can get the normalised counts.</p>
<p>ğŸ¬ Estimate the factors for normalisation and store them in the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Look at the factors (just for information):</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  S30_C_1   S30_C_2   S30_C_3   S30_F_1   S30_F_2   S30_F_3 
0.8812200 0.9454600 1.2989886 1.0881870 1.0518961 0.8322894 </code></pre>
</div>
</div>
<p>The normalised counts will be useful to use later. To get the normalised counts we again used the <code>counts()</code> function but this time we use the <code>normalized=TRUE</code> argument.</p>
<p>ğŸ¬ Save the normalised to a matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">normalised_counts</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will write the normalised counts to a file so that we can use them in the future.</p>
<p>ğŸ¬ Make a dataframe of the normalised counts, adding a column for the gene ids at the same time:</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">s30_normalised_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">normalised_counts</span>,</span>
<span>                                    xenbase_gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">normalised_counts</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="differential-expression-analysis-1" class="level3"><h3 class="anchored" data-anchor-id="differential-expression-analysis-1">4. Differential expression analysis</h3>
<p>We use the <code>DESeq()</code> function to do the differential expression analysis. This function fits the statistical model to the data and then uses the model to calculate the significance of the difference between the treatments. It again stores the results in the DESseqDataSet object. Note that the differential expression needs the raw (unnormalised counts) as it does its own normalisation as part of the process.</p>
<p>ğŸ¬ Run the differential expression analysis and store the results in the same object:</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function will take only a few moments to run on this data but can take longer for bigger datasets.</p>
<p>We need to define the contrasts we want to test. We want to test the difference between the treatments so we will define the contrast as <code>FGF</code> and <code>control</code>.</p>
<p>ğŸ¬ Define the contrast:</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">contrast_fgf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"treatment"</span>, <span class="st">"FGF"</span>, <span class="st">"control"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that <code>treatment</code> is the name of the column in the metadata dataframe and <code>FGF</code> and <code>control</code> are the names of the levels in the <code>treatment</code> column. By putting them in the order <code>FGF</code> , <code>control</code> we are saying the fold change will be FGF / control. This means:</p>
<ul>
<li>positive log fold changes indicate FGF &gt; control and</li>
<li>negative log fold changes indicates control &gt; FGF.</li>
</ul>
<p>If we had put them in the order <code>control</code>, <code>FGF</code> we would have the reverse.</p>
<p>ğŸ¬ Extract the results from the DESseqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results_fgf</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                       contrast <span class="op">=</span> <span class="va">contrast_fgf</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This will give us the log<sub>2</sub> fold change, the <em>p</em>-value and the adjusted <em>p</em>-value between the control and the FGF-treatment for each gene.</p>
<p>ğŸ¬ Put the results in a dataframe and add the gene ids as a column:</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">s30_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">results_fgf</span>,</span>
<span>                          xenbase_gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">results_fgf</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is useful to have the normalised counts and the statistical results in one dataframe.</p>
<p>ğŸ¬ Merge the two dataframes:</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># merge the results with the normalised counts</span></span>
<span><span class="va">s30_results</span> <span class="op">&lt;-</span> <span class="va">s30_normalised_counts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">s30_results</span>, by <span class="op">=</span> <span class="st">"xenbase_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#add-gene-information">Add gene information</a>.</p>
</section></section><section id="arabidopisis-2" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis-2">ğŸ„ Arabidopisis</h2>
<p>DE instructions</p>
<p>Now go to <a href="#add-gene-information">Add gene information</a>.</p>
</section><section id="leishmania-mexicana-2" class="level2"><h2 class="anchored" data-anchor-id="leishmania-mexicana-2">ğŸ’‰ Leishmania mexicana</h2>
<p>DE instructions</p>
<p>Now go to <a href="#add-gene-information">Add gene information</a>.</p>
</section><section id="stem-cells-2" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-2">ğŸ­ Stem cells</h2>
<p>DE instructions</p>
<p>Now go to <a href="#add-gene-information">Add gene information</a>.</p>
</section></section><section id="add-gene-information" class="level1"><h1>Add gene information</h1>
<section id="frog-development-3" class="level2"><h2 class="anchored" data-anchor-id="frog-development-3">ğŸ¸ Frog development</h2>
<ul>
<li><p>I got the information from the <a href="https://www.xenbase.org/xenbase/static-xenbase/ftpDatafiles.jsp">Xenbase information pages</a> under Data Reports | Gene Information</p></li>
<li><p>This is listed: Xenbase Gene Product Information [readme] <a href="https://download.xenbase.org/xenbase/GenePageReports/xenbase.gpi.gz">gzipped gpi (tab separated)</a></p></li>
<li><p>Click on the readme link to see the file format and columns</p></li>
<li><p>I downloaded <a href="https://download.xenbase.org/xenbase/GenePageReports/xenbase.gpi.gz">xenbase.gpi.gz</a>, unzipped it, removed header lines and the <em>Xenopus tropicalis</em> (taxon:8364) entries and saved it as <a href="meta/xenbase_info.xlsx">xenbase_info.xlsx</a></p></li>
</ul>
<p>If you want to emulate what I did you can use the following commands in the terminal after downloading the file:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> xenbase.gpi.gz</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> xenbase.gpi</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="ex">q</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><code>gunzip</code> unzips the file and <code>less</code> allows you to view the file. <code>q</code> quits the viewer. You will see the header lines and that the file contains both <em>Xenopus tropicalis</em> and <em>Xenopus laevis</em>. I read the file in with <code>read_tsv</code> (skipping the first header lines) then filtered out the <em>Xenopus tropicalis</em> entries, dropped some columns and saved the file as an excel file.</p>
<p>However, I have already done this for you and saved the file as <a href="meta/xenbase_info.xlsx">xenbase_info.xlsx</a> in the <code>meta</code> folder. We will import this file and join it to the results dataframe.</p>
<p>ğŸ¬ Load the <strong><code>readxl</code></strong> <span class="citation" data-cites="readxl">(<a href="#ref-readxl" role="doc-biblioref">Wickham and Bryan 2023</a>)</span> package:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org">readxl</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Import the Xenbase gene information file:</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_excel</a></span><span class="op">(</span><span class="st">"meta/xenbase_info.xlsx"</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should view the resulting dataframe to see what information is available. You can use <code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code> or <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code>.</p>
<p>ğŸ¬ Merge the gene information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># join the gene info with the results</span></span>
<span><span class="va">s30_results</span> <span class="op">&lt;-</span> <span class="va">s30_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">gene_info</span>, by <span class="op">=</span> <span class="st">"xenbase_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Save the results to a file:</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">s30_results</span>, file <span class="op">=</span> <span class="st">"results/s30_results.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="arabidopisis-3" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis-3">ğŸ„ Arabidopisis</h2>
</section><section id="leishmania-mexicana-3" class="level2"><h2 class="anchored" data-anchor-id="leishmania-mexicana-3">ğŸ’‰ Leishmania mexicana</h2>
</section><section id="stem-cells-3" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-3">ğŸ­ Stem cells</h2>
</section></section><section id="look-after-future-you" class="level1"><h1>ğŸ¤— Look after future you!</h1>
<p>ğŸ¬ Go through you script and tidy up. You might need to :</p>
<ul>
<li>collect together library statements at the top of the script</li>
<li>remove code that you needed to start today but which wouldnâ€™t be needed if you were running the script from the top (e.g., importing)</li>
<li>edit your comments for clarity</li>
<li>rename variables for consistency or clarity</li>
<li>remove house keeping or exploratory code</li>
<li>restyle code, add code section headers etc</li>
</ul></section><section id="finished" class="level1"><h1>ğŸ¥³ Finished</h1>
<p>Well Done!</p>
</section><section id="independent-study-following-the-workshop" class="level1"><h1>Independent study following the workshop</h1>
<p><a href="../../transcriptomics/week-4/study_after_workshop.html">Consolidate</a></p>
</section><section id="the-code-file" class="level1"><h1>The Code file</h1>
<p>These contain all the code needed in the workshop even where it is not visible on the webpage.</p>
<p>The <a href="../../transcriptomics/week-4/workshop.html">workshop.qmd</a> file is the file I use to compile the practical. Qmd stands for Quarto markdown. It allows code and ordinary text to be interleaved to produce well-formatted reports including webpages. Right-click on the link and choose Save-As to download. You will be able to open the Qmd file in RStudio. Alternatively, <a href="https://github.com/3mmaRand/BIO00088H-data/blob/main/transcriptomics/week-4/workshop.qmd">View in Browser</a>. Coding and thinking answers are marked with <code>#---CODING ANSWER---</code> and <code>#---THINKING ANSWER---</code></p>
<p>Pages made with R <span class="citation" data-cites="R-core">(<a href="#ref-R-core" role="doc-biblioref">R Core Team 2024</a>)</span>, Quarto <span class="citation" data-cites="Allaire_Quarto_2024">(<a href="#ref-Allaire_Quarto_2024" role="doc-biblioref">Allaire et al. 2024</a>)</span>, <code>knitr</code> <span class="citation" data-cites="knitr1 knitr2 knitr3">(<a href="#ref-knitr1" role="doc-biblioref">Xie 2024</a>, <a href="#ref-knitr2" role="doc-biblioref">2015</a>, <a href="#ref-knitr3" role="doc-biblioref">2014</a>)</span>, <code>kableExtra</code> <span class="citation" data-cites="kableExtra">(<a href="#ref-kableExtra" role="doc-biblioref">Zhu 2021</a>)</span></p>
</section><section id="references" class="level1">



</section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Allaire_Quarto_2024" class="csl-entry" role="listitem">
Allaire, J. J., Charles Teague, Carlos Scheidegger, Yihui Xie, and Christophe Dervieux. 2024. <span>â€œ<span>Quarto</span>.â€</span> <a href="https://doi.org/10.5281/zenodo.5960048">https://doi.org/10.5281/zenodo.5960048</a>.
</div>
<div id="ref-DESeq2" class="csl-entry" role="listitem">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>â€œModerated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.â€</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-scran" class="csl-entry" role="listitem">
Lun, Aaron T. L., Davis J. McCarthy, and John C. Marioni. 2016. <span>â€œA Step-by-Step Workflow for Low-Level Analysis of Single-Cell RNA-Seq Data with Bioconductor.â€</span> <em>F1000Res.</em> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-R-core" class="csl-entry" role="listitem">
R Core Team. 2024. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-conflicted" class="csl-entry" role="listitem">
Wickham, Hadley. 2023. <em>Conflicted: An Alternative Conflict Resolution Strategy</em>. <a href="https://conflicted.r-lib.org/">https://conflicted.r-lib.org/</a>.
</div>
<div id="ref-tidyverse" class="csl-entry" role="listitem">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy Dâ€™Agostino McGowan, Romain FranÃ§ois, Garrett Grolemund, et al. 2019. <span>â€œWelcome to the <span class="nocase">tidyverse</span>.â€</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-readxl" class="csl-entry" role="listitem">
Wickham, Hadley, and Jennifer Bryan. 2023. <em>Readxl: Read Excel Files</em>. <a href="https://readxl.tidyverse.org">https://readxl.tidyverse.org</a>.
</div>
<div id="ref-knitr3" class="csl-entry" role="listitem">
Xie, Yihui. 2014. <span>â€œKnitr: A Comprehensive Tool for Reproducible Research in <span>R</span>.â€</span> In <em>Implementing Reproducible Computational Research</em>, edited by Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman; Hall/CRC.
</div>
<div id="ref-knitr2" class="csl-entry" role="listitem">
â€”â€”â€”. 2015. <em>Dynamic Documents with <span>R</span> and Knitr</em>. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-knitr1" class="csl-entry" role="listitem">
â€”â€”â€”. 2024. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in r</em>. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-kableExtra" class="csl-entry" role="listitem">
Zhu, Hao. 2021. <span>â€œkableExtra: Construct Complex Table with â€™Kableâ€™ and Pipe Syntax.â€</span> <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>
<ol>
<li id="fn1"><p><a href="https://www.bioconductor.org/">Bioconductor</a> is a project that develops and supports R packages for bioinformatics.<a href="#fnref1" class="footnote-back" role="doc-backlink">â†©ï¸</a></p></li>
</ol></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("^(?:http:|https:)\/\/3mmarand\.github\.io\/BIO00088H-data");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Rand E (2024). Data Analysis for Group Project. <a href="https://3mmarand.github.io/BIO00088H-data/" class="uri">https://3mmarand.github.io/BIO00088H-data/</a>.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/er_13">
      <i class="bi bi-twitter" role="img" aria-label="Twitter">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/3mmaRand/BIO00088H-data">
      <i class="bi bi-github" role="img" aria-label="BIO00088H-data GitHub">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>