<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Emma Rand">
<title>Workshop â€“ Data Analysis for Group Project</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-TVF95J593T"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-TVF95J593T', { 'anonymize_ip': true});
</script><link rel="stylesheet" href="../../styles.css">
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Data Analysis for Group Project</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Welcome!</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../core/core.html"> 
<span class="menu-text">Core Supporting Info</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../transcriptomics/transcriptomics.html" aria-current="page"> 
<span class="menu-text">Transcriptomics</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../images/images.html"> 
<span class="menu-text">Image Analysis</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../structures/structures.html"> 
<span class="menu-text">Structure Analysis</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <a href="https://twitter.com/er_13" title="Emma's Twitter" class="quarto-navigation-tool px-1" aria-label="Emma's Twitter"><i class="bi bi-twitter"></i></a>
    <a href="https://github.com/3mmaRand/BIO00088H-data" title="BIO00088H-data GitHub" class="quarto-navigation-tool px-1" aria-label="BIO00088H-data GitHub"><i class="bi bi-github"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/overview.html">Week 4: Statistical Analysis</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/workshop.html">Workshop</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/transcriptomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Transcriptomics Data Analysis for Group Project</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 3: Hello data!</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 4: Statistical Analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/workshop.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 5: Visualising and Interpreting</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul>
<li><a href="#session-overview" id="toc-session-overview" class="nav-link" data-scroll-target="#session-overview">Session overview</a></li>
  </ul>
</li>
  <li>
<a href="#set-up" id="toc-set-up" class="nav-link" data-scroll-target="#set-up">Set up</a>
  <ul>
<li><a href="#frog-development" id="toc-frog-development" class="nav-link" data-scroll-target="#frog-development">ğŸ¸ Frog development</a></li>
  <li><a href="#arabidopisis" id="toc-arabidopisis" class="nav-link" data-scroll-target="#arabidopisis">ğŸ„ <em>Arabidopisis</em></a></li>
  <li><a href="#leishmania" id="toc-leishmania" class="nav-link" data-scroll-target="#leishmania">ğŸ’‰ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells" id="toc-stem-cells" class="nav-link" data-scroll-target="#stem-cells">ğŸ­ Stem cells</a></li>
  <li><a href="#everyone" id="toc-everyone" class="nav-link" data-scroll-target="#everyone">Everyone</a></li>
  </ul>
</li>
  <li>
<a href="#import" id="toc-import" class="nav-link" data-scroll-target="#import">Import</a>
  <ul>
<li><a href="#frog-development-1" id="toc-frog-development-1" class="nav-link" data-scroll-target="#frog-development-1">ğŸ¸ Frog development</a></li>
  <li><a href="#arabidopisis-1" id="toc-arabidopisis-1" class="nav-link" data-scroll-target="#arabidopisis-1">ğŸ„ <em>Arabidopisis</em></a></li>
  <li><a href="#leishmania-1" id="toc-leishmania-1" class="nav-link" data-scroll-target="#leishmania-1">ğŸ’‰ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells-1" id="toc-stem-cells-1" class="nav-link" data-scroll-target="#stem-cells-1">ğŸ­ Stem cells</a></li>
  </ul>
</li>
  <li>
<a href="#differential-expression-analysis" id="toc-differential-expression-analysis" class="nav-link" data-scroll-target="#differential-expression-analysis">Differential Expression Analysis</a>
  <ul>
<li>
<a href="#frog-development-2" id="toc-frog-development-2" class="nav-link" data-scroll-target="#frog-development-2">ğŸ¸ Frog development</a>
  <ul class="collapse">
<li><a href="#genes-expressed-in-one-treatment" id="toc-genes-expressed-in-one-treatment" class="nav-link" data-scroll-target="#genes-expressed-in-one-treatment">1. Genes expressed in one treatment</a></li>
  <li><a href="#create-deseqdataset-object" id="toc-create-deseqdataset-object" class="nav-link" data-scroll-target="#create-deseqdataset-object">2. Create DESeqDataSet object</a></li>
  <li><a href="#prepare-the-normalised-counts" id="toc-prepare-the-normalised-counts" class="nav-link" data-scroll-target="#prepare-the-normalised-counts">3. Prepare the normalised counts</a></li>
  <li><a href="#differential-expression-analysis-1" id="toc-differential-expression-analysis-1" class="nav-link" data-scroll-target="#differential-expression-analysis-1">4. Differential expression analysis</a></li>
  </ul>
</li>
  <li>
<a href="#arabidopisis-2" id="toc-arabidopisis-2" class="nav-link" data-scroll-target="#arabidopisis-2">ğŸ„ <em>Arabidopisis</em></a>
  <ul class="collapse">
<li><a href="#genes-expressed-in-one-treatment-1" id="toc-genes-expressed-in-one-treatment-1" class="nav-link" data-scroll-target="#genes-expressed-in-one-treatment-1">1. Genes expressed in one treatment</a></li>
  <li><a href="#create-deseqdataset-object-1" id="toc-create-deseqdataset-object-1" class="nav-link" data-scroll-target="#create-deseqdataset-object-1">2. Create DESeqDataSet object</a></li>
  <li><a href="#prepare-the-normalised-counts-1" id="toc-prepare-the-normalised-counts-1" class="nav-link" data-scroll-target="#prepare-the-normalised-counts-1">3. Prepare the normalised counts</a></li>
  <li><a href="#differential-expression-analysis-2" id="toc-differential-expression-analysis-2" class="nav-link" data-scroll-target="#differential-expression-analysis-2">4. Differential expression analysis</a></li>
  </ul>
</li>
  <li>
<a href="#leishmania-2" id="toc-leishmania-2" class="nav-link" data-scroll-target="#leishmania-2">ğŸ’‰ <em>Leishmania</em></a>
  <ul class="collapse">
<li><a href="#genes-expressed-in-one-treatment-2" id="toc-genes-expressed-in-one-treatment-2" class="nav-link" data-scroll-target="#genes-expressed-in-one-treatment-2">1. Genes expressed in one treatment</a></li>
  <li><a href="#create-deseqdataset-object-2" id="toc-create-deseqdataset-object-2" class="nav-link" data-scroll-target="#create-deseqdataset-object-2">2. Create DESeqDataSet object</a></li>
  <li><a href="#prepare-the-normalised-counts-2" id="toc-prepare-the-normalised-counts-2" class="nav-link" data-scroll-target="#prepare-the-normalised-counts-2">3. Prepare the normalised counts</a></li>
  <li><a href="#differential-expression-analysis-3" id="toc-differential-expression-analysis-3" class="nav-link" data-scroll-target="#differential-expression-analysis-3">4. Differential expression analysis</a></li>
  </ul>
</li>
  <li>
<a href="#stem-cells-2" id="toc-stem-cells-2" class="nav-link" data-scroll-target="#stem-cells-2">ğŸ­ Stem cells</a>
  <ul class="collapse">
<li><a href="#genes-expressed-in-one-cell-type" id="toc-genes-expressed-in-one-cell-type" class="nav-link" data-scroll-target="#genes-expressed-in-one-cell-type">1. Genes expressed in one cell type</a></li>
  <li><a href="#prepare-the-data-for-analysis-with-scran" id="toc-prepare-the-data-for-analysis-with-scran" class="nav-link" data-scroll-target="#prepare-the-data-for-analysis-with-scran">2. Prepare the data for analysis with <strong><code>scran</code></strong></a></li>
  <li><a href="#differential-expression-analysis-4" id="toc-differential-expression-analysis-4" class="nav-link" data-scroll-target="#differential-expression-analysis-4">3. Differential expression analysis</a></li>
  </ul>
</li>
  </ul>
</li>
  <li>
<a href="#add-gene-information" id="toc-add-gene-information" class="nav-link" data-scroll-target="#add-gene-information">Add gene information</a>
  <ul>
<li><a href="#frog-development-3" id="toc-frog-development-3" class="nav-link" data-scroll-target="#frog-development-3">ğŸ¸ Frog development</a></li>
  <li><a href="#arabidopisis-3" id="toc-arabidopisis-3" class="nav-link" data-scroll-target="#arabidopisis-3">ğŸ„ <em>Arabidopisis</em></a></li>
  <li><a href="#leishmania-3" id="toc-leishmania-3" class="nav-link" data-scroll-target="#leishmania-3">ğŸ’‰ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells-3" id="toc-stem-cells-3" class="nav-link" data-scroll-target="#stem-cells-3">ğŸ­ Stem cells</a></li>
  </ul>
</li>
  <li><a href="#look-after-future-you" id="toc-look-after-future-you" class="nav-link" data-scroll-target="#look-after-future-you">ğŸ¤— Look after future you!</a></li>
  <li><a href="#finished" id="toc-finished" class="nav-link" data-scroll-target="#finished">ğŸ¥³ Finished</a></li>
  <li><a href="#independent-study-following-the-workshop" id="toc-independent-study-following-the-workshop" class="nav-link" data-scroll-target="#independent-study-following-the-workshop">Independent study following the workshop</a></li>
  <li><a href="#the-code-file" id="toc-the-code-file" class="nav-link" data-scroll-target="#the-code-file">The Code file</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/overview.html">Week 4: Statistical Analysis</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/workshop.html">Workshop</a></li></ol></nav><div class="quarto-title">
<h1 class="title">Workshop</h1>
<p class="subtitle lead">Transcripttranscriptomics 2: Statistical Analysis</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Emma Rand </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">25 October, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header><section id="introduction" class="level1"><h1>Introduction</h1>
<section id="session-overview" class="level2"><h2 class="anchored" data-anchor-id="session-overview">Session overview</h2>
<p>In the workshop, you will learn how to perform differential expression analysis on raw counts using <strong><code>DESeq2</code></strong> <span class="citation" data-cites="DESeq2">(<a href="#ref-DESeq2" role="doc-biblioref">Love, Huber, and Anders 2014</a>)</span> or on logged normalised expression values using <strong><code>scran</code></strong> <span class="citation" data-cites="scran">(<a href="#ref-scran" role="doc-biblioref">Lun, McCarthy, and Marioni 2016</a>)</span> or both.</p>
</section></section><section id="set-up" class="level1"><h1>Set up</h1>
<section id="frog-development" class="level2"><h2 class="anchored" data-anchor-id="frog-development">ğŸ¸ Frog development</h2>
<p>ğŸ¬ Open the <code>frogs-88H</code> RStudio Project and the <code>cont-fgf-s30.R</code> script.</p>
</section><section id="arabidopisis" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis">ğŸ„ <em>Arabidopisis</em>
</h2>
<p>ğŸ¬ Open the <code>arab-88H</code> RStudio Project and the <code>suff-def-wild.R</code> script.</p>
</section><section id="leishmania" class="level2"><h2 class="anchored" data-anchor-id="leishmania">ğŸ’‰ <em>Leishmania</em>
</h2>
<p>ğŸ¬ Open the <code>leish-88H</code> RStudio Project and the <code>pro-meta.R</code> script.</p>
</section><section id="stem-cells" class="level2"><h2 class="anchored" data-anchor-id="stem-cells">ğŸ­ Stem cells</h2>
<p>ğŸ¬ Open the <code>mice-88H</code> RStudio Project and the <code>hspc-prog.R</code> script.</p>
</section><section id="everyone" class="level2"><h2 class="anchored" data-anchor-id="everyone">Everyone</h2>
<p>ğŸ¬ Make a new folder <code>results</code> in the project directory.</p>
<p>This is where we will save our results.</p>
<p>ğŸ¬ Load <strong><code>tidyverse</code></strong> <span class="citation" data-cites="tidyverse">(<a href="#ref-tidyverse" role="doc-biblioref">Wickham et al. 2019</a>)</span> You most likely have this code at the top of `your script already.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€
âœ” dplyr     1.1.3     âœ” readr     2.1.4
âœ” forcats   1.0.0     âœ” stringr   1.5.0
âœ” ggplot2   3.4.3     âœ” tibble    3.2.1
âœ” lubridate 1.9.3     âœ” tidyr     1.3.0
âœ” purrr     1.0.2     
â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€
âœ– dplyr::filter() masks stats::filter()
âœ– dplyr::lag()    masks stats::lag()
â„¹ Use the conflicted package to force all conflicts to become errors</code></pre>
<p>Have you ever stopped to think about this message? It is telling us that there are functions in the <code>dplyr</code> package that have the same name as functions in the <code>stats</code> package and that R will use the <code>dplyr</code> version. As this is what you want, this has always been fine. It still is fine in this case. However, as you start to load more packages, you will want to know if you are using a function from a package that has the same name as a function in another loaded package. This is where the <strong><code>conflicted</code></strong> <span class="citation" data-cites="conflicted">(<a href="#ref-conflicted" role="doc-biblioref">Wickham 2023</a>)</span> package comes in. <strong><code>Conflicted</code></strong> will warn you when you are using a function that has the same name as a function in another package. You can then choose which function to use.</p>
<p>ğŸ¬ Load the <strong><code>conflicted</code></strong> package:</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://conflicted.r-lib.org/">conflicted</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Instead of getting a warning every time you are using a function that has a function with the same name in another package, we can declare a preference for one function over another. This is useful for the functions you use a lot or ones where you are certain you always want to use a particular function.</p>
<p>For example, to always use the <strong><code>dplyr</code></strong> version of <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> by default you can add this to the top of your script:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://conflicted.r-lib.org/reference/conflicts_prefer.html">conflicts_prefer</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will also want to ensure that we are using the <code><a href="https://generics.r-lib.org/reference/setops.html">setdiff()</a></code> function from the <strong><code>GenomicRanges</code></strong> package which is used by <strong><code>DESeq2</code></strong>.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://conflicted.r-lib.org/reference/conflicts_prefer.html">conflicts_prefer</a></span><span class="op">(</span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html">setdiff</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="import" class="level1"><h1>Import</h1>
<section id="frog-development-1" class="level2"><h2 class="anchored" data-anchor-id="frog-development-1">ğŸ¸ Frog development</h2>
<p>We need to import the S30 data that were filtered to remove genes with 4, 5 or 6 zeros and those where the total counts was less than 20.</p>
<p>ğŸ¬ Import the data from the <code>data-processed</code> folder.</p>
<p>Now go to <a href="#frog-development-2">Differential Expression Analysis</a>.</p>
</section><section id="arabidopisis-1" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis-1">ğŸ„ <em>Arabidopisis</em>
</h2>
<p>We need to import the wildtype data that were filtered to remove genes with 3 or 4 zeros and those where the total counts was less than 20.</p>
<p>ğŸ¬ Import the data from the <code>data-processed</code> folder.</p>
<p>Now go to <a href="#arabidopisis-2">Differential Expression Analysis</a>.</p>
</section><section id="leishmania-1" class="level2"><h2 class="anchored" data-anchor-id="leishmania-1">ğŸ’‰ <em>Leishmania</em>
</h2>
<p>We need to import the procyclic- and metacyclic-promastigote data that were filtered to remove genes with 4, 5 or 6 zeros and those where the total counts was less than 20.</p>
<p>ğŸ¬ Import the data from the <code>data-processed</code> folder.</p>
<p>Now go to <a href="#leishmania-2">Differential Expression Analysis</a>.</p>
</section><section id="stem-cells-1" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-1">ğŸ­ Stem cells</h2>
<p>We need to import the progenitor and HSPC cell data that were combined.</p>
<p>ğŸ¬ Import the data from the <code>data-processed</code> folder..</p>
<p>Now go to <a href="#stem-cells-2">Differential Expression Analysis</a>.</p>
</section></section><section id="differential-expression-analysis" class="level1"><h1>Differential Expression Analysis</h1>
<section id="frog-development-2" class="level2"><h2 class="anchored" data-anchor-id="frog-development-2">ğŸ¸ Frog development</h2>
<p>These are the steps we will take</p>
<ol type="1">
<li>Find the genes that are expressed in only one treatment group.</li>
<li>Create a DESeqDataSet object. This is a special object that is used by the <strong><code>DESeq2</code></strong> package</li>
<li>Prepare the normalised counts from the DESeqDataSet object.</li>
<li>Do differential expression analysis on the genes. This needs to be done on the raw counts.</li>
</ol>
<p>All but the first step are done with the <strong><code>DESeq2</code></strong> package</p>
<section id="genes-expressed-in-one-treatment" class="level3"><h3 class="anchored" data-anchor-id="genes-expressed-in-one-treatment">1. Genes expressed in one treatment</h3>
<p>The genes expressed in only one treatment group are those with zeros in all three replicates in one group and non-zero values in all three replicates in the other group. For example, those shown here:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 26%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">xenbase_gene_id</th>
<th style="text-align: right;">S30_C_1</th>
<th style="text-align: right;">S30_C_2</th>
<th style="text-align: right;">S30_C_3</th>
<th style="text-align: right;">S30_F_1</th>
<th style="text-align: right;">S30_F_2</th>
<th style="text-align: right;">S30_F_3</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">XB-GENE-1018260</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">16</td>
</tr>
<tr class="even">
<td style="text-align: left;">XB-GENE-17330117</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">17</td>
</tr>
<tr class="odd">
<td style="text-align: left;">XB-GENE-17332184</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">6</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We will use <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> to find these genes.</p>
<p>ğŸ¬ Find the genes that are expressed only in the FGF-treated group:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">s30_fgf_only</span> <span class="op">&lt;-</span> <span class="va">s30_filtered</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">S30_C_1</span> <span class="op">==</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_C_2</span> <span class="op">==</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_C_3</span> <span class="op">==</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_F_1</span> <span class="op">&gt;</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_F_2</span> <span class="op">&gt;</span> <span class="fl">0</span>, </span>
<span>         <span class="va">S30_F_3</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>â“ How many genes are expressed only in the FGF-treated group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 26 genes expressed only in the FGF-treated group. -->
<p>ğŸ¬ Now you find any genes that are expressed only in the control group.</p>
<p>â“ How many genes are expressed only in the control group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 0 genes expressed only in the control group. -->
<p>â“ Do the results make sense to you in light of what you know about the biology?</p>
<!-- #---THINKING ANSWER--- -->
<!-- Since FGF is a growth factor, it is likely to induce the expression  -->
<!-- of genes. Therefore, it is not surprising to me that there are no  -->
<!-- genes expressed in the control group but not in the FGF-treated  -->
<!-- group. However, I'm not an expert in developmental biology. -->
<!-- Domain expertise is important when analysing data so  -->
<!-- discuss your results with Betsy and your group. -->
<p>ğŸ¬ Write all the genes that are expressed one group only to file (saved in <code>results</code>)</p>
</section><section id="create-deseqdataset-object" class="level3"><h3 class="anchored" data-anchor-id="create-deseqdataset-object">2. Create DESeqDataSet object</h3>
<p>ğŸ¬ Load the DESeq2 package:</p>
<p>A DEseqDataSet object is a custom data type that is used by <strong><code>DESeq2</code></strong>. Custom data types are common in the Bioconductor<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> packages. They are used to store data in a way that is useful for the analysis. These data types typically have data, transformed data, metadata and experimental designs within them.</p>
<p>To create a DESeqDataSet object, we need to provide three things:</p>
<ul>
<li>The raw counts - these are in <code>s30_filtered</code>
</li>
<li>The meta data which gives information about the samples and which treatment groups they belong to</li>
<li>A design matrix which captures the design of the statistical model.</li>
</ul>
<p>The counts must in a <em>matrix</em> rather than a dataframe. Unlike a dataframe, a matrix has columns of all the same type. That is, it will contain only the counts. The gene ids are given as row names rather than a column. The <code><a href="https://rdrr.io/r/base/matrix.html">matrix()</a></code> function will create a matrix from a dataframe of columns of the same type and the <code><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">select()</a></code> function can be used to remove the gene ids column.</p>
<p>ğŸ¬ Create a matrix of the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">s30_count_mat</span> <span class="op">&lt;-</span> <span class="va">s30_filtered</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">xenbase_gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Add the gene ids as row names to the matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># add the row names to the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">s30_count_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">s30_filtered</span><span class="op">$</span><span class="va">xenbase_gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You might want to view the matrix (click on it in your environment pane).</p>
<p>The metadata are in a file, <a href="meta/frog_meta_data.txt">frog_meta_data.txt</a>. This is a tab-delimited file. The first column is the sample name and the other columns give the â€œtreatmentsâ€. In this case, the treatments stage (with three levels) and treatment (with two levels).</p>
<p>ğŸ¬ Make a folder called <code>meta</code> and save the file to it.</p>
<p>ğŸ¬ Read the metadata into a dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_table.html">read_table</a></span><span class="op">(</span><span class="st">"meta/frog_meta_data.txt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Examine the resulting dataframe.</p>
<p>We need to add the sample names as row names to the metadata dataframe. This is because the DESeqDataSet object will use the row names to match the samples in the metadata to the samples in the counts matrix.</p>
<p>ğŸ¬ Add the sample names as row names to the metadata dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are dealing only with the S30 data so we need to remove the samples that are not in the S30 data.</p>
<p>ğŸ¬ Filter the metadata to keep only the S30 information:</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta_s30</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">stage</span> <span class="op">==</span> <span class="st">"stage_30"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now create the DESeqDataSet object. The design formula describes the statistical model. You should recognise the form from previous work. The <code>~</code> can be read as â€œexplain byâ€ and on its right hand side are the explanatory variables. That is, the model is counts explained by treatment and sibling_rep. We are interested in the difference between the treatments but we include <code>sibling_rep</code> to account for the fact that the data are paired. The condition of interest should go at the end of the design formula.</p>
<p>Note that:</p>
<ul>
<li>The names of the columns in the count matrix have to exactly match the names of the rows in the metadata dataframe. They also need to be in the same order.</li>
<li>The names of the explanatory variables in the design formula have to match the names of columns in the metadata.</li>
</ul>
<p>ğŸ¬ Create the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="va">s30_count_mat</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">meta_s30</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span> <span class="va">sibling_rep</span> <span class="op">+</span> <span class="va">treatment</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The warning â€œWarning: some variables in design formula are characters, converting to factorsâ€ just means that the variable type of treatment and sibling_rep in the metadata dataframe are â€œcharâ€ and they have been converted into the factors.</p>
<p>To help you understand what the <code>DESeqDataSet</code> object we have called <code>dds</code> contains, we can look its contents</p>
<p>The counts are in <code>dds@assays@data@listData[["counts"]]</code> and the metadata are in <code>dds@colData</code> but the easiest way to see them is to use the <code>counts()</code> and <code>colData()</code> functions from the <strong><code>DESeq2</code></strong> package.</p>
<p>ğŸ¬ View the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should be able to see that this is the same as in <code>s30_count_mat</code>.</p>
<p>ğŸ¬ View the column information:</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 3 columns
              stage treatment sibling_rep
        &lt;character&gt;  &lt;factor&gt;    &lt;factor&gt;
S30_C_1    stage_30   control       one  
S30_C_2    stage_30   control       two  
S30_C_3    stage_30   control       three
S30_F_1    stage_30   FGF           one  
S30_F_2    stage_30   FGF           two  
S30_F_3    stage_30   FGF           three</code></pre>
</div>
</div>
<p>You should be able to see this is the same as in <code>meta_s30</code>.</p>
</section><section id="prepare-the-normalised-counts" class="level3"><h3 class="anchored" data-anchor-id="prepare-the-normalised-counts">3. Prepare the normalised counts</h3>
<p>The normalised counts are the counts that have been transformed to account for the library size (i.e., the total number of reads in a sample) and the gene length. We have to first estimate the normalisation factors and store them in the DESeqDataSet object and then we can get the normalised counts.</p>
<p>ğŸ¬ Estimate the factors for normalisation and store them in the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Look at the factors (just for information):</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  S30_C_1   S30_C_2   S30_C_3   S30_F_1   S30_F_2   S30_F_3 
0.8812200 0.9454600 1.2989886 1.0881870 1.0518961 0.8322894 </code></pre>
</div>
</div>
<p>The normalised counts will be useful to use later. To get the normalised counts we again used the <code>counts()</code> function but this time we use the <code>normalized=TRUE</code> argument.</p>
<p>ğŸ¬ Save the normalised to a matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">normalised_counts</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Make a dataframe of the normalised counts, adding a column for the gene ids at the same time:</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">s30_normalised_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">normalised_counts</span>,</span>
<span>                                    xenbase_gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">normalised_counts</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="differential-expression-analysis-1" class="level3"><h3 class="anchored" data-anchor-id="differential-expression-analysis-1">4. Differential expression analysis</h3>
<p>We use the <code>DESeq()</code> function to do the differential expression analysis. This function fits the statistical model to the data and then uses the model to calculate the significance of the difference between the treatments. It again stores the results in the DESseqDataSet object. Note that the differential expression needs the raw (unnormalised counts) as it does its own normalisation as part of the process.</p>
<p>ğŸ¬ Run the differential expression analysis and store the results in the same object:</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function will take only a few moments to run on this data but can take longer for bigger datasets.</p>
<p>We need to define the contrasts we want to test. We want to test the difference between the treatments so we will define the contrast as <code>FGF</code> and <code>control</code>.</p>
<p>ğŸ¬ Define the contrast:</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">contrast_fgf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"treatment"</span>, <span class="st">"FGF"</span>, <span class="st">"control"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that <code>treatment</code> is the name of the column in the metadata dataframe and <code>FGF</code> and <code>control</code> are the names of the levels in the <code>treatment</code> column. By putting them in the order <code>FGF</code> , <code>control</code> we are saying the fold change will be FGF / control. This means:</p>
<ul>
<li>positive log fold changes indicate FGF &gt; control and</li>
<li>negative log fold changes indicates control &gt; FGF.</li>
</ul>
<p>If we had put them in the order <code>control</code>, <code>FGF</code> we would have the reverse.</p>
<p>ğŸ¬ Extract the results from the DESseqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results_fgf</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                       contrast <span class="op">=</span> <span class="va">contrast_fgf</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This will give us the log<sub>2</sub> fold change, the <em>p</em>-value and the adjusted <em>p</em>-value for the comparison between the control and the FGF-treatment for each gene.</p>
<p>ğŸ¬ Put the results in a dataframe and add the gene ids as a column:</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">s30_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">results_fgf</span>,</span>
<span>                          xenbase_gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">results_fgf</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is useful to have the normalised counts and the statistical results in one dataframe.</p>
<p>ğŸ¬ Merge the two dataframes:</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># merge the results with the normalised counts</span></span>
<span><span class="va">s30_results</span> <span class="op">&lt;-</span> <span class="va">s30_normalised_counts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">s30_results</span>, by <span class="op">=</span> <span class="st">"xenbase_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#frog-development-3">Add gene information</a>.</p>
</section></section><section id="arabidopisis-2" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis-2">ğŸ„ <em>Arabidopisis</em>
</h2>
<p>These are the steps we will take</p>
<ol type="1">
<li>Find the genes that are expressed in only one treatment group.</li>
<li>Create a DESeqDataSet object. This is a special object that is used by the <strong><code>DESeq2</code></strong> package</li>
<li>Prepare the normalised counts from the DESeqDataSet object.</li>
<li>Do differential expression analysis on the genes. This needs to be done on the raw counts.</li>
</ol>
<p>All but the first step are done with the <strong><code>DESeq2</code></strong> package</p>
<section id="genes-expressed-in-one-treatment-1" class="level3"><h3 class="anchored" data-anchor-id="genes-expressed-in-one-treatment-1">1. Genes expressed in one treatment</h3>
<p>The genes expressed in only one treatment group are those with zeros in both replicates in one group and non-zero values in both replicates in the other group. For example, those shown here:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 11%">
<col style="width: 22%">
<col style="width: 22%">
<col style="width: 22%">
<col style="width: 22%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">gene_id</th>
<th style="text-align: right;">SRX028956_wild_suf</th>
<th style="text-align: right;">SRX028957_wild_def</th>
<th style="text-align: right;">SRX028960_wild_suf</th>
<th style="text-align: right;">SRX028961_wild_def</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">AT1G04513</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">25</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">AT1G22610</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">AT1G26290</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">23</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">AT1G59810</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">16</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">AT2G44130</td>
<td style="text-align: right;">28</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We will use <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> to find these genes.</p>
<p>ğŸ¬ Find the genes that are expressed only in the sufficient copper group:</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_suf_only</span> <span class="op">&lt;-</span> <span class="va">wild_filtered</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">SRX028961_wild_def</span> <span class="op">==</span> <span class="fl">0</span>,</span>
<span>         <span class="va">SRX028957_wild_def</span> <span class="op">==</span> <span class="fl">0</span>,</span>
<span>         <span class="va">SRX028960_wild_suf</span> <span class="op">&gt;</span> <span class="fl">0</span>,</span>
<span>         <span class="va">SRX028956_wild_suf</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>â“ How many genes are expressed only in the sufficient copper group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 19 genes expressed only in the sufficient copper group. -->
<p>ğŸ¬ Now you find any genes that are expressed only in the deficient copper group.</p>
<p>â“ How many genes are expressed only in the deficient copper group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 3 genes expressed only in the deficient copper group. -->
<p>â“ Do the results make sense to you in light of what you know about the biology?</p>
<!-- #---THINKING ANSWER--- -->
<!-- This is wildtype plant so perhaps normal processes do not happen  -->
<!-- when copper is deficient and because the plant does not have 
<!-- adaptations for deficient copper, no compensatory processes kick in. <!-- <!-- However, I'm not an expert in Arabidopsis, or plants! -->
<!-- Domain expertise is important when analysing data so  -->
<!-- discuss your results with Liz and your group. -->
<p>ğŸ¬ Write all the genes that are expressed one group only to file (saved in <code>results</code>)</p>
</section><section id="create-deseqdataset-object-1" class="level3"><h3 class="anchored" data-anchor-id="create-deseqdataset-object-1">2. Create DESeqDataSet object</h3>
<p>ğŸ¬ Load the DESeq2 package:</p>
<p>A DEseqDataSet object is a custom data type that is used by <strong><code>DESeq2</code></strong>. Custom data types are common in the Bioconductor<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> packages. They are used to store data in a way that is useful for the analysis. These data types typically have data, transformed data, metadata and experimental designs within them.</p>
<p>To create a DESeqDataSet object, we need to provide three things:</p>
<ul>
<li>The raw counts - these are in <code>wild_filtered</code>
</li>
<li>The meta data which gives information about the samples and which treatment groups they belong to</li>
<li>A design matrix which captures the design of the statistical model.</li>
</ul>
<p>The counts must in a <em>matrix</em> rather than a dataframe. Unlike a dataframe, a matrix has columns of all the same type. That is, it will contain only the counts. The gene ids are given as row names rather than a column. The <code><a href="https://rdrr.io/r/base/matrix.html">matrix()</a></code> function will create a matrix from a dataframe of columns of the same type and the <code><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">select()</a></code> function can be used to remove the gene ids column.</p>
<p>ğŸ¬ Create a matrix of the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_count_mat</span> <span class="op">&lt;-</span> <span class="va">wild_filtered</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Add the gene ids as row names to the matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># add the row names to the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">wild_count_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">wild_filtered</span><span class="op">$</span><span class="va">gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You might want to view the matrix (click on it in your environment pane).</p>
<p>The metadata are in a file, <a href="meta/arab_meta_data.txt">arab_meta_data.txt</a>. This is a tab-delimited file. The first column is the sample name and the other columns give the â€œtreatmentsâ€. In this case, the treatments genotype (with two levels) and copper (with two levels).</p>
<p>ğŸ¬ Make a folder called <code>meta</code> and save the file to it.</p>
<p>ğŸ¬ Read the metadata into a dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_table.html">read_table</a></span><span class="op">(</span><span class="st">"meta/arab_meta_data.txt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Examine the resulting dataframe.</p>
<p>We need to add the sample names as row names to the metadata dataframe. This is because the DESeqDataSet object will use the row names to match the samples in the metadata to the samples in the counts matrix.</p>
<p>ğŸ¬ Add the sample names as row names to the metadata dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are dealing only with the wild data so we need to remove the samples that are not in the wild data.</p>
<p>ğŸ¬ Filter the metadata to keep only the wild information:</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta_wild</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">genotype</span> <span class="op">==</span> <span class="st">"wt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now create the DESeqDataSet object. The design formula describes the statistical model. You should recognise the form from previous work. The <code>~</code> can be read as â€œexplain byâ€ and on its right hand side are the explanatory variables. That is, the model is counts explained by copper status.</p>
<p>Note that:</p>
<ul>
<li>The names of the columns in the count matrix have to exactly match the names of the rows in the metadata dataframe. They also need to be in the same order.</li>
<li>The names of the explanatory variables in the design formula have to match the names of columns in the metadata.</li>
</ul>
<p>ğŸ¬ Create the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span><span class="va">wild_count_mat</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">meta_wild</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span> <span class="va">copper</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The warning â€œWarning: some variables in design formula are characters, converting to factorsâ€ just means that the variable type of copper in the metadata dataframe is â€œcharâ€ and it has been converted into a factor type.</p>
<p>To help you understand what the <code>DESeqDataSet</code> object we have called <code>dds</code> contains, we can look its contents</p>
<p>The counts are in <code>dds@assays@data@listData[["counts"]]</code> and the metadata are in <code>dds@colData</code> but the easiest way to see them is to use the <code>counts()</code> and <code>colData()</code> functions from the <strong><code>DESeq2</code></strong> package.</p>
<p>ğŸ¬ View the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should be able to see that this is the same as in <code>wild_count_mat</code>.</p>
<p>ğŸ¬ View the column information:</p>
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 2 columns
                      genotype     copper
                   &lt;character&gt;   &lt;factor&gt;
SRX028956_wild_suf          wt sufficient
SRX028957_wild_def          wt deficient 
SRX028960_wild_suf          wt sufficient
SRX028961_wild_def          wt deficient </code></pre>
</div>
</div>
<p>You should be able to see this is the same as in <code>meta_wild</code>.</p>
</section><section id="prepare-the-normalised-counts-1" class="level3"><h3 class="anchored" data-anchor-id="prepare-the-normalised-counts-1">3. Prepare the normalised counts</h3>
<p>The normalised counts are the counts that have been transformed to account for the library size (i.e., the total number of reads in a sample) and the gene length. We have to first estimate the normalisation factors and store them in the DESeqDataSet object and then we can get the normalised counts.</p>
<p>ğŸ¬ Estimate the factors for normalisation and store them in the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Look at the factors (just for information):</p>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SRX028956_wild_suf SRX028957_wild_def SRX028960_wild_suf SRX028961_wild_def 
         0.8200020          0.4653024          2.3002428          1.1965924 </code></pre>
</div>
</div>
<p>The normalised counts will be useful to use later. To get the normalised counts we again used the <code>counts()</code> function but this time we use the <code>normalized=TRUE</code> argument.</p>
<p>ğŸ¬ Save the normalised to a matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">normalised_counts</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Make a dataframe of the normalised counts, adding a column for the gene ids at the same time:</p>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_normalised_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">normalised_counts</span>,</span>
<span>                                    gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">normalised_counts</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="differential-expression-analysis-2" class="level3"><h3 class="anchored" data-anchor-id="differential-expression-analysis-2">4. Differential expression analysis</h3>
<p>We use the <code>DESeq()</code> function to do the differential expression analysis. This function fits the statistical model to the data and then uses the model to calculate the significance of the difference between the treatments. It again stores the results in the DESseqDataSet object. Note that the differential expression needs the raw (unnormalised counts) as it does its own normalisation as part of the process.</p>
<p>ğŸ¬ Run the differential expression analysis and store the results in the same object:</p>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function will take only a few moments to run on this data but can take longer for bigger datasets.</p>
<p>We need to define the contrasts we want to test. We want to test the difference between the treatments so we will define the contrast as <code>sufficient</code> and <code>deficient</code>.</p>
<p>ğŸ¬ Define the contrast:</p>
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">contrast_suf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"copper"</span>, <span class="st">"sufficient"</span>, <span class="st">"deficient"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that <code>copper</code> is the name of the column in the metadata dataframe and <code>sufficient</code> and <code>deficient</code> are the names of the levels in the <code>copper</code> column. By putting them in the order <code>sufficient</code> , <code>deficient</code> we are saying the fold change will be sufficient / deficient. This means:</p>
<ul>
<li>positive log fold changes indicate sufficient &gt; deficient and</li>
<li>negative log fold changes indicates deficient &gt; sufficient.</li>
</ul>
<p>If we had put them in the order <code>deficient</code>, <code>sufficient</code> we would have the reverse.</p>
<p>ğŸ¬ Extract the results from the DESseqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results_suf</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                       contrast <span class="op">=</span> <span class="va">contrast_suf</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This will give us the log<sub>2</sub> fold change, the <em>p</em>-value and the adjusted <em>p</em>-value for the comparison between the sufficient- and<br>
deficient-copper for each gene.</p>
<p>ğŸ¬ Put the results in a dataframe and add the gene ids as a column:</p>
<div class="cell">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">results_suf</span>,</span>
<span>                          gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">results_suf</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is useful to have the normalised counts and the statistical results in one dataframe.</p>
<p>ğŸ¬ Merge the two dataframes:</p>
<div class="cell">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># merge the results with the normalised counts</span></span>
<span><span class="va">wild_results</span> <span class="op">&lt;-</span> <span class="va">wild_normalised_counts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">wild_results</span>, by <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#arabidopisis-3">Add gene information</a>.</p>
</section></section><section id="leishmania-2" class="level2"><h2 class="anchored" data-anchor-id="leishmania-2">ğŸ’‰ <em>Leishmania</em>
</h2>
<p>These are the steps we will take</p>
<ol type="1">
<li>Find the genes that are expressed in only one treatment group.</li>
<li>Create a DESeqDataSet object. This is a special object that is used by the <strong><code>DESeq2</code></strong> package</li>
<li>Prepare the normalised counts from the DESeqDataSet object.</li>
<li>Do differential expression analysis on the genes. This needs to be done on the raw counts.</li>
</ol>
<p>All but the first step are done with the <strong><code>DESeq2</code></strong> package</p>
<section id="genes-expressed-in-one-treatment-2" class="level3"><h3 class="anchored" data-anchor-id="genes-expressed-in-one-treatment-2">1. Genes expressed in one treatment</h3>
<p>The genes expressed in only one treatment group are those with zeros in all replicates in one group and non-zero values in all replicates in the other group.</p>
<p>We will use <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> to find these genes.</p>
<p>ğŸ¬ Find the genes that are expressed only at the procyclic-promastigote stage:</p>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_pro_only</span> <span class="op">&lt;-</span> <span class="va">pro_meta_filtered</span>  <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">lm_pro_1</span> <span class="op">&gt;</span> <span class="fl">0</span>,</span>
<span>         <span class="va">lm_pro_2</span> <span class="op">&gt;</span> <span class="fl">0</span>,</span>
<span>         <span class="va">lm_pro_3</span> <span class="op">&gt;</span> <span class="fl">0</span>,</span>
<span>         <span class="va">lm_meta_1</span> <span class="op">==</span> <span class="fl">0</span>,</span>
<span>         <span class="va">lm_meta_2</span> <span class="op">==</span> <span class="fl">0</span>,</span>
<span>         <span class="va">lm_meta_2</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>â“ How many genes are expressed only in the procyclic-promastigote stage group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 0 genes expressed only in the procyclic-promastigote stage -->
<p>ğŸ¬ Now you find any genes that are expressed only at the metacyclic stage</p>
<p>â“ How many genes are expressed only at the metacyclic stage?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 0 genes expressed only in the meta stage group. -->
<p>â“ Do the results make sense to you in light of what you know about the biology?</p>
<!-- #---THINKING ANSWER--- -->
<!-- I have no clue here! -->
<!-- Domain expertise is important when analysing data so  -->
<!-- discuss your results with Pegine and your group. -->
<p>ğŸ¬ Write all the genes that are expressed one group only to file (saved in <code>results</code>)</p>
</section><section id="create-deseqdataset-object-2" class="level3"><h3 class="anchored" data-anchor-id="create-deseqdataset-object-2">2. Create DESeqDataSet object</h3>
<p>ğŸ¬ Load the DESeq2 package:</p>
<p>A DEseqDataSet object is a custom data type that is used by <strong><code>DESeq2</code></strong>. Custom data types are common in the Bioconductor<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> packages. They are used to store data in a way that is useful for the analysis. These data types typically have data, transformed data, metadata and experimental designs within them.</p>
<p>To create a DESeqDataSet object, we need to provide three things:</p>
<ul>
<li>The raw counts - these are in <code>pro_meta_filtered</code>
</li>
<li>The meta data which gives information about the samples and which treatment groups they belong to</li>
<li>A design matrix which captures the design of the statistical model.</li>
</ul>
<p>The counts must in a <em>matrix</em> rather than a dataframe. Unlike a dataframe, a matrix has columns of all the same type. That is, it will contain only the counts. The gene ids are given as row names rather than a column. The <code><a href="https://rdrr.io/r/base/matrix.html">matrix()</a></code> function will create a matrix from a dataframe of columns of the same type and the <code><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">select()</a></code> function can be used to remove the gene ids column.</p>
<p>ğŸ¬ Create a matrix of the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_count_mat</span> <span class="op">&lt;-</span> <span class="va">pro_meta_filtered</span>  <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Add the gene ids as row names to the matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># add the row names to the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">pro_meta_count_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pro_meta_filtered</span><span class="op">$</span><span class="va">gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You might want to view the matrix (click on it in your environment pane).</p>
<p>The metadata are in a file, <a href="meta/leish_meta_data.txt">leish_meta_data.txt</a>. This is a tab-delimited file. The first column is the sample name and the other columns give the â€œtreatmentsâ€. In this case, the treatment is stage (with three levels).</p>
<p>ğŸ¬ Make a folder called <code>meta</code> and save the file to it.</p>
<p>ğŸ¬ Read the metadata into a dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_table.html">read_table</a></span><span class="op">(</span><span class="st">"meta/leish_meta_data.txt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Examine the resulting dataframe.</p>
<p>We need to add the sample names as row names to the metadata dataframe. This is because the DESeqDataSet object will use the row names to match the samples in the metadata to the samples in the counts matrix.</p>
<p>ğŸ¬ Add the sample names as row names to the metadata dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are dealing only with the wild data so we need to remove the samples that are not in the wild data.</p>
<p>ğŸ¬ Filter the metadata to keep only the procyclic and metacyclic information:</p>
<div class="cell">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta_pro_meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">stage</span> <span class="op">!=</span> <span class="st">"amastigotes"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now create the DESeqDataSet object. The design formula describes the statistical model. You should recognise the form from previous work. The <code>~</code> can be read as â€œexplain byâ€ and on its right hand side are the explanatory variables. That is, the model is counts explained by stage status.</p>
<p>Note that:</p>
<ul>
<li>The names of the columns in the count matrix have to exactly match the names of the rows in the metadata dataframe. They also need to be in the same order.</li>
<li>The names of the explanatory variables in the design formula have to match the names of columns in the metadata.</li>
</ul>
<p>ğŸ¬ Create the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span><span class="va">pro_meta_count_mat</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">meta_pro_meta</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span> <span class="va">stage</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The warning â€œWarning: some variables in design formula are characters, converting to factorsâ€ just means that the variable type of stage in the metadata dataframe is â€œcharâ€ and it has been converted into a factor type.</p>
<p>To help you understand what the <code>DESeqDataSet</code> object we have called <code>dds</code> contains, we can look its contents</p>
<p>The counts are in <code>dds@assays@data@listData[["counts"]]</code> and the metadata are in <code>dds@colData</code> but the easiest way to see them is to use the <code>counts()</code> and <code>colData()</code> functions from the <strong><code>DESeq2</code></strong> package.</p>
<p>ğŸ¬ View the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should be able to see that this is the same as in <code>pro_meta_count_mat</code>.</p>
<p>ğŸ¬ View the column information:</p>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 2 columns
               stage replicate
            &lt;factor&gt; &lt;numeric&gt;
lm_pro_1  procyclic          1
lm_pro_2  procyclic          2
lm_pro_3  procyclic          3
lm_meta_1 metacyclic         1
lm_meta_2 metacyclic         2
lm_meta_3 metacyclic         3</code></pre>
</div>
</div>
<p>You should be able to see this is the same as in <code>meta_pro_meta</code>.</p>
</section><section id="prepare-the-normalised-counts-2" class="level3"><h3 class="anchored" data-anchor-id="prepare-the-normalised-counts-2">3. Prepare the normalised counts</h3>
<p>The normalised counts are the counts that have been transformed to account for the library size (i.e., the total number of reads in a sample) and the gene length. We have to first estimate the normalisation factors and store them in the DESeqDataSet object and then we can get the normalised counts.</p>
<p>ğŸ¬ Estimate the factors for normalisation and store them in the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Look at the factors (just for information):</p>
<div class="cell">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> lm_pro_1  lm_pro_2  lm_pro_3 lm_meta_1 lm_meta_2 lm_meta_3 
1.3029351 0.9158157 0.9943186 0.7849299 0.8443586 1.3250409 </code></pre>
</div>
</div>
<p>The normalised counts will be useful to use later. To get the normalised counts we again used the <code>counts()</code> function but this time we use the <code>normalized=TRUE</code> argument.</p>
<p>ğŸ¬ Save the normalised to a matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">normalised_counts</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Make a dataframe of the normalised counts, adding a column for the gene ids at the same time:</p>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_normalised_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">normalised_counts</span>,</span>
<span>                                    gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">normalised_counts</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="differential-expression-analysis-3" class="level3"><h3 class="anchored" data-anchor-id="differential-expression-analysis-3">4. Differential expression analysis</h3>
<p>We use the <code>DESeq()</code> function to do the differential expression analysis. This function fits the statistical model to the data and then uses the model to calculate the significance of the difference between the treatments. It again stores the results in the DESseqDataSet object. Note that the differential expression needs the raw (unnormalised counts) as it does its own normalisation as part of the process.</p>
<p>ğŸ¬ Run the differential expression analysis and store the results in the same object:</p>
<div class="cell">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function will take only a few moments to run on this data but can take longer for bigger datasets.</p>
<p>We need to define the contrasts we want to test. We want to test the difference between the treatments so we will define the contrast as <code>procyclic</code> and <code>metacyclic</code>.</p>
<p>ğŸ¬ Define the contrast:</p>
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">contrast_pro_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stage"</span>, <span class="st">"procyclic"</span>, <span class="st">"metacyclic"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that <code>stage</code> is the name of the column in the metadata dataframe and <code>procyclic</code> and <code>metacyclic</code> are the names of the levels in the <code>stage</code> column. By putting them in the order <code>procyclic</code> , <code>metacyclic</code> we are saying the fold change will be procyclic / metacyclic. This means:</p>
<ul>
<li>positive log fold changes indicate procyclic &gt; metacyclic and</li>
<li>negative log fold changes indicates metacyclic &gt; procyclic.</li>
</ul>
<p>If we had put them in the order <code>metacyclic</code>, <code>procyclic</code> we would have the reverse.</p>
<p>ğŸ¬ Extract the results from the DESseqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results_pro_meta</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                       contrast <span class="op">=</span> <span class="va">contrast_pro_meta</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This will give us the log<sub>2</sub> fold change, the <em>p</em>-value and the adjusted <em>p</em>-value for the comparison between procyclic and metacyclic stage for each gene</p>
<p>ğŸ¬ Put the results in a dataframe and add the gene ids as a column:</p>
<div class="cell">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">results_pro_meta</span>,</span>
<span>                          gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">results_pro_meta</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is useful to have the normalised counts and the statistical results in one dataframe.</p>
<p>ğŸ¬ Merge the two dataframes:</p>
<div class="cell">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># merge the results with the normalised counts</span></span>
<span><span class="va">pro_meta_results</span> <span class="op">&lt;-</span> <span class="va">pro_meta_normalised_counts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">pro_meta_results</span>, by <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#leishmania-3">Add gene information</a>.</p>
</section></section><section id="stem-cells-2" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-2">ğŸ­ Stem cells</h2>
<p>These are the steps we will take</p>
<ol type="1">
<li>Find the genes that are expressed in only one cell type (the prog or the hspc).</li>
<li>Prepare the data for differential expression analysis with the <strong><code>scran</code></strong> package.</li>
<li>Do differential expression analysis on the genes using the <strong><code>scran</code></strong> package. This needs to be done on the logged normalised counts.</li>
</ol>
<section id="genes-expressed-in-one-cell-type" class="level3"><h3 class="anchored" data-anchor-id="genes-expressed-in-one-cell-type">1. Genes expressed in one cell type</h3>
<p>The genes expressed in only cell type are those with zeros in all the cells of the other type. We can find these by summing the expression values for each gene across the cells of one type and filtering for those that are zero.</p>
<p>To do row wise aggregates such as the sum across rows we can use the <code><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise()</a></code> function. <code><a href="https://dplyr.tidyverse.org/reference/c_across.html">c_across()</a></code> allows us to use the colon notation to select columns. This is very useful when you have a lot of columns because it would be annoying to have to list all of them. <code>HSPC_001:HSPC_852</code> means all the columns from <code>HSPC_001</code> to <code>HSPC_852</code>.</p>
<p>ğŸ¬ Find the genes expressed only in progenitor cells (<em>i.e.</em>, those that are 0 in every HSPC cell):</p>
<div class="cell">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/c_across.html">c_across</a></span><span class="op">(</span><span class="va">HSPC_001</span><span class="op">:</span><span class="va">HSPC_852</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 0 Ã— 1,500
# Rowwise: 
# â„¹ 1,500 variables: ensembl_gene_id &lt;chr&gt;, HSPC_001 &lt;dbl&gt;, HSPC_002 &lt;dbl&gt;,
#   HSPC_003 &lt;dbl&gt;, HSPC_004 &lt;dbl&gt;, HSPC_006 &lt;dbl&gt;, HSPC_008 &lt;dbl&gt;,
#   HSPC_009 &lt;dbl&gt;, HSPC_011 &lt;dbl&gt;, HSPC_012 &lt;dbl&gt;, HSPC_014 &lt;dbl&gt;,
#   HSPC_015 &lt;dbl&gt;, HSPC_016 &lt;dbl&gt;, HSPC_017 &lt;dbl&gt;, HSPC_018 &lt;dbl&gt;,
#   HSPC_020 &lt;dbl&gt;, HSPC_021 &lt;dbl&gt;, HSPC_022 &lt;dbl&gt;, HSPC_023 &lt;dbl&gt;,
#   HSPC_024 &lt;dbl&gt;, HSPC_025 &lt;dbl&gt;, HSPC_026 &lt;dbl&gt;, HSPC_027 &lt;dbl&gt;,
#   HSPC_028 &lt;dbl&gt;, HSPC_030 &lt;dbl&gt;, HSPC_031 &lt;dbl&gt;, HSPC_033 &lt;dbl&gt;, â€¦</code></pre>
</div>
</div>
<p>We already know from last weekâ€™s work that there are no genes that are zero across all the cells (both types). If we did not know that, we would need to add <code>|&gt; filter(sum(c_across(Prog_001:Prog_852)) != 0)</code></p>
<p>meaning zero in all the HSPC but not zero in all the Prog</p>
<p>â“ How many genes are expressed only in the progenitor cells only</p>
<!-- #---THINKING ANSWER--- -->
<!-- No genes are expressed only in the progenitor cells. -->
<p>ğŸ¬ Now you find any genes that are expressed only in the HSPC cells.</p>
<p>â“ How many genes are expressed only in the HSPC cells?</p>
<!-- #---THINKING ANSWER--- -->
<!-- No genes genes are expressed only in the HSPC cells? -->
<p>â“ Do the results make sense to you in light of what you know about the biology?</p>
<!-- #---THINKING ANSWER--- -->
<!-- I'm not sure what to expect here. However, when there are many  -->
<!-- "replicates" it is less likely that a gene will be expressed  -->
<!-- in only one group. It's more likely that a gene will be -->
<!-- expressed in both groups but at different levels. -->
<!-- Domain expertise is important when analysing data so  -->
<!-- discuss your results with Jillian and your group. -->
<p>ğŸ¬ Write all the genes that are expressed one cell type only to file (saved in <code>results</code>)</p>
</section><section id="prepare-the-data-for-analysis-with-scran" class="level3"><h3 class="anchored" data-anchor-id="prepare-the-data-for-analysis-with-scran">2. Prepare the data for analysis with <strong><code>scran</code></strong>
</h3>
<p><strong><code>scran</code></strong> can use a matrix or a dataframe of counts but these must be log normalised counts. If using a dataframe, the columns must only contain the expression values (not the gene ids). The rows can be named to retain the gene ids.</p>
<p><code>hspc_prog</code> is a dataframe so we will use the ensembl gene ids to name the rows and remove the gene ids from the dataframe.</p>
<p>ğŸ¬ Add the gene ids as the row names:</p>
<div class="cell">
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog</span> <span class="op">&lt;-</span> <span class="va">hspc_prog</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Like <strong><code>DESeq2</code></strong>, <strong><code>scran</code></strong> needs metadata to define which columns were in which group. Instead of having this is a file, we will create a vector that indicates which column belongs to which cell type.</p>
<p>ğŸ¬ Create a vector that indicates which column belongs to which cell type:</p>
<div class="cell">
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">n_hspc</span> <span class="op">&lt;-</span> <span class="fl">701</span></span>
<span><span class="va">n_prog</span> <span class="op">&lt;-</span> <span class="fl">798</span></span>
<span></span>
<span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"hspc"</span>,<span class="st">"prog"</span><span class="op">)</span>, </span>
<span>                 times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">n_hspc</span>, <span class="va">n_prog</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The number of times each cell type is repeated is the number of cells of that type. Do check that the length of the <code>cell_type</code> vector is the same as the number of columns in the <code>hspc_prog</code> dataframe.</p>
</section><section id="differential-expression-analysis-4" class="level3"><h3 class="anchored" data-anchor-id="differential-expression-analysis-4">3. Differential expression analysis</h3>
<p>ğŸ¬ Load the <strong><code>scran</code></strong> package:</p>
<p>Differential expression is carried out with the <code>findMarkers()</code> function. It takes two arguments. The first argument is the dataframe containing the data and the second argument is the vector indicating which columns are in which cell type. Make sure that the order of the cell types in the vector is appropriate for the order of the columns in the dataframe.</p>
<p>ğŸ¬ Run the differential expression analysis:</p>
<div class="cell">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results_hspc_prog</span> <span class="op">&lt;-</span> <span class="fu">findMarkers</span><span class="op">(</span><span class="va">hspc_prog</span>, </span>
<span>                             <span class="va">cell_type</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output is a list object which, rather unnecessarily, includes two dataframes. This is not really necessary as the results are the same except for the fold change having a different sign.</p>
<ul>
<li>
<p>The dataframe <code>results_hspc_prog$prog</code> is log prog - log hspc (<em>i.e.</em>,Prog/HSPC). This means:</p>
<ul>
<li>Positive fold change: prog is higher than hspc</li>
<li>Negative fold change: hspc is higher than prog</li>
</ul>
</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>The results_hspc_prog$prog dataframe</caption>
<colgroup>
<col style="width: 24%">
<col style="width: 5%">
<col style="width: 10%">
<col style="width: 5%">
<col style="width: 17%">
<col style="width: 13%">
<col style="width: 24%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Top</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">FDR</th>
<th style="text-align: right;">summary.logFC</th>
<th style="text-align: right;">logFC.hspc</th>
<th style="text-align: left;">ensembl_gene_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000028639</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1.596910</td>
<td style="text-align: right;">1.596910</td>
<td style="text-align: left;">ENSMUSG00000028639</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000024053</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.035165</td>
<td style="text-align: right;">3.035165</td>
<td style="text-align: left;">ENSMUSG00000024053</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000041329</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.261056</td>
<td style="text-align: right;">3.261056</td>
<td style="text-align: left;">ENSMUSG00000041329</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000030336</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-2.146491</td>
<td style="text-align: right;">-2.146491</td>
<td style="text-align: left;">ENSMUSG00000030336</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000016494</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.056730</td>
<td style="text-align: right;">-3.056730</td>
<td style="text-align: left;">ENSMUSG00000016494</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000002808</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.000810</td>
<td style="text-align: right;">3.000810</td>
<td style="text-align: left;">ENSMUSG00000002808</td>
</tr>
</tbody>
</table>
</div>
</div>
<ul>
<li>
<p>The dataframe <code>results_hspc_prog$hspc</code> is log hspc - log prog (<em>i.e.</em>, HSPC/Prog). This means:</p>
<ul>
<li>Positive fold change: hspc is higher than prog</li>
<li>Negative fold change: prog is higher than hspc</li>
</ul>
</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>The results_hspc_prog$hspc dataframe. Notice the sign of the fold change is the other way</caption>
<colgroup>
<col style="width: 24%">
<col style="width: 5%">
<col style="width: 10%">
<col style="width: 5%">
<col style="width: 17%">
<col style="width: 13%">
<col style="width: 24%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Top</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">FDR</th>
<th style="text-align: right;">summary.logFC</th>
<th style="text-align: right;">logFC.prog</th>
<th style="text-align: left;">ensembl_gene_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000028639</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-1.596910</td>
<td style="text-align: right;">-1.596910</td>
<td style="text-align: left;">ENSMUSG00000028639</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000024053</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.035165</td>
<td style="text-align: right;">-3.035165</td>
<td style="text-align: left;">ENSMUSG00000024053</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000041329</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.261056</td>
<td style="text-align: right;">-3.261056</td>
<td style="text-align: left;">ENSMUSG00000041329</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000030336</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2.146491</td>
<td style="text-align: right;">2.146491</td>
<td style="text-align: left;">ENSMUSG00000030336</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000016494</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.056730</td>
<td style="text-align: right;">3.056730</td>
<td style="text-align: left;">ENSMUSG00000016494</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000002808</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.000810</td>
<td style="text-align: right;">-3.000810</td>
<td style="text-align: left;">ENSMUSG00000002808</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We only need one of these results dataframes and we will use the <code>prog</code> one. It does not matter which one you use but you do need keep the direction of the foldchange in mind when interpreting the results.</p>
<p>Notice that the dataframes are ordered by significance rather than the original gene order.</p>
<p>It is useful to have the normalised counts and the statistical results in one dataframe to which we will add the gene information from Ensembl. Having all the information together will make it easier to interpret the results and select genes of interest. We will need to extract the results from the list object, add the gene ids and then join with the normalised counts.</p>
<p>ğŸ¬ Extract the results dataframe from the list object and add the gene ids as a column:</p>
<div class="cell">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">results_hspc_prog</span><span class="op">$</span><span class="va">prog</span>, </span>
<span>                                ensembl_gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">results_hspc_prog</span><span class="op">$</span><span class="va">prog</span><span class="op">)</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Return the ensembl gene ids as a column to the normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog</span> <span class="op">&lt;-</span> <span class="va">hspc_prog</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Merge the results dataframe with the normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># merge the results with the normalised counts</span></span>
<span><span class="va">hspc_prog_results</span> <span class="op">&lt;-</span> <span class="va">hspc_prog_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">hspc_prog</span>, by <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#stem-cells-3">Add gene information</a>.</p>
</section></section></section><section id="add-gene-information" class="level1"><h1>Add gene information</h1>
<section id="frog-development-3" class="level2"><h2 class="anchored" data-anchor-id="frog-development-3">ğŸ¸ Frog development</h2>
<ul>
<li><p>I got the information from the <a href="https://www.xenbase.org/xenbase/static-xenbase/ftpDatafiles.jsp">Xenbase information pages</a> under Data Reports | Gene Information</p></li>
<li><p>This is listed: Xenbase Gene Product Information [readme] <a href="https://download.xenbase.org/xenbase/GenePageReports/xenbase.gpi.gz">gzipped gpi (tab separated)</a></p></li>
<li><p>Click on the readme link to see the file format and columns</p></li>
<li><p>I downloaded <a href="https://download.xenbase.org/xenbase/GenePageReports/xenbase.gpi.gz">xenbase.gpi.gz</a>, unzipped it, removed header lines and the <em>Xenopus tropicalis</em> (taxon:8364) entries and saved it as <a href="meta/xenbase_info.xlsx">xenbase_info.xlsx</a></p></li>
</ul>
<p>If you want to emulate what I did you can use the following commands in the terminal after downloading the file:</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> xenbase.gpi.gz</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> xenbase.gpi</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a><span class="ex">q</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><code>gunzip</code> unzips the file and <code>less</code> allows you to view the file. <code>q</code> quits the viewer. You will see the header lines and that the file contains both <em>Xenopus tropicalis</em> and <em>Xenopus laevis</em>. I read the file in with <code>read_tsv</code> (skipping the first header lines) then filtered out the <em>Xenopus tropicalis</em> entries, dropped some columns and saved the file as an excel file.</p>
<p>However, I have already done this for you and saved the file as <a href="meta/xenbase_info.xlsx">xenbase_info.xlsx</a> in the <code>meta</code> folder. We will import this file and join it to the results dataframe.</p>
<p>ğŸ¬ Load the <strong><code>readxl</code></strong> <span class="citation" data-cites="readxl">(<a href="#ref-readxl" role="doc-biblioref">Wickham and Bryan 2023</a>)</span> package:</p>
<div class="cell">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org">readxl</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Import the Xenbase gene information file:</p>
<div class="cell">
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_excel</a></span><span class="op">(</span><span class="st">"meta/xenbase_info.xlsx"</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should view the resulting dataframe to see what information is available. You can use <code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code> or <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code>.</p>
<p>ğŸ¬ Merge the gene information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># join the gene info with the results</span></span>
<span><span class="va">s30_results</span> <span class="op">&lt;-</span> <span class="va">s30_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">gene_info</span>, by <span class="op">=</span> <span class="st">"xenbase_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Save the results to a file:</p>
<div class="cell">
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">s30_results</span>, file <span class="op">=</span> <span class="st">"results/s30_results.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="arabidopisis-3" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis-3">ğŸ„ <em>Arabidopisis</em>
</h2>
<p><a href="https://www.ensembl.org/index.html">Ensembl</a> <span class="citation" data-cites="martin2023 birney2004">(<a href="#ref-martin2023" role="doc-biblioref">Martin et al. 2023</a>; <a href="#ref-birney2004" role="doc-biblioref">Birney et al. 2004</a>)</span>is a bioinformatics project to organise all the biological information around the sequences of large genomes. The are a large number of databases and <a href="https://www.ensembl.org/info/data/biomart/index.html">BioMart</a> <span class="citation" data-cites="smedley2009">(<a href="#ref-smedley2009" role="doc-biblioref">Smedley et al. 2009</a>)</span> provides a consistent interface to the material. There are web-based tools to use these but the R package <strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#ref-biomaRt1" role="doc-biblioref">Durinck et al. 2009</a>, <a href="#ref-biomaRt2" role="doc-biblioref">2005</a>)</span> gives you programmatic access making it easier to integrate information into R dataframes.</p>
<p>ğŸ¬ Load the <strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#ref-biomaRt1" role="doc-biblioref">Durinck et al. 2009</a>, <a href="#ref-biomaRt2" role="doc-biblioref">2005</a>)</span> package:</p>
<div class="cell">
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/grimbough/biomaRt">biomaRt</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <strong><code>biomaRt</code></strong> package includes a function to list all the available datasets</p>
<p>ğŸ¬ List the Ensembl â€œmartsâ€ available:</p>
<div class="cell">
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listEnsembl.html">listEnsemblGenomes</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              biomart                        version
1       protists_mart      Ensembl Protists Genes 60
2 protists_variations Ensembl Protists Variations 60
3          fungi_mart         Ensembl Fungi Genes 60
4    fungi_variations    Ensembl Fungi Variations 60
5        metazoa_mart       Ensembl Metazoa Genes 60
6  metazoa_variations  Ensembl Metazoa Variations 60
7         plants_mart        Ensembl Plants Genes 60
8   plants_variations   Ensembl Plants Variations 60</code></pre>
</div>
</div>
<p><code>plants_mart</code> looks like the one we want. We can see what genomes are available with names like â€œArabidopsisâ€ in this mart using the <code><a href="https://rdrr.io/pkg/biomaRt/man/listDatasets.html">searchDatasets()</a></code> function.</p>
<p>ğŸ¬</p>
<div class="cell">
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listDatasets.html">searchDatasets</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useEnsembl.html">useEnsemblGenomes</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"plants_mart"</span><span class="op">)</span>, </span>
<span>               pattern <span class="op">=</span> <span class="st">"Arabidopsis"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             dataset                         description version
4   ahalleri_eg_gene Arabidopsis halleri genes (Ahal2.2) Ahal2.2
6    alyrata_eg_gene    Arabidopsis lyrata genes (v.1.0)   v.1.0
11 athaliana_eg_gene Arabidopsis thaliana genes (TAIR10)  TAIR10</code></pre>
</div>
</div>
<p><code>athaliana_eg_gene</code> is the Arabidopsis thaliana genes (TAIR10) dataset we want.</p>
<p>ğŸ¬ Connect to the <code>athaliana_eg_gene</code> database in <code>plants_mart</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useEnsembl.html">useEnsemblGenomes</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"plants_mart"</span>,</span>
<span>                             dataset <span class="op">=</span> <span class="st">"athaliana_eg_gene"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ See the the types of information we can retrieve:</p>
<div class="cell">
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listAttributes.html">listAttributes</a></span><span class="op">(</span>mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are many (1,714!) possible bits of information (attributes) that can be obtained.</p>
<p>We use the <code><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM()</a></code> function to retrieve information from the database. The <code>filters</code> argument is used to specified what kind of identifier we are supplying in <code>values</code> to retrieve information. The <code>attributes</code> argument is used to select the information we want to retrieve. The <code>values</code> argument is used to specify the identifiers. The mart argument is used to specify the connection we created.</p>
<p>ğŸ¬ Get the the gene name and a description. We also retreive the gene id so we can later join the information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span>                   attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensembl_gene_id"</span>,</span>
<span>                                  <span class="st">"external_gene_name"</span>,</span>
<span>                                  <span class="st">"description"</span><span class="op">)</span>,</span>
<span>                   values <span class="op">=</span> <span class="va">wild_results</span><span class="op">$</span><span class="va">gene_id</span>,</span>
<span>                   mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should view the resulting dataframe to see what information is available. You can use <code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code> or <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code>.</p>
<p>ğŸ¬ Merge the gene information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># join the gene info with the results</span></span>
<span><span class="va">wild_results</span> <span class="op">&lt;-</span> <span class="va">wild_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">gene_info</span>,</span>
<span>            by <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join_by.html">join_by</a></span><span class="op">(</span><span class="va">gene_id</span> <span class="op">==</span> <span class="va">ensembl_gene_id</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Save the results to a file:</p>
<div class="cell">
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">wild_results</span>, file <span class="op">=</span> <span class="st">"results/wild_results.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="leishmania-3" class="level2"><h2 class="anchored" data-anchor-id="leishmania-3">ğŸ’‰ <em>Leishmania</em>
</h2>
<ul>
<li><p>I got the information from <a href="https://tritrypdb.org/tritrypdb/app/">TriTrypDB</a><br>
which is a functional genomic resource for the Trypanosomatidae and Plasmodidae</p></li>
<li><p><a href="https://tritrypdb.org/tritrypdb/app/downloads">https://tritrypdb.org/tritrypdb/app/downloads</a> section</p></li>
<li><p>I downloaded the <a href="https://tritrypdb.org/common/downloads/release-68/LmexicanaMHOMGT2001U1103/gff/data/TriTrypDB-68_LmexicanaMHOMGT2001U1103.gff"><em>L. mexicana</em> MHOM/GT/2001/U1103 Full GFF</a> and extracted the gene information and saved it as <a href="meta/leishmania_mex.xlsx">leishmania_mex.xlsx</a></p></li>
</ul>
<p>We will import this file and join it to the results dataframe.</p>
<p>ğŸ¬ Load the <strong><code>readxl</code></strong> <span class="citation" data-cites="readxl">(<a href="#ref-readxl" role="doc-biblioref">Wickham and Bryan 2023</a>)</span> package:</p>
<div class="cell">
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org">readxl</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Import the Xenbase gene information file:</p>
<div class="cell">
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_excel</a></span><span class="op">(</span><span class="st">"meta/leishmania_mex.xlsx"</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should view the resulting dataframe to see what information is available. You can use <code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code> or <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code>.</p>
<p>ğŸ¬ Merge the gene information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># join the gene info with the results</span></span>
<span><span class="va">pro_meta_results</span> <span class="op">&lt;-</span> <span class="va">pro_meta_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">gene_info</span>, by <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Save the results to a file:</p>
<div class="cell">
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">pro_meta_results</span>, file <span class="op">=</span> <span class="st">"results/pro_meta_results.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="stem-cells-3" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-3">ğŸ­ Stem cells</h2>
<p><a href="https://www.ensembl.org/index.html">Ensembl</a> <span class="citation" data-cites="martin2023 birney2004">(<a href="#ref-martin2023" role="doc-biblioref">Martin et al. 2023</a>; <a href="#ref-birney2004" role="doc-biblioref">Birney et al. 2004</a>)</span>is a bioinformatics project to organise all the biological information around the sequences of large genomes. The are a large number of databases but <a href="https://www.ensembl.org/info/data/biomart/index.html">BioMart</a> <span class="citation" data-cites="smedley2009">(<a href="#ref-smedley2009" role="doc-biblioref">Smedley et al. 2009</a>)</span> provides a consistent interface to the material. There are web-based tools to use these but the R package <strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#ref-biomaRt1" role="doc-biblioref">Durinck et al. 2009</a>, <a href="#ref-biomaRt2" role="doc-biblioref">2005</a>)</span> gives you programmatic access making it easier to integrate information into R dataframes</p>
<p>ğŸ¬ Load the <strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#ref-biomaRt1" role="doc-biblioref">Durinck et al. 2009</a>, <a href="#ref-biomaRt2" role="doc-biblioref">2005</a>)</span> package:</p>
<div class="cell">
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/grimbough/biomaRt">biomaRt</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Connect to the mouse database</p>
<div class="cell">
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Connect to the mouse database</span></span>
<span></span>
<span><span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"ensembl"</span>, </span>
<span>                   dataset <span class="op">=</span> <span class="st">"mmusculus_gene_ensembl"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ See information we can retrieve:</p>
<div class="cell">
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># See what information we can retrieve</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listAttributes.html">listAttributes</a></span><span class="op">(</span>mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are many (2,988!) possible bits of information (attributes) that can be obtained.</p>
<p>We use the <code><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM()</a></code> function to retrieve information from the database. The <code>filters</code> argument is used to specified what kind of identifier we are supplying in <code>values</code> to retrieve information. The <code>attributes</code> argument is used to select the information we want to retrieve. The <code>values</code> argument is used to specify the identifiers. The mart argument is used to specify the connection we created.</p>
<p>ğŸ¬ Get the the gene name and a description. We also retrieve the gene id so we can later join the information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span>                   attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensembl_gene_id"</span>,</span>
<span>                                  <span class="st">"external_gene_name"</span>,</span>
<span>                                  <span class="st">"description"</span><span class="op">)</span>,</span>
<span>                   values <span class="op">=</span> <span class="va">hspc_prog_results</span><span class="op">$</span><span class="va">ensembl_gene_id</span>,</span>
<span>                   mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should view the resulting dataframe to see what information is available. You can use <code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code> or <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code>. Notice the dataframe returned only has 279 rows - one of the ids does not have information.</p>
<p>ğŸ¬ Merge the gene information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># join the gene info with the results</span></span>
<span><span class="va">hspc_prog_results</span> <span class="op">&lt;-</span> <span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">gene_info</span>, by <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Save the results to a file:</p>
<div class="cell">
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">hspc_prog_results</span>, file <span class="op">=</span> <span class="st">"results/hspc_prog_results.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="look-after-future-you" class="level1"><h1>ğŸ¤— Look after future you!</h1>
<p>ğŸ¬ Go through you script and tidy up. You might need to :</p>
<ul>
<li>collect together library statements at the top of the script</li>
<li>remove code that you needed to start today but which wouldnâ€™t be needed if you were running the script from the top (e.g., importing)</li>
<li>edit your comments for clarity</li>
<li>rename variables for consistency or clarity</li>
<li>remove house keeping or exploratory code</li>
<li>restyle code, add code section headers etc</li>
</ul></section><section id="finished" class="level1"><h1>ğŸ¥³ Finished</h1>
<p>Well Done!</p>
</section><section id="independent-study-following-the-workshop" class="level1"><h1>Independent study following the workshop</h1>
<p><a href="../../transcriptomics/week-4/study_after_workshop.html">Consolidate</a></p>
</section><section id="the-code-file" class="level1"><h1>The Code file</h1>
<p>These contain all the code needed in the workshop even where it is not visible on the webpage.</p>
<p>The <a href="../../transcriptomics/week-4/workshop.html">workshop.qmd</a> file is the file I use to compile the practical. Qmd stands for Quarto markdown. It allows code and ordinary text to be interleaved to produce well-formatted reports including webpages. Right-click on the link and choose Save-As to download. You will be able to open the Qmd file in RStudio. Alternatively, <a href="https://github.com/3mmaRand/BIO00088H-data/blob/main/transcriptomics/week-4/workshop.qmd">View in Browser</a>. Coding and thinking answers are marked with <code>#---CODING ANSWER---</code> and <code>#---THINKING ANSWER---</code></p>
<p>Pages made with R <span class="citation" data-cites="R-core">(<a href="#ref-R-core" role="doc-biblioref">R Core Team 2024</a>)</span>, Quarto <span class="citation" data-cites="Allaire_Quarto_2024">(<a href="#ref-Allaire_Quarto_2024" role="doc-biblioref">Allaire et al. 2024</a>)</span>, <code>knitr</code> <span class="citation" data-cites="knitr1 knitr2 knitr3">(<a href="#ref-knitr1" role="doc-biblioref">Xie 2024</a>, <a href="#ref-knitr2" role="doc-biblioref">2015</a>, <a href="#ref-knitr3" role="doc-biblioref">2014</a>)</span>, <code>kableExtra</code> <span class="citation" data-cites="kableExtra">(<a href="#ref-kableExtra" role="doc-biblioref">Zhu 2021</a>)</span></p>
</section><section id="references" class="level1">



</section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Allaire_Quarto_2024" class="csl-entry" role="listitem">
Allaire, J. J., Charles Teague, Carlos Scheidegger, Yihui Xie, and Christophe Dervieux. 2024. <span>â€œ<span>Quarto</span>.â€</span> <a href="https://doi.org/10.5281/zenodo.5960048">https://doi.org/10.5281/zenodo.5960048</a>.
</div>
<div id="ref-birney2004" class="csl-entry" role="listitem">
Birney, Ewan, T. Daniel Andrews, Paul Bevan, Mario Caccamo, Yuan Chen, Laura Clarke, Guy Coates, et al. 2004. <span>â€œAn Overview of Ensembl.â€</span> <em>Genome Research</em> 14 (5): 925â€“28. <a href="https://doi.org/10.1101/gr.1860604">https://doi.org/10.1101/gr.1860604</a>.
</div>
<div id="ref-biomaRt2" class="csl-entry" role="listitem">
Durinck, Steffen, Yves Moreau, Arek Kasprzyk, Sean Davis, Bart De Moor, Alvis Brazma, and Wolfgang Huber. 2005. <span>â€œBioMart and Bioconductor: A Powerful Link Between Biological Databases and Microarray Data Analysis.â€</span> <em>Bioinformatics</em> 21: 3439â€“40.
</div>
<div id="ref-biomaRt1" class="csl-entry" role="listitem">
Durinck, Steffen, Paul T. Spellman, Ewan Birney, and Wolfgang Huber. 2009. <span>â€œMapping Identifiers for the Integration of Genomic Datasets with the r/Bioconductor Package biomaRt.â€</span> <em>Nature Protocols</em> 4: 1184â€“91.
</div>
<div id="ref-DESeq2" class="csl-entry" role="listitem">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>â€œModerated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.â€</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-scran" class="csl-entry" role="listitem">
Lun, Aaron T. L., Davis J. McCarthy, and John C. Marioni. 2016. <span>â€œA Step-by-Step Workflow for Low-Level Analysis of Single-Cell RNA-Seq Data with Bioconductor.â€</span> <em>F1000Res.</em> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-martin2023" class="csl-entry" role="listitem">
Martin, Fergal J, M Ridwan Amode, Alisha Aneja, Olanrewaju Austine-Orimoloye, Andrey&nbsp;G Azov, If Barnes, Arne Becker, et al. 2023. <span>â€œEnsembl 2023.â€</span> <em>Nucleic Acids Research</em> 51 (D1): D933â€“41. <a href="https://doi.org/10.1093/nar/gkac958">https://doi.org/10.1093/nar/gkac958</a>.
</div>
<div id="ref-R-core" class="csl-entry" role="listitem">
R Core Team. 2024. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-smedley2009" class="csl-entry" role="listitem">
Smedley, Damian, Syed Haider, Benoit Ballester, Richard Holland, Darin London, Gudmundur Thorisson, and Arek Kasprzyk. 2009. <span>â€œBioMart <span></span> Biological Queries Made Easy.â€</span> <em>BMC Genomics</em> 10 (1): 22. <a href="https://doi.org/10.1186/1471-2164-10-22">https://doi.org/10.1186/1471-2164-10-22</a>.
</div>
<div id="ref-conflicted" class="csl-entry" role="listitem">
Wickham, Hadley. 2023. <em>Conflicted: An Alternative Conflict Resolution Strategy</em>. <a href="https://conflicted.r-lib.org/">https://conflicted.r-lib.org/</a>.
</div>
<div id="ref-tidyverse" class="csl-entry" role="listitem">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy Dâ€™Agostino McGowan, Romain FranÃ§ois, Garrett Grolemund, et al. 2019. <span>â€œWelcome to the <span class="nocase">tidyverse</span>.â€</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-readxl" class="csl-entry" role="listitem">
Wickham, Hadley, and Jennifer Bryan. 2023. <em>Readxl: Read Excel Files</em>. <a href="https://readxl.tidyverse.org">https://readxl.tidyverse.org</a>.
</div>
<div id="ref-knitr3" class="csl-entry" role="listitem">
Xie, Yihui. 2014. <span>â€œKnitr: A Comprehensive Tool for Reproducible Research in <span>R</span>.â€</span> In <em>Implementing Reproducible Computational Research</em>, edited by Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman; Hall/CRC.
</div>
<div id="ref-knitr2" class="csl-entry" role="listitem">
â€”â€”â€”. 2015. <em>Dynamic Documents with <span>R</span> and Knitr</em>. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-knitr1" class="csl-entry" role="listitem">
â€”â€”â€”. 2024. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in r</em>. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-kableExtra" class="csl-entry" role="listitem">
Zhu, Hao. 2021. <span>â€œkableExtra: Construct Complex Table with â€™Kableâ€™ and Pipe Syntax.â€</span> <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>
<ol>
<li id="fn1"><p><a href="https://www.bioconductor.org/">Bioconductor</a> is a project that develops and supports R packages for bioinformatics.<a href="#fnref1" class="footnote-back" role="doc-backlink">â†©ï¸</a></p></li>
<li id="fn2"><p><a href="https://www.bioconductor.org/">Bioconductor</a> is a project that develops and supports R packages for bioinformatics.<a href="#fnref2" class="footnote-back" role="doc-backlink">â†©ï¸</a></p></li>
<li id="fn3"><p><a href="https://www.bioconductor.org/">Bioconductor</a> is a project that develops and supports R packages for bioinformatics.<a href="#fnref3" class="footnote-back" role="doc-backlink">â†©ï¸</a></p></li>
</ol></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("^(?:http:|https:)\/\/3mmarand\.github\.io\/BIO00088H-data");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Rand E (2024). Data Analysis for Group Project. <a href="https://3mmarand.github.io/BIO00088H-data/" class="uri">https://3mmarand.github.io/BIO00088H-data/</a>.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/er_13">
      <i class="bi bi-twitter" role="img" aria-label="Twitter">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/3mmaRand/BIO00088H-data">
      <i class="bi bi-github" role="img" aria-label="BIO00088H-data GitHub">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>