<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.33">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Emma Rand">
<title>Workshop â€“ Data Analysis for Group Project</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-bc185b5c5bdbcb35c2eb49d8a876ef70.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" class="quarto-color-scheme-extra" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-72ac799c4abbddf31cb463dd8fc99f5d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark-25ac7be22ffa75a8e8541e30a6dcc9a3.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="../../site_libs/bootstrap/bootstrap-72ac799c4abbddf31cb463dd8fc99f5d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme-extra" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-TVF95J593T"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-TVF95J593T', { 'anonymize_ip': true});
</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar floating nav-fixed quarto-light"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = false;
    const darkModeDefault = authorPrefersDark;
      document.querySelector('link#quarto-text-highlighting-styles.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
      document.querySelector('link#quarto-bootstrap.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
      window.dispatchEvent(new Event('resize'));
    };
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Data Analysis for Group Project</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Welcome!</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../core/core.html"> 
<span class="menu-text">Core Supporting Info</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../transcriptomics/transcriptomics.html" aria-current="page"> 
<span class="menu-text">Transcriptomics</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../images/images.html"> 
<span class="menu-text">Image Analysis</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../structures/structures.html"> 
<span class="menu-text">Structure Analysis</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools tools-wide">
    <a href="https://twitter.com/er_13" title="Emma's Twitter" class="quarto-navigation-tool px-1" aria-label="Emma's Twitter"><i class="bi bi-twitter"></i></a>
    <a href="https://github.com/3mmaRand/BIO00088H-data" title="BIO00088H-data GitHub" class="quarto-navigation-tool px-1" aria-label="BIO00088H-data GitHub"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/overview.html">Week 4: Statistical Analysis</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/workshop.html">Workshop</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/transcriptomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Transcriptomics Data Analysis for Group Project</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 3: Hello data!</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 4: Statistical Analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/workshop.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 5: Visualising and Interpreting</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Semester 2: Extras</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/semester-2/stem_cell_variant.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Additional ğŸ­ stem cell labels</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Kellyâ€™s Project</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/kelly/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul>
<li><a href="#session-overview" id="toc-session-overview" class="nav-link" data-scroll-target="#session-overview">Session overview</a></li>
  </ul>
</li>
  <li>
<a href="#set-up" id="toc-set-up" class="nav-link" data-scroll-target="#set-up">Set up</a>
  <ul>
<li><a href="#arabidopisis" id="toc-arabidopisis" class="nav-link" data-scroll-target="#arabidopisis">ğŸ„ <em>Arabidopisis</em></a></li>
  <li><a href="#leishmania" id="toc-leishmania" class="nav-link" data-scroll-target="#leishmania">ğŸ’‰ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells" id="toc-stem-cells" class="nav-link" data-scroll-target="#stem-cells">ğŸ­ Stem cells</a></li>
  <li><a href="#everyone" id="toc-everyone" class="nav-link" data-scroll-target="#everyone">Everyone</a></li>
  </ul>
</li>
  <li>
<a href="#import" id="toc-import" class="nav-link" data-scroll-target="#import">Import</a>
  <ul>
<li><a href="#arabidopisis-1" id="toc-arabidopisis-1" class="nav-link" data-scroll-target="#arabidopisis-1">ğŸ„ <em>Arabidopisis</em></a></li>
  <li><a href="#leishmania-1" id="toc-leishmania-1" class="nav-link" data-scroll-target="#leishmania-1">ğŸ’‰ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells-1" id="toc-stem-cells-1" class="nav-link" data-scroll-target="#stem-cells-1">ğŸ­ Stem cells</a></li>
  </ul>
</li>
  <li>
<a href="#differential-expression-analysis" id="toc-differential-expression-analysis" class="nav-link" data-scroll-target="#differential-expression-analysis">Differential Expression Analysis</a>
  <ul>
<li>
<a href="#arabidopisis-differential" id="toc-arabidopisis-differential" class="nav-link" data-scroll-target="#arabidopisis-differential">ğŸ„ <em>Arabidopisis</em></a>
  <ul class="collapse">
<li><a href="#genes-expressed-in-one-treatment" id="toc-genes-expressed-in-one-treatment" class="nav-link" data-scroll-target="#genes-expressed-in-one-treatment">1. Genes expressed in one treatment</a></li>
  <li><a href="#create-deseqdataset-object" id="toc-create-deseqdataset-object" class="nav-link" data-scroll-target="#create-deseqdataset-object">2. Create DESeqDataSet object</a></li>
  <li><a href="#prepare-the-normalised-counts" id="toc-prepare-the-normalised-counts" class="nav-link" data-scroll-target="#prepare-the-normalised-counts">3. Prepare the normalised counts</a></li>
  <li><a href="#differential-expression-analysis-1" id="toc-differential-expression-analysis-1" class="nav-link" data-scroll-target="#differential-expression-analysis-1">4. Differential expression analysis</a></li>
  </ul>
</li>
  <li>
<a href="#leishmania-differential" id="toc-leishmania-differential" class="nav-link" data-scroll-target="#leishmania-differential">ğŸ’‰ <em>Leishmania</em></a>
  <ul class="collapse">
<li><a href="#genes-expressed-in-one-treatment-1" id="toc-genes-expressed-in-one-treatment-1" class="nav-link" data-scroll-target="#genes-expressed-in-one-treatment-1">1. Genes expressed in one treatment</a></li>
  <li><a href="#create-deseqdataset-object-1" id="toc-create-deseqdataset-object-1" class="nav-link" data-scroll-target="#create-deseqdataset-object-1">2. Create DESeqDataSet object</a></li>
  <li><a href="#prepare-the-normalised-counts-1" id="toc-prepare-the-normalised-counts-1" class="nav-link" data-scroll-target="#prepare-the-normalised-counts-1">3. Prepare the normalised counts</a></li>
  <li><a href="#differential-expression-analysis-2" id="toc-differential-expression-analysis-2" class="nav-link" data-scroll-target="#differential-expression-analysis-2">4. Differential expression analysis</a></li>
  </ul>
</li>
  <li>
<a href="#stem-cells-differential" id="toc-stem-cells-differential" class="nav-link" data-scroll-target="#stem-cells-differential">ğŸ­ Stem cells</a>
  <ul class="collapse">
<li><a href="#genes-expressed-in-one-cell-type" id="toc-genes-expressed-in-one-cell-type" class="nav-link" data-scroll-target="#genes-expressed-in-one-cell-type">1. Genes expressed in one cell type</a></li>
  <li><a href="#prepare-the-data-for-analysis-with-scran" id="toc-prepare-the-data-for-analysis-with-scran" class="nav-link" data-scroll-target="#prepare-the-data-for-analysis-with-scran">2. Prepare the data for analysis with <strong><code>scran</code></strong></a></li>
  <li><a href="#differential-expression-analysis-3" id="toc-differential-expression-analysis-3" class="nav-link" data-scroll-target="#differential-expression-analysis-3">3. Differential expression analysis</a></li>
  </ul>
</li>
  </ul>
</li>
  <li>
<a href="#add-gene-information" id="toc-add-gene-information" class="nav-link" data-scroll-target="#add-gene-information">Add gene information</a>
  <ul>
<li><a href="#arabidopisis-add-info" id="toc-arabidopisis-add-info" class="nav-link" data-scroll-target="#arabidopisis-add-info">ğŸ„ <em>Arabidopisis</em></a></li>
  <li><a href="#leishmania-add-info" id="toc-leishmania-add-info" class="nav-link" data-scroll-target="#leishmania-add-info">ğŸ’‰ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells-add-info" id="toc-stem-cells-add-info" class="nav-link" data-scroll-target="#stem-cells-add-info">ğŸ­ Stem cells</a></li>
  </ul>
</li>
  <li><a href="#look-after-future-you" id="toc-look-after-future-you" class="nav-link" data-scroll-target="#look-after-future-you">ğŸ¤— Look after future you!</a></li>
  <li><a href="#finished" id="toc-finished" class="nav-link" data-scroll-target="#finished">ğŸ¥³ Finished</a></li>
  <li><a href="#independent-study-following-the-workshop" id="toc-independent-study-following-the-workshop" class="nav-link" data-scroll-target="#independent-study-following-the-workshop">Independent study following the workshop</a></li>
  <li><a href="#the-code-file" id="toc-the-code-file" class="nav-link" data-scroll-target="#the-code-file">The Code file</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/overview.html">Week 4: Statistical Analysis</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/week-4/workshop.html">Workshop</a></li></ol></nav><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Workshop</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
<p class="subtitle lead">Transcripttranscriptomics 2: Statistical Analysis</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Emma Rand </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">15 August, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header><section id="introduction" class="level1"><h1>Introduction</h1>
<section id="session-overview" class="level2"><h2 class="anchored" data-anchor-id="session-overview">Session overview</h2>
<p>In the workshop, you will learn how to perform differential expression analysis on raw counts using <strong><code>DESeq2</code></strong> <span class="citation" data-cites="DESeq2">(<a href="#ref-DESeq2" role="doc-biblioref">Love, Huber, and Anders 2014</a>)</span> or on logged normalised expression values using <strong><code>scran</code></strong> <span class="citation" data-cites="scran">(<a href="#ref-scran" role="doc-biblioref">Lun, McCarthy, and Marioni 2016</a>)</span> or both.</p>
</section></section><section id="set-up" class="level1"><h1>Set up</h1>
<section id="arabidopisis" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis">ğŸ„ <em>Arabidopisis</em>
</h2>
<p>ğŸ¬ Open the <code>arab-88H</code> RStudio Project and the <code>cont-low-root.R</code> script.</p>
</section><section id="leishmania" class="level2"><h2 class="anchored" data-anchor-id="leishmania">ğŸ’‰ <em>Leishmania</em>
</h2>
<p>ğŸ¬ Open the <code>leish-88H</code> RStudio Project and the <code>pro-meta.R</code> script.</p>
</section><section id="stem-cells" class="level2"><h2 class="anchored" data-anchor-id="stem-cells">ğŸ­ Stem cells</h2>
<p>ğŸ¬ Open the <code>mice-88H</code> RStudio Project and the <code>hspc-prog.R</code> script.</p>
</section><section id="everyone" class="level2"><h2 class="anchored" data-anchor-id="everyone">Everyone</h2>
<p>ğŸ¬ Make a new folder <code>results</code> in the project directory.</p>
<p>This is where we will save our results.</p>
<p>ğŸ¬ Load <strong><code>tidyverse</code></strong> <span class="citation" data-cites="tidyverse">(<a href="#ref-tidyverse" role="doc-biblioref">Wickham et al. 2019</a>)</span> You most likely have this code at the top of your script already.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€
âœ” dplyr     1.1.3     âœ” readr     2.1.4
âœ” forcats   1.0.0     âœ” stringr   1.5.0
âœ” ggplot2   3.4.3     âœ” tibble    3.2.1
âœ” lubridate 1.9.3     âœ” tidyr     1.3.0
âœ” purrr     1.0.2     
â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€
âœ– dplyr::filter() masks stats::filter()
âœ– dplyr::lag()    masks stats::lag()
â„¹ Use the conflicted package to force all conflicts to become errors</code></pre>
<p>Have you ever stopped to think about this message? It is telling us that there are functions in the <code>dplyr</code> package that have the same name as functions in the <code>stats</code> package and that R will use the <code>dplyr</code> version. As this is what you want, this has always been fine. It still is fine in this case. However, as you start to load more packages, you will want to know if you are using a function from a package that has the same name as a function in another loaded package. This is where the <strong><code>conflicted</code></strong> <span class="citation" data-cites="conflicted">(<a href="#ref-conflicted" role="doc-biblioref">Wickham 2023</a>)</span> package comes in. <strong><code>Conflicted</code></strong> will warn you when you are using a function that has the same name as a function in another package. You can then choose which function to use.</p>
<p>ğŸ¬ Load the <strong><code>conflicted</code></strong> package:</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://conflicted.r-lib.org/">conflicted</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Instead of getting a warning every time you are using a function that has a function with the same name in another package, we can declare a preference for one function over another. This is useful for the functions you use a lot or ones where you are certain you always want to use a particular function.</p>
<p>For example, to always use the <strong><code>dplyr</code></strong> version of <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> by default you can add this to the top of your script:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://conflicted.r-lib.org/reference/conflicts_prefer.html">conflicts_prefer</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will also want to ensure that we are using the <code><a href="https://generics.r-lib.org/reference/setops.html">setdiff()</a></code> function from the <strong><code>GenomicRanges</code></strong> package which is used by <strong><code>DESeq2</code></strong>.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://conflicted.r-lib.org/reference/conflicts_prefer.html">conflicts_prefer</a></span><span class="op">(</span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="va"><a href="https://generics.r-lib.org/reference/setops.html">setdiff</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="import" class="level1"><h1>Import</h1>
<section id="arabidopisis-1" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis-1">ğŸ„ <em>Arabidopisis</em>
</h2>
<p>We need to import the root tissue data that were filtered to remove genes with 3 or 4 zeros and those where the total counts was less than 20.</p>
<p>ğŸ¬ Import the data from the <code>data-processed</code> folder.</p>
<p>Now go to <a href="#arabidopisis-differential">Differential Expression Analysis</a>.</p>
</section><section id="leishmania-1" class="level2"><h2 class="anchored" data-anchor-id="leishmania-1">ğŸ’‰ <em>Leishmania</em>
</h2>
<p>We need to import the procyclic- and metacyclic-promastigote data that were filtered to remove genes with 4, 5 or 6 zeros and those where the total counts was less than 20.</p>
<p>ğŸ¬ Import the data from the <code>data-processed</code> folder.</p>
<p>Now go to <a href="#leishmania-differential">Differential Expression Analysis</a>.</p>
</section><section id="stem-cells-1" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-1">ğŸ­ Stem cells</h2>
<p>We need to import the progenitor and HSPC cell data that were combined.</p>
<p>ğŸ¬ Import the data from the <code>data-processed</code> folder..</p>
<p>Now go to <a href="#stem-cells-differential">Differential Expression Analysis</a>.</p>
</section></section><section id="differential-expression-analysis" class="level1"><h1>Differential Expression Analysis</h1>
<section id="arabidopisis-differential" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis-differential">ğŸ„ <em>Arabidopisis</em>
</h2>
<p>These are the steps we will take</p>
<ol type="1">
<li>Find the genes that are expressed in only one treatment group.</li>
<li>Create a DESeqDataSet object. This is a special object that is used by the <strong><code>DESeq2</code></strong> package</li>
<li>Prepare the normalised counts from the DESeqDataSet object.</li>
<li>Do differential expression analysis on the genes. This needs to be done on the raw counts.</li>
</ol>
<p>All but the first step are done with the <strong><code>DESeq2</code></strong> package</p>
<section id="genes-expressed-in-one-treatment" class="level3"><h3 class="anchored" data-anchor-id="genes-expressed-in-one-treatment">1. Genes expressed in one treatment</h3>
<p>The genes expressed in only one treatment group are those with zeros in all replicates in one group and non-zero values in all replicates in the other group. For example, those shown here:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">gene_id</th>
<th style="text-align: left;">gene_name</th>
<th style="text-align: right;">CTR1</th>
<th style="text-align: right;">CTR2</th>
<th style="text-align: right;">CTR3</th>
<th style="text-align: right;">CTR4</th>
<th style="text-align: right;">CTR5</th>
<th style="text-align: right;">CTR6</th>
<th style="text-align: right;">LWR1</th>
<th style="text-align: right;">LWR2</th>
<th style="text-align: right;">LWR3</th>
<th style="text-align: right;">LWR4</th>
<th style="text-align: right;">LWR5</th>
<th style="text-align: right;">LWR6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">AT1G27090</td>
<td style="text-align: left;">AT1G27090</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">16</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">14</td>
</tr>
<tr class="even">
<td style="text-align: left;">AT1G14950</td>
<td style="text-align: left;">AT1G14950</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">17</td>
</tr>
<tr class="odd">
<td style="text-align: left;">AT1G80160</td>
<td style="text-align: left;">GLYI7</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">33</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">27</td>
</tr>
<tr class="even">
<td style="text-align: left;">AT1G18700</td>
<td style="text-align: left;">AT1G18700</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">AT5G39775</td>
<td style="text-align: left;">AT5G39775</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">29</td>
<td style="text-align: right;">16</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We will use <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> to find these genes.</p>
<p>ğŸ¬ Find the genes that are expressed only in the control group:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root_lowni_only</span> <span class="op">&lt;-</span> <span class="va">root_filtered</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">if_all</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="st">"CTR"</span><span class="op">)</span>, \<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">if_all</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="st">"LWR"</span><span class="op">)</span>, \<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This code creates a new data frame called <code>root_lowni_only</code> that contains only the rows from root_filtered where:</p>
<ul>
<li>All CTR columns have a value of 0</li>
<li>All LWR columns have a value greater than 0</li>
</ul>
<p><code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> keeps rows that match the conditions you give it. Here we give two conditions:</p>
<ul>
<li>All columns whose names start with â€œCTRâ€ must have a value of 0.</li>
<li>All columns whose names start with â€œLWRâ€ must have a value greater than 0.</li>
</ul>
<p>The <code><a href="https://dplyr.tidyverse.org/reference/across.html">if_all()</a></code> function applies a test to each selected column (those starting with â€œCTRâ€ or â€œLWRâ€) for each row. It returns <code>TRUE</code> only if all the selected columns pass the test.</p>
<p>The bit <code>\(x) x == 0</code> is an anonymous function â€” a function without a name. In base R syntax, itâ€™s equivalent to: <code>function(x) x == 0</code> Here, <code>x</code> will be the values from one column at a time, and the function checks whether they are equal to 0. We use an anonymous function because the condition needs to be applied separately to each column, but itâ€™s simple enough that thereâ€™s no need to define a full named function elsewhere in the script.</p>
<p>â“ How many genes are expressed only in the low nickel group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 16 genes expressed only in low nickel group?. -->
<p>Note: There may have been more genes with counts in the low nickel group that were removed when we filtered on the total number of counts being less that 50.</p>
<p>ğŸ¬ Now you find any genes that are expressed only in the control group.</p>
<p>â“ How many genes are expressed only in the control group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 7 genes expressed only in the control group. -->
<p>â“ Do the results make sense to you in light of what you know about the biology?</p>
<!-- #---THINKING ANSWER--- -->
<!-- We might expect more genes to be switched on cope with nickel -->
<!-- than in control conditions -->
<!-- However, I'm not an expert in Arabidopsis, or plants! -->
<!-- Domain expertise is important when analysing data so  -->
<!-- discuss your results with Liz and your group. -->
<p>ğŸ¬ Write all the genes that are expressed one group only to file (saved in <code>results</code>)</p>
</section><section id="create-deseqdataset-object" class="level3"><h3 class="anchored" data-anchor-id="create-deseqdataset-object">2. Create DESeqDataSet object</h3>
<p>ğŸ¬ Load the DESeq2 package:</p>
<p>A DEseqDataSet object is a custom data type that is used by <strong><code>DESeq2</code></strong>. Custom data types are common in the Bioconductor<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> packages. They are used to store data in a way that is useful for the analysis. These data types typically have data, transformed data, metadata and experimental designs within them.</p>
<p>To create a DESeqDataSet object, we need to provide three things:</p>
<ul>
<li>The raw counts - these are in <code>root_filtered</code>
</li>
<li>The meta data which gives information about the samples and which treatment groups they belong to</li>
<li>A design matrix which captures the design of the statistical model.</li>
</ul>
<p>The counts must in a <em>matrix</em> rather than a dataframe. Unlike a dataframe, a matrix has columns of all the same type. That is, it will contain only the counts. The gene ids are given as row names rather than a column. The <code><a href="https://rdrr.io/r/base/matrix.html">matrix()</a></code> function will create a matrix from a dataframe of columns of the same type and the <code><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">select()</a></code> function can be used to remove the gene ids column.</p>
<p>ğŸ¬ Create a matrix of the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root_count_mat</span> <span class="op">&lt;-</span> <span class="va">root_filtered</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">gene_id</span>, <span class="op">-</span><span class="va">gene_name</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Add the gene ids as row names to the matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># add the row names to the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">root_count_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">root_filtered</span><span class="op">$</span><span class="va">gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You might want to view the matrix (click on it in your environment pane).</p>
<p>The metadata are in a file, <a href="meta/arab_meta_data.txt">arab_meta_data.txt</a>. This is a tab-delimited file. The first column is the sample name and the other columns give the â€œtreatmentsâ€. In this case, the treatments genotype (with two levels) and copper (with two levels).</p>
<p>ğŸ¬ Make a folder called <code>meta</code> and save the file to it.</p>
<p>ğŸ¬ Read the metadata into a dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_table.html">read_table</a></span><span class="op">(</span><span class="st">"meta/arab_meta_data.txt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Examine the resulting dataframe.</p>
<p>We need to add the sample names as row names to the metadata dataframe. This is because the DESeqDataSet object will use the row names to match the samples in the metadata to the samples in the counts matrix.</p>
<p>ğŸ¬ Add the sample names as row names to the metadata dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are dealing only with the root data so we need to remove the samples that are not in the root data.</p>
<p>ğŸ¬ Filter the metadata to keep only the root information:</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta_root</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"root"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now create the DESeqDataSet object. The design formula describes the statistical model. You should recognise the form from previous work. The <code>~</code> can be read as â€œexplain byâ€ and on its right hand side are the explanatory variables. That is, the model is counts explained by nickel status.</p>
<p>Note that:</p>
<ul>
<li>The names of the columns in the count matrix have to exactly match the names of the rows in the metadata dataframe. They also need to be in the same order.</li>
<li>The names of the explanatory variables in the design formula have to match the names of columns in the metadata.</li>
</ul>
<p>ğŸ¬ Create the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span><span class="va">root_count_mat</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">meta_root</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span> <span class="va">nickel</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The warning â€œWarning: some variables in design formula are characters, converting to factorsâ€ just means that the variable type of nickel in the metadata dataframe is â€œcharâ€ and it has been converted into a factor type.</p>
<p>To help you understand what the <code>DESeqDataSet</code> object we have called <code>dds</code> contains, we can look its contents</p>
<p>The counts are in <code>dds@assays@data@listData[["counts"]]</code> and the metadata are in <code>dds@colData</code> but the easiest way to see them is to use the <code>counts()</code> and <code>colData()</code> functions from the <strong><code>DESeq2</code></strong> package.</p>
<p>ğŸ¬ View the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should be able to see that this is the same as in <code>root_count_mat</code>.</p>
<p>ğŸ¬ View the column information:</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 12 rows and 2 columns
          tissue   nickel
     &lt;character&gt; &lt;factor&gt;
CTR1        root  control
CTR2        root  control
CTR3        root  control
CTR4        root  control
CTR5        root  control
...          ...      ...
LWR2        root   low_ni
LWR3        root   low_ni
LWR4        root   low_ni
LWR5        root   low_ni
LWR6        root   low_ni</code></pre>
</div>
</div>
<p>You should be able to see this is the same as in <code>meta_root</code>.</p>
</section><section id="prepare-the-normalised-counts" class="level3"><h3 class="anchored" data-anchor-id="prepare-the-normalised-counts">3. Prepare the normalised counts</h3>
<p>The normalised counts are the counts that have been transformed to account for the library size (i.e., the total number of reads in a sample) and the gene length. We have to first estimate the normalisation factors and store them in the DESeqDataSet object and then we can get the normalised counts.</p>
<p>ğŸ¬ Estimate the factors for normalisation and store them in the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Look at the factors (just for information):</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     CTR1      CTR2      CTR3      CTR4      CTR5      CTR6      LWR1      LWR2 
1.0242216 1.3312269 1.0268583 1.1672719 1.0272988 1.0104053 0.9386242 0.8341973 
     LWR3      LWR4      LWR5      LWR6 
0.8622105 0.9117686 1.0260255 0.9740288 </code></pre>
</div>
</div>
<p>The normalised counts will be useful to use later. To get the normalised counts we again used the <code>counts()</code> function but this time we use the <code>normalized=TRUE</code> argument.</p>
<p>ğŸ¬ Save the normalised to a matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">normalised_counts</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Make a dataframe of the normalised counts, adding a column for the gene ids at the same time:</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root_normalised_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">normalised_counts</span>,</span>
<span>                                    gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">normalised_counts</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="differential-expression-analysis-1" class="level3"><h3 class="anchored" data-anchor-id="differential-expression-analysis-1">4. Differential expression analysis</h3>
<p>We use the <code>DESeq()</code> function to do the differential expression analysis. This function fits the statistical model to the data and then uses the model to calculate the significance of the difference between the treatments. It again stores the results in the DESseqDataSet object. Note that the differential expression needs the raw (unnormalised counts) as it does its own normalisation as part of the process.</p>
<p>ğŸ¬ Run the differential expression analysis and store the results in the same object:</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function will take only a few moments to run on this data but can take longer for bigger datasets.</p>
<p>We need to define the contrasts we want to test. We want to test the difference between the treatments so we will define the contrast as <code>control</code> and <code>low_ni</code>.</p>
<p>ğŸ¬ Define the contrast:</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">contrast_suf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nickel"</span>, <span class="st">"control"</span>, <span class="st">"low_ni"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that <code>nickel</code> is the name of the column in the metadata dataframe and <code>control</code> and <code>low_ni</code> are the names of the levels in the <code>nickel</code> column. By putting them in the order <code>control</code> , <code>low_ni</code> we are saying the fold change will be control / low_ni. This means:</p>
<ul>
<li>positive log fold changes indicate control &gt; low_ni and</li>
<li>negative log fold changes indicates low_ni &gt; control.</li>
</ul>
<p>If we had put them in the order <code>low_ni</code>, <code>control</code> we would have the reverse.</p>
<p>ğŸ¬ Extract the results from the DESseqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results_suf</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                       contrast <span class="op">=</span> <span class="va">contrast_suf</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This will give us the log<sub>2</sub> fold change, the <em>p</em>-value and the adjusted <em>p</em>-value for the comparison between the control and low_ni for each gene.</p>
<p>ğŸ¬ Put the results in a dataframe and add the gene ids as a column:</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">results_suf</span>,</span>
<span>                          gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">results_suf</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is useful to have the normalised counts and the statistical results in one dataframe.</p>
<p>ğŸ¬ Merge the two dataframes:</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># merge the results with the normalised counts</span></span>
<span><span class="va">root_results</span> <span class="op">&lt;-</span> <span class="va">root_normalised_counts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">root_results</span>, by <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#arabidopisis-add-info">Add gene information</a>.</p>
</section></section><section id="leishmania-differential" class="level2"><h2 class="anchored" data-anchor-id="leishmania-differential">ğŸ’‰ <em>Leishmania</em>
</h2>
<p>These are the steps we will take</p>
<ol type="1">
<li>Find the genes that are expressed in only one treatment group.</li>
<li>Create a DESeqDataSet object. This is a special object that is used by the <strong><code>DESeq2</code></strong> package</li>
<li>Prepare the normalised counts from the DESeqDataSet object.</li>
<li>Do differential expression analysis on the genes. This needs to be done on the raw counts.</li>
</ol>
<p>All but the first step are done with the <strong><code>DESeq2</code></strong> package</p>
<section id="genes-expressed-in-one-treatment-1" class="level3"><h3 class="anchored" data-anchor-id="genes-expressed-in-one-treatment-1">1. Genes expressed in one treatment</h3>
<p>The genes expressed in only one treatment group are those with zeros in all replicates in one group and non-zero values in all replicates in the other group.</p>
<p>We will use <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> to find these genes.</p>
<p>ğŸ¬ Find the genes that are expressed only at the procyclic-promastigote stage:</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_pro_only</span> <span class="op">&lt;-</span> <span class="va">pro_meta_filtered</span>  <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">lm_pro_1</span> <span class="op">&gt;</span> <span class="fl">0</span>,</span>
<span>         <span class="va">lm_pro_2</span> <span class="op">&gt;</span> <span class="fl">0</span>,</span>
<span>         <span class="va">lm_pro_3</span> <span class="op">&gt;</span> <span class="fl">0</span>,</span>
<span>         <span class="va">lm_meta_1</span> <span class="op">==</span> <span class="fl">0</span>,</span>
<span>         <span class="va">lm_meta_2</span> <span class="op">==</span> <span class="fl">0</span>,</span>
<span>         <span class="va">lm_meta_2</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>â“ How many genes are expressed only in the procyclic-promastigote stage group?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 0 genes expressed only in the procyclic-promastigote stage -->
<p>ğŸ¬ Now you find any genes that are expressed only at the metacyclic stage</p>
<p>â“ How many genes are expressed only at the metacyclic stage?</p>
<!-- #---THINKING ANSWER--- -->
<!-- There are 0 genes expressed only in the meta stage group. -->
<p>â“ Do the results make sense to you in light of what you know about the biology?</p>
<!-- #---THINKING ANSWER--- -->
<!-- I have no clue here! -->
<!-- Domain expertise is important when analysing data so  -->
<!-- discuss your results with Pegine and your group. -->
<p>ğŸ¬ Write all the genes that are expressed one group only to file (saved in <code>results</code>)</p>
</section><section id="create-deseqdataset-object-1" class="level3"><h3 class="anchored" data-anchor-id="create-deseqdataset-object-1">2. Create DESeqDataSet object</h3>
<p>ğŸ¬ Load the DESeq2 package:</p>
<p>A DEseqDataSet object is a custom data type that is used by <strong><code>DESeq2</code></strong>. Custom data types are common in the Bioconductor<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> packages. They are used to store data in a way that is useful for the analysis. These data types typically have data, transformed data, metadata and experimental designs within them.</p>
<p>To create a DESeqDataSet object, we need to provide three things:</p>
<ul>
<li>The raw counts - these are in <code>pro_meta_filtered</code>
</li>
<li>The meta data which gives information about the samples and which treatment groups they belong to</li>
<li>A design matrix which captures the design of the statistical model.</li>
</ul>
<p>The counts must in a <em>matrix</em> rather than a dataframe. Unlike a dataframe, a matrix has columns of all the same type. That is, it will contain only the counts. The gene ids are given as row names rather than a column. The <code><a href="https://rdrr.io/r/base/matrix.html">matrix()</a></code> function will create a matrix from a dataframe of columns of the same type and the <code><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">select()</a></code> function can be used to remove the gene ids column.</p>
<p>ğŸ¬ Create a matrix of the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_count_mat</span> <span class="op">&lt;-</span> <span class="va">pro_meta_filtered</span>  <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Add the gene ids as row names to the matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># add the row names to the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">pro_meta_count_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pro_meta_filtered</span><span class="op">$</span><span class="va">gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You might want to view the matrix (click on it in your environment pane).</p>
<p>The metadata are in a file, <a href="meta/leish_meta_data.txt">leish_meta_data.txt</a>. This is a tab-delimited file. The first column is the sample name and the other columns give the â€œtreatmentsâ€. In this case, the treatment is stage (with three levels).</p>
<p>ğŸ¬ Make a folder called <code>meta</code> and save the file to it.</p>
<p>ğŸ¬ Read the metadata into a dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_table.html">read_table</a></span><span class="op">(</span><span class="st">"meta/leish_meta_data.txt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Examine the resulting dataframe.</p>
<p>We need to add the sample names as row names to the metadata dataframe. This is because the DESeqDataSet object will use the row names to match the samples in the metadata to the samples in the counts matrix.</p>
<p>ğŸ¬ Add the sample names as row names to the metadata dataframe:</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are dealing only with the wild data so we need to remove the samples that are not in the wild data.</p>
<p>ğŸ¬ Filter the metadata to keep only the procyclic and metacyclic information:</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">meta_pro_meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">stage</span> <span class="op">!=</span> <span class="st">"amastigotes"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now create the DESeqDataSet object. The design formula describes the statistical model. You should recognise the form from previous work. The <code>~</code> can be read as â€œexplain byâ€ and on its right hand side are the explanatory variables. That is, the model is counts explained by stage status.</p>
<p>Note that:</p>
<ul>
<li>The names of the columns in the count matrix have to exactly match the names of the rows in the metadata dataframe. They also need to be in the same order.</li>
<li>The names of the explanatory variables in the design formula have to match the names of columns in the metadata.</li>
</ul>
<p>ğŸ¬ Create the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span><span class="va">pro_meta_count_mat</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">meta_pro_meta</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span> <span class="va">stage</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The warning â€œWarning: some variables in design formula are characters, converting to factorsâ€ just means that the variable type of stage in the metadata dataframe is â€œcharâ€ and it has been converted into a factor type.</p>
<p>To help you understand what the <code>DESeqDataSet</code> object we have called <code>dds</code> contains, we can look its contents</p>
<p>The counts are in <code>dds@assays@data@listData[["counts"]]</code> and the metadata are in <code>dds@colData</code> but the easiest way to see them is to use the <code>counts()</code> and <code>colData()</code> functions from the <strong><code>DESeq2</code></strong> package.</p>
<p>ğŸ¬ View the counts:</p>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should be able to see that this is the same as in <code>pro_meta_count_mat</code>.</p>
<p>ğŸ¬ View the column information:</p>
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 2 columns
               stage replicate
            &lt;factor&gt; &lt;numeric&gt;
lm_pro_1  procyclic          1
lm_pro_2  procyclic          2
lm_pro_3  procyclic          3
lm_meta_1 metacyclic         1
lm_meta_2 metacyclic         2
lm_meta_3 metacyclic         3</code></pre>
</div>
</div>
<p>You should be able to see this is the same as in <code>meta_pro_meta</code>.</p>
</section><section id="prepare-the-normalised-counts-1" class="level3"><h3 class="anchored" data-anchor-id="prepare-the-normalised-counts-1">3. Prepare the normalised counts</h3>
<p>The normalised counts are the counts that have been transformed to account for the library size (i.e., the total number of reads in a sample) and the gene length. We have to first estimate the normalisation factors and store them in the DESeqDataSet object and then we can get the normalised counts.</p>
<p>ğŸ¬ Estimate the factors for normalisation and store them in the DESeqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Look at the factors (just for information):</p>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> lm_pro_1  lm_pro_2  lm_pro_3 lm_meta_1 lm_meta_2 lm_meta_3 
1.3029351 0.9158157 0.9943186 0.7849299 0.8443586 1.3250409 </code></pre>
</div>
</div>
<p>The normalised counts will be useful to use later. To get the normalised counts we again used the <code>counts()</code> function but this time we use the <code>normalized=TRUE</code> argument.</p>
<p>ğŸ¬ Save the normalised to a matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">normalised_counts</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">dds</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Make a dataframe of the normalised counts, adding a column for the gene ids at the same time:</p>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_normalised_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">normalised_counts</span>,</span>
<span>                                    gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">normalised_counts</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="differential-expression-analysis-2" class="level3"><h3 class="anchored" data-anchor-id="differential-expression-analysis-2">4. Differential expression analysis</h3>
<p>We use the <code>DESeq()</code> function to do the differential expression analysis. This function fits the statistical model to the data and then uses the model to calculate the significance of the difference between the treatments. It again stores the results in the DESseqDataSet object. Note that the differential expression needs the raw (unnormalised counts) as it does its own normalisation as part of the process.</p>
<p>ğŸ¬ Run the differential expression analysis and store the results in the same object:</p>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function will take only a few moments to run on this data but can take longer for bigger datasets.</p>
<p>We need to define the contrasts we want to test. We want to test the difference between the treatments so we will define the contrast as <code>procyclic</code> and <code>metacyclic</code>.</p>
<p>ğŸ¬ Define the contrast:</p>
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">contrast_pro_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stage"</span>, <span class="st">"procyclic"</span>, <span class="st">"metacyclic"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that <code>stage</code> is the name of the column in the metadata dataframe and <code>procyclic</code> and <code>metacyclic</code> are the names of the levels in the <code>stage</code> column. By putting them in the order <code>procyclic</code> , <code>metacyclic</code> we are saying the fold change will be procyclic / metacyclic. This means:</p>
<ul>
<li>positive log fold changes indicate procyclic &gt; metacyclic and</li>
<li>negative log fold changes indicates metacyclic &gt; procyclic.</li>
</ul>
<p>If we had put them in the order <code>metacyclic</code>, <code>procyclic</code> we would have the reverse.</p>
<p>ğŸ¬ Extract the results from the DESseqDataSet object:</p>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results_pro_meta</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                       contrast <span class="op">=</span> <span class="va">contrast_pro_meta</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This will give us the log<sub>2</sub> fold change, the <em>p</em>-value and the adjusted <em>p</em>-value for the comparison between procyclic and metacyclic stage for each gene</p>
<p>ğŸ¬ Put the results in a dataframe and add the gene ids as a column:</p>
<div class="cell">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">results_pro_meta</span>,</span>
<span>                          gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">results_pro_meta</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is useful to have the normalised counts and the statistical results in one dataframe.</p>
<p>ğŸ¬ Merge the two dataframes:</p>
<div class="cell">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># merge the results with the normalised counts</span></span>
<span><span class="va">pro_meta_results</span> <span class="op">&lt;-</span> <span class="va">pro_meta_normalised_counts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">pro_meta_results</span>, by <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#leishmania-add-info">Add gene information</a>.</p>
</section></section><section id="stem-cells-differential" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-differential">ğŸ­ Stem cells</h2>
<p>These are the steps we will take</p>
<ol type="1">
<li>Find the genes that are expressed in only one cell type (the prog or the hspc).</li>
<li>Prepare the data for differential expression analysis with the <strong><code>scran</code></strong> package.</li>
<li>Do differential expression analysis on the genes using the <strong><code>scran</code></strong> package. This needs to be done on the logged normalised counts.</li>
</ol>
<section id="genes-expressed-in-one-cell-type" class="level3"><h3 class="anchored" data-anchor-id="genes-expressed-in-one-cell-type">1. Genes expressed in one cell type</h3>
<p>The genes expressed in only cell type are those with zeros in all the cells of the other type. We can find these by summing the expression values for each gene across the cells of one type and filtering for those that are zero.</p>
<p>To do row wise aggregates such as the sum across rows we can use the <code><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise()</a></code> function. <code><a href="https://dplyr.tidyverse.org/reference/c_across.html">c_across()</a></code> allows us to use the colon notation to select columns. This is very useful when you have a lot of columns because it would be annoying to have to list all of them. <code>HSPC_001:HSPC_852</code> means all the columns from <code>HSPC_001</code> to <code>HSPC_852</code>.</p>
<p>ğŸ¬ Find the genes expressed only in progenitor cells (<em>i.e.</em>, those that are 0 in every HSPC cell):</p>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/c_across.html">c_across</a></span><span class="op">(</span><span class="va">HSPC_001</span><span class="op">:</span><span class="va">HSPC_852</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 0 Ã— 1,500
# Rowwise: 
# â„¹ 1,500 variables: ensembl_gene_id &lt;chr&gt;, HSPC_001 &lt;dbl&gt;, HSPC_002 &lt;dbl&gt;,
#   HSPC_003 &lt;dbl&gt;, HSPC_004 &lt;dbl&gt;, HSPC_006 &lt;dbl&gt;, HSPC_008 &lt;dbl&gt;,
#   HSPC_009 &lt;dbl&gt;, HSPC_011 &lt;dbl&gt;, HSPC_012 &lt;dbl&gt;, HSPC_014 &lt;dbl&gt;,
#   HSPC_015 &lt;dbl&gt;, HSPC_016 &lt;dbl&gt;, HSPC_017 &lt;dbl&gt;, HSPC_018 &lt;dbl&gt;,
#   HSPC_020 &lt;dbl&gt;, HSPC_021 &lt;dbl&gt;, HSPC_022 &lt;dbl&gt;, HSPC_023 &lt;dbl&gt;,
#   HSPC_024 &lt;dbl&gt;, HSPC_025 &lt;dbl&gt;, HSPC_026 &lt;dbl&gt;, HSPC_027 &lt;dbl&gt;,
#   HSPC_028 &lt;dbl&gt;, HSPC_030 &lt;dbl&gt;, HSPC_031 &lt;dbl&gt;, HSPC_033 &lt;dbl&gt;, â€¦</code></pre>
</div>
</div>
<p>We already know from last weekâ€™s work that there are no genes that are zero across all the cells (both types). If we did not know that, we would need to add <code>|&gt; filter(sum(c_across(Prog_001:Prog_852)) != 0)</code></p>
<p>meaning zero in all the HSPC but not zero in all the Prog</p>
<p>â“ How many genes are expressed only in the progenitor cells only</p>
<!-- #---THINKING ANSWER--- -->
<!-- No genes are expressed only in the progenitor cells. -->
<p>ğŸ¬ Now you find any genes that are expressed only in the HSPC cells.</p>
<p>â“ How many genes are expressed only in the HSPC cells?</p>
<!-- #---THINKING ANSWER--- -->
<!-- No genes genes are expressed only in the HSPC cells? -->
<p>â“ Do the results make sense to you in light of what you know about the biology?</p>
<!-- #---THINKING ANSWER--- -->
<!-- I'm not sure what to expect here. However, when there are many  -->
<!-- "replicates" it is less likely that a gene will be expressed  -->
<!-- in only one group. It's more likely that a gene will be -->
<!-- expressed in both groups but at different levels. -->
<!-- Domain expertise is important when analysing data so  -->
<!-- discuss your results with Jillian and your group. -->
<p>ğŸ¬ Write all the genes that are expressed one cell type only to file (saved in <code>results</code>)</p>
</section><section id="prepare-the-data-for-analysis-with-scran" class="level3"><h3 class="anchored" data-anchor-id="prepare-the-data-for-analysis-with-scran">2. Prepare the data for analysis with <strong><code>scran</code></strong>
</h3>
<p><strong><code>scran</code></strong> can use a matrix or a dataframe of counts but these must be log normalised counts. If using a dataframe, the columns must only contain the expression values (not the gene ids). The rows can be named to retain the gene ids.</p>
<p><code>hspc_prog</code> is a dataframe so we will use the ensembl gene ids to name the rows and remove the gene ids from the dataframe.</p>
<p>ğŸ¬ Add the gene ids as the row names:</p>
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog</span> <span class="op">&lt;-</span> <span class="va">hspc_prog</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Like <strong><code>DESeq2</code></strong>, <strong><code>scran</code></strong> needs metadata to define which columns were in which group. Instead of having this is a file, we will create a vector that indicates which column belongs to which cell type.</p>
<p>ğŸ¬ Create a vector that indicates which column belongs to which cell type:</p>
<div class="cell">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">n_hspc</span> <span class="op">&lt;-</span> <span class="fl">701</span></span>
<span><span class="va">n_prog</span> <span class="op">&lt;-</span> <span class="fl">798</span></span>
<span></span>
<span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"hspc"</span>,<span class="st">"prog"</span><span class="op">)</span>, </span>
<span>                 times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">n_hspc</span>, <span class="va">n_prog</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The number of times each cell type is repeated is the number of cells of that type. Do check that the length of the <code>cell_type</code> vector is the same as the number of columns in the <code>hspc_prog</code> dataframe.</p>
</section><section id="differential-expression-analysis-3" class="level3"><h3 class="anchored" data-anchor-id="differential-expression-analysis-3">3. Differential expression analysis</h3>
<p>ğŸ¬ Load the <strong><code>scran</code></strong> package:</p>
<p>Differential expression is carried out with the <code>findMarkers()</code> function. It takes two arguments. The first argument is the dataframe containing the data and the second argument is the vector indicating which columns are in which cell type. Make sure that the order of the cell types in the vector is appropriate for the order of the columns in the dataframe.</p>
<p>ğŸ¬ Run the differential expression analysis:</p>
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results_hspc_prog</span> <span class="op">&lt;-</span> <span class="fu">findMarkers</span><span class="op">(</span><span class="va">hspc_prog</span>, </span>
<span>                             <span class="va">cell_type</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output is a list object which, rather unnecessarily, includes two dataframes. This is not really necessary as the results are the same except for the fold change having a different sign.</p>
<ul>
<li>
<p>The dataframe <code>results_hspc_prog$prog</code> is log prog - log hspc (<em>i.e.</em>,Prog/HSPC). This means:</p>
<ul>
<li>Positive fold change: prog is higher than hspc</li>
<li>Negative fold change: hspc is higher than prog</li>
</ul>
</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>The results_hspc_prog$prog dataframe</caption>
<colgroup>
<col style="width: 24%">
<col style="width: 5%">
<col style="width: 10%">
<col style="width: 5%">
<col style="width: 17%">
<col style="width: 13%">
<col style="width: 24%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Top</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">FDR</th>
<th style="text-align: right;">summary.logFC</th>
<th style="text-align: right;">logFC.hspc</th>
<th style="text-align: left;">ensembl_gene_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000028639</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1.596910</td>
<td style="text-align: right;">1.596910</td>
<td style="text-align: left;">ENSMUSG00000028639</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000024053</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.035165</td>
<td style="text-align: right;">3.035165</td>
<td style="text-align: left;">ENSMUSG00000024053</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000041329</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.261056</td>
<td style="text-align: right;">3.261056</td>
<td style="text-align: left;">ENSMUSG00000041329</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000030336</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-2.146491</td>
<td style="text-align: right;">-2.146491</td>
<td style="text-align: left;">ENSMUSG00000030336</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000016494</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.056730</td>
<td style="text-align: right;">-3.056730</td>
<td style="text-align: left;">ENSMUSG00000016494</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000002808</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.000810</td>
<td style="text-align: right;">3.000810</td>
<td style="text-align: left;">ENSMUSG00000002808</td>
</tr>
</tbody>
</table>
</div>
</div>
<ul>
<li>
<p>The dataframe <code>results_hspc_prog$hspc</code> is log hspc - log prog (<em>i.e.</em>, HSPC/Prog). This means:</p>
<ul>
<li>Positive fold change: hspc is higher than prog</li>
<li>Negative fold change: prog is higher than hspc</li>
</ul>
</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>The results_hspc_prog$hspc dataframe. Notice the sign of the fold change is the other way</caption>
<colgroup>
<col style="width: 24%">
<col style="width: 5%">
<col style="width: 10%">
<col style="width: 5%">
<col style="width: 17%">
<col style="width: 13%">
<col style="width: 24%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Top</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">FDR</th>
<th style="text-align: right;">summary.logFC</th>
<th style="text-align: right;">logFC.prog</th>
<th style="text-align: left;">ensembl_gene_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000028639</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-1.596910</td>
<td style="text-align: right;">-1.596910</td>
<td style="text-align: left;">ENSMUSG00000028639</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000024053</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.035165</td>
<td style="text-align: right;">-3.035165</td>
<td style="text-align: left;">ENSMUSG00000024053</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000041329</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.261056</td>
<td style="text-align: right;">-3.261056</td>
<td style="text-align: left;">ENSMUSG00000041329</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000030336</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2.146491</td>
<td style="text-align: right;">2.146491</td>
<td style="text-align: left;">ENSMUSG00000030336</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ENSMUSG00000016494</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.056730</td>
<td style="text-align: right;">3.056730</td>
<td style="text-align: left;">ENSMUSG00000016494</td>
</tr>
<tr class="even">
<td style="text-align: left;">ENSMUSG00000002808</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-3.000810</td>
<td style="text-align: right;">-3.000810</td>
<td style="text-align: left;">ENSMUSG00000002808</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We only need one of these results dataframes and we will use the <code>prog</code> one. It does not matter which one you use but you do need keep the direction of the foldchange in mind when interpreting the results.</p>
<p>Notice that the dataframes are ordered by significance rather than the original gene order.</p>
<p>It is useful to have the normalised counts and the statistical results in one dataframe to which we will add the gene information from Ensembl. Having all the information together will make it easier to interpret the results and select genes of interest. We will need to extract the results from the list object, add the gene ids and then join with the normalised counts.</p>
<p>ğŸ¬ Extract the results dataframe from the list object and add the gene ids as a column:</p>
<div class="cell">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">results_hspc_prog</span><span class="op">$</span><span class="va">prog</span>, </span>
<span>                                ensembl_gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">results_hspc_prog</span><span class="op">$</span><span class="va">prog</span><span class="op">)</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Return the ensembl gene ids as a column to the normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog</span> <span class="op">&lt;-</span> <span class="va">hspc_prog</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Merge the results dataframe with the normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># merge the results with the normalised counts</span></span>
<span><span class="va">hspc_prog_results</span> <span class="op">&lt;-</span> <span class="va">hspc_prog_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">hspc_prog</span>, by <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#stem-add-info">Add gene information</a>.</p>
</section></section></section><section id="add-gene-information" class="level1"><h1>Add gene information</h1>
<section id="arabidopisis-add-info" class="level2"><h2 class="anchored" data-anchor-id="arabidopisis-add-info">ğŸ„ <em>Arabidopisis</em>
</h2>
<p><a href="https://www.ensembl.org/index.html">Ensembl</a> <span class="citation" data-cites="martin2023 birney2004">(<a href="#ref-martin2023" role="doc-biblioref">Martin et al. 2023</a>; <a href="#ref-birney2004" role="doc-biblioref">Birney et al. 2004</a>)</span>is a bioinformatics project to organise all the biological information around the sequences of large genomes. The are a large number of databases and <a href="https://www.ensembl.org/info/data/biomart/index.html">BioMart</a> <span class="citation" data-cites="smedley2009">(<a href="#ref-smedley2009" role="doc-biblioref">Smedley et al. 2009</a>)</span> provides a consistent interface to the material. There are web-based tools to use these but the R package <strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#ref-biomaRt1" role="doc-biblioref">Durinck et al. 2009</a>, <a href="#ref-biomaRt2" role="doc-biblioref">2005</a>)</span> gives you programmatic access making it easier to integrate information into R dataframes.</p>
<p>ğŸ¬ Load the <strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#ref-biomaRt1" role="doc-biblioref">Durinck et al. 2009</a>, <a href="#ref-biomaRt2" role="doc-biblioref">2005</a>)</span> package:</p>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Huber-group-EMBL/biomaRt">biomaRt</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <strong><code>biomaRt</code></strong> package includes a function to list all the available datasets</p>
<p>ğŸ¬ List the Ensembl â€œmartsâ€ available:</p>
<div class="cell">
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listEnsembl.html">listEnsemblGenomes</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              biomart                        version
1       protists_mart      Ensembl Protists Genes 61
2 protists_variations Ensembl Protists Variations 61
3          fungi_mart         Ensembl Fungi Genes 61
4    fungi_variations    Ensembl Fungi Variations 61
5        metazoa_mart       Ensembl Metazoa Genes 61
6  metazoa_variations  Ensembl Metazoa Variations 61
7         plants_mart        Ensembl Plants Genes 61
8   plants_variations   Ensembl Plants Variations 61</code></pre>
</div>
</div>
<p><code>plants_mart</code> looks like the one we want. We can see what genomes are available with names like â€œArabidopsisâ€ in this mart using the <code><a href="https://rdrr.io/pkg/biomaRt/man/listDatasets.html">searchDatasets()</a></code> function.</p>
<p>ğŸ¬</p>
<div class="cell">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listDatasets.html">searchDatasets</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useEnsembl.html">useEnsemblGenomes</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"plants_mart"</span><span class="op">)</span>, </span>
<span>               pattern <span class="op">=</span> <span class="st">"Arabidopsis"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             dataset                         description version
4   ahalleri_eg_gene Arabidopsis halleri genes (Ahal2.2) Ahal2.2
6    alyrata_eg_gene    Arabidopsis lyrata genes (v.1.0)   v.1.0
11 athaliana_eg_gene Arabidopsis thaliana genes (TAIR10)  TAIR10</code></pre>
</div>
</div>
<p><code>athaliana_eg_gene</code> is the Arabidopsis thaliana genes (TAIR10) dataset we want.</p>
<p>ğŸ¬ Connect to the <code>athaliana_eg_gene</code> database in <code>plants_mart</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useEnsembl.html">useEnsemblGenomes</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"plants_mart"</span>,</span>
<span>                             dataset <span class="op">=</span> <span class="st">"athaliana_eg_gene"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ See the the types of information we can retrieve:</p>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listAttributes.html">listAttributes</a></span><span class="op">(</span>mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are many (1,796!) possible bits of information (attributes) that can be obtained.</p>
<p>We use the <code><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM()</a></code> function to retrieve information from the database. The <code>filters</code> argument is used to specified what kind of identifier we are supplying in <code>values</code> to retrieve information. The <code>attributes</code> argument is used to select the information we want to retrieve. The <code>values</code> argument is used to specify the identifiers. The mart argument is used to specify the connection we created.</p>
<p>ğŸ¬ Get the the gene name and a description. We also retreive the gene id so we can later join the information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span>                   attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensembl_gene_id"</span>,</span>
<span>                                  <span class="st">"external_gene_name"</span>,</span>
<span>                                  <span class="st">"description"</span><span class="op">)</span>,</span>
<span>                   values <span class="op">=</span> <span class="va">root_results</span><span class="op">$</span><span class="va">gene_id</span>,</span>
<span>                   mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should view the resulting dataframe to see what information is available. You can use <code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code> or <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code>.</p>
<p>ğŸ¬ Merge the gene information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># join the gene info with the results</span></span>
<span><span class="va">root_results</span> <span class="op">&lt;-</span> <span class="va">root_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">gene_info</span>,</span>
<span>            by <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join_by.html">join_by</a></span><span class="op">(</span><span class="va">gene_id</span> <span class="op">==</span> <span class="va">ensembl_gene_id</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Save the results to a file:</p>
<div class="cell">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">root_results</span>, file <span class="op">=</span> <span class="st">"results/root_results.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#look-after-future-you">Look after future you</a>.</p>
</section><section id="leishmania-add-info" class="level2"><h2 class="anchored" data-anchor-id="leishmania-add-info">ğŸ’‰ <em>Leishmania</em>
</h2>
<ul>
<li><p>I got the information from <a href="https://tritrypdb.org/tritrypdb/app/">TriTrypDB</a><br>
which is a functional genomic resource for the Trypanosomatidae and Plasmodidae</p></li>
<li><p><a href="https://tritrypdb.org/tritrypdb/app/downloads" class="uri">https://tritrypdb.org/tritrypdb/app/downloads</a> section</p></li>
<li><p>I downloaded the <a href="https://tritrypdb.org/common/downloads/release-68/LmexicanaMHOMGT2001U1103/gff/data/TriTrypDB-68_LmexicanaMHOMGT2001U1103.gff"><em>L. mexicana</em> MHOM/GT/2001/U1103 Full GFF</a> and extracted the gene information and saved it as <a href="meta/leishmania_mex.xlsx">leishmania_mex.xlsx</a></p></li>
</ul>
<p>We will import this file and join it to the results dataframe.</p>
<p>ğŸ¬ Load the <strong><code>readxl</code></strong> <span class="citation" data-cites="readxl">(<a href="#ref-readxl" role="doc-biblioref">Wickham and Bryan 2023</a>)</span> package:</p>
<div class="cell">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org">readxl</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Import the Xenbase gene information file:</p>
<div class="cell">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_excel</a></span><span class="op">(</span><span class="st">"meta/leishmania_mex.xlsx"</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in read_excel("meta/leishmania_mex.xlsx"): could not find function "read_excel"</code></pre>
</div>
</div>
<p>You should view the resulting dataframe to see what information is available. You can use <code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code> or <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code>.</p>
<p>ğŸ¬ Merge the gene information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># join the gene info with the results</span></span>
<span><span class="va">pro_meta_results</span> <span class="op">&lt;-</span> <span class="va">pro_meta_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">gene_info</span>, by <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in `left_join()`:
! Join columns in `y` must be present in the data.
âœ– Problem with `gene_id`.</code></pre>
</div>
</div>
<p>ğŸ¬ Save the results to a file:</p>
<div class="cell">
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">pro_meta_results</span>, file <span class="op">=</span> <span class="st">"results/pro_meta_results.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#look-after-future-you">Look after future you</a>.</p>
</section><section id="stem-cells-add-info" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-add-info">ğŸ­ Stem cells</h2>
<p><a href="https://www.ensembl.org/index.html">Ensembl</a> <span class="citation" data-cites="martin2023 birney2004">(<a href="#ref-martin2023" role="doc-biblioref">Martin et al. 2023</a>; <a href="#ref-birney2004" role="doc-biblioref">Birney et al. 2004</a>)</span>is a bioinformatics project to organise all the biological information around the sequences of large genomes. The are a large number of databases but <a href="https://www.ensembl.org/info/data/biomart/index.html">BioMart</a> <span class="citation" data-cites="smedley2009">(<a href="#ref-smedley2009" role="doc-biblioref">Smedley et al. 2009</a>)</span> provides a consistent interface to the material. There are web-based tools to use these but the R package <strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#ref-biomaRt1" role="doc-biblioref">Durinck et al. 2009</a>, <a href="#ref-biomaRt2" role="doc-biblioref">2005</a>)</span> gives you programmatic access making it easier to integrate information into R dataframes</p>
<p>ğŸ¬ Load the <strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#ref-biomaRt1" role="doc-biblioref">Durinck et al. 2009</a>, <a href="#ref-biomaRt2" role="doc-biblioref">2005</a>)</span> package:</p>
<div class="cell">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Huber-group-EMBL/biomaRt">biomaRt</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Connect to the mouse database</p>
<div class="cell">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Connect to the mouse database</span></span>
<span></span>
<span><span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"ensembl"</span>, </span>
<span>                   dataset <span class="op">=</span> <span class="st">"mmusculus_gene_ensembl"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ See information we can retrieve:</p>
<div class="cell">
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># See what information we can retrieve</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listAttributes.html">listAttributes</a></span><span class="op">(</span>mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are many (2,988!) possible bits of information (attributes) that can be obtained.</p>
<p>We use the <code><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM()</a></code> function to retrieve information from the database. The <code>filters</code> argument is used to specified what kind of identifier we are supplying in <code>values</code> to retrieve information. The <code>attributes</code> argument is used to select the information we want to retrieve. The <code>values</code> argument is used to specify the identifiers. The mart argument is used to specify the connection we created.</p>
<p>ğŸ¬ Get the the gene name and a description. We also retrieve the gene id so we can later join the information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span>                   attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensembl_gene_id"</span>,</span>
<span>                                  <span class="st">"external_gene_name"</span>,</span>
<span>                                  <span class="st">"description"</span><span class="op">)</span>,</span>
<span>                   values <span class="op">=</span> <span class="va">hspc_prog_results</span><span class="op">$</span><span class="va">ensembl_gene_id</span>,</span>
<span>                   mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You should view the resulting dataframe to see what information is available. You can use <code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code> or <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code>. Notice the dataframe returned only has 279 rows - one of the ids does not have information.</p>
<p>ğŸ¬ Merge the gene information with the results:</p>
<div class="cell">
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># join the gene info with the results</span></span>
<span><span class="va">hspc_prog_results</span> <span class="op">&lt;-</span> <span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">gene_info</span>, by <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Save the results to a file:</p>
<div class="cell">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">hspc_prog_results</span>, file <span class="op">=</span> <span class="st">"results/hspc_prog_results.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now go to <a href="#look-after-future-you">Look after future you</a>.</p>
</section></section><section id="look-after-future-you" class="level1"><h1>ğŸ¤— Look after future you!</h1>
<p>ğŸ¬ Go through you script and tidy up. You might need to :</p>
<ul>
<li>collect together library statements at the top of the script</li>
<li>remove code that you needed to start today but which wouldnâ€™t be needed if you were running the script from the top (e.g., importing)</li>
<li>edit your comments for clarity</li>
<li>rename variables for consistency or clarity</li>
<li>remove house keeping or exploratory code</li>
<li>restyle code, add code section headers etc</li>
</ul></section><section id="finished" class="level1"><h1>ğŸ¥³ Finished</h1>
<p>Well Done!</p>
</section><section id="independent-study-following-the-workshop" class="level1"><h1>Independent study following the workshop</h1>
<p><a href="../../transcriptomics/week-4/study_after_workshop.html">Consolidate</a></p>
</section><section id="the-code-file" class="level1"><h1>The Code file</h1>
<p>These contain all the code needed in the workshop even where it is not visible on the webpage.</p>
<p>The <a href="../../transcriptomics/week-4/workshop.html">workshop.qmd</a> file is the file I use to compile the practical. Qmd stands for Quarto markdown. It allows code and ordinary text to be interleaved to produce well-formatted reports including webpages. View the source code for this workshop using the <code>&lt;/&gt; Code</code> button at the top of the page. Coding and thinking answers are marked with <code>#---CODING ANSWER---</code> and <code>#---THINKING ANSWER---</code></p>
<p>Pages made with R <span class="citation" data-cites="R-core">(<a href="#ref-R-core" role="doc-biblioref">R Core Team 2024</a>)</span>, Quarto <span class="citation" data-cites="Allaire_Quarto_2024">(<a href="#ref-Allaire_Quarto_2024" role="doc-biblioref">Allaire et al. 2024</a>)</span>, <code>knitr</code> <span class="citation" data-cites="knitr1 knitr2 knitr3">(<a href="#ref-knitr1" role="doc-biblioref">Xie 2024</a>, <a href="#ref-knitr2" role="doc-biblioref">2015</a>, <a href="#ref-knitr3" role="doc-biblioref">2014</a>)</span>, <code>kableExtra</code> <span class="citation" data-cites="kableExtra">(<a href="#ref-kableExtra" role="doc-biblioref">Zhu 2021</a>)</span></p>
</section><section id="references" class="level1">


<!-- -->


</section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Allaire_Quarto_2024" class="csl-entry" role="listitem">
Allaire, J. J., Charles Teague, Carlos Scheidegger, Yihui Xie, and Christophe Dervieux. 2024. <span>â€œ<span>Quarto</span>.â€</span> <a href="https://doi.org/10.5281/zenodo.5960048">https://doi.org/10.5281/zenodo.5960048</a>.
</div>
<div id="ref-birney2004" class="csl-entry" role="listitem">
Birney, Ewan, T. Daniel Andrews, Paul Bevan, Mario Caccamo, Yuan Chen, Laura Clarke, Guy Coates, et al. 2004. <span>â€œAn Overview of Ensembl.â€</span> <em>Genome Research</em> 14 (5): 925â€“28. <a href="https://doi.org/10.1101/gr.1860604">https://doi.org/10.1101/gr.1860604</a>.
</div>
<div id="ref-biomaRt2" class="csl-entry" role="listitem">
Durinck, Steffen, Yves Moreau, Arek Kasprzyk, Sean Davis, Bart De Moor, Alvis Brazma, and Wolfgang Huber. 2005. <span>â€œBioMart and Bioconductor: A Powerful Link Between Biological Databases and Microarray Data Analysis.â€</span> <em>Bioinformatics</em> 21: 3439â€“40.
</div>
<div id="ref-biomaRt1" class="csl-entry" role="listitem">
Durinck, Steffen, Paul T. Spellman, Ewan Birney, and Wolfgang Huber. 2009. <span>â€œMapping Identifiers for the Integration of Genomic Datasets with the r/Bioconductor Package biomaRt.â€</span> <em>Nature Protocols</em> 4: 1184â€“91.
</div>
<div id="ref-DESeq2" class="csl-entry" role="listitem">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>â€œModerated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.â€</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-scran" class="csl-entry" role="listitem">
Lun, Aaron T. L., Davis J. McCarthy, and John C. Marioni. 2016. <span>â€œA Step-by-Step Workflow for Low-Level Analysis of Single-Cell RNA-Seq Data with Bioconductor.â€</span> <em>F1000Res.</em> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-martin2023" class="csl-entry" role="listitem">
Martin, Fergal J, M Ridwan Amode, Alisha Aneja, Olanrewaju Austine-Orimoloye, Andrey&nbsp;G Azov, If Barnes, Arne Becker, et al. 2023. <span>â€œEnsembl 2023.â€</span> <em>Nucleic Acids Research</em> 51 (D1): D933â€“41. <a href="https://doi.org/10.1093/nar/gkac958">https://doi.org/10.1093/nar/gkac958</a>.
</div>
<div id="ref-R-core" class="csl-entry" role="listitem">
R Core Team. 2024. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-smedley2009" class="csl-entry" role="listitem">
Smedley, Damian, Syed Haider, Benoit Ballester, Richard Holland, Darin London, Gudmundur Thorisson, and Arek Kasprzyk. 2009. <span>â€œBioMart <span></span> Biological Queries Made Easy.â€</span> <em>BMC Genomics</em> 10 (1): 22. <a href="https://doi.org/10.1186/1471-2164-10-22">https://doi.org/10.1186/1471-2164-10-22</a>.
</div>
<div id="ref-conflicted" class="csl-entry" role="listitem">
Wickham, Hadley. 2023. <em>Conflicted: An Alternative Conflict Resolution Strategy</em>. <a href="https://conflicted.r-lib.org/">https://conflicted.r-lib.org/</a>.
</div>
<div id="ref-tidyverse" class="csl-entry" role="listitem">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy Dâ€™Agostino McGowan, Romain FranÃ§ois, Garrett Grolemund, et al. 2019. <span>â€œWelcome to the <span class="nocase">tidyverse</span>.â€</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-readxl" class="csl-entry" role="listitem">
Wickham, Hadley, and Jennifer Bryan. 2023. <em>Readxl: Read Excel Files</em>. <a href="https://readxl.tidyverse.org">https://readxl.tidyverse.org</a>.
</div>
<div id="ref-knitr3" class="csl-entry" role="listitem">
Xie, Yihui. 2014. <span>â€œKnitr: A Comprehensive Tool for Reproducible Research in <span>R</span>.â€</span> In <em>Implementing Reproducible Computational Research</em>, edited by Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman; Hall/CRC.
</div>
<div id="ref-knitr2" class="csl-entry" role="listitem">
â€”â€”â€”. 2015. <em>Dynamic Documents with <span>R</span> and Knitr</em>. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-knitr1" class="csl-entry" role="listitem">
â€”â€”â€”. 2024. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in r</em>. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-kableExtra" class="csl-entry" role="listitem">
Zhu, Hao. 2021. <span>â€œkableExtra: Construct Complex Table with â€™Kableâ€™ and Pipe Syntax.â€</span> <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>
<ol>
<li id="fn1"><p><a href="https://www.bioconductor.org/">Bioconductor</a> is a project that develops and supports R packages for bioinformatics.<a href="#fnref1" class="footnote-back" role="doc-backlink">â†©ï¸</a></p></li>
<li id="fn2"><p><a href="https://www.bioconductor.org/">Bioconductor</a> is a project that develops and supports R packages for bioinformatics.<a href="#fnref2" class="footnote-back" role="doc-backlink">â†©ï¸</a></p></li>
</ol></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("^(?:http:|https:)\/\/3mmarand\.github\.io\/BIO00088H-data");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb76" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Workshop"</span></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="an">subtitle:</span><span class="co"> "Transcripttranscriptomics 2: Statistical Analysis"</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Emma Rand"</span></span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a><span class="an">toc:</span><span class="co"> true</span></span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a><span class="an">toc-depth:</span><span class="co"> 4</span></span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a><span class="an">toc-location:</span><span class="co"> right</span></span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span></span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a><span class="co">  echo: true</span></span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a><span class="co">  include: true</span></span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a><span class="co">  error: true</span></span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a><span class="an">bibliography:</span><span class="co"> ../../references.bib</span></span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span><span class="co"> </span></span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown: </span></span>
<span id="cb76-15"><a href="#cb76-15" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb76-16"><a href="#cb76-16" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb76-17"><a href="#cb76-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-18"><a href="#cb76-18" aria-hidden="true" tabindex="-1"></a><span class="fu"># Introduction</span></span>
<span id="cb76-19"><a href="#cb76-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-20"><a href="#cb76-20" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session overview</span></span>
<span id="cb76-21"><a href="#cb76-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-22"><a href="#cb76-22" aria-hidden="true" tabindex="-1"></a>In the workshop, you will learn how to perform differential expression</span>
<span id="cb76-23"><a href="#cb76-23" aria-hidden="true" tabindex="-1"></a>analysis on raw counts using **`DESeq2`** <span class="co">[</span><span class="ot">@DESeq2</span><span class="co">]</span> or on logged</span>
<span id="cb76-24"><a href="#cb76-24" aria-hidden="true" tabindex="-1"></a>normalised expression values using **`scran`** <span class="co">[</span><span class="ot">@scran</span><span class="co">]</span> or both.</span>
<span id="cb76-25"><a href="#cb76-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-26"><a href="#cb76-26" aria-hidden="true" tabindex="-1"></a><span class="fu"># Set up</span></span>
<span id="cb76-27"><a href="#cb76-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-28"><a href="#cb76-28" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ„ *Arabidopisis*</span></span>
<span id="cb76-29"><a href="#cb76-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-30"><a href="#cb76-30" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Open the <span class="in">`arab-88H`</span> RStudio Project and the <span class="in">`cont-low-root.R`</span> script.</span>
<span id="cb76-31"><a href="#cb76-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-32"><a href="#cb76-32" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ’‰ *Leishmania*</span></span>
<span id="cb76-33"><a href="#cb76-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-34"><a href="#cb76-34" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Open the <span class="in">`leish-88H`</span> RStudio Project and the <span class="in">`pro-meta.R`</span> script.</span>
<span id="cb76-35"><a href="#cb76-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-36"><a href="#cb76-36" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ­ Stem cells</span></span>
<span id="cb76-37"><a href="#cb76-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-38"><a href="#cb76-38" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Open the <span class="in">`mice-88H`</span> RStudio Project and the <span class="in">`hspc-prog.R`</span> script.</span>
<span id="cb76-39"><a href="#cb76-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-40"><a href="#cb76-40" aria-hidden="true" tabindex="-1"></a><span class="fu">## Everyone</span></span>
<span id="cb76-41"><a href="#cb76-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-42"><a href="#cb76-42" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Make a new folder <span class="in">`results`</span> in the project directory.</span>
<span id="cb76-43"><a href="#cb76-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-44"><a href="#cb76-44" aria-hidden="true" tabindex="-1"></a>This is where we will save our results.</span>
<span id="cb76-45"><a href="#cb76-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-46"><a href="#cb76-46" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load **`tidyverse`** <span class="co">[</span><span class="ot">@tidyverse</span><span class="co">]</span> You most likely have this code at</span>
<span id="cb76-47"><a href="#cb76-47" aria-hidden="true" tabindex="-1"></a>the top of your script already.</span>
<span id="cb76-48"><a href="#cb76-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-51"><a href="#cb76-51" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-52"><a href="#cb76-52" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb76-53"><a href="#cb76-53" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-54"><a href="#cb76-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-55"><a href="#cb76-55" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb76-56"><a href="#cb76-56" aria-hidden="true" tabindex="-1"></a><span class="in">â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€</span></span>
<span id="cb76-57"><a href="#cb76-57" aria-hidden="true" tabindex="-1"></a><span class="in">âœ” dplyr     1.1.3     âœ” readr     2.1.4</span></span>
<span id="cb76-58"><a href="#cb76-58" aria-hidden="true" tabindex="-1"></a><span class="in">âœ” forcats   1.0.0     âœ” stringr   1.5.0</span></span>
<span id="cb76-59"><a href="#cb76-59" aria-hidden="true" tabindex="-1"></a><span class="in">âœ” ggplot2   3.4.3     âœ” tibble    3.2.1</span></span>
<span id="cb76-60"><a href="#cb76-60" aria-hidden="true" tabindex="-1"></a><span class="in">âœ” lubridate 1.9.3     âœ” tidyr     1.3.0</span></span>
<span id="cb76-61"><a href="#cb76-61" aria-hidden="true" tabindex="-1"></a><span class="in">âœ” purrr     1.0.2     </span></span>
<span id="cb76-62"><a href="#cb76-62" aria-hidden="true" tabindex="-1"></a><span class="in">â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€</span></span>
<span id="cb76-63"><a href="#cb76-63" aria-hidden="true" tabindex="-1"></a><span class="in">âœ– dplyr::filter() masks stats::filter()</span></span>
<span id="cb76-64"><a href="#cb76-64" aria-hidden="true" tabindex="-1"></a><span class="in">âœ– dplyr::lag()    masks stats::lag()</span></span>
<span id="cb76-65"><a href="#cb76-65" aria-hidden="true" tabindex="-1"></a><span class="in">â„¹ Use the conflicted package to force all conflicts to become errors</span></span>
<span id="cb76-66"><a href="#cb76-66" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-67"><a href="#cb76-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-68"><a href="#cb76-68" aria-hidden="true" tabindex="-1"></a>Have you ever stopped to think about this message? It is telling us that</span>
<span id="cb76-69"><a href="#cb76-69" aria-hidden="true" tabindex="-1"></a>there are functions in the <span class="in">`dplyr`</span> package that have the same name as</span>
<span id="cb76-70"><a href="#cb76-70" aria-hidden="true" tabindex="-1"></a>functions in the <span class="in">`stats`</span> package and that R will use the <span class="in">`dplyr`</span></span>
<span id="cb76-71"><a href="#cb76-71" aria-hidden="true" tabindex="-1"></a>version. As this is what you want, this has always been fine. It still</span>
<span id="cb76-72"><a href="#cb76-72" aria-hidden="true" tabindex="-1"></a>is fine in this case. However, as you start to load more packages, you</span>
<span id="cb76-73"><a href="#cb76-73" aria-hidden="true" tabindex="-1"></a>will want to know if you are using a function from a package that has</span>
<span id="cb76-74"><a href="#cb76-74" aria-hidden="true" tabindex="-1"></a>the same name as a function in another loaded package. This is where the</span>
<span id="cb76-75"><a href="#cb76-75" aria-hidden="true" tabindex="-1"></a>**`conflicted`** [@conflicted] package comes in. **`Conflicted`** will</span>
<span id="cb76-76"><a href="#cb76-76" aria-hidden="true" tabindex="-1"></a>warn you when you are using a function that has the same name as a</span>
<span id="cb76-77"><a href="#cb76-77" aria-hidden="true" tabindex="-1"></a>function in another package. You can then choose which function to use.</span>
<span id="cb76-78"><a href="#cb76-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-79"><a href="#cb76-79" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load the **`conflicted`** package:</span>
<span id="cb76-80"><a href="#cb76-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-83"><a href="#cb76-83" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-84"><a href="#cb76-84" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(conflicted)</span>
<span id="cb76-85"><a href="#cb76-85" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-86"><a href="#cb76-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-87"><a href="#cb76-87" aria-hidden="true" tabindex="-1"></a>Instead of getting a warning every time you are using a function that</span>
<span id="cb76-88"><a href="#cb76-88" aria-hidden="true" tabindex="-1"></a>has a function with the same name in another package, we can declare a</span>
<span id="cb76-89"><a href="#cb76-89" aria-hidden="true" tabindex="-1"></a>preference for one function over another. This is useful for the</span>
<span id="cb76-90"><a href="#cb76-90" aria-hidden="true" tabindex="-1"></a>functions you use a lot or ones where you are certain you always want to</span>
<span id="cb76-91"><a href="#cb76-91" aria-hidden="true" tabindex="-1"></a>use a particular function.</span>
<span id="cb76-92"><a href="#cb76-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-93"><a href="#cb76-93" aria-hidden="true" tabindex="-1"></a>For example, to always use the **`dplyr`** version of <span class="in">`filter()`</span> by</span>
<span id="cb76-94"><a href="#cb76-94" aria-hidden="true" tabindex="-1"></a>default you can add this to the top of your script:</span>
<span id="cb76-95"><a href="#cb76-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-98"><a href="#cb76-98" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-99"><a href="#cb76-99" aria-hidden="true" tabindex="-1"></a><span class="fu">conflicts_prefer</span>(dplyr<span class="sc">::</span>filter)</span>
<span id="cb76-100"><a href="#cb76-100" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-101"><a href="#cb76-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-102"><a href="#cb76-102" aria-hidden="true" tabindex="-1"></a>We will also want to ensure that we are using the <span class="in">`setdiff()`</span> function</span>
<span id="cb76-103"><a href="#cb76-103" aria-hidden="true" tabindex="-1"></a>from the **`GenomicRanges`** package which is used by **`DESeq2`**.</span>
<span id="cb76-104"><a href="#cb76-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-107"><a href="#cb76-107" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-108"><a href="#cb76-108" aria-hidden="true" tabindex="-1"></a><span class="fu">conflicts_prefer</span>(GenomicRanges<span class="sc">::</span>setdiff)</span>
<span id="cb76-109"><a href="#cb76-109" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-110"><a href="#cb76-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-111"><a href="#cb76-111" aria-hidden="true" tabindex="-1"></a><span class="fu"># Import</span></span>
<span id="cb76-112"><a href="#cb76-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-113"><a href="#cb76-113" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ„ *Arabidopisis*</span></span>
<span id="cb76-114"><a href="#cb76-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-115"><a href="#cb76-115" aria-hidden="true" tabindex="-1"></a>We need to import the root tissue data that were filtered to remove</span>
<span id="cb76-116"><a href="#cb76-116" aria-hidden="true" tabindex="-1"></a>genes with 3 or 4 zeros and those where the total counts was less than</span>
<span id="cb76-117"><a href="#cb76-117" aria-hidden="true" tabindex="-1"></a>20.</span>
<span id="cb76-118"><a href="#cb76-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-119"><a href="#cb76-119" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Import the data from the <span class="in">`data-processed`</span> folder.</span>
<span id="cb76-120"><a href="#cb76-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-123"><a href="#cb76-123" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-124"><a href="#cb76-124" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-125"><a href="#cb76-125" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-126"><a href="#cb76-126" aria-hidden="true" tabindex="-1"></a>root_filtered <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-processed/root_filtered.csv"</span>)</span>
<span id="cb76-127"><a href="#cb76-127" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-128"><a href="#cb76-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-129"><a href="#cb76-129" aria-hidden="true" tabindex="-1"></a>Now go to [Differential Expression</span>
<span id="cb76-130"><a href="#cb76-130" aria-hidden="true" tabindex="-1"></a>Analysis](#arabidopisis-differential).</span>
<span id="cb76-131"><a href="#cb76-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-132"><a href="#cb76-132" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ’‰ *Leishmania*</span></span>
<span id="cb76-133"><a href="#cb76-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-134"><a href="#cb76-134" aria-hidden="true" tabindex="-1"></a>We need to import the procyclic- and metacyclic-promastigote data that</span>
<span id="cb76-135"><a href="#cb76-135" aria-hidden="true" tabindex="-1"></a>were filtered to remove genes with 4, 5 or 6 zeros and those where the</span>
<span id="cb76-136"><a href="#cb76-136" aria-hidden="true" tabindex="-1"></a>total counts was less than 20.</span>
<span id="cb76-137"><a href="#cb76-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-138"><a href="#cb76-138" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Import the data from the <span class="in">`data-processed`</span> folder.</span>
<span id="cb76-139"><a href="#cb76-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-142"><a href="#cb76-142" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-143"><a href="#cb76-143" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-144"><a href="#cb76-144" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-145"><a href="#cb76-145" aria-hidden="true" tabindex="-1"></a>pro_meta_filtered <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-processed/pro_meta_filtered.csv"</span>)</span>
<span id="cb76-146"><a href="#cb76-146" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-147"><a href="#cb76-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-148"><a href="#cb76-148" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Differential Expression Analysis</span><span class="co">](#leishmania-differential)</span>.</span>
<span id="cb76-149"><a href="#cb76-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-150"><a href="#cb76-150" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ­ Stem cells</span></span>
<span id="cb76-151"><a href="#cb76-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-152"><a href="#cb76-152" aria-hidden="true" tabindex="-1"></a>We need to import the progenitor and HSPC cell data that were combined.</span>
<span id="cb76-153"><a href="#cb76-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-154"><a href="#cb76-154" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Import the data from the <span class="in">`data-processed`</span> folder..</span>
<span id="cb76-155"><a href="#cb76-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-158"><a href="#cb76-158" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-159"><a href="#cb76-159" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-160"><a href="#cb76-160" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-161"><a href="#cb76-161" aria-hidden="true" tabindex="-1"></a>hspc_prog <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-processed/hspc_prog.csv"</span>)</span>
<span id="cb76-162"><a href="#cb76-162" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-163"><a href="#cb76-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-164"><a href="#cb76-164" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Differential Expression Analysis</span><span class="co">](#stem-cells-differential)</span>.</span>
<span id="cb76-165"><a href="#cb76-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-166"><a href="#cb76-166" aria-hidden="true" tabindex="-1"></a><span class="fu"># Differential Expression Analysis</span></span>
<span id="cb76-167"><a href="#cb76-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-168"><a href="#cb76-168" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ„ *Arabidopisis* {#arabidopisis-differential}</span></span>
<span id="cb76-169"><a href="#cb76-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-170"><a href="#cb76-170" aria-hidden="true" tabindex="-1"></a>These are the steps we will take</span>
<span id="cb76-171"><a href="#cb76-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-172"><a href="#cb76-172" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Find the genes that are expressed in only one treatment group.</span>
<span id="cb76-173"><a href="#cb76-173" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Create a DESeqDataSet object. This is a special object that is used</span>
<span id="cb76-174"><a href="#cb76-174" aria-hidden="true" tabindex="-1"></a>    by the **`DESeq2`** package</span>
<span id="cb76-175"><a href="#cb76-175" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Prepare the normalised counts from the DESeqDataSet object.</span>
<span id="cb76-176"><a href="#cb76-176" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Do differential expression analysis on the genes. This needs to be</span>
<span id="cb76-177"><a href="#cb76-177" aria-hidden="true" tabindex="-1"></a>    done on the raw counts.</span>
<span id="cb76-178"><a href="#cb76-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-179"><a href="#cb76-179" aria-hidden="true" tabindex="-1"></a>All but the first step are done with the **`DESeq2`** package</span>
<span id="cb76-180"><a href="#cb76-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-181"><a href="#cb76-181" aria-hidden="true" tabindex="-1"></a><span class="fu">### 1. Genes expressed in one treatment</span></span>
<span id="cb76-182"><a href="#cb76-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-183"><a href="#cb76-183" aria-hidden="true" tabindex="-1"></a>The genes expressed in only one treatment group are those with zeros in</span>
<span id="cb76-184"><a href="#cb76-184" aria-hidden="true" tabindex="-1"></a>all replicates in one group and non-zero values in all replicates in the</span>
<span id="cb76-185"><a href="#cb76-185" aria-hidden="true" tabindex="-1"></a>other group. For example, those shown here:</span>
<span id="cb76-186"><a href="#cb76-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-189"><a href="#cb76-189" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-190"><a href="#cb76-190" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-191"><a href="#cb76-191" aria-hidden="true" tabindex="-1"></a><span class="co"># sort the data by the values in the control column</span></span>
<span id="cb76-192"><a href="#cb76-192" aria-hidden="true" tabindex="-1"></a>root_filtered <span class="sc">|&gt;</span>  </span>
<span id="cb76-193"><a href="#cb76-193" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(CTR1 <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">|&gt;</span> </span>
<span id="cb76-194"><a href="#cb76-194" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(CTR2 <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">|&gt;</span></span>
<span id="cb76-195"><a href="#cb76-195" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(CTR3 <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">|&gt;</span></span>
<span id="cb76-196"><a href="#cb76-196" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(CTR4 <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">|&gt;</span></span>
<span id="cb76-197"><a href="#cb76-197" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(CTR5 <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">|&gt;</span></span>
<span id="cb76-198"><a href="#cb76-198" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(CTR6 <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">|&gt;</span></span>
<span id="cb76-199"><a href="#cb76-199" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">5</span>) <span class="sc">|&gt;</span></span>
<span id="cb76-200"><a href="#cb76-200" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span>
<span id="cb76-201"><a href="#cb76-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-202"><a href="#cb76-202" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-203"><a href="#cb76-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-204"><a href="#cb76-204" aria-hidden="true" tabindex="-1"></a>We will use <span class="in">`filter()`</span> to find these genes.</span>
<span id="cb76-205"><a href="#cb76-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-206"><a href="#cb76-206" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Find the genes that are expressed only in the control group:</span>
<span id="cb76-207"><a href="#cb76-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-210"><a href="#cb76-210" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-211"><a href="#cb76-211" aria-hidden="true" tabindex="-1"></a>root_lowni_only <span class="ot">&lt;-</span> root_filtered <span class="sc">|&gt;</span></span>
<span id="cb76-212"><a href="#cb76-212" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">if_all</span>(<span class="fu">starts_with</span>(<span class="st">"CTR"</span>), \(x) x <span class="sc">==</span> <span class="dv">0</span>),</span>
<span id="cb76-213"><a href="#cb76-213" aria-hidden="true" tabindex="-1"></a>         <span class="fu">if_all</span>(<span class="fu">starts_with</span>(<span class="st">"LWR"</span>), \(x) x <span class="sc">&gt;</span> <span class="dv">0</span>) )</span>
<span id="cb76-214"><a href="#cb76-214" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-215"><a href="#cb76-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-216"><a href="#cb76-216" aria-hidden="true" tabindex="-1"></a>This code creates a new data frame called <span class="in">`root_lowni_only`</span> that</span>
<span id="cb76-217"><a href="#cb76-217" aria-hidden="true" tabindex="-1"></a>contains only the rows from root_filtered where:</span>
<span id="cb76-218"><a href="#cb76-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-219"><a href="#cb76-219" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>All CTR columns have a value of 0</span>
<span id="cb76-220"><a href="#cb76-220" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>All LWR columns have a value greater than 0</span>
<span id="cb76-221"><a href="#cb76-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-222"><a href="#cb76-222" aria-hidden="true" tabindex="-1"></a><span class="in">`filter()`</span> keeps rows that match the conditions you give it. Here we</span>
<span id="cb76-223"><a href="#cb76-223" aria-hidden="true" tabindex="-1"></a>give two conditions:</span>
<span id="cb76-224"><a href="#cb76-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-225"><a href="#cb76-225" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>All columns whose names start with "CTR" must have a value of 0.</span>
<span id="cb76-226"><a href="#cb76-226" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>All columns whose names start with "LWR" must have a value greater</span>
<span id="cb76-227"><a href="#cb76-227" aria-hidden="true" tabindex="-1"></a>    than 0.</span>
<span id="cb76-228"><a href="#cb76-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-229"><a href="#cb76-229" aria-hidden="true" tabindex="-1"></a>The <span class="in">`if_all()`</span> function applies a test to each selected column (those</span>
<span id="cb76-230"><a href="#cb76-230" aria-hidden="true" tabindex="-1"></a>starting with "CTR" or "LWR") for each row. It returns <span class="in">`TRUE`</span> only if</span>
<span id="cb76-231"><a href="#cb76-231" aria-hidden="true" tabindex="-1"></a>all the selected columns pass the test.</span>
<span id="cb76-232"><a href="#cb76-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-233"><a href="#cb76-233" aria-hidden="true" tabindex="-1"></a>The bit <span class="in">`\(x) x == 0`</span> is an anonymous function â€” a function without a</span>
<span id="cb76-234"><a href="#cb76-234" aria-hidden="true" tabindex="-1"></a>name. In base R syntax, itâ€™s equivalent to: <span class="in">`function(x) x == 0`</span> Here,</span>
<span id="cb76-235"><a href="#cb76-235" aria-hidden="true" tabindex="-1"></a><span class="in">`x`</span> will be the values from one column at a time, and the function</span>
<span id="cb76-236"><a href="#cb76-236" aria-hidden="true" tabindex="-1"></a>checks whether they are equal to 0. We use an anonymous function because</span>
<span id="cb76-237"><a href="#cb76-237" aria-hidden="true" tabindex="-1"></a>the condition needs to be applied separately to each column, but itâ€™s</span>
<span id="cb76-238"><a href="#cb76-238" aria-hidden="true" tabindex="-1"></a>simple enough that thereâ€™s no need to define a full named function</span>
<span id="cb76-239"><a href="#cb76-239" aria-hidden="true" tabindex="-1"></a>elsewhere in the script.</span>
<span id="cb76-240"><a href="#cb76-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-241"><a href="#cb76-241" aria-hidden="true" tabindex="-1"></a>â“ How many genes are expressed only in the low nickel group?</span>
<span id="cb76-242"><a href="#cb76-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-243"><a href="#cb76-243" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb76-244"><a href="#cb76-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-245"><a href="#cb76-245" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- There are 16 genes expressed only in low nickel group?. --&gt;</span></span>
<span id="cb76-246"><a href="#cb76-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-247"><a href="#cb76-247" aria-hidden="true" tabindex="-1"></a>Note: There may have been more genes with counts in the low nickel group</span>
<span id="cb76-248"><a href="#cb76-248" aria-hidden="true" tabindex="-1"></a>that were removed when we filtered on the total number of counts being</span>
<span id="cb76-249"><a href="#cb76-249" aria-hidden="true" tabindex="-1"></a>less that 50.</span>
<span id="cb76-250"><a href="#cb76-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-251"><a href="#cb76-251" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Now you find any genes that are expressed only in the control group.</span>
<span id="cb76-252"><a href="#cb76-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-255"><a href="#cb76-255" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-256"><a href="#cb76-256" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-257"><a href="#cb76-257" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-258"><a href="#cb76-258" aria-hidden="true" tabindex="-1"></a>root_cont_only <span class="ot">&lt;-</span> root_filtered <span class="sc">|&gt;</span></span>
<span id="cb76-259"><a href="#cb76-259" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">if_all</span>(<span class="fu">starts_with</span>(<span class="st">"CTR"</span>), \(x) x <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb76-260"><a href="#cb76-260" aria-hidden="true" tabindex="-1"></a>         <span class="fu">if_all</span>(<span class="fu">starts_with</span>(<span class="st">"LWR"</span>), \(x) x <span class="sc">==</span> <span class="dv">0</span>) )</span>
<span id="cb76-261"><a href="#cb76-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-262"><a href="#cb76-262" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-263"><a href="#cb76-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-264"><a href="#cb76-264" aria-hidden="true" tabindex="-1"></a>â“ How many genes are expressed only in the control group?</span>
<span id="cb76-265"><a href="#cb76-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-266"><a href="#cb76-266" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb76-267"><a href="#cb76-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-268"><a href="#cb76-268" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- There are 7 genes expressed only in the control group. --&gt;</span></span>
<span id="cb76-269"><a href="#cb76-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-270"><a href="#cb76-270" aria-hidden="true" tabindex="-1"></a>â“ Do the results make sense to you in light of what you know about the</span>
<span id="cb76-271"><a href="#cb76-271" aria-hidden="true" tabindex="-1"></a>biology?</span>
<span id="cb76-272"><a href="#cb76-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-273"><a href="#cb76-273" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb76-274"><a href="#cb76-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-275"><a href="#cb76-275" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- We might expect more genes to be switched on cope with nickel --&gt;</span></span>
<span id="cb76-276"><a href="#cb76-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-277"><a href="#cb76-277" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- than in control conditions --&gt;</span></span>
<span id="cb76-278"><a href="#cb76-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-279"><a href="#cb76-279" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- However, I'm not an expert in Arabidopsis, or plants! --&gt;</span></span>
<span id="cb76-280"><a href="#cb76-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-281"><a href="#cb76-281" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- Domain expertise is important when analysing data so  --&gt;</span></span>
<span id="cb76-282"><a href="#cb76-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-283"><a href="#cb76-283" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- discuss your results with Liz and your group. --&gt;</span></span>
<span id="cb76-284"><a href="#cb76-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-285"><a href="#cb76-285" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Write all the genes that are expressed one group only to file (saved</span>
<span id="cb76-286"><a href="#cb76-286" aria-hidden="true" tabindex="-1"></a>in <span class="in">`results`</span>)</span>
<span id="cb76-287"><a href="#cb76-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-290"><a href="#cb76-290" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-291"><a href="#cb76-291" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-292"><a href="#cb76-292" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-293"><a href="#cb76-293" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(root_lowni_only, <span class="st">"results/root_lowni_only.csv"</span>)</span>
<span id="cb76-294"><a href="#cb76-294" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(root_cont_only, <span class="st">"results/root_cont_only.csv"</span>)</span>
<span id="cb76-295"><a href="#cb76-295" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-296"><a href="#cb76-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-297"><a href="#cb76-297" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2. Create DESeqDataSet object</span></span>
<span id="cb76-298"><a href="#cb76-298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-299"><a href="#cb76-299" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load the DESeq2 package:</span>
<span id="cb76-300"><a href="#cb76-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-303"><a href="#cb76-303" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-304"><a href="#cb76-304" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-305"><a href="#cb76-305" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-306"><a href="#cb76-306" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb76-307"><a href="#cb76-307" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-308"><a href="#cb76-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-309"><a href="#cb76-309" aria-hidden="true" tabindex="-1"></a>A DEseqDataSet object is a custom data type that is used by</span>
<span id="cb76-310"><a href="#cb76-310" aria-hidden="true" tabindex="-1"></a>**`DESeq2`**. Custom data types are common in the Bioconductor<span class="ot">[^1]</span></span>
<span id="cb76-311"><a href="#cb76-311" aria-hidden="true" tabindex="-1"></a>packages. They are used to store data in a way that is useful for the</span>
<span id="cb76-312"><a href="#cb76-312" aria-hidden="true" tabindex="-1"></a>analysis. These data types typically have data, transformed data,</span>
<span id="cb76-313"><a href="#cb76-313" aria-hidden="true" tabindex="-1"></a>metadata and experimental designs within them.</span>
<span id="cb76-314"><a href="#cb76-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-315"><a href="#cb76-315" aria-hidden="true" tabindex="-1"></a><span class="ot">[^1]: </span><span class="co">[</span><span class="ot">Bioconductor</span><span class="co">](https://www.bioconductor.org/)</span> is a project that</span>
<span id="cb76-316"><a href="#cb76-316" aria-hidden="true" tabindex="-1"></a>    develops and supports R packages for bioinformatics.</span>
<span id="cb76-317"><a href="#cb76-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-318"><a href="#cb76-318" aria-hidden="true" tabindex="-1"></a>To create a DESeqDataSet object, we need to provide three things:</span>
<span id="cb76-319"><a href="#cb76-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-320"><a href="#cb76-320" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The raw counts - these are in <span class="in">`root_filtered`</span></span>
<span id="cb76-321"><a href="#cb76-321" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The meta data which gives information about the samples and which</span>
<span id="cb76-322"><a href="#cb76-322" aria-hidden="true" tabindex="-1"></a>    treatment groups they belong to</span>
<span id="cb76-323"><a href="#cb76-323" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>A design matrix which captures the design of the statistical model.</span>
<span id="cb76-324"><a href="#cb76-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-325"><a href="#cb76-325" aria-hidden="true" tabindex="-1"></a>The counts must in a *matrix* rather than a dataframe. Unlike a</span>
<span id="cb76-326"><a href="#cb76-326" aria-hidden="true" tabindex="-1"></a>dataframe, a matrix has columns of all the same type. That is, it will</span>
<span id="cb76-327"><a href="#cb76-327" aria-hidden="true" tabindex="-1"></a>contain only the counts. The gene ids are given as row names rather than</span>
<span id="cb76-328"><a href="#cb76-328" aria-hidden="true" tabindex="-1"></a>a column. The <span class="in">`matrix()`</span> function will create a matrix from a dataframe</span>
<span id="cb76-329"><a href="#cb76-329" aria-hidden="true" tabindex="-1"></a>of columns of the same type and the <span class="in">`select()`</span> function can be used to</span>
<span id="cb76-330"><a href="#cb76-330" aria-hidden="true" tabindex="-1"></a>remove the gene ids column.</span>
<span id="cb76-331"><a href="#cb76-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-332"><a href="#cb76-332" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a matrix of the counts:</span>
<span id="cb76-333"><a href="#cb76-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-336"><a href="#cb76-336" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-337"><a href="#cb76-337" aria-hidden="true" tabindex="-1"></a>root_count_mat <span class="ot">&lt;-</span> root_filtered <span class="sc">|&gt;</span></span>
<span id="cb76-338"><a href="#cb76-338" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>gene_id, <span class="sc">-</span>gene_name) <span class="sc">|&gt;</span></span>
<span id="cb76-339"><a href="#cb76-339" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span>
<span id="cb76-340"><a href="#cb76-340" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-341"><a href="#cb76-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-342"><a href="#cb76-342" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add the gene ids as row names to the matrix:</span>
<span id="cb76-343"><a href="#cb76-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-346"><a href="#cb76-346" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-347"><a href="#cb76-347" aria-hidden="true" tabindex="-1"></a><span class="co"># add the row names to the matrix</span></span>
<span id="cb76-348"><a href="#cb76-348" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(root_count_mat) <span class="ot">&lt;-</span> root_filtered<span class="sc">$</span>gene_id</span>
<span id="cb76-349"><a href="#cb76-349" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-350"><a href="#cb76-350" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-351"><a href="#cb76-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-352"><a href="#cb76-352" aria-hidden="true" tabindex="-1"></a>You might want to view the matrix (click on it in your environment</span>
<span id="cb76-353"><a href="#cb76-353" aria-hidden="true" tabindex="-1"></a>pane).</span>
<span id="cb76-354"><a href="#cb76-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-355"><a href="#cb76-355" aria-hidden="true" tabindex="-1"></a>The metadata are in a file,</span>
<span id="cb76-356"><a href="#cb76-356" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">arab_meta_data.txt</span><span class="co">](meta/arab_meta_data.txt)</span>. This is a tab-delimited</span>
<span id="cb76-357"><a href="#cb76-357" aria-hidden="true" tabindex="-1"></a>file. The first column is the sample name and the other columns give the</span>
<span id="cb76-358"><a href="#cb76-358" aria-hidden="true" tabindex="-1"></a>"treatments". In this case, the treatments genotype (with two levels)</span>
<span id="cb76-359"><a href="#cb76-359" aria-hidden="true" tabindex="-1"></a>and copper (with two levels).</span>
<span id="cb76-360"><a href="#cb76-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-361"><a href="#cb76-361" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Make a folder called <span class="in">`meta`</span> and save the file to it.</span>
<span id="cb76-362"><a href="#cb76-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-363"><a href="#cb76-363" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Read the metadata into a dataframe:</span>
<span id="cb76-364"><a href="#cb76-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-367"><a href="#cb76-367" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-368"><a href="#cb76-368" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read_table</span>(<span class="st">"meta/arab_meta_data.txt"</span>)</span>
<span id="cb76-369"><a href="#cb76-369" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-370"><a href="#cb76-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-371"><a href="#cb76-371" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Examine the resulting dataframe.</span>
<span id="cb76-372"><a href="#cb76-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-373"><a href="#cb76-373" aria-hidden="true" tabindex="-1"></a>We need to add the sample names as row names to the metadata dataframe.</span>
<span id="cb76-374"><a href="#cb76-374" aria-hidden="true" tabindex="-1"></a>This is because the DESeqDataSet object will use the row names to match</span>
<span id="cb76-375"><a href="#cb76-375" aria-hidden="true" tabindex="-1"></a>the samples in the metadata to the samples in the counts matrix.</span>
<span id="cb76-376"><a href="#cb76-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-377"><a href="#cb76-377" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add the sample names as row names to the metadata dataframe:</span>
<span id="cb76-378"><a href="#cb76-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-381"><a href="#cb76-381" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-382"><a href="#cb76-382" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> meta <span class="sc">|&gt;</span></span>
<span id="cb76-383"><a href="#cb76-383" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"sample_id"</span>)</span>
<span id="cb76-384"><a href="#cb76-384" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-385"><a href="#cb76-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-386"><a href="#cb76-386" aria-hidden="true" tabindex="-1"></a>We are dealing only with the root data so we need to remove the samples</span>
<span id="cb76-387"><a href="#cb76-387" aria-hidden="true" tabindex="-1"></a>that are not in the root data.</span>
<span id="cb76-388"><a href="#cb76-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-389"><a href="#cb76-389" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Filter the metadata to keep only the root information:</span>
<span id="cb76-390"><a href="#cb76-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-393"><a href="#cb76-393" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-394"><a href="#cb76-394" aria-hidden="true" tabindex="-1"></a>meta_root <span class="ot">&lt;-</span> meta <span class="sc">|&gt;</span></span>
<span id="cb76-395"><a href="#cb76-395" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(tissue <span class="sc">==</span> <span class="st">"root"</span>)</span>
<span id="cb76-396"><a href="#cb76-396" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-397"><a href="#cb76-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-398"><a href="#cb76-398" aria-hidden="true" tabindex="-1"></a>We can now create the DESeqDataSet object. The design formula describes</span>
<span id="cb76-399"><a href="#cb76-399" aria-hidden="true" tabindex="-1"></a>the statistical model. You should recognise the form from previous work.</span>
<span id="cb76-400"><a href="#cb76-400" aria-hidden="true" tabindex="-1"></a>The <span class="in">`~`</span> can be read as "explain by" and on its right hand side are the</span>
<span id="cb76-401"><a href="#cb76-401" aria-hidden="true" tabindex="-1"></a>explanatory variables. That is, the model is counts explained by nickel</span>
<span id="cb76-402"><a href="#cb76-402" aria-hidden="true" tabindex="-1"></a>status.</span>
<span id="cb76-403"><a href="#cb76-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-404"><a href="#cb76-404" aria-hidden="true" tabindex="-1"></a>Note that:</span>
<span id="cb76-405"><a href="#cb76-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-406"><a href="#cb76-406" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The names of the columns in the count matrix have to exactly match</span>
<span id="cb76-407"><a href="#cb76-407" aria-hidden="true" tabindex="-1"></a>    the names of the rows in the metadata dataframe. They also need to</span>
<span id="cb76-408"><a href="#cb76-408" aria-hidden="true" tabindex="-1"></a>    be in the same order.</span>
<span id="cb76-409"><a href="#cb76-409" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The names of the explanatory variables in the design formula have to</span>
<span id="cb76-410"><a href="#cb76-410" aria-hidden="true" tabindex="-1"></a>    match the names of columns in the metadata.</span>
<span id="cb76-411"><a href="#cb76-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-412"><a href="#cb76-412" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create the DESeqDataSet object:</span>
<span id="cb76-413"><a href="#cb76-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-416"><a href="#cb76-416" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-417"><a href="#cb76-417" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(root_count_mat,</span>
<span id="cb76-418"><a href="#cb76-418" aria-hidden="true" tabindex="-1"></a>                              <span class="at">colData =</span> meta_root,</span>
<span id="cb76-419"><a href="#cb76-419" aria-hidden="true" tabindex="-1"></a>                              <span class="at">design =</span> <span class="sc">~</span> nickel)</span>
<span id="cb76-420"><a href="#cb76-420" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-421"><a href="#cb76-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-422"><a href="#cb76-422" aria-hidden="true" tabindex="-1"></a>The warning "Warning: some variables in design formula are characters,</span>
<span id="cb76-423"><a href="#cb76-423" aria-hidden="true" tabindex="-1"></a>converting to factors" just means that the variable type of nickel in</span>
<span id="cb76-424"><a href="#cb76-424" aria-hidden="true" tabindex="-1"></a>the metadata dataframe is "char" and it has been converted into a factor</span>
<span id="cb76-425"><a href="#cb76-425" aria-hidden="true" tabindex="-1"></a>type.</span>
<span id="cb76-426"><a href="#cb76-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-427"><a href="#cb76-427" aria-hidden="true" tabindex="-1"></a>To help you understand what the <span class="in">`DESeqDataSet`</span> object we have called</span>
<span id="cb76-428"><a href="#cb76-428" aria-hidden="true" tabindex="-1"></a><span class="in">`dds`</span> contains, we can look its contents</span>
<span id="cb76-429"><a href="#cb76-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-430"><a href="#cb76-430" aria-hidden="true" tabindex="-1"></a>The counts are in <span class="in">`dds@assays@data@listData[["counts"]]`</span> and the</span>
<span id="cb76-431"><a href="#cb76-431" aria-hidden="true" tabindex="-1"></a>metadata are in <span class="in">`dds@colData`</span> but the easiest way to see them is to use</span>
<span id="cb76-432"><a href="#cb76-432" aria-hidden="true" tabindex="-1"></a>the <span class="in">`counts()`</span> and <span class="in">`colData()`</span> functions from the **`DESeq2`** package.</span>
<span id="cb76-433"><a href="#cb76-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-434"><a href="#cb76-434" aria-hidden="true" tabindex="-1"></a>ğŸ¬ View the counts:</span>
<span id="cb76-435"><a href="#cb76-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-438"><a href="#cb76-438" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-439"><a href="#cb76-439" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb76-440"><a href="#cb76-440" aria-hidden="true" tabindex="-1"></a><span class="fu">counts</span>(dds) <span class="sc">|&gt;</span> <span class="fu">View</span>()</span>
<span id="cb76-441"><a href="#cb76-441" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-442"><a href="#cb76-442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-443"><a href="#cb76-443" aria-hidden="true" tabindex="-1"></a>You should be able to see that this is the same as in <span class="in">`root_count_mat`</span>.</span>
<span id="cb76-444"><a href="#cb76-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-445"><a href="#cb76-445" aria-hidden="true" tabindex="-1"></a>ğŸ¬ View the column information:</span>
<span id="cb76-446"><a href="#cb76-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-449"><a href="#cb76-449" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-450"><a href="#cb76-450" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(dds)</span>
<span id="cb76-451"><a href="#cb76-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-452"><a href="#cb76-452" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-453"><a href="#cb76-453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-454"><a href="#cb76-454" aria-hidden="true" tabindex="-1"></a>You should be able to see this is the same as in <span class="in">`meta_root`</span>.</span>
<span id="cb76-455"><a href="#cb76-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-456"><a href="#cb76-456" aria-hidden="true" tabindex="-1"></a><span class="fu">### 3. Prepare the normalised counts</span></span>
<span id="cb76-457"><a href="#cb76-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-458"><a href="#cb76-458" aria-hidden="true" tabindex="-1"></a>The normalised counts are the counts that have been transformed to</span>
<span id="cb76-459"><a href="#cb76-459" aria-hidden="true" tabindex="-1"></a>account for the library size (i.e., the total number of reads in a</span>
<span id="cb76-460"><a href="#cb76-460" aria-hidden="true" tabindex="-1"></a>sample) and the gene length. We have to first estimate the normalisation</span>
<span id="cb76-461"><a href="#cb76-461" aria-hidden="true" tabindex="-1"></a>factors and store them in the DESeqDataSet object and then we can get</span>
<span id="cb76-462"><a href="#cb76-462" aria-hidden="true" tabindex="-1"></a>the normalised counts.</span>
<span id="cb76-463"><a href="#cb76-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-464"><a href="#cb76-464" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Estimate the factors for normalisation and store them in the</span>
<span id="cb76-465"><a href="#cb76-465" aria-hidden="true" tabindex="-1"></a>DESeqDataSet object:</span>
<span id="cb76-466"><a href="#cb76-466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-469"><a href="#cb76-469" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-470"><a href="#cb76-470" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(dds)</span>
<span id="cb76-471"><a href="#cb76-471" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-472"><a href="#cb76-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-473"><a href="#cb76-473" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Look at the factors (just for information):</span>
<span id="cb76-474"><a href="#cb76-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-477"><a href="#cb76-477" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-478"><a href="#cb76-478" aria-hidden="true" tabindex="-1"></a><span class="fu">sizeFactors</span>(dds)</span>
<span id="cb76-479"><a href="#cb76-479" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-480"><a href="#cb76-480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-481"><a href="#cb76-481" aria-hidden="true" tabindex="-1"></a>The normalised counts will be useful to use later. To get the normalised</span>
<span id="cb76-482"><a href="#cb76-482" aria-hidden="true" tabindex="-1"></a>counts we again used the <span class="in">`counts()`</span> function but this time we use the</span>
<span id="cb76-483"><a href="#cb76-483" aria-hidden="true" tabindex="-1"></a><span class="in">`normalized=TRUE`</span> argument.</span>
<span id="cb76-484"><a href="#cb76-484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-485"><a href="#cb76-485" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Save the normalised to a matrix:</span>
<span id="cb76-486"><a href="#cb76-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-489"><a href="#cb76-489" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-490"><a href="#cb76-490" aria-hidden="true" tabindex="-1"></a>normalised_counts <span class="ot">&lt;-</span> <span class="fu">counts</span>(dds, <span class="at">normalized =</span> <span class="cn">TRUE</span>)</span>
<span id="cb76-491"><a href="#cb76-491" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-492"><a href="#cb76-492" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-493"><a href="#cb76-493" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Make a dataframe of the normalised counts, adding a column for the</span>
<span id="cb76-494"><a href="#cb76-494" aria-hidden="true" tabindex="-1"></a>gene ids at the same time:</span>
<span id="cb76-495"><a href="#cb76-495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-498"><a href="#cb76-498" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-499"><a href="#cb76-499" aria-hidden="true" tabindex="-1"></a>root_normalised_counts <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(normalised_counts,</span>
<span id="cb76-500"><a href="#cb76-500" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">gene_id =</span> <span class="fu">row.names</span>(normalised_counts))</span>
<span id="cb76-501"><a href="#cb76-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-502"><a href="#cb76-502" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-503"><a href="#cb76-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-504"><a href="#cb76-504" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4. Differential expression analysis</span></span>
<span id="cb76-505"><a href="#cb76-505" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-506"><a href="#cb76-506" aria-hidden="true" tabindex="-1"></a>We use the <span class="in">`DESeq()`</span> function to do the differential expression</span>
<span id="cb76-507"><a href="#cb76-507" aria-hidden="true" tabindex="-1"></a>analysis. This function fits the statistical model to the data and then</span>
<span id="cb76-508"><a href="#cb76-508" aria-hidden="true" tabindex="-1"></a>uses the model to calculate the significance of the difference between</span>
<span id="cb76-509"><a href="#cb76-509" aria-hidden="true" tabindex="-1"></a>the treatments. It again stores the results in the DESseqDataSet object.</span>
<span id="cb76-510"><a href="#cb76-510" aria-hidden="true" tabindex="-1"></a>Note that the differential expression needs the raw (unnormalised</span>
<span id="cb76-511"><a href="#cb76-511" aria-hidden="true" tabindex="-1"></a>counts) as it does its own normalisation as part of the process.</span>
<span id="cb76-512"><a href="#cb76-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-513"><a href="#cb76-513" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Run the differential expression analysis and store the results in the</span>
<span id="cb76-514"><a href="#cb76-514" aria-hidden="true" tabindex="-1"></a>same object:</span>
<span id="cb76-515"><a href="#cb76-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-518"><a href="#cb76-518" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-519"><a href="#cb76-519" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb76-520"><a href="#cb76-520" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-521"><a href="#cb76-521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-522"><a href="#cb76-522" aria-hidden="true" tabindex="-1"></a>The function will take only a few moments to run on this data but can</span>
<span id="cb76-523"><a href="#cb76-523" aria-hidden="true" tabindex="-1"></a>take longer for bigger datasets.</span>
<span id="cb76-524"><a href="#cb76-524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-525"><a href="#cb76-525" aria-hidden="true" tabindex="-1"></a>We need to define the contrasts we want to test. We want to test the</span>
<span id="cb76-526"><a href="#cb76-526" aria-hidden="true" tabindex="-1"></a>difference between the treatments so we will define the contrast as</span>
<span id="cb76-527"><a href="#cb76-527" aria-hidden="true" tabindex="-1"></a><span class="in">`control`</span> and <span class="in">`low_ni`</span>.</span>
<span id="cb76-528"><a href="#cb76-528" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-529"><a href="#cb76-529" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Define the contrast:</span>
<span id="cb76-530"><a href="#cb76-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-533"><a href="#cb76-533" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-534"><a href="#cb76-534" aria-hidden="true" tabindex="-1"></a>contrast_suf <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"nickel"</span>, <span class="st">"control"</span>, <span class="st">"low_ni"</span>)</span>
<span id="cb76-535"><a href="#cb76-535" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-536"><a href="#cb76-536" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-537"><a href="#cb76-537" aria-hidden="true" tabindex="-1"></a>Note that <span class="in">`nickel`</span> is the name of the column in the metadata dataframe</span>
<span id="cb76-538"><a href="#cb76-538" aria-hidden="true" tabindex="-1"></a>and <span class="in">`control`</span> and <span class="in">`low_ni`</span> are the names of the levels in the <span class="in">`nickel`</span></span>
<span id="cb76-539"><a href="#cb76-539" aria-hidden="true" tabindex="-1"></a>column. By putting them in the order <span class="in">`control`</span> , <span class="in">`low_ni`</span> we are saying</span>
<span id="cb76-540"><a href="#cb76-540" aria-hidden="true" tabindex="-1"></a>the fold change will be control / low_ni. This means:</span>
<span id="cb76-541"><a href="#cb76-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-542"><a href="#cb76-542" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>positive log fold changes indicate control <span class="sc">\&gt;</span> low_ni and</span>
<span id="cb76-543"><a href="#cb76-543" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>negative log fold changes indicates low_ni <span class="sc">\&gt;</span> control.</span>
<span id="cb76-544"><a href="#cb76-544" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-545"><a href="#cb76-545" aria-hidden="true" tabindex="-1"></a>If we had put them in the order <span class="in">`low_ni`</span>, <span class="in">`control`</span> we would have the</span>
<span id="cb76-546"><a href="#cb76-546" aria-hidden="true" tabindex="-1"></a>reverse.</span>
<span id="cb76-547"><a href="#cb76-547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-548"><a href="#cb76-548" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Extract the results from the DESseqDataSet object:</span>
<span id="cb76-549"><a href="#cb76-549" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-552"><a href="#cb76-552" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-553"><a href="#cb76-553" aria-hidden="true" tabindex="-1"></a>results_suf <span class="ot">&lt;-</span> <span class="fu">results</span>(dds,</span>
<span id="cb76-554"><a href="#cb76-554" aria-hidden="true" tabindex="-1"></a>                       <span class="at">contrast =</span> contrast_suf)</span>
<span id="cb76-555"><a href="#cb76-555" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-556"><a href="#cb76-556" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-557"><a href="#cb76-557" aria-hidden="true" tabindex="-1"></a>This will give us the log~2~ fold change, the *p*-value and the adjusted</span>
<span id="cb76-558"><a href="#cb76-558" aria-hidden="true" tabindex="-1"></a>*p*-value for the comparison between the control and low_ni for each</span>
<span id="cb76-559"><a href="#cb76-559" aria-hidden="true" tabindex="-1"></a>gene.</span>
<span id="cb76-560"><a href="#cb76-560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-561"><a href="#cb76-561" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Put the results in a dataframe and add the gene ids as a column:</span>
<span id="cb76-562"><a href="#cb76-562" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-565"><a href="#cb76-565" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-566"><a href="#cb76-566" aria-hidden="true" tabindex="-1"></a>root_results <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(results_suf,</span>
<span id="cb76-567"><a href="#cb76-567" aria-hidden="true" tabindex="-1"></a>                          <span class="at">gene_id =</span> <span class="fu">row.names</span>(results_suf))</span>
<span id="cb76-568"><a href="#cb76-568" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-569"><a href="#cb76-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-570"><a href="#cb76-570" aria-hidden="true" tabindex="-1"></a>It is useful to have the normalised counts and the statistical results</span>
<span id="cb76-571"><a href="#cb76-571" aria-hidden="true" tabindex="-1"></a>in one dataframe.</span>
<span id="cb76-572"><a href="#cb76-572" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-573"><a href="#cb76-573" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Merge the two dataframes:</span>
<span id="cb76-574"><a href="#cb76-574" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-577"><a href="#cb76-577" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-578"><a href="#cb76-578" aria-hidden="true" tabindex="-1"></a><span class="co"># merge the results with the normalised counts</span></span>
<span id="cb76-579"><a href="#cb76-579" aria-hidden="true" tabindex="-1"></a>root_results <span class="ot">&lt;-</span> root_normalised_counts <span class="sc">|&gt;</span></span>
<span id="cb76-580"><a href="#cb76-580" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(root_results, <span class="at">by =</span> <span class="st">"gene_id"</span>)</span>
<span id="cb76-581"><a href="#cb76-581" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-582"><a href="#cb76-582" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-583"><a href="#cb76-583" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Add gene information</span><span class="co">](#arabidopisis-add-info)</span>.</span>
<span id="cb76-584"><a href="#cb76-584" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-585"><a href="#cb76-585" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ’‰ *Leishmania* {#leishmania-differential}</span></span>
<span id="cb76-586"><a href="#cb76-586" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-587"><a href="#cb76-587" aria-hidden="true" tabindex="-1"></a>These are the steps we will take</span>
<span id="cb76-588"><a href="#cb76-588" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-589"><a href="#cb76-589" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Find the genes that are expressed in only one treatment group.</span>
<span id="cb76-590"><a href="#cb76-590" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Create a DESeqDataSet object. This is a special object that is used</span>
<span id="cb76-591"><a href="#cb76-591" aria-hidden="true" tabindex="-1"></a>    by the **`DESeq2`** package</span>
<span id="cb76-592"><a href="#cb76-592" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Prepare the normalised counts from the DESeqDataSet object.</span>
<span id="cb76-593"><a href="#cb76-593" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Do differential expression analysis on the genes. This needs to be</span>
<span id="cb76-594"><a href="#cb76-594" aria-hidden="true" tabindex="-1"></a>    done on the raw counts.</span>
<span id="cb76-595"><a href="#cb76-595" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-596"><a href="#cb76-596" aria-hidden="true" tabindex="-1"></a>All but the first step are done with the **`DESeq2`** package</span>
<span id="cb76-597"><a href="#cb76-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-598"><a href="#cb76-598" aria-hidden="true" tabindex="-1"></a><span class="fu">### 1. Genes expressed in one treatment</span></span>
<span id="cb76-599"><a href="#cb76-599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-600"><a href="#cb76-600" aria-hidden="true" tabindex="-1"></a>The genes expressed in only one treatment group are those with zeros in</span>
<span id="cb76-601"><a href="#cb76-601" aria-hidden="true" tabindex="-1"></a>all replicates in one group and non-zero values in all replicates in the</span>
<span id="cb76-602"><a href="#cb76-602" aria-hidden="true" tabindex="-1"></a>other group.</span>
<span id="cb76-603"><a href="#cb76-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-604"><a href="#cb76-604" aria-hidden="true" tabindex="-1"></a>We will use <span class="in">`filter()`</span> to find these genes.</span>
<span id="cb76-605"><a href="#cb76-605" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-606"><a href="#cb76-606" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Find the genes that are expressed only at the procyclic-promastigote</span>
<span id="cb76-607"><a href="#cb76-607" aria-hidden="true" tabindex="-1"></a>stage:</span>
<span id="cb76-608"><a href="#cb76-608" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-611"><a href="#cb76-611" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-612"><a href="#cb76-612" aria-hidden="true" tabindex="-1"></a>pro_meta_pro_only <span class="ot">&lt;-</span> pro_meta_filtered  <span class="sc">|&gt;</span></span>
<span id="cb76-613"><a href="#cb76-613" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(lm_pro_1 <span class="sc">&gt;</span> <span class="dv">0</span>,</span>
<span id="cb76-614"><a href="#cb76-614" aria-hidden="true" tabindex="-1"></a>         lm_pro_2 <span class="sc">&gt;</span> <span class="dv">0</span>,</span>
<span id="cb76-615"><a href="#cb76-615" aria-hidden="true" tabindex="-1"></a>         lm_pro_3 <span class="sc">&gt;</span> <span class="dv">0</span>,</span>
<span id="cb76-616"><a href="#cb76-616" aria-hidden="true" tabindex="-1"></a>         lm_meta_1 <span class="sc">==</span> <span class="dv">0</span>,</span>
<span id="cb76-617"><a href="#cb76-617" aria-hidden="true" tabindex="-1"></a>         lm_meta_2 <span class="sc">==</span> <span class="dv">0</span>,</span>
<span id="cb76-618"><a href="#cb76-618" aria-hidden="true" tabindex="-1"></a>         lm_meta_2 <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb76-619"><a href="#cb76-619" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-620"><a href="#cb76-620" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-621"><a href="#cb76-621" aria-hidden="true" tabindex="-1"></a>â“ How many genes are expressed only in the procyclic-promastigote stage</span>
<span id="cb76-622"><a href="#cb76-622" aria-hidden="true" tabindex="-1"></a>group?</span>
<span id="cb76-623"><a href="#cb76-623" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-624"><a href="#cb76-624" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb76-625"><a href="#cb76-625" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-626"><a href="#cb76-626" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- There are 0 genes expressed only in the procyclic-promastigote stage --&gt;</span></span>
<span id="cb76-627"><a href="#cb76-627" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-628"><a href="#cb76-628" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Now you find any genes that are expressed only at the metacyclic</span>
<span id="cb76-629"><a href="#cb76-629" aria-hidden="true" tabindex="-1"></a>stage</span>
<span id="cb76-630"><a href="#cb76-630" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-633"><a href="#cb76-633" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-634"><a href="#cb76-634" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-635"><a href="#cb76-635" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-636"><a href="#cb76-636" aria-hidden="true" tabindex="-1"></a>pro_meta_meta_only <span class="ot">&lt;-</span> pro_meta_filtered  <span class="sc">|&gt;</span></span>
<span id="cb76-637"><a href="#cb76-637" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(lm_pro_1 <span class="sc">==</span> <span class="dv">0</span>,</span>
<span id="cb76-638"><a href="#cb76-638" aria-hidden="true" tabindex="-1"></a>         lm_pro_2 <span class="sc">==</span> <span class="dv">0</span>,</span>
<span id="cb76-639"><a href="#cb76-639" aria-hidden="true" tabindex="-1"></a>         lm_pro_3 <span class="sc">==</span> <span class="dv">0</span>,</span>
<span id="cb76-640"><a href="#cb76-640" aria-hidden="true" tabindex="-1"></a>         lm_meta_1 <span class="sc">&gt;</span> <span class="dv">0</span>,</span>
<span id="cb76-641"><a href="#cb76-641" aria-hidden="true" tabindex="-1"></a>         lm_meta_2 <span class="sc">&gt;</span> <span class="dv">0</span>,</span>
<span id="cb76-642"><a href="#cb76-642" aria-hidden="true" tabindex="-1"></a>         lm_meta_2 <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb76-643"><a href="#cb76-643" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-644"><a href="#cb76-644" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-645"><a href="#cb76-645" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-646"><a href="#cb76-646" aria-hidden="true" tabindex="-1"></a>â“ How many genes are expressed only at the metacyclic stage?</span>
<span id="cb76-647"><a href="#cb76-647" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-648"><a href="#cb76-648" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb76-649"><a href="#cb76-649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-650"><a href="#cb76-650" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- There are 0 genes expressed only in the meta stage group. --&gt;</span></span>
<span id="cb76-651"><a href="#cb76-651" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-652"><a href="#cb76-652" aria-hidden="true" tabindex="-1"></a>â“ Do the results make sense to you in light of what you know about the</span>
<span id="cb76-653"><a href="#cb76-653" aria-hidden="true" tabindex="-1"></a>biology?</span>
<span id="cb76-654"><a href="#cb76-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-655"><a href="#cb76-655" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb76-656"><a href="#cb76-656" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-657"><a href="#cb76-657" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- I have no clue here! --&gt;</span></span>
<span id="cb76-658"><a href="#cb76-658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-659"><a href="#cb76-659" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- Domain expertise is important when analysing data so  --&gt;</span></span>
<span id="cb76-660"><a href="#cb76-660" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-661"><a href="#cb76-661" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- discuss your results with Pegine and your group. --&gt;</span></span>
<span id="cb76-662"><a href="#cb76-662" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-663"><a href="#cb76-663" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Write all the genes that are expressed one group only to file (saved</span>
<span id="cb76-664"><a href="#cb76-664" aria-hidden="true" tabindex="-1"></a>in <span class="in">`results`</span>)</span>
<span id="cb76-665"><a href="#cb76-665" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-668"><a href="#cb76-668" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-669"><a href="#cb76-669" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-670"><a href="#cb76-670" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb76-671"><a href="#cb76-671" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-672"><a href="#cb76-672" aria-hidden="true" tabindex="-1"></a><span class="co"># If we had any genes expressed in only one group we would write </span></span>
<span id="cb76-673"><a href="#cb76-673" aria-hidden="true" tabindex="-1"></a><span class="co"># them to a file like this. </span></span>
<span id="cb76-674"><a href="#cb76-674" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(pro_meta_pro_only, <span class="st">"results/pro_meta_pro_only.csv"</span>)</span>
<span id="cb76-675"><a href="#cb76-675" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(pro_meta_meta_only, <span class="st">"results/pro_meta_meta_only.csv"</span>)</span>
<span id="cb76-676"><a href="#cb76-676" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-677"><a href="#cb76-677" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-678"><a href="#cb76-678" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2. Create DESeqDataSet object</span></span>
<span id="cb76-679"><a href="#cb76-679" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-680"><a href="#cb76-680" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load the DESeq2 package:</span>
<span id="cb76-681"><a href="#cb76-681" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-684"><a href="#cb76-684" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-685"><a href="#cb76-685" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-686"><a href="#cb76-686" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb76-687"><a href="#cb76-687" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-688"><a href="#cb76-688" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb76-689"><a href="#cb76-689" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-690"><a href="#cb76-690" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-691"><a href="#cb76-691" aria-hidden="true" tabindex="-1"></a>A DEseqDataSet object is a custom data type that is used by</span>
<span id="cb76-692"><a href="#cb76-692" aria-hidden="true" tabindex="-1"></a>**`DESeq2`**. Custom data types are common in the Bioconductor<span class="ot">[^2]</span></span>
<span id="cb76-693"><a href="#cb76-693" aria-hidden="true" tabindex="-1"></a>packages. They are used to store data in a way that is useful for the</span>
<span id="cb76-694"><a href="#cb76-694" aria-hidden="true" tabindex="-1"></a>analysis. These data types typically have data, transformed data,</span>
<span id="cb76-695"><a href="#cb76-695" aria-hidden="true" tabindex="-1"></a>metadata and experimental designs within them.</span>
<span id="cb76-696"><a href="#cb76-696" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-697"><a href="#cb76-697" aria-hidden="true" tabindex="-1"></a><span class="ot">[^2]: </span><span class="co">[</span><span class="ot">Bioconductor</span><span class="co">](https://www.bioconductor.org/)</span> is a project that</span>
<span id="cb76-698"><a href="#cb76-698" aria-hidden="true" tabindex="-1"></a>    develops and supports R packages for bioinformatics.</span>
<span id="cb76-699"><a href="#cb76-699" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-700"><a href="#cb76-700" aria-hidden="true" tabindex="-1"></a>To create a DESeqDataSet object, we need to provide three things:</span>
<span id="cb76-701"><a href="#cb76-701" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-702"><a href="#cb76-702" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The raw counts - these are in <span class="in">`pro_meta_filtered`</span></span>
<span id="cb76-703"><a href="#cb76-703" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The meta data which gives information about the samples and which</span>
<span id="cb76-704"><a href="#cb76-704" aria-hidden="true" tabindex="-1"></a>    treatment groups they belong to</span>
<span id="cb76-705"><a href="#cb76-705" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>A design matrix which captures the design of the statistical model.</span>
<span id="cb76-706"><a href="#cb76-706" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-707"><a href="#cb76-707" aria-hidden="true" tabindex="-1"></a>The counts must in a *matrix* rather than a dataframe. Unlike a</span>
<span id="cb76-708"><a href="#cb76-708" aria-hidden="true" tabindex="-1"></a>dataframe, a matrix has columns of all the same type. That is, it will</span>
<span id="cb76-709"><a href="#cb76-709" aria-hidden="true" tabindex="-1"></a>contain only the counts. The gene ids are given as row names rather than</span>
<span id="cb76-710"><a href="#cb76-710" aria-hidden="true" tabindex="-1"></a>a column. The <span class="in">`matrix()`</span> function will create a matrix from a dataframe</span>
<span id="cb76-711"><a href="#cb76-711" aria-hidden="true" tabindex="-1"></a>of columns of the same type and the <span class="in">`select()`</span> function can be used to</span>
<span id="cb76-712"><a href="#cb76-712" aria-hidden="true" tabindex="-1"></a>remove the gene ids column.</span>
<span id="cb76-713"><a href="#cb76-713" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-714"><a href="#cb76-714" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a matrix of the counts:</span>
<span id="cb76-715"><a href="#cb76-715" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-718"><a href="#cb76-718" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-719"><a href="#cb76-719" aria-hidden="true" tabindex="-1"></a>pro_meta_count_mat <span class="ot">&lt;-</span> pro_meta_filtered  <span class="sc">|&gt;</span></span>
<span id="cb76-720"><a href="#cb76-720" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>gene_id) <span class="sc">|&gt;</span></span>
<span id="cb76-721"><a href="#cb76-721" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span>
<span id="cb76-722"><a href="#cb76-722" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-723"><a href="#cb76-723" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-724"><a href="#cb76-724" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add the gene ids as row names to the matrix:</span>
<span id="cb76-725"><a href="#cb76-725" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-728"><a href="#cb76-728" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-729"><a href="#cb76-729" aria-hidden="true" tabindex="-1"></a><span class="co"># add the row names to the matrix</span></span>
<span id="cb76-730"><a href="#cb76-730" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pro_meta_count_mat) <span class="ot">&lt;-</span> pro_meta_filtered<span class="sc">$</span>gene_id</span>
<span id="cb76-731"><a href="#cb76-731" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-732"><a href="#cb76-732" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-733"><a href="#cb76-733" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-734"><a href="#cb76-734" aria-hidden="true" tabindex="-1"></a>You might want to view the matrix (click on it in your environment</span>
<span id="cb76-735"><a href="#cb76-735" aria-hidden="true" tabindex="-1"></a>pane).</span>
<span id="cb76-736"><a href="#cb76-736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-737"><a href="#cb76-737" aria-hidden="true" tabindex="-1"></a>The metadata are in a file,</span>
<span id="cb76-738"><a href="#cb76-738" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">leish_meta_data.txt</span><span class="co">](meta/leish_meta_data.txt)</span>. This is a tab-delimited</span>
<span id="cb76-739"><a href="#cb76-739" aria-hidden="true" tabindex="-1"></a>file. The first column is the sample name and the other columns give the</span>
<span id="cb76-740"><a href="#cb76-740" aria-hidden="true" tabindex="-1"></a>"treatments". In this case, the treatment is stage (with three levels).</span>
<span id="cb76-741"><a href="#cb76-741" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-742"><a href="#cb76-742" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Make a folder called <span class="in">`meta`</span> and save the file to it.</span>
<span id="cb76-743"><a href="#cb76-743" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-744"><a href="#cb76-744" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Read the metadata into a dataframe:</span>
<span id="cb76-745"><a href="#cb76-745" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-748"><a href="#cb76-748" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-749"><a href="#cb76-749" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read_table</span>(<span class="st">"meta/leish_meta_data.txt"</span>)</span>
<span id="cb76-750"><a href="#cb76-750" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-751"><a href="#cb76-751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-752"><a href="#cb76-752" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Examine the resulting dataframe.</span>
<span id="cb76-753"><a href="#cb76-753" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-754"><a href="#cb76-754" aria-hidden="true" tabindex="-1"></a>We need to add the sample names as row names to the metadata dataframe.</span>
<span id="cb76-755"><a href="#cb76-755" aria-hidden="true" tabindex="-1"></a>This is because the DESeqDataSet object will use the row names to match</span>
<span id="cb76-756"><a href="#cb76-756" aria-hidden="true" tabindex="-1"></a>the samples in the metadata to the samples in the counts matrix.</span>
<span id="cb76-757"><a href="#cb76-757" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-758"><a href="#cb76-758" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add the sample names as row names to the metadata dataframe:</span>
<span id="cb76-759"><a href="#cb76-759" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-762"><a href="#cb76-762" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-763"><a href="#cb76-763" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> meta <span class="sc">|&gt;</span></span>
<span id="cb76-764"><a href="#cb76-764" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"sample_id"</span>)</span>
<span id="cb76-765"><a href="#cb76-765" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-766"><a href="#cb76-766" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-767"><a href="#cb76-767" aria-hidden="true" tabindex="-1"></a>We are dealing only with the wild data so we need to remove the samples</span>
<span id="cb76-768"><a href="#cb76-768" aria-hidden="true" tabindex="-1"></a>that are not in the wild data.</span>
<span id="cb76-769"><a href="#cb76-769" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-770"><a href="#cb76-770" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Filter the metadata to keep only the procyclic and metacyclic</span>
<span id="cb76-771"><a href="#cb76-771" aria-hidden="true" tabindex="-1"></a>information:</span>
<span id="cb76-772"><a href="#cb76-772" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-775"><a href="#cb76-775" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-776"><a href="#cb76-776" aria-hidden="true" tabindex="-1"></a>meta_pro_meta <span class="ot">&lt;-</span> meta <span class="sc">|&gt;</span></span>
<span id="cb76-777"><a href="#cb76-777" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(stage <span class="sc">!=</span> <span class="st">"amastigotes"</span>)</span>
<span id="cb76-778"><a href="#cb76-778" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-779"><a href="#cb76-779" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-780"><a href="#cb76-780" aria-hidden="true" tabindex="-1"></a>We can now create the DESeqDataSet object. The design formula describes</span>
<span id="cb76-781"><a href="#cb76-781" aria-hidden="true" tabindex="-1"></a>the statistical model. You should recognise the form from previous work.</span>
<span id="cb76-782"><a href="#cb76-782" aria-hidden="true" tabindex="-1"></a>The <span class="in">`~`</span> can be read as "explain by" and on its right hand side are the</span>
<span id="cb76-783"><a href="#cb76-783" aria-hidden="true" tabindex="-1"></a>explanatory variables. That is, the model is counts explained by stage</span>
<span id="cb76-784"><a href="#cb76-784" aria-hidden="true" tabindex="-1"></a>status.</span>
<span id="cb76-785"><a href="#cb76-785" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-786"><a href="#cb76-786" aria-hidden="true" tabindex="-1"></a>Note that:</span>
<span id="cb76-787"><a href="#cb76-787" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-788"><a href="#cb76-788" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The names of the columns in the count matrix have to exactly match</span>
<span id="cb76-789"><a href="#cb76-789" aria-hidden="true" tabindex="-1"></a>    the names of the rows in the metadata dataframe. They also need to</span>
<span id="cb76-790"><a href="#cb76-790" aria-hidden="true" tabindex="-1"></a>    be in the same order.</span>
<span id="cb76-791"><a href="#cb76-791" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The names of the explanatory variables in the design formula have to</span>
<span id="cb76-792"><a href="#cb76-792" aria-hidden="true" tabindex="-1"></a>    match the names of columns in the metadata.</span>
<span id="cb76-793"><a href="#cb76-793" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-794"><a href="#cb76-794" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create the DESeqDataSet object:</span>
<span id="cb76-795"><a href="#cb76-795" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-798"><a href="#cb76-798" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-799"><a href="#cb76-799" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(pro_meta_count_mat,</span>
<span id="cb76-800"><a href="#cb76-800" aria-hidden="true" tabindex="-1"></a>                              <span class="at">colData =</span> meta_pro_meta,</span>
<span id="cb76-801"><a href="#cb76-801" aria-hidden="true" tabindex="-1"></a>                              <span class="at">design =</span> <span class="sc">~</span> stage)</span>
<span id="cb76-802"><a href="#cb76-802" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-803"><a href="#cb76-803" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-804"><a href="#cb76-804" aria-hidden="true" tabindex="-1"></a>The warning "Warning: some variables in design formula are characters,</span>
<span id="cb76-805"><a href="#cb76-805" aria-hidden="true" tabindex="-1"></a>converting to factors" just means that the variable type of stage in the</span>
<span id="cb76-806"><a href="#cb76-806" aria-hidden="true" tabindex="-1"></a>metadata dataframe is "char" and it has been converted into a factor</span>
<span id="cb76-807"><a href="#cb76-807" aria-hidden="true" tabindex="-1"></a>type.</span>
<span id="cb76-808"><a href="#cb76-808" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-809"><a href="#cb76-809" aria-hidden="true" tabindex="-1"></a>To help you understand what the <span class="in">`DESeqDataSet`</span> object we have called</span>
<span id="cb76-810"><a href="#cb76-810" aria-hidden="true" tabindex="-1"></a><span class="in">`dds`</span> contains, we can look its contents</span>
<span id="cb76-811"><a href="#cb76-811" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-812"><a href="#cb76-812" aria-hidden="true" tabindex="-1"></a>The counts are in <span class="in">`dds@assays@data@listData[["counts"]]`</span> and the</span>
<span id="cb76-813"><a href="#cb76-813" aria-hidden="true" tabindex="-1"></a>metadata are in <span class="in">`dds@colData`</span> but the easiest way to see them is to use</span>
<span id="cb76-814"><a href="#cb76-814" aria-hidden="true" tabindex="-1"></a>the <span class="in">`counts()`</span> and <span class="in">`colData()`</span> functions from the **`DESeq2`** package.</span>
<span id="cb76-815"><a href="#cb76-815" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-816"><a href="#cb76-816" aria-hidden="true" tabindex="-1"></a>ğŸ¬ View the counts:</span>
<span id="cb76-817"><a href="#cb76-817" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-820"><a href="#cb76-820" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-821"><a href="#cb76-821" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb76-822"><a href="#cb76-822" aria-hidden="true" tabindex="-1"></a><span class="fu">counts</span>(dds) <span class="sc">|&gt;</span> <span class="fu">View</span>()</span>
<span id="cb76-823"><a href="#cb76-823" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-824"><a href="#cb76-824" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-825"><a href="#cb76-825" aria-hidden="true" tabindex="-1"></a>You should be able to see that this is the same as in</span>
<span id="cb76-826"><a href="#cb76-826" aria-hidden="true" tabindex="-1"></a><span class="in">`pro_meta_count_mat`</span>.</span>
<span id="cb76-827"><a href="#cb76-827" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-828"><a href="#cb76-828" aria-hidden="true" tabindex="-1"></a>ğŸ¬ View the column information:</span>
<span id="cb76-829"><a href="#cb76-829" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-832"><a href="#cb76-832" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-833"><a href="#cb76-833" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(dds)</span>
<span id="cb76-834"><a href="#cb76-834" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-835"><a href="#cb76-835" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-836"><a href="#cb76-836" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-837"><a href="#cb76-837" aria-hidden="true" tabindex="-1"></a>You should be able to see this is the same as in <span class="in">`meta_pro_meta`</span>.</span>
<span id="cb76-838"><a href="#cb76-838" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-839"><a href="#cb76-839" aria-hidden="true" tabindex="-1"></a><span class="fu">### 3. Prepare the normalised counts</span></span>
<span id="cb76-840"><a href="#cb76-840" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-841"><a href="#cb76-841" aria-hidden="true" tabindex="-1"></a>The normalised counts are the counts that have been transformed to</span>
<span id="cb76-842"><a href="#cb76-842" aria-hidden="true" tabindex="-1"></a>account for the library size (i.e., the total number of reads in a</span>
<span id="cb76-843"><a href="#cb76-843" aria-hidden="true" tabindex="-1"></a>sample) and the gene length. We have to first estimate the normalisation</span>
<span id="cb76-844"><a href="#cb76-844" aria-hidden="true" tabindex="-1"></a>factors and store them in the DESeqDataSet object and then we can get</span>
<span id="cb76-845"><a href="#cb76-845" aria-hidden="true" tabindex="-1"></a>the normalised counts.</span>
<span id="cb76-846"><a href="#cb76-846" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-847"><a href="#cb76-847" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Estimate the factors for normalisation and store them in the</span>
<span id="cb76-848"><a href="#cb76-848" aria-hidden="true" tabindex="-1"></a>DESeqDataSet object:</span>
<span id="cb76-849"><a href="#cb76-849" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-852"><a href="#cb76-852" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-853"><a href="#cb76-853" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(dds)</span>
<span id="cb76-854"><a href="#cb76-854" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-855"><a href="#cb76-855" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-856"><a href="#cb76-856" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Look at the factors (just for information):</span>
<span id="cb76-857"><a href="#cb76-857" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-860"><a href="#cb76-860" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-861"><a href="#cb76-861" aria-hidden="true" tabindex="-1"></a><span class="fu">sizeFactors</span>(dds)</span>
<span id="cb76-862"><a href="#cb76-862" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-863"><a href="#cb76-863" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-864"><a href="#cb76-864" aria-hidden="true" tabindex="-1"></a>The normalised counts will be useful to use later. To get the normalised</span>
<span id="cb76-865"><a href="#cb76-865" aria-hidden="true" tabindex="-1"></a>counts we again used the <span class="in">`counts()`</span> function but this time we use the</span>
<span id="cb76-866"><a href="#cb76-866" aria-hidden="true" tabindex="-1"></a><span class="in">`normalized=TRUE`</span> argument.</span>
<span id="cb76-867"><a href="#cb76-867" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-868"><a href="#cb76-868" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Save the normalised to a matrix:</span>
<span id="cb76-869"><a href="#cb76-869" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-872"><a href="#cb76-872" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-873"><a href="#cb76-873" aria-hidden="true" tabindex="-1"></a>normalised_counts <span class="ot">&lt;-</span> <span class="fu">counts</span>(dds, <span class="at">normalized =</span> <span class="cn">TRUE</span>)</span>
<span id="cb76-874"><a href="#cb76-874" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-875"><a href="#cb76-875" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-876"><a href="#cb76-876" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Make a dataframe of the normalised counts, adding a column for the</span>
<span id="cb76-877"><a href="#cb76-877" aria-hidden="true" tabindex="-1"></a>gene ids at the same time:</span>
<span id="cb76-878"><a href="#cb76-878" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-881"><a href="#cb76-881" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-882"><a href="#cb76-882" aria-hidden="true" tabindex="-1"></a>pro_meta_normalised_counts <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(normalised_counts,</span>
<span id="cb76-883"><a href="#cb76-883" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">gene_id =</span> <span class="fu">row.names</span>(normalised_counts))</span>
<span id="cb76-884"><a href="#cb76-884" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-885"><a href="#cb76-885" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-886"><a href="#cb76-886" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-887"><a href="#cb76-887" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4. Differential expression analysis</span></span>
<span id="cb76-888"><a href="#cb76-888" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-889"><a href="#cb76-889" aria-hidden="true" tabindex="-1"></a>We use the <span class="in">`DESeq()`</span> function to do the differential expression</span>
<span id="cb76-890"><a href="#cb76-890" aria-hidden="true" tabindex="-1"></a>analysis. This function fits the statistical model to the data and then</span>
<span id="cb76-891"><a href="#cb76-891" aria-hidden="true" tabindex="-1"></a>uses the model to calculate the significance of the difference between</span>
<span id="cb76-892"><a href="#cb76-892" aria-hidden="true" tabindex="-1"></a>the treatments. It again stores the results in the DESseqDataSet object.</span>
<span id="cb76-893"><a href="#cb76-893" aria-hidden="true" tabindex="-1"></a>Note that the differential expression needs the raw (unnormalised</span>
<span id="cb76-894"><a href="#cb76-894" aria-hidden="true" tabindex="-1"></a>counts) as it does its own normalisation as part of the process.</span>
<span id="cb76-895"><a href="#cb76-895" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-896"><a href="#cb76-896" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Run the differential expression analysis and store the results in the</span>
<span id="cb76-897"><a href="#cb76-897" aria-hidden="true" tabindex="-1"></a>same object:</span>
<span id="cb76-898"><a href="#cb76-898" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-901"><a href="#cb76-901" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-902"><a href="#cb76-902" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb76-903"><a href="#cb76-903" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-904"><a href="#cb76-904" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-905"><a href="#cb76-905" aria-hidden="true" tabindex="-1"></a>The function will take only a few moments to run on this data but can</span>
<span id="cb76-906"><a href="#cb76-906" aria-hidden="true" tabindex="-1"></a>take longer for bigger datasets.</span>
<span id="cb76-907"><a href="#cb76-907" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-908"><a href="#cb76-908" aria-hidden="true" tabindex="-1"></a>We need to define the contrasts we want to test. We want to test the</span>
<span id="cb76-909"><a href="#cb76-909" aria-hidden="true" tabindex="-1"></a>difference between the treatments so we will define the contrast as</span>
<span id="cb76-910"><a href="#cb76-910" aria-hidden="true" tabindex="-1"></a><span class="in">`procyclic`</span> and <span class="in">`metacyclic`</span>.</span>
<span id="cb76-911"><a href="#cb76-911" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-912"><a href="#cb76-912" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Define the contrast:</span>
<span id="cb76-913"><a href="#cb76-913" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-916"><a href="#cb76-916" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-917"><a href="#cb76-917" aria-hidden="true" tabindex="-1"></a>contrast_pro_meta <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"stage"</span>, <span class="st">"procyclic"</span>, <span class="st">"metacyclic"</span>)</span>
<span id="cb76-918"><a href="#cb76-918" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-919"><a href="#cb76-919" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-920"><a href="#cb76-920" aria-hidden="true" tabindex="-1"></a>Note that <span class="in">`stage`</span> is the name of the column in the metadata dataframe</span>
<span id="cb76-921"><a href="#cb76-921" aria-hidden="true" tabindex="-1"></a>and <span class="in">`procyclic`</span> and <span class="in">`metacyclic`</span> are the names of the levels in the</span>
<span id="cb76-922"><a href="#cb76-922" aria-hidden="true" tabindex="-1"></a><span class="in">`stage`</span> column. By putting them in the order <span class="in">`procyclic`</span> , <span class="in">`metacyclic`</span></span>
<span id="cb76-923"><a href="#cb76-923" aria-hidden="true" tabindex="-1"></a>we are saying the fold change will be procyclic / metacyclic. This</span>
<span id="cb76-924"><a href="#cb76-924" aria-hidden="true" tabindex="-1"></a>means:</span>
<span id="cb76-925"><a href="#cb76-925" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-926"><a href="#cb76-926" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>positive log fold changes indicate procyclic <span class="sc">\&gt;</span> metacyclic and</span>
<span id="cb76-927"><a href="#cb76-927" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>negative log fold changes indicates metacyclic <span class="sc">\&gt;</span> procyclic.</span>
<span id="cb76-928"><a href="#cb76-928" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-929"><a href="#cb76-929" aria-hidden="true" tabindex="-1"></a>If we had put them in the order <span class="in">`metacyclic`</span>, <span class="in">`procyclic`</span> we would have</span>
<span id="cb76-930"><a href="#cb76-930" aria-hidden="true" tabindex="-1"></a>the reverse.</span>
<span id="cb76-931"><a href="#cb76-931" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-932"><a href="#cb76-932" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Extract the results from the DESseqDataSet object:</span>
<span id="cb76-933"><a href="#cb76-933" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-936"><a href="#cb76-936" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-937"><a href="#cb76-937" aria-hidden="true" tabindex="-1"></a>results_pro_meta <span class="ot">&lt;-</span> <span class="fu">results</span>(dds,</span>
<span id="cb76-938"><a href="#cb76-938" aria-hidden="true" tabindex="-1"></a>                       <span class="at">contrast =</span> contrast_pro_meta)</span>
<span id="cb76-939"><a href="#cb76-939" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-940"><a href="#cb76-940" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-941"><a href="#cb76-941" aria-hidden="true" tabindex="-1"></a>This will give us the log~2~ fold change, the *p*-value and the adjusted</span>
<span id="cb76-942"><a href="#cb76-942" aria-hidden="true" tabindex="-1"></a>*p*-value for the comparison between procyclic and metacyclic stage for</span>
<span id="cb76-943"><a href="#cb76-943" aria-hidden="true" tabindex="-1"></a>each gene</span>
<span id="cb76-944"><a href="#cb76-944" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-945"><a href="#cb76-945" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Put the results in a dataframe and add the gene ids as a column:</span>
<span id="cb76-946"><a href="#cb76-946" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-949"><a href="#cb76-949" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-950"><a href="#cb76-950" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(results_pro_meta,</span>
<span id="cb76-951"><a href="#cb76-951" aria-hidden="true" tabindex="-1"></a>                          <span class="at">gene_id =</span> <span class="fu">row.names</span>(results_pro_meta))</span>
<span id="cb76-952"><a href="#cb76-952" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-953"><a href="#cb76-953" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-954"><a href="#cb76-954" aria-hidden="true" tabindex="-1"></a>It is useful to have the normalised counts and the statistical results</span>
<span id="cb76-955"><a href="#cb76-955" aria-hidden="true" tabindex="-1"></a>in one dataframe.</span>
<span id="cb76-956"><a href="#cb76-956" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-957"><a href="#cb76-957" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Merge the two dataframes:</span>
<span id="cb76-958"><a href="#cb76-958" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-961"><a href="#cb76-961" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-962"><a href="#cb76-962" aria-hidden="true" tabindex="-1"></a><span class="co"># merge the results with the normalised counts</span></span>
<span id="cb76-963"><a href="#cb76-963" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="ot">&lt;-</span> pro_meta_normalised_counts <span class="sc">|&gt;</span></span>
<span id="cb76-964"><a href="#cb76-964" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(pro_meta_results, <span class="at">by =</span> <span class="st">"gene_id"</span>)</span>
<span id="cb76-965"><a href="#cb76-965" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-966"><a href="#cb76-966" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-967"><a href="#cb76-967" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Add gene information</span><span class="co">](#leishmania-add-info)</span>.</span>
<span id="cb76-968"><a href="#cb76-968" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-969"><a href="#cb76-969" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ­ Stem cells {#stem-cells-differential}</span></span>
<span id="cb76-970"><a href="#cb76-970" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-971"><a href="#cb76-971" aria-hidden="true" tabindex="-1"></a>These are the steps we will take</span>
<span id="cb76-972"><a href="#cb76-972" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-973"><a href="#cb76-973" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Find the genes that are expressed in only one cell type (the prog or</span>
<span id="cb76-974"><a href="#cb76-974" aria-hidden="true" tabindex="-1"></a>    the hspc).</span>
<span id="cb76-975"><a href="#cb76-975" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Prepare the data for differential expression analysis with the</span>
<span id="cb76-976"><a href="#cb76-976" aria-hidden="true" tabindex="-1"></a>    **`scran`** package.</span>
<span id="cb76-977"><a href="#cb76-977" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Do differential expression analysis on the genes using the</span>
<span id="cb76-978"><a href="#cb76-978" aria-hidden="true" tabindex="-1"></a>    **`scran`** package. This needs to be done on the logged normalised</span>
<span id="cb76-979"><a href="#cb76-979" aria-hidden="true" tabindex="-1"></a>    counts.</span>
<span id="cb76-980"><a href="#cb76-980" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-981"><a href="#cb76-981" aria-hidden="true" tabindex="-1"></a><span class="fu">### 1. Genes expressed in one cell type</span></span>
<span id="cb76-982"><a href="#cb76-982" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-983"><a href="#cb76-983" aria-hidden="true" tabindex="-1"></a>The genes expressed in only cell type are those with zeros in all the</span>
<span id="cb76-984"><a href="#cb76-984" aria-hidden="true" tabindex="-1"></a>cells of the other type. We can find these by summing the expression</span>
<span id="cb76-985"><a href="#cb76-985" aria-hidden="true" tabindex="-1"></a>values for each gene across the cells of one type and filtering for</span>
<span id="cb76-986"><a href="#cb76-986" aria-hidden="true" tabindex="-1"></a>those that are zero.</span>
<span id="cb76-987"><a href="#cb76-987" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-988"><a href="#cb76-988" aria-hidden="true" tabindex="-1"></a>To do row wise aggregates such as the sum across rows we can use the</span>
<span id="cb76-989"><a href="#cb76-989" aria-hidden="true" tabindex="-1"></a><span class="in">`rowwise()`</span> function. <span class="in">`c_across()`</span> allows us to use the colon notation</span>
<span id="cb76-990"><a href="#cb76-990" aria-hidden="true" tabindex="-1"></a>to select columns. This is very useful when you have a lot of columns</span>
<span id="cb76-991"><a href="#cb76-991" aria-hidden="true" tabindex="-1"></a>because it would be annoying to have to list all of them.</span>
<span id="cb76-992"><a href="#cb76-992" aria-hidden="true" tabindex="-1"></a><span class="in">`HSPC_001:HSPC_852`</span> means all the columns from <span class="in">`HSPC_001`</span> to <span class="in">`HSPC_852`</span>.</span>
<span id="cb76-993"><a href="#cb76-993" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-994"><a href="#cb76-994" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Find the genes expressed only in progenitor cells (*i.e.*, those that</span>
<span id="cb76-995"><a href="#cb76-995" aria-hidden="true" tabindex="-1"></a>are 0 in every HSPC cell):</span>
<span id="cb76-996"><a href="#cb76-996" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-999"><a href="#cb76-999" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1000"><a href="#cb76-1000" aria-hidden="true" tabindex="-1"></a>hspc_prog <span class="sc">|&gt;</span> </span>
<span id="cb76-1001"><a href="#cb76-1001" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">|&gt;</span> </span>
<span id="cb76-1002"><a href="#cb76-1002" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">sum</span>(<span class="fu">c_across</span>(HSPC_001<span class="sc">:</span>HSPC_852)) <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb76-1003"><a href="#cb76-1003" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1004"><a href="#cb76-1004" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1005"><a href="#cb76-1005" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1006"><a href="#cb76-1006" aria-hidden="true" tabindex="-1"></a>We already know from last week's work that there are no genes that are</span>
<span id="cb76-1007"><a href="#cb76-1007" aria-hidden="true" tabindex="-1"></a>zero across all the cells (both types). If we did not know that, we</span>
<span id="cb76-1008"><a href="#cb76-1008" aria-hidden="true" tabindex="-1"></a>would need to add <span class="in">`|&gt; filter(sum(c_across(Prog_001:Prog_852)) != 0)`</span></span>
<span id="cb76-1009"><a href="#cb76-1009" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1010"><a href="#cb76-1010" aria-hidden="true" tabindex="-1"></a>meaning zero in all the HSPC but not zero in all the Prog</span>
<span id="cb76-1011"><a href="#cb76-1011" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1012"><a href="#cb76-1012" aria-hidden="true" tabindex="-1"></a>â“ How many genes are expressed only in the progenitor cells only</span>
<span id="cb76-1013"><a href="#cb76-1013" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1014"><a href="#cb76-1014" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb76-1015"><a href="#cb76-1015" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1016"><a href="#cb76-1016" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- No genes are expressed only in the progenitor cells. --&gt;</span></span>
<span id="cb76-1017"><a href="#cb76-1017" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1018"><a href="#cb76-1018" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Now you find any genes that are expressed only in the HSPC cells.</span>
<span id="cb76-1019"><a href="#cb76-1019" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1022"><a href="#cb76-1022" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1023"><a href="#cb76-1023" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb76-1024"><a href="#cb76-1024" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-1025"><a href="#cb76-1025" aria-hidden="true" tabindex="-1"></a>hspc_prog <span class="sc">|&gt;</span> </span>
<span id="cb76-1026"><a href="#cb76-1026" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">|&gt;</span> </span>
<span id="cb76-1027"><a href="#cb76-1027" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">sum</span>(<span class="fu">c_across</span>(Prog_001<span class="sc">:</span>Prog_852)) <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb76-1028"><a href="#cb76-1028" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1029"><a href="#cb76-1029" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1030"><a href="#cb76-1030" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1031"><a href="#cb76-1031" aria-hidden="true" tabindex="-1"></a>â“ How many genes are expressed only in the HSPC cells?</span>
<span id="cb76-1032"><a href="#cb76-1032" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1033"><a href="#cb76-1033" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb76-1034"><a href="#cb76-1034" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1035"><a href="#cb76-1035" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- No genes genes are expressed only in the HSPC cells? --&gt;</span></span>
<span id="cb76-1036"><a href="#cb76-1036" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1037"><a href="#cb76-1037" aria-hidden="true" tabindex="-1"></a>â“ Do the results make sense to you in light of what you know about the</span>
<span id="cb76-1038"><a href="#cb76-1038" aria-hidden="true" tabindex="-1"></a>biology?</span>
<span id="cb76-1039"><a href="#cb76-1039" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1040"><a href="#cb76-1040" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb76-1041"><a href="#cb76-1041" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1042"><a href="#cb76-1042" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- I'm not sure what to expect here. However, when there are many  --&gt;</span></span>
<span id="cb76-1043"><a href="#cb76-1043" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1044"><a href="#cb76-1044" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- "replicates" it is less likely that a gene will be expressed  --&gt;</span></span>
<span id="cb76-1045"><a href="#cb76-1045" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1046"><a href="#cb76-1046" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- in only one group. It's more likely that a gene will be --&gt;</span></span>
<span id="cb76-1047"><a href="#cb76-1047" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1048"><a href="#cb76-1048" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- expressed in both groups but at different levels. --&gt;</span></span>
<span id="cb76-1049"><a href="#cb76-1049" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1050"><a href="#cb76-1050" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- Domain expertise is important when analysing data so  --&gt;</span></span>
<span id="cb76-1051"><a href="#cb76-1051" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1052"><a href="#cb76-1052" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- discuss your results with Jillian and your group. --&gt;</span></span>
<span id="cb76-1053"><a href="#cb76-1053" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1054"><a href="#cb76-1054" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Write all the genes that are expressed one cell type only to file</span>
<span id="cb76-1055"><a href="#cb76-1055" aria-hidden="true" tabindex="-1"></a>(saved in <span class="in">`results`</span>)</span>
<span id="cb76-1056"><a href="#cb76-1056" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1059"><a href="#cb76-1059" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1060"><a href="#cb76-1060" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-1061"><a href="#cb76-1061" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb76-1062"><a href="#cb76-1062" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb76-1063"><a href="#cb76-1063" aria-hidden="true" tabindex="-1"></a><span class="co"># If we had any genes expressed in only one cell type we would write </span></span>
<span id="cb76-1064"><a href="#cb76-1064" aria-hidden="true" tabindex="-1"></a><span class="co"># them to a file like this. </span></span>
<span id="cb76-1065"><a href="#cb76-1065" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(hspc_prog_hspc_only, <span class="st">"results/hspc_prog_hspc_only.csv"</span>)</span>
<span id="cb76-1066"><a href="#cb76-1066" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(hspc_prog_prog_only, <span class="st">"results/hspc_prog_prog_only.csv"</span>)</span>
<span id="cb76-1067"><a href="#cb76-1067" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1068"><a href="#cb76-1068" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1069"><a href="#cb76-1069" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2. Prepare the data for analysis with **`scran`**</span></span>
<span id="cb76-1070"><a href="#cb76-1070" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1071"><a href="#cb76-1071" aria-hidden="true" tabindex="-1"></a>**`scran`** can use a matrix or a dataframe of counts but these must be</span>
<span id="cb76-1072"><a href="#cb76-1072" aria-hidden="true" tabindex="-1"></a>log normalised counts. If using a dataframe, the columns must only</span>
<span id="cb76-1073"><a href="#cb76-1073" aria-hidden="true" tabindex="-1"></a>contain the expression values (not the gene ids). The rows can be named</span>
<span id="cb76-1074"><a href="#cb76-1074" aria-hidden="true" tabindex="-1"></a>to retain the gene ids.</span>
<span id="cb76-1075"><a href="#cb76-1075" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1076"><a href="#cb76-1076" aria-hidden="true" tabindex="-1"></a><span class="in">`hspc_prog`</span> is a dataframe so we will use the ensembl gene ids to name</span>
<span id="cb76-1077"><a href="#cb76-1077" aria-hidden="true" tabindex="-1"></a>the rows and remove the gene ids from the dataframe.</span>
<span id="cb76-1078"><a href="#cb76-1078" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1079"><a href="#cb76-1079" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add the gene ids as the row names:</span>
<span id="cb76-1080"><a href="#cb76-1080" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1083"><a href="#cb76-1083" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1084"><a href="#cb76-1084" aria-hidden="true" tabindex="-1"></a>hspc_prog <span class="ot">&lt;-</span> hspc_prog <span class="sc">|&gt;</span></span>
<span id="cb76-1085"><a href="#cb76-1085" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"ensembl_gene_id"</span>)</span>
<span id="cb76-1086"><a href="#cb76-1086" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1087"><a href="#cb76-1087" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1088"><a href="#cb76-1088" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1089"><a href="#cb76-1089" aria-hidden="true" tabindex="-1"></a>Like **`DESeq2`**, **`scran`** needs metadata to define which columns</span>
<span id="cb76-1090"><a href="#cb76-1090" aria-hidden="true" tabindex="-1"></a>were in which group. Instead of having this is a file, we will create a</span>
<span id="cb76-1091"><a href="#cb76-1091" aria-hidden="true" tabindex="-1"></a>vector that indicates which column belongs to which cell type.</span>
<span id="cb76-1092"><a href="#cb76-1092" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1093"><a href="#cb76-1093" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a vector that indicates which column belongs to which cell</span>
<span id="cb76-1094"><a href="#cb76-1094" aria-hidden="true" tabindex="-1"></a>type:</span>
<span id="cb76-1095"><a href="#cb76-1095" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1098"><a href="#cb76-1098" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1099"><a href="#cb76-1099" aria-hidden="true" tabindex="-1"></a>n_hspc <span class="ot">&lt;-</span> <span class="dv">701</span></span>
<span id="cb76-1100"><a href="#cb76-1100" aria-hidden="true" tabindex="-1"></a>n_prog <span class="ot">&lt;-</span> <span class="dv">798</span></span>
<span id="cb76-1101"><a href="#cb76-1101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1102"><a href="#cb76-1102" aria-hidden="true" tabindex="-1"></a>cell_type <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"hspc"</span>,<span class="st">"prog"</span>), </span>
<span id="cb76-1103"><a href="#cb76-1103" aria-hidden="true" tabindex="-1"></a>                 <span class="at">times =</span> <span class="fu">c</span>(n_hspc, n_prog))</span>
<span id="cb76-1104"><a href="#cb76-1104" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1105"><a href="#cb76-1105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1106"><a href="#cb76-1106" aria-hidden="true" tabindex="-1"></a>The number of times each cell type is repeated is the number of cells of</span>
<span id="cb76-1107"><a href="#cb76-1107" aria-hidden="true" tabindex="-1"></a>that type. Do check that the length of the <span class="in">`cell_type`</span> vector is the</span>
<span id="cb76-1108"><a href="#cb76-1108" aria-hidden="true" tabindex="-1"></a>same as the number of columns in the <span class="in">`hspc_prog`</span> dataframe.</span>
<span id="cb76-1109"><a href="#cb76-1109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1110"><a href="#cb76-1110" aria-hidden="true" tabindex="-1"></a><span class="fu">### 3. Differential expression analysis</span></span>
<span id="cb76-1111"><a href="#cb76-1111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1112"><a href="#cb76-1112" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load the **`scran`** package:</span>
<span id="cb76-1113"><a href="#cb76-1113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1116"><a href="#cb76-1116" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1117"><a href="#cb76-1117" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-1118"><a href="#cb76-1118" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb76-1119"><a href="#cb76-1119" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1120"><a href="#cb76-1120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1121"><a href="#cb76-1121" aria-hidden="true" tabindex="-1"></a>Differential expression is carried out with the <span class="in">`findMarkers()`</span></span>
<span id="cb76-1122"><a href="#cb76-1122" aria-hidden="true" tabindex="-1"></a>function. It takes two arguments. The first argument is the dataframe</span>
<span id="cb76-1123"><a href="#cb76-1123" aria-hidden="true" tabindex="-1"></a>containing the data and the second argument is the vector indicating</span>
<span id="cb76-1124"><a href="#cb76-1124" aria-hidden="true" tabindex="-1"></a>which columns are in which cell type. Make sure that the order of the</span>
<span id="cb76-1125"><a href="#cb76-1125" aria-hidden="true" tabindex="-1"></a>cell types in the vector is appropriate for the order of the columns in</span>
<span id="cb76-1126"><a href="#cb76-1126" aria-hidden="true" tabindex="-1"></a>the dataframe.</span>
<span id="cb76-1127"><a href="#cb76-1127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1128"><a href="#cb76-1128" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Run the differential expression analysis:</span>
<span id="cb76-1129"><a href="#cb76-1129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1132"><a href="#cb76-1132" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1133"><a href="#cb76-1133" aria-hidden="true" tabindex="-1"></a>results_hspc_prog <span class="ot">&lt;-</span> <span class="fu">findMarkers</span>(hspc_prog, </span>
<span id="cb76-1134"><a href="#cb76-1134" aria-hidden="true" tabindex="-1"></a>                             cell_type)</span>
<span id="cb76-1135"><a href="#cb76-1135" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1136"><a href="#cb76-1136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1137"><a href="#cb76-1137" aria-hidden="true" tabindex="-1"></a>The output is a list object which, rather unnecessarily, includes two</span>
<span id="cb76-1138"><a href="#cb76-1138" aria-hidden="true" tabindex="-1"></a>dataframes. This is not really necessary as the results are the same</span>
<span id="cb76-1139"><a href="#cb76-1139" aria-hidden="true" tabindex="-1"></a>except for the fold change having a different sign.</span>
<span id="cb76-1140"><a href="#cb76-1140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1141"><a href="#cb76-1141" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The dataframe <span class="in">`results_hspc_prog$prog`</span> is log prog - log hspc</span>
<span id="cb76-1142"><a href="#cb76-1142" aria-hidden="true" tabindex="-1"></a>    (*i.e.*,Prog/HSPC). This means:</span>
<span id="cb76-1143"><a href="#cb76-1143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1144"><a href="#cb76-1144" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Positive fold change: prog is higher than hspc</span>
<span id="cb76-1145"><a href="#cb76-1145" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Negative fold change: hspc is higher than prog</span>
<span id="cb76-1146"><a href="#cb76-1146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1149"><a href="#cb76-1149" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1150"><a href="#cb76-1150" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-1151"><a href="#cb76-1151" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(results_hspc_prog<span class="sc">$</span>prog, </span>
<span id="cb76-1152"><a href="#cb76-1152" aria-hidden="true" tabindex="-1"></a>           <span class="at">ensembl_gene_id =</span> <span class="fu">row.names</span>(results_hspc_prog<span class="sc">$</span>prog)) <span class="sc">|&gt;</span></span>
<span id="cb76-1153"><a href="#cb76-1153" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>() <span class="sc">|&gt;</span> </span>
<span id="cb76-1154"><a href="#cb76-1154" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="at">cap =</span> <span class="st">"The results_hspc_prog$prog dataframe"</span>)</span>
<span id="cb76-1155"><a href="#cb76-1155" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1156"><a href="#cb76-1156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1157"><a href="#cb76-1157" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The dataframe <span class="in">`results_hspc_prog$hspc`</span> is log hspc - log prog</span>
<span id="cb76-1158"><a href="#cb76-1158" aria-hidden="true" tabindex="-1"></a>    (*i.e.*, HSPC/Prog). This means:</span>
<span id="cb76-1159"><a href="#cb76-1159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1160"><a href="#cb76-1160" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Positive fold change: hspc is higher than prog</span>
<span id="cb76-1161"><a href="#cb76-1161" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Negative fold change: prog is higher than hspc</span>
<span id="cb76-1162"><a href="#cb76-1162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1165"><a href="#cb76-1165" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1166"><a href="#cb76-1166" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb76-1167"><a href="#cb76-1167" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(results_hspc_prog<span class="sc">$</span>hspc, </span>
<span id="cb76-1168"><a href="#cb76-1168" aria-hidden="true" tabindex="-1"></a>           <span class="at">ensembl_gene_id =</span> <span class="fu">row.names</span>(results_hspc_prog<span class="sc">$</span>hspc)) <span class="sc">|&gt;</span></span>
<span id="cb76-1169"><a href="#cb76-1169" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>() <span class="sc">|&gt;</span> </span>
<span id="cb76-1170"><a href="#cb76-1170" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="at">cap =</span> <span class="st">"The results_hspc_prog$hspc dataframe. Notice the sign of the fold change is the other way"</span>)</span>
<span id="cb76-1171"><a href="#cb76-1171" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1172"><a href="#cb76-1172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1173"><a href="#cb76-1173" aria-hidden="true" tabindex="-1"></a>We only need one of these results dataframes and we will use the <span class="in">`prog`</span></span>
<span id="cb76-1174"><a href="#cb76-1174" aria-hidden="true" tabindex="-1"></a>one. It does not matter which one you use but you do need keep the</span>
<span id="cb76-1175"><a href="#cb76-1175" aria-hidden="true" tabindex="-1"></a>direction of the foldchange in mind when interpreting the results.</span>
<span id="cb76-1176"><a href="#cb76-1176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1177"><a href="#cb76-1177" aria-hidden="true" tabindex="-1"></a>Notice that the dataframes are ordered by significance rather than the</span>
<span id="cb76-1178"><a href="#cb76-1178" aria-hidden="true" tabindex="-1"></a>original gene order.</span>
<span id="cb76-1179"><a href="#cb76-1179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1180"><a href="#cb76-1180" aria-hidden="true" tabindex="-1"></a>It is useful to have the normalised counts and the statistical results</span>
<span id="cb76-1181"><a href="#cb76-1181" aria-hidden="true" tabindex="-1"></a>in one dataframe to which we will add the gene information from Ensembl.</span>
<span id="cb76-1182"><a href="#cb76-1182" aria-hidden="true" tabindex="-1"></a>Having all the information together will make it easier to interpret the</span>
<span id="cb76-1183"><a href="#cb76-1183" aria-hidden="true" tabindex="-1"></a>results and select genes of interest. We will need to extract the</span>
<span id="cb76-1184"><a href="#cb76-1184" aria-hidden="true" tabindex="-1"></a>results from the list object, add the gene ids and then join with the</span>
<span id="cb76-1185"><a href="#cb76-1185" aria-hidden="true" tabindex="-1"></a>normalised counts.</span>
<span id="cb76-1186"><a href="#cb76-1186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1187"><a href="#cb76-1187" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Extract the results dataframe from the list object and add the gene</span>
<span id="cb76-1188"><a href="#cb76-1188" aria-hidden="true" tabindex="-1"></a>ids as a column:</span>
<span id="cb76-1189"><a href="#cb76-1189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1192"><a href="#cb76-1192" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1193"><a href="#cb76-1193" aria-hidden="true" tabindex="-1"></a>hspc_prog_results <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(results_hspc_prog<span class="sc">$</span>prog, </span>
<span id="cb76-1194"><a href="#cb76-1194" aria-hidden="true" tabindex="-1"></a>                                <span class="at">ensembl_gene_id =</span> <span class="fu">row.names</span>(results_hspc_prog<span class="sc">$</span>prog)) </span>
<span id="cb76-1195"><a href="#cb76-1195" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1196"><a href="#cb76-1196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1197"><a href="#cb76-1197" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Return the ensembl gene ids as a column to the normalised counts:</span>
<span id="cb76-1198"><a href="#cb76-1198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1201"><a href="#cb76-1201" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1202"><a href="#cb76-1202" aria-hidden="true" tabindex="-1"></a>hspc_prog <span class="ot">&lt;-</span> hspc_prog <span class="sc">|&gt;</span></span>
<span id="cb76-1203"><a href="#cb76-1203" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"ensembl_gene_id"</span>)</span>
<span id="cb76-1204"><a href="#cb76-1204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1205"><a href="#cb76-1205" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1206"><a href="#cb76-1206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1207"><a href="#cb76-1207" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Merge the results dataframe with the normalised counts:</span>
<span id="cb76-1208"><a href="#cb76-1208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1211"><a href="#cb76-1211" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1212"><a href="#cb76-1212" aria-hidden="true" tabindex="-1"></a><span class="co"># merge the results with the normalised counts</span></span>
<span id="cb76-1213"><a href="#cb76-1213" aria-hidden="true" tabindex="-1"></a>hspc_prog_results <span class="ot">&lt;-</span> hspc_prog_results <span class="sc">|&gt;</span></span>
<span id="cb76-1214"><a href="#cb76-1214" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(hspc_prog, <span class="at">by =</span> <span class="st">"ensembl_gene_id"</span>)</span>
<span id="cb76-1215"><a href="#cb76-1215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1216"><a href="#cb76-1216" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1217"><a href="#cb76-1217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1218"><a href="#cb76-1218" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Add gene information</span><span class="co">](#stem-add-info)</span>.</span>
<span id="cb76-1219"><a href="#cb76-1219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1220"><a href="#cb76-1220" aria-hidden="true" tabindex="-1"></a><span class="fu"># Add gene information</span></span>
<span id="cb76-1221"><a href="#cb76-1221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1222"><a href="#cb76-1222" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ„ *Arabidopisis* {#arabidopisis-add-info}</span></span>
<span id="cb76-1223"><a href="#cb76-1223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1224"><a href="#cb76-1224" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Ensembl</span><span class="co">](https://www.ensembl.org/index.html)</span> [@martin2023;</span>
<span id="cb76-1225"><a href="#cb76-1225" aria-hidden="true" tabindex="-1"></a>@birney2004]is a bioinformatics project to organise all the biological</span>
<span id="cb76-1226"><a href="#cb76-1226" aria-hidden="true" tabindex="-1"></a>information around the sequences of large genomes. The are a large</span>
<span id="cb76-1227"><a href="#cb76-1227" aria-hidden="true" tabindex="-1"></a>number of databases and</span>
<span id="cb76-1228"><a href="#cb76-1228" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">BioMart</span><span class="co">](https://www.ensembl.org/info/data/biomart/index.html)</span></span>
<span id="cb76-1229"><a href="#cb76-1229" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@smedley2009</span><span class="co">]</span> provides a consistent interface to the material. There</span>
<span id="cb76-1230"><a href="#cb76-1230" aria-hidden="true" tabindex="-1"></a>are web-based tools to use these but the R package **`biomaRt`**</span>
<span id="cb76-1231"><a href="#cb76-1231" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@biomaRt1; @biomaRt2</span><span class="co">]</span> gives you programmatic access making it easier to</span>
<span id="cb76-1232"><a href="#cb76-1232" aria-hidden="true" tabindex="-1"></a>integrate information into R dataframes.</span>
<span id="cb76-1233"><a href="#cb76-1233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1234"><a href="#cb76-1234" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load the **`biomaRt`** <span class="co">[</span><span class="ot">@biomaRt1; @biomaRt2</span><span class="co">]</span> package:</span>
<span id="cb76-1235"><a href="#cb76-1235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1238"><a href="#cb76-1238" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1239"><a href="#cb76-1239" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb76-1240"><a href="#cb76-1240" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1241"><a href="#cb76-1241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1242"><a href="#cb76-1242" aria-hidden="true" tabindex="-1"></a>The **`biomaRt`** package includes a function to list all the available</span>
<span id="cb76-1243"><a href="#cb76-1243" aria-hidden="true" tabindex="-1"></a>datasets</span>
<span id="cb76-1244"><a href="#cb76-1244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1245"><a href="#cb76-1245" aria-hidden="true" tabindex="-1"></a>ğŸ¬ List the Ensembl "marts" available:</span>
<span id="cb76-1246"><a href="#cb76-1246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1249"><a href="#cb76-1249" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1250"><a href="#cb76-1250" aria-hidden="true" tabindex="-1"></a><span class="fu">listEnsemblGenomes</span>()</span>
<span id="cb76-1251"><a href="#cb76-1251" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1252"><a href="#cb76-1252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1253"><a href="#cb76-1253" aria-hidden="true" tabindex="-1"></a><span class="in">`plants_mart`</span> looks like the one we want. We can see what genomes are</span>
<span id="cb76-1254"><a href="#cb76-1254" aria-hidden="true" tabindex="-1"></a>available with names like "Arabidopsis" in this mart using the</span>
<span id="cb76-1255"><a href="#cb76-1255" aria-hidden="true" tabindex="-1"></a><span class="in">`searchDatasets()`</span> function.</span>
<span id="cb76-1256"><a href="#cb76-1256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1257"><a href="#cb76-1257" aria-hidden="true" tabindex="-1"></a>ğŸ¬</span>
<span id="cb76-1258"><a href="#cb76-1258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1261"><a href="#cb76-1261" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1262"><a href="#cb76-1262" aria-hidden="true" tabindex="-1"></a><span class="fu">searchDatasets</span>(<span class="fu">useEnsemblGenomes</span>(<span class="at">biomart =</span> <span class="st">"plants_mart"</span>), </span>
<span id="cb76-1263"><a href="#cb76-1263" aria-hidden="true" tabindex="-1"></a>               <span class="at">pattern =</span> <span class="st">"Arabidopsis"</span>)</span>
<span id="cb76-1264"><a href="#cb76-1264" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1265"><a href="#cb76-1265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1266"><a href="#cb76-1266" aria-hidden="true" tabindex="-1"></a><span class="in">`athaliana_eg_gene`</span> is the Arabidopsis thaliana genes (TAIR10) dataset</span>
<span id="cb76-1267"><a href="#cb76-1267" aria-hidden="true" tabindex="-1"></a>we want.</span>
<span id="cb76-1268"><a href="#cb76-1268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1269"><a href="#cb76-1269" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Connect to the <span class="in">`athaliana_eg_gene`</span> database in <span class="in">`plants_mart`</span>:</span>
<span id="cb76-1270"><a href="#cb76-1270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1273"><a href="#cb76-1273" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1274"><a href="#cb76-1274" aria-hidden="true" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> <span class="fu">useEnsemblGenomes</span>(<span class="at">biomart =</span> <span class="st">"plants_mart"</span>,</span>
<span id="cb76-1275"><a href="#cb76-1275" aria-hidden="true" tabindex="-1"></a>                             <span class="at">dataset =</span> <span class="st">"athaliana_eg_gene"</span>)</span>
<span id="cb76-1276"><a href="#cb76-1276" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1277"><a href="#cb76-1277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1278"><a href="#cb76-1278" aria-hidden="true" tabindex="-1"></a>ğŸ¬ See the the types of information we can retrieve:</span>
<span id="cb76-1279"><a href="#cb76-1279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1282"><a href="#cb76-1282" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1283"><a href="#cb76-1283" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb76-1284"><a href="#cb76-1284" aria-hidden="true" tabindex="-1"></a><span class="fu">listAttributes</span>(<span class="at">mart =</span> ensembl) <span class="sc">|&gt;</span> <span class="fu">View</span>()</span>
<span id="cb76-1285"><a href="#cb76-1285" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1286"><a href="#cb76-1286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1287"><a href="#cb76-1287" aria-hidden="true" tabindex="-1"></a>There are many (1,796!) possible bits of information (attributes) that</span>
<span id="cb76-1288"><a href="#cb76-1288" aria-hidden="true" tabindex="-1"></a>can be obtained.</span>
<span id="cb76-1289"><a href="#cb76-1289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1290"><a href="#cb76-1290" aria-hidden="true" tabindex="-1"></a>We use the <span class="in">`getBM()`</span> function to retrieve information from the database.</span>
<span id="cb76-1291"><a href="#cb76-1291" aria-hidden="true" tabindex="-1"></a>The <span class="in">`filters`</span> argument is used to specified what kind of identifier we</span>
<span id="cb76-1292"><a href="#cb76-1292" aria-hidden="true" tabindex="-1"></a>are supplying in <span class="in">`values`</span> to retrieve information. The <span class="in">`attributes`</span></span>
<span id="cb76-1293"><a href="#cb76-1293" aria-hidden="true" tabindex="-1"></a>argument is used to select the information we want to retrieve. The</span>
<span id="cb76-1294"><a href="#cb76-1294" aria-hidden="true" tabindex="-1"></a><span class="in">`values`</span> argument is used to specify the identifiers. The mart argument</span>
<span id="cb76-1295"><a href="#cb76-1295" aria-hidden="true" tabindex="-1"></a>is used to specify the connection we created.</span>
<span id="cb76-1296"><a href="#cb76-1296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1297"><a href="#cb76-1297" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Get the the gene name and a description. We also retreive the gene id</span>
<span id="cb76-1298"><a href="#cb76-1298" aria-hidden="true" tabindex="-1"></a>so we can later join the information with the results:</span>
<span id="cb76-1299"><a href="#cb76-1299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1302"><a href="#cb76-1302" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1303"><a href="#cb76-1303" aria-hidden="true" tabindex="-1"></a>gene_info <span class="ot">&lt;-</span> <span class="fu">getBM</span>(<span class="at">filters =</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb76-1304"><a href="#cb76-1304" aria-hidden="true" tabindex="-1"></a>                   <span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb76-1305"><a href="#cb76-1305" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"external_gene_name"</span>,</span>
<span id="cb76-1306"><a href="#cb76-1306" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"description"</span>),</span>
<span id="cb76-1307"><a href="#cb76-1307" aria-hidden="true" tabindex="-1"></a>                   <span class="at">values =</span> root_results<span class="sc">$</span>gene_id,</span>
<span id="cb76-1308"><a href="#cb76-1308" aria-hidden="true" tabindex="-1"></a>                   <span class="at">mart =</span> ensembl)</span>
<span id="cb76-1309"><a href="#cb76-1309" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1310"><a href="#cb76-1310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1311"><a href="#cb76-1311" aria-hidden="true" tabindex="-1"></a>You should view the resulting dataframe to see what information is</span>
<span id="cb76-1312"><a href="#cb76-1312" aria-hidden="true" tabindex="-1"></a>available. You can use <span class="in">`glimpse()`</span> or <span class="in">`View()`</span>.</span>
<span id="cb76-1313"><a href="#cb76-1313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1314"><a href="#cb76-1314" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Merge the gene information with the results:</span>
<span id="cb76-1315"><a href="#cb76-1315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1318"><a href="#cb76-1318" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1319"><a href="#cb76-1319" aria-hidden="true" tabindex="-1"></a><span class="co"># join the gene info with the results</span></span>
<span id="cb76-1320"><a href="#cb76-1320" aria-hidden="true" tabindex="-1"></a>root_results <span class="ot">&lt;-</span> root_results <span class="sc">|&gt;</span></span>
<span id="cb76-1321"><a href="#cb76-1321" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(gene_info,</span>
<span id="cb76-1322"><a href="#cb76-1322" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">join_by</span>(gene_id <span class="sc">==</span> ensembl_gene_id))</span>
<span id="cb76-1323"><a href="#cb76-1323" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1324"><a href="#cb76-1324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1325"><a href="#cb76-1325" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Save the results to a file:</span>
<span id="cb76-1326"><a href="#cb76-1326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1329"><a href="#cb76-1329" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1330"><a href="#cb76-1330" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(root_results, <span class="at">file =</span> <span class="st">"results/root_results.csv"</span>)</span>
<span id="cb76-1331"><a href="#cb76-1331" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1332"><a href="#cb76-1332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1333"><a href="#cb76-1333" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Look after future you</span><span class="co">](#look-after-future-you)</span>.</span>
<span id="cb76-1334"><a href="#cb76-1334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1335"><a href="#cb76-1335" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ’‰ *Leishmania* {#leishmania-add-info}</span></span>
<span id="cb76-1336"><a href="#cb76-1336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1337"><a href="#cb76-1337" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>I got the information from</span>
<span id="cb76-1338"><a href="#cb76-1338" aria-hidden="true" tabindex="-1"></a>    <span class="co">[</span><span class="ot">TriTrypDB</span><span class="co">](https://tritrypdb.org/tritrypdb/app/)</span>\</span>
<span id="cb76-1339"><a href="#cb76-1339" aria-hidden="true" tabindex="-1"></a>    which is a functional genomic resource for the Trypanosomatidae and</span>
<span id="cb76-1340"><a href="#cb76-1340" aria-hidden="true" tabindex="-1"></a>    Plasmodidae</span>
<span id="cb76-1341"><a href="#cb76-1341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1342"><a href="#cb76-1342" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="ot">&lt;https://tritrypdb.org/tritrypdb/app/downloads&gt;</span> section</span>
<span id="cb76-1343"><a href="#cb76-1343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1344"><a href="#cb76-1344" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>I downloaded the [*L. mexicana* MHOM/GT/2001/U1103 Full</span>
<span id="cb76-1345"><a href="#cb76-1345" aria-hidden="true" tabindex="-1"></a>    GFF](https://tritrypdb.org/common/downloads/release-68/LmexicanaMHOMGT2001U1103/gff/data/TriTrypDB-68_LmexicanaMHOMGT2001U1103.gff)</span>
<span id="cb76-1346"><a href="#cb76-1346" aria-hidden="true" tabindex="-1"></a>    and extracted the gene information and saved it as</span>
<span id="cb76-1347"><a href="#cb76-1347" aria-hidden="true" tabindex="-1"></a>    <span class="co">[</span><span class="ot">leishmania_mex.xlsx</span><span class="co">](meta/leishmania_mex.xlsx)</span></span>
<span id="cb76-1348"><a href="#cb76-1348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1349"><a href="#cb76-1349" aria-hidden="true" tabindex="-1"></a>We will import this file and join it to the results dataframe.</span>
<span id="cb76-1350"><a href="#cb76-1350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1351"><a href="#cb76-1351" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load the **`readxl`** <span class="co">[</span><span class="ot">@readxl</span><span class="co">]</span> package:</span>
<span id="cb76-1352"><a href="#cb76-1352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1355"><a href="#cb76-1355" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1356"><a href="#cb76-1356" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb76-1357"><a href="#cb76-1357" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb76-1358"><a href="#cb76-1358" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1359"><a href="#cb76-1359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1360"><a href="#cb76-1360" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Import the Xenbase gene information file:</span>
<span id="cb76-1361"><a href="#cb76-1361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1364"><a href="#cb76-1364" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1365"><a href="#cb76-1365" aria-hidden="true" tabindex="-1"></a>gene_info <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">"meta/leishmania_mex.xlsx"</span>) </span>
<span id="cb76-1366"><a href="#cb76-1366" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1367"><a href="#cb76-1367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1368"><a href="#cb76-1368" aria-hidden="true" tabindex="-1"></a>You should view the resulting dataframe to see what information is</span>
<span id="cb76-1369"><a href="#cb76-1369" aria-hidden="true" tabindex="-1"></a>available. You can use <span class="in">`glimpse()`</span> or <span class="in">`View()`</span>.</span>
<span id="cb76-1370"><a href="#cb76-1370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1371"><a href="#cb76-1371" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Merge the gene information with the results:</span>
<span id="cb76-1372"><a href="#cb76-1372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1375"><a href="#cb76-1375" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1376"><a href="#cb76-1376" aria-hidden="true" tabindex="-1"></a><span class="co"># join the gene info with the results</span></span>
<span id="cb76-1377"><a href="#cb76-1377" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="ot">&lt;-</span> pro_meta_results <span class="sc">|&gt;</span></span>
<span id="cb76-1378"><a href="#cb76-1378" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(gene_info, <span class="at">by =</span> <span class="st">"gene_id"</span>)</span>
<span id="cb76-1379"><a href="#cb76-1379" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1380"><a href="#cb76-1380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1381"><a href="#cb76-1381" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Save the results to a file:</span>
<span id="cb76-1382"><a href="#cb76-1382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1385"><a href="#cb76-1385" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1386"><a href="#cb76-1386" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(pro_meta_results, <span class="at">file =</span> <span class="st">"results/pro_meta_results.csv"</span>)</span>
<span id="cb76-1387"><a href="#cb76-1387" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1388"><a href="#cb76-1388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1389"><a href="#cb76-1389" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Look after future you</span><span class="co">](#look-after-future-you)</span>.</span>
<span id="cb76-1390"><a href="#cb76-1390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1391"><a href="#cb76-1391" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ­ Stem cells {#stem-cells-add-info}</span></span>
<span id="cb76-1392"><a href="#cb76-1392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1393"><a href="#cb76-1393" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Ensembl</span><span class="co">](https://www.ensembl.org/index.html)</span> [@martin2023;</span>
<span id="cb76-1394"><a href="#cb76-1394" aria-hidden="true" tabindex="-1"></a>@birney2004]is a bioinformatics project to organise all the biological</span>
<span id="cb76-1395"><a href="#cb76-1395" aria-hidden="true" tabindex="-1"></a>information around the sequences of large genomes. The are a large</span>
<span id="cb76-1396"><a href="#cb76-1396" aria-hidden="true" tabindex="-1"></a>number of databases but</span>
<span id="cb76-1397"><a href="#cb76-1397" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">BioMart</span><span class="co">](https://www.ensembl.org/info/data/biomart/index.html)</span></span>
<span id="cb76-1398"><a href="#cb76-1398" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@smedley2009</span><span class="co">]</span> provides a consistent interface to the material. There</span>
<span id="cb76-1399"><a href="#cb76-1399" aria-hidden="true" tabindex="-1"></a>are web-based tools to use these but the R package **`biomaRt`**</span>
<span id="cb76-1400"><a href="#cb76-1400" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@biomaRt1; @biomaRt2</span><span class="co">]</span> gives you programmatic access making it easier to</span>
<span id="cb76-1401"><a href="#cb76-1401" aria-hidden="true" tabindex="-1"></a>integrate information into R dataframes</span>
<span id="cb76-1402"><a href="#cb76-1402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1403"><a href="#cb76-1403" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load the **`biomaRt`** <span class="co">[</span><span class="ot">@biomaRt1; @biomaRt2</span><span class="co">]</span> package:</span>
<span id="cb76-1404"><a href="#cb76-1404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1407"><a href="#cb76-1407" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1408"><a href="#cb76-1408" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb76-1409"><a href="#cb76-1409" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb76-1410"><a href="#cb76-1410" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1411"><a href="#cb76-1411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1412"><a href="#cb76-1412" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Connect to the mouse database</span>
<span id="cb76-1413"><a href="#cb76-1413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1416"><a href="#cb76-1416" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1417"><a href="#cb76-1417" aria-hidden="true" tabindex="-1"></a><span class="co"># Connect to the mouse database</span></span>
<span id="cb76-1418"><a href="#cb76-1418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1419"><a href="#cb76-1419" aria-hidden="true" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> <span class="fu">useMart</span>(<span class="at">biomart =</span> <span class="st">"ensembl"</span>, </span>
<span id="cb76-1420"><a href="#cb76-1420" aria-hidden="true" tabindex="-1"></a>                   <span class="at">dataset =</span> <span class="st">"mmusculus_gene_ensembl"</span>)</span>
<span id="cb76-1421"><a href="#cb76-1421" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1422"><a href="#cb76-1422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1423"><a href="#cb76-1423" aria-hidden="true" tabindex="-1"></a>ğŸ¬ See information we can retrieve:</span>
<span id="cb76-1424"><a href="#cb76-1424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1427"><a href="#cb76-1427" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1428"><a href="#cb76-1428" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb76-1429"><a href="#cb76-1429" aria-hidden="true" tabindex="-1"></a><span class="co"># See what information we can retrieve</span></span>
<span id="cb76-1430"><a href="#cb76-1430" aria-hidden="true" tabindex="-1"></a><span class="fu">listAttributes</span>(<span class="at">mart =</span> ensembl) <span class="sc">|&gt;</span> <span class="fu">View</span>()</span>
<span id="cb76-1431"><a href="#cb76-1431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1432"><a href="#cb76-1432" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1433"><a href="#cb76-1433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1434"><a href="#cb76-1434" aria-hidden="true" tabindex="-1"></a>There are many (2,988!) possible bits of information (attributes) that</span>
<span id="cb76-1435"><a href="#cb76-1435" aria-hidden="true" tabindex="-1"></a>can be obtained.</span>
<span id="cb76-1436"><a href="#cb76-1436" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1437"><a href="#cb76-1437" aria-hidden="true" tabindex="-1"></a>We use the <span class="in">`getBM()`</span> function to retrieve information from the database.</span>
<span id="cb76-1438"><a href="#cb76-1438" aria-hidden="true" tabindex="-1"></a>The <span class="in">`filters`</span> argument is used to specified what kind of identifier we</span>
<span id="cb76-1439"><a href="#cb76-1439" aria-hidden="true" tabindex="-1"></a>are supplying in <span class="in">`values`</span> to retrieve information. The <span class="in">`attributes`</span></span>
<span id="cb76-1440"><a href="#cb76-1440" aria-hidden="true" tabindex="-1"></a>argument is used to select the information we want to retrieve. The</span>
<span id="cb76-1441"><a href="#cb76-1441" aria-hidden="true" tabindex="-1"></a><span class="in">`values`</span> argument is used to specify the identifiers. The mart argument</span>
<span id="cb76-1442"><a href="#cb76-1442" aria-hidden="true" tabindex="-1"></a>is used to specify the connection we created.</span>
<span id="cb76-1443"><a href="#cb76-1443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1444"><a href="#cb76-1444" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Get the the gene name and a description. We also retrieve the gene id</span>
<span id="cb76-1445"><a href="#cb76-1445" aria-hidden="true" tabindex="-1"></a>so we can later join the information with the results:</span>
<span id="cb76-1446"><a href="#cb76-1446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1449"><a href="#cb76-1449" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1450"><a href="#cb76-1450" aria-hidden="true" tabindex="-1"></a>gene_info <span class="ot">&lt;-</span> <span class="fu">getBM</span>(<span class="at">filters =</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb76-1451"><a href="#cb76-1451" aria-hidden="true" tabindex="-1"></a>                   <span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb76-1452"><a href="#cb76-1452" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"external_gene_name"</span>,</span>
<span id="cb76-1453"><a href="#cb76-1453" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"description"</span>),</span>
<span id="cb76-1454"><a href="#cb76-1454" aria-hidden="true" tabindex="-1"></a>                   <span class="at">values =</span> hspc_prog_results<span class="sc">$</span>ensembl_gene_id,</span>
<span id="cb76-1455"><a href="#cb76-1455" aria-hidden="true" tabindex="-1"></a>                   <span class="at">mart =</span> ensembl)</span>
<span id="cb76-1456"><a href="#cb76-1456" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1457"><a href="#cb76-1457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1458"><a href="#cb76-1458" aria-hidden="true" tabindex="-1"></a>You should view the resulting dataframe to see what information is</span>
<span id="cb76-1459"><a href="#cb76-1459" aria-hidden="true" tabindex="-1"></a>available. You can use <span class="in">`glimpse()`</span> or <span class="in">`View()`</span>. Notice the dataframe</span>
<span id="cb76-1460"><a href="#cb76-1460" aria-hidden="true" tabindex="-1"></a>returned only has 279 rows - one of the ids does not have information.</span>
<span id="cb76-1461"><a href="#cb76-1461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1462"><a href="#cb76-1462" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Merge the gene information with the results:</span>
<span id="cb76-1463"><a href="#cb76-1463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1466"><a href="#cb76-1466" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1467"><a href="#cb76-1467" aria-hidden="true" tabindex="-1"></a><span class="co"># join the gene info with the results</span></span>
<span id="cb76-1468"><a href="#cb76-1468" aria-hidden="true" tabindex="-1"></a>hspc_prog_results <span class="ot">&lt;-</span> hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb76-1469"><a href="#cb76-1469" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(gene_info, <span class="at">by =</span> <span class="st">"ensembl_gene_id"</span>)</span>
<span id="cb76-1470"><a href="#cb76-1470" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1471"><a href="#cb76-1471" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1472"><a href="#cb76-1472" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Save the results to a file:</span>
<span id="cb76-1473"><a href="#cb76-1473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1476"><a href="#cb76-1476" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb76-1477"><a href="#cb76-1477" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(hspc_prog_results, <span class="at">file =</span> <span class="st">"results/hspc_prog_results.csv"</span>)</span>
<span id="cb76-1478"><a href="#cb76-1478" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb76-1479"><a href="#cb76-1479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1480"><a href="#cb76-1480" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Look after future you</span><span class="co">](#look-after-future-you)</span>.</span>
<span id="cb76-1481"><a href="#cb76-1481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1482"><a href="#cb76-1482" aria-hidden="true" tabindex="-1"></a><span class="fu"># ğŸ¤— Look after future you! {#look-after-future-you}</span></span>
<span id="cb76-1483"><a href="#cb76-1483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1484"><a href="#cb76-1484" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Go through you script and tidy up. You might need to :</span>
<span id="cb76-1485"><a href="#cb76-1485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1486"><a href="#cb76-1486" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>collect together library statements at the top of the script</span>
<span id="cb76-1487"><a href="#cb76-1487" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>remove code that you needed to start today but which wouldn't be</span>
<span id="cb76-1488"><a href="#cb76-1488" aria-hidden="true" tabindex="-1"></a>    needed if you were running the script from the top (e.g., importing)</span>
<span id="cb76-1489"><a href="#cb76-1489" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>edit your comments for clarity</span>
<span id="cb76-1490"><a href="#cb76-1490" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>rename variables for consistency or clarity</span>
<span id="cb76-1491"><a href="#cb76-1491" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>remove house keeping or exploratory code</span>
<span id="cb76-1492"><a href="#cb76-1492" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>restyle code, add code section headers etc</span>
<span id="cb76-1493"><a href="#cb76-1493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1494"><a href="#cb76-1494" aria-hidden="true" tabindex="-1"></a><span class="fu"># ğŸ¥³ Finished</span></span>
<span id="cb76-1495"><a href="#cb76-1495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1496"><a href="#cb76-1496" aria-hidden="true" tabindex="-1"></a>Well Done!</span>
<span id="cb76-1497"><a href="#cb76-1497" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1498"><a href="#cb76-1498" aria-hidden="true" tabindex="-1"></a><span class="fu"># Independent study following the workshop</span></span>
<span id="cb76-1499"><a href="#cb76-1499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1500"><a href="#cb76-1500" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Consolidate</span><span class="co">](study_after_workshop.qmd)</span></span>
<span id="cb76-1501"><a href="#cb76-1501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1502"><a href="#cb76-1502" aria-hidden="true" tabindex="-1"></a><span class="fu"># The Code file</span></span>
<span id="cb76-1503"><a href="#cb76-1503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1504"><a href="#cb76-1504" aria-hidden="true" tabindex="-1"></a>These contain all the code needed in the workshop even where it is not</span>
<span id="cb76-1505"><a href="#cb76-1505" aria-hidden="true" tabindex="-1"></a>visible on the webpage.</span>
<span id="cb76-1506"><a href="#cb76-1506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1507"><a href="#cb76-1507" aria-hidden="true" tabindex="-1"></a>The <span class="co">[</span><span class="ot">workshop.qmd</span><span class="co">](workshop.qmd)</span> file is the file I use to compile the</span>
<span id="cb76-1508"><a href="#cb76-1508" aria-hidden="true" tabindex="-1"></a>practical. Qmd stands for Quarto markdown. It allows code and ordinary</span>
<span id="cb76-1509"><a href="#cb76-1509" aria-hidden="true" tabindex="-1"></a>text to be interleaved to produce well-formatted reports including</span>
<span id="cb76-1510"><a href="#cb76-1510" aria-hidden="true" tabindex="-1"></a>webpages. View the source code for this workshop using the <span class="in">`&lt;/&gt; Code`</span></span>
<span id="cb76-1511"><a href="#cb76-1511" aria-hidden="true" tabindex="-1"></a>button at the top of the page. Coding and thinking answers are marked</span>
<span id="cb76-1512"><a href="#cb76-1512" aria-hidden="true" tabindex="-1"></a>with <span class="in">`#---CODING ANSWER---`</span> and <span class="in">`#---THINKING ANSWER---`</span></span>
<span id="cb76-1513"><a href="#cb76-1513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1514"><a href="#cb76-1514" aria-hidden="true" tabindex="-1"></a>Pages made with R <span class="co">[</span><span class="ot">@R-core</span><span class="co">]</span>, Quarto <span class="co">[</span><span class="ot">@Allaire_Quarto_2024</span><span class="co">]</span>, <span class="in">`knitr`</span></span>
<span id="cb76-1515"><a href="#cb76-1515" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@knitr1; @knitr2; @knitr3</span><span class="co">]</span>, <span class="in">`kableExtra`</span> <span class="co">[</span><span class="ot">@kableExtra</span><span class="co">]</span></span>
<span id="cb76-1516"><a href="#cb76-1516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-1517"><a href="#cb76-1517" aria-hidden="true" tabindex="-1"></a><span class="fu"># References</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Rand E (2024). Data Analysis for Group Project. <a href="https://3mmarand.github.io/BIO00088H-data/" class="uri">https://3mmarand.github.io/BIO00088H-data/</a>.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/er_13">
      <i class="bi bi-twitter" role="img" aria-label="Twitter">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/3mmaRand/BIO00088H-data">
      <i class="bi bi-github" role="img" aria-label="BIO00088H-data GitHub">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer><script>
document.addEventListener("DOMContentLoaded", function() {
  // grab all links under â€œOther Linksâ€ in the right-hand TOC
  const links = document.querySelectorAll('.quarto-other-links ul li a');
  links.forEach(function(a) {
    // find the one whose text is exactly â€œMake a suggestionâ€
    if (a.textContent.trim() === 'Make a suggestion') {
      const pageUrl   = encodeURIComponent(window.location.href);
      const pageTitle = encodeURIComponent(document.title);
      a.href = 'https://github.com/3mmaRand/BIO00088H-data/issues/new'
             + '?title=' + pageTitle
             + '&body='  + pageUrl;
      a.target = '_blank';
    }
  });
});
</script>


</body></html>