<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.33">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Emma Rand">
<title>Self study – Data Analysis for Group Project</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-bc185b5c5bdbcb35c2eb49d8a876ef70.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" class="quarto-color-scheme-extra" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-72ac799c4abbddf31cb463dd8fc99f5d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark-25ac7be22ffa75a8e8541e30a6dcc9a3.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="../../site_libs/bootstrap/bootstrap-72ac799c4abbddf31cb463dd8fc99f5d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme-extra" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-TVF95J593T"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-TVF95J593T', { 'anonymize_ip': true});
</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar floating nav-fixed quarto-light"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = false;
    const darkModeDefault = authorPrefersDark;
      document.querySelector('link#quarto-text-highlighting-styles.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
      document.querySelector('link#quarto-bootstrap.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
      window.dispatchEvent(new Event('resize'));
    };
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Data Analysis for Group Project</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Welcome!</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../core/core.html"> 
<span class="menu-text">Core Supporting Info</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../transcriptomics/transcriptomics.html" aria-current="page"> 
<span class="menu-text">Transcriptomics</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../images/images.html"> 
<span class="menu-text">Image Analysis</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../structures/structures.html"> 
<span class="menu-text">Structure Analysis</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <a href="https://github.com/3mmaRand/BIO00088H-data" title="BIO00088H-data GitHub" class="quarto-navigation-tool px-1" aria-label="BIO00088H-data GitHub"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/semester-2/stem_cell_variant.html">Semester 2: Extras</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/semester-2/stem_cell_variant.html">Additional 🐭 stem cell labels</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/transcriptomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Transcriptomics Data Analysis for Group Project</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 3: Hello data!</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 4: Statistical Analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 5: Visualising and Interpreting</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Semester 2: Extras</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/semester-2/stem_cell_variant.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Additional 🐭 stem cell labels</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Kelly’s Project</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/kelly/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#set-up" id="toc-set-up" class="nav-link" data-scroll-target="#set-up">Set up</a></li>
  <li>
<a href="#import-and-combine-data" id="toc-import-and-combine-data" class="nav-link" data-scroll-target="#import-and-combine-data">Import and combine data</a>
  <ul>
<li><a href="#the-raw-data" id="toc-the-raw-data" class="nav-link" data-scroll-target="#the-raw-data">The raw data</a></li>
  <li><a href="#gene-information" id="toc-gene-information" class="nav-link" data-scroll-target="#gene-information">Gene information</a></li>
  <li><a href="#type-categorisation-data" id="toc-type-categorisation-data" class="nav-link" data-scroll-target="#type-categorisation-data">12-type categorisation data</a></li>
  </ul>
</li>
  <li>
<a href="#expression-of-a-specific-gene-in-each-cell-type" id="toc-expression-of-a-specific-gene-in-each-cell-type" class="nav-link" data-scroll-target="#expression-of-a-specific-gene-in-each-cell-type">Expression of a specific gene in each cell type</a>
  <ul>
<li><a href="#pivot-to-long-form" id="toc-pivot-to-long-form" class="nav-link" data-scroll-target="#pivot-to-long-form">Pivot to long form</a></li>
  <li>
<a href="#add-the-two-types-of-categorisation-to-the-expression-data" id="toc-add-the-two-types-of-categorisation-to-the-expression-data" class="nav-link" data-scroll-target="#add-the-two-types-of-categorisation-to-the-expression-data">Add the two types of categorisation to the expression data</a>
  <ul class="collapse">
<li><a href="#plot-gene-expression-under-3-type-categorisation" id="toc-plot-gene-expression-under-3-type-categorisation" class="nav-link" data-scroll-target="#plot-gene-expression-under-3-type-categorisation">Plot gene expression under 3-type categorisation</a></li>
  <li><a href="#plot-expression-under-12-type-categorisation" id="toc-plot-expression-under-12-type-categorisation" class="nav-link" data-scroll-target="#plot-expression-under-12-type-categorisation">Plot expression under 12-type categorisation</a></li>
  </ul>
</li>
  </ul>
</li>
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/3mmaRand/BIO00088H-data/edit/main/transcriptomics/semester-2/stem_cell_variant.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li></ul></div><div class="quarto-other-links"><h2>Other Links</h2><ul><li><a href="#"><i class="bi bi-chat-left-dots"></i>Make a suggestion</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/semester-2/stem_cell_variant.html">Semester 2: Extras</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/semester-2/stem_cell_variant.html">Additional 🐭 stem cell labels</a></li></ol></nav><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Self study</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
<p class="subtitle lead">Semester 2: Additional 🐭 stem cell labels</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Emma Rand </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">19 August, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header><section id="introduction" class="level1"><h1>Introduction</h1>
<p>We have analysed the <span class="citation" data-cites="nestorowa2016">Nestorowa et al. (<a href="#ref-nestorowa2016" role="doc-biblioref">2016</a>)</span> stem cells categorised as LTHSC, HSPC, or Prog. The cell types were determined from flowcytometry markers. Additional markers allow the cells to be categorised into 12 other types (including unknown): <code>CMP</code>, <code>GMP</code>, <code>LMPP</code>, <code>LTHSC</code>, <code>LTHSC-ESLAM</code>, <code>MEP</code>, <code>MPP_broad</code>, <code>MPP1</code>, <code>MPP2</code>, <code>MPP3</code>, <code>STHSC</code>, <code>unknown</code>. These cell types are mutually exclusive, meaning that a cell can only be one of these types.</p>
<p>In this self-study, you will:</p>
<ul>
<li>import all the expression data and gene information for the three cell types</li>
<li>import the data that gives the additional labels</li>
<li>combine the data and pivot it to have the expression values in a single column</li>
<li>plot the expression of one gene, <code>Emb</code>, in each cell type for the 3-type categorisation and the 12-type categorisation</li>
<li>conduct a PCA on all the data and plot the first two components using colour/shapes for the 12-type categorisation, the 3-type categorisation and the expression of <code>Emb</code>.</li>
</ul></section><section id="set-up" class="level1"><h1>Set up</h1>
<p>🎬 Open your <code>mice-88H</code> RStudio Project.</p>
<p>🎬 Save <a href="data-raw/er_cell_types.csv">er_cell_types.csv</a>.</p>
<p>🎬 Open a new or existing script</p>
<p>🎬 Load <code>tidyverse</code> <span class="citation" data-cites="tidyverse">(<a href="#ref-tidyverse" role="doc-biblioref">Wickham et al. 2019</a>)</span>.</p>
</section><section id="import-and-combine-data" class="level1"><h1>Import and combine data</h1>
<section id="the-raw-data" class="level2"><h2 class="anchored" data-anchor-id="the-raw-data">The raw data</h2>
<p>Import the data for the LT-HSC, HSPC and the Progenitor cells.</p>
<p>🎬 Import <a href="data-raw/surfaceome_lthsc.csv">surfaceome_lthsc.csv</a>, <a href="data-raw/surfaceome_hspc.csv">surfaceome_hspc.csv</a> and <a href="data-raw/surfaceome_prog.csv">surfaceome_hspc.csv</a></p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 🐭 import the three datasets</span></span>
<span><span class="va">lthsc</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data-raw/surfaceome_lthsc.csv"</span><span class="op">)</span></span>
<span><span class="va">hspc</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data-raw/surfaceome_hspc.csv"</span><span class="op">)</span></span>
<span><span class="va">prog</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data-raw/surfaceome_prog.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note there are 155 LT-HSC + 701 HSPC + 798 Progenitor = 1654 total cells.</p>
<p>We will need to join the three datasets using <code>ensembl_gene_id</code> to match the rows.</p>
<p>🎬 Combine the three datasets by <code>ensembl_gene_id</code> and save the result as <code>cell_expr</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#  combine the three datasets</span></span>
<span><span class="va">cell_expr</span> <span class="op">&lt;-</span> <span class="va">hspc</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">prog</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">lthsc</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="gene-information" class="level2"><h2 class="anchored" data-anchor-id="gene-information">Gene information</h2>
<p>You will likely want the gene information in the dataframe along with the <code>ensembl_gene_id</code>. Rather the connecting again to <a href="https://www.ensembl.org/index.html">Ensembl</a> <span class="citation" data-cites="martin2023 birney2004">(<a href="#ref-martin2023" role="doc-biblioref">Martin et al. 2023</a>; <a href="#ref-birney2004" role="doc-biblioref">Birney et al. 2004</a>)</span> and <a href="https://www.ensembl.org/info/data/biomart/index.html">BioMart</a> <span class="citation" data-cites="smedley2009">(<a href="#ref-smedley2009" role="doc-biblioref">Smedley et al. 2009</a>)</span> using the R package <strong><code>biomaRt</code></strong> <span class="citation" data-cites="biomaRt1 biomaRt2">(<a href="#ref-biomaRt1" role="doc-biblioref">Durinck et al. 2009</a>, <a href="#ref-biomaRt2" role="doc-biblioref">2005</a>)</span>, which means loading more packages, managing conflicts and using more compute, we can use the information we already have in the <a href="results/hspc_prog_results.csv">results/hspc_prog_results.csv</a> file. This works because the same 280 genes are in the three raw data files and this results file.</p>
<p>We need only the <code>ensembl_gene_id</code>, (to join the information), the <code>external_gene_name</code> and the <code>description</code>.</p>
<p>🎬 Import the gene information from <code>results/hspc_prog_results.csv</code> and save as <code>gene_info</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 🐭 import the gene information</span></span>
<span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"results/hspc_prog_results.csv"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">ensembl_gene_id</span>, <span class="va">external_gene_name</span>, <span class="va">description</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>🎬 Join the <code>gene_info</code> to the <code>cell_expr</code> data.</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 🐭 join the gene information to the cell_expr data</span></span>
<span><span class="va">cell_expr</span> <span class="op">&lt;-</span> <span class="va">cell_expr</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">gene_info</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"ensembl_gene_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="type-categorisation-data" class="level2"><h2 class="anchored" data-anchor-id="type-categorisation-data">12-type categorisation data</h2>
<p>The information in the <a href="data-raw/er_cell_types.csv">er_cell_types.csv</a> has 14 columns. The first column gives the cell_id (e.g., <code>HSPC_001</code>) the next 12 columns are named with the cell type or unknown and contain a 0 or 1. These columns are useful if you want to compare (plot or DE), for example, CMP cells vs others. The last column gives the cell type which is useful to include <em>all</em> the labels in a plot.</p>
<p>🎬 Import <a href="data-raw/er_cell_types.csv">er_cell_types.csv</a></p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data-raw/er_cell_types.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>🎬 Check the dataframe has the number of rows and columns you were expecting and that column types and names are as expected.</p>
<p>🎬 Examine the number of cells of each type:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cell</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">cell_type_12</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 2
# Groups:   cell_type_12 [12]
   cell_type_12     n
   &lt;chr&gt;        &lt;int&gt;
 1 CMP            317
 2 GMP            120
 3 LMPP           246
 4 LTHSC          233
 5 LTHSC-ESLAM     38
 6 MEP            354
 7 MPP1            27
 8 MPP2            11
 9 MPP3            60
10 MPP_broad      202
11 STHSC           36
12 unknown         10</code></pre>
</div>
</div>
</section></section><section id="expression-of-a-specific-gene-in-each-cell-type" class="level1"><h1>Expression of a specific gene in each cell type</h1>
<section id="pivot-to-long-form" class="level2"><h2 class="anchored" data-anchor-id="pivot-to-long-form">Pivot to long form</h2>
<p>The <code>cell_expr</code> data has the expression of each gene in each cell. The <code>cell_ids</code> are in columns and the gene are in rows.</p>
<p>We will need to pivot the <code>cell_expr</code> data to have the expression values in a single column with additional columns giving the cell_id and gene.</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 🐭 pivot the cell_expr data</span></span>
<span><span class="va">cell_expr_long</span> <span class="op">&lt;-</span> <span class="va">cell_expr</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">ensembl_gene_id</span>, </span>
<span>                         <span class="va">external_gene_name</span>, </span>
<span>                         <span class="va">description</span><span class="op">)</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"cell_id"</span>, </span>
<span>               values_to <span class="op">=</span> <span class="st">"expression"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that this data frame will have: [280 genes * 1654 cells = 463120] rows.</p>
</section><section id="add-the-two-types-of-categorisation-to-the-expression-data" class="level2"><h2 class="anchored" data-anchor-id="add-the-two-types-of-categorisation-to-the-expression-data">Add the two types of categorisation to the expression data</h2>
<p>We also want to add the two types of categorisation (the 3-type and the 12-types) to the <code>cell_expr_long</code> data. To add the 3-type categorisation we extract the information in the <code>cell_id</code> column into two columns: one with the cell type and one with the number. We did this before in making a <a href="../week-5/workshop.html#stem-cells-2">PCA plot</a></p>
<p>🎬 Extract the cell type and cell number from the <code>cell_id</code> column (keeping the <code>cell_id</code> column):</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cell_expr_long</span> <span class="op">&lt;-</span> <span class="va">cell_expr_long</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">extract</span><span class="op">(</span><span class="va">cell_id</span>, </span>
<span>          remove <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"cell_type_3"</span>, <span class="st">"cell_number"</span><span class="op">)</span>,</span>
<span>          <span class="st">"([a-zA-Z.]{4,6})_([0-9]{3})"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s just check we have the right number of cells in each of the three groups:</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cell_expr_long</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">cell_type_3</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">summarise</span><span class="op">(</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">/</span> <span class="fl">280</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  cell_type_3 `n()/280`
  &lt;chr&gt;           &lt;dbl&gt;
1 HSPC              701
2 LT.HSC            155
3 Prog              798</code></pre>
</div>
</div>
<p>To add the 12-type categorisation we need to join the <code>cell_expr_long</code> data to the <code>cell</code> data on the <code>cell_id</code></p>
<p>🎬 Add the 12-type categorisation to the <code>cell_expr_long</code> data.</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cell_expr_long</span> <span class="op">&lt;-</span> <span class="va">cell_expr_long</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">cell</span> <span class="op">|&gt;</span> <span class="fu">select</span><span class="op">(</span><span class="va">cell_id</span>, <span class="va">cell_type_12</span><span class="op">)</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"cell_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="plot-gene-expression-under-3-type-categorisation" class="level3"><h3 class="anchored" data-anchor-id="plot-gene-expression-under-3-type-categorisation">Plot gene expression under 3-type categorisation</h3>
<p>I have chosen the gene <code>Emb</code> to plot the expression in each cell type. I prefer the use of violin plots and points because multimodal distributions are more obvious than in boxplots</p>
<p>🎬 Plot the expression of <code>Emb</code> in each cell type for the 3-type categorisation.</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 🐭 plot the expression of Emb in each cell type</span></span>
<span><span class="va">cell_expr_long</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">external_gene_name</span> <span class="op">==</span> <span class="st">"Emb"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">cell_type_3</span>, y <span class="op">=</span> <span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_violin</span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.4</span>, alpha <span class="op">=</span> <span class="fl">0.2</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="stem_cell_variant_files/figure-html/unnamed-chunk-13-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="stem_cell_variant_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section><section id="plot-expression-under-12-type-categorisation" class="level3"><h3 class="anchored" data-anchor-id="plot-expression-under-12-type-categorisation">Plot expression under 12-type categorisation</h3>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 🐭 plot the expression of Emb in each cell type</span></span>
<span><span class="va">cell_expr_long</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">external_gene_name</span> <span class="op">==</span> <span class="st">"Emb"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">cell_type_12</span>, y <span class="op">=</span> <span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_violin</span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.2</span>, alpha <span class="op">=</span> <span class="fl">0.2</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="stem_cell_variant_files/figure-html/unnamed-chunk-14-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="stem_cell_variant_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section></section></section><section id="pca" class="level1"><h1>PCA</h1>
<p>We will carry out the the same steps to do the PCA:</p>
<ul>
<li>Select the log<sub>2</sub> transformed normalised counts. You can only use numerical data in PCA.</li>
<li>Transpose our data. We have genes in rows and samples in columns (this is common for gene expression data). However, to treat the genes as variables, PCA expects samples in rows and genes in columns.</li>
<li>Add the gene names as column names in the transposed data</li>
<li>Perform the PCA</li>
<li>Extract the scores on the first two principal components and label the data</li>
<li>Plot the the first two principal components as a scatter plot</li>
</ul>
<p>Refer to the <a href="../week-5/workshop.html#stem-cells-2">workshop for more explanation on PCA</a>.</p>
<p>🎬 Transpose the log<sub>2</sub> transformed normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cell_expr_trans</span> <span class="op">&lt;-</span> <span class="va">cell_expr</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu">starts_with</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HSPC_"</span>, <span class="st">"Prog_"</span>, <span class="st">"LT.HSC_"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>🎬 Set the column names to the gene ids:</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cell_expr_trans</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cell_expr</span><span class="op">$</span><span class="va">ensembl_gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>🎬 Perform PCA on the log<sub>2</sub> transformed normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="va">cell_expr_trans</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span>rank. <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To plot the PC1 against PC2 we will need to extract the PC1 and PC2 “scores” from the PCA object and add labels for the cells. Our PCA object is called <code>pca</code> and the scores are in pca$x. The cells labels will come from the row names of the transformed data.</p>
<p>🎬 Create a dataframe of the PC1 and PC2 scores (in <code>pca$x</code>) and add the cell ids:</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">x</span>,</span>
<span>                           cell_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">cell_expr_trans</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>🎬 Extract the cell type and cell number from the <code>cell_id</code> column (keeping the <code>cell_id</code> column):</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">&lt;-</span> <span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">extract</span><span class="op">(</span><span class="va">cell_id</span>, </span>
<span>          remove <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"cell_type_3"</span>, <span class="st">"cell_number"</span><span class="op">)</span>,</span>
<span>          <span class="st">"([a-zA-Z.]{4,6})_([0-9]{3})"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>🎬 Add the 12-type categorisation to the <code>pca_labelled</code> data.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">&lt;-</span> <span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">cell</span> <span class="op">|&gt;</span> <span class="fu">select</span><span class="op">(</span><span class="va">cell_id</span>, <span class="va">cell_type_12</span><span class="op">)</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"cell_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>🎬 Add the expression of <code>Emb</code> to <code>pca_labelled</code> data. I’ve also renamed the expression column from the filtered <code>cell_expr_long</code> so it is <code>Emb_expr</code> in the <code>pca_labelled</code> data.</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">&lt;-</span> <span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">cell_expr_long</span> <span class="op">|&gt;</span> </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">external_gene_name</span> <span class="op">==</span> <span class="st">"Emb"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>              <span class="fu">select</span><span class="op">(</span><span class="va">cell_id</span>, <span class="st">"Emb_expr"</span> <span class="op">=</span> <span class="va">expression</span><span class="op">)</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"cell_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>🎬 Plot PC2 against PC1 and colour by cell_type_3 type:</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, </span>
<span>             colour <span class="op">=</span> <span class="va">cell_type_3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_colour_viridis_d</span><span class="op">(</span>end <span class="op">=</span> <span class="fl">0.95</span>, begin <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>                         name <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="stem_cell_variant_files/figure-html/unnamed-chunk-22-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="stem_cell_variant_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>🎬 Plot PC2 against PC1 and colour by cell_type_12 type:</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, </span>
<span>             colour <span class="op">=</span> <span class="va">cell_type_12</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_colour_viridis_d</span><span class="op">(</span>name <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="stem_cell_variant_files/figure-html/unnamed-chunk-23-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="stem_cell_variant_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>🎬 Plot PC2 against PC1 and colour by expression of Emb and shape by cell_type_3 type:</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, </span>
<span>             colour <span class="op">=</span> <span class="va">Emb_expr</span>,</span>
<span>             shape <span class="op">=</span> <span class="va">cell_type_3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_colour_viridis_c</span><span class="op">(</span>end <span class="op">=</span> <span class="fl">0.95</span>, begin <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>                         name <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="stem_cell_variant_files/figure-html/unnamed-chunk-24-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="stem_cell_variant_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>🎬 Easier to see as facets?:</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, </span>
<span>             colour <span class="op">=</span> <span class="va">Emb_expr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_colour_viridis_c</span><span class="op">(</span>end <span class="op">=</span> <span class="fl">0.95</span>, begin <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>                         name <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">cell_type_3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><a href="stem_cell_variant_files/figure-html/unnamed-chunk-25-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="stem_cell_variant_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Pages made with R <span class="citation" data-cites="R-core">(<a href="#ref-R-core" role="doc-biblioref">R Core Team 2024</a>)</span>, Quarto <span class="citation" data-cites="Allaire_Quarto_2024">(<a href="#ref-Allaire_Quarto_2024" role="doc-biblioref">Allaire et al. 2024</a>)</span>, <code>knitr</code> <span class="citation" data-cites="knitr1 knitr2 knitr3">(<a href="#ref-knitr1" role="doc-biblioref">Xie 2024</a>, <a href="#ref-knitr2" role="doc-biblioref">2015</a>, <a href="#ref-knitr3" role="doc-biblioref">2014</a>)</span>, <code>kableExtra</code> <span class="citation" data-cites="kableExtra">(<a href="#ref-kableExtra" role="doc-biblioref">Zhu 2021</a>)</span></p>
</section><section id="references" class="level1">


<!-- -->


</section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Allaire_Quarto_2024" class="csl-entry" role="listitem">
Allaire, J. J., Charles Teague, Carlos Scheidegger, Yihui Xie, and Christophe Dervieux. 2024. <span>“<span>Quarto</span>.”</span> <a href="https://doi.org/10.5281/zenodo.5960048">https://doi.org/10.5281/zenodo.5960048</a>.
</div>
<div id="ref-birney2004" class="csl-entry" role="listitem">
Birney, Ewan, T. Daniel Andrews, Paul Bevan, Mario Caccamo, Yuan Chen, Laura Clarke, Guy Coates, et al. 2004. <span>“An Overview of Ensembl.”</span> <em>Genome Research</em> 14 (5): 925–28. <a href="https://doi.org/10.1101/gr.1860604">https://doi.org/10.1101/gr.1860604</a>.
</div>
<div id="ref-biomaRt2" class="csl-entry" role="listitem">
Durinck, Steffen, Yves Moreau, Arek Kasprzyk, Sean Davis, Bart De Moor, Alvis Brazma, and Wolfgang Huber. 2005. <span>“BioMart and Bioconductor: A Powerful Link Between Biological Databases and Microarray Data Analysis.”</span> <em>Bioinformatics</em> 21: 3439–40.
</div>
<div id="ref-biomaRt1" class="csl-entry" role="listitem">
Durinck, Steffen, Paul T. Spellman, Ewan Birney, and Wolfgang Huber. 2009. <span>“Mapping Identifiers for the Integration of Genomic Datasets with the r/Bioconductor Package biomaRt.”</span> <em>Nature Protocols</em> 4: 1184–91.
</div>
<div id="ref-martin2023" class="csl-entry" role="listitem">
Martin, Fergal J, M Ridwan Amode, Alisha Aneja, Olanrewaju Austine-Orimoloye, Andrey&nbsp;G Azov, If Barnes, Arne Becker, et al. 2023. <span>“Ensembl 2023.”</span> <em>Nucleic Acids Research</em> 51 (D1): D933–41. <a href="https://doi.org/10.1093/nar/gkac958">https://doi.org/10.1093/nar/gkac958</a>.
</div>
<div id="ref-nestorowa2016" class="csl-entry" role="listitem">
Nestorowa, Sonia, Fiona K. Hamey, Blanca Pijuan Sala, Evangelia Diamanti, Mairi Shepherd, Elisa Laurenti, Nicola K. Wilson, David G. Kent, and Berthold Göttgens. 2016. <span>“A Single-Cell Resolution Map of Mouse Hematopoietic Stem and Progenitor Cell Differentiation.”</span> <em>Blood</em> 128 (8): e20–31. <a href="https://doi.org/10.1182/blood-2016-05-716480">https://doi.org/10.1182/blood-2016-05-716480</a>.
</div>
<div id="ref-R-core" class="csl-entry" role="listitem">
R Core Team. 2024. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-smedley2009" class="csl-entry" role="listitem">
Smedley, Damian, Syed Haider, Benoit Ballester, Richard Holland, Darin London, Gudmundur Thorisson, and Arek Kasprzyk. 2009. <span>“BioMart <span></span> Biological Queries Made Easy.”</span> <em>BMC Genomics</em> 10 (1): 22. <a href="https://doi.org/10.1186/1471-2164-10-22">https://doi.org/10.1186/1471-2164-10-22</a>.
</div>
<div id="ref-tidyverse" class="csl-entry" role="listitem">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. <span>“Welcome to the <span class="nocase">tidyverse</span>.”</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-knitr3" class="csl-entry" role="listitem">
Xie, Yihui. 2014. <span>“Knitr: A Comprehensive Tool for Reproducible Research in <span>R</span>.”</span> In <em>Implementing Reproducible Computational Research</em>, edited by Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman; Hall/CRC.
</div>
<div id="ref-knitr2" class="csl-entry" role="listitem">
———. 2015. <em>Dynamic Documents with <span>R</span> and Knitr</em>. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-knitr1" class="csl-entry" role="listitem">
———. 2024. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in r</em>. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-kableExtra" class="csl-entry" role="listitem">
Zhu, Hao. 2021. <span>“kableExtra: Construct Complex Table with ’Kable’ and Pipe Syntax.”</span> <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>.
</div>
</div></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("^(?:http:|https:)\/\/3mmarand\.github\.io\/BIO00088H-data");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb26" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Self study"</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="an">subtitle:</span><span class="co"> "Semester 2: Additional 🐭 stem cell labels"</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Emma Rand"</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="an">toc:</span><span class="co"> true</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="an">toc-depth:</span><span class="co"> 4</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a><span class="an">toc-location:</span><span class="co"> right</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a><span class="co">  echo: true</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a><span class="co">  include: true</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a><span class="co">  error: true</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a><span class="an">bibliography:</span><span class="co"> ../../references.bib</span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span><span class="co"> </span></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown: </span></span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a><span class="fu"># Introduction</span></span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>We have analysed the @nestorowa2016 stem cells categorised as LTHSC,</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>HSPC, or Prog. The cell types were determined from flowcytometry</span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>markers. Additional markers allow the cells to be categorised into 12</span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a>other types (including unknown): <span class="in">`CMP`</span>, <span class="in">`GMP`</span>, <span class="in">`LMPP`</span>, <span class="in">`LTHSC`</span>,</span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a><span class="in">`LTHSC-ESLAM`</span>, <span class="in">`MEP`</span>, <span class="in">`MPP_broad`</span>, <span class="in">`MPP1`</span>, <span class="in">`MPP2`</span>, <span class="in">`MPP3`</span>, <span class="in">`STHSC`</span>,</span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a><span class="in">`unknown`</span>. These cell types are mutually exclusive, meaning that a cell</span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a>can only be one of these types.</span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-28"><a href="#cb26-28" aria-hidden="true" tabindex="-1"></a>In this self-study, you will:</span>
<span id="cb26-29"><a href="#cb26-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-30"><a href="#cb26-30" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>import all the expression data and gene information for the three</span>
<span id="cb26-31"><a href="#cb26-31" aria-hidden="true" tabindex="-1"></a>    cell types</span>
<span id="cb26-32"><a href="#cb26-32" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>import the data that gives the additional labels</span>
<span id="cb26-33"><a href="#cb26-33" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>combine the data and pivot it to have the expression values in a</span>
<span id="cb26-34"><a href="#cb26-34" aria-hidden="true" tabindex="-1"></a>    single column</span>
<span id="cb26-35"><a href="#cb26-35" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>plot the expression of one gene, <span class="in">`Emb`</span>, in each cell type for the</span>
<span id="cb26-36"><a href="#cb26-36" aria-hidden="true" tabindex="-1"></a>    3-type categorisation and the 12-type categorisation</span>
<span id="cb26-37"><a href="#cb26-37" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>conduct a PCA on all the data and plot the first two components</span>
<span id="cb26-38"><a href="#cb26-38" aria-hidden="true" tabindex="-1"></a>    using colour/shapes for the 12-type categorisation, the 3-type</span>
<span id="cb26-39"><a href="#cb26-39" aria-hidden="true" tabindex="-1"></a>    categorisation and the expression of <span class="in">`Emb`</span>.</span>
<span id="cb26-40"><a href="#cb26-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-41"><a href="#cb26-41" aria-hidden="true" tabindex="-1"></a><span class="fu"># Set up</span></span>
<span id="cb26-42"><a href="#cb26-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-43"><a href="#cb26-43" aria-hidden="true" tabindex="-1"></a>🎬 Open your <span class="in">`mice-88H`</span> RStudio Project.</span>
<span id="cb26-44"><a href="#cb26-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-45"><a href="#cb26-45" aria-hidden="true" tabindex="-1"></a>🎬 Save <span class="co">[</span><span class="ot">er_cell_types.csv</span><span class="co">](data-raw/er_cell_types.csv)</span>.</span>
<span id="cb26-46"><a href="#cb26-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-47"><a href="#cb26-47" aria-hidden="true" tabindex="-1"></a>🎬 Open a new or existing script</span>
<span id="cb26-48"><a href="#cb26-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-49"><a href="#cb26-49" aria-hidden="true" tabindex="-1"></a>🎬 Load <span class="in">`tidyverse`</span> <span class="co">[</span><span class="ot">@tidyverse</span><span class="co">]</span>.</span>
<span id="cb26-50"><a href="#cb26-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-53"><a href="#cb26-53" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-54"><a href="#cb26-54" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb26-55"><a href="#cb26-55" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb26-56"><a href="#cb26-56" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-57"><a href="#cb26-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-58"><a href="#cb26-58" aria-hidden="true" tabindex="-1"></a><span class="fu"># Import and combine data</span></span>
<span id="cb26-59"><a href="#cb26-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-60"><a href="#cb26-60" aria-hidden="true" tabindex="-1"></a><span class="fu">## The raw data</span></span>
<span id="cb26-61"><a href="#cb26-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-62"><a href="#cb26-62" aria-hidden="true" tabindex="-1"></a>Import the data for the LT-HSC, HSPC and the Progenitor cells.</span>
<span id="cb26-63"><a href="#cb26-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-64"><a href="#cb26-64" aria-hidden="true" tabindex="-1"></a>🎬 Import <span class="co">[</span><span class="ot">surfaceome_lthsc.csv</span><span class="co">](data-raw/surfaceome_lthsc.csv)</span>,</span>
<span id="cb26-65"><a href="#cb26-65" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">surfaceome_hspc.csv</span><span class="co">](data-raw/surfaceome_hspc.csv)</span> and</span>
<span id="cb26-66"><a href="#cb26-66" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">surfaceome_hspc.csv</span><span class="co">](data-raw/surfaceome_prog.csv)</span></span>
<span id="cb26-67"><a href="#cb26-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-70"><a href="#cb26-70" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-71"><a href="#cb26-71" aria-hidden="true" tabindex="-1"></a><span class="co"># 🐭 import the three datasets</span></span>
<span id="cb26-72"><a href="#cb26-72" aria-hidden="true" tabindex="-1"></a>lthsc <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/surfaceome_lthsc.csv"</span>)</span>
<span id="cb26-73"><a href="#cb26-73" aria-hidden="true" tabindex="-1"></a>hspc <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/surfaceome_hspc.csv"</span>)</span>
<span id="cb26-74"><a href="#cb26-74" aria-hidden="true" tabindex="-1"></a>prog <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/surfaceome_prog.csv"</span>)</span>
<span id="cb26-75"><a href="#cb26-75" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-76"><a href="#cb26-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-79"><a href="#cb26-79" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-80"><a href="#cb26-80" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb26-81"><a href="#cb26-81" aria-hidden="true" tabindex="-1"></a><span class="co"># number of columns in each dataset</span></span>
<span id="cb26-82"><a href="#cb26-82" aria-hidden="true" tabindex="-1"></a>nlthsc <span class="ot">&lt;-</span> <span class="fu">ncol</span>(lthsc) <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb26-83"><a href="#cb26-83" aria-hidden="true" tabindex="-1"></a>nhspc <span class="ot">&lt;-</span> <span class="fu">ncol</span>(hspc) <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb26-84"><a href="#cb26-84" aria-hidden="true" tabindex="-1"></a>nprog <span class="ot">&lt;-</span> <span class="fu">ncol</span>(prog) <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb26-85"><a href="#cb26-85" aria-hidden="true" tabindex="-1"></a>ntotal <span class="ot">&lt;-</span> nlthsc <span class="sc">+</span> nhspc <span class="sc">+</span> nprog</span>
<span id="cb26-86"><a href="#cb26-86" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-87"><a href="#cb26-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-88"><a href="#cb26-88" aria-hidden="true" tabindex="-1"></a>Note there are <span class="in">`r nlthsc`</span> LT-HSC + <span class="in">`r nhspc`</span> HSPC + <span class="in">`r nprog`</span> Progenitor</span>
<span id="cb26-89"><a href="#cb26-89" aria-hidden="true" tabindex="-1"></a>= <span class="in">`r ntotal`</span> total cells.</span>
<span id="cb26-90"><a href="#cb26-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-91"><a href="#cb26-91" aria-hidden="true" tabindex="-1"></a>We will need to join the three datasets using <span class="in">`ensembl_gene_id`</span> to match</span>
<span id="cb26-92"><a href="#cb26-92" aria-hidden="true" tabindex="-1"></a>the rows.</span>
<span id="cb26-93"><a href="#cb26-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-94"><a href="#cb26-94" aria-hidden="true" tabindex="-1"></a>🎬 Combine the three datasets by <span class="in">`ensembl_gene_id`</span> and save the result</span>
<span id="cb26-95"><a href="#cb26-95" aria-hidden="true" tabindex="-1"></a>as <span class="in">`cell_expr`</span>.</span>
<span id="cb26-96"><a href="#cb26-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-99"><a href="#cb26-99" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-100"><a href="#cb26-100" aria-hidden="true" tabindex="-1"></a><span class="co">#  combine the three datasets</span></span>
<span id="cb26-101"><a href="#cb26-101" aria-hidden="true" tabindex="-1"></a>cell_expr <span class="ot">&lt;-</span> hspc <span class="sc">|&gt;</span></span>
<span id="cb26-102"><a href="#cb26-102" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(prog, </span>
<span id="cb26-103"><a href="#cb26-103" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"ensembl_gene_id"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb26-104"><a href="#cb26-104" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(lthsc, </span>
<span id="cb26-105"><a href="#cb26-105" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"ensembl_gene_id"</span>)</span>
<span id="cb26-106"><a href="#cb26-106" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-107"><a href="#cb26-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-108"><a href="#cb26-108" aria-hidden="true" tabindex="-1"></a><span class="fu">## Gene information</span></span>
<span id="cb26-109"><a href="#cb26-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-110"><a href="#cb26-110" aria-hidden="true" tabindex="-1"></a>You will likely want the gene information in the dataframe along with</span>
<span id="cb26-111"><a href="#cb26-111" aria-hidden="true" tabindex="-1"></a>the <span class="in">`ensembl_gene_id`</span>. Rather the connecting again to</span>
<span id="cb26-112"><a href="#cb26-112" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Ensembl</span><span class="co">](https://www.ensembl.org/index.html)</span> <span class="co">[</span><span class="ot">@martin2023; @birney2004</span><span class="co">]</span></span>
<span id="cb26-113"><a href="#cb26-113" aria-hidden="true" tabindex="-1"></a>and <span class="co">[</span><span class="ot">BioMart</span><span class="co">](https://www.ensembl.org/info/data/biomart/index.html)</span></span>
<span id="cb26-114"><a href="#cb26-114" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@smedley2009</span><span class="co">]</span> using the R package **`biomaRt`** <span class="co">[</span><span class="ot">@biomaRt1; @biomaRt2</span><span class="co">]</span>,</span>
<span id="cb26-115"><a href="#cb26-115" aria-hidden="true" tabindex="-1"></a>which means loading more packages, managing conflicts and using more</span>
<span id="cb26-116"><a href="#cb26-116" aria-hidden="true" tabindex="-1"></a>compute, we can use the information we already have in the</span>
<span id="cb26-117"><a href="#cb26-117" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">results/hspc_prog_results.csv</span><span class="co">](results/hspc_prog_results.csv)</span> file.</span>
<span id="cb26-118"><a href="#cb26-118" aria-hidden="true" tabindex="-1"></a>This works because the same 280 genes are in the three raw data files</span>
<span id="cb26-119"><a href="#cb26-119" aria-hidden="true" tabindex="-1"></a>and this results file.</span>
<span id="cb26-120"><a href="#cb26-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-121"><a href="#cb26-121" aria-hidden="true" tabindex="-1"></a>We need only the <span class="in">`ensembl_gene_id`</span>, (to join the information), the</span>
<span id="cb26-122"><a href="#cb26-122" aria-hidden="true" tabindex="-1"></a><span class="in">`external_gene_name`</span> and the <span class="in">`description`</span>.</span>
<span id="cb26-123"><a href="#cb26-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-124"><a href="#cb26-124" aria-hidden="true" tabindex="-1"></a>🎬 Import the gene information from <span class="in">`results/hspc_prog_results.csv`</span> and</span>
<span id="cb26-125"><a href="#cb26-125" aria-hidden="true" tabindex="-1"></a>save as <span class="in">`gene_info`</span>.</span>
<span id="cb26-126"><a href="#cb26-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-129"><a href="#cb26-129" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-130"><a href="#cb26-130" aria-hidden="true" tabindex="-1"></a><span class="co"># 🐭 import the gene information</span></span>
<span id="cb26-131"><a href="#cb26-131" aria-hidden="true" tabindex="-1"></a>gene_info <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"results/hspc_prog_results.csv"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb26-132"><a href="#cb26-132" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(ensembl_gene_id, external_gene_name, description)</span>
<span id="cb26-133"><a href="#cb26-133" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-134"><a href="#cb26-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-135"><a href="#cb26-135" aria-hidden="true" tabindex="-1"></a>🎬 Join the <span class="in">`gene_info`</span> to the <span class="in">`cell_expr`</span> data.</span>
<span id="cb26-136"><a href="#cb26-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-139"><a href="#cb26-139" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-140"><a href="#cb26-140" aria-hidden="true" tabindex="-1"></a><span class="co"># 🐭 join the gene information to the cell_expr data</span></span>
<span id="cb26-141"><a href="#cb26-141" aria-hidden="true" tabindex="-1"></a>cell_expr <span class="ot">&lt;-</span> cell_expr <span class="sc">|&gt;</span> </span>
<span id="cb26-142"><a href="#cb26-142" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(gene_info, </span>
<span id="cb26-143"><a href="#cb26-143" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"ensembl_gene_id"</span>)</span>
<span id="cb26-144"><a href="#cb26-144" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-145"><a href="#cb26-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-146"><a href="#cb26-146" aria-hidden="true" tabindex="-1"></a><span class="fu">## 12-type categorisation data</span></span>
<span id="cb26-147"><a href="#cb26-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-148"><a href="#cb26-148" aria-hidden="true" tabindex="-1"></a>The information in the <span class="co">[</span><span class="ot">er_cell_types.csv</span><span class="co">](data-raw/er_cell_types.csv)</span></span>
<span id="cb26-149"><a href="#cb26-149" aria-hidden="true" tabindex="-1"></a>has 14 columns. The first column gives the cell_id (e.g., <span class="in">`HSPC_001`</span>)</span>
<span id="cb26-150"><a href="#cb26-150" aria-hidden="true" tabindex="-1"></a>the next 12 columns are named with the cell type or unknown and contain</span>
<span id="cb26-151"><a href="#cb26-151" aria-hidden="true" tabindex="-1"></a>a 0 or 1. These columns are useful if you want to compare (plot or DE),</span>
<span id="cb26-152"><a href="#cb26-152" aria-hidden="true" tabindex="-1"></a>for example, CMP cells vs others. The last column gives the cell type </span>
<span id="cb26-153"><a href="#cb26-153" aria-hidden="true" tabindex="-1"></a>which is useful to include *all* the labels in a plot.</span>
<span id="cb26-154"><a href="#cb26-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-155"><a href="#cb26-155" aria-hidden="true" tabindex="-1"></a>🎬 Import <span class="co">[</span><span class="ot">er_cell_types.csv</span><span class="co">](data-raw/er_cell_types.csv)</span></span>
<span id="cb26-156"><a href="#cb26-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-159"><a href="#cb26-159" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-160"><a href="#cb26-160" aria-hidden="true" tabindex="-1"></a>cell <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data-raw/er_cell_types.csv"</span>)</span>
<span id="cb26-161"><a href="#cb26-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-162"><a href="#cb26-162" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-163"><a href="#cb26-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-164"><a href="#cb26-164" aria-hidden="true" tabindex="-1"></a>🎬 Check the dataframe has the number of rows and columns you were</span>
<span id="cb26-165"><a href="#cb26-165" aria-hidden="true" tabindex="-1"></a>expecting and that column types and names are as expected.</span>
<span id="cb26-166"><a href="#cb26-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-167"><a href="#cb26-167" aria-hidden="true" tabindex="-1"></a>🎬 Examine the number of cells of each type:</span>
<span id="cb26-168"><a href="#cb26-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-171"><a href="#cb26-171" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-172"><a href="#cb26-172" aria-hidden="true" tabindex="-1"></a>cell <span class="sc">|&gt;</span> </span>
<span id="cb26-173"><a href="#cb26-173" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cell_type_12) <span class="sc">|&gt;</span></span>
<span id="cb26-174"><a href="#cb26-174" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>()</span>
<span id="cb26-175"><a href="#cb26-175" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-176"><a href="#cb26-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-177"><a href="#cb26-177" aria-hidden="true" tabindex="-1"></a><span class="fu"># Expression of a specific gene in each cell type</span></span>
<span id="cb26-178"><a href="#cb26-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-179"><a href="#cb26-179" aria-hidden="true" tabindex="-1"></a><span class="fu">## Pivot to long form</span></span>
<span id="cb26-180"><a href="#cb26-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-181"><a href="#cb26-181" aria-hidden="true" tabindex="-1"></a>The <span class="in">`cell_expr`</span> data has the expression of each gene in each cell. The</span>
<span id="cb26-182"><a href="#cb26-182" aria-hidden="true" tabindex="-1"></a><span class="in">`cell_ids`</span> are in columns and the gene are in rows.</span>
<span id="cb26-183"><a href="#cb26-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-184"><a href="#cb26-184" aria-hidden="true" tabindex="-1"></a>We will need to pivot the <span class="in">`cell_expr`</span> data to have the expression values</span>
<span id="cb26-185"><a href="#cb26-185" aria-hidden="true" tabindex="-1"></a>in a single column with additional columns giving the cell_id and gene.</span>
<span id="cb26-186"><a href="#cb26-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-189"><a href="#cb26-189" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-190"><a href="#cb26-190" aria-hidden="true" tabindex="-1"></a><span class="co"># 🐭 pivot the cell_expr data</span></span>
<span id="cb26-191"><a href="#cb26-191" aria-hidden="true" tabindex="-1"></a>cell_expr_long <span class="ot">&lt;-</span> cell_expr <span class="sc">|&gt;</span> </span>
<span id="cb26-192"><a href="#cb26-192" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(ensembl_gene_id, </span>
<span id="cb26-193"><a href="#cb26-193" aria-hidden="true" tabindex="-1"></a>                         external_gene_name, </span>
<span id="cb26-194"><a href="#cb26-194" aria-hidden="true" tabindex="-1"></a>                         description),</span>
<span id="cb26-195"><a href="#cb26-195" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"cell_id"</span>, </span>
<span id="cb26-196"><a href="#cb26-196" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"expression"</span>)</span>
<span id="cb26-197"><a href="#cb26-197" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-198"><a href="#cb26-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-199"><a href="#cb26-199" aria-hidden="true" tabindex="-1"></a>Note that this data frame will have:</span>
<span id="cb26-200"><a href="#cb26-200" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">280 genes * 1654 cells =  463120</span><span class="co">]</span> rows.</span>
<span id="cb26-201"><a href="#cb26-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-202"><a href="#cb26-202" aria-hidden="true" tabindex="-1"></a><span class="fu">## Add the two types of categorisation to the expression data</span></span>
<span id="cb26-203"><a href="#cb26-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-204"><a href="#cb26-204" aria-hidden="true" tabindex="-1"></a>We also want to add the two types of categorisation (the 3-type and the</span>
<span id="cb26-205"><a href="#cb26-205" aria-hidden="true" tabindex="-1"></a>12-types) to the <span class="in">`cell_expr_long`</span> data. To add the 3-type categorisation</span>
<span id="cb26-206"><a href="#cb26-206" aria-hidden="true" tabindex="-1"></a>we extract the information in the <span class="in">`cell_id`</span> column into two columns: one</span>
<span id="cb26-207"><a href="#cb26-207" aria-hidden="true" tabindex="-1"></a>with the cell type and one with the number. We did this before in making</span>
<span id="cb26-208"><a href="#cb26-208" aria-hidden="true" tabindex="-1"></a>a <span class="co">[</span><span class="ot">PCA plot</span><span class="co">](../week-5/workshop.html#stem-cells-2)</span></span>
<span id="cb26-209"><a href="#cb26-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-210"><a href="#cb26-210" aria-hidden="true" tabindex="-1"></a>🎬 Extract the cell type and cell number from the <span class="in">`cell_id`</span> column</span>
<span id="cb26-211"><a href="#cb26-211" aria-hidden="true" tabindex="-1"></a>(keeping the <span class="in">`cell_id`</span> column):</span>
<span id="cb26-212"><a href="#cb26-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-215"><a href="#cb26-215" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-216"><a href="#cb26-216" aria-hidden="true" tabindex="-1"></a>cell_expr_long <span class="ot">&lt;-</span> cell_expr_long <span class="sc">|&gt;</span> </span>
<span id="cb26-217"><a href="#cb26-217" aria-hidden="true" tabindex="-1"></a>  <span class="fu">extract</span>(cell_id, </span>
<span id="cb26-218"><a href="#cb26-218" aria-hidden="true" tabindex="-1"></a>          <span class="at">remove =</span> <span class="cn">FALSE</span>,</span>
<span id="cb26-219"><a href="#cb26-219" aria-hidden="true" tabindex="-1"></a>          <span class="fu">c</span>(<span class="st">"cell_type_3"</span>, <span class="st">"cell_number"</span>),</span>
<span id="cb26-220"><a href="#cb26-220" aria-hidden="true" tabindex="-1"></a>          <span class="st">"([a-zA-Z.]{4,6})_([0-9]{3})"</span>)</span>
<span id="cb26-221"><a href="#cb26-221" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-222"><a href="#cb26-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-223"><a href="#cb26-223" aria-hidden="true" tabindex="-1"></a>Let's just check we have the right number of cells in each of the three</span>
<span id="cb26-224"><a href="#cb26-224" aria-hidden="true" tabindex="-1"></a>groups:</span>
<span id="cb26-225"><a href="#cb26-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-228"><a href="#cb26-228" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-229"><a href="#cb26-229" aria-hidden="true" tabindex="-1"></a>cell_expr_long <span class="sc">|&gt;</span> </span>
<span id="cb26-230"><a href="#cb26-230" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cell_type_3) <span class="sc">|&gt;</span> </span>
<span id="cb26-231"><a href="#cb26-231" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">n</span>()<span class="sc">/</span> <span class="dv">280</span>)</span>
<span id="cb26-232"><a href="#cb26-232" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-233"><a href="#cb26-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-234"><a href="#cb26-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-235"><a href="#cb26-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-236"><a href="#cb26-236" aria-hidden="true" tabindex="-1"></a>To add the 12-type categorisation we need to join the <span class="in">`cell_expr_long`</span></span>
<span id="cb26-237"><a href="#cb26-237" aria-hidden="true" tabindex="-1"></a>data to the <span class="in">`cell`</span> data on the <span class="in">`cell_id`</span></span>
<span id="cb26-238"><a href="#cb26-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-239"><a href="#cb26-239" aria-hidden="true" tabindex="-1"></a>🎬 Add the 12-type categorisation to the <span class="in">`cell_expr_long`</span> data.</span>
<span id="cb26-240"><a href="#cb26-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-243"><a href="#cb26-243" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-244"><a href="#cb26-244" aria-hidden="true" tabindex="-1"></a>cell_expr_long <span class="ot">&lt;-</span> cell_expr_long <span class="sc">|&gt;</span> </span>
<span id="cb26-245"><a href="#cb26-245" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(cell <span class="sc">|&gt;</span> <span class="fu">select</span>(cell_id, cell_type_12), </span>
<span id="cb26-246"><a href="#cb26-246" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"cell_id"</span>)</span>
<span id="cb26-247"><a href="#cb26-247" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-248"><a href="#cb26-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-249"><a href="#cb26-249" aria-hidden="true" tabindex="-1"></a><span class="fu">### Plot gene expression under 3-type categorisation</span></span>
<span id="cb26-250"><a href="#cb26-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-251"><a href="#cb26-251" aria-hidden="true" tabindex="-1"></a>I have chosen the gene <span class="in">`Emb`</span> to plot the expression in each cell type. I</span>
<span id="cb26-252"><a href="#cb26-252" aria-hidden="true" tabindex="-1"></a>prefer the use of violin plots and points because multimodal</span>
<span id="cb26-253"><a href="#cb26-253" aria-hidden="true" tabindex="-1"></a>distributions are more obvious than in boxplots</span>
<span id="cb26-254"><a href="#cb26-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-255"><a href="#cb26-255" aria-hidden="true" tabindex="-1"></a>🎬 Plot the expression of <span class="in">`Emb`</span> in each cell type for the 3-type</span>
<span id="cb26-256"><a href="#cb26-256" aria-hidden="true" tabindex="-1"></a>categorisation.</span>
<span id="cb26-257"><a href="#cb26-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-260"><a href="#cb26-260" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-261"><a href="#cb26-261" aria-hidden="true" tabindex="-1"></a><span class="co"># 🐭 plot the expression of Emb in each cell type</span></span>
<span id="cb26-262"><a href="#cb26-262" aria-hidden="true" tabindex="-1"></a>cell_expr_long <span class="sc">|&gt;</span> </span>
<span id="cb26-263"><a href="#cb26-263" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(external_gene_name <span class="sc">==</span> <span class="st">"Emb"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb26-264"><a href="#cb26-264" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> cell_type_3, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb26-265"><a href="#cb26-265" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb26-266"><a href="#cb26-266" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.4</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">pch =</span> <span class="dv">16</span>) <span class="sc">+</span></span>
<span id="cb26-267"><a href="#cb26-267" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb26-268"><a href="#cb26-268" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-269"><a href="#cb26-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-270"><a href="#cb26-270" aria-hidden="true" tabindex="-1"></a><span class="fu">### Plot expression under 12-type categorisation</span></span>
<span id="cb26-271"><a href="#cb26-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-274"><a href="#cb26-274" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-275"><a href="#cb26-275" aria-hidden="true" tabindex="-1"></a><span class="co"># 🐭 plot the expression of Emb in each cell type</span></span>
<span id="cb26-276"><a href="#cb26-276" aria-hidden="true" tabindex="-1"></a>cell_expr_long <span class="sc">|&gt;</span> </span>
<span id="cb26-277"><a href="#cb26-277" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(external_gene_name <span class="sc">==</span> <span class="st">"Emb"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb26-278"><a href="#cb26-278" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> cell_type_12, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb26-279"><a href="#cb26-279" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb26-280"><a href="#cb26-280" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.2</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">pch =</span> <span class="dv">16</span>) <span class="sc">+</span></span>
<span id="cb26-281"><a href="#cb26-281" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb26-282"><a href="#cb26-282" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-283"><a href="#cb26-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-284"><a href="#cb26-284" aria-hidden="true" tabindex="-1"></a><span class="fu"># PCA</span></span>
<span id="cb26-285"><a href="#cb26-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-286"><a href="#cb26-286" aria-hidden="true" tabindex="-1"></a>We will carry out the the same steps to do the PCA:</span>
<span id="cb26-287"><a href="#cb26-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-288"><a href="#cb26-288" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Select the log~2~ transformed normalised counts. You can only use</span>
<span id="cb26-289"><a href="#cb26-289" aria-hidden="true" tabindex="-1"></a>    numerical data in PCA.</span>
<span id="cb26-290"><a href="#cb26-290" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Transpose our data. We have genes in rows and samples in columns</span>
<span id="cb26-291"><a href="#cb26-291" aria-hidden="true" tabindex="-1"></a>    (this is common for gene expression data). However, to treat the</span>
<span id="cb26-292"><a href="#cb26-292" aria-hidden="true" tabindex="-1"></a>    genes as variables, PCA expects samples in rows and genes in</span>
<span id="cb26-293"><a href="#cb26-293" aria-hidden="true" tabindex="-1"></a>    columns.</span>
<span id="cb26-294"><a href="#cb26-294" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Add the gene names as column names in the transposed data</span>
<span id="cb26-295"><a href="#cb26-295" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Perform the PCA</span>
<span id="cb26-296"><a href="#cb26-296" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Extract the scores on the first two principal components and label</span>
<span id="cb26-297"><a href="#cb26-297" aria-hidden="true" tabindex="-1"></a>    the data</span>
<span id="cb26-298"><a href="#cb26-298" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Plot the the first two principal components as a scatter plot</span>
<span id="cb26-299"><a href="#cb26-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-300"><a href="#cb26-300" aria-hidden="true" tabindex="-1"></a>Refer to the [workshop for more explanation on</span>
<span id="cb26-301"><a href="#cb26-301" aria-hidden="true" tabindex="-1"></a>PCA](../week-5/workshop.html#stem-cells-2).</span>
<span id="cb26-302"><a href="#cb26-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-303"><a href="#cb26-303" aria-hidden="true" tabindex="-1"></a>🎬 Transpose the log~2~ transformed normalised counts:</span>
<span id="cb26-304"><a href="#cb26-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-307"><a href="#cb26-307" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-308"><a href="#cb26-308" aria-hidden="true" tabindex="-1"></a>cell_expr_trans <span class="ot">&lt;-</span> cell_expr <span class="sc">|&gt;</span> </span>
<span id="cb26-309"><a href="#cb26-309" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">starts_with</span>(<span class="fu">c</span>(<span class="st">"HSPC_"</span>, <span class="st">"Prog_"</span>, <span class="st">"LT.HSC_"</span>))) <span class="sc">|&gt;</span></span>
<span id="cb26-310"><a href="#cb26-310" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">|&gt;</span> </span>
<span id="cb26-311"><a href="#cb26-311" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>()</span>
<span id="cb26-312"><a href="#cb26-312" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-313"><a href="#cb26-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-314"><a href="#cb26-314" aria-hidden="true" tabindex="-1"></a>🎬 Set the column names to the gene ids:</span>
<span id="cb26-315"><a href="#cb26-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-318"><a href="#cb26-318" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-319"><a href="#cb26-319" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cell_expr_trans) <span class="ot">&lt;-</span> cell_expr<span class="sc">$</span>ensembl_gene_id</span>
<span id="cb26-320"><a href="#cb26-320" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-321"><a href="#cb26-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-322"><a href="#cb26-322" aria-hidden="true" tabindex="-1"></a>🎬 Perform PCA on the log~2~ transformed normalised counts:</span>
<span id="cb26-323"><a href="#cb26-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-326"><a href="#cb26-326" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-327"><a href="#cb26-327" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> cell_expr_trans <span class="sc">|&gt;</span></span>
<span id="cb26-328"><a href="#cb26-328" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prcomp</span>(<span class="at">rank. =</span> <span class="dv">8</span>)</span>
<span id="cb26-329"><a href="#cb26-329" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-330"><a href="#cb26-330" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-331"><a href="#cb26-331" aria-hidden="true" tabindex="-1"></a>To plot the PC1 against PC2 we will need to extract the PC1 and PC2</span>
<span id="cb26-332"><a href="#cb26-332" aria-hidden="true" tabindex="-1"></a>"scores" from the PCA object and add labels for the cells. Our PCA</span>
<span id="cb26-333"><a href="#cb26-333" aria-hidden="true" tabindex="-1"></a>object is called <span class="in">`pca`</span> and the scores are in pca\$x. The cells labels</span>
<span id="cb26-334"><a href="#cb26-334" aria-hidden="true" tabindex="-1"></a>will come from the row names of the transformed data.</span>
<span id="cb26-335"><a href="#cb26-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-336"><a href="#cb26-336" aria-hidden="true" tabindex="-1"></a>🎬 Create a dataframe of the PC1 and PC2 scores (in <span class="in">`pca$x`</span>) and add the</span>
<span id="cb26-337"><a href="#cb26-337" aria-hidden="true" tabindex="-1"></a>cell ids:</span>
<span id="cb26-338"><a href="#cb26-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-341"><a href="#cb26-341" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-342"><a href="#cb26-342" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(pca<span class="sc">$</span>x,</span>
<span id="cb26-343"><a href="#cb26-343" aria-hidden="true" tabindex="-1"></a>                           <span class="at">cell_id =</span> <span class="fu">row.names</span>(cell_expr_trans))</span>
<span id="cb26-344"><a href="#cb26-344" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-345"><a href="#cb26-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-346"><a href="#cb26-346" aria-hidden="true" tabindex="-1"></a>🎬 Extract the cell type and cell number from the <span class="in">`cell_id`</span> column</span>
<span id="cb26-347"><a href="#cb26-347" aria-hidden="true" tabindex="-1"></a>(keeping the <span class="in">`cell_id`</span> column):</span>
<span id="cb26-348"><a href="#cb26-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-351"><a href="#cb26-351" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-352"><a href="#cb26-352" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="ot">&lt;-</span> pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb26-353"><a href="#cb26-353" aria-hidden="true" tabindex="-1"></a>  <span class="fu">extract</span>(cell_id, </span>
<span id="cb26-354"><a href="#cb26-354" aria-hidden="true" tabindex="-1"></a>          <span class="at">remove =</span> <span class="cn">FALSE</span>,</span>
<span id="cb26-355"><a href="#cb26-355" aria-hidden="true" tabindex="-1"></a>          <span class="fu">c</span>(<span class="st">"cell_type_3"</span>, <span class="st">"cell_number"</span>),</span>
<span id="cb26-356"><a href="#cb26-356" aria-hidden="true" tabindex="-1"></a>          <span class="st">"([a-zA-Z.]{4,6})_([0-9]{3})"</span>)</span>
<span id="cb26-357"><a href="#cb26-357" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-358"><a href="#cb26-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-359"><a href="#cb26-359" aria-hidden="true" tabindex="-1"></a>🎬 Add the 12-type categorisation to the <span class="in">`pca_labelled`</span> data.</span>
<span id="cb26-360"><a href="#cb26-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-363"><a href="#cb26-363" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-364"><a href="#cb26-364" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="ot">&lt;-</span> pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb26-365"><a href="#cb26-365" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(cell <span class="sc">|&gt;</span> <span class="fu">select</span>(cell_id, cell_type_12), </span>
<span id="cb26-366"><a href="#cb26-366" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"cell_id"</span>)</span>
<span id="cb26-367"><a href="#cb26-367" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-368"><a href="#cb26-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-369"><a href="#cb26-369" aria-hidden="true" tabindex="-1"></a>🎬 Add the expression of <span class="in">`Emb`</span> to <span class="in">`pca_labelled`</span> data. I've also renamed</span>
<span id="cb26-370"><a href="#cb26-370" aria-hidden="true" tabindex="-1"></a>the expression column from the filtered <span class="in">`cell_expr_long`</span> so it is</span>
<span id="cb26-371"><a href="#cb26-371" aria-hidden="true" tabindex="-1"></a><span class="in">`Emb_expr`</span> in the <span class="in">`pca_labelled`</span> data.</span>
<span id="cb26-372"><a href="#cb26-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-375"><a href="#cb26-375" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-376"><a href="#cb26-376" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="ot">&lt;-</span> pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb26-377"><a href="#cb26-377" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(cell_expr_long <span class="sc">|&gt;</span> </span>
<span id="cb26-378"><a href="#cb26-378" aria-hidden="true" tabindex="-1"></a>              <span class="fu">filter</span>(external_gene_name <span class="sc">==</span> <span class="st">"Emb"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb26-379"><a href="#cb26-379" aria-hidden="true" tabindex="-1"></a>              <span class="fu">select</span>(cell_id, <span class="st">"Emb_expr"</span> <span class="ot">=</span> expression), </span>
<span id="cb26-380"><a href="#cb26-380" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"cell_id"</span>)</span>
<span id="cb26-381"><a href="#cb26-381" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-382"><a href="#cb26-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-383"><a href="#cb26-383" aria-hidden="true" tabindex="-1"></a>🎬 Plot PC2 against PC1 and colour by cell_type_3 type:</span>
<span id="cb26-384"><a href="#cb26-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-387"><a href="#cb26-387" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-388"><a href="#cb26-388" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb26-389"><a href="#cb26-389" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, </span>
<span id="cb26-390"><a href="#cb26-390" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> cell_type_3)) <span class="sc">+</span></span>
<span id="cb26-391"><a href="#cb26-391" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb26-392"><a href="#cb26-392" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_viridis_d</span>(<span class="at">end =</span> <span class="fl">0.95</span>, <span class="at">begin =</span> <span class="fl">0.15</span>,</span>
<span id="cb26-393"><a href="#cb26-393" aria-hidden="true" tabindex="-1"></a>                         <span class="at">name =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb26-394"><a href="#cb26-394" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb26-395"><a href="#cb26-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-396"><a href="#cb26-396" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-397"><a href="#cb26-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-398"><a href="#cb26-398" aria-hidden="true" tabindex="-1"></a>🎬 Plot PC2 against PC1 and colour by cell_type_12 type:</span>
<span id="cb26-399"><a href="#cb26-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-402"><a href="#cb26-402" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-403"><a href="#cb26-403" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb26-404"><a href="#cb26-404" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, </span>
<span id="cb26-405"><a href="#cb26-405" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> cell_type_12)) <span class="sc">+</span></span>
<span id="cb26-406"><a href="#cb26-406" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb26-407"><a href="#cb26-407" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_viridis_d</span>(<span class="at">name =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb26-408"><a href="#cb26-408" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb26-409"><a href="#cb26-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-410"><a href="#cb26-410" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-411"><a href="#cb26-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-412"><a href="#cb26-412" aria-hidden="true" tabindex="-1"></a>🎬 Plot PC2 against PC1 and colour by expression of Emb and shape by</span>
<span id="cb26-413"><a href="#cb26-413" aria-hidden="true" tabindex="-1"></a>cell_type_3 type:</span>
<span id="cb26-414"><a href="#cb26-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-417"><a href="#cb26-417" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-418"><a href="#cb26-418" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb26-419"><a href="#cb26-419" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, </span>
<span id="cb26-420"><a href="#cb26-420" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> Emb_expr,</span>
<span id="cb26-421"><a href="#cb26-421" aria-hidden="true" tabindex="-1"></a>             <span class="at">shape =</span> cell_type_3)) <span class="sc">+</span></span>
<span id="cb26-422"><a href="#cb26-422" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-423"><a href="#cb26-423" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_viridis_c</span>(<span class="at">end =</span> <span class="fl">0.95</span>, <span class="at">begin =</span> <span class="fl">0.15</span>,</span>
<span id="cb26-424"><a href="#cb26-424" aria-hidden="true" tabindex="-1"></a>                         <span class="at">name =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb26-425"><a href="#cb26-425" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb26-426"><a href="#cb26-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-427"><a href="#cb26-427" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-428"><a href="#cb26-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-429"><a href="#cb26-429" aria-hidden="true" tabindex="-1"></a>🎬 Easier to see as facets?:</span>
<span id="cb26-430"><a href="#cb26-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-433"><a href="#cb26-433" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-434"><a href="#cb26-434" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb26-435"><a href="#cb26-435" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, </span>
<span id="cb26-436"><a href="#cb26-436" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> Emb_expr)) <span class="sc">+</span></span>
<span id="cb26-437"><a href="#cb26-437" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-438"><a href="#cb26-438" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_viridis_c</span>(<span class="at">end =</span> <span class="fl">0.95</span>, <span class="at">begin =</span> <span class="fl">0.15</span>,</span>
<span id="cb26-439"><a href="#cb26-439" aria-hidden="true" tabindex="-1"></a>                         <span class="at">name =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb26-440"><a href="#cb26-440" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>cell_type_3)</span>
<span id="cb26-441"><a href="#cb26-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-442"><a href="#cb26-442" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-443"><a href="#cb26-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-444"><a href="#cb26-444" aria-hidden="true" tabindex="-1"></a>Pages made with R <span class="co">[</span><span class="ot">@R-core</span><span class="co">]</span>, Quarto <span class="co">[</span><span class="ot">@Allaire_Quarto_2024</span><span class="co">]</span>, <span class="in">`knitr`</span></span>
<span id="cb26-445"><a href="#cb26-445" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@knitr1; @knitr2; @knitr3</span><span class="co">]</span>, <span class="in">`kableExtra`</span> <span class="co">[</span><span class="ot">@kableExtra</span><span class="co">]</span></span>
<span id="cb26-446"><a href="#cb26-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-447"><a href="#cb26-447" aria-hidden="true" tabindex="-1"></a><span class="fu"># References</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Rand E (2024). Data Analysis for Group Project. <a href="https://3mmarand.github.io/BIO00088H-data/" class="uri">https://3mmarand.github.io/BIO00088H-data/</a>.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/3mmaRand/BIO00088H-data/edit/main/transcriptomics/semester-2/stem_cell_variant.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li></ul></div></div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
    <a class="nav-link" href="https://github.com/3mmaRand">
<p>Emma on GitHub</p>
</a>
  </li>  
</ul>
</div>
  </div>
</footer><script>
document.addEventListener("DOMContentLoaded", function() {
  // grab all links under “Other Links” in the right-hand TOC
  const links = document.querySelectorAll('.quarto-other-links ul li a');
  links.forEach(function(a) {
    // find the one whose text is exactly “Make a suggestion”
    if (a.textContent.trim() === 'Make a suggestion') {
      const pageUrl   = encodeURIComponent(window.location.href);
      const pageTitle = encodeURIComponent(document.title);
      a.href = 'https://github.com/3mmaRand/BIO00088H-data/issues/new'
             + '?title=' + pageTitle
             + '&body='  + pageUrl;
      a.target = '_blank';
    }
  });
});
</script><script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>


</body></html>