---
title: "Independent Study to prepare for workshop"
subtitle: "Transcriptomics 3: Visualising"
author: "Emma Rand"
format: 
  revealjs:
    footer: "üîó [About Transcriptomics 3: Visualising](https://3mmarand.github.io/BIO00088H-data/transcriptomics/week-5/overview.html)"
    slide-number: true
    chalkboard: true
    code-link: true
    code-line-numbers: false
bibliography: ../../references.bib
editor: 
  markdown: 
    wrap: 72
---

```{r}
#| include: false
library(tidyverse)

```

## Overview

In these slides we will:

::: incremental
-   Check where you are

-   learn some concepts used omics visualisation

    -   Principle Component Analysis (PCA)
    -   Volcano plots
    -   Heatmaps

-   Find out what packages to install before the workshop
:::

# Where should you be?

## What we did in Transcriptomics 2: Statistical Analysis

::: incremental
-   carried out differential expression analysis

-   found genes not expressed at all, or expressed in one group only

-   Saved results files
:::

## Where should you be?

After the [Transcriptomics 2: üëã Statistical Analysis
Workshop](../week-4/workshop.html) including:

-   [ü§ó Look after future
    you!](../week-4/workshop.html#look-after-future-you) and

-   the [Independent Study to
    consolidate](../week-4/study_after_workshop.html), you should have:

## üê∏ Frog development

::: {style="font-size: 70%;"}
-   An RStudio Project called `frogs-88H` which contains:
    -   data-raw/ with `xlaevis_counts_S14.csv`, `xlaevis_counts_S20.csv`, 
        `xlaevis_counts_S30.csv` 
    -   data-processed with `s30_filtered.csv` and equivalent for S14 
        *OR* S20
    -   results/ with `s30_fgf_only.csv` (there were no control only 
        genes in s30), `s30_results.csv` and equivalent for S14 *OR* S20)   
    -   Two scripts called `cont-fgf-s30.R` and either `cont-fgf-s20.R`
        *OR* `cont-fgf-s14.R`
:::

Files should be organised into folders. Code should well commented and
easy to read.


## üéÑ Arabidopisis

## üíâ Leishmania mexicana


## üê≠ Stem cells

::: {style="font-size: 70%;"}
-   An RStudio Project called `mice-88H` which contains
    -   Raw data (hspc, prog, lthsc)
    -   Processed data (`hspc_summary_gene.csv`,
        `hspc_summary_samp.csv`, `prog_summary_gene.csv`,
        `prog_summary_samp.csv`, `lthsc_summary_gene.csv`,
        `lthsc_summary_samp.csv`)
-   Results files (`prog_hspc_results.csv` and an equivalent for lthsc
    vs prog or hspc vs lthsc)
-   Two scripts called `hspc-prog.R` and either `hspc-lthsc.R` *OR*
    `prog-lthsc.R`
:::

Files should be organised into folders. Code should well commented and
easy to read.


## If you do not have those

Go through:

-   [Transcriptomics 2: Statistical Analysis](../week-4/workshop.html) including:

-   [ü§ó Look after future
    you!](../week-4/workshop.html#look-after-future-you) and

-   the [Independent Study to
    consolidate](../week-4/study_after_workshop.html)

# Examine the results files

## All results files

Remind yourself of the key columns in any of the results files:

-   normalised counts for each sample/cell
-   a log~2~ fold change
-   an unadjusted *p*-value
-   a *p* value adjusted for multiple testing (called `FDR` or `padj`)
-   a gene id
-   other information about each gene

## üê∏ , üéÑ , üíâ results files

-   `baseMean` is the mean of the normalised counts for the gene across
    all samples
-   `lfcSE` standard error of the fold change
-   `stat` is the test statistic (the Wald statistic)



## üê≠ Stem cells


-   Top is the rank of the gene ordered by the *p*-value (smallest
    first)
-   `summary.logFC` and `logFC.hspc` give the same value (in this case
    since comparing two cell types)



# Plots

## What is the purpose of a Transcriptomics plot?

::: incremental
-   In general, we plot data to help us summarise and understand it

-   This is especially import for transcriptomics data where we have a very large
    number of variables and often a large number of observations

-   We will look at two plots very commonly used in transcriptomics analysis:
    Principal Component Analysis (PCA) plot and Volcano Plots
:::

# Principal Component Analysis (PCA)

## PCA

::: incremental
-   Principal Component Analysis is an unsupervised machine learning
    technique

-   Unsupervised methods[^1] are unsupervised in that they do not
    use/optimise to a particular output. The goal is to uncover
    structure. They do not test hypotheses

-   It is often used to visualise high dimensional data because it is a
    dimension reduction technique
:::

[^1]: You may wish to read a previous introduction to unsupervised
    methods I have written [An introduction to Machine Learning:
    Unsupervised
    methods](https://3mmarand.github.io/BIO00058M-Data-science-2020/slides/05_intro_to_ML_unsupervised.html#1)
    [@rand2021]

## PCA

::: incremental
-   Takes a large number of continuous variables (like gene
    expression) and reduces them to a smaller number of variables
    (called principal components) that explain most of the variation in
    the data

-   The principal components can be plotted to see how samples cluster
    together
:::

## PCA
::: {style="font-size: 70%;"}
-   To understand the logic of PCA, imagine we might plotting the
    expression of one gene against that of another

::: {layout-ncol="2"}
![Samples](images/why_pca_frog.png){width="300"}

![Cells](images/why_pca_mouse.png){width="300"}
:::

This gives us some in insight in how the sample/cells cluster. But we have 
a lot of genes (even for the stem cells) to consider. How do we know if 
the pair we use is typical? How can we consider al the genes at once?

:::

## PCA

::: {style="font-size: 70%;"}
-   PCA is a solution for this - It takes a large number of continuous
    variables (like gene expression) and reduces them to a smaller
    number of "principal components" that explain most of the variation
    in the data.

::: {layout-ncol="2"}

![Samples](images/frog-s30-pca.png){width="300"}

![Cells](images/prog_hspc-pca.png){width="300"}
:::


:::
## PCA

We have done PCA after differential expression, but often PCA might 
is one of the first exploratory steps because it gives you an idea 
whether you expect general patterns in gene expression that distinguish 
groups.


# Volcano plots

## Volcano plots

::: incremental
-   Volcano plots often used to visualise the results of differential
    expression analysis

-   They are just a scatter of the adjusted *p* value against the fold
    change....

-   almost - the we actually plot the negative log of the adjusted *p*
    value against the log fold change
:::

## Volcano plots

::: {style="font-size: 70%;"}
-   This is because small probabilities are important, large ones are 
    not so the axis is counter intuitive because Small *p*-values 
    (i.e., significant values) are at the bottom of the axis)

-   And since *p*-values range from 1 to very tiny the important
    points are all squashed at the bottom of the axis
:::

![Volcano plot padj against fold change](images/volcano-why-1.png)

## Volcano plots

::: {style="font-size: 70%;"}
-   Plotting the negative log of the adjusted *p*-value means that the
    values are spread out, *and* most significant are at the top of 
    the axis
:::

![Volcano plot -log(adjusted p) against fold change](images/volcano-why-2.png)


## Visualisations

-   Should be done on normalised data so meaningful comparisons can be
    made

-   The üê≠ stem cell data were already log~2~normalised

-   The other datasets were normalised by the DE method and we saved 
    the values to the results files

## Packages

This packages is on the University computers which you can access on 
campus or [remotely using the VDS](https://www.york.ac.uk/it-services/tools/vds/)

If you want to use your own machine you will need to install the package.
::: {style="font-size: 60%;"}

Install **`ggrepel`** from CRAN in the the normal way:

```{r}
#| eval: false
install.packages("ggrepel")

```

This package allows you to label points on a plot without them overlapping.

# Workshops

## Workshops

-   Transcriptomics 1: Hello data Getting to know the data. Checking the
    distributions of values overall, across rows and columns to
    check things are as we expect and detect rows/columns that need to
    be removed

-   Transcriptomics 2: Statistical Analysis. Identifying which genes are
    differentially expressed between treatments. This is the main
    analysis step. We will use different methods for bulk and single
    cell data.

-   Transcriptomics 3: Visualising. Principal Component 
    Analysis (PCA) volcano plots to visualise the results of the 

## References
