<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.33">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Emma Rand">
<title>Workshop â€“ Data Analysis for Group Project</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-bc185b5c5bdbcb35c2eb49d8a876ef70.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" class="quarto-color-scheme-extra" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-72ac799c4abbddf31cb463dd8fc99f5d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark-25ac7be22ffa75a8e8541e30a6dcc9a3.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="../../site_libs/bootstrap/bootstrap-72ac799c4abbddf31cb463dd8fc99f5d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme-extra" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-TVF95J593T"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-TVF95J593T', { 'anonymize_ip': true});
</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar floating nav-fixed quarto-light"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = false;
    const darkModeDefault = authorPrefersDark;
      document.querySelector('link#quarto-text-highlighting-styles.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
      document.querySelector('link#quarto-bootstrap.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
      window.dispatchEvent(new Event('resize'));
    };
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Data Analysis for Group Project</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Welcome!</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../core/core.html"> 
<span class="menu-text">Core Supporting Info</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../transcriptomics/transcriptomics.html" aria-current="page"> 
<span class="menu-text">Transcriptomics</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../images/images.html"> 
<span class="menu-text">Image Analysis</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../structures/structures.html"> 
<span class="menu-text">Structure Analysis</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools tools-wide">
    <a href="https://twitter.com/er_13" title="Emma's Twitter" class="quarto-navigation-tool px-1" aria-label="Emma's Twitter"><i class="bi bi-twitter"></i></a>
    <a href="https://github.com/3mmaRand/BIO00088H-data" title="BIO00088H-data GitHub" class="quarto-navigation-tool px-1" aria-label="BIO00088H-data GitHub"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/week-5/overview.html">Week 5: Visualising and Interpreting</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/week-5/workshop.html">Workshop</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/transcriptomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Transcriptomics Data Analysis for Group Project</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 3: Hello data!</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-3/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 4: Statistical Analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-4/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Week 5: Visualising and Interpreting</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_before_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prepare!</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/workshop.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/week-5/study_after_workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Consolidate!</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Semester 2: Extras</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/semester-2/stem_cell_variant.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Additional ğŸ­ stem cell labels</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Kellyâ€™s Project</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../transcriptomics/kelly/workshop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workshop</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul>
<li><a href="#session-overview" id="toc-session-overview" class="nav-link" data-scroll-target="#session-overview">Session overview</a></li>
  </ul>
</li>
  <li>
<a href="#set-up" id="toc-set-up" class="nav-link" data-scroll-target="#set-up">Set up</a>
  <ul>
<li><a href="#arabidopsis" id="toc-arabidopsis" class="nav-link" data-scroll-target="#arabidopsis">ğŸ„ <em>Arabidopsis</em></a></li>
  <li><a href="#leishmania" id="toc-leishmania" class="nav-link" data-scroll-target="#leishmania">ğŸ’‰ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells" id="toc-stem-cells" class="nav-link" data-scroll-target="#stem-cells">ğŸ­ Stem cells</a></li>
  <li><a href="#everyone" id="toc-everyone" class="nav-link" data-scroll-target="#everyone">Everyone</a></li>
  </ul>
</li>
  <li>
<a href="#import" id="toc-import" class="nav-link" data-scroll-target="#import">Import</a>
  <ul>
<li><a href="#everyone-1" id="toc-everyone-1" class="nav-link" data-scroll-target="#everyone-1">Everyone</a></li>
  </ul>
</li>
  <li>
<a href="#log2-transform-the-normalised-counts" id="toc-log2-transform-the-normalised-counts" class="nav-link" data-scroll-target="#log2-transform-the-normalised-counts">log<sub>2</sub> transform the normalised counts</a>
  <ul>
<li><a href="#arabidopsis-and-leishmania" id="toc-arabidopsis-and-leishmania" class="nav-link" data-scroll-target="#arabidopsis-and-leishmania">ğŸ„ <em>Arabidopsis</em> and ğŸ’‰ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells-1" id="toc-stem-cells-1" class="nav-link" data-scroll-target="#stem-cells-1">ğŸ­ Stem cells</a></li>
  <li><a href="#everyone-2" id="toc-everyone-2" class="nav-link" data-scroll-target="#everyone-2">Everyone</a></li>
  </ul>
</li>
  <li>
<a href="#write-the-significant-genes-to-file" id="toc-write-the-significant-genes-to-file" class="nav-link" data-scroll-target="#write-the-significant-genes-to-file">Write the significant genes to file</a>
  <ul>
<li><a href="#everyone-3" id="toc-everyone-3" class="nav-link" data-scroll-target="#everyone-3">Everyone</a></li>
  </ul>
</li>
  <li>
<a href="#principal-component-analysis-pca" id="toc-principal-component-analysis-pca" class="nav-link" data-scroll-target="#principal-component-analysis-pca">Principal Component Analysis (PCA)</a>
  <ul>
<li><a href="#arabidopsis-1" id="toc-arabidopsis-1" class="nav-link" data-scroll-target="#arabidopsis-1">ğŸ„ <em>Arabidopsis</em></a></li>
  <li><a href="#leishmania-1" id="toc-leishmania-1" class="nav-link" data-scroll-target="#leishmania-1">ğŸ’‰ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells-2" id="toc-stem-cells-2" class="nav-link" data-scroll-target="#stem-cells-2">ğŸ­ Stem cells</a></li>
  </ul>
</li>
  <li>
<a href="#visualise-all-the-results-with-a-volcano-plot" id="toc-visualise-all-the-results-with-a-volcano-plot" class="nav-link" data-scroll-target="#visualise-all-the-results-with-a-volcano-plot">Visualise all the results with a volcano plot</a>
  <ul>
<li><a href="#arabidopsis-2" id="toc-arabidopsis-2" class="nav-link" data-scroll-target="#arabidopsis-2">ğŸ„ <em>Arabidopsis</em></a></li>
  <li><a href="#leishmania-2" id="toc-leishmania-2" class="nav-link" data-scroll-target="#leishmania-2">ğŸ’‰ <em>Leishmania</em></a></li>
  <li><a href="#stem-cells-3" id="toc-stem-cells-3" class="nav-link" data-scroll-target="#stem-cells-3">ğŸ­ Stem cells</a></li>
  </ul>
</li>
  <li><a href="#save-your-plots" id="toc-save-your-plots" class="nav-link" data-scroll-target="#save-your-plots">Save your plots</a></li>
  <li><a href="#look-after-future-you" id="toc-look-after-future-you" class="nav-link" data-scroll-target="#look-after-future-you">ğŸ¤— Look after future you!</a></li>
  <li><a href="#finished" id="toc-finished" class="nav-link" data-scroll-target="#finished">ğŸ¥³ Finished</a></li>
  <li><a href="#independent-study-following-the-workshop" id="toc-independent-study-following-the-workshop" class="nav-link" data-scroll-target="#independent-study-following-the-workshop">Independent study following the workshop</a></li>
  <li><a href="#the-code-file" id="toc-the-code-file" class="nav-link" data-scroll-target="#the-code-file">The Code file</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../transcriptomics/week-5/overview.html">Week 5: Visualising and Interpreting</a></li><li class="breadcrumb-item"><a href="../../transcriptomics/week-5/workshop.html">Workshop</a></li></ol></nav><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Workshop</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
<p class="subtitle lead">Transcriptomics 3: Visualising</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Emma Rand </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">13 August, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header><section id="introduction" class="level1"><h1>Introduction</h1>
<section id="session-overview" class="level2"><h2 class="anchored" data-anchor-id="session-overview">Session overview</h2>
<p>In the workshop, you will learn how to conduct and plot a Principle Component Analysis (PCA) as well as how to create a nicely formatted Volcano plot. You will also save significant genes to file to make it easier to identify genes of interest. <del>and perform Gene Ontology (GO) term enrichment analysis.</del></p>
</section></section><section id="set-up" class="level1"><h1>Set up</h1>
<section id="arabidopsis" class="level2"><h2 class="anchored" data-anchor-id="arabidopsis">ğŸ„ <em>Arabidopsis</em>
</h2>
<p>ğŸ¬ Open the <code>arabi-88H</code> RStudio Project and the <code>wildsuf-wilddef.R</code> script.</p>
</section><section id="leishmania" class="level2"><h2 class="anchored" data-anchor-id="leishmania">ğŸ’‰ <em>Leishmania</em>
</h2>
<p>ğŸ¬ Open the <code>leish-88H</code> RStudio Project and the <code>pro-meta.R</code> script.</p>
</section><section id="stem-cells" class="level2"><h2 class="anchored" data-anchor-id="stem-cells">ğŸ­ Stem cells</h2>
<p>ğŸ¬ Open the <code>mice-88H</code> RStudio Project and the <code>hspc-prog.R</code> script.</p>
</section><section id="everyone" class="level2"><h2 class="anchored" data-anchor-id="everyone">Everyone</h2>
<p>ğŸ¬ Make a new folder <code>figures</code> in the project directory.</p>
<p>This is where we will save our figure files</p>
<p>ğŸ¬ Load <strong><code>tidyverse</code></strong> <span class="citation" data-cites="tidyverse">(<a href="#ref-tidyverse" role="doc-biblioref">Wickham et al. 2019</a>)</span> and <strong><code>conflicted</code></strong> <span class="citation" data-cites="conflicted">(<a href="#ref-conflicted" role="doc-biblioref">Wickham 2023</a>)</span>. You most likely have this code at the top of your script already.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://conflicted.r-lib.org/">conflicted</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€
âœ” dplyr     1.1.3     âœ” readr     2.1.4
âœ” forcats   1.0.0     âœ” stringr   1.5.0
âœ” ggplot2   3.4.3     âœ” tibble    3.2.1
âœ” lubridate 1.9.3     âœ” tidyr     1.3.0
âœ” purrr     1.0.2     
â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€
âœ– dplyr::filter() masks stats::filter()
âœ– dplyr::lag()    masks stats::lag()
â„¹ Use the conflicted package to force all conflicts to become errors</code></pre>
<p>I recommend you set the <strong><code>dplyr</code></strong> versions of <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> to use by default</p>
<p>ğŸ¬ Use the <strong><code>dplyr</code></strong> version of <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> by default:</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://conflicted.r-lib.org/reference/conflicts_prefer.html">conflicts_prefer</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://conflicted.r-lib.org/reference/conflicts_prefer.html">conflicts_prefer</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="import" class="level1"><h1>Import</h1>
<section id="everyone-1" class="level2"><h2 class="anchored" data-anchor-id="everyone-1">Everyone</h2>
<p>ğŸ¬ Import your results data. This should be a file in the <code>results</code> folder called <code>xxxx_results.csv</code> where xxxx indicates the comparison you made.</p>
<p>ğŸ¬ Remind yourself what is in the rows and columns and the structure of the dataframes (perhaps using <code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code>)</p>
<!-- #---THINKING ANSWER--- -->
<!-- You have data with 13 columns and 26355 rows. The rows are the genes. -->
<!-- The columns are the normalised counts for for each of the 4 -->
<!-- samples, the mean of the normalised counts, the gene id, -->
<!-- the gene name and a description. -->
<!-- You also have the results of a statistical comparison between -->
<!-- the Cu sufficient and Cu deficient conditions, the log fold change,  -->
<!-- the test statistic, the p-value and the adjusted p-value -->
<!-- #---THINKING ANSWER--- -->
<!-- You have data with 14 columns and 8437 rows. The rows are the genes. -->
<!-- The columns are the normalised counts for for each of the 6 -->
<!-- samples, the mean of the normalised counts, the gene id, -->
<!-- and a description. -->
<!-- You also have the results of a statistical comparison between -->
<!-- the procyclic and metacyclic promastigote stages, the log  -->
<!-- fold change, the test statistic, the p-value and the  -->
<!-- adjusted p-value -->
<!-- #---THINKING ANSWER--- -->
<!-- You have data with 1507 columns and 280 rows. The rows are the genes. -->
<!-- The columns are the normalised counts for each of the cell -->
<!-- (1499 HPSC and and Prog cells), the ensembl_gene_id, a gene name and a -->
<!-- description. -->
<!-- You also have the results of a statistical comparison between -->
<!-- the HSPC and progenitor cell types, the log fold change (twice!), -->
<!-- the p-value and the adjusted p-value (FDR) -->
<p>When we do PCA we will want to label the samples with their treatment for figures. For ğŸ¸ Frog development, ğŸ„ <em>Arabidopsis</em> and ğŸ’‰ <em>Leishmania</em>, this labelling information is most easily added using the metadata. You will need to filter for only the samples in the comparison that was made in the results file.</p>
<p>You may need to refer back to the <a href="../week-4/workshop.html">Week 4 Statistical Analysis workshop</a> (in section 2. Create DESeqDataSet object) to remind yourself how to import and filter the metadata you need.</p>
<p>ğŸ¬ Import the metadata that maps the sample names to treatments. Remember to select only the samples for comparison that was made.</p>
<p>For ğŸ­ Stem cells, cell types are encoded in the column names. We will do some <a href="https://en.wikipedia.org/wiki/Regular_expression">regular expression</a> magic to extract the cell type from the column names.</p>
</section></section><section id="log2-transform-the-normalised-counts" class="level1"><h1>log<sub>2</sub> transform the normalised counts</h1>
<p>We use the normalised counts for data visualisations so that the comparisons are meaningful. Since the fold changes are given is log<sub>2</sub> it is useful to log<sub>2</sub> transform the normalised counts too. We will add columns to the dataframe with these transformed values. Since we have some counts of 0 we will add a tiny amount to avoid <code>-Inf</code> values.</p>
<p>log<sub>2</sub> transformation would be applied to one column like this:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># DO NOT DO</span></span>
<span><span class="co"># log2 transform the counts plus a tiny amount to avoid log(0)</span></span>
<span><span class="va">dataframe</span> <span class="op">&lt;-</span> <span class="va">dataframe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>log2_mycolumn <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">mycolumn</span> <span class="op">+</span> <span class="fl">0.001</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are going to use a wonderful bit of R wizardry to apply a transformation to multiple columns. This is the <code><a href="https://dplyr.tidyverse.org/reference/across.html">across()</a></code> function which has three arguments:</p>
<p><code>across(.cols, .fns, .names)</code></p>
<p>where:</p>
<ul>
<li>
<code>.cols</code> is the selection of columns to transform</li>
<li>
<code>.fns</code> is the function we want to apply to the selected columns</li>
<li>
<code>.names</code> is the naming convention for the new columns</li>
</ul>
<p>The general form of the code you need is:</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># DO NOT DO</span></span>
<span><span class="co"># log2 transform the counts plus a tiny amount to avoid log(0)</span></span>
<span><span class="va">xxxx_results</span> <span class="op">&lt;-</span> <span class="va">xxxx_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="st">"pattern"</span><span class="op">)</span>, </span>
<span>                \<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">x</span> <span class="op">+</span> <span class="fl">0.001</span><span class="op">)</span>,</span>
<span>                .names <span class="op">=</span> <span class="st">"log2_{.col}"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>where:</p>
<ul>
<li>
<code>xxxx_results</code> is the name of the dataframe of results</li>
<li>
<code>pattern</code> matches the starting letters for all of the normalised counts so that <code>starts_with("pattern")</code> gives the selection of columns to transform</li>
<li>the bit after the <code>\(x)</code> is the function we want to apply to the selected columns</li>
<li>the <code>\(x)</code> means it is an â€œanonymousâ€ function which means we donâ€™t have to define a function name.</li>
<li>
<code>"log2_{.col}"</code> means the columns will have the same name (in the <code>.col</code>) but with the prefix <code>log2_</code> added.</li>
</ul>
<p>You can read more about <code><a href="https://dplyr.tidyverse.org/reference/across.html">across()</a></code> and anonymous functions from my <a href="https://posit-conf-2024.github.io/programming-r/">posit::conf(2024) workshop</a></p>
<section id="arabidopsis-and-leishmania" class="level2"><h2 class="anchored" data-anchor-id="arabidopsis-and-leishmania">ğŸ„ <em>Arabidopsis</em> and ğŸ’‰ <em>Leishmania</em>
</h2>
<p>ğŸ¬ Design the code to log<sub>2</sub> transform the normalised counts using the template given</p>
<p>I recommend viewing the dataframe to see the new columns. Check you have the expected number of columns.</p>
</section><section id="stem-cells-1" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-1">ğŸ­ Stem cells</h2>
<p>You do not need to apply this transformation because the data is already log<sub>2</sub> transformed.</p>
</section><section id="everyone-2" class="level2"><h2 class="anchored" data-anchor-id="everyone-2">Everyone</h2>
<p>We now all have dataframes with all the information we need: normalised counts, log<sub>2</sub> normalised counts, statistical comparisons with fold changes and <em>p</em>-values, and information about the gene.</p>
</section></section><section id="write-the-significant-genes-to-file" class="level1"><h1>Write the significant genes to file</h1>
<section id="everyone-3" class="level2"><h2 class="anchored" data-anchor-id="everyone-3">Everyone</h2>
<p>We will create dataframe of the significant genes and write them to file. This is subset from the results file but will make it a little easier to examine and select genes of interest.</p>
<p>The general form of the code you need is:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># DO NOT DO</span></span>
<span><span class="co"># create a dataframe of genes significant at 0.05 level</span></span>
<span><span class="va">xxxx_results_sig0.05</span> <span class="op">&lt;-</span> <span class="va">xxxx_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">pvalue</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that you determine the significance level using the adjusted <em>p</em>-values rather than the uncorrected <em>p</em>-values. This column is name <code>padj</code> in <strong><code>DESeq2</code></strong> output and <code>FDR</code> in <strong><code>scran</code></strong> output.</p>
<p>ğŸ¬ Create a dataframe of the genes significant at the 0.05 level using <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code>. You will need to know the name of column with the adjusted <em>p</em>-values.</p>
<p>â“How many genes are significant at the 0.05 levels?</p>
<!-- #---THINKING ANSWER--- -->
<!-- ğŸ„ Arabidopsis  -->
<!-- 12 genes are significant at the 0.05 level. -->
<!-- ğŸ’‰ leishmania  -->
<!-- 3252 genes are significant at the 0.05 level. -->
<!-- I recommend using a more stringent cut-off such as 0.01  -->
<!-- ğŸ­ stemcells -->
<!-- 182 genes are significant at the 0.05 level. -->
<p>If you have a very large number of genes significant at the 0.05 level, you may want to consider a more stringent cut-off such as 0.01.</p>
<p>ğŸ¬ Write the dataframe to a csv file. I recommend using the same file name as you used for the dataframe.</p>
</section></section><section id="principal-component-analysis-pca" class="level1"><h1>Principal Component Analysis (PCA)</h1>
<p>We have many genes in our datasets. PCA will allow us to plot our samples in the â€œgene expressionâ€ space so we can see if replicates with the same treatment cluster together as we would expect. PCA is a dimension reduction technique that finds the directions of maximum variance in the data. We are doing PCA after our statistical analysis but often it is one of the first techniques used to explore the data. If we do not see treatment effects in a PCA plot then we are not likely to see them in the statistical analysis. PCA can also help you identify any outlier replicates. The PCA is best conducted on the normalised counts or the log<sub>2</sub> transformed normalised counts. We will use the log<sub>2</sub> transformed normalised counts.</p>
<p>We will carry out the following steps to do the PCA:</p>
<ul>
<li>Select the log<sub>2</sub> transformed normalised counts. You can only use numerical data in PCA.</li>
<li>Transpose our data. We have genes in rows and samples in columns (this is common for gene expression data). However, to treatment the genes as variables, PCA expects samples in rows and genes in columns.</li>
<li>Add the gene names as column names in the transposed data</li>
<li>Perform the PCA</li>
<li>Extract the scores on the first two principal components and label the data</li>
<li>Plot the the first two principal components as a scatter plot</li>
</ul>
<p>In each case we will use only the samples for the comparison that was made in the PCA. However, it would be informative to do PCA on all the samples you have and that is something you may want to consider in your independent study.</p>
<p>Now go to PCA for:</p>
<ul>
<li><a href="#arabidopsis-1">ğŸ„ Arabidopsis</a></li>
<li><a href="#leishmania-1">ğŸ’‰ Leishmania</a></li>
<li><a href="#stem-cells-2">ğŸ­ Stem cells</a></li>
</ul>
<section id="arabidopsis-1" class="level2"><h2 class="anchored" data-anchor-id="arabidopsis-1">ğŸ„ <em>Arabidopsis</em>
</h2>
<p>ğŸ¬ Transpose the log<sub>2</sub> transformed normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_log2_trans</span> <span class="op">&lt;-</span> <span class="va">wild_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="st">"log2_"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We have used the <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> function to select all the columns that start with <code>log2_</code>. We then use the <code><a href="https://rdrr.io/r/base/t.html">t()</a></code> function to transpose the dataframe. We then convert the resulting matrix to a dataframe using <code><a href="https://rdrr.io/r/base/data.frame.html">data.frame()</a></code>. If you view that dataframe youâ€™ll see it has default column name which we can fix using <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> to set the column names to the gene ids.</p>
<p>ğŸ¬ Set the column names to the gene ids:</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">wild_log2_trans</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">wild_results</span><span class="op">$</span><span class="va">gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Perform PCA on the log<sub>2</sub> transformed normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="va">wild_log2_trans</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span>rank. <span class="op">=</span> <span class="fl">4</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>scale</code> argument tells <code><a href="https://rdrr.io/r/stats/prcomp.html">prcomp()</a></code> to scale the data before doing the PCA. This is important when the variables are on different scales to stop variables with large values dominating the PCA. The <code>rank.</code> argument tells <code><a href="https://rdrr.io/r/stats/prcomp.html">prcomp()</a></code> to only calculate the first 4 principal components. This is useful for visualisation as we can only plot in 2 or 3 dimensions. We can see the results of the PCA by viewing the <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> of the <code>pca</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Importance of components:
                            PC1     PC2     PC3       PC4
Standard deviation     123.0171 85.6822 62.2924 2.186e-13
Proportion of Variance   0.5742  0.2786  0.1472 0.000e+00
Cumulative Proportion    0.5742  0.8528  1.0000 1.000e+00</code></pre>
</div>
</div>
<p>The Proportion of Variance tells us how much of the variance is explained by each component. We can see that the first component explains 0.5742 of the variance, the second 0.2786, and the third 0.1472. Together the first three components explain nearly 100% of the total variance in the data. Plotting PC1 against PC2 will capture about 92% of the variance which is very likely very much better than we would get plotting any two genes against each other. To plot the PC1 against PC2 we will need to extract the PC1 and PC2 â€œscoresâ€ from the PCA object and add labels for the samples. Those labels will come from the row names of the transformed data which has the sample ids and from the metadata.</p>
<p>ğŸ¬ Create a vector of the sample ids from the row names. These include the <code>log2</code> prefix which we can removed for labelling:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">wild_log2_trans</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="st">"log2_"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You might want to check the result.</p>
<p>Now we will extract the PC1 and PC2 scores from the PCA object and add. Our PCA object is called <code>pca</code> and the scores are in pca$x. We will create a dataframe of the scores and add the sample ids.</p>
<p>ğŸ¬ Create a dataframe of PC1 and PC2 scores and add the sample ids:</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">x</span>,</span>
<span>                           <span class="va">sample_id</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Merge with the metadata so we can label points by treatment and sibling pair:</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">&lt;-</span> <span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">meta_wild</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since the metadata contained the sample ids, it was especially important to remove the <code>log2_</code> from the row names so that the join would work.</p>
<p>The dataframe should look like this:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 14%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 5%">
<col style="width: 25%">
<col style="width: 12%">
<col style="width: 14%">
</colgroup>
<thead><tr class="header">
<th style="text-align: right;">PC1</th>
<th style="text-align: right;">PC2</th>
<th style="text-align: right;">PC3</th>
<th style="text-align: right;">PC4</th>
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">genotype</th>
<th style="text-align: left;">copper</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">-91.23038</td>
<td style="text-align: right;">105.92675</td>
<td style="text-align: right;">25.80774</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">SRX028956_wild_suf</td>
<td style="text-align: left;">wt</td>
<td style="text-align: left;">sufficient</td>
</tr>
<tr class="even">
<td style="text-align: right;">-120.00085</td>
<td style="text-align: right;">-95.79115</td>
<td style="text-align: right;">-13.72618</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">SRX028957_wild_def</td>
<td style="text-align: left;">wt</td>
<td style="text-align: left;">deficient</td>
</tr>
<tr class="odd">
<td style="text-align: right;">92.06667</td>
<td style="text-align: right;">23.00841</td>
<td style="text-align: right;">-79.23090</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">SRX028960_wild_suf</td>
<td style="text-align: left;">wt</td>
<td style="text-align: left;">sufficient</td>
</tr>
<tr class="even">
<td style="text-align: right;">119.16456</td>
<td style="text-align: right;">-33.14401</td>
<td style="text-align: right;">67.14935</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">SRX028961_wild_def</td>
<td style="text-align: left;">wt</td>
<td style="text-align: left;">deficient</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The next task is to plot PC2 against PC1 and colour by copper conditions. This is just a scatterplot so we can use <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point()</a></code>. We will use colour to indicate the copper conditions.</p>
<p>ğŸ¬ Plot PC2 against PC1 and colour by copper conditions:</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC3</span>, y <span class="op">=</span> <span class="va">PC2</span>, </span>
<span>             colour <span class="op">=</span> <span class="va">copper</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We do not see particularly good separation between treatments, though perhaps there is some separation between the copper sufficient and copper deficient on PC2. It is also difficult to see if the replicates cluster together (the treatments separate) when there are only two reps. You can also try plotting PC3 or PC4.</p>
<p>I prefer to customise the colours and shapes. I especially like the<br>
viridis colour scales which provide colour scales that are perceptually uniform in both colour and black-and-white. They are also designed to be perceived by viewers with common forms of colour blindness. See <a href="https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html#introduction">Introduction to viridis</a> for more information.</p>
<p><code>ggplot</code> provides functions to access the viridis scales. Here I use <code><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_d()</a></code>. The d stands for discrete. The function <code><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_c()</a></code> would be used for continuous data. Iâ€™ve used the default â€œviridisâ€ (or â€œDâ€) option (do ?scale_fill_viridis_d for all the options) and used the <code>begin</code> and <code>end</code> arguments to control the range of colour - I have set the range to be from 0.15 to 0.95 the avoid the strongest contrast. I have also set the <code>name</code> argument to provide a label for the legend.</p>
<p>ğŸ¬ Customise the PC2 against PC1 plot:</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, </span>
<span>             colour <span class="op">=</span> <span class="va">copper</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_colour_viridis_d</a></span><span class="op">(</span>end <span class="op">=</span> <span class="fl">0.95</span>, begin <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>                         name <span class="op">=</span> <span class="st">"Copper"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now go to Volcano plots for <a href="#arabidopsis-2">ğŸ„ Arabidopsis</a></p>
</section><section id="leishmania-1" class="level2"><h2 class="anchored" data-anchor-id="leishmania-1">ğŸ’‰ <em>Leishmania</em>
</h2>
<p>ğŸ¬ Transpose the log<sub>2</sub> transformed normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_log2_trans</span> <span class="op">&lt;-</span> <span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="st">"log2_"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We have used the <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> function to select all the columns that start with <code>log2_</code>. We then use the <code><a href="https://rdrr.io/r/base/t.html">t()</a></code> function to transpose the dataframe. We then convert the resulting matrix to a dataframe using <code><a href="https://rdrr.io/r/base/data.frame.html">data.frame()</a></code>. If you view that dataframe youâ€™ll see it has default column name which we can fix using <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> to set the column names to the gene ids.</p>
<p>ğŸ¬ Set the column names to the gene ids:</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">pro_meta_log2_trans</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pro_meta_results</span><span class="op">$</span><span class="va">gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Perform PCA on the log<sub>2</sub> transformed normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="va">pro_meta_log2_trans</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span>rank. <span class="op">=</span> <span class="fl">4</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>scale</code> argument tells <code><a href="https://rdrr.io/r/stats/prcomp.html">prcomp()</a></code> to scale the data before doing the PCA. This is important when the variables are on different scales to stop variables with large values dominating the PCA. The <code>rank.</code> argument tells <code><a href="https://rdrr.io/r/stats/prcomp.html">prcomp()</a></code> to only calculate the first 4 principal components. This is useful for visualisation as we can only plot in 2 or 3 dimensions. We can see the results of the PCA by viewing the <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> of the <code>pca</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Importance of first k=4 (out of 6) components:
                           PC1     PC2     PC3      PC4
Standard deviation     66.0744 41.9741 29.4077 27.61377
Proportion of Variance  0.5175  0.2088  0.1025  0.09038
Cumulative Proportion   0.5175  0.7263  0.8288  0.91916</code></pre>
</div>
</div>
<p>The Proportion of Variance tells us how much of the variance is explained by each component. We can see that the first component explains 0.5175 of the variance, the second 0.2088, and the third 0.1025. Together the first three components explain nearly 92% of the total variance in the data. Plotting PC1 against PC2 will capture about 66% of the variance which is likely very much better than we would get plotting any two genes against each other. To plot the PC1 against PC2 we will need to extract the PC1 and PC2 â€œscoresâ€ from the PCA object and add labels for the samples. Those labels will come from the row names of the transformed data which has the sample ids and from the metadata.</p>
<p>ğŸ¬ Create a vector of the sample ids from the row names. These include the <code>log2</code> prefix which we can removed for labelling:</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">pro_meta_log2_trans</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="st">"log2_"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You might want to check the result.</p>
<p>Now we will extract the PC1 and PC2 scores from the PCA object and add. Our PCA object is called <code>pca</code> and the scores are in pca$x. We will create a dataframe of the scores and add the sample ids.</p>
<p>ğŸ¬ Create a dataframe of PC1 and PC2 scores and add the sample ids:</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">x</span>,</span>
<span>                           <span class="va">sample_id</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Merge with the metadata so we can label points by life cycle stage:</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">&lt;-</span> <span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">meta_pro_meta</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since the metadata contained the sample ids, it was especially important to remove the <code>log2_</code> from the row names so that the join would work.</p>
<p>The dataframe should look like this:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 14%">
<col style="width: 14%">
<col style="width: 16%">
<col style="width: 13%">
<col style="width: 14%">
<col style="width: 13%">
</colgroup>
<thead><tr class="header">
<th style="text-align: right;">PC1</th>
<th style="text-align: right;">PC2</th>
<th style="text-align: right;">PC3</th>
<th style="text-align: right;">PC4</th>
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">stage</th>
<th style="text-align: right;">replicate</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">-59.42107</td>
<td style="text-align: right;">-2.922229</td>
<td style="text-align: right;">-11.043368</td>
<td style="text-align: right;">-13.4641060</td>
<td style="text-align: left;">lm_pro_1</td>
<td style="text-align: left;">procyclic</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: right;">-63.25818</td>
<td style="text-align: right;">-15.985856</td>
<td style="text-align: right;">-36.248191</td>
<td style="text-align: right;">0.7025242</td>
<td style="text-align: left;">lm_pro_2</td>
<td style="text-align: left;">procyclic</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: right;">-57.92065</td>
<td style="text-align: right;">25.679645</td>
<td style="text-align: right;">47.373874</td>
<td style="text-align: right;">13.7893070</td>
<td style="text-align: left;">lm_pro_3</td>
<td style="text-align: left;">procyclic</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: right;">56.20326</td>
<td style="text-align: right;">-59.707787</td>
<td style="text-align: right;">5.958727</td>
<td style="text-align: right;">31.5528028</td>
<td style="text-align: left;">lm_meta_1</td>
<td style="text-align: left;">metacyclic</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: right;">57.41205</td>
<td style="text-align: right;">-11.738157</td>
<td style="text-align: right;">14.157242</td>
<td style="text-align: right;">-47.2285609</td>
<td style="text-align: left;">lm_meta_2</td>
<td style="text-align: left;">metacyclic</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: right;">66.98459</td>
<td style="text-align: right;">64.674383</td>
<td style="text-align: right;">-20.198284</td>
<td style="text-align: right;">14.6480329</td>
<td style="text-align: left;">lm_meta_3</td>
<td style="text-align: left;">metacyclic</td>
<td style="text-align: right;">3</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The next task is to plot PC2 against PC1 and colour by sibling pair. This is just a scatterplot so we can use <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point()</a></code>. We will use colour to indicate the life cycle stage.</p>
<p>ğŸ¬ Plot PC2 against PC1 and colour by copper conditions:</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, </span>
<span>             colour <span class="op">=</span> <span class="va">stage</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>There is a good separation between treatments on PCA1. The replicates do seem to cluster together. You can also try plotting PC3 or PC4.</p>
<p>I prefer to customise the colours. I especially like the<br>
viridis colour scales which provide colour scales that are perceptually uniform in both colour and black-and-white. They are also designed to be perceived by viewers with common forms of colour blindness. See <a href="https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html#introduction">Introduction to viridis</a> for more information.</p>
<p><code>ggplot</code> provides functions to access the viridis scales. Here I use <code><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_d()</a></code>. The d stands for discrete. The function <code><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_c()</a></code> would be used for continuous data. Iâ€™ve used the default â€œviridisâ€ (or â€œDâ€) option (do ?scale_fill_viridis_d for all the options) and used the <code>begin</code> and <code>end</code> arguments to control the range of colour - I have set the range to be from 0.15 to 0.95 the avoid the strongest contrast. I have also set the <code>name</code> argument to provide a label for the legend.</p>
<p>ğŸ¬ Plot PC2 against PC1 and colour by life stage:</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, </span>
<span>             colour <span class="op">=</span> <span class="va">stage</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_colour_viridis_d</a></span><span class="op">(</span>end <span class="op">=</span> <span class="fl">0.95</span>, begin <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>                         name <span class="op">=</span> <span class="st">"Stage"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now go to Volcano plots for <a href="#leishmania-2">ğŸ’‰ Leishmania</a></p>
</section><section id="stem-cells-2" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-2">ğŸ­ Stem cells</h2>
<p>ğŸ¬ Transpose the log<sub>2</sub> transformed normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_trans</span> <span class="op">&lt;-</span> <span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HSPC_"</span>, <span class="st">"Prog_"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We have used the <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> function to select all the columns that start with <code>Prog_</code> or <code>HSPC_</code>. We then use the <code><a href="https://rdrr.io/r/base/t.html">t()</a></code> function to transpose the dataframe. We then convert the resulting matrix to a dataframe using <code><a href="https://rdrr.io/r/base/data.frame.html">data.frame()</a></code>. If you view that dataframe youâ€™ll see it has default column name which we can fix using <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> to set the column names to the gene ids.</p>
<p>ğŸ¬ Set the column names to the gene ids:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">hspc_prog_trans</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">hspc_prog_results</span><span class="op">$</span><span class="va">ensembl_gene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Perform PCA on the log<sub>2</sub> transformed normalised counts:</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="va">hspc_prog_trans</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span>rank. <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>rank.</code> argument tells <code><a href="https://rdrr.io/r/stats/prcomp.html">prcomp()</a></code> to only calculate the first 8 principal components. This is useful for visualisation as we can only plot in 2 or 3 dimensions. We can see the results of the PCA by viewing the <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> of the <code>pca</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Importance of first k=8 (out of 280) components:
                           PC1     PC2     PC3     PC4     PC5     PC6     PC7
Standard deviation     12.5612 8.36646 5.98988 5.41386 4.55730 4.06142 3.84444
Proportion of Variance  0.1099 0.04874 0.02498 0.02041 0.01446 0.01149 0.01029
Cumulative Proportion   0.1099 0.15861 0.18359 0.20400 0.21846 0.22995 0.24024
                           PC8
Standard deviation     3.70848
Proportion of Variance 0.00958
Cumulative Proportion  0.24982</code></pre>
</div>
</div>
<p>The Proportion of Variance tells us how much of the variance is explained by each component. We can see that the first component explains 0.1099 of the variance, the second 0.04874, and the third 0.02498. Together the first three components explain 18% of the total variance in the data. This is not <em>that</em> high but it is also quite a lot better than than we would get plotting any two genes randomly chosen against each other.</p>
<p>To plot the PC1 against PC2 we will need to extract the PC1 and PC2 â€œscoresâ€ from the PCA object and add labels for the cells. Our PCA object is called <code>pca</code> and the scores are in pca$x. The cells labels will come from the row names of the transformed data.</p>
<p>ğŸ¬ Create a dataframe of the PC1 and PC2 scores (in <code>pca$x</code>) and add the cell ids:</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">x</span>,</span>
<span>                           cell_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">hspc_prog_trans</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It will be helpful to add a column for the cell type so we can label points. One way to do this is to extract the information in the cell_id column into two columns: one with the complete cell id and one with just the cell type. This extraction is done with <a href="https://en.wikipedia.org/wiki/Regular_expression">Regular Expression</a> magic.</p>
<p>ğŸ¬ Extract the cell type and cell number from the <code>cell_id</code> column (keeping the <code>cell_id</code> column):</p>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">&lt;-</span> <span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/extract.html">extract</a></span><span class="op">(</span><span class="va">cell_id</span>, </span>
<span>          remove <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"cell_type"</span>, <span class="st">"cell_number"</span><span class="op">)</span>,</span>
<span>          <span class="st">"([a-zA-Z]{4})_([0-9]{3})"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What this code does is take what is in the <code>cell_id</code> column (something like <code>Prog_001</code> or <code>HSPC_001</code>) and split it into two columns (â€œcell_typeâ€ and â€œcell_numberâ€). The reason why we want to do that is to colour the points by cell type. We would not want to use <code>cell_id</code> to colour the points because each cell id is unique and that would be 1000+ colours. The last argument in the <code><a href="https://tidyr.tidyverse.org/reference/extract.html">extract()</a></code> function is the pattern to match described with a <a href="https://en.wikipedia.org/wiki/Regular_expression">regular expression</a>. Three patterns are being matched, and two of those are in brackets meaning they are kept to fill the two new columns.</p>
<p>The first pattern is <code>([a-zA-Z]{4})</code></p>
<ul>
<li>it is brackets because we want to keep it and put it in <code>cell_type</code>
</li>
<li>
<code>[a-zA-Z]</code> means any lower (<code>a-z</code>) or upper case letter (<code>A-Z</code>).</li>
<li>The square brackets means any of the characters in the square brackets will be matched</li>
<li>
<code>{4}</code> means 4 of them.</li>
</ul>
<p>So the first pattern inside the first (â€¦) will match exactly 4 upper or lower case letters (like Prog or HSPC)</p>
<p>The second pattern is <code>_</code> to match the underscore in every cell id that separates the cell type from the number. It is not in brackets because we do not want to keep it.</p>
<p>The third pattern is <code>([0-9]{3})</code></p>
<ul>
<li>
<code>[0-9]</code> means any number</li>
<li>
<code>{3}</code> means 3 of them.</li>
</ul>
<p>So the second pattern inside the second (â€¦) will match exactly 3 numbers (like 001 or 851).</p>
<p><strong>Important</strong>: Prog and HPSC have 4 letters. The column names, LT.HSPC_ have 6 characters and includes a dot. You will need to adjust the regex when make comparison between LT-HSC and other cell types. The pattern to match the <code>LT.HSC</code> as well as the <code>Prog</code> and <code>HSPC</code> is <code>([a-zA-Z.]{4,6})</code>. Note the dot inside the square brackets and numbers meaning 4 or 6 of. The pattern to match the underscore and the cell number is the same.</p>
<p>The top of the dataframe should look like this (but with more decimal places)</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 10%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 10%">
<col style="width: 11%">
<col style="width: 13%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">PC1</th>
<th style="text-align: right;">PC2</th>
<th style="text-align: right;">PC3</th>
<th style="text-align: right;">PC4</th>
<th style="text-align: right;">PC5</th>
<th style="text-align: right;">PC6</th>
<th style="text-align: right;">PC7</th>
<th style="text-align: right;">PC8</th>
<th style="text-align: left;">cell_id</th>
<th style="text-align: left;">cell_type</th>
<th style="text-align: left;">cell_number</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">HSPC_001</td>
<td style="text-align: right;">-2.62</td>
<td style="text-align: right;">-15.39</td>
<td style="text-align: right;">-7.94</td>
<td style="text-align: right;">-4.87</td>
<td style="text-align: right;">-2.52</td>
<td style="text-align: right;">4.29</td>
<td style="text-align: right;">-1.69</td>
<td style="text-align: right;">4.45</td>
<td style="text-align: left;">HSPC_001</td>
<td style="text-align: left;">HSPC</td>
<td style="text-align: left;">001</td>
</tr>
<tr class="even">
<td style="text-align: left;">HSPC_002</td>
<td style="text-align: right;">0.70</td>
<td style="text-align: right;">-15.22</td>
<td style="text-align: right;">0.13</td>
<td style="text-align: right;">3.74</td>
<td style="text-align: right;">-6.47</td>
<td style="text-align: right;">-1.70</td>
<td style="text-align: right;">0.83</td>
<td style="text-align: right;">-3.35</td>
<td style="text-align: left;">HSPC_002</td>
<td style="text-align: left;">HSPC</td>
<td style="text-align: left;">002</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HSPC_003</td>
<td style="text-align: right;">-15.52</td>
<td style="text-align: right;">7.80</td>
<td style="text-align: right;">0.53</td>
<td style="text-align: right;">-4.04</td>
<td style="text-align: right;">0.10</td>
<td style="text-align: right;">-1.48</td>
<td style="text-align: right;">-0.32</td>
<td style="text-align: right;">-7.40</td>
<td style="text-align: left;">HSPC_003</td>
<td style="text-align: left;">HSPC</td>
<td style="text-align: left;">003</td>
</tr>
<tr class="even">
<td style="text-align: left;">HSPC_004</td>
<td style="text-align: right;">-4.11</td>
<td style="text-align: right;">-7.04</td>
<td style="text-align: right;">0.45</td>
<td style="text-align: right;">10.58</td>
<td style="text-align: right;">-3.09</td>
<td style="text-align: right;">-8.30</td>
<td style="text-align: right;">-6.01</td>
<td style="text-align: right;">2.27</td>
<td style="text-align: left;">HSPC_004</td>
<td style="text-align: left;">HSPC</td>
<td style="text-align: left;">004</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HSPC_006</td>
<td style="text-align: right;">-7.89</td>
<td style="text-align: right;">5.15</td>
<td style="text-align: right;">-2.95</td>
<td style="text-align: right;">12.45</td>
<td style="text-align: right;">4.77</td>
<td style="text-align: right;">4.63</td>
<td style="text-align: right;">5.46</td>
<td style="text-align: right;">1.97</td>
<td style="text-align: left;">HSPC_006</td>
<td style="text-align: left;">HSPC</td>
<td style="text-align: left;">006</td>
</tr>
<tr class="even">
<td style="text-align: left;">HSPC_008</td>
<td style="text-align: right;">-10.29</td>
<td style="text-align: right;">-8.06</td>
<td style="text-align: right;">-8.33</td>
<td style="text-align: right;">-2.69</td>
<td style="text-align: right;">1.86</td>
<td style="text-align: right;">-2.42</td>
<td style="text-align: right;">0.03</td>
<td style="text-align: right;">-2.01</td>
<td style="text-align: left;">HSPC_008</td>
<td style="text-align: left;">HSPC</td>
<td style="text-align: left;">008</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The next task is to plot PC2 against PC1 and colour by cell type. This is just a scatterplot so we can use <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point()</a></code>. We will use colour to indicate the cell type.</p>
<p>ğŸ¬ Plot PC2 against PC1 and colour by cell type:</p>
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, </span>
<span>             colour <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>There is a good clustering of cell types but plenty of overlap. You can also try plotting PC3 or PC4 (or others).</p>
<p>I prefer to customise the colours. I especially like the viridis colour scales which provide colour scales that are perceptually uniform in both colour and black-and-white. They are also designed to be perceived by viewers with common forms of colour blindness. See <a href="https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html#introduction">Introduction to viridis</a> for more information.</p>
<p><code>ggplot</code> provides functions to access the viridis scales. Here I use <code><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_d()</a></code>. The d stands for discrete used because cell type is a discrete variable. The function <code><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_fill_viridis_c()</a></code> would be used for continuous data. Iâ€™ve used the default â€œviridisâ€ (or â€œDâ€) option (do ?scale_fill_viridis_d for all the options) and used the <code>begin</code> and <code>end</code> arguments to control the range of colour - I have set the range to be from 0.15 to 0.95 the avoid the strongest contrast. I have also set the <code>name</code> argument to NULL because that the legend refers to cell types is obvious.</p>
<p>ğŸ¬ Plot PC2 against PC1 and colour by cell type:</p>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_labelled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, </span>
<span>             colour <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_colour_viridis_d</a></span><span class="op">(</span>end <span class="op">=</span> <span class="fl">0.95</span>, begin <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>                         name <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now go to Volcano plots for <a href="#stem-cells-3">ğŸ­ Stem cells</a></p>
</section></section><section id="visualise-all-the-results-with-a-volcano-plot" class="level1"><h1>Visualise all the results with a volcano plot</h1>
<p>A volcano plot is a scatterplot that shows statistical significance (p-value) versus magnitude of change (fold change). As the <a href="study_before_workshop.html">independent study</a> explained, we plot -log<sub>10</sub>(adjusted <em>p</em>-value) on the <em>y</em>-axis so that the most significant genes are uppermost.</p>
<section id="arabidopsis-2" class="level2"><h2 class="anchored" data-anchor-id="arabidopsis-2">ğŸ„ <em>Arabidopsis</em>
</h2>
<p>We will add a column to the results dataframe that contains the -log<sub>10</sub>(padj). You could perform this transformation within the plot command without adding a column to the data if you prefer.</p>
<p>ğŸ¬ Add a column to the results dataframe that contains the -log<sub>10</sub>(padj):</p>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_results</span> <span class="op">&lt;-</span> <span class="va">wild_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>log10_padj <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Create a volcano plot of the results:</p>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_padj</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Our dashed lines are at -log<sub>10</sub>(0.05) and log<sub>2</sub>(2) and log<sub>2</sub>(-2) to make more clear which genes (points) are significantly different between the copper sufficient and deficient conditions samples and have a fold change of at least 2.</p>
<p>In most cases, people colour the points to show that the quadrants. I like to add columns to the dataframe to indicate if the gene is significant and if the fold change is large and use those variables in the plot.</p>
<p>ğŸ¬ Add columns to the results dataframe to indicate if the gene is significant and if the fold change is large:</p>
<div class="cell">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_results</span> <span class="op">&lt;-</span> <span class="va">wild_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>sig <span class="op">=</span> <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span>,</span>
<span>         bigfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The use of <code><a href="https://rdrr.io/r/base/MathFun.html">abs()</a></code> (absolute) means genes with a fold change of at least 2 in either direction will be considered to have a large fold change.</p>
<p>Now we can colour the points by these new columns. I use <code><a href="https://rdrr.io/r/base/interaction.html">interaction()</a></code> to create four categories:</p>
<ul>
<li>not significant and not large fold change (FF)</li>
<li>significant and not large fold change (TF)</li>
<li>not significant and large fold (FT)</li>
<li>significant and large fold change (TT)</li>
</ul>
<p>And I use <code><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual()</a></code> to set the colours for these categories.</p>
<p>NOTE: there are no â€œsignificant and not large fold change (TF)â€ in this case.This means we do not need four colours. I have put the â€œpinkâ€ colour in the code but commented it out.</p>
<p>ğŸ¬ Create a volcano plot of the results with the points coloured by significance and fold change:</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_padj</span>, </span>
<span>             colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">sig</span>, <span class="va">bigfc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gray"</span>, </span>
<span>                                <span class="co"># "pink",</span></span>
<span>                                 <span class="st">"gray30"</span>,</span>
<span>                                 <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>For exploring the data, I like add labels to all the significant genes with a large fold change so I can very quickly identity them. The <code>ggrepel</code> package has a function <code><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel()</a></code> that is useful for adding labels so that they donâ€™t overlap.</p>
<p>ğŸ¬ Load the package:</p>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/">ggrepel</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Add labels to the significant genes with a large fold change:</p>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_padj</span>, </span>
<span>             colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">sig</span>, <span class="va">bigfc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gray"</span>, </span>
<span>                               <span class="co">#  "pink",</span></span>
<span>                                 <span class="st">"gray30"</span>,</span>
<span>                                 <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">wild_results</span> <span class="op">|&gt;</span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">bigfc</span> <span class="op">==</span> <span class="cn">TRUE</span>, <span class="va">sig</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">external_gene_name</span><span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                  max.overlaps <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in geom_text_repel(data = filter(wild_results, bigfc == TRUE, sig == : could not find function "geom_text_repel"</code></pre>
</div>
</div>
<p>Notice that I have used <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> label only the genes that are both significant and have a large fold change. In systems you are familiar with, this labelling is very informative and can help you quickly identify common themes. Key to interpreting the volcano plot is to remember that positive fold changes means the gene is up-regulated in the sufficient conditions and negative fold changes means the gene is down-regulated (i.e., higher in the deficient). This was determined by the order of the treatments in the <a href="../week-4/workshop.html#differential-expression-analysis-2">contrast used in the DESeq2 analysis</a></p>
<p>If you do forget which way round you did the comparison, you can always examine the results dataframe to see which of the treatments seem to be higher for the positive fold changes.</p>
<p>When you have a gene of interest, you may wish to label it, and only it, on the plot. This is done in the same way except that you filter the data to only include the gene of interest. I have used and then use <code><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_label_repel()</a></code> rather than <code><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel()</a></code> to put the label in a box and nudged itâ€™s position to get a line connecting the point and the label. I have also increased the size of the point.</p>
<p>ğŸ¬ Add a label to one gene of interest (hoxb9.S) and :</p>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">wild_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_padj</span>, </span>
<span>             colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">sig</span>, <span class="va">bigfc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gray"</span>, </span>
<span>                               <span class="co">#  "pink",</span></span>
<span>                                 <span class="st">"gray30"</span>,</span>
<span>                                 <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_label_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">wild_results</span> <span class="op">|&gt;</span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">external_gene_name</span> <span class="op">==</span> <span class="st">"FRO4"</span><span class="op">)</span>,</span>
<span>                  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">external_gene_name</span><span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                  nudge_x <span class="op">=</span> <span class="fl">.5</span>,</span>
<span>                  nudge_y <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">wild_results</span> <span class="op">|&gt;</span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">external_gene_name</span> <span class="op">==</span> <span class="st">"FRO4"</span><span class="op">)</span>,</span>
<span>                size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in geom_label_repel(data = filter(wild_results, external_gene_name == : could not find function "geom_label_repel"</code></pre>
</div>
</div>
<p>Should you want to label more than one gene, you will need to use (for example): <code>filter(external_gene_name %in% c("FRO4", "FRO5"))</code></p>
<p>Now go to <a href="#save-your-plots">Save your plots</a></p>
</section><section id="leishmania-2" class="level2"><h2 class="anchored" data-anchor-id="leishmania-2">ğŸ’‰ <em>Leishmania</em>
</h2>
<p>We will add a column to the results dataframe that contains the -log<sub>10</sub>(padj). You could perform this transformation within the plot command without adding a column to the data if you prefer.</p>
<p>ğŸ¬ Add a column to the results dataframe that contains the -log<sub>10</sub>(padj):</p>
<div class="cell">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_results</span> <span class="op">&lt;-</span> <span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>log10_padj <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Create a volcano plot of the results:</p>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_padj</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-60-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Our dashed lines are at -log<sub>10</sub>(0.05) and log<sub>2</sub>(2) and log<sub>2</sub>(-2) to make more clear which genes (points) are significantly different between the life stages and have a fold change of at least 2. Whilst we have very many genes that are significant, we have fewer that are significant and have a large fold change.</p>
<p>In most cases, people colour the points to show that the quadrants. I like to add columns to the dataframe to indicate if the gene is significant and if the fold change is large and use those variables in the plot.</p>
<p>ğŸ¬ Add columns to the results dataframe to indicate if the gene is significant and if the fold change is large:</p>
<div class="cell">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_results</span> <span class="op">&lt;-</span> <span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>sig <span class="op">=</span> <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span>,</span>
<span>         bigfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The use of <code><a href="https://rdrr.io/r/base/MathFun.html">abs()</a></code> (absolute) means genes with a fold change of at least 2 in either direction will be considered to have a large fold change.</p>
<p>Now we can colour the points by these new columns. I use <code><a href="https://rdrr.io/r/base/interaction.html">interaction()</a></code> to create four categories:</p>
<ul>
<li>not significant and not large fold change (FF)</li>
<li>significant and not large fold change (TF)</li>
<li>not significant and large fold (FT)</li>
<li>significant and large fold change (TT)</li>
</ul>
<p>And I use <code><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual()</a></code> to set the colours for these categories.</p>
<p>ğŸ¬ Create a volcano plot of the results with the points coloured by significance and fold change:</p>
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_padj</span>, </span>
<span>             colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">sig</span>, <span class="va">bigfc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gray"</span>, </span>
<span>                                 <span class="st">"pink"</span>,</span>
<span>                                 <span class="st">"gray30"</span>,</span>
<span>                                 <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-62-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>For exploring the data, I like add labels to all the significant genes with a large fold change so I can very quickly identity them. The <code>ggrepel</code> package has a function <code><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel()</a></code> that is useful for adding labels so that they donâ€™t overlap.</p>
<p>ğŸ¬ Load the package:</p>
<div class="cell">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/">ggrepel</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Add labels to the significant genes with a large fold change:</p>
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_padj</span>, </span>
<span>             colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">sig</span>, <span class="va">bigfc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gray"</span>, </span>
<span>                                 <span class="st">"pink"</span>,</span>
<span>                                 <span class="st">"gray30"</span>,</span>
<span>                                 <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">bigfc</span> <span class="op">==</span> <span class="cn">TRUE</span>, <span class="va">sig</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">description</span><span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                  max.overlaps <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in geom_text_repel(data = filter(pro_meta_results, bigfc == TRUE, : could not find function "geom_text_repel"</code></pre>
</div>
</div>
<p>Notice that I have used <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> label only the genes that are both significant and have a large fold change. In systems you are familiar with, this labelling is very informative and can help you quickly identify common themes. However, in this case, we do not have good annotation for the genes. We can label only with the gene id or the description. Many of the descriptions are</p>
<p>ğŸ¬ Add labels to the significant genes with a large fold change only where description doesnâ€™t contain â€œhypotheticalâ€ or â€œunspecified :</p>
<div class="cell">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_padj</span>, </span>
<span>             colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">sig</span>, <span class="va">bigfc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gray"</span>, </span>
<span>                                 <span class="st">"pink"</span>,</span>
<span>                                 <span class="st">"gray30"</span>,</span>
<span>                                 <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">bigfc</span> <span class="op">==</span> <span class="cn">TRUE</span>, <span class="va">sig</span> <span class="op">==</span> <span class="cn">TRUE</span>,</span>
<span>                           <span class="op">!</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html">str_detect</a></span><span class="op">(</span><span class="va">description</span>, <span class="st">"hypothetical|unspecified"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">description</span><span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                  max.overlaps <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in geom_text_repel(data = filter(pro_meta_results, bigfc == TRUE, : could not find function "geom_text_repel"</code></pre>
</div>
</div>
<p>Key to interpreting the volcano plot is to remember that positive fold changes means the gene is up-regulated in the procyclic stage and negative fold changes means the gene is down-regulated (i.e., higher in the metacyclic). This was determined by the order of the treatments in the <a href="../week-4/workshop.html#differential-expression-analysis-3">contrast used in the DESeq2 analysis</a></p>
<p>If you do forget which way round you did the comparison, you can always examine the results dataframe to see which of the treatments seem to be higher for the positive fold changes.</p>
<p>When you have a gene of interest, you may wish to label it, and only it, on the plot. This is done in the same way except that you filter the data to only include the gene of interest. I have used and then use <code><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_label_repel()</a></code> rather than <code><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel()</a></code> to put the label in a box and nudged itâ€™s position to get a line connecting the point and the label. I have also increased the size of the point.</p>
<p>ğŸ¬ Add a label to one gene of interest (elongation factor 1-alpha) and :</p>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_padj</span>, </span>
<span>             colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">sig</span>, <span class="va">bigfc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gray"</span>, </span>
<span>                                 <span class="st">"pink"</span>,</span>
<span>                                 <span class="st">"gray30"</span>,</span>
<span>                                 <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_label_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">description</span> <span class="op">==</span> <span class="st">"elongation factor 1-alpha"</span><span class="op">)</span>,</span>
<span>                  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">description</span><span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                  nudge_x <span class="op">=</span> <span class="fl">.5</span>,</span>
<span>                  nudge_y <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">pro_meta_results</span> <span class="op">|&gt;</span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">description</span> <span class="op">==</span> <span class="st">"elongation factor 1-alpha"</span><span class="op">)</span>,</span>
<span>                size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in geom_label_repel(data = filter(pro_meta_results, description == : could not find function "geom_label_repel"</code></pre>
</div>
</div>
<p>Should you want to label more than one gene, you will need to use (for example): <code>filter(description %in% c("elongation factor 1-alpha", "ADP/ATP translocase 1 putative"))</code></p>
<p>Now go to <a href="#save-your-plots">Save your plots</a></p>
</section><section id="stem-cells-3" class="level2"><h2 class="anchored" data-anchor-id="stem-cells-3">ğŸ­ Stem cells</h2>
<p>We will add a column to the results dataframe that contains the -log<sub>10</sub>(FDR). You could perform this transformation within the plot command without adding a column to the data if you prefer.</p>
<p>ğŸ¬ Add a column to the results dataframe that contains the -log<sub>10</sub>(FDR):</p>
<div class="cell">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_results</span> <span class="op">&lt;-</span> <span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>log10_FDR <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">FDR</span><span class="op">)</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Create a volcano plot of the results:</p>
<div class="cell">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">summary.logFC</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_FDR</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-68-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Our dashed lines are at -log<sub>10</sub>(0.05) and log<sub>2</sub>(2) and log<sub>2</sub>(-2) to make more clear which genes (points) are significantly different between the cell types and have a fold change of at least 2.</p>
<p>In most cases, people colour the points to show that the quadrants. I like to add columns to the dataframe to indicate if the gene is significant and if the fold change is large and use those variables in the plot.</p>
<p>ğŸ¬ Add columns to the results dataframe to indicate if the gene is significant and if the fold change is large:</p>
<div class="cell">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_results</span> <span class="op">&lt;-</span> <span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>sig <span class="op">=</span> <span class="va">FDR</span> <span class="op">&lt;=</span> <span class="fl">0.05</span>,</span>
<span>         bigfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">summary.logFC</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The use of <code><a href="https://rdrr.io/r/base/MathFun.html">abs()</a></code> (absolute) means genes with a fold change of at least 2 in either direction will be considered to have a large fold change.</p>
<p>Now we can colour the points by these new columns. I use <code><a href="https://rdrr.io/r/base/interaction.html">interaction()</a></code> to create four categories:</p>
<ul>
<li>not significant and not large fold change (FF)</li>
<li>significant and not large fold change (TF)</li>
<li>not significant and large fold (FT)</li>
<li>significant and large fold change (TT)</li>
</ul>
<p>And I use <code><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual()</a></code> to set the colours for these categories.</p>
<p>NOTE: there are no â€œnot significant and large fold change (FT)â€ in this case. This means we do not need four colours. I have put the â€œgray30â€ colour in the code but commented it out.</p>
<p>ğŸ¬ Create a volcano plot of the results with the points coloured by significance and fold change:</p>
<div class="cell">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">summary.logFC</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_FDR</span>, </span>
<span>             colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">sig</span>, <span class="va">bigfc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gray"</span>, </span>
<span>                                 <span class="st">"pink"</span>,</span>
<span>                                <span class="co"># "gray30",</span></span>
<span>                                 <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-70-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>For exploring the data, I like add labels to all the significant genes with a large fold change so I can very quickly identity them. The <code>ggrepel</code> package has a function <code><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel()</a></code> that is useful for adding labels so that they donâ€™t overlap.</p>
<p>ğŸ¬ Load the package:</p>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/">ggrepel</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ğŸ¬ Add labels to the significant genes with a large fold change:</p>
<div class="cell">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">summary.logFC</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_FDR</span>, </span>
<span>             colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">sig</span>, <span class="va">bigfc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gray"</span>, </span>
<span>                                 <span class="st">"pink"</span>,</span>
<span>                              <span class="co">#   "gray30",</span></span>
<span>                                 <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">bigfc</span> <span class="op">==</span> <span class="cn">TRUE</span>, <span class="va">sig</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">external_gene_name</span><span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                  max.overlaps <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-72-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Notice that I have used <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> label only the genes that are both significant and have a large fold change. In systems you are familiar with, this labelling is very informative and can help you quickly identify common themes. Key to interpreting the volcano plot is to remember that positive fold changes means the gene is up-regulated in the Prog and negative fold changes means the gene is down-regulated (i.e., higher HSPC). This was determined by us choosing <a href="../week-4/workshop.html#differential-expression-analysis-4">the results_hspc_prog$prog dataframe from the list object</a></p>
<p>If you do forget which way round you did the comparison, you can always examine the gene summary dataframe to see which of the treatments seem to be higher for the positive fold changes.</p>
<p>When you have a gene of interest, you may wish to label it on the plot. This is done in the same way except that you filter the data to only include the gene of interest. I have used and then use <code><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_label_repel()</a></code> rather than <code><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel()</a></code> to put the label in a box and nudged itâ€™s position to get a line connecting the point and the label. I have also increased the size of the point.</p>
<p>ğŸ¬ Add a label to one gene of interest (Procr) and :</p>
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">summary.logFC</span>, </span>
<span>             y <span class="op">=</span> <span class="va">log10_FDR</span>, </span>
<span>             colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">sig</span>, <span class="va">bigfc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gray"</span>, </span>
<span>                                 <span class="st">"pink"</span>,</span>
<span>                                <span class="co"># "gray30",</span></span>
<span>                                 <span class="st">"deeppink"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_label_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">external_gene_name</span> <span class="op">==</span> <span class="st">"Procr"</span><span class="op">)</span>,</span>
<span>                  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">external_gene_name</span><span class="op">)</span>,</span>
<span>                  size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                  nudge_x <span class="op">=</span> <span class="fl">.5</span>,</span>
<span>                  nudge_y <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">hspc_prog_results</span> <span class="op">|&gt;</span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">external_gene_name</span> <span class="op">==</span> <span class="st">"Procr"</span><span class="op">)</span>,</span>
<span>                size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop_files/figure-html/unnamed-chunk-73-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Should you want to label more than one gene, you will need to use (for example): <code>filter(external_gene_name %in% c("Procr", "Emb"))</code></p>
<p>Now go to <a href="#save-your-plots">Save your plots</a></p>
</section></section><section id="save-your-plots" class="level1"><h1>Save your plots</h1>
<p>Once you have finished designing your plots, you should save them using <code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave()</a></code>. You will want to assign the plot to a variable and use code similar to this:</p>
<div class="cell">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="st">"figures/my-informative-file-name.png"</span>,</span>
<span>       plot <span class="op">=</span> <span class="va">vol</span>,</span>
<span>       height <span class="op">=</span> <span class="fl">4.5</span>, </span>
<span>       width <span class="op">=</span> <span class="fl">4.5</span>,</span>
<span>       units <span class="op">=</span> <span class="st">"in"</span>,</span>
<span>       device <span class="op">=</span> <span class="st">"png"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>I recommend saving your figure in the approximate size it will appear in your report, that is, donâ€™t resize it in word/google docs. This because the font will then be an appropriate size for the report. For journals, we often have to creat figure files of very high resolution. You can do this by increasing the <code>height</code> and <code>width</code> and then resizing the figure in the document. however, this requires resizing the all font and lines in the figure. I also recommend saving it as a png file as this is a lossless format.</p>
</section><section id="look-after-future-you" class="level1"><h1>ğŸ¤— Look after future you!</h1>
<p>ğŸ¬ Go through your script and tidy up. I would suggest restarting R and trying to run the full pipeline from start to finish. You might need to:</p>
<ul>
<li>collect together library statements at the top of the script</li>
<li>remove code that you needed to start today but which wouldnâ€™t be needed running the script from the top (e.g., importing the results)</li>
<li>edit your comments for clarity</li>
<li>rename variables for consistency or clarity</li>
<li>remove house keeping or exploratory code</li>
<li>restyle code, add code section headers etc</li>
</ul></section><section id="finished" class="level1"><h1>ğŸ¥³ Finished</h1>
<p>Well Done!</p>
</section><section id="independent-study-following-the-workshop" class="level1"><h1>Independent study following the workshop</h1>
<p><a href="../../transcriptomics/week-5/study_after_workshop.html">Consolidate</a></p>
</section><section id="the-code-file" class="level1"><h1>The Code file</h1>
<p>These contain all the code needed in the workshop even where it is not visible on the webpage.</p>
<p>The <a href="../../transcriptomics/week-5/workshop.html">workshop.qmd</a> file is the file I use to compile the practical. Qmd stands for Quarto markdown. It allows code and ordinary text to be interleaved to produce well-formatted reports including webpages. Right-click on the link and choose Save-As to download. You will be able to open the Qmd file in RStudio. Alternatively, <a href="https://github.com/3mmaRand/BIO00088H-data/blob/main/transcriptomics/week-5/workshop.qmd">View in Browser</a>. Coding and thinking answers are marked with <code>#---CODING ANSWER---</code> and <code>#---THINKING ANSWER---</code></p>
<p>Pages made with R <span class="citation" data-cites="R-core">(<a href="#ref-R-core" role="doc-biblioref">R Core Team 2024</a>)</span>, Quarto <span class="citation" data-cites="Allaire_Quarto_2024">(<a href="#ref-Allaire_Quarto_2024" role="doc-biblioref">Allaire et al. 2024</a>)</span>, <code>knitr</code> <span class="citation" data-cites="knitr1 knitr2 knitr3">(<a href="#ref-knitr1" role="doc-biblioref">Xie 2024</a>, <a href="#ref-knitr2" role="doc-biblioref">2015</a>, <a href="#ref-knitr3" role="doc-biblioref">2014</a>)</span>, <code>kableExtra</code> <span class="citation" data-cites="kableExtra">(<a href="#ref-kableExtra" role="doc-biblioref">Zhu 2021</a>)</span></p>
</section><section id="references" class="level1">


<!-- -->


</section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Allaire_Quarto_2024" class="csl-entry" role="listitem">
Allaire, J. J., Charles Teague, Carlos Scheidegger, Yihui Xie, and Christophe Dervieux. 2024. <span>â€œ<span>Quarto</span>.â€</span> <a href="https://doi.org/10.5281/zenodo.5960048">https://doi.org/10.5281/zenodo.5960048</a>.
</div>
<div id="ref-R-core" class="csl-entry" role="listitem">
R Core Team. 2024. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-conflicted" class="csl-entry" role="listitem">
Wickham, Hadley. 2023. <em>Conflicted: An Alternative Conflict Resolution Strategy</em>. <a href="https://conflicted.r-lib.org/">https://conflicted.r-lib.org/</a>.
</div>
<div id="ref-tidyverse" class="csl-entry" role="listitem">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy Dâ€™Agostino McGowan, Romain FranÃ§ois, Garrett Grolemund, et al. 2019. <span>â€œWelcome to the <span class="nocase">tidyverse</span>.â€</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-knitr3" class="csl-entry" role="listitem">
Xie, Yihui. 2014. <span>â€œKnitr: A Comprehensive Tool for Reproducible Research in <span>R</span>.â€</span> In <em>Implementing Reproducible Computational Research</em>, edited by Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman; Hall/CRC.
</div>
<div id="ref-knitr2" class="csl-entry" role="listitem">
â€”â€”â€”. 2015. <em>Dynamic Documents with <span>R</span> and Knitr</em>. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-knitr1" class="csl-entry" role="listitem">
â€”â€”â€”. 2024. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in r</em>. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-kableExtra" class="csl-entry" role="listitem">
Zhu, Hao. 2021. <span>â€œkableExtra: Construct Complex Table with â€™Kableâ€™ and Pipe Syntax.â€</span> <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>.
</div>
</div></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("^(?:http:|https:)\/\/3mmarand\.github\.io\/BIO00088H-data");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb64" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Workshop"</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="an">subtitle:</span><span class="co"> "Transcriptomics 3: Visualising"</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Emma Rand"</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a><span class="an">toc:</span><span class="co"> true</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a><span class="an">toc-depth:</span><span class="co"> 4</span></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a><span class="an">toc-location:</span><span class="co"> right</span></span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a><span class="co">  echo: true</span></span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a><span class="co">  include: true</span></span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a><span class="co">  error: true</span></span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a><span class="an">bibliography:</span><span class="co"> ../../references.bib</span></span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span><span class="co"> </span></span>
<span id="cb64-14"><a href="#cb64-14" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown: </span></span>
<span id="cb64-15"><a href="#cb64-15" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb64-16"><a href="#cb64-16" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb64-17"><a href="#cb64-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-18"><a href="#cb64-18" aria-hidden="true" tabindex="-1"></a><span class="fu"># Introduction</span></span>
<span id="cb64-19"><a href="#cb64-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-20"><a href="#cb64-20" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session overview</span></span>
<span id="cb64-21"><a href="#cb64-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-22"><a href="#cb64-22" aria-hidden="true" tabindex="-1"></a>In the workshop, you will learn how to conduct and plot a Principle</span>
<span id="cb64-23"><a href="#cb64-23" aria-hidden="true" tabindex="-1"></a>Component Analysis (PCA) as well as how to create a nicely formatted</span>
<span id="cb64-24"><a href="#cb64-24" aria-hidden="true" tabindex="-1"></a>Volcano plot. You will also save significant genes to file to make it</span>
<span id="cb64-25"><a href="#cb64-25" aria-hidden="true" tabindex="-1"></a>easier to identify genes of interest. ~~and perform Gene Ontology (GO)</span>
<span id="cb64-26"><a href="#cb64-26" aria-hidden="true" tabindex="-1"></a>term enrichment analysis.~~</span>
<span id="cb64-27"><a href="#cb64-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-28"><a href="#cb64-28" aria-hidden="true" tabindex="-1"></a><span class="fu"># Set up</span></span>
<span id="cb64-29"><a href="#cb64-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-30"><a href="#cb64-30" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ„ *Arabidopsis*</span></span>
<span id="cb64-31"><a href="#cb64-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-32"><a href="#cb64-32" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Open the <span class="in">`arabi-88H`</span> RStudio Project and the <span class="in">`wildsuf-wilddef.R`</span></span>
<span id="cb64-33"><a href="#cb64-33" aria-hidden="true" tabindex="-1"></a>script.</span>
<span id="cb64-34"><a href="#cb64-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-35"><a href="#cb64-35" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ’‰ *Leishmania*</span></span>
<span id="cb64-36"><a href="#cb64-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-37"><a href="#cb64-37" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Open the <span class="in">`leish-88H`</span> RStudio Project and the <span class="in">`pro-meta.R`</span> script.</span>
<span id="cb64-38"><a href="#cb64-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-39"><a href="#cb64-39" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ­ Stem cells</span></span>
<span id="cb64-40"><a href="#cb64-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-41"><a href="#cb64-41" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Open the <span class="in">`mice-88H`</span> RStudio Project and the <span class="in">`hspc-prog.R`</span> script.</span>
<span id="cb64-42"><a href="#cb64-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-43"><a href="#cb64-43" aria-hidden="true" tabindex="-1"></a><span class="fu">## Everyone</span></span>
<span id="cb64-44"><a href="#cb64-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-45"><a href="#cb64-45" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Make a new folder <span class="in">`figures`</span> in the project directory.</span>
<span id="cb64-46"><a href="#cb64-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-47"><a href="#cb64-47" aria-hidden="true" tabindex="-1"></a>This is where we will save our figure files</span>
<span id="cb64-48"><a href="#cb64-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-49"><a href="#cb64-49" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load **`tidyverse`** [@tidyverse] and **`conflicted`** <span class="co">[</span><span class="ot">@conflicted</span><span class="co">]</span>.</span>
<span id="cb64-50"><a href="#cb64-50" aria-hidden="true" tabindex="-1"></a>You most likely have this code at the top of your script already.</span>
<span id="cb64-51"><a href="#cb64-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-54"><a href="#cb64-54" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-55"><a href="#cb64-55" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb64-56"><a href="#cb64-56" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(conflicted)</span>
<span id="cb64-57"><a href="#cb64-57" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-58"><a href="#cb64-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-59"><a href="#cb64-59" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb64-60"><a href="#cb64-60" aria-hidden="true" tabindex="-1"></a><span class="in">â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€</span></span>
<span id="cb64-61"><a href="#cb64-61" aria-hidden="true" tabindex="-1"></a><span class="in">âœ” dplyr     1.1.3     âœ” readr     2.1.4</span></span>
<span id="cb64-62"><a href="#cb64-62" aria-hidden="true" tabindex="-1"></a><span class="in">âœ” forcats   1.0.0     âœ” stringr   1.5.0</span></span>
<span id="cb64-63"><a href="#cb64-63" aria-hidden="true" tabindex="-1"></a><span class="in">âœ” ggplot2   3.4.3     âœ” tibble    3.2.1</span></span>
<span id="cb64-64"><a href="#cb64-64" aria-hidden="true" tabindex="-1"></a><span class="in">âœ” lubridate 1.9.3     âœ” tidyr     1.3.0</span></span>
<span id="cb64-65"><a href="#cb64-65" aria-hidden="true" tabindex="-1"></a><span class="in">âœ” purrr     1.0.2     </span></span>
<span id="cb64-66"><a href="#cb64-66" aria-hidden="true" tabindex="-1"></a><span class="in">â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€</span></span>
<span id="cb64-67"><a href="#cb64-67" aria-hidden="true" tabindex="-1"></a><span class="in">âœ– dplyr::filter() masks stats::filter()</span></span>
<span id="cb64-68"><a href="#cb64-68" aria-hidden="true" tabindex="-1"></a><span class="in">âœ– dplyr::lag()    masks stats::lag()</span></span>
<span id="cb64-69"><a href="#cb64-69" aria-hidden="true" tabindex="-1"></a><span class="in">â„¹ Use the conflicted package to force all conflicts to become errors</span></span>
<span id="cb64-70"><a href="#cb64-70" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-71"><a href="#cb64-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-72"><a href="#cb64-72" aria-hidden="true" tabindex="-1"></a>I recommend you set the **`dplyr`** versions of <span class="in">`filter()`</span> and</span>
<span id="cb64-73"><a href="#cb64-73" aria-hidden="true" tabindex="-1"></a><span class="in">`select()`</span> to use by default</span>
<span id="cb64-74"><a href="#cb64-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-75"><a href="#cb64-75" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Use the **`dplyr`** version of <span class="in">`filter()`</span> by default:</span>
<span id="cb64-76"><a href="#cb64-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-79"><a href="#cb64-79" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-80"><a href="#cb64-80" aria-hidden="true" tabindex="-1"></a><span class="fu">conflicts_prefer</span>(dplyr<span class="sc">::</span>filter)</span>
<span id="cb64-81"><a href="#cb64-81" aria-hidden="true" tabindex="-1"></a><span class="fu">conflicts_prefer</span>(dplyr<span class="sc">::</span>select)</span>
<span id="cb64-82"><a href="#cb64-82" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-83"><a href="#cb64-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-84"><a href="#cb64-84" aria-hidden="true" tabindex="-1"></a><span class="fu"># Import</span></span>
<span id="cb64-85"><a href="#cb64-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-86"><a href="#cb64-86" aria-hidden="true" tabindex="-1"></a><span class="fu">## Everyone</span></span>
<span id="cb64-87"><a href="#cb64-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-88"><a href="#cb64-88" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Import your results data. This should be a file in the <span class="in">`results`</span></span>
<span id="cb64-89"><a href="#cb64-89" aria-hidden="true" tabindex="-1"></a>folder called <span class="in">`xxxx_results.csv`</span> where xxxx indicates the comparison you</span>
<span id="cb64-90"><a href="#cb64-90" aria-hidden="true" tabindex="-1"></a>made.</span>
<span id="cb64-91"><a href="#cb64-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-94"><a href="#cb64-94" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-95"><a href="#cb64-95" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb64-96"><a href="#cb64-96" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-97"><a href="#cb64-97" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ„  Arabidopisis</span></span>
<span id="cb64-98"><a href="#cb64-98" aria-hidden="true" tabindex="-1"></a><span class="co"># import results data</span></span>
<span id="cb64-99"><a href="#cb64-99" aria-hidden="true" tabindex="-1"></a>wild_results <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"results/wild_results.csv"</span>)</span>
<span id="cb64-100"><a href="#cb64-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-101"><a href="#cb64-101" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-102"><a href="#cb64-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-105"><a href="#cb64-105" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-106"><a href="#cb64-106" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb64-107"><a href="#cb64-107" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-108"><a href="#cb64-108" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ’‰ leihsmania</span></span>
<span id="cb64-109"><a href="#cb64-109" aria-hidden="true" tabindex="-1"></a><span class="co"># import results data</span></span>
<span id="cb64-110"><a href="#cb64-110" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"results/pro_meta_results.csv"</span>)</span>
<span id="cb64-111"><a href="#cb64-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-112"><a href="#cb64-112" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-113"><a href="#cb64-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-116"><a href="#cb64-116" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-117"><a href="#cb64-117" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb64-118"><a href="#cb64-118" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-119"><a href="#cb64-119" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ­ stem cells</span></span>
<span id="cb64-120"><a href="#cb64-120" aria-hidden="true" tabindex="-1"></a><span class="co"># import results data</span></span>
<span id="cb64-121"><a href="#cb64-121" aria-hidden="true" tabindex="-1"></a>hspc_prog_results <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"results/hspc_prog_results.csv"</span>)</span>
<span id="cb64-122"><a href="#cb64-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-123"><a href="#cb64-123" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-124"><a href="#cb64-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-125"><a href="#cb64-125" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Remind yourself what is in the rows and columns and the structure of</span>
<span id="cb64-126"><a href="#cb64-126" aria-hidden="true" tabindex="-1"></a>the dataframes (perhaps using <span class="in">`glimpse()`</span>)</span>
<span id="cb64-127"><a href="#cb64-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-130"><a href="#cb64-130" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-131"><a href="#cb64-131" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-132"><a href="#cb64-132" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-133"><a href="#cb64-133" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ„  Arabidopisis</span></span>
<span id="cb64-134"><a href="#cb64-134" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(wild_results)</span>
<span id="cb64-135"><a href="#cb64-135" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-136"><a href="#cb64-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-137"><a href="#cb64-137" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb64-138"><a href="#cb64-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-139"><a href="#cb64-139" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- You have data with 13 columns and 26355 rows. The rows are the genes. --&gt;</span></span>
<span id="cb64-140"><a href="#cb64-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-141"><a href="#cb64-141" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- The columns are the normalised counts for for each of the 4 --&gt;</span></span>
<span id="cb64-142"><a href="#cb64-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-143"><a href="#cb64-143" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- samples, the mean of the normalised counts, the gene id, --&gt;</span></span>
<span id="cb64-144"><a href="#cb64-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-145"><a href="#cb64-145" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- the gene name and a description. --&gt;</span></span>
<span id="cb64-146"><a href="#cb64-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-147"><a href="#cb64-147" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- You also have the results of a statistical comparison between --&gt;</span></span>
<span id="cb64-148"><a href="#cb64-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-149"><a href="#cb64-149" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- the Cu sufficient and Cu deficient conditions, the log fold change,  --&gt;</span></span>
<span id="cb64-150"><a href="#cb64-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-151"><a href="#cb64-151" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- the test statistic, the p-value and the adjusted p-value --&gt;</span></span>
<span id="cb64-152"><a href="#cb64-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-155"><a href="#cb64-155" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-156"><a href="#cb64-156" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-157"><a href="#cb64-157" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-158"><a href="#cb64-158" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ’‰ Leishmania</span></span>
<span id="cb64-159"><a href="#cb64-159" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(pro_meta_results)</span>
<span id="cb64-160"><a href="#cb64-160" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-161"><a href="#cb64-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-162"><a href="#cb64-162" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb64-163"><a href="#cb64-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-164"><a href="#cb64-164" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- You have data with 14 columns and 8437 rows. The rows are the genes. --&gt;</span></span>
<span id="cb64-165"><a href="#cb64-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-166"><a href="#cb64-166" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- The columns are the normalised counts for for each of the 6 --&gt;</span></span>
<span id="cb64-167"><a href="#cb64-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-168"><a href="#cb64-168" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- samples, the mean of the normalised counts, the gene id, --&gt;</span></span>
<span id="cb64-169"><a href="#cb64-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-170"><a href="#cb64-170" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- and a description. --&gt;</span></span>
<span id="cb64-171"><a href="#cb64-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-172"><a href="#cb64-172" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- You also have the results of a statistical comparison between --&gt;</span></span>
<span id="cb64-173"><a href="#cb64-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-174"><a href="#cb64-174" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- the procyclic and metacyclic promastigote stages, the log  --&gt;</span></span>
<span id="cb64-175"><a href="#cb64-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-176"><a href="#cb64-176" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- fold change, the test statistic, the p-value and the  --&gt;</span></span>
<span id="cb64-177"><a href="#cb64-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-178"><a href="#cb64-178" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- adjusted p-value --&gt;</span></span>
<span id="cb64-179"><a href="#cb64-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-182"><a href="#cb64-182" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-183"><a href="#cb64-183" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-184"><a href="#cb64-184" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-185"><a href="#cb64-185" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ­ stem cells</span></span>
<span id="cb64-186"><a href="#cb64-186" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(hspc_prog_results)</span>
<span id="cb64-187"><a href="#cb64-187" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-188"><a href="#cb64-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-189"><a href="#cb64-189" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb64-190"><a href="#cb64-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-191"><a href="#cb64-191" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- You have data with 1507 columns and 280 rows. The rows are the genes. --&gt;</span></span>
<span id="cb64-192"><a href="#cb64-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-193"><a href="#cb64-193" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- The columns are the normalised counts for each of the cell --&gt;</span></span>
<span id="cb64-194"><a href="#cb64-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-195"><a href="#cb64-195" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- (1499 HPSC and and Prog cells), the ensembl_gene_id, a gene name and a --&gt;</span></span>
<span id="cb64-196"><a href="#cb64-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-197"><a href="#cb64-197" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- description. --&gt;</span></span>
<span id="cb64-198"><a href="#cb64-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-199"><a href="#cb64-199" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- You also have the results of a statistical comparison between --&gt;</span></span>
<span id="cb64-200"><a href="#cb64-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-201"><a href="#cb64-201" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- the HSPC and progenitor cell types, the log fold change (twice!), --&gt;</span></span>
<span id="cb64-202"><a href="#cb64-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-203"><a href="#cb64-203" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- the p-value and the adjusted p-value (FDR) --&gt;</span></span>
<span id="cb64-204"><a href="#cb64-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-205"><a href="#cb64-205" aria-hidden="true" tabindex="-1"></a>When we do PCA we will want to label the samples with their treatment</span>
<span id="cb64-206"><a href="#cb64-206" aria-hidden="true" tabindex="-1"></a>for figures. For ğŸ¸ Frog development, ğŸ„ *Arabidopsis* and ğŸ’‰</span>
<span id="cb64-207"><a href="#cb64-207" aria-hidden="true" tabindex="-1"></a>*Leishmania*, this labelling information is most easily added using the</span>
<span id="cb64-208"><a href="#cb64-208" aria-hidden="true" tabindex="-1"></a>metadata. You will need to filter for only the samples in the comparison</span>
<span id="cb64-209"><a href="#cb64-209" aria-hidden="true" tabindex="-1"></a>that was made in the results file.</span>
<span id="cb64-210"><a href="#cb64-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-211"><a href="#cb64-211" aria-hidden="true" tabindex="-1"></a>You may need to refer back to the [Week 4 Statistical Analysis</span>
<span id="cb64-212"><a href="#cb64-212" aria-hidden="true" tabindex="-1"></a>workshop](../week-4/workshop.html) (in section 2. Create DESeqDataSet</span>
<span id="cb64-213"><a href="#cb64-213" aria-hidden="true" tabindex="-1"></a>object) to remind yourself how to import and filter the metadata you</span>
<span id="cb64-214"><a href="#cb64-214" aria-hidden="true" tabindex="-1"></a>need.</span>
<span id="cb64-215"><a href="#cb64-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-216"><a href="#cb64-216" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Import the metadata that maps the sample names to treatments.</span>
<span id="cb64-217"><a href="#cb64-217" aria-hidden="true" tabindex="-1"></a>Remember to select only the samples for comparison that was made.</span>
<span id="cb64-218"><a href="#cb64-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-221"><a href="#cb64-221" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-222"><a href="#cb64-222" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb64-223"><a href="#cb64-223" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-224"><a href="#cb64-224" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ„  Arabidopisis</span></span>
<span id="cb64-225"><a href="#cb64-225" aria-hidden="true" tabindex="-1"></a><span class="co"># Import metadata that maps the sample names to treatments</span></span>
<span id="cb64-226"><a href="#cb64-226" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read_table</span>(<span class="st">"meta/arab_meta_data.txt"</span>)</span>
<span id="cb64-227"><a href="#cb64-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-228"><a href="#cb64-228" aria-hidden="true" tabindex="-1"></a><span class="co"># We only need the wild</span></span>
<span id="cb64-229"><a href="#cb64-229" aria-hidden="true" tabindex="-1"></a>meta_wild <span class="ot">&lt;-</span> meta <span class="sc">|&gt;</span></span>
<span id="cb64-230"><a href="#cb64-230" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(genotype <span class="sc">==</span> <span class="st">"wt"</span>)</span>
<span id="cb64-231"><a href="#cb64-231" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-232"><a href="#cb64-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-235"><a href="#cb64-235" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-236"><a href="#cb64-236" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb64-237"><a href="#cb64-237" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-238"><a href="#cb64-238" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ’‰ Leishmania</span></span>
<span id="cb64-239"><a href="#cb64-239" aria-hidden="true" tabindex="-1"></a><span class="co"># Import metadata that maps the sample names to treatments</span></span>
<span id="cb64-240"><a href="#cb64-240" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read_table</span>(<span class="st">"meta/leish_meta_data.txt"</span>)</span>
<span id="cb64-241"><a href="#cb64-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-242"><a href="#cb64-242" aria-hidden="true" tabindex="-1"></a><span class="co"># We only need the procyclic and metacyclic promastigote stages</span></span>
<span id="cb64-243"><a href="#cb64-243" aria-hidden="true" tabindex="-1"></a>meta_pro_meta <span class="ot">&lt;-</span> meta <span class="sc">|&gt;</span></span>
<span id="cb64-244"><a href="#cb64-244" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(stage <span class="sc">!=</span> <span class="st">"amastigotes"</span>)</span>
<span id="cb64-245"><a href="#cb64-245" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-246"><a href="#cb64-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-247"><a href="#cb64-247" aria-hidden="true" tabindex="-1"></a>For ğŸ­ Stem cells, cell types are encoded in the column names. We will</span>
<span id="cb64-248"><a href="#cb64-248" aria-hidden="true" tabindex="-1"></a>do some [regular</span>
<span id="cb64-249"><a href="#cb64-249" aria-hidden="true" tabindex="-1"></a>expression](https://en.wikipedia.org/wiki/Regular_expression) magic to</span>
<span id="cb64-250"><a href="#cb64-250" aria-hidden="true" tabindex="-1"></a>extract the cell type from the column names.</span>
<span id="cb64-251"><a href="#cb64-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-252"><a href="#cb64-252" aria-hidden="true" tabindex="-1"></a><span class="fu"># log~2~ transform the normalised counts</span></span>
<span id="cb64-253"><a href="#cb64-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-254"><a href="#cb64-254" aria-hidden="true" tabindex="-1"></a>We use the normalised counts for data visualisations so that the</span>
<span id="cb64-255"><a href="#cb64-255" aria-hidden="true" tabindex="-1"></a>comparisons are meaningful. Since the fold changes are given is log~2~</span>
<span id="cb64-256"><a href="#cb64-256" aria-hidden="true" tabindex="-1"></a>it is useful to log~2~ transform the normalised counts too. We will add</span>
<span id="cb64-257"><a href="#cb64-257" aria-hidden="true" tabindex="-1"></a>columns to the dataframe with these transformed values. Since we have</span>
<span id="cb64-258"><a href="#cb64-258" aria-hidden="true" tabindex="-1"></a>some counts of 0 we will add a tiny amount to avoid <span class="in">`-Inf`</span> values.</span>
<span id="cb64-259"><a href="#cb64-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-260"><a href="#cb64-260" aria-hidden="true" tabindex="-1"></a>log~2~ transformation would be applied to one column like this:</span>
<span id="cb64-261"><a href="#cb64-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-264"><a href="#cb64-264" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-265"><a href="#cb64-265" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb64-266"><a href="#cb64-266" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT DO</span></span>
<span id="cb64-267"><a href="#cb64-267" aria-hidden="true" tabindex="-1"></a><span class="co"># log2 transform the counts plus a tiny amount to avoid log(0)</span></span>
<span id="cb64-268"><a href="#cb64-268" aria-hidden="true" tabindex="-1"></a>dataframe <span class="ot">&lt;-</span> dataframe <span class="sc">|&gt;</span></span>
<span id="cb64-269"><a href="#cb64-269" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">log2_mycolumn =</span> <span class="fu">log2</span>(mycolumn <span class="sc">+</span> <span class="fl">0.001</span>))</span>
<span id="cb64-270"><a href="#cb64-270" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-271"><a href="#cb64-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-272"><a href="#cb64-272" aria-hidden="true" tabindex="-1"></a>We are going to use a wonderful bit of R wizardry to apply a</span>
<span id="cb64-273"><a href="#cb64-273" aria-hidden="true" tabindex="-1"></a>transformation to multiple columns. This is the <span class="in">`across()`</span> function</span>
<span id="cb64-274"><a href="#cb64-274" aria-hidden="true" tabindex="-1"></a>which has three arguments:</span>
<span id="cb64-275"><a href="#cb64-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-276"><a href="#cb64-276" aria-hidden="true" tabindex="-1"></a><span class="in">`across(.cols, .fns, .names)`</span></span>
<span id="cb64-277"><a href="#cb64-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-278"><a href="#cb64-278" aria-hidden="true" tabindex="-1"></a>where:</span>
<span id="cb64-279"><a href="#cb64-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-280"><a href="#cb64-280" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`.cols`</span> is the selection of columns to transform</span>
<span id="cb64-281"><a href="#cb64-281" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`.fns`</span> is the function we want to apply to the selected columns</span>
<span id="cb64-282"><a href="#cb64-282" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`.names`</span> is the naming convention for the new columns</span>
<span id="cb64-283"><a href="#cb64-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-284"><a href="#cb64-284" aria-hidden="true" tabindex="-1"></a>The general form of the code you need is:</span>
<span id="cb64-285"><a href="#cb64-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-288"><a href="#cb64-288" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-289"><a href="#cb64-289" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb64-290"><a href="#cb64-290" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT DO</span></span>
<span id="cb64-291"><a href="#cb64-291" aria-hidden="true" tabindex="-1"></a><span class="co"># log2 transform the counts plus a tiny amount to avoid log(0)</span></span>
<span id="cb64-292"><a href="#cb64-292" aria-hidden="true" tabindex="-1"></a>xxxx_results <span class="ot">&lt;-</span> xxxx_results <span class="sc">|&gt;</span></span>
<span id="cb64-293"><a href="#cb64-293" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">starts_with</span>(<span class="st">"pattern"</span>), </span>
<span id="cb64-294"><a href="#cb64-294" aria-hidden="true" tabindex="-1"></a>                \(x) <span class="fu">log2</span>(x <span class="sc">+</span> <span class="fl">0.001</span>),</span>
<span id="cb64-295"><a href="#cb64-295" aria-hidden="true" tabindex="-1"></a>                <span class="at">.names =</span> <span class="st">"log2_{.col}"</span>))</span>
<span id="cb64-296"><a href="#cb64-296" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-297"><a href="#cb64-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-298"><a href="#cb64-298" aria-hidden="true" tabindex="-1"></a>where:</span>
<span id="cb64-299"><a href="#cb64-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-300"><a href="#cb64-300" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`xxxx_results`</span> is the name of the dataframe of results</span>
<span id="cb64-301"><a href="#cb64-301" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`pattern`</span> matches the starting letters for all of the normalised</span>
<span id="cb64-302"><a href="#cb64-302" aria-hidden="true" tabindex="-1"></a>    counts so that <span class="in">`starts_with("pattern")`</span> gives the selection of</span>
<span id="cb64-303"><a href="#cb64-303" aria-hidden="true" tabindex="-1"></a>    columns to transform</span>
<span id="cb64-304"><a href="#cb64-304" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the bit after the <span class="in">`\(x)`</span> is the function we want to apply to the</span>
<span id="cb64-305"><a href="#cb64-305" aria-hidden="true" tabindex="-1"></a>    selected columns</span>
<span id="cb64-306"><a href="#cb64-306" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the <span class="in">`\(x)`</span> means it is an "anonymous" function which means we don't</span>
<span id="cb64-307"><a href="#cb64-307" aria-hidden="true" tabindex="-1"></a>    have to define a function name.</span>
<span id="cb64-308"><a href="#cb64-308" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`"log2_{.col}"`</span> means the columns will have the same name (in the</span>
<span id="cb64-309"><a href="#cb64-309" aria-hidden="true" tabindex="-1"></a>    <span class="in">`.col`</span>) but with the prefix <span class="in">`log2_`</span> added.</span>
<span id="cb64-310"><a href="#cb64-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-311"><a href="#cb64-311" aria-hidden="true" tabindex="-1"></a>You can read more about <span class="in">`across()`</span> and anonymous functions from my</span>
<span id="cb64-312"><a href="#cb64-312" aria-hidden="true" tabindex="-1"></a>[posit::conf(2024)</span>
<span id="cb64-313"><a href="#cb64-313" aria-hidden="true" tabindex="-1"></a>workshop](https://posit-conf-2024.github.io/programming-r/)</span>
<span id="cb64-314"><a href="#cb64-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-315"><a href="#cb64-315" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ„ *Arabidopsis* and ğŸ’‰ *Leishmania*</span></span>
<span id="cb64-316"><a href="#cb64-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-317"><a href="#cb64-317" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Design the code to log~2~ transform the normalised counts using the</span>
<span id="cb64-318"><a href="#cb64-318" aria-hidden="true" tabindex="-1"></a>template given</span>
<span id="cb64-319"><a href="#cb64-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-322"><a href="#cb64-322" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-323"><a href="#cb64-323" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-324"><a href="#cb64-324" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-325"><a href="#cb64-325" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ„ Arabidopsis</span></span>
<span id="cb64-326"><a href="#cb64-326" aria-hidden="true" tabindex="-1"></a><span class="co"># log2 transform the counts plus a tiny amount to avoid log(0)</span></span>
<span id="cb64-327"><a href="#cb64-327" aria-hidden="true" tabindex="-1"></a>wild_results <span class="ot">&lt;-</span> wild_results <span class="sc">|&gt;</span></span>
<span id="cb64-328"><a href="#cb64-328" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">starts_with</span>(<span class="st">"SRX"</span>),</span>
<span id="cb64-329"><a href="#cb64-329" aria-hidden="true" tabindex="-1"></a>                \(x) <span class="fu">log2</span>(x <span class="sc">+</span> <span class="fl">0.001</span>),</span>
<span id="cb64-330"><a href="#cb64-330" aria-hidden="true" tabindex="-1"></a>                <span class="at">.names =</span> <span class="st">"log2_{.col}"</span>))</span>
<span id="cb64-331"><a href="#cb64-331" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-332"><a href="#cb64-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-335"><a href="#cb64-335" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-336"><a href="#cb64-336" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-337"><a href="#cb64-337" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-338"><a href="#cb64-338" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ’‰ Leishmania</span></span>
<span id="cb64-339"><a href="#cb64-339" aria-hidden="true" tabindex="-1"></a><span class="co"># log2 transform the counts plus a tiny amount to avoid log(0)</span></span>
<span id="cb64-340"><a href="#cb64-340" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="ot">&lt;-</span> pro_meta_results <span class="sc">|&gt;</span></span>
<span id="cb64-341"><a href="#cb64-341" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">starts_with</span>(<span class="st">"lm_"</span>), </span>
<span id="cb64-342"><a href="#cb64-342" aria-hidden="true" tabindex="-1"></a>                \(x) <span class="fu">log2</span>(x <span class="sc">+</span> <span class="fl">0.001</span>),</span>
<span id="cb64-343"><a href="#cb64-343" aria-hidden="true" tabindex="-1"></a>                <span class="at">.names =</span> <span class="st">"log2_{.col}"</span>))</span>
<span id="cb64-344"><a href="#cb64-344" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-345"><a href="#cb64-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-346"><a href="#cb64-346" aria-hidden="true" tabindex="-1"></a>I recommend viewing the dataframe to see the new columns. Check you have</span>
<span id="cb64-347"><a href="#cb64-347" aria-hidden="true" tabindex="-1"></a>the expected number of columns.</span>
<span id="cb64-348"><a href="#cb64-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-349"><a href="#cb64-349" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ­ Stem cells</span></span>
<span id="cb64-350"><a href="#cb64-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-351"><a href="#cb64-351" aria-hidden="true" tabindex="-1"></a>You do not need to apply this transformation because the data is already</span>
<span id="cb64-352"><a href="#cb64-352" aria-hidden="true" tabindex="-1"></a>log~2~ transformed.</span>
<span id="cb64-353"><a href="#cb64-353" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-354"><a href="#cb64-354" aria-hidden="true" tabindex="-1"></a><span class="fu">## Everyone</span></span>
<span id="cb64-355"><a href="#cb64-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-356"><a href="#cb64-356" aria-hidden="true" tabindex="-1"></a>We now all have dataframes with all the information we need: normalised</span>
<span id="cb64-357"><a href="#cb64-357" aria-hidden="true" tabindex="-1"></a>counts, log~2~ normalised counts, statistical comparisons with fold</span>
<span id="cb64-358"><a href="#cb64-358" aria-hidden="true" tabindex="-1"></a>changes and *p*-values, and information about the gene.</span>
<span id="cb64-359"><a href="#cb64-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-360"><a href="#cb64-360" aria-hidden="true" tabindex="-1"></a><span class="fu"># Write the significant genes to file</span></span>
<span id="cb64-361"><a href="#cb64-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-362"><a href="#cb64-362" aria-hidden="true" tabindex="-1"></a><span class="fu">## Everyone</span></span>
<span id="cb64-363"><a href="#cb64-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-364"><a href="#cb64-364" aria-hidden="true" tabindex="-1"></a>We will create dataframe of the significant genes and write them to</span>
<span id="cb64-365"><a href="#cb64-365" aria-hidden="true" tabindex="-1"></a>file. This is subset from the results file but will make it a little</span>
<span id="cb64-366"><a href="#cb64-366" aria-hidden="true" tabindex="-1"></a>easier to examine and select genes of interest.</span>
<span id="cb64-367"><a href="#cb64-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-368"><a href="#cb64-368" aria-hidden="true" tabindex="-1"></a>The general form of the code you need is:</span>
<span id="cb64-369"><a href="#cb64-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-372"><a href="#cb64-372" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-373"><a href="#cb64-373" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb64-374"><a href="#cb64-374" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT DO</span></span>
<span id="cb64-375"><a href="#cb64-375" aria-hidden="true" tabindex="-1"></a><span class="co"># create a dataframe of genes significant at 0.05 level</span></span>
<span id="cb64-376"><a href="#cb64-376" aria-hidden="true" tabindex="-1"></a>xxxx_results_sig0<span class="fl">.05</span> <span class="ot">&lt;-</span> xxxx_results <span class="sc">|&gt;</span> </span>
<span id="cb64-377"><a href="#cb64-377" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(pvalue <span class="sc">&lt;=</span> <span class="fl">0.05</span>)</span>
<span id="cb64-378"><a href="#cb64-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-379"><a href="#cb64-379" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-380"><a href="#cb64-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-381"><a href="#cb64-381" aria-hidden="true" tabindex="-1"></a>Note that you determine the significance level using the adjusted</span>
<span id="cb64-382"><a href="#cb64-382" aria-hidden="true" tabindex="-1"></a>*p*-values rather than the uncorrected *p*-values. This column is name</span>
<span id="cb64-383"><a href="#cb64-383" aria-hidden="true" tabindex="-1"></a><span class="in">`padj`</span> in **`DESeq2`** output and `FDR` in **`scran`** output.</span>
<span id="cb64-384"><a href="#cb64-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-385"><a href="#cb64-385" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a dataframe of the genes significant at the 0.05 level using</span>
<span id="cb64-386"><a href="#cb64-386" aria-hidden="true" tabindex="-1"></a><span class="in">`filter()`</span>. You will need to know the name of column with the adjusted</span>
<span id="cb64-387"><a href="#cb64-387" aria-hidden="true" tabindex="-1"></a>*p*-values.</span>
<span id="cb64-388"><a href="#cb64-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-391"><a href="#cb64-391" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-392"><a href="#cb64-392" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-393"><a href="#cb64-393" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-394"><a href="#cb64-394" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ„ Arabidopsis</span></span>
<span id="cb64-395"><a href="#cb64-395" aria-hidden="true" tabindex="-1"></a><span class="co"># create a dataframe of genes significant at 0.05 level</span></span>
<span id="cb64-396"><a href="#cb64-396" aria-hidden="true" tabindex="-1"></a>wild_results_sig0<span class="fl">.05</span> <span class="ot">&lt;-</span> wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-397"><a href="#cb64-397" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(padj <span class="sc">&lt;=</span> <span class="fl">0.05</span>)</span>
<span id="cb64-398"><a href="#cb64-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-399"><a href="#cb64-399" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-400"><a href="#cb64-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-403"><a href="#cb64-403" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-404"><a href="#cb64-404" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-405"><a href="#cb64-405" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-406"><a href="#cb64-406" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ’‰ Leishmania</span></span>
<span id="cb64-407"><a href="#cb64-407" aria-hidden="true" tabindex="-1"></a><span class="co"># create a dataframe of genes significant at 0.05 level</span></span>
<span id="cb64-408"><a href="#cb64-408" aria-hidden="true" tabindex="-1"></a>pro_meta_results_sig0<span class="fl">.05</span> <span class="ot">&lt;-</span> pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-409"><a href="#cb64-409" aria-hidden="true" tabindex="-1"></a>   <span class="fu">filter</span>(padj <span class="sc">&lt;=</span> <span class="fl">0.05</span>)</span>
<span id="cb64-410"><a href="#cb64-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-411"><a href="#cb64-411" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-412"><a href="#cb64-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-415"><a href="#cb64-415" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-416"><a href="#cb64-416" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-417"><a href="#cb64-417" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-418"><a href="#cb64-418" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ­ Stem cells</span></span>
<span id="cb64-419"><a href="#cb64-419" aria-hidden="true" tabindex="-1"></a><span class="co"># create a dataframe of genes significant at 0.05 level</span></span>
<span id="cb64-420"><a href="#cb64-420" aria-hidden="true" tabindex="-1"></a> hspc_prog_results_sig0<span class="fl">.05</span> <span class="ot">&lt;-</span> hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-421"><a href="#cb64-421" aria-hidden="true" tabindex="-1"></a>   <span class="fu">filter</span>(FDR <span class="sc">&lt;=</span> <span class="fl">0.05</span>)</span>
<span id="cb64-422"><a href="#cb64-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-423"><a href="#cb64-423" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-424"><a href="#cb64-424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-425"><a href="#cb64-425" aria-hidden="true" tabindex="-1"></a>â“How many genes are significant at the 0.05 levels?</span>
<span id="cb64-426"><a href="#cb64-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-427"><a href="#cb64-427" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #---THINKING ANSWER--- --&gt;</span></span>
<span id="cb64-428"><a href="#cb64-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-429"><a href="#cb64-429" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ğŸ„ Arabidopsis  --&gt;</span></span>
<span id="cb64-430"><a href="#cb64-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-431"><a href="#cb64-431" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 12 genes are significant at the 0.05 level. --&gt;</span></span>
<span id="cb64-432"><a href="#cb64-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-433"><a href="#cb64-433" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ğŸ’‰ leishmania  --&gt;</span></span>
<span id="cb64-434"><a href="#cb64-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-435"><a href="#cb64-435" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 3252 genes are significant at the 0.05 level. --&gt;</span></span>
<span id="cb64-436"><a href="#cb64-436" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-437"><a href="#cb64-437" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- I recommend using a more stringent cut-off such as 0.01  --&gt;</span></span>
<span id="cb64-438"><a href="#cb64-438" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-439"><a href="#cb64-439" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ğŸ­ stemcells --&gt;</span></span>
<span id="cb64-440"><a href="#cb64-440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-441"><a href="#cb64-441" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 182 genes are significant at the 0.05 level. --&gt;</span></span>
<span id="cb64-442"><a href="#cb64-442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-443"><a href="#cb64-443" aria-hidden="true" tabindex="-1"></a>If you have a very large number of genes significant at the 0.05 level,</span>
<span id="cb64-444"><a href="#cb64-444" aria-hidden="true" tabindex="-1"></a>you may want to consider a more stringent cut-off such as 0.01.</span>
<span id="cb64-445"><a href="#cb64-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-448"><a href="#cb64-448" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-449"><a href="#cb64-449" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-450"><a href="#cb64-450" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-451"><a href="#cb64-451" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ’‰ Leishmania</span></span>
<span id="cb64-452"><a href="#cb64-452" aria-hidden="true" tabindex="-1"></a><span class="co"># create a dataframe of genes significant at 0.01 level</span></span>
<span id="cb64-453"><a href="#cb64-453" aria-hidden="true" tabindex="-1"></a>pro_meta_results_sig0<span class="fl">.01</span> <span class="ot">&lt;-</span> pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-454"><a href="#cb64-454" aria-hidden="true" tabindex="-1"></a>   <span class="fu">filter</span>(padj <span class="sc">&lt;=</span> <span class="fl">0.01</span>)</span>
<span id="cb64-455"><a href="#cb64-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-456"><a href="#cb64-456" aria-hidden="true" tabindex="-1"></a><span class="co"># there are still 2370 genes significant at the 0.01 level</span></span>
<span id="cb64-457"><a href="#cb64-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-458"><a href="#cb64-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-459"><a href="#cb64-459" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-460"><a href="#cb64-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-461"><a href="#cb64-461" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Write the dataframe to a csv file. I recommend using the same file</span>
<span id="cb64-462"><a href="#cb64-462" aria-hidden="true" tabindex="-1"></a>name as you used for the dataframe.</span>
<span id="cb64-463"><a href="#cb64-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-466"><a href="#cb64-466" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-467"><a href="#cb64-467" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-468"><a href="#cb64-468" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-469"><a href="#cb64-469" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ„ Arabidopsis</span></span>
<span id="cb64-470"><a href="#cb64-470" aria-hidden="true" tabindex="-1"></a><span class="co"># write to csv file</span></span>
<span id="cb64-471"><a href="#cb64-471" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(wild_results_sig0<span class="fl">.05</span>, </span>
<span id="cb64-472"><a href="#cb64-472" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"results/wild_results_sig0.05.csv"</span>)</span>
<span id="cb64-473"><a href="#cb64-473" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-474"><a href="#cb64-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-477"><a href="#cb64-477" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-478"><a href="#cb64-478" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-479"><a href="#cb64-479" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-480"><a href="#cb64-480" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ’‰ Leishmania</span></span>
<span id="cb64-481"><a href="#cb64-481" aria-hidden="true" tabindex="-1"></a><span class="co"># write to csv file</span></span>
<span id="cb64-482"><a href="#cb64-482" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(pro_meta_results_sig0<span class="fl">.05</span>, </span>
<span id="cb64-483"><a href="#cb64-483" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"results/pro_meta_results_sig0.05.csv"</span>)</span>
<span id="cb64-484"><a href="#cb64-484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-485"><a href="#cb64-485" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(pro_meta_results_sig0<span class="fl">.01</span>, </span>
<span id="cb64-486"><a href="#cb64-486" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"results/pro_meta_results_sig0.01.csv"</span>)</span>
<span id="cb64-487"><a href="#cb64-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-488"><a href="#cb64-488" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-489"><a href="#cb64-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-492"><a href="#cb64-492" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-493"><a href="#cb64-493" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb64-494"><a href="#cb64-494" aria-hidden="true" tabindex="-1"></a><span class="co">#---CODING ANSWER---</span></span>
<span id="cb64-495"><a href="#cb64-495" aria-hidden="true" tabindex="-1"></a><span class="co"># ğŸ­ Stem cells</span></span>
<span id="cb64-496"><a href="#cb64-496" aria-hidden="true" tabindex="-1"></a><span class="co"># write to csv file</span></span>
<span id="cb64-497"><a href="#cb64-497" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(hspc_prog_results_sig0<span class="fl">.05</span>, </span>
<span id="cb64-498"><a href="#cb64-498" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"results/hspc_prog_results_sig0.05.csv"</span>)</span>
<span id="cb64-499"><a href="#cb64-499" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-500"><a href="#cb64-500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-501"><a href="#cb64-501" aria-hidden="true" tabindex="-1"></a><span class="fu"># Principal Component Analysis (PCA)</span></span>
<span id="cb64-502"><a href="#cb64-502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-503"><a href="#cb64-503" aria-hidden="true" tabindex="-1"></a>We have many genes in our datasets. PCA will allow us to plot our</span>
<span id="cb64-504"><a href="#cb64-504" aria-hidden="true" tabindex="-1"></a>samples in the "gene expression" space so we can see if replicates with</span>
<span id="cb64-505"><a href="#cb64-505" aria-hidden="true" tabindex="-1"></a>the same treatment cluster together as we would expect. PCA is a</span>
<span id="cb64-506"><a href="#cb64-506" aria-hidden="true" tabindex="-1"></a>dimension reduction technique that finds the directions of maximum</span>
<span id="cb64-507"><a href="#cb64-507" aria-hidden="true" tabindex="-1"></a>variance in the data. We are doing PCA after our statistical analysis</span>
<span id="cb64-508"><a href="#cb64-508" aria-hidden="true" tabindex="-1"></a>but often it is one of the first techniques used to explore the data. If</span>
<span id="cb64-509"><a href="#cb64-509" aria-hidden="true" tabindex="-1"></a>we do not see treatment effects in a PCA plot then we are not likely to</span>
<span id="cb64-510"><a href="#cb64-510" aria-hidden="true" tabindex="-1"></a>see them in the statistical analysis. PCA can also help you identify any</span>
<span id="cb64-511"><a href="#cb64-511" aria-hidden="true" tabindex="-1"></a>outlier replicates. The PCA is best conducted on the normalised counts</span>
<span id="cb64-512"><a href="#cb64-512" aria-hidden="true" tabindex="-1"></a>or the log~2~ transformed normalised counts. We will use the log~2~</span>
<span id="cb64-513"><a href="#cb64-513" aria-hidden="true" tabindex="-1"></a>transformed normalised counts.</span>
<span id="cb64-514"><a href="#cb64-514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-515"><a href="#cb64-515" aria-hidden="true" tabindex="-1"></a>We will carry out the following steps to do the PCA:</span>
<span id="cb64-516"><a href="#cb64-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-517"><a href="#cb64-517" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Select the log~2~ transformed normalised counts. You can only use</span>
<span id="cb64-518"><a href="#cb64-518" aria-hidden="true" tabindex="-1"></a>    numerical data in PCA.</span>
<span id="cb64-519"><a href="#cb64-519" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Transpose our data. We have genes in rows and samples in columns</span>
<span id="cb64-520"><a href="#cb64-520" aria-hidden="true" tabindex="-1"></a>    (this is common for gene expression data). However, to treatment the</span>
<span id="cb64-521"><a href="#cb64-521" aria-hidden="true" tabindex="-1"></a>    genes as variables, PCA expects samples in rows and genes in</span>
<span id="cb64-522"><a href="#cb64-522" aria-hidden="true" tabindex="-1"></a>    columns.</span>
<span id="cb64-523"><a href="#cb64-523" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Add the gene names as column names in the transposed data</span>
<span id="cb64-524"><a href="#cb64-524" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Perform the PCA</span>
<span id="cb64-525"><a href="#cb64-525" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Extract the scores on the first two principal components and label</span>
<span id="cb64-526"><a href="#cb64-526" aria-hidden="true" tabindex="-1"></a>    the data</span>
<span id="cb64-527"><a href="#cb64-527" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Plot the the first two principal components as a scatter plot</span>
<span id="cb64-528"><a href="#cb64-528" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-529"><a href="#cb64-529" aria-hidden="true" tabindex="-1"></a>In each case we will use only the samples for the comparison that was</span>
<span id="cb64-530"><a href="#cb64-530" aria-hidden="true" tabindex="-1"></a>made in the PCA. However, it would be informative to do PCA on all the</span>
<span id="cb64-531"><a href="#cb64-531" aria-hidden="true" tabindex="-1"></a>samples you have and that is something you may want to consider in your</span>
<span id="cb64-532"><a href="#cb64-532" aria-hidden="true" tabindex="-1"></a>independent study.</span>
<span id="cb64-533"><a href="#cb64-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-534"><a href="#cb64-534" aria-hidden="true" tabindex="-1"></a>Now go to PCA for:</span>
<span id="cb64-535"><a href="#cb64-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-536"><a href="#cb64-536" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">ğŸ„ Arabidopsis</span><span class="co">](#arabidopsis-1)</span></span>
<span id="cb64-537"><a href="#cb64-537" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">ğŸ’‰ Leishmania</span><span class="co">](#leishmania-1)</span></span>
<span id="cb64-538"><a href="#cb64-538" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">ğŸ­ Stem cells</span><span class="co">](#stem-cells-2)</span></span>
<span id="cb64-539"><a href="#cb64-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-540"><a href="#cb64-540" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ„ *Arabidopsis* {#arabidopsis-1}</span></span>
<span id="cb64-541"><a href="#cb64-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-542"><a href="#cb64-542" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Transpose the log~2~ transformed normalised counts:</span>
<span id="cb64-543"><a href="#cb64-543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-546"><a href="#cb64-546" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-547"><a href="#cb64-547" aria-hidden="true" tabindex="-1"></a>wild_log2_trans <span class="ot">&lt;-</span> wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-548"><a href="#cb64-548" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">starts_with</span>(<span class="st">"log2_"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb64-549"><a href="#cb64-549" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">|&gt;</span> </span>
<span id="cb64-550"><a href="#cb64-550" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>()</span>
<span id="cb64-551"><a href="#cb64-551" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-552"><a href="#cb64-552" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-553"><a href="#cb64-553" aria-hidden="true" tabindex="-1"></a>We have used the <span class="in">`select()`</span> function to select all the columns that</span>
<span id="cb64-554"><a href="#cb64-554" aria-hidden="true" tabindex="-1"></a>start with <span class="in">`log2_`</span>. We then use the <span class="in">`t()`</span> function to transpose the</span>
<span id="cb64-555"><a href="#cb64-555" aria-hidden="true" tabindex="-1"></a>dataframe. We then convert the resulting matrix to a dataframe using</span>
<span id="cb64-556"><a href="#cb64-556" aria-hidden="true" tabindex="-1"></a><span class="in">`data.frame()`</span>. If you view that dataframe you'll see it has default</span>
<span id="cb64-557"><a href="#cb64-557" aria-hidden="true" tabindex="-1"></a>column name which we can fix using <span class="in">`colnames()`</span> to set the column names</span>
<span id="cb64-558"><a href="#cb64-558" aria-hidden="true" tabindex="-1"></a>to the gene ids.</span>
<span id="cb64-559"><a href="#cb64-559" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-560"><a href="#cb64-560" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Set the column names to the gene ids:</span>
<span id="cb64-561"><a href="#cb64-561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-564"><a href="#cb64-564" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-565"><a href="#cb64-565" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(wild_log2_trans) <span class="ot">&lt;-</span> wild_results<span class="sc">$</span>gene_id</span>
<span id="cb64-566"><a href="#cb64-566" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-567"><a href="#cb64-567" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-568"><a href="#cb64-568" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Perform PCA on the log~2~ transformed normalised counts:</span>
<span id="cb64-569"><a href="#cb64-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-572"><a href="#cb64-572" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-573"><a href="#cb64-573" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> wild_log2_trans <span class="sc">|&gt;</span></span>
<span id="cb64-574"><a href="#cb64-574" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prcomp</span>(<span class="at">rank. =</span> <span class="dv">4</span>, <span class="at">scale =</span> <span class="cn">TRUE</span>)</span>
<span id="cb64-575"><a href="#cb64-575" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-576"><a href="#cb64-576" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-577"><a href="#cb64-577" aria-hidden="true" tabindex="-1"></a>The <span class="in">`scale`</span> argument tells <span class="in">`prcomp()`</span> to scale the data before doing the</span>
<span id="cb64-578"><a href="#cb64-578" aria-hidden="true" tabindex="-1"></a>PCA. This is important when the variables are on different scales to</span>
<span id="cb64-579"><a href="#cb64-579" aria-hidden="true" tabindex="-1"></a>stop variables with large values dominating the PCA. The <span class="in">`rank.`</span></span>
<span id="cb64-580"><a href="#cb64-580" aria-hidden="true" tabindex="-1"></a>argument tells <span class="in">`prcomp()`</span> to only calculate the first 4 principal</span>
<span id="cb64-581"><a href="#cb64-581" aria-hidden="true" tabindex="-1"></a>components. This is useful for visualisation as we can only plot in 2 or</span>
<span id="cb64-582"><a href="#cb64-582" aria-hidden="true" tabindex="-1"></a>3 dimensions. We can see the results of the PCA by viewing the</span>
<span id="cb64-583"><a href="#cb64-583" aria-hidden="true" tabindex="-1"></a><span class="in">`summary()`</span> of the <span class="in">`pca`</span> object.</span>
<span id="cb64-584"><a href="#cb64-584" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-587"><a href="#cb64-587" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-588"><a href="#cb64-588" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pca)</span>
<span id="cb64-589"><a href="#cb64-589" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-590"><a href="#cb64-590" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-591"><a href="#cb64-591" aria-hidden="true" tabindex="-1"></a>The Proportion of Variance tells us how much of the variance is</span>
<span id="cb64-592"><a href="#cb64-592" aria-hidden="true" tabindex="-1"></a>explained by each component. We can see that the first component</span>
<span id="cb64-593"><a href="#cb64-593" aria-hidden="true" tabindex="-1"></a>explains 0.5742 of the variance, the second 0.2786, and the third</span>
<span id="cb64-594"><a href="#cb64-594" aria-hidden="true" tabindex="-1"></a>0.1472. Together the first three components explain nearly 100% of the</span>
<span id="cb64-595"><a href="#cb64-595" aria-hidden="true" tabindex="-1"></a>total variance in the data. Plotting PC1 against PC2 will capture about</span>
<span id="cb64-596"><a href="#cb64-596" aria-hidden="true" tabindex="-1"></a>92% of the variance which is very likely very much better than we would</span>
<span id="cb64-597"><a href="#cb64-597" aria-hidden="true" tabindex="-1"></a>get plotting any two genes against each other. To plot the PC1 against</span>
<span id="cb64-598"><a href="#cb64-598" aria-hidden="true" tabindex="-1"></a>PC2 we will need to extract the PC1 and PC2 "scores" from the PCA object</span>
<span id="cb64-599"><a href="#cb64-599" aria-hidden="true" tabindex="-1"></a>and add labels for the samples. Those labels will come from the row</span>
<span id="cb64-600"><a href="#cb64-600" aria-hidden="true" tabindex="-1"></a>names of the transformed data which has the sample ids and from the</span>
<span id="cb64-601"><a href="#cb64-601" aria-hidden="true" tabindex="-1"></a>metadata.</span>
<span id="cb64-602"><a href="#cb64-602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-603"><a href="#cb64-603" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a vector of the sample ids from the row names. These include</span>
<span id="cb64-604"><a href="#cb64-604" aria-hidden="true" tabindex="-1"></a>the <span class="in">`log2`</span> prefix which we can removed for labelling:</span>
<span id="cb64-605"><a href="#cb64-605" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-608"><a href="#cb64-608" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-609"><a href="#cb64-609" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">&lt;-</span> <span class="fu">row.names</span>(wild_log2_trans) <span class="sc">|&gt;</span> <span class="fu">str_remove</span>(<span class="st">"log2_"</span>)</span>
<span id="cb64-610"><a href="#cb64-610" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-611"><a href="#cb64-611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-612"><a href="#cb64-612" aria-hidden="true" tabindex="-1"></a>You might want to check the result.</span>
<span id="cb64-613"><a href="#cb64-613" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-614"><a href="#cb64-614" aria-hidden="true" tabindex="-1"></a>Now we will extract the PC1 and PC2 scores from the PCA object and add.</span>
<span id="cb64-615"><a href="#cb64-615" aria-hidden="true" tabindex="-1"></a>Our PCA object is called <span class="in">`pca`</span> and the scores are in pca\$x. We will</span>
<span id="cb64-616"><a href="#cb64-616" aria-hidden="true" tabindex="-1"></a>create a dataframe of the scores and add the sample ids.</span>
<span id="cb64-617"><a href="#cb64-617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-618"><a href="#cb64-618" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a dataframe of PC1 and PC2 scores and add the sample ids:</span>
<span id="cb64-619"><a href="#cb64-619" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-622"><a href="#cb64-622" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-623"><a href="#cb64-623" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(pca<span class="sc">$</span>x,</span>
<span id="cb64-624"><a href="#cb64-624" aria-hidden="true" tabindex="-1"></a>                           sample_id)</span>
<span id="cb64-625"><a href="#cb64-625" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-626"><a href="#cb64-626" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-627"><a href="#cb64-627" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Merge with the metadata so we can label points by treatment and</span>
<span id="cb64-628"><a href="#cb64-628" aria-hidden="true" tabindex="-1"></a>sibling pair:</span>
<span id="cb64-629"><a href="#cb64-629" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-632"><a href="#cb64-632" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-633"><a href="#cb64-633" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="ot">&lt;-</span> pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb64-634"><a href="#cb64-634" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(meta_wild, </span>
<span id="cb64-635"><a href="#cb64-635" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb64-636"><a href="#cb64-636" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-637"><a href="#cb64-637" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-638"><a href="#cb64-638" aria-hidden="true" tabindex="-1"></a>Since the metadata contained the sample ids, it was especially important</span>
<span id="cb64-639"><a href="#cb64-639" aria-hidden="true" tabindex="-1"></a>to remove the <span class="in">`log2_`</span> from the row names so that the join would work.</span>
<span id="cb64-640"><a href="#cb64-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-641"><a href="#cb64-641" aria-hidden="true" tabindex="-1"></a>The dataframe should look like this:</span>
<span id="cb64-642"><a href="#cb64-642" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-645"><a href="#cb64-645" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-646"><a href="#cb64-646" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb64-647"><a href="#cb64-647" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(pca_labelled)</span>
<span id="cb64-648"><a href="#cb64-648" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-649"><a href="#cb64-649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-650"><a href="#cb64-650" aria-hidden="true" tabindex="-1"></a>The next task is to plot PC2 against PC1 and colour by copper</span>
<span id="cb64-651"><a href="#cb64-651" aria-hidden="true" tabindex="-1"></a>conditions. This is just a scatterplot so we can use <span class="in">`geom_point()`</span>. We</span>
<span id="cb64-652"><a href="#cb64-652" aria-hidden="true" tabindex="-1"></a>will use colour to indicate the copper conditions.</span>
<span id="cb64-653"><a href="#cb64-653" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-654"><a href="#cb64-654" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Plot PC2 against PC1 and colour by copper conditions:</span>
<span id="cb64-655"><a href="#cb64-655" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-658"><a href="#cb64-658" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-659"><a href="#cb64-659" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb64-660"><a href="#cb64-660" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC3, <span class="at">y =</span> PC2, </span>
<span id="cb64-661"><a href="#cb64-661" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> copper)) <span class="sc">+</span></span>
<span id="cb64-662"><a href="#cb64-662" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb64-663"><a href="#cb64-663" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb64-664"><a href="#cb64-664" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-665"><a href="#cb64-665" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-666"><a href="#cb64-666" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-667"><a href="#cb64-667" aria-hidden="true" tabindex="-1"></a>We do not see particularly good separation between treatments, though</span>
<span id="cb64-668"><a href="#cb64-668" aria-hidden="true" tabindex="-1"></a>perhaps there is some separation between the copper sufficient and</span>
<span id="cb64-669"><a href="#cb64-669" aria-hidden="true" tabindex="-1"></a>copper deficient on PC2. It is also difficult to see if the replicates</span>
<span id="cb64-670"><a href="#cb64-670" aria-hidden="true" tabindex="-1"></a>cluster together (the treatments separate) when there are only two reps.</span>
<span id="cb64-671"><a href="#cb64-671" aria-hidden="true" tabindex="-1"></a>You can also try plotting PC3 or PC4.</span>
<span id="cb64-672"><a href="#cb64-672" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-673"><a href="#cb64-673" aria-hidden="true" tabindex="-1"></a>I prefer to customise the colours and shapes. I especially like the\</span>
<span id="cb64-674"><a href="#cb64-674" aria-hidden="true" tabindex="-1"></a>viridis colour scales which provide colour scales that are perceptually</span>
<span id="cb64-675"><a href="#cb64-675" aria-hidden="true" tabindex="-1"></a>uniform in both colour and black-and-white. They are also designed to be</span>
<span id="cb64-676"><a href="#cb64-676" aria-hidden="true" tabindex="-1"></a>perceived by viewers with common forms of colour blindness. See</span>
<span id="cb64-677"><a href="#cb64-677" aria-hidden="true" tabindex="-1"></a>[Introduction to</span>
<span id="cb64-678"><a href="#cb64-678" aria-hidden="true" tabindex="-1"></a>viridis](https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html#introduction)</span>
<span id="cb64-679"><a href="#cb64-679" aria-hidden="true" tabindex="-1"></a>for more information.</span>
<span id="cb64-680"><a href="#cb64-680" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-681"><a href="#cb64-681" aria-hidden="true" tabindex="-1"></a><span class="in">`ggplot`</span> provides functions to access the viridis scales. Here I use</span>
<span id="cb64-682"><a href="#cb64-682" aria-hidden="true" tabindex="-1"></a><span class="in">`scale_fill_viridis_d()`</span>. The d stands for discrete. The function</span>
<span id="cb64-683"><a href="#cb64-683" aria-hidden="true" tabindex="-1"></a><span class="in">`scale_fill_viridis_c()`</span> would be used for continuous data. Iâ€™ve used</span>
<span id="cb64-684"><a href="#cb64-684" aria-hidden="true" tabindex="-1"></a>the default â€œviridisâ€ (or â€œDâ€) option (do ?scale_fill_viridis_d for all</span>
<span id="cb64-685"><a href="#cb64-685" aria-hidden="true" tabindex="-1"></a>the options) and used the <span class="in">`begin`</span> and <span class="in">`end`</span> arguments to control the</span>
<span id="cb64-686"><a href="#cb64-686" aria-hidden="true" tabindex="-1"></a>range of colour - I have set the range to be from 0.15 to 0.95 the avoid</span>
<span id="cb64-687"><a href="#cb64-687" aria-hidden="true" tabindex="-1"></a>the strongest contrast. I have also set the <span class="in">`name`</span> argument to provide a</span>
<span id="cb64-688"><a href="#cb64-688" aria-hidden="true" tabindex="-1"></a>label for the legend.</span>
<span id="cb64-689"><a href="#cb64-689" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-690"><a href="#cb64-690" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Customise the PC2 against PC1 plot:</span>
<span id="cb64-691"><a href="#cb64-691" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-694"><a href="#cb64-694" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-695"><a href="#cb64-695" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb64-696"><a href="#cb64-696" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, </span>
<span id="cb64-697"><a href="#cb64-697" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> copper)) <span class="sc">+</span></span>
<span id="cb64-698"><a href="#cb64-698" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb64-699"><a href="#cb64-699" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_viridis_d</span>(<span class="at">end =</span> <span class="fl">0.95</span>, <span class="at">begin =</span> <span class="fl">0.15</span>,</span>
<span id="cb64-700"><a href="#cb64-700" aria-hidden="true" tabindex="-1"></a>                         <span class="at">name =</span> <span class="st">"Copper"</span>) <span class="sc">+</span></span>
<span id="cb64-701"><a href="#cb64-701" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb64-702"><a href="#cb64-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-703"><a href="#cb64-703" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-704"><a href="#cb64-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-705"><a href="#cb64-705" aria-hidden="true" tabindex="-1"></a>Now go to Volcano plots for <span class="co">[</span><span class="ot">ğŸ„ Arabidopsis</span><span class="co">](#arabidopsis-2)</span></span>
<span id="cb64-706"><a href="#cb64-706" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-707"><a href="#cb64-707" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ’‰ *Leishmania* {#leishmania-1}</span></span>
<span id="cb64-708"><a href="#cb64-708" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-709"><a href="#cb64-709" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Transpose the log~2~ transformed normalised counts:</span>
<span id="cb64-710"><a href="#cb64-710" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-713"><a href="#cb64-713" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-714"><a href="#cb64-714" aria-hidden="true" tabindex="-1"></a>pro_meta_log2_trans <span class="ot">&lt;-</span> pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-715"><a href="#cb64-715" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">starts_with</span>(<span class="st">"log2_"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb64-716"><a href="#cb64-716" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">|&gt;</span> </span>
<span id="cb64-717"><a href="#cb64-717" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>()</span>
<span id="cb64-718"><a href="#cb64-718" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-719"><a href="#cb64-719" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-720"><a href="#cb64-720" aria-hidden="true" tabindex="-1"></a>We have used the <span class="in">`select()`</span> function to select all the columns that</span>
<span id="cb64-721"><a href="#cb64-721" aria-hidden="true" tabindex="-1"></a>start with <span class="in">`log2_`</span>. We then use the <span class="in">`t()`</span> function to transpose the</span>
<span id="cb64-722"><a href="#cb64-722" aria-hidden="true" tabindex="-1"></a>dataframe. We then convert the resulting matrix to a dataframe using</span>
<span id="cb64-723"><a href="#cb64-723" aria-hidden="true" tabindex="-1"></a><span class="in">`data.frame()`</span>. If you view that dataframe you'll see it has default</span>
<span id="cb64-724"><a href="#cb64-724" aria-hidden="true" tabindex="-1"></a>column name which we can fix using <span class="in">`colnames()`</span> to set the column names</span>
<span id="cb64-725"><a href="#cb64-725" aria-hidden="true" tabindex="-1"></a>to the gene ids.</span>
<span id="cb64-726"><a href="#cb64-726" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-727"><a href="#cb64-727" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Set the column names to the gene ids:</span>
<span id="cb64-728"><a href="#cb64-728" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-731"><a href="#cb64-731" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-732"><a href="#cb64-732" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pro_meta_log2_trans) <span class="ot">&lt;-</span> pro_meta_results<span class="sc">$</span>gene_id</span>
<span id="cb64-733"><a href="#cb64-733" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-734"><a href="#cb64-734" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-735"><a href="#cb64-735" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Perform PCA on the log~2~ transformed normalised counts:</span>
<span id="cb64-736"><a href="#cb64-736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-739"><a href="#cb64-739" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-740"><a href="#cb64-740" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> pro_meta_log2_trans <span class="sc">|&gt;</span></span>
<span id="cb64-741"><a href="#cb64-741" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prcomp</span>(<span class="at">rank. =</span> <span class="dv">4</span>, <span class="at">scale =</span> <span class="cn">TRUE</span>) </span>
<span id="cb64-742"><a href="#cb64-742" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-743"><a href="#cb64-743" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-744"><a href="#cb64-744" aria-hidden="true" tabindex="-1"></a>The <span class="in">`scale`</span> argument tells <span class="in">`prcomp()`</span> to scale the data before doing the</span>
<span id="cb64-745"><a href="#cb64-745" aria-hidden="true" tabindex="-1"></a>PCA. This is important when the variables are on different scales to</span>
<span id="cb64-746"><a href="#cb64-746" aria-hidden="true" tabindex="-1"></a>stop variables with large values dominating the PCA. The <span class="in">`rank.`</span></span>
<span id="cb64-747"><a href="#cb64-747" aria-hidden="true" tabindex="-1"></a>argument tells <span class="in">`prcomp()`</span> to only calculate the first 4 principal</span>
<span id="cb64-748"><a href="#cb64-748" aria-hidden="true" tabindex="-1"></a>components. This is useful for visualisation as we can only plot in 2 or</span>
<span id="cb64-749"><a href="#cb64-749" aria-hidden="true" tabindex="-1"></a>3 dimensions. We can see the results of the PCA by viewing the</span>
<span id="cb64-750"><a href="#cb64-750" aria-hidden="true" tabindex="-1"></a><span class="in">`summary()`</span> of the <span class="in">`pca`</span> object.</span>
<span id="cb64-751"><a href="#cb64-751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-754"><a href="#cb64-754" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-755"><a href="#cb64-755" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pca)</span>
<span id="cb64-756"><a href="#cb64-756" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-757"><a href="#cb64-757" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-758"><a href="#cb64-758" aria-hidden="true" tabindex="-1"></a>The Proportion of Variance tells us how much of the variance is</span>
<span id="cb64-759"><a href="#cb64-759" aria-hidden="true" tabindex="-1"></a>explained by each component. We can see that the first component</span>
<span id="cb64-760"><a href="#cb64-760" aria-hidden="true" tabindex="-1"></a>explains 0.5175 of the variance, the second 0.2088, and the third</span>
<span id="cb64-761"><a href="#cb64-761" aria-hidden="true" tabindex="-1"></a>0.1025. Together the first three components explain nearly 92% of the</span>
<span id="cb64-762"><a href="#cb64-762" aria-hidden="true" tabindex="-1"></a>total variance in the data. Plotting PC1 against PC2 will capture about</span>
<span id="cb64-763"><a href="#cb64-763" aria-hidden="true" tabindex="-1"></a>66% of the variance which is likely very much better than we would get</span>
<span id="cb64-764"><a href="#cb64-764" aria-hidden="true" tabindex="-1"></a>plotting any two genes against each other. To plot the PC1 against PC2</span>
<span id="cb64-765"><a href="#cb64-765" aria-hidden="true" tabindex="-1"></a>we will need to extract the PC1 and PC2 "scores" from the PCA object and</span>
<span id="cb64-766"><a href="#cb64-766" aria-hidden="true" tabindex="-1"></a>add labels for the samples. Those labels will come from the row names of</span>
<span id="cb64-767"><a href="#cb64-767" aria-hidden="true" tabindex="-1"></a>the transformed data which has the sample ids and from the metadata.</span>
<span id="cb64-768"><a href="#cb64-768" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-769"><a href="#cb64-769" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a vector of the sample ids from the row names. These include</span>
<span id="cb64-770"><a href="#cb64-770" aria-hidden="true" tabindex="-1"></a>the <span class="in">`log2`</span> prefix which we can removed for labelling:</span>
<span id="cb64-771"><a href="#cb64-771" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-774"><a href="#cb64-774" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-775"><a href="#cb64-775" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">&lt;-</span> <span class="fu">row.names</span>(pro_meta_log2_trans) <span class="sc">|&gt;</span> <span class="fu">str_remove</span>(<span class="st">"log2_"</span>)</span>
<span id="cb64-776"><a href="#cb64-776" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-777"><a href="#cb64-777" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-778"><a href="#cb64-778" aria-hidden="true" tabindex="-1"></a>You might want to check the result.</span>
<span id="cb64-779"><a href="#cb64-779" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-780"><a href="#cb64-780" aria-hidden="true" tabindex="-1"></a>Now we will extract the PC1 and PC2 scores from the PCA object and add.</span>
<span id="cb64-781"><a href="#cb64-781" aria-hidden="true" tabindex="-1"></a>Our PCA object is called <span class="in">`pca`</span> and the scores are in pca\$x. We will</span>
<span id="cb64-782"><a href="#cb64-782" aria-hidden="true" tabindex="-1"></a>create a dataframe of the scores and add the sample ids.</span>
<span id="cb64-783"><a href="#cb64-783" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-784"><a href="#cb64-784" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a dataframe of PC1 and PC2 scores and add the sample ids:</span>
<span id="cb64-785"><a href="#cb64-785" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-788"><a href="#cb64-788" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-789"><a href="#cb64-789" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(pca<span class="sc">$</span>x,</span>
<span id="cb64-790"><a href="#cb64-790" aria-hidden="true" tabindex="-1"></a>                           sample_id)</span>
<span id="cb64-791"><a href="#cb64-791" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-792"><a href="#cb64-792" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-793"><a href="#cb64-793" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Merge with the metadata so we can label points by life cycle stage:</span>
<span id="cb64-794"><a href="#cb64-794" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-797"><a href="#cb64-797" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-798"><a href="#cb64-798" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="ot">&lt;-</span> pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb64-799"><a href="#cb64-799" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(meta_pro_meta, </span>
<span id="cb64-800"><a href="#cb64-800" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb64-801"><a href="#cb64-801" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-802"><a href="#cb64-802" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-803"><a href="#cb64-803" aria-hidden="true" tabindex="-1"></a>Since the metadata contained the sample ids, it was especially important</span>
<span id="cb64-804"><a href="#cb64-804" aria-hidden="true" tabindex="-1"></a>to remove the <span class="in">`log2_`</span> from the row names so that the join would work.</span>
<span id="cb64-805"><a href="#cb64-805" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-806"><a href="#cb64-806" aria-hidden="true" tabindex="-1"></a>The dataframe should look like this:</span>
<span id="cb64-807"><a href="#cb64-807" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-810"><a href="#cb64-810" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-811"><a href="#cb64-811" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb64-812"><a href="#cb64-812" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(pca_labelled)</span>
<span id="cb64-813"><a href="#cb64-813" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-814"><a href="#cb64-814" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-815"><a href="#cb64-815" aria-hidden="true" tabindex="-1"></a>The next task is to plot PC2 against PC1 and colour by sibling pair.</span>
<span id="cb64-816"><a href="#cb64-816" aria-hidden="true" tabindex="-1"></a>This is just a scatterplot so we can use <span class="in">`geom_point()`</span>. We will use</span>
<span id="cb64-817"><a href="#cb64-817" aria-hidden="true" tabindex="-1"></a>colour to indicate the life cycle stage.</span>
<span id="cb64-818"><a href="#cb64-818" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-819"><a href="#cb64-819" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Plot PC2 against PC1 and colour by copper conditions:</span>
<span id="cb64-820"><a href="#cb64-820" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-823"><a href="#cb64-823" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-824"><a href="#cb64-824" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb64-825"><a href="#cb64-825" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, </span>
<span id="cb64-826"><a href="#cb64-826" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> stage)) <span class="sc">+</span></span>
<span id="cb64-827"><a href="#cb64-827" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb64-828"><a href="#cb64-828" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb64-829"><a href="#cb64-829" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-830"><a href="#cb64-830" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-831"><a href="#cb64-831" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-832"><a href="#cb64-832" aria-hidden="true" tabindex="-1"></a>There is a good separation between treatments on PCA1. The replicates do</span>
<span id="cb64-833"><a href="#cb64-833" aria-hidden="true" tabindex="-1"></a>seem to cluster together. You can also try plotting PC3 or PC4.</span>
<span id="cb64-834"><a href="#cb64-834" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-835"><a href="#cb64-835" aria-hidden="true" tabindex="-1"></a>I prefer to customise the colours. I especially like the\</span>
<span id="cb64-836"><a href="#cb64-836" aria-hidden="true" tabindex="-1"></a>viridis colour scales which provide colour scales that are perceptually</span>
<span id="cb64-837"><a href="#cb64-837" aria-hidden="true" tabindex="-1"></a>uniform in both colour and black-and-white. They are also designed to be</span>
<span id="cb64-838"><a href="#cb64-838" aria-hidden="true" tabindex="-1"></a>perceived by viewers with common forms of colour blindness. See</span>
<span id="cb64-839"><a href="#cb64-839" aria-hidden="true" tabindex="-1"></a>[Introduction to</span>
<span id="cb64-840"><a href="#cb64-840" aria-hidden="true" tabindex="-1"></a>viridis](https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html#introduction)</span>
<span id="cb64-841"><a href="#cb64-841" aria-hidden="true" tabindex="-1"></a>for more information.</span>
<span id="cb64-842"><a href="#cb64-842" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-843"><a href="#cb64-843" aria-hidden="true" tabindex="-1"></a><span class="in">`ggplot`</span> provides functions to access the viridis scales. Here I use</span>
<span id="cb64-844"><a href="#cb64-844" aria-hidden="true" tabindex="-1"></a><span class="in">`scale_fill_viridis_d()`</span>. The d stands for discrete. The function</span>
<span id="cb64-845"><a href="#cb64-845" aria-hidden="true" tabindex="-1"></a><span class="in">`scale_fill_viridis_c()`</span> would be used for continuous data. Iâ€™ve used</span>
<span id="cb64-846"><a href="#cb64-846" aria-hidden="true" tabindex="-1"></a>the default â€œviridisâ€ (or â€œDâ€) option (do ?scale_fill_viridis_d for all</span>
<span id="cb64-847"><a href="#cb64-847" aria-hidden="true" tabindex="-1"></a>the options) and used the <span class="in">`begin`</span> and <span class="in">`end`</span> arguments to control the</span>
<span id="cb64-848"><a href="#cb64-848" aria-hidden="true" tabindex="-1"></a>range of colour - I have set the range to be from 0.15 to 0.95 the avoid</span>
<span id="cb64-849"><a href="#cb64-849" aria-hidden="true" tabindex="-1"></a>the strongest contrast. I have also set the <span class="in">`name`</span> argument to provide a</span>
<span id="cb64-850"><a href="#cb64-850" aria-hidden="true" tabindex="-1"></a>label for the legend.</span>
<span id="cb64-851"><a href="#cb64-851" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-852"><a href="#cb64-852" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Plot PC2 against PC1 and colour by life stage:</span>
<span id="cb64-853"><a href="#cb64-853" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-856"><a href="#cb64-856" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-857"><a href="#cb64-857" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb64-858"><a href="#cb64-858" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, </span>
<span id="cb64-859"><a href="#cb64-859" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> stage)) <span class="sc">+</span></span>
<span id="cb64-860"><a href="#cb64-860" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb64-861"><a href="#cb64-861" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_viridis_d</span>(<span class="at">end =</span> <span class="fl">0.95</span>, <span class="at">begin =</span> <span class="fl">0.15</span>,</span>
<span id="cb64-862"><a href="#cb64-862" aria-hidden="true" tabindex="-1"></a>                         <span class="at">name =</span> <span class="st">"Stage"</span>) <span class="sc">+</span></span>
<span id="cb64-863"><a href="#cb64-863" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb64-864"><a href="#cb64-864" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-865"><a href="#cb64-865" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-866"><a href="#cb64-866" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-867"><a href="#cb64-867" aria-hidden="true" tabindex="-1"></a>Now go to Volcano plots for <span class="co">[</span><span class="ot">ğŸ’‰ Leishmania</span><span class="co">](#leishmania-2)</span></span>
<span id="cb64-868"><a href="#cb64-868" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-869"><a href="#cb64-869" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ­ Stem cells {#stem-cells-2}</span></span>
<span id="cb64-870"><a href="#cb64-870" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-871"><a href="#cb64-871" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Transpose the log~2~ transformed normalised counts:</span>
<span id="cb64-872"><a href="#cb64-872" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-875"><a href="#cb64-875" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-876"><a href="#cb64-876" aria-hidden="true" tabindex="-1"></a>hspc_prog_trans <span class="ot">&lt;-</span> hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-877"><a href="#cb64-877" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">starts_with</span>(<span class="fu">c</span>(<span class="st">"HSPC_"</span>, <span class="st">"Prog_"</span>))) <span class="sc">|&gt;</span></span>
<span id="cb64-878"><a href="#cb64-878" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">|&gt;</span> </span>
<span id="cb64-879"><a href="#cb64-879" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>()</span>
<span id="cb64-880"><a href="#cb64-880" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-881"><a href="#cb64-881" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-882"><a href="#cb64-882" aria-hidden="true" tabindex="-1"></a>We have used the <span class="in">`select()`</span> function to select all the columns that</span>
<span id="cb64-883"><a href="#cb64-883" aria-hidden="true" tabindex="-1"></a>start with <span class="in">`Prog_`</span> or <span class="in">`HSPC_`</span>. We then use the <span class="in">`t()`</span> function to</span>
<span id="cb64-884"><a href="#cb64-884" aria-hidden="true" tabindex="-1"></a>transpose the dataframe. We then convert the resulting matrix to a</span>
<span id="cb64-885"><a href="#cb64-885" aria-hidden="true" tabindex="-1"></a>dataframe using <span class="in">`data.frame()`</span>. If you view that dataframe you'll see it</span>
<span id="cb64-886"><a href="#cb64-886" aria-hidden="true" tabindex="-1"></a>has default column name which we can fix using <span class="in">`colnames()`</span> to set the</span>
<span id="cb64-887"><a href="#cb64-887" aria-hidden="true" tabindex="-1"></a>column names to the gene ids.</span>
<span id="cb64-888"><a href="#cb64-888" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-889"><a href="#cb64-889" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Set the column names to the gene ids:</span>
<span id="cb64-890"><a href="#cb64-890" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-893"><a href="#cb64-893" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-894"><a href="#cb64-894" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(hspc_prog_trans) <span class="ot">&lt;-</span> hspc_prog_results<span class="sc">$</span>ensembl_gene_id</span>
<span id="cb64-895"><a href="#cb64-895" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-896"><a href="#cb64-896" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-897"><a href="#cb64-897" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Perform PCA on the log~2~ transformed normalised counts:</span>
<span id="cb64-898"><a href="#cb64-898" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-901"><a href="#cb64-901" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-902"><a href="#cb64-902" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> hspc_prog_trans <span class="sc">|&gt;</span></span>
<span id="cb64-903"><a href="#cb64-903" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prcomp</span>(<span class="at">rank. =</span> <span class="dv">8</span>)</span>
<span id="cb64-904"><a href="#cb64-904" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-905"><a href="#cb64-905" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-906"><a href="#cb64-906" aria-hidden="true" tabindex="-1"></a>The <span class="in">`rank.`</span> argument tells <span class="in">`prcomp()`</span> to only calculate the first 8</span>
<span id="cb64-907"><a href="#cb64-907" aria-hidden="true" tabindex="-1"></a>principal components. This is useful for visualisation as we can only</span>
<span id="cb64-908"><a href="#cb64-908" aria-hidden="true" tabindex="-1"></a>plot in 2 or 3 dimensions. We can see the results of the PCA by viewing</span>
<span id="cb64-909"><a href="#cb64-909" aria-hidden="true" tabindex="-1"></a>the <span class="in">`summary()`</span> of the <span class="in">`pca`</span> object.</span>
<span id="cb64-910"><a href="#cb64-910" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-913"><a href="#cb64-913" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-914"><a href="#cb64-914" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pca)</span>
<span id="cb64-915"><a href="#cb64-915" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-916"><a href="#cb64-916" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-917"><a href="#cb64-917" aria-hidden="true" tabindex="-1"></a>The Proportion of Variance tells us how much of the variance is</span>
<span id="cb64-918"><a href="#cb64-918" aria-hidden="true" tabindex="-1"></a>explained by each component. We can see that the first component</span>
<span id="cb64-919"><a href="#cb64-919" aria-hidden="true" tabindex="-1"></a>explains 0.1099 of the variance, the second 0.04874, and the third</span>
<span id="cb64-920"><a href="#cb64-920" aria-hidden="true" tabindex="-1"></a>0.02498. Together the first three components explain 18% of the total</span>
<span id="cb64-921"><a href="#cb64-921" aria-hidden="true" tabindex="-1"></a>variance in the data. This is not *that* high but it is also quite a lot</span>
<span id="cb64-922"><a href="#cb64-922" aria-hidden="true" tabindex="-1"></a>better than than we would get plotting any two genes randomly chosen</span>
<span id="cb64-923"><a href="#cb64-923" aria-hidden="true" tabindex="-1"></a>against each other.</span>
<span id="cb64-924"><a href="#cb64-924" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-925"><a href="#cb64-925" aria-hidden="true" tabindex="-1"></a>To plot the PC1 against PC2 we will need to extract the PC1 and PC2</span>
<span id="cb64-926"><a href="#cb64-926" aria-hidden="true" tabindex="-1"></a>"scores" from the PCA object and add labels for the cells. Our PCA</span>
<span id="cb64-927"><a href="#cb64-927" aria-hidden="true" tabindex="-1"></a>object is called <span class="in">`pca`</span> and the scores are in pca\$x. The cells labels</span>
<span id="cb64-928"><a href="#cb64-928" aria-hidden="true" tabindex="-1"></a>will come from the row names of the transformed data.</span>
<span id="cb64-929"><a href="#cb64-929" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-930"><a href="#cb64-930" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a dataframe of the PC1 and PC2 scores (in <span class="in">`pca$x`</span>) and add the</span>
<span id="cb64-931"><a href="#cb64-931" aria-hidden="true" tabindex="-1"></a>cell ids:</span>
<span id="cb64-932"><a href="#cb64-932" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-935"><a href="#cb64-935" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-936"><a href="#cb64-936" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(pca<span class="sc">$</span>x,</span>
<span id="cb64-937"><a href="#cb64-937" aria-hidden="true" tabindex="-1"></a>                           <span class="at">cell_id =</span> <span class="fu">row.names</span>(hspc_prog_trans))</span>
<span id="cb64-938"><a href="#cb64-938" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-939"><a href="#cb64-939" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-940"><a href="#cb64-940" aria-hidden="true" tabindex="-1"></a>It will be helpful to add a column for the cell type so we can label</span>
<span id="cb64-941"><a href="#cb64-941" aria-hidden="true" tabindex="-1"></a>points. One way to do this is to extract the information in the cell_id</span>
<span id="cb64-942"><a href="#cb64-942" aria-hidden="true" tabindex="-1"></a>column into two columns: one with the complete cell id and one with just</span>
<span id="cb64-943"><a href="#cb64-943" aria-hidden="true" tabindex="-1"></a>the cell type. This extraction is done with [Regular</span>
<span id="cb64-944"><a href="#cb64-944" aria-hidden="true" tabindex="-1"></a>Expression](https://en.wikipedia.org/wiki/Regular_expression) magic.</span>
<span id="cb64-945"><a href="#cb64-945" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-946"><a href="#cb64-946" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Extract the cell type and cell number from the <span class="in">`cell_id`</span> column</span>
<span id="cb64-947"><a href="#cb64-947" aria-hidden="true" tabindex="-1"></a>(keeping the <span class="in">`cell_id`</span> column):</span>
<span id="cb64-948"><a href="#cb64-948" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-951"><a href="#cb64-951" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-952"><a href="#cb64-952" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="ot">&lt;-</span> pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb64-953"><a href="#cb64-953" aria-hidden="true" tabindex="-1"></a>  <span class="fu">extract</span>(cell_id, </span>
<span id="cb64-954"><a href="#cb64-954" aria-hidden="true" tabindex="-1"></a>          <span class="at">remove =</span> <span class="cn">FALSE</span>,</span>
<span id="cb64-955"><a href="#cb64-955" aria-hidden="true" tabindex="-1"></a>          <span class="fu">c</span>(<span class="st">"cell_type"</span>, <span class="st">"cell_number"</span>),</span>
<span id="cb64-956"><a href="#cb64-956" aria-hidden="true" tabindex="-1"></a>          <span class="st">"([a-zA-Z]{4})_([0-9]{3})"</span>)</span>
<span id="cb64-957"><a href="#cb64-957" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-958"><a href="#cb64-958" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-959"><a href="#cb64-959" aria-hidden="true" tabindex="-1"></a>What this code does is take what is in the <span class="in">`cell_id`</span> column (something</span>
<span id="cb64-960"><a href="#cb64-960" aria-hidden="true" tabindex="-1"></a>like <span class="in">`Prog_001`</span> or <span class="in">`HSPC_001`</span>) and split it into two columns</span>
<span id="cb64-961"><a href="#cb64-961" aria-hidden="true" tabindex="-1"></a>("cell_type" and "cell_number"). The reason why we want to do that is to</span>
<span id="cb64-962"><a href="#cb64-962" aria-hidden="true" tabindex="-1"></a>colour the points by cell type. We would not want to use <span class="in">`cell_id`</span> to</span>
<span id="cb64-963"><a href="#cb64-963" aria-hidden="true" tabindex="-1"></a>colour the points because each cell id is unique and that would be 1000+</span>
<span id="cb64-964"><a href="#cb64-964" aria-hidden="true" tabindex="-1"></a>colours. The last argument in the <span class="in">`extract()`</span> function is the pattern to</span>
<span id="cb64-965"><a href="#cb64-965" aria-hidden="true" tabindex="-1"></a>match described with a [regular</span>
<span id="cb64-966"><a href="#cb64-966" aria-hidden="true" tabindex="-1"></a>expression](https://en.wikipedia.org/wiki/Regular_expression). Three</span>
<span id="cb64-967"><a href="#cb64-967" aria-hidden="true" tabindex="-1"></a>patterns are being matched, and two of those are in brackets meaning</span>
<span id="cb64-968"><a href="#cb64-968" aria-hidden="true" tabindex="-1"></a>they are kept to fill the two new columns.</span>
<span id="cb64-969"><a href="#cb64-969" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-970"><a href="#cb64-970" aria-hidden="true" tabindex="-1"></a>The first pattern is <span class="in">`([a-zA-Z]{4})`</span></span>
<span id="cb64-971"><a href="#cb64-971" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-972"><a href="#cb64-972" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>it is brackets because we want to keep it and put it in <span class="in">`cell_type`</span></span>
<span id="cb64-973"><a href="#cb64-973" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`[a-zA-Z]`</span> means any lower (<span class="in">`a-z`</span>) or upper case letter (<span class="in">`A-Z`</span>).</span>
<span id="cb64-974"><a href="#cb64-974" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The square brackets means any of the characters in the square</span>
<span id="cb64-975"><a href="#cb64-975" aria-hidden="true" tabindex="-1"></a>    brackets will be matched</span>
<span id="cb64-976"><a href="#cb64-976" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`{4}`</span> means 4 of them.</span>
<span id="cb64-977"><a href="#cb64-977" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-978"><a href="#cb64-978" aria-hidden="true" tabindex="-1"></a>So the first pattern inside the first (...) will match exactly 4 upper</span>
<span id="cb64-979"><a href="#cb64-979" aria-hidden="true" tabindex="-1"></a>or lower case letters (like Prog or HSPC)</span>
<span id="cb64-980"><a href="#cb64-980" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-981"><a href="#cb64-981" aria-hidden="true" tabindex="-1"></a>The second pattern is <span class="in">`_`</span> to match the underscore in every cell id that</span>
<span id="cb64-982"><a href="#cb64-982" aria-hidden="true" tabindex="-1"></a>separates the cell type from the number. It is not in brackets because</span>
<span id="cb64-983"><a href="#cb64-983" aria-hidden="true" tabindex="-1"></a>we do not want to keep it.</span>
<span id="cb64-984"><a href="#cb64-984" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-985"><a href="#cb64-985" aria-hidden="true" tabindex="-1"></a>The third pattern is <span class="in">`([0-9]{3})`</span></span>
<span id="cb64-986"><a href="#cb64-986" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-987"><a href="#cb64-987" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`[0-9]`</span> means any number</span>
<span id="cb64-988"><a href="#cb64-988" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`{3}`</span> means 3 of them.</span>
<span id="cb64-989"><a href="#cb64-989" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-990"><a href="#cb64-990" aria-hidden="true" tabindex="-1"></a>So the second pattern inside the second (...) will match exactly 3</span>
<span id="cb64-991"><a href="#cb64-991" aria-hidden="true" tabindex="-1"></a>numbers (like 001 or 851).</span>
<span id="cb64-992"><a href="#cb64-992" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-993"><a href="#cb64-993" aria-hidden="true" tabindex="-1"></a>**Important**: Prog and HPSC have 4 letters. The column names, LT.HSPC<span class="sc">\_</span></span>
<span id="cb64-994"><a href="#cb64-994" aria-hidden="true" tabindex="-1"></a>have 6 characters and includes a dot. You will need to adjust the regex</span>
<span id="cb64-995"><a href="#cb64-995" aria-hidden="true" tabindex="-1"></a>when make comparison between LT-HSC and other cell types. The pattern to</span>
<span id="cb64-996"><a href="#cb64-996" aria-hidden="true" tabindex="-1"></a>match the <span class="in">`LT.HSC`</span> as well as the <span class="in">`Prog`</span> and <span class="in">`HSPC`</span> is</span>
<span id="cb64-997"><a href="#cb64-997" aria-hidden="true" tabindex="-1"></a><span class="in">`([a-zA-Z.]{4,6})`</span>. Note the dot inside the square brackets and numbers</span>
<span id="cb64-998"><a href="#cb64-998" aria-hidden="true" tabindex="-1"></a>meaning 4 or 6 of. The pattern to match the underscore and the cell</span>
<span id="cb64-999"><a href="#cb64-999" aria-hidden="true" tabindex="-1"></a>number is the same.</span>
<span id="cb64-1000"><a href="#cb64-1000" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1001"><a href="#cb64-1001" aria-hidden="true" tabindex="-1"></a>The top of the dataframe should look like this (but with more decimal</span>
<span id="cb64-1002"><a href="#cb64-1002" aria-hidden="true" tabindex="-1"></a>places)</span>
<span id="cb64-1003"><a href="#cb64-1003" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1006"><a href="#cb64-1006" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1007"><a href="#cb64-1007" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb64-1008"><a href="#cb64-1008" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(pca_labelled), <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb64-1009"><a href="#cb64-1009" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1010"><a href="#cb64-1010" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1011"><a href="#cb64-1011" aria-hidden="true" tabindex="-1"></a>The next task is to plot PC2 against PC1 and colour by cell type. This</span>
<span id="cb64-1012"><a href="#cb64-1012" aria-hidden="true" tabindex="-1"></a>is just a scatterplot so we can use <span class="in">`geom_point()`</span>. We will use colour</span>
<span id="cb64-1013"><a href="#cb64-1013" aria-hidden="true" tabindex="-1"></a>to indicate the cell type.</span>
<span id="cb64-1014"><a href="#cb64-1014" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1015"><a href="#cb64-1015" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Plot PC2 against PC1 and colour by cell type:</span>
<span id="cb64-1016"><a href="#cb64-1016" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1019"><a href="#cb64-1019" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1020"><a href="#cb64-1020" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb64-1021"><a href="#cb64-1021" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, </span>
<span id="cb64-1022"><a href="#cb64-1022" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> cell_type)) <span class="sc">+</span></span>
<span id="cb64-1023"><a href="#cb64-1023" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb64-1024"><a href="#cb64-1024" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb64-1025"><a href="#cb64-1025" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1026"><a href="#cb64-1026" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1027"><a href="#cb64-1027" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1028"><a href="#cb64-1028" aria-hidden="true" tabindex="-1"></a>There is a good clustering of cell types but plenty of overlap. You can</span>
<span id="cb64-1029"><a href="#cb64-1029" aria-hidden="true" tabindex="-1"></a>also try plotting PC3 or PC4 (or others).</span>
<span id="cb64-1030"><a href="#cb64-1030" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1031"><a href="#cb64-1031" aria-hidden="true" tabindex="-1"></a>I prefer to customise the colours. I especially like the viridis colour</span>
<span id="cb64-1032"><a href="#cb64-1032" aria-hidden="true" tabindex="-1"></a>scales which provide colour scales that are perceptually uniform in both</span>
<span id="cb64-1033"><a href="#cb64-1033" aria-hidden="true" tabindex="-1"></a>colour and black-and-white. They are also designed to be perceived by</span>
<span id="cb64-1034"><a href="#cb64-1034" aria-hidden="true" tabindex="-1"></a>viewers with common forms of colour blindness. See [Introduction to</span>
<span id="cb64-1035"><a href="#cb64-1035" aria-hidden="true" tabindex="-1"></a>viridis](https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html#introduction)</span>
<span id="cb64-1036"><a href="#cb64-1036" aria-hidden="true" tabindex="-1"></a>for more information.</span>
<span id="cb64-1037"><a href="#cb64-1037" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1038"><a href="#cb64-1038" aria-hidden="true" tabindex="-1"></a><span class="in">`ggplot`</span> provides functions to access the viridis scales. Here I use</span>
<span id="cb64-1039"><a href="#cb64-1039" aria-hidden="true" tabindex="-1"></a><span class="in">`scale_fill_viridis_d()`</span>. The d stands for discrete used because cell</span>
<span id="cb64-1040"><a href="#cb64-1040" aria-hidden="true" tabindex="-1"></a>type is a discrete variable. The function <span class="in">`scale_fill_viridis_c()`</span> would</span>
<span id="cb64-1041"><a href="#cb64-1041" aria-hidden="true" tabindex="-1"></a>be used for continuous data. Iâ€™ve used the default â€œviridisâ€ (or â€œDâ€)</span>
<span id="cb64-1042"><a href="#cb64-1042" aria-hidden="true" tabindex="-1"></a>option (do ?scale_fill_viridis_d for all the options) and used the</span>
<span id="cb64-1043"><a href="#cb64-1043" aria-hidden="true" tabindex="-1"></a><span class="in">`begin`</span> and <span class="in">`end`</span> arguments to control the range of colour - I have set</span>
<span id="cb64-1044"><a href="#cb64-1044" aria-hidden="true" tabindex="-1"></a>the range to be from 0.15 to 0.95 the avoid the strongest contrast. I</span>
<span id="cb64-1045"><a href="#cb64-1045" aria-hidden="true" tabindex="-1"></a>have also set the <span class="in">`name`</span> argument to NULL because that the legend refers</span>
<span id="cb64-1046"><a href="#cb64-1046" aria-hidden="true" tabindex="-1"></a>to cell types is obvious.</span>
<span id="cb64-1047"><a href="#cb64-1047" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1048"><a href="#cb64-1048" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Plot PC2 against PC1 and colour by cell type:</span>
<span id="cb64-1049"><a href="#cb64-1049" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1052"><a href="#cb64-1052" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1053"><a href="#cb64-1053" aria-hidden="true" tabindex="-1"></a>pca_labelled <span class="sc">|&gt;</span> </span>
<span id="cb64-1054"><a href="#cb64-1054" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, </span>
<span id="cb64-1055"><a href="#cb64-1055" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> cell_type)) <span class="sc">+</span></span>
<span id="cb64-1056"><a href="#cb64-1056" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb64-1057"><a href="#cb64-1057" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_viridis_d</span>(<span class="at">end =</span> <span class="fl">0.95</span>, <span class="at">begin =</span> <span class="fl">0.15</span>,</span>
<span id="cb64-1058"><a href="#cb64-1058" aria-hidden="true" tabindex="-1"></a>                         <span class="at">name =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb64-1059"><a href="#cb64-1059" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb64-1060"><a href="#cb64-1060" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1061"><a href="#cb64-1061" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1062"><a href="#cb64-1062" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1063"><a href="#cb64-1063" aria-hidden="true" tabindex="-1"></a>Now go to Volcano plots for <span class="co">[</span><span class="ot">ğŸ­ Stem cells</span><span class="co">](#stem-cells-3)</span></span>
<span id="cb64-1064"><a href="#cb64-1064" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1065"><a href="#cb64-1065" aria-hidden="true" tabindex="-1"></a><span class="fu"># Visualise all the results with a volcano plot</span></span>
<span id="cb64-1066"><a href="#cb64-1066" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1067"><a href="#cb64-1067" aria-hidden="true" tabindex="-1"></a>A volcano plot is a scatterplot that shows statistical significance</span>
<span id="cb64-1068"><a href="#cb64-1068" aria-hidden="true" tabindex="-1"></a>(p-value) versus magnitude of change (fold change). As the [independent</span>
<span id="cb64-1069"><a href="#cb64-1069" aria-hidden="true" tabindex="-1"></a>study](study_before_workshop.html) explained, we plot -log~10~(adjusted</span>
<span id="cb64-1070"><a href="#cb64-1070" aria-hidden="true" tabindex="-1"></a>*p*-value) on the *y*-axis so that the most significant genes are</span>
<span id="cb64-1071"><a href="#cb64-1071" aria-hidden="true" tabindex="-1"></a>uppermost.</span>
<span id="cb64-1072"><a href="#cb64-1072" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1073"><a href="#cb64-1073" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ„ *Arabidopsis* {#arabidopsis-2}</span></span>
<span id="cb64-1074"><a href="#cb64-1074" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1075"><a href="#cb64-1075" aria-hidden="true" tabindex="-1"></a>We will add a column to the results dataframe that contains the</span>
<span id="cb64-1076"><a href="#cb64-1076" aria-hidden="true" tabindex="-1"></a>-log~10~(padj). You could perform this transformation within the plot</span>
<span id="cb64-1077"><a href="#cb64-1077" aria-hidden="true" tabindex="-1"></a>command without adding a column to the data if you prefer.</span>
<span id="cb64-1078"><a href="#cb64-1078" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1079"><a href="#cb64-1079" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add a column to the results dataframe that contains the</span>
<span id="cb64-1080"><a href="#cb64-1080" aria-hidden="true" tabindex="-1"></a>-log~10~(padj):</span>
<span id="cb64-1081"><a href="#cb64-1081" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1084"><a href="#cb64-1084" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1085"><a href="#cb64-1085" aria-hidden="true" tabindex="-1"></a>wild_results <span class="ot">&lt;-</span> wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1086"><a href="#cb64-1086" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">log10_padj =</span> <span class="sc">-</span><span class="fu">log10</span>(padj)) </span>
<span id="cb64-1087"><a href="#cb64-1087" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1088"><a href="#cb64-1088" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1089"><a href="#cb64-1089" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a volcano plot of the results:</span>
<span id="cb64-1090"><a href="#cb64-1090" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1093"><a href="#cb64-1093" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1094"><a href="#cb64-1094" aria-hidden="true" tabindex="-1"></a>wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1095"><a href="#cb64-1095" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, </span>
<span id="cb64-1096"><a href="#cb64-1096" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_padj)) <span class="sc">+</span></span>
<span id="cb64-1097"><a href="#cb64-1097" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1098"><a href="#cb64-1098" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1099"><a href="#cb64-1099" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1100"><a href="#cb64-1100" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1101"><a href="#cb64-1101" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1102"><a href="#cb64-1102" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1103"><a href="#cb64-1103" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1104"><a href="#cb64-1104" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1105"><a href="#cb64-1105" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1106"><a href="#cb64-1106" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1107"><a href="#cb64-1107" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1108"><a href="#cb64-1108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1109"><a href="#cb64-1109" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1110"><a href="#cb64-1110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1111"><a href="#cb64-1111" aria-hidden="true" tabindex="-1"></a>Our dashed lines are at -log~10~(0.05) and log~2~(2) and log~2~(-2) to</span>
<span id="cb64-1112"><a href="#cb64-1112" aria-hidden="true" tabindex="-1"></a>make more clear which genes (points) are significantly different between</span>
<span id="cb64-1113"><a href="#cb64-1113" aria-hidden="true" tabindex="-1"></a>the copper sufficient and deficient conditions samples and have a fold</span>
<span id="cb64-1114"><a href="#cb64-1114" aria-hidden="true" tabindex="-1"></a>change of at least 2.</span>
<span id="cb64-1115"><a href="#cb64-1115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1116"><a href="#cb64-1116" aria-hidden="true" tabindex="-1"></a>In most cases, people colour the points to show that the quadrants. I</span>
<span id="cb64-1117"><a href="#cb64-1117" aria-hidden="true" tabindex="-1"></a>like to add columns to the dataframe to indicate if the gene is</span>
<span id="cb64-1118"><a href="#cb64-1118" aria-hidden="true" tabindex="-1"></a>significant and if the fold change is large and use those variables in</span>
<span id="cb64-1119"><a href="#cb64-1119" aria-hidden="true" tabindex="-1"></a>the plot.</span>
<span id="cb64-1120"><a href="#cb64-1120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1121"><a href="#cb64-1121" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add columns to the results dataframe to indicate if the gene is</span>
<span id="cb64-1122"><a href="#cb64-1122" aria-hidden="true" tabindex="-1"></a>significant and if the fold change is large:</span>
<span id="cb64-1123"><a href="#cb64-1123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1126"><a href="#cb64-1126" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1127"><a href="#cb64-1127" aria-hidden="true" tabindex="-1"></a>wild_results <span class="ot">&lt;-</span> wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1128"><a href="#cb64-1128" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sig =</span> padj <span class="sc">&lt;=</span> <span class="fl">0.05</span>,</span>
<span id="cb64-1129"><a href="#cb64-1129" aria-hidden="true" tabindex="-1"></a>         <span class="at">bigfc =</span> <span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;=</span> <span class="dv">2</span>) </span>
<span id="cb64-1130"><a href="#cb64-1130" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1131"><a href="#cb64-1131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1132"><a href="#cb64-1132" aria-hidden="true" tabindex="-1"></a>The use of <span class="in">`abs()`</span> (absolute) means genes with a fold change of at least</span>
<span id="cb64-1133"><a href="#cb64-1133" aria-hidden="true" tabindex="-1"></a>2 in either direction will be considered to have a large fold change.</span>
<span id="cb64-1134"><a href="#cb64-1134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1135"><a href="#cb64-1135" aria-hidden="true" tabindex="-1"></a>Now we can colour the points by these new columns. I use <span class="in">`interaction()`</span></span>
<span id="cb64-1136"><a href="#cb64-1136" aria-hidden="true" tabindex="-1"></a>to create four categories:</span>
<span id="cb64-1137"><a href="#cb64-1137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1138"><a href="#cb64-1138" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>not significant and not large fold change (FF)</span>
<span id="cb64-1139"><a href="#cb64-1139" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>significant and not large fold change (TF)</span>
<span id="cb64-1140"><a href="#cb64-1140" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>not significant and large fold (FT)</span>
<span id="cb64-1141"><a href="#cb64-1141" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>significant and large fold change (TT)</span>
<span id="cb64-1142"><a href="#cb64-1142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1143"><a href="#cb64-1143" aria-hidden="true" tabindex="-1"></a>And I use <span class="in">`scale_colour_manual()`</span> to set the colours for these</span>
<span id="cb64-1144"><a href="#cb64-1144" aria-hidden="true" tabindex="-1"></a>categories.</span>
<span id="cb64-1145"><a href="#cb64-1145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1146"><a href="#cb64-1146" aria-hidden="true" tabindex="-1"></a>NOTE: there are no "significant and not large fold change (TF)" in this</span>
<span id="cb64-1147"><a href="#cb64-1147" aria-hidden="true" tabindex="-1"></a>case.This means we do not need four colours. I have put the "pink"</span>
<span id="cb64-1148"><a href="#cb64-1148" aria-hidden="true" tabindex="-1"></a>colour in the code but commented it out.</span>
<span id="cb64-1149"><a href="#cb64-1149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1150"><a href="#cb64-1150" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a volcano plot of the results with the points coloured by</span>
<span id="cb64-1151"><a href="#cb64-1151" aria-hidden="true" tabindex="-1"></a>significance and fold change:</span>
<span id="cb64-1152"><a href="#cb64-1152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1155"><a href="#cb64-1155" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1156"><a href="#cb64-1156" aria-hidden="true" tabindex="-1"></a>wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1157"><a href="#cb64-1157" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, </span>
<span id="cb64-1158"><a href="#cb64-1158" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_padj, </span>
<span id="cb64-1159"><a href="#cb64-1159" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> <span class="fu">interaction</span>(sig, bigfc))) <span class="sc">+</span></span>
<span id="cb64-1160"><a href="#cb64-1160" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1161"><a href="#cb64-1161" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1162"><a href="#cb64-1162" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1163"><a href="#cb64-1163" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1164"><a href="#cb64-1164" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1165"><a href="#cb64-1165" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1166"><a href="#cb64-1166" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1167"><a href="#cb64-1167" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1168"><a href="#cb64-1168" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1169"><a href="#cb64-1169" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, </span>
<span id="cb64-1170"><a href="#cb64-1170" aria-hidden="true" tabindex="-1"></a>                                <span class="co"># "pink",</span></span>
<span id="cb64-1171"><a href="#cb64-1171" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"gray30"</span>,</span>
<span id="cb64-1172"><a href="#cb64-1172" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"deeppink"</span>)) <span class="sc">+</span></span>
<span id="cb64-1173"><a href="#cb64-1173" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1174"><a href="#cb64-1174" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1175"><a href="#cb64-1175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1176"><a href="#cb64-1176" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1177"><a href="#cb64-1177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1178"><a href="#cb64-1178" aria-hidden="true" tabindex="-1"></a>For exploring the data, I like add labels to all the significant genes</span>
<span id="cb64-1179"><a href="#cb64-1179" aria-hidden="true" tabindex="-1"></a>with a large fold change so I can very quickly identity them. The</span>
<span id="cb64-1180"><a href="#cb64-1180" aria-hidden="true" tabindex="-1"></a><span class="in">`ggrepel`</span> package has a function <span class="in">`geom_text_repel()`</span> that is useful for</span>
<span id="cb64-1181"><a href="#cb64-1181" aria-hidden="true" tabindex="-1"></a>adding labels so that they don't overlap.</span>
<span id="cb64-1182"><a href="#cb64-1182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1183"><a href="#cb64-1183" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load the package:</span>
<span id="cb64-1184"><a href="#cb64-1184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1187"><a href="#cb64-1187" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1188"><a href="#cb64-1188" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb64-1189"><a href="#cb64-1189" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb64-1190"><a href="#cb64-1190" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1191"><a href="#cb64-1191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1192"><a href="#cb64-1192" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add labels to the significant genes with a large fold change:</span>
<span id="cb64-1193"><a href="#cb64-1193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1196"><a href="#cb64-1196" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1197"><a href="#cb64-1197" aria-hidden="true" tabindex="-1"></a>wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1198"><a href="#cb64-1198" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, </span>
<span id="cb64-1199"><a href="#cb64-1199" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_padj, </span>
<span id="cb64-1200"><a href="#cb64-1200" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> <span class="fu">interaction</span>(sig, bigfc))) <span class="sc">+</span></span>
<span id="cb64-1201"><a href="#cb64-1201" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1202"><a href="#cb64-1202" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1203"><a href="#cb64-1203" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1204"><a href="#cb64-1204" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1205"><a href="#cb64-1205" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1206"><a href="#cb64-1206" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1207"><a href="#cb64-1207" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1208"><a href="#cb64-1208" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1209"><a href="#cb64-1209" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1210"><a href="#cb64-1210" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, </span>
<span id="cb64-1211"><a href="#cb64-1211" aria-hidden="true" tabindex="-1"></a>                               <span class="co">#  "pink",</span></span>
<span id="cb64-1212"><a href="#cb64-1212" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"gray30"</span>,</span>
<span id="cb64-1213"><a href="#cb64-1213" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"deeppink"</span>)) <span class="sc">+</span></span>
<span id="cb64-1214"><a href="#cb64-1214" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">data =</span> wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1215"><a href="#cb64-1215" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(bigfc <span class="sc">==</span> <span class="cn">TRUE</span>, sig <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb64-1216"><a href="#cb64-1216" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">label =</span> external_gene_name),</span>
<span id="cb64-1217"><a href="#cb64-1217" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="dv">3</span>,</span>
<span id="cb64-1218"><a href="#cb64-1218" aria-hidden="true" tabindex="-1"></a>                  <span class="at">max.overlaps =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb64-1219"><a href="#cb64-1219" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1220"><a href="#cb64-1220" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1221"><a href="#cb64-1221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1222"><a href="#cb64-1222" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1223"><a href="#cb64-1223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1224"><a href="#cb64-1224" aria-hidden="true" tabindex="-1"></a>Notice that I have used <span class="in">`filter()`</span> label only the genes that are both</span>
<span id="cb64-1225"><a href="#cb64-1225" aria-hidden="true" tabindex="-1"></a>significant and have a large fold change. In systems you are familiar</span>
<span id="cb64-1226"><a href="#cb64-1226" aria-hidden="true" tabindex="-1"></a>with, this labelling is very informative and can help you quickly</span>
<span id="cb64-1227"><a href="#cb64-1227" aria-hidden="true" tabindex="-1"></a>identify common themes. Key to interpreting the volcano plot is to</span>
<span id="cb64-1228"><a href="#cb64-1228" aria-hidden="true" tabindex="-1"></a>remember that positive fold changes means the gene is up-regulated in</span>
<span id="cb64-1229"><a href="#cb64-1229" aria-hidden="true" tabindex="-1"></a>the sufficient conditions and negative fold changes means the gene is</span>
<span id="cb64-1230"><a href="#cb64-1230" aria-hidden="true" tabindex="-1"></a>down-regulated (i.e., higher in the deficient). This was determined by</span>
<span id="cb64-1231"><a href="#cb64-1231" aria-hidden="true" tabindex="-1"></a>the order of the treatments in the [contrast used in the DESeq2</span>
<span id="cb64-1232"><a href="#cb64-1232" aria-hidden="true" tabindex="-1"></a>analysis](../week-4/workshop.html#differential-expression-analysis-2)</span>
<span id="cb64-1233"><a href="#cb64-1233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1234"><a href="#cb64-1234" aria-hidden="true" tabindex="-1"></a>If you do forget which way round you did the comparison, you can always</span>
<span id="cb64-1235"><a href="#cb64-1235" aria-hidden="true" tabindex="-1"></a>examine the results dataframe to see which of the treatments seem to be</span>
<span id="cb64-1236"><a href="#cb64-1236" aria-hidden="true" tabindex="-1"></a>higher for the positive fold changes.</span>
<span id="cb64-1237"><a href="#cb64-1237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1238"><a href="#cb64-1238" aria-hidden="true" tabindex="-1"></a>When you have a gene of interest, you may wish to label it, and only it,</span>
<span id="cb64-1239"><a href="#cb64-1239" aria-hidden="true" tabindex="-1"></a>on the plot. This is done in the same way except that you filter the</span>
<span id="cb64-1240"><a href="#cb64-1240" aria-hidden="true" tabindex="-1"></a>data to only include the gene of interest. I have used and then use</span>
<span id="cb64-1241"><a href="#cb64-1241" aria-hidden="true" tabindex="-1"></a><span class="in">`geom_label_repel()`</span> rather than <span class="in">`geom_text_repel()`</span> to put the label in</span>
<span id="cb64-1242"><a href="#cb64-1242" aria-hidden="true" tabindex="-1"></a>a box and nudged it's position to get a line connecting the point and</span>
<span id="cb64-1243"><a href="#cb64-1243" aria-hidden="true" tabindex="-1"></a>the label. I have also increased the size of the point.</span>
<span id="cb64-1244"><a href="#cb64-1244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1245"><a href="#cb64-1245" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add a label to one gene of interest (hoxb9.S) and :</span>
<span id="cb64-1246"><a href="#cb64-1246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1249"><a href="#cb64-1249" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1250"><a href="#cb64-1250" aria-hidden="true" tabindex="-1"></a>wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1251"><a href="#cb64-1251" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, </span>
<span id="cb64-1252"><a href="#cb64-1252" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_padj, </span>
<span id="cb64-1253"><a href="#cb64-1253" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> <span class="fu">interaction</span>(sig, bigfc))) <span class="sc">+</span></span>
<span id="cb64-1254"><a href="#cb64-1254" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1255"><a href="#cb64-1255" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1256"><a href="#cb64-1256" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1257"><a href="#cb64-1257" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1258"><a href="#cb64-1258" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1259"><a href="#cb64-1259" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1260"><a href="#cb64-1260" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1261"><a href="#cb64-1261" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1262"><a href="#cb64-1262" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1263"><a href="#cb64-1263" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, </span>
<span id="cb64-1264"><a href="#cb64-1264" aria-hidden="true" tabindex="-1"></a>                               <span class="co">#  "pink",</span></span>
<span id="cb64-1265"><a href="#cb64-1265" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"gray30"</span>,</span>
<span id="cb64-1266"><a href="#cb64-1266" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"deeppink"</span>)) <span class="sc">+</span></span>
<span id="cb64-1267"><a href="#cb64-1267" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(<span class="at">data =</span> wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1268"><a href="#cb64-1268" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(external_gene_name <span class="sc">==</span> <span class="st">"FRO4"</span>),</span>
<span id="cb64-1269"><a href="#cb64-1269" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">label =</span> external_gene_name),</span>
<span id="cb64-1270"><a href="#cb64-1270" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="dv">4</span>,</span>
<span id="cb64-1271"><a href="#cb64-1271" aria-hidden="true" tabindex="-1"></a>                  <span class="at">nudge_x =</span> .<span class="dv">5</span>,</span>
<span id="cb64-1272"><a href="#cb64-1272" aria-hidden="true" tabindex="-1"></a>                  <span class="at">nudge_y =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb64-1273"><a href="#cb64-1273" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> wild_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1274"><a href="#cb64-1274" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(external_gene_name <span class="sc">==</span> <span class="st">"FRO4"</span>),</span>
<span id="cb64-1275"><a href="#cb64-1275" aria-hidden="true" tabindex="-1"></a>                <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb64-1276"><a href="#cb64-1276" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1277"><a href="#cb64-1277" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1278"><a href="#cb64-1278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1279"><a href="#cb64-1279" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1280"><a href="#cb64-1280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1281"><a href="#cb64-1281" aria-hidden="true" tabindex="-1"></a>Should you want to label more than one gene, you will need to use (for</span>
<span id="cb64-1282"><a href="#cb64-1282" aria-hidden="true" tabindex="-1"></a>example): <span class="in">`filter(external_gene_name %in% c("FRO4", "FRO5"))`</span></span>
<span id="cb64-1283"><a href="#cb64-1283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1284"><a href="#cb64-1284" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Save your plots</span><span class="co">](#save-your-plots)</span></span>
<span id="cb64-1285"><a href="#cb64-1285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1286"><a href="#cb64-1286" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ’‰ *Leishmania* {#leishmania-2}</span></span>
<span id="cb64-1287"><a href="#cb64-1287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1288"><a href="#cb64-1288" aria-hidden="true" tabindex="-1"></a>We will add a column to the results dataframe that contains the</span>
<span id="cb64-1289"><a href="#cb64-1289" aria-hidden="true" tabindex="-1"></a>-log~10~(padj). You could perform this transformation within the plot</span>
<span id="cb64-1290"><a href="#cb64-1290" aria-hidden="true" tabindex="-1"></a>command without adding a column to the data if you prefer.</span>
<span id="cb64-1291"><a href="#cb64-1291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1292"><a href="#cb64-1292" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add a column to the results dataframe that contains the</span>
<span id="cb64-1293"><a href="#cb64-1293" aria-hidden="true" tabindex="-1"></a>-log~10~(padj):</span>
<span id="cb64-1294"><a href="#cb64-1294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1297"><a href="#cb64-1297" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1298"><a href="#cb64-1298" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="ot">&lt;-</span> pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1299"><a href="#cb64-1299" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">log10_padj =</span> <span class="sc">-</span><span class="fu">log10</span>(padj)) </span>
<span id="cb64-1300"><a href="#cb64-1300" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1301"><a href="#cb64-1301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1302"><a href="#cb64-1302" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a volcano plot of the results:</span>
<span id="cb64-1303"><a href="#cb64-1303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1306"><a href="#cb64-1306" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1307"><a href="#cb64-1307" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1308"><a href="#cb64-1308" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, </span>
<span id="cb64-1309"><a href="#cb64-1309" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_padj)) <span class="sc">+</span></span>
<span id="cb64-1310"><a href="#cb64-1310" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1311"><a href="#cb64-1311" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1312"><a href="#cb64-1312" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1313"><a href="#cb64-1313" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1314"><a href="#cb64-1314" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1315"><a href="#cb64-1315" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1316"><a href="#cb64-1316" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1317"><a href="#cb64-1317" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1318"><a href="#cb64-1318" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1319"><a href="#cb64-1319" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1320"><a href="#cb64-1320" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1321"><a href="#cb64-1321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1322"><a href="#cb64-1322" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1323"><a href="#cb64-1323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1324"><a href="#cb64-1324" aria-hidden="true" tabindex="-1"></a>Our dashed lines are at -log~10~(0.05) and log~2~(2) and log~2~(-2) to</span>
<span id="cb64-1325"><a href="#cb64-1325" aria-hidden="true" tabindex="-1"></a>make more clear which genes (points) are significantly different between</span>
<span id="cb64-1326"><a href="#cb64-1326" aria-hidden="true" tabindex="-1"></a>the life stages and have a fold change of at least 2. Whilst we have</span>
<span id="cb64-1327"><a href="#cb64-1327" aria-hidden="true" tabindex="-1"></a>very many genes that are significant, we have fewer that are significant</span>
<span id="cb64-1328"><a href="#cb64-1328" aria-hidden="true" tabindex="-1"></a>and have a large fold change.</span>
<span id="cb64-1329"><a href="#cb64-1329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1330"><a href="#cb64-1330" aria-hidden="true" tabindex="-1"></a>In most cases, people colour the points to show that the quadrants. I</span>
<span id="cb64-1331"><a href="#cb64-1331" aria-hidden="true" tabindex="-1"></a>like to add columns to the dataframe to indicate if the gene is</span>
<span id="cb64-1332"><a href="#cb64-1332" aria-hidden="true" tabindex="-1"></a>significant and if the fold change is large and use those variables in</span>
<span id="cb64-1333"><a href="#cb64-1333" aria-hidden="true" tabindex="-1"></a>the plot.</span>
<span id="cb64-1334"><a href="#cb64-1334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1335"><a href="#cb64-1335" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add columns to the results dataframe to indicate if the gene is</span>
<span id="cb64-1336"><a href="#cb64-1336" aria-hidden="true" tabindex="-1"></a>significant and if the fold change is large:</span>
<span id="cb64-1337"><a href="#cb64-1337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1340"><a href="#cb64-1340" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1341"><a href="#cb64-1341" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="ot">&lt;-</span> pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1342"><a href="#cb64-1342" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sig =</span> padj <span class="sc">&lt;=</span> <span class="fl">0.05</span>,</span>
<span id="cb64-1343"><a href="#cb64-1343" aria-hidden="true" tabindex="-1"></a>         <span class="at">bigfc =</span> <span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;=</span> <span class="dv">2</span>) </span>
<span id="cb64-1344"><a href="#cb64-1344" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1345"><a href="#cb64-1345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1346"><a href="#cb64-1346" aria-hidden="true" tabindex="-1"></a>The use of <span class="in">`abs()`</span> (absolute) means genes with a fold change of at least</span>
<span id="cb64-1347"><a href="#cb64-1347" aria-hidden="true" tabindex="-1"></a>2 in either direction will be considered to have a large fold change.</span>
<span id="cb64-1348"><a href="#cb64-1348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1349"><a href="#cb64-1349" aria-hidden="true" tabindex="-1"></a>Now we can colour the points by these new columns. I use <span class="in">`interaction()`</span></span>
<span id="cb64-1350"><a href="#cb64-1350" aria-hidden="true" tabindex="-1"></a>to create four categories:</span>
<span id="cb64-1351"><a href="#cb64-1351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1352"><a href="#cb64-1352" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>not significant and not large fold change (FF)</span>
<span id="cb64-1353"><a href="#cb64-1353" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>significant and not large fold change (TF)</span>
<span id="cb64-1354"><a href="#cb64-1354" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>not significant and large fold (FT)</span>
<span id="cb64-1355"><a href="#cb64-1355" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>significant and large fold change (TT)</span>
<span id="cb64-1356"><a href="#cb64-1356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1357"><a href="#cb64-1357" aria-hidden="true" tabindex="-1"></a>And I use <span class="in">`scale_colour_manual()`</span> to set the colours for these</span>
<span id="cb64-1358"><a href="#cb64-1358" aria-hidden="true" tabindex="-1"></a>categories.</span>
<span id="cb64-1359"><a href="#cb64-1359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1360"><a href="#cb64-1360" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a volcano plot of the results with the points coloured by</span>
<span id="cb64-1361"><a href="#cb64-1361" aria-hidden="true" tabindex="-1"></a>significance and fold change:</span>
<span id="cb64-1362"><a href="#cb64-1362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1365"><a href="#cb64-1365" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1366"><a href="#cb64-1366" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1367"><a href="#cb64-1367" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, </span>
<span id="cb64-1368"><a href="#cb64-1368" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_padj, </span>
<span id="cb64-1369"><a href="#cb64-1369" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> <span class="fu">interaction</span>(sig, bigfc))) <span class="sc">+</span></span>
<span id="cb64-1370"><a href="#cb64-1370" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1371"><a href="#cb64-1371" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1372"><a href="#cb64-1372" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1373"><a href="#cb64-1373" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1374"><a href="#cb64-1374" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1375"><a href="#cb64-1375" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1376"><a href="#cb64-1376" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1377"><a href="#cb64-1377" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1378"><a href="#cb64-1378" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1379"><a href="#cb64-1379" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, </span>
<span id="cb64-1380"><a href="#cb64-1380" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"pink"</span>,</span>
<span id="cb64-1381"><a href="#cb64-1381" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"gray30"</span>,</span>
<span id="cb64-1382"><a href="#cb64-1382" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"deeppink"</span>)) <span class="sc">+</span></span>
<span id="cb64-1383"><a href="#cb64-1383" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1384"><a href="#cb64-1384" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1385"><a href="#cb64-1385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1386"><a href="#cb64-1386" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1387"><a href="#cb64-1387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1388"><a href="#cb64-1388" aria-hidden="true" tabindex="-1"></a>For exploring the data, I like add labels to all the significant genes</span>
<span id="cb64-1389"><a href="#cb64-1389" aria-hidden="true" tabindex="-1"></a>with a large fold change so I can very quickly identity them. The</span>
<span id="cb64-1390"><a href="#cb64-1390" aria-hidden="true" tabindex="-1"></a><span class="in">`ggrepel`</span> package has a function <span class="in">`geom_text_repel()`</span> that is useful for</span>
<span id="cb64-1391"><a href="#cb64-1391" aria-hidden="true" tabindex="-1"></a>adding labels so that they don't overlap.</span>
<span id="cb64-1392"><a href="#cb64-1392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1393"><a href="#cb64-1393" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load the package:</span>
<span id="cb64-1394"><a href="#cb64-1394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1397"><a href="#cb64-1397" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1398"><a href="#cb64-1398" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb64-1399"><a href="#cb64-1399" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb64-1400"><a href="#cb64-1400" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1401"><a href="#cb64-1401" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1402"><a href="#cb64-1402" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add labels to the significant genes with a large fold change:</span>
<span id="cb64-1403"><a href="#cb64-1403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1406"><a href="#cb64-1406" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1407"><a href="#cb64-1407" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1408"><a href="#cb64-1408" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, </span>
<span id="cb64-1409"><a href="#cb64-1409" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_padj, </span>
<span id="cb64-1410"><a href="#cb64-1410" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> <span class="fu">interaction</span>(sig, bigfc))) <span class="sc">+</span></span>
<span id="cb64-1411"><a href="#cb64-1411" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1412"><a href="#cb64-1412" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1413"><a href="#cb64-1413" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1414"><a href="#cb64-1414" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1415"><a href="#cb64-1415" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1416"><a href="#cb64-1416" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1417"><a href="#cb64-1417" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1418"><a href="#cb64-1418" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1419"><a href="#cb64-1419" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1420"><a href="#cb64-1420" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, </span>
<span id="cb64-1421"><a href="#cb64-1421" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"pink"</span>,</span>
<span id="cb64-1422"><a href="#cb64-1422" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"gray30"</span>,</span>
<span id="cb64-1423"><a href="#cb64-1423" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"deeppink"</span>)) <span class="sc">+</span></span>
<span id="cb64-1424"><a href="#cb64-1424" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">data =</span> pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1425"><a href="#cb64-1425" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(bigfc <span class="sc">==</span> <span class="cn">TRUE</span>, sig <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb64-1426"><a href="#cb64-1426" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">label =</span> description),</span>
<span id="cb64-1427"><a href="#cb64-1427" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="dv">3</span>,</span>
<span id="cb64-1428"><a href="#cb64-1428" aria-hidden="true" tabindex="-1"></a>                  <span class="at">max.overlaps =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb64-1429"><a href="#cb64-1429" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1430"><a href="#cb64-1430" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1431"><a href="#cb64-1431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1432"><a href="#cb64-1432" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1433"><a href="#cb64-1433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1434"><a href="#cb64-1434" aria-hidden="true" tabindex="-1"></a>Notice that I have used <span class="in">`filter()`</span> label only the genes that are both</span>
<span id="cb64-1435"><a href="#cb64-1435" aria-hidden="true" tabindex="-1"></a>significant and have a large fold change. In systems you are familiar</span>
<span id="cb64-1436"><a href="#cb64-1436" aria-hidden="true" tabindex="-1"></a>with, this labelling is very informative and can help you quickly</span>
<span id="cb64-1437"><a href="#cb64-1437" aria-hidden="true" tabindex="-1"></a>identify common themes. However, in this case, we do not have good</span>
<span id="cb64-1438"><a href="#cb64-1438" aria-hidden="true" tabindex="-1"></a>annotation for the genes. We can label only with the gene id or the</span>
<span id="cb64-1439"><a href="#cb64-1439" aria-hidden="true" tabindex="-1"></a>description. Many of the descriptions are</span>
<span id="cb64-1440"><a href="#cb64-1440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1441"><a href="#cb64-1441" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add labels to the significant genes with a large fold change only</span>
<span id="cb64-1442"><a href="#cb64-1442" aria-hidden="true" tabindex="-1"></a>where description doesn't contain "hypothetical" or "unspecified :</span>
<span id="cb64-1443"><a href="#cb64-1443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1446"><a href="#cb64-1446" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1447"><a href="#cb64-1447" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1448"><a href="#cb64-1448" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, </span>
<span id="cb64-1449"><a href="#cb64-1449" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_padj, </span>
<span id="cb64-1450"><a href="#cb64-1450" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> <span class="fu">interaction</span>(sig, bigfc))) <span class="sc">+</span></span>
<span id="cb64-1451"><a href="#cb64-1451" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1452"><a href="#cb64-1452" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1453"><a href="#cb64-1453" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1454"><a href="#cb64-1454" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1455"><a href="#cb64-1455" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1456"><a href="#cb64-1456" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1457"><a href="#cb64-1457" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1458"><a href="#cb64-1458" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1459"><a href="#cb64-1459" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1460"><a href="#cb64-1460" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, </span>
<span id="cb64-1461"><a href="#cb64-1461" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"pink"</span>,</span>
<span id="cb64-1462"><a href="#cb64-1462" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"gray30"</span>,</span>
<span id="cb64-1463"><a href="#cb64-1463" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"deeppink"</span>)) <span class="sc">+</span></span>
<span id="cb64-1464"><a href="#cb64-1464" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">data =</span> pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1465"><a href="#cb64-1465" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(bigfc <span class="sc">==</span> <span class="cn">TRUE</span>, sig <span class="sc">==</span> <span class="cn">TRUE</span>,</span>
<span id="cb64-1466"><a href="#cb64-1466" aria-hidden="true" tabindex="-1"></a>                           <span class="sc">!</span><span class="fu">str_detect</span>(description, <span class="st">"hypothetical|unspecified"</span>)),</span>
<span id="cb64-1467"><a href="#cb64-1467" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">label =</span> description),</span>
<span id="cb64-1468"><a href="#cb64-1468" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="dv">3</span>,</span>
<span id="cb64-1469"><a href="#cb64-1469" aria-hidden="true" tabindex="-1"></a>                  <span class="at">max.overlaps =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb64-1470"><a href="#cb64-1470" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1471"><a href="#cb64-1471" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1472"><a href="#cb64-1472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1473"><a href="#cb64-1473" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1474"><a href="#cb64-1474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1475"><a href="#cb64-1475" aria-hidden="true" tabindex="-1"></a>Key to interpreting the volcano plot is to remember that positive fold</span>
<span id="cb64-1476"><a href="#cb64-1476" aria-hidden="true" tabindex="-1"></a>changes means the gene is up-regulated in the procyclic stage and</span>
<span id="cb64-1477"><a href="#cb64-1477" aria-hidden="true" tabindex="-1"></a>negative fold changes means the gene is down-regulated (i.e., higher in</span>
<span id="cb64-1478"><a href="#cb64-1478" aria-hidden="true" tabindex="-1"></a>the metacyclic). This was determined by the order of the treatments in</span>
<span id="cb64-1479"><a href="#cb64-1479" aria-hidden="true" tabindex="-1"></a>the [contrast used in the DESeq2</span>
<span id="cb64-1480"><a href="#cb64-1480" aria-hidden="true" tabindex="-1"></a>analysis](../week-4/workshop.html#differential-expression-analysis-3)</span>
<span id="cb64-1481"><a href="#cb64-1481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1482"><a href="#cb64-1482" aria-hidden="true" tabindex="-1"></a>If you do forget which way round you did the comparison, you can always</span>
<span id="cb64-1483"><a href="#cb64-1483" aria-hidden="true" tabindex="-1"></a>examine the results dataframe to see which of the treatments seem to be</span>
<span id="cb64-1484"><a href="#cb64-1484" aria-hidden="true" tabindex="-1"></a>higher for the positive fold changes.</span>
<span id="cb64-1485"><a href="#cb64-1485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1486"><a href="#cb64-1486" aria-hidden="true" tabindex="-1"></a>When you have a gene of interest, you may wish to label it, and only it,</span>
<span id="cb64-1487"><a href="#cb64-1487" aria-hidden="true" tabindex="-1"></a>on the plot. This is done in the same way except that you filter the</span>
<span id="cb64-1488"><a href="#cb64-1488" aria-hidden="true" tabindex="-1"></a>data to only include the gene of interest. I have used and then use</span>
<span id="cb64-1489"><a href="#cb64-1489" aria-hidden="true" tabindex="-1"></a><span class="in">`geom_label_repel()`</span> rather than <span class="in">`geom_text_repel()`</span> to put the label in</span>
<span id="cb64-1490"><a href="#cb64-1490" aria-hidden="true" tabindex="-1"></a>a box and nudged it's position to get a line connecting the point and</span>
<span id="cb64-1491"><a href="#cb64-1491" aria-hidden="true" tabindex="-1"></a>the label. I have also increased the size of the point.</span>
<span id="cb64-1492"><a href="#cb64-1492" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1493"><a href="#cb64-1493" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add a label to one gene of interest (elongation factor 1-alpha) and :</span>
<span id="cb64-1494"><a href="#cb64-1494" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1497"><a href="#cb64-1497" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1498"><a href="#cb64-1498" aria-hidden="true" tabindex="-1"></a>pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1499"><a href="#cb64-1499" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, </span>
<span id="cb64-1500"><a href="#cb64-1500" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_padj, </span>
<span id="cb64-1501"><a href="#cb64-1501" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> <span class="fu">interaction</span>(sig, bigfc))) <span class="sc">+</span></span>
<span id="cb64-1502"><a href="#cb64-1502" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1503"><a href="#cb64-1503" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1504"><a href="#cb64-1504" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1505"><a href="#cb64-1505" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1506"><a href="#cb64-1506" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1507"><a href="#cb64-1507" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1508"><a href="#cb64-1508" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1509"><a href="#cb64-1509" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1510"><a href="#cb64-1510" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1511"><a href="#cb64-1511" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, </span>
<span id="cb64-1512"><a href="#cb64-1512" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"pink"</span>,</span>
<span id="cb64-1513"><a href="#cb64-1513" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"gray30"</span>,</span>
<span id="cb64-1514"><a href="#cb64-1514" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"deeppink"</span>)) <span class="sc">+</span></span>
<span id="cb64-1515"><a href="#cb64-1515" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(<span class="at">data =</span> pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1516"><a href="#cb64-1516" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(description <span class="sc">==</span> <span class="st">"elongation factor 1-alpha"</span>),</span>
<span id="cb64-1517"><a href="#cb64-1517" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">label =</span> description),</span>
<span id="cb64-1518"><a href="#cb64-1518" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="dv">4</span>,</span>
<span id="cb64-1519"><a href="#cb64-1519" aria-hidden="true" tabindex="-1"></a>                  <span class="at">nudge_x =</span> .<span class="dv">5</span>,</span>
<span id="cb64-1520"><a href="#cb64-1520" aria-hidden="true" tabindex="-1"></a>                  <span class="at">nudge_y =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb64-1521"><a href="#cb64-1521" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> pro_meta_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1522"><a href="#cb64-1522" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(description <span class="sc">==</span> <span class="st">"elongation factor 1-alpha"</span>),</span>
<span id="cb64-1523"><a href="#cb64-1523" aria-hidden="true" tabindex="-1"></a>                <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb64-1524"><a href="#cb64-1524" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1525"><a href="#cb64-1525" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1526"><a href="#cb64-1526" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1527"><a href="#cb64-1527" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1528"><a href="#cb64-1528" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1529"><a href="#cb64-1529" aria-hidden="true" tabindex="-1"></a>Should you want to label more than one gene, you will need to use (for</span>
<span id="cb64-1530"><a href="#cb64-1530" aria-hidden="true" tabindex="-1"></a>example):</span>
<span id="cb64-1531"><a href="#cb64-1531" aria-hidden="true" tabindex="-1"></a><span class="in">`filter(description %in% c("elongation factor 1-alpha", "ADP/ATP translocase 1 putative"))`</span></span>
<span id="cb64-1532"><a href="#cb64-1532" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1533"><a href="#cb64-1533" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Save your plots</span><span class="co">](#save-your-plots)</span></span>
<span id="cb64-1534"><a href="#cb64-1534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1535"><a href="#cb64-1535" aria-hidden="true" tabindex="-1"></a><span class="fu">## ğŸ­ Stem cells {#stem-cells-3}</span></span>
<span id="cb64-1536"><a href="#cb64-1536" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1537"><a href="#cb64-1537" aria-hidden="true" tabindex="-1"></a>We will add a column to the results dataframe that contains the</span>
<span id="cb64-1538"><a href="#cb64-1538" aria-hidden="true" tabindex="-1"></a>-log~10~(FDR). You could perform this transformation within the plot</span>
<span id="cb64-1539"><a href="#cb64-1539" aria-hidden="true" tabindex="-1"></a>command without adding a column to the data if you prefer.</span>
<span id="cb64-1540"><a href="#cb64-1540" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1541"><a href="#cb64-1541" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add a column to the results dataframe that contains the</span>
<span id="cb64-1542"><a href="#cb64-1542" aria-hidden="true" tabindex="-1"></a>-log~10~(FDR):</span>
<span id="cb64-1543"><a href="#cb64-1543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1546"><a href="#cb64-1546" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1547"><a href="#cb64-1547" aria-hidden="true" tabindex="-1"></a>hspc_prog_results <span class="ot">&lt;-</span> hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1548"><a href="#cb64-1548" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">log10_FDR =</span> <span class="sc">-</span><span class="fu">log10</span>(FDR)) </span>
<span id="cb64-1549"><a href="#cb64-1549" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1550"><a href="#cb64-1550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1551"><a href="#cb64-1551" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a volcano plot of the results:</span>
<span id="cb64-1552"><a href="#cb64-1552" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1555"><a href="#cb64-1555" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1556"><a href="#cb64-1556" aria-hidden="true" tabindex="-1"></a>hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1557"><a href="#cb64-1557" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> summary.logFC, </span>
<span id="cb64-1558"><a href="#cb64-1558" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_FDR)) <span class="sc">+</span></span>
<span id="cb64-1559"><a href="#cb64-1559" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1560"><a href="#cb64-1560" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1561"><a href="#cb64-1561" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1562"><a href="#cb64-1562" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1563"><a href="#cb64-1563" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1564"><a href="#cb64-1564" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1565"><a href="#cb64-1565" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1566"><a href="#cb64-1566" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1567"><a href="#cb64-1567" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1568"><a href="#cb64-1568" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1569"><a href="#cb64-1569" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1570"><a href="#cb64-1570" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1571"><a href="#cb64-1571" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1572"><a href="#cb64-1572" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1573"><a href="#cb64-1573" aria-hidden="true" tabindex="-1"></a>Our dashed lines are at -log~10~(0.05) and log~2~(2) and log~2~(-2) to</span>
<span id="cb64-1574"><a href="#cb64-1574" aria-hidden="true" tabindex="-1"></a>make more clear which genes (points) are significantly different between</span>
<span id="cb64-1575"><a href="#cb64-1575" aria-hidden="true" tabindex="-1"></a>the cell types and have a fold change of at least 2.</span>
<span id="cb64-1576"><a href="#cb64-1576" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1577"><a href="#cb64-1577" aria-hidden="true" tabindex="-1"></a>In most cases, people colour the points to show that the quadrants. I</span>
<span id="cb64-1578"><a href="#cb64-1578" aria-hidden="true" tabindex="-1"></a>like to add columns to the dataframe to indicate if the gene is</span>
<span id="cb64-1579"><a href="#cb64-1579" aria-hidden="true" tabindex="-1"></a>significant and if the fold change is large and use those variables in</span>
<span id="cb64-1580"><a href="#cb64-1580" aria-hidden="true" tabindex="-1"></a>the plot.</span>
<span id="cb64-1581"><a href="#cb64-1581" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1582"><a href="#cb64-1582" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add columns to the results dataframe to indicate if the gene is</span>
<span id="cb64-1583"><a href="#cb64-1583" aria-hidden="true" tabindex="-1"></a>significant and if the fold change is large:</span>
<span id="cb64-1584"><a href="#cb64-1584" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1587"><a href="#cb64-1587" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1588"><a href="#cb64-1588" aria-hidden="true" tabindex="-1"></a>hspc_prog_results <span class="ot">&lt;-</span> hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1589"><a href="#cb64-1589" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sig =</span> FDR <span class="sc">&lt;=</span> <span class="fl">0.05</span>,</span>
<span id="cb64-1590"><a href="#cb64-1590" aria-hidden="true" tabindex="-1"></a>         <span class="at">bigfc =</span> <span class="fu">abs</span>(summary.logFC) <span class="sc">&gt;=</span> <span class="dv">2</span>) </span>
<span id="cb64-1591"><a href="#cb64-1591" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1592"><a href="#cb64-1592" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1593"><a href="#cb64-1593" aria-hidden="true" tabindex="-1"></a>The use of <span class="in">`abs()`</span> (absolute) means genes with a fold change of at least</span>
<span id="cb64-1594"><a href="#cb64-1594" aria-hidden="true" tabindex="-1"></a>2 in either direction will be considered to have a large fold change.</span>
<span id="cb64-1595"><a href="#cb64-1595" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1596"><a href="#cb64-1596" aria-hidden="true" tabindex="-1"></a>Now we can colour the points by these new columns. I use <span class="in">`interaction()`</span></span>
<span id="cb64-1597"><a href="#cb64-1597" aria-hidden="true" tabindex="-1"></a>to create four categories:</span>
<span id="cb64-1598"><a href="#cb64-1598" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1599"><a href="#cb64-1599" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>not significant and not large fold change (FF)</span>
<span id="cb64-1600"><a href="#cb64-1600" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>significant and not large fold change (TF)</span>
<span id="cb64-1601"><a href="#cb64-1601" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>not significant and large fold (FT)</span>
<span id="cb64-1602"><a href="#cb64-1602" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>significant and large fold change (TT)</span>
<span id="cb64-1603"><a href="#cb64-1603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1604"><a href="#cb64-1604" aria-hidden="true" tabindex="-1"></a>And I use <span class="in">`scale_colour_manual()`</span> to set the colours for these</span>
<span id="cb64-1605"><a href="#cb64-1605" aria-hidden="true" tabindex="-1"></a>categories.</span>
<span id="cb64-1606"><a href="#cb64-1606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1607"><a href="#cb64-1607" aria-hidden="true" tabindex="-1"></a>NOTE: there are no "not significant and large fold change (FT)" in this</span>
<span id="cb64-1608"><a href="#cb64-1608" aria-hidden="true" tabindex="-1"></a>case. This means we do not need four colours. I have put the "gray30"</span>
<span id="cb64-1609"><a href="#cb64-1609" aria-hidden="true" tabindex="-1"></a>colour in the code but commented it out.</span>
<span id="cb64-1610"><a href="#cb64-1610" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1611"><a href="#cb64-1611" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Create a volcano plot of the results with the points coloured by</span>
<span id="cb64-1612"><a href="#cb64-1612" aria-hidden="true" tabindex="-1"></a>significance and fold change:</span>
<span id="cb64-1613"><a href="#cb64-1613" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1616"><a href="#cb64-1616" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1617"><a href="#cb64-1617" aria-hidden="true" tabindex="-1"></a>hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1618"><a href="#cb64-1618" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> summary.logFC, </span>
<span id="cb64-1619"><a href="#cb64-1619" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_FDR, </span>
<span id="cb64-1620"><a href="#cb64-1620" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> <span class="fu">interaction</span>(sig, bigfc))) <span class="sc">+</span></span>
<span id="cb64-1621"><a href="#cb64-1621" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1622"><a href="#cb64-1622" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1623"><a href="#cb64-1623" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1624"><a href="#cb64-1624" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1625"><a href="#cb64-1625" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1626"><a href="#cb64-1626" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1627"><a href="#cb64-1627" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1628"><a href="#cb64-1628" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1629"><a href="#cb64-1629" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1630"><a href="#cb64-1630" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, </span>
<span id="cb64-1631"><a href="#cb64-1631" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"pink"</span>,</span>
<span id="cb64-1632"><a href="#cb64-1632" aria-hidden="true" tabindex="-1"></a>                                <span class="co"># "gray30",</span></span>
<span id="cb64-1633"><a href="#cb64-1633" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"deeppink"</span>)) <span class="sc">+</span></span>
<span id="cb64-1634"><a href="#cb64-1634" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1635"><a href="#cb64-1635" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1636"><a href="#cb64-1636" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1637"><a href="#cb64-1637" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1638"><a href="#cb64-1638" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1639"><a href="#cb64-1639" aria-hidden="true" tabindex="-1"></a>For exploring the data, I like add labels to all the significant genes</span>
<span id="cb64-1640"><a href="#cb64-1640" aria-hidden="true" tabindex="-1"></a>with a large fold change so I can very quickly identity them. The</span>
<span id="cb64-1641"><a href="#cb64-1641" aria-hidden="true" tabindex="-1"></a><span class="in">`ggrepel`</span> package has a function <span class="in">`geom_text_repel()`</span> that is useful for</span>
<span id="cb64-1642"><a href="#cb64-1642" aria-hidden="true" tabindex="-1"></a>adding labels so that they don't overlap.</span>
<span id="cb64-1643"><a href="#cb64-1643" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1644"><a href="#cb64-1644" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Load the package:</span>
<span id="cb64-1645"><a href="#cb64-1645" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1648"><a href="#cb64-1648" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1649"><a href="#cb64-1649" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb64-1650"><a href="#cb64-1650" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1651"><a href="#cb64-1651" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1652"><a href="#cb64-1652" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add labels to the significant genes with a large fold change:</span>
<span id="cb64-1653"><a href="#cb64-1653" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1656"><a href="#cb64-1656" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1657"><a href="#cb64-1657" aria-hidden="true" tabindex="-1"></a>hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1658"><a href="#cb64-1658" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> summary.logFC, </span>
<span id="cb64-1659"><a href="#cb64-1659" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_FDR, </span>
<span id="cb64-1660"><a href="#cb64-1660" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> <span class="fu">interaction</span>(sig, bigfc))) <span class="sc">+</span></span>
<span id="cb64-1661"><a href="#cb64-1661" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1662"><a href="#cb64-1662" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1663"><a href="#cb64-1663" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1664"><a href="#cb64-1664" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1665"><a href="#cb64-1665" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1666"><a href="#cb64-1666" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1667"><a href="#cb64-1667" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1668"><a href="#cb64-1668" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1669"><a href="#cb64-1669" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1670"><a href="#cb64-1670" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, </span>
<span id="cb64-1671"><a href="#cb64-1671" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"pink"</span>,</span>
<span id="cb64-1672"><a href="#cb64-1672" aria-hidden="true" tabindex="-1"></a>                              <span class="co">#   "gray30",</span></span>
<span id="cb64-1673"><a href="#cb64-1673" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"deeppink"</span>)) <span class="sc">+</span></span>
<span id="cb64-1674"><a href="#cb64-1674" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">data =</span> hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1675"><a href="#cb64-1675" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(bigfc <span class="sc">==</span> <span class="cn">TRUE</span>, sig <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb64-1676"><a href="#cb64-1676" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">label =</span> external_gene_name),</span>
<span id="cb64-1677"><a href="#cb64-1677" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="dv">3</span>,</span>
<span id="cb64-1678"><a href="#cb64-1678" aria-hidden="true" tabindex="-1"></a>                  <span class="at">max.overlaps =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb64-1679"><a href="#cb64-1679" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1680"><a href="#cb64-1680" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1681"><a href="#cb64-1681" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1682"><a href="#cb64-1682" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1683"><a href="#cb64-1683" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1684"><a href="#cb64-1684" aria-hidden="true" tabindex="-1"></a>Notice that I have used <span class="in">`filter()`</span> label only the genes that are both</span>
<span id="cb64-1685"><a href="#cb64-1685" aria-hidden="true" tabindex="-1"></a>significant and have a large fold change. In systems you are familiar</span>
<span id="cb64-1686"><a href="#cb64-1686" aria-hidden="true" tabindex="-1"></a>with, this labelling is very informative and can help you quickly</span>
<span id="cb64-1687"><a href="#cb64-1687" aria-hidden="true" tabindex="-1"></a>identify common themes. Key to interpreting the volcano plot is to</span>
<span id="cb64-1688"><a href="#cb64-1688" aria-hidden="true" tabindex="-1"></a>remember that positive fold changes means the gene is up-regulated in</span>
<span id="cb64-1689"><a href="#cb64-1689" aria-hidden="true" tabindex="-1"></a>the Prog and negative fold changes means the gene is down-regulated</span>
<span id="cb64-1690"><a href="#cb64-1690" aria-hidden="true" tabindex="-1"></a>(i.e., higher HSPC). This was determined by us choosing [the</span>
<span id="cb64-1691"><a href="#cb64-1691" aria-hidden="true" tabindex="-1"></a>results_hspc_prog\$prog dataframe from the list</span>
<span id="cb64-1692"><a href="#cb64-1692" aria-hidden="true" tabindex="-1"></a>object](../week-4/workshop.html#differential-expression-analysis-4)</span>
<span id="cb64-1693"><a href="#cb64-1693" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1694"><a href="#cb64-1694" aria-hidden="true" tabindex="-1"></a>If you do forget which way round you did the comparison, you can always</span>
<span id="cb64-1695"><a href="#cb64-1695" aria-hidden="true" tabindex="-1"></a>examine the gene summary dataframe to see which of the treatments seem</span>
<span id="cb64-1696"><a href="#cb64-1696" aria-hidden="true" tabindex="-1"></a>to be higher for the positive fold changes.</span>
<span id="cb64-1697"><a href="#cb64-1697" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1698"><a href="#cb64-1698" aria-hidden="true" tabindex="-1"></a>When you have a gene of interest, you may wish to label it on the plot.</span>
<span id="cb64-1699"><a href="#cb64-1699" aria-hidden="true" tabindex="-1"></a>This is done in the same way except that you filter the data to only</span>
<span id="cb64-1700"><a href="#cb64-1700" aria-hidden="true" tabindex="-1"></a>include the gene of interest. I have used and then use</span>
<span id="cb64-1701"><a href="#cb64-1701" aria-hidden="true" tabindex="-1"></a><span class="in">`geom_label_repel()`</span> rather than <span class="in">`geom_text_repel()`</span> to put the label in</span>
<span id="cb64-1702"><a href="#cb64-1702" aria-hidden="true" tabindex="-1"></a>a box and nudged it's position to get a line connecting the point and</span>
<span id="cb64-1703"><a href="#cb64-1703" aria-hidden="true" tabindex="-1"></a>the label. I have also increased the size of the point.</span>
<span id="cb64-1704"><a href="#cb64-1704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1705"><a href="#cb64-1705" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Add a label to one gene of interest (Procr) and :</span>
<span id="cb64-1706"><a href="#cb64-1706" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1709"><a href="#cb64-1709" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1710"><a href="#cb64-1710" aria-hidden="true" tabindex="-1"></a>hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1711"><a href="#cb64-1711" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> summary.logFC, </span>
<span id="cb64-1712"><a href="#cb64-1712" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> log10_FDR, </span>
<span id="cb64-1713"><a href="#cb64-1713" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour =</span> <span class="fu">interaction</span>(sig, bigfc))) <span class="sc">+</span></span>
<span id="cb64-1714"><a href="#cb64-1714" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb64-1715"><a href="#cb64-1715" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), </span>
<span id="cb64-1716"><a href="#cb64-1716" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1717"><a href="#cb64-1717" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, </span>
<span id="cb64-1718"><a href="#cb64-1718" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1719"><a href="#cb64-1719" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, </span>
<span id="cb64-1720"><a href="#cb64-1720" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb64-1721"><a href="#cb64-1721" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1722"><a href="#cb64-1722" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb64-1723"><a href="#cb64-1723" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray"</span>, </span>
<span id="cb64-1724"><a href="#cb64-1724" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"pink"</span>,</span>
<span id="cb64-1725"><a href="#cb64-1725" aria-hidden="true" tabindex="-1"></a>                                <span class="co"># "gray30",</span></span>
<span id="cb64-1726"><a href="#cb64-1726" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"deeppink"</span>)) <span class="sc">+</span></span>
<span id="cb64-1727"><a href="#cb64-1727" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(<span class="at">data =</span> hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1728"><a href="#cb64-1728" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(external_gene_name <span class="sc">==</span> <span class="st">"Procr"</span>),</span>
<span id="cb64-1729"><a href="#cb64-1729" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">label =</span> external_gene_name),</span>
<span id="cb64-1730"><a href="#cb64-1730" aria-hidden="true" tabindex="-1"></a>                  <span class="at">size =</span> <span class="dv">4</span>,</span>
<span id="cb64-1731"><a href="#cb64-1731" aria-hidden="true" tabindex="-1"></a>                  <span class="at">nudge_x =</span> .<span class="dv">5</span>,</span>
<span id="cb64-1732"><a href="#cb64-1732" aria-hidden="true" tabindex="-1"></a>                  <span class="at">nudge_y =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb64-1733"><a href="#cb64-1733" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> hspc_prog_results <span class="sc">|&gt;</span> </span>
<span id="cb64-1734"><a href="#cb64-1734" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">filter</span>(external_gene_name <span class="sc">==</span> <span class="st">"Procr"</span>),</span>
<span id="cb64-1735"><a href="#cb64-1735" aria-hidden="true" tabindex="-1"></a>                <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb64-1736"><a href="#cb64-1736" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-1737"><a href="#cb64-1737" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb64-1738"><a href="#cb64-1738" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1739"><a href="#cb64-1739" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1740"><a href="#cb64-1740" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1741"><a href="#cb64-1741" aria-hidden="true" tabindex="-1"></a>Should you want to label more than one gene, you will need to use (for</span>
<span id="cb64-1742"><a href="#cb64-1742" aria-hidden="true" tabindex="-1"></a>example): <span class="in">`filter(external_gene_name %in% c("Procr", "Emb"))`</span></span>
<span id="cb64-1743"><a href="#cb64-1743" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1744"><a href="#cb64-1744" aria-hidden="true" tabindex="-1"></a>Now go to <span class="co">[</span><span class="ot">Save your plots</span><span class="co">](#save-your-plots)</span></span>
<span id="cb64-1745"><a href="#cb64-1745" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1746"><a href="#cb64-1746" aria-hidden="true" tabindex="-1"></a><span class="fu"># Save your plots {#save-your-plots}</span></span>
<span id="cb64-1747"><a href="#cb64-1747" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1748"><a href="#cb64-1748" aria-hidden="true" tabindex="-1"></a>Once you have finished designing your plots, you should save them using</span>
<span id="cb64-1749"><a href="#cb64-1749" aria-hidden="true" tabindex="-1"></a><span class="in">`ggsave()`</span>. You will want to assign the plot to a variable and use code</span>
<span id="cb64-1750"><a href="#cb64-1750" aria-hidden="true" tabindex="-1"></a>similar to this:</span>
<span id="cb64-1751"><a href="#cb64-1751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1754"><a href="#cb64-1754" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb64-1755"><a href="#cb64-1755" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb64-1756"><a href="#cb64-1756" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"figures/my-informative-file-name.png"</span>,</span>
<span id="cb64-1757"><a href="#cb64-1757" aria-hidden="true" tabindex="-1"></a>       <span class="at">plot =</span> vol,</span>
<span id="cb64-1758"><a href="#cb64-1758" aria-hidden="true" tabindex="-1"></a>       <span class="at">height =</span> <span class="fl">4.5</span>, </span>
<span id="cb64-1759"><a href="#cb64-1759" aria-hidden="true" tabindex="-1"></a>       <span class="at">width =</span> <span class="fl">4.5</span>,</span>
<span id="cb64-1760"><a href="#cb64-1760" aria-hidden="true" tabindex="-1"></a>       <span class="at">units =</span> <span class="st">"in"</span>,</span>
<span id="cb64-1761"><a href="#cb64-1761" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">"png"</span>)</span>
<span id="cb64-1762"><a href="#cb64-1762" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb64-1763"><a href="#cb64-1763" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1764"><a href="#cb64-1764" aria-hidden="true" tabindex="-1"></a>I recommend saving your figure in the approximate size it will appear in</span>
<span id="cb64-1765"><a href="#cb64-1765" aria-hidden="true" tabindex="-1"></a>your report, that is, don't resize it in word/google docs. This because</span>
<span id="cb64-1766"><a href="#cb64-1766" aria-hidden="true" tabindex="-1"></a>the font will then be an appropriate size for the report. For journals,</span>
<span id="cb64-1767"><a href="#cb64-1767" aria-hidden="true" tabindex="-1"></a>we often have to creat figure files of very high resolution. You can do</span>
<span id="cb64-1768"><a href="#cb64-1768" aria-hidden="true" tabindex="-1"></a>this by increasing the <span class="in">`height`</span> and <span class="in">`width`</span> and then resizing the figure</span>
<span id="cb64-1769"><a href="#cb64-1769" aria-hidden="true" tabindex="-1"></a>in the document. however, this requires resizing the all font and lines</span>
<span id="cb64-1770"><a href="#cb64-1770" aria-hidden="true" tabindex="-1"></a>in the figure. I also recommend saving it as a png file as this is a</span>
<span id="cb64-1771"><a href="#cb64-1771" aria-hidden="true" tabindex="-1"></a>lossless format.</span>
<span id="cb64-1772"><a href="#cb64-1772" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1773"><a href="#cb64-1773" aria-hidden="true" tabindex="-1"></a><span class="fu"># ğŸ¤— Look after future you!</span></span>
<span id="cb64-1774"><a href="#cb64-1774" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1775"><a href="#cb64-1775" aria-hidden="true" tabindex="-1"></a>ğŸ¬ Go through your script and tidy up. I would suggest restarting R and</span>
<span id="cb64-1776"><a href="#cb64-1776" aria-hidden="true" tabindex="-1"></a>trying to run the full pipeline from start to finish. You might need to:</span>
<span id="cb64-1777"><a href="#cb64-1777" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1778"><a href="#cb64-1778" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>collect together library statements at the top of the script</span>
<span id="cb64-1779"><a href="#cb64-1779" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>remove code that you needed to start today but which wouldn't be</span>
<span id="cb64-1780"><a href="#cb64-1780" aria-hidden="true" tabindex="-1"></a>    needed running the script from the top (e.g., importing the results)</span>
<span id="cb64-1781"><a href="#cb64-1781" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>edit your comments for clarity</span>
<span id="cb64-1782"><a href="#cb64-1782" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>rename variables for consistency or clarity</span>
<span id="cb64-1783"><a href="#cb64-1783" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>remove house keeping or exploratory code</span>
<span id="cb64-1784"><a href="#cb64-1784" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>restyle code, add code section headers etc</span>
<span id="cb64-1785"><a href="#cb64-1785" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1786"><a href="#cb64-1786" aria-hidden="true" tabindex="-1"></a><span class="fu"># ğŸ¥³ Finished</span></span>
<span id="cb64-1787"><a href="#cb64-1787" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1788"><a href="#cb64-1788" aria-hidden="true" tabindex="-1"></a>Well Done!</span>
<span id="cb64-1789"><a href="#cb64-1789" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1790"><a href="#cb64-1790" aria-hidden="true" tabindex="-1"></a><span class="fu"># Independent study following the workshop</span></span>
<span id="cb64-1791"><a href="#cb64-1791" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1792"><a href="#cb64-1792" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Consolidate</span><span class="co">](study_after_workshop.qmd)</span></span>
<span id="cb64-1793"><a href="#cb64-1793" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1794"><a href="#cb64-1794" aria-hidden="true" tabindex="-1"></a><span class="fu"># The Code file</span></span>
<span id="cb64-1795"><a href="#cb64-1795" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1796"><a href="#cb64-1796" aria-hidden="true" tabindex="-1"></a>These contain all the code needed in the workshop even where it is not</span>
<span id="cb64-1797"><a href="#cb64-1797" aria-hidden="true" tabindex="-1"></a>visible on the webpage.</span>
<span id="cb64-1798"><a href="#cb64-1798" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1799"><a href="#cb64-1799" aria-hidden="true" tabindex="-1"></a>The <span class="co">[</span><span class="ot">workshop.qmd</span><span class="co">](workshop.qmd)</span> file is the file I use to compile the</span>
<span id="cb64-1800"><a href="#cb64-1800" aria-hidden="true" tabindex="-1"></a>practical. Qmd stands for Quarto markdown. It allows code and ordinary</span>
<span id="cb64-1801"><a href="#cb64-1801" aria-hidden="true" tabindex="-1"></a>text to be interleaved to produce well-formatted reports including</span>
<span id="cb64-1802"><a href="#cb64-1802" aria-hidden="true" tabindex="-1"></a>webpages. Right-click on the link and choose Save-As to download. You</span>
<span id="cb64-1803"><a href="#cb64-1803" aria-hidden="true" tabindex="-1"></a>will be able to open the Qmd file in RStudio. Alternatively, [View in</span>
<span id="cb64-1804"><a href="#cb64-1804" aria-hidden="true" tabindex="-1"></a>Browser](https://github.com/3mmaRand/BIO00088H-data/blob/main/transcriptomics/week-5/workshop.qmd).</span>
<span id="cb64-1805"><a href="#cb64-1805" aria-hidden="true" tabindex="-1"></a>Coding and thinking answers are marked with <span class="in">`#---CODING ANSWER---`</span> and</span>
<span id="cb64-1806"><a href="#cb64-1806" aria-hidden="true" tabindex="-1"></a><span class="in">`#---THINKING ANSWER---`</span></span>
<span id="cb64-1807"><a href="#cb64-1807" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1808"><a href="#cb64-1808" aria-hidden="true" tabindex="-1"></a>Pages made with R <span class="co">[</span><span class="ot">@R-core</span><span class="co">]</span>, Quarto <span class="co">[</span><span class="ot">@Allaire_Quarto_2024</span><span class="co">]</span>, <span class="in">`knitr`</span></span>
<span id="cb64-1809"><a href="#cb64-1809" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">@knitr1; @knitr2; @knitr3</span><span class="co">]</span>, <span class="in">`kableExtra`</span> <span class="co">[</span><span class="ot">@kableExtra</span><span class="co">]</span></span>
<span id="cb64-1810"><a href="#cb64-1810" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-1811"><a href="#cb64-1811" aria-hidden="true" tabindex="-1"></a><span class="fu"># References</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Rand E (2024). Data Analysis for Group Project. <a href="https://3mmarand.github.io/BIO00088H-data/" class="uri">https://3mmarand.github.io/BIO00088H-data/</a>.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/er_13">
      <i class="bi bi-twitter" role="img" aria-label="Twitter">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/3mmaRand/BIO00088H-data">
      <i class="bi bi-github" role="img" aria-label="BIO00088H-data GitHub">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer><script>
document.addEventListener("DOMContentLoaded", function() {
  // grab all links under â€œOther Linksâ€ in the right-hand TOC
  const links = document.querySelectorAll('.quarto-other-links ul li a');
  links.forEach(function(a) {
    // find the one whose text is exactly â€œMake a suggestionâ€
    if (a.textContent.trim() === 'Make a suggestion') {
      const pageUrl   = encodeURIComponent(window.location.href);
      const pageTitle = encodeURIComponent(document.title);
      a.href = 'https://github.com/3mmaRand/BIO00088H-data/issues/new'
             + '?title=' + pageTitle
             + '&body='  + pageUrl;
      a.target = '_blank';
    }
  });
});
</script>


</body></html>